// jibo-emotion-system/src/sensing/sensors/MotorFaultSensor.ts
import {Sensor} from './Sensor';
import {AxisState} from 'jibo-common-types';
import {EmotionSystem} from "../../EmotionSystem";

export {}

/**
 * @internal
 */
export class MotorFaultEvent {
    public timestamp: number;
    public type: string; //'body', 'neck' or 'pelvis'

    constructor(type = 'neck') {
        this.timestamp = Date.now();
        this.type = type;
    }
}

/**
 * @internal
 */
type FaultHandler = (data: AxisState) => void;

/**
 * @internal
 */
export class MotorFaultSensor extends Sensor<MotorFaultEvent> {

    static SENSOR_NAME = 'MotorFaultSensor';
    public readyForFault = true; // flag that makes sure we dont overreact to a single fault
    private faultListenerOn: FaultHandler;
    private faultListenerOff: FaultHandler;

    constructor(parent: EmotionSystem) {
        super(MotorFaultSensor.SENSOR_NAME, true, parent);
    }

    init() {
        super.init();
        this.faultListenerOn = (data: AxisState) => {
            // ready to handle a fault
            if (this.readyForFault) {
                this.readyForFault = false;
                this.supplyData(new MotorFaultEvent('neck'));
            }
        };
        this.faultListenerOff = () => {
            this.readyForFault = true;
        };
        // we know a motor fault happened
        this.parent.jibo.system.events.axisFaultOn.on(this.faultListenerOn);
        this.parent.jibo.system.events.axisFaultOff.on(this.faultListenerOff);
    }

    /**
     * Deallocate and clean up any resources that were created.
     */
    dispose() {
        super.dispose();
        this.parent.jibo.system.events.axisFaultOn.removeListener(this.faultListenerOn);
        this.parent.jibo.system.events.axisFaultOff.removeListener(this.faultListenerOff);
    }

}
