// jibo-emotion-system/src/sensing/sensors/Sensor.ts
/*
 * The base class for each sensor object
 */

import {Event, EventContainer} from 'jibo-typed-events';
import {EmotionSystem} from "../../EmotionSystem";

export {}

// a wrapper class that includes the name of the Sensor that generated the data, plus the data itself
/**
 * @internal
 */
export class SensorDataWrapper<SensorDataType> {

    constructor(public sensorName: string, public sensorData: SensorDataType) {

    }
}

/**
 * @internal
 */
export class SensorEventContainer<SensorDataType> extends EventContainer {
    newDataEvent = new Event<SensorDataWrapper<SensorDataType>>('New Data Event from Sensor.');
}

/**
 * @internal
 */
export abstract class Sensor<SensorDataType> {

    protected lastData: SensorDataType;

    public events = new SensorEventContainer<SensorDataType>();

    constructor(public name: string, private emitNewDataEvent = false, protected parent: EmotionSystem) {
        return;
    }

    /**
     * Can be used to retrieve the last supplied data point
     * @returns {T}
     */
    getLastData() {
        return this.lastData;
    }

    /**
     * This method should be called to provide the latest data available to
     * this sensor
     * @param {T} data
     */
    supplyData(data: SensorDataType) {
        this.lastData = data;
        if (this.emitNewDataEvent) {
            const sensorData = new SensorDataWrapper<SensorDataType>(this.name, data);
            this.events.newDataEvent.emit(sensorData);
        }
    }

    /**
     * life cycle of the sensor.
     */
    init() {
        return;
    };

    /**
     * Cleans up components, listeners, that were attached during setup or
     * life cycle of the sensor.
     */
    dispose() {
        this.events.removeAllListeners();
        return;
    }
}
