// jibo-emotion-system/src/sensing/sensors/TouchSensor.ts
import {Sensor} from './Sensor';
import {Handler} from "jibo-typed-events/dts/Event";
import {EmotionSystem} from "../../EmotionSystem";
import {TouchOptions, TouchTypes} from "../../common/Types";

import parentLog from '../../log';
const log = parentLog.createChild('Sensor.Touch');

export {}

/**
 * @internal
 */
export class TouchEvent {

    public timestamp: number;
    public type: string;

    constructor(type = 'HeadTouch') {
        this.timestamp = Date.now();
        this.type = type;
    }
}

/**
 * @internal
 */
export class TouchSensor extends Sensor<TouchEvent> {

    static SENSOR_NAME = 'TouchSensor';
    public touchHandler:Handler<boolean[]>;

    constructor(parent: EmotionSystem) {
        super(TouchSensor.SENSOR_NAME, true, parent);
        // Things to think about: Gesture recognition
        // Single tap, multiple taps, rubbing
    }

    init() {
        super.init();
        // accept any type b/c bug in typings in jibo rn.
        this.touchHandler = (pads: boolean[]=[]) => {
            log.debug("sensing touch", pads);
            let options: TouchOptions = {
                randomOrientation: false
            };
            const rightTouches = pads[0] || pads[1] || pads[2];
            const leftTouches = pads[3] || pads[4] || pads[5];

            if (rightTouches) {
                options.type = TouchTypes.HEAD_RIGHT;
            }
            else if (leftTouches) {
                options.type = TouchTypes.HEAD_LEFT;
            }
            if (pads[5] && !pads[0] && !pads[1] && !pads[2] && !pads[3] && !pads[4] && !pads[6]) {
                options.type = TouchTypes.TICKLE;
                this.jiboWasTickled(options);
            }
            else {
                this.jiboWasTouched(options);
            }
        };
        if (this.parent.jibo.action.events) {
            this.parent.jibo.action.events.headPat.on(this.touchHandler);
        }
    }

    /**
     * Deallocate and clean up any resources that were created.
     */
    dispose() {
        super.dispose();
        if (this.parent.jibo.action.events) {
            this.parent.jibo.action.events.headPat.removeListener(this.touchHandler);
        }
    }

    /**
     * Triggered whenever multiple jibo touch sensors are activated.
     */
    public jiboWasTouched(options:TouchOptions) {
        log.debug("jiboWasTouched", options);
        this.supplyData(new TouchEvent(options.type));
    }

    /**
     * Triggered whenever only  pad 5 is touched.
     */
    public jiboWasTickled(options:TouchOptions) {
        this.supplyData(new TouchEvent(options.type));
    }
}
