// jibo-emotion-system/src/appraisal/rules/ASRTextInputRule.ts
import {AppraisalRule} from './AppraisalRule';
import {SensorySystem} from '../../sensing/SensorySystem';
import {ASRInputSensor} from '../../sensing/sensors/ASRInputSensor';
import {SensorDataWrapper} from '../../sensing/sensors/Sensor';
import {EmotionTrace} from '../../emotion/EmotionTrace';
import {EmotionSystem} from "../../EmotionSystem";
import {EmotionPoint} from "../../common/Types";

import parentLog from '../../log';
const log = parentLog.createChild('Appraisal.ASRTextInputRule');

export {}

/**
 * A list of words that Jibo should assume is good.
 * @const {Array.<String>}
 * @internal
 */
const DESIRABLE_WORDS: string[] = [
    'good',
    'great',
    'love',
    'awesome',
    'amazing',
    'like',
    'wonderful',
    'happy',
    'lovely',
    'beautiful',
    'handsome',
    'gorgeous',
];

/**
 * Pattern used to detect words that Jibo inteprets as good and desirable.
 * @const {RegExp}
 * @internal
 */
const DESIRABLE_WORD_PATTERN = new RegExp(DESIRABLE_WORDS.join('|'), 'g');


/**
 * A list of words that Jibo should assume is bad.
 * @const {Array.<String>}
 * @internal
 */
const UNDESIRED_WORDS: string[] = [
    'hate',
    'horrible',
    'dumb',
    'stupid',
    'ugly',
    'bad',
    'suck',
    'bitch',
    'fuck',
    'shit',
    'crap',
    'moron',
    'idiot',
    'wanker',
];

/**
 * Pattern used to detect words that Jibo interprets as negative/bad.
 * @const {RegExp}
 * @internal
 */
const UNDESIRABLE_WORD_PATTERN = new RegExp(UNDESIRED_WORDS.join('|'), 'g');

/**
 * @internal
 */
export class ASRTextInputRule extends AppraisalRule {

    private asrSensor: ASRInputSensor;

    /** @override */
    constructor(name: string, parent: EmotionSystem) {
        super(name, parent);
    }

    init(sensorySystem: SensorySystem) {
        this.asrSensor = <ASRInputSensor>sensorySystem.getSensor(ASRInputSensor.SENSOR_NAME);
        if (!this.asrSensor) {
            throw new Error(`ASR sensor is required, but doesn't exist.`);
        }
        this.asrSensor.events.newDataEvent.on((sensorDataWrapper: SensorDataWrapper<string>) => {
            this.createAndRegisterImpact(sensorDataWrapper);
        });
    }

    /**
     * Function that receives a sensorDataWrapper from a sensor
     * @param sensorDataWrapper
     */
    createAndRegisterImpact(sensorDataWrapper: SensorDataWrapper<string>) {
        log.debug('Appraising ASR data:', sensorDataWrapper.sensorData);
        let impactRatio = this.appraiseData(sensorDataWrapper.sensorData);
        let myTrace = new EmotionTrace([sensorDataWrapper], this, new EmotionPoint(impactRatio));
        if (impactRatio >= .25 || impactRatio <= .25) {
            log.debug(`impactRatio was ${impactRatio}`);
            myTrace.emotionalImpact = new EmotionPoint(impactRatio);
            this.events.impactTriggered.emit(myTrace);
        } else {
            log.debug(`impactRatio was ${impactRatio}: not impactful enough to make a difference`);
        }
    }

    /** @override */
    appraiseData(txtInput: string): number {
        let badWords = txtInput.match(UNDESIRABLE_WORD_PATTERN) || [];
        let goodWords = txtInput.match(DESIRABLE_WORD_PATTERN) || [];
        let totalWordCount = txtInput.split(/[;.,\s+]/g).length;
        let wordEffectValue = goodWords.length - badWords.length;
        return wordEffectValue / totalWordCount;
    }

    dispose() {
        this.asrSensor.events.removeAllListeners();
    }

}
