// jibo-emotion-system/src/appraisal/rules/ExternalImpactRule.ts
import {AppraisalRule} from './AppraisalRule';
import {SensorySystem} from '../../sensing/SensorySystem';
import {ExternalImpactSensor} from '../../sensing/sensors/ExternalImpactSensor';
import {SensorDataWrapper} from '../../sensing/sensors/Sensor';
import {EmotionTrace} from '../../emotion/EmotionTrace';
import {EmotionSystem} from "../../EmotionSystem";
import {EmotionPoint} from "../../common/Types";

import parentLog from '../../log';
const log = parentLog.createChild('Appraisal.ExternalImpactRule');

export {}

/**
 * @internal
 */
export class ExternalImpactRule extends AppraisalRule {

    private sensor: ExternalImpactSensor;

    constructor(name: string, parent: EmotionSystem) {
        super(name, parent);
    }

    /** @override */
    init(sensorySystem: SensorySystem) {
        this.sensor = <ExternalImpactSensor>sensorySystem.getSensor(ExternalImpactSensor.SENSOR_NAME);
        if (!this.sensor) {
            throw new Error(`Skill sensor is required, but doesn't exist.`);
        }
        this.sensor.events.newDataEvent.on((sensorDataWrapper: SensorDataWrapper<EmotionPoint>) => {
            this.createAndRegisterImpact(sensorDataWrapper);
        });
    }

    /**
     * Function that receives a sensorDataWrapper from a sensor
     * @param sensorDataWrapper
     */
    createAndRegisterImpact(sensorDataWrapper: SensorDataWrapper<EmotionPoint>) {
        log.debug("creating impact", sensorDataWrapper);
        let myTrace = new EmotionTrace([sensorDataWrapper], this, this.appraiseData(sensorDataWrapper.sensorData));
        this.events.impactTriggered.emit(myTrace);

    }

    /** @override */
    appraiseData(skillData: EmotionPoint): EmotionPoint {
        return skillData;
    }

    dispose() {
        this.sensor.events.removeAllListeners();
    }

}
