// jibo-client-framework/lib/jibo-client-framework.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboClientFramework = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
class ClientRemoteObject extends events_1.EventEmitter {
    constructor(client, options) {
        super();
        this.client = client;
        this.options = options;
        client.instances[options.instanceId] = this;
        this.instanceId = options.instanceId;
    }
    sendMessage(methodName, args, sendAndForget = false) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.sendMessage(this.instanceId, methodName, args, sendAndForget);
        });
    }
    destroy() {
        delete this.client.instances[this.instanceId];
    }
}
exports.default = ClientRemoteObject;

},{"events":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const http = require("http");
const log_1 = require("./log");
const log = log_1.default.createChild('HTTPClient');
class HTTPClient extends events_1.EventEmitter {
    constructor(host, port) {
        super();
        this.host = host;
        this.port = port;
    }
    sendRequest(method, path, body, callback) {
        let options = {
            host: this.host,
            port: this.port,
            path: path,
            method: method,
            headers: {
                "Content-Type": "application/json",
                "Content-Length": body.length
            }
        };
        let request = http.request(options, (response) => {
            response.setEncoding('utf8');
            let data = '';
            response.on('data', (chunk) => {
                data += chunk.toString();
            });
            response.on('end', () => {
                let err;
                let json;
                if (data.length > 0) {
                    try {
                        json = JSON.parse(data);
                    }
                    catch (e) {
                        if (data.length < 500) {
                            log.error('could not parse JSON from request', `${this.host}:${this.port}${path}`, data);
                        }
                        else {
                            log.error('could not parse JSON from request', `${this.host}:${this.port}${path}`);
                        }
                        err = e;
                    }
                }
                if (err) {
                    callback(err);
                }
                else {
                    if (response.statusCode < 200 || response.statusCode > 299) {
                        err = new Error('HTTP Error Code ' + response.statusCode + ' ' + response.statusMessage);
                    }
                    callback(err, json);
                }
            });
        });
        request.on('error', (err) => {
            callback(new Error('problem with request: ' + err.message));
        });
        request.write(body);
        request.end();
    }
    postJSON(path, json, callback) {
        this.sendRequest('post', path, JSON.stringify(json), callback);
    }
    getJSON(path, json, callback) {
        this.sendRequest('get', path, JSON.stringify(json), callback);
    }
    get(path, callback) {
        let request = http.get('http://' + this.host + ':' + this.port + path, (response) => {
            let data = '';
            response.on('data', (chunk) => {
                data += chunk.toString();
            });
            response.on('end', () => {
                callback(null, JSON.parse(data));
            });
        });
        request.on('error', (e) => {
            callback(new Error(`Got error: ${e.message}`));
        });
    }
}
exports.default = HTTPClient;

},{"./log":7,"events":undefined,"http":undefined}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const WSClient_1 = require("./WSClient");
const RegistryClient_1 = require("./RegistryClient");
const log_1 = require("./log");
const log = log_1.default.createChild('NotificationsDispatcher');
;
class NotificationsDispatcher extends events_1.EventEmitter {
    constructor(host, port) {
        super();
        if (!RegistryClient_1.default.instance && host && port) {
            RegistryClient_1.default.createInstance(host, port);
        }
    }
    init(callback) {
        RegistryClient_1.default.instance.getRecordByName('server', (err, record) => {
            log.iferr(err, 'Error getting "server" service record');
            if (!err) {
                this._setupSockets(`ws://${record.host}:${record.port}`);
            }
            callback(err);
        });
    }
    _processNotification(message) {
        log.info('process notification', message);
        if (message && message.payload && message.payload.name) {
            log.info('emitting', message.payload.name, message.payload.payload);
            this.emit(message.payload.name, message.payload.payload);
        }
    }
    _processStatus(message) {
        if (message && message.status) {
            if (message.status !== this.lastStatus) {
                this.lastStatus = message.status;
                this.emit('StatusChanged', message.status);
                if (message.status === 1) {
                    this.emit('StatusConnected');
                }
                if (message.status === 2) {
                    this.emit('StatusDisconnected');
                }
            }
        }
    }
    _setupSockets(url) {
        if (!this.notificationsSocket) {
            let socketUrl = url + '/server/notifications';
            this.notificationsSocket = new WSClient_1.default(socketUrl);
            this.notificationsSocket.on('message', this._processNotification.bind(this));
        }
        if (!this.statusSocket) {
            let socketUrl = url + '/server/notifications/status';
            this.statusSocket = new WSClient_1.default(socketUrl);
            this.statusSocket.on('message', this._processStatus.bind(this));
        }
    }
}
exports.default = NotificationsDispatcher;

},{"./RegistryClient":4,"./WSClient":6,"./log":7,"events":undefined}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HTTPClient_1 = require("./HTTPClient");
class RegistryClient extends HTTPClient_1.default {
    constructor(host, port) {
        super(host, port);
        this.host = host;
        this.port = port;
        if (RegistryClient._instance) {
            throw new Error('Registry Client is a singleton');
        }
        RegistryClient._instance = this;
    }
    static createInstance(host, port) {
        return new RegistryClient(host, port);
    }
    static get instance() {
        return RegistryClient._instance;
    }
    addNewRecord(record, callback) {
        this.sendRequest('PUT', '/registry', JSON.stringify(record), callback);
    }
    editRecord(record, callback) {
        this.sendRequest('POST', '/registry', JSON.stringify(record), callback);
    }
    deleteRecord(record, callback) {
        this.sendRequest('DELETE', '/registry', JSON.stringify(record), callback);
    }
    getRecords(callback) {
        this.get('/registry', (error, records) => {
            callback(error, records.records);
        });
    }
    getRecordByName(serviceName, callback) {
        this.getRecords((err, records) => {
            if (err) {
                callback(err);
            }
            else if (!records) {
                callback(new Error('no records from registry'));
            }
            else {
                let found = false;
                for (let i = 0; i < records.length; i++) {
                    if (records[i].name === serviceName) {
                        found = true;
                        callback(null, records[i]);
                        break;
                    }
                }
                if (!found) {
                    callback(new Error('no record for service "' + serviceName + '" found in registry'));
                }
            }
        });
    }
}
exports.default = RegistryClient;

},{"./HTTPClient":2}],5:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const jibo_cai_utils_1 = require("jibo-cai-utils");
const events_1 = require("events");
const WSClient_1 = require("./WSClient");
const log_1 = require("./log");
const log = log_1.default.createChild('RemoteClient');
class RemoteClient extends events_1.EventEmitter {
    constructor(token) {
        super();
        this.token = token;
        this.instances = new Map();
        this._callbackHandlers = new Map();
        this._rejectionHandlers = new Map();
        this._isInitialized = false;
        this.onMessage = this.onMessage.bind(this);
        this._messageMap = {
            event: this.onEvent.bind(this),
            reply: this.onReply.bind(this),
            request: this.onRequest.bind(this),
            error: this.onError.bind(this)
        };
    }
    init(port = 11111) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._isInitialized) {
                return;
            }
            this.port = port;
            this._isInitialized = true;
            this._client = new WSClient_1.default(`ws://127.0.0.1:${this.port}`);
            this._client.on('reopen', () => {
                this.emit('reopen');
            });
            this._client.on('message', this.onMessage);
            yield jibo_cai_utils_1.PromiseUtils.promisify(cb => this._client.on('open', cb));
        });
    }
    sendMessage(instanceId, methodName, args = [], sendAndForget = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = {
                type: 'request',
                messageId: this.getMessageId(),
                instanceId,
                methodName,
                args,
                sendAndForget
            };
            this._client.send(JSON.stringify(data));
            if (!sendAndForget) {
                return yield this.waitForReply(data.messageId);
            }
        });
    }
    destroy() {
        if (this._client) {
            this._client.removeListener('message', this.onMessage);
            this._client = undefined;
        }
    }
    waitForReply(messageId) {
        return new Promise((resolve, reject) => {
            this._callbackHandlers.set(messageId, (value) => {
                resolve(value);
            });
            this._rejectionHandlers.set(messageId, (err) => {
                reject(err);
            });
        });
    }
    onMessage(messageBase) {
        this._messageMap[messageBase.type](messageBase);
    }
    onEvent(message) {
        const instance = this.instances[message.instanceId];
        if (!instance) {
            log.warn(`Cannot emit event ${message.args[0]} on instance id ${message.instanceId}`);
        }
        else if (instance instanceof events_1.EventEmitter) {
            instance.emit.apply(instance, message.args);
        }
        else {
            throw new Error(`Tried to emit an event on ${instance}, but
                it's not an EventEmitter`);
        }
    }
    onRequest(message) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const instance = this.instances[message.instanceId];
                if (!instance) {
                    log.warn(`Could not call method ${message.methodName} on id ${message.instanceId} with args ${message.args}`);
                    return;
                }
                const ret = yield instance[message.methodName].apply(instance, message.args);
                if (!message.sendAndForget) {
                    const reply = {
                        type: 'reply',
                        messageId: message.messageId,
                        value: ret
                    };
                    this._client.send(JSON.stringify(reply));
                }
            }
            catch (err) {
                let messageText = 'Unspeficied Error';
                if (err instanceof Error) {
                    messageText = err.message;
                }
                else if (typeof err === 'string') {
                    messageText = err;
                }
                const errorMessage = {
                    type: 'error',
                    messageId: message.messageId,
                    message: messageText
                };
                log.error(errorMessage);
                this._client.send(JSON.stringify(errorMessage));
            }
        });
    }
    onReply(message) {
        if (this._callbackHandlers.has(message.messageId)) {
            this._callbackHandlers.get(message.messageId)(message.value);
            this._callbackHandlers.delete(message.messageId);
            this._rejectionHandlers.delete(message.messageId);
        }
    }
    onError(message) {
        if (this._rejectionHandlers.has(message.messageId)) {
            this._rejectionHandlers.get(message.messageId)(message.message);
            this._callbackHandlers.delete(message.messageId);
            this._rejectionHandlers.delete(message.messageId);
        }
    }
    getMessageId() {
        return RemoteClient._messageId++;
    }
}
RemoteClient._messageId = 0;
exports.default = RemoteClient;

},{"./WSClient":6,"./log":7,"events":undefined,"jibo-cai-utils":undefined}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const WebSocket = require("ws-without-native");
const events_1 = require("events");
const log_1 = require("./log");
const log = log_1.default.createChild('WSClient');
class HTTPWSClient extends events_1.EventEmitter {
    constructor(host) {
        super();
        this.host = host;
        this.isReconnecting = false;
        this.onStartReconnect = this.onStartReconnect.bind(this);
        this.onError = this.onError.bind(this);
        this.onClose = this.onClose.bind(this);
        this.onMessage = this.onMessage.bind(this);
        this.connect = this.connect.bind(this);
        this.onOpen = this.onOpen.bind(this);
        this.connect();
    }
    send(json) {
        if (typeof json !== 'string') {
            try {
                json = JSON.stringify(json);
            }
            catch (e) {
                log.error('JSON.stringify: ', e);
                json = null;
            }
        }
        if (this.socket) {
            this.socket.send(json);
        }
    }
    connect() {
        log.debug('connect');
        this.socket = new WebSocket(this.host, { perMessageDeflate: false });
        this.socket.on('error', this.onError);
        this.socket.on('close', this.onClose);
        this.socket.on('message', this.onMessage);
        this.socket.on('open', this.onOpen);
    }
    onStartReconnect() {
        this.isReconnecting = true;
        this.socket.removeAllListeners();
        this.socket = null;
        setTimeout(this.connect, 1000);
    }
    onError(err) {
        log.warn('socket error', err);
        this.onStartReconnect();
    }
    onClose(code, message) {
        log.warn('socket closed', code, message);
        this.emit('close');
        this.onStartReconnect();
    }
    onMessage(data) {
        let command;
        try {
            command = JSON.parse(data.replace(/\bNaN\b/g, "0"));
        }
        catch (e) {
            const r = /at position (\d+)/g;
            const result = r.exec(e.message);
            if (data.length < 500) {
                if (result) {
                    let pos = parseInt(result[1]);
                    log.error('Invalid JSON sent to jibo client on host ' + this.host + ' at position ' + pos, data);
                }
                else {
                    log.error('Invalid JSON sent to jibo client on host ' + this.host, data);
                }
            }
            else {
                if (result) {
                    const padding = 250;
                    let pos = parseInt(result[1]);
                    data = data.substr(Math.max(0, pos - padding), Math.min(data.length - 1, pos + padding));
                    if (pos > padding) {
                        pos = padding;
                    }
                    log.error('Invalid JSON sent to jibo client on host' + this.host + ' at position ' + pos + ' of snippet', data);
                }
                else {
                    log.error('Invalid JSON sent to jibo client on host ' + this.host);
                }
            }
        }
        if (command) {
            this.emit('message', command);
        }
    }
    onOpen() {
        this.emit('open');
        if (this.isReconnecting) {
            this.isReconnecting = false;
            this.emit('reopen');
        }
    }
}
exports.default = HTTPWSClient;

},{"./log":7,"events":undefined,"ws-without-native":undefined}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Log = require("jibo-log");
exports.default = new Log('SF.Client');

},{"jibo-log":undefined}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HTTPClient_1 = require("./HTTPClient");
exports.HTTPClient = HTTPClient_1.default;
const WSClient_1 = require("./WSClient");
exports.WSClient = WSClient_1.default;
const RegistryClient_1 = require("./RegistryClient");
exports.RegistryClient = RegistryClient_1.default;
const ClientRemoteObject_1 = require("./ClientRemoteObject");
exports.ClientRemoteObject = ClientRemoteObject_1.default;
const RemoteClient_1 = require("./RemoteClient");
exports.RemoteClient = RemoteClient_1.default;
const NotificationsDispatcher_1 = require("./NotificationsDispatcher");
exports.NotificationsDispatcher = NotificationsDispatcher_1.default;

},{"./ClientRemoteObject":1,"./HTTPClient":2,"./NotificationsDispatcher":3,"./RegistryClient":4,"./RemoteClient":5,"./WSClient":6}]},{},[8])(8)
});
