// jibo-emotion-system/src/common/Types.ts
/**
 * An enum of default emotion names provided by the emotion system.
 * ```
 * import { EmotionName } from 'jibo-emotion-system';
 * const emotion = jibo.emotion.getNearestEmotion();
 * if (emotion.name ==== EmotionName.SAD) {
 *     // do a sad thing
 * }
 * ```
 * @memberof jibo.emotion
 * @typedef EmotionName
 * @type {enum}
 * @property {string} PLEASED Jibo's present emotional values are closer to (1, 0) than any other default emotion.
 * @property {string} SAD Jibo's present emotional values are closer to (-1, 0) than any other default emotion.
 * @property {string} NEUTRAL Jibo's present emotional values are closer to (0, 0) than any other default emotion.
 * @property {string} CONFIDENT Jibo's present emotional values are closer to (0, 1) than any other default emotion.
 * @property {string} UNCERTAIN Jibo's present emotional values are closer to (0, -1) than any other default emotion.
 * @property {string} PESSIMISTIC Jibo's present emotional values are closer to (-1, 1) than any other default emotion.
 * @property {string} FRUSTRATED Jibo's present emotional values are closer to (-1, -1) than any other default emotion.
 * @property {string} HOPEFUL Jibo's present emotional values are closer to (-1, 1) than any other default emotion.
 * @property {string} JOYFUL Jibo's present emotional values are closer to (1, 1) than any other default emotion.
 * @property {string} BASELINE Jibo's present emotional values are closer to (0.45, 0.2) than any other default emotion.
 */
export type EmotionName = 'PLEASED' |'SAD' |'NEUTRAL' |'CONFIDENT' |'UNCERTAIN' |'PESSIMISTIC' |'FRUSTRATED' |'HOPEFUL' |'JOYFUL' | 'BASELINE';
export namespace EmotionName {
    export const PLEASED: EmotionName = 'PLEASED';
    export const SAD: EmotionName = 'SAD';
    export const NEUTRAL: EmotionName = 'NEUTRAL';
    export const CONFIDENT: EmotionName = 'CONFIDENT';
    export const UNCERTAIN: EmotionName = 'UNCERTAIN';
    export const PESSIMISTIC: EmotionName = 'PESSIMISTIC';
    export const FRUSTRATED: EmotionName = 'FRUSTRATED';
    export const HOPEFUL: EmotionName = 'HOPEFUL';
    export const JOYFUL: EmotionName = 'JOYFUL';
    export const BASELINE: EmotionName = 'BASELINE';
}

/**
 * @class jibo.emotion.EmotionPoint
 * @prop {number} [valence=0]
 * @prop {number} [confidence=0]
 */
export class EmotionPoint {
    constructor(public valence: number = 0, public confidence: number = 0) {
    }
}

/**
 * An enum of atomic Emotional Impacts that skills can use via the skill sensor
 * to influence the valence and confidence of jibo's emotions.
 * ```
 * import { EmotionName, ImpactSize, EmotionImpact } from 'jibo-emotion-system';
 * const impact: ImpactSize = {
 *     valence: ImpactSize.HIGH_POS,
 *     confidence: ImpactSize.MEDIUM_POS
 * }
 * jibo.emotion.triggerImpact(impact);
 * const emotion = jibo.emotion.getNearestEmotion();
 * // emotion.name should equal EmotionName.JOYFUL
 * ```
 * @memberof jibo.emotion
 * @typedef ImpactSize
 * @property {string} NONE Impact an axis of Jibo's emotions with no effect.
 * @property {string} LOW_POS Impact an axis of Jibo's emotions with low positive effect.
 * @property {string} MEDIUM_POS Impact an axis of Jibo's emotions with medium positive effect.
 * @property {string} HIGH_POS Impact an axis of Jibo's emotions with high positive effect.
 * @property {string} LOW_NEG Impact an axis of Jibo's emotions with low negative effect.
 * @property {string} MEDIUM_NEG Impact an axis of Jibo's emotions with medium negative effect.
 * @property {string} HIGH_NEG Impact an axis of Jibo's emotions with high negative effect.
 */
export type ImpactSize = 'NONE' |'LOW_POS' |'MEDIUM_POS' | 'HIGH_POS' | 'LOW_NEG'| 'MEDIUM_NEG' | 'HIGH_NEG';
export namespace ImpactSize {
    export const NONE: ImpactSize = 'NONE';
    export const LOW_POS: ImpactSize = 'LOW_POS';
    export const MEDIUM_POS: ImpactSize = 'MEDIUM_POS';
    export const HIGH_POS: ImpactSize = 'HIGH_POS';
    export const LOW_NEG: ImpactSize = 'LOW_NEG';
    export const MEDIUM_NEG: ImpactSize = 'MEDIUM_NEG';
    export const HIGH_NEG: ImpactSize = 'HIGH_NEG';
}

/**
 * Interface to impact Jibo's emotional state via `jibo.emotion.triggerImpact()`
 * ```
 * import { EmotionName, ImpactSize, EmotionImpact } from 'jibo-emotion-system';
 * const impact: EmotionImpact = {
 *     valence: ImpactSize.HIGH_POS,
 *     confidence: ImpactSize.MEDIUM_POS
 * }
 * jibo.emotion.triggerImpact(impact);
 * const emotion = jibo.emotion.getNearestEmotion();
 * // emotion.name should equal EmotionName.JOYFUL
 * ```
 * @memberof jibo.emotion
 * @typedef EmotionImpact
 * @property {jibo.emotion.ImpactSize} [valence=NONE] The impact size to impact Jibo's emotional valence.
 * @property {jibo.emotion.ImpactSize} [confidence=NONE] The impact size to impact Jibo's emotional confidence.
 */
export interface EmotionImpact {
    valence?:ImpactSize;
    confidence?: ImpactSize;
}

/**
 * Constructor options for the Jibo emotion system, right now
 * this is sparse but provides framework for future in which we
 * might want to create own update loop, a la action system
 * @memberof jibo.emotion
 * @typedef InitOptions
 * @property {Object} jibo - The Jibo runtime
 */
export interface InitOptions {
    jibo: any;
}


/**
 * A set of rule names used to appraise sensor inputs
 * @internal
 */
export type RuleNames = 'ASRTextInputRule' |'TouchRule' |'SkillInputRule' |'MotorFaultRule' |'MimRule' | 'IdentityRule';
/**
 * A set of rule names used to appraise sensor inputs
 * @internal
 */
export namespace RuleNames {
    export const ASRTextInputRule: RuleNames = 'ASRTextInputRule';
    export const TouchRule: RuleNames = 'TouchRule';
    export const SkillInputRule: RuleNames = 'SkillInputRule';
    export const MotorFaultRule: RuleNames = 'MotorFaultRule';
    export const MimRule: RuleNames = 'MimRule';
    export const IdentityRule: RuleNames = 'IdentityRule';
}

/**
 * Touch types
 * @internal
 */
export type TouchTypes = 'HEAD_LEFT' | 'HEAD_RIGHT' | 'TICKLE';
/**
 * Touch types
 * @internal
 */
export namespace TouchTypes {
    export const HEAD_LEFT: TouchTypes = 'HEAD_LEFT';
    export const HEAD_RIGHT: TouchTypes = 'HEAD_RIGHT';
    export const TICKLE: TouchTypes = 'TICKLE';
}

/**
 * Identity Events
 * @internal
 */
export type IdentityEvents = 'PRESENCE_STARTED' | 'ID_ACQUIRED';
/**
 * Identity Events
 * @internal
 */
export namespace IdentityEvents {
    export const PRESENCE_STARTED: IdentityEvents = 'PRESENCE_STARTED';
    export const ID_ACQUIRED: IdentityEvents = 'ID_ACQUIRED';
}

/**
 * Options we send to jibo.action.addPlayAnimationGoal
 * @memberof jibo.emotion
 * @typedef AnimationOptions
 * @property {string} [category] The category of anim to be selected
 * @property {string} [name] The name of the goal to be played
 * @property {string[]} [includeMeta] Any additional metadata around selection of animations
 * @property {boolean} [randomOrientation] Whether orientation should be random
 * @internal
 * @private
 */
export interface AnimationOptions {
    category?: string;
    name?: string;
    includeMeta?: string[];
    randomOrientation?: boolean;
}

/**
 * Options for touch events
 * @memberof jibo.emotion
 * @typedef TouchOptions
 * @property {jibo.emotion.TouchTypes} [type] The type of touch detected
 * @property {boolean} [randomOrientation] Whether orientation should be random
 * @internal
 * @private
 */
export interface TouchOptions {
    type?: TouchTypes;
    randomOrientation?: boolean;
}


/**
 * Constant impacts of appraisal rules, in one central location
 * @memberof jibo.emotion
 * @typedef RuleImpactValues
 * @property {number} PRESENCE_STARTED_VALENCE the VALENCE impact of a PRESENCE_STARTED event.
 * @property {number} PRESENCE_STARTED_CONFIDENCE the CONFIDENCE impact of a PRESENCE_STARTED event.
 * @property {number} ID_ACQUIRED_VALENCE the VALENCE impact of a ID_ACQUIRED event.
 * @property {number} ID_ACQUIRED_CONFIDENCE the CONFIDENCE impact of a ID_ACQUIRED event.
 * @property {number} HEAD_TOUCH_VALENCE the VALENCE impact of a HEAD_TOUCH event.
 * @property {number} HEAD_TOUCH_CONFIDENCE the CONFIDENCE impact of a HEAD_TOUCH event.
 * @property {number} TICKLE_VALENCE the VALENCE impact of a TICKLE event.
 * @property {number} TICKLE_CONFIDENCE the CONFIDENCE impact of a TICKLE event.
 * @property {number} MOTOR_FAULT_VALENCE the VALENCE impact of a MOTOR_FAULT event.
 * @property {number} MOTOR_FAULT_CONFIDENCE the CONFIDENCE impact of a MOTOR_FAULT event.
 * @property {number} MIM_ERROR_VALENCE the VALENCE impact of a MIM_ERROR event.
 * @property {number} MIM_ERROR_CONFIDENCE the CONFIDENCE impact of a MIM_ERROR event.
 * @internal
 * @private
 */
export enum RuleImpactValues {
    PRESENCE_STARTED_VALENCE = 0.01,
    PRESENCE_STARTED_CONFIDENCE = 0.01,
    ID_ACQUIRED_VALENCE = 0.01,
    ID_ACQUIRED_CONFIDENCE = 0.01,
    HEAD_TOUCH_VALENCE = 0.2,
    HEAD_TOUCH_CONFIDENCE = 0.1,
    TICKLE_VALENCE = 1,
    TICKLE_CONFIDENCE = 1,
    MOTOR_FAULT_VALENCE = -2,
    MOTOR_FAULT_CONFIDENCE = -2,
    MIM_ERROR_VALENCE = 0,
    MIM_ERROR_CONFIDENCE = -1
}
