// jibo-flow-core/lib/jibo-flow-core.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboFlowCore = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var Activity = (function () {
    function Activity(name, id, className, assetPack, options) {
        this.name = name;
        this.id = id;
        this.class = className;
        this.assetPack = assetPack;
        this.options = options;
    }
    Activity.prototype.set_transitions = function (transitions) {
        this.transitions = transitions;
    };
    Activity.prototype.set_exceptions = function (exceptions) {
        this.exceptions = exceptions;
    };
    Activity.prototype.set_procedure = function (procedure) {
        this.procedure = procedure;
    };
    return Activity;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Activity;

},{}],2:[function(require,module,exports){
"use strict";
var FlowRoot = (function () {
    function FlowRoot(loader) {
        this.procedures = {};
        this.global_inits = {};
        this.loader = loader;
    }
    FlowRoot.prototype.setStartInfoFromUri = function (uri) {
        this.startUri = this.loader.getUri(uri);
        this.startProcedureOrName = this.loader.getStartProcedureOrName(uri);
    };
    FlowRoot.prototype.loadProcedure = function (currentUri, procedureObjectOrName) {
        return this.loader.loadProcedureRelative(currentUri, procedureObjectOrName);
    };
    FlowRoot.prototype.getStartActivity = function () {
        return this.getStartProcedure().getBeginActivity();
    };
    FlowRoot.prototype.getStartProcedure = function () {
        return this.loader.loadProcedure(this.startUri);
    };
    FlowRoot.prototype.getStartProcedureOrName = function () {
        return this.startProcedureOrName;
    };
    return FlowRoot;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = FlowRoot;

},{}],3:[function(require,module,exports){
"use strict";
var FlowRoot_1 = require('./FlowRoot');
var TreeLoader_1 = require('./loaders/TreeLoader');
var GojsLoader_1 = require('./loaders/GojsLoader');
var FlowRootFactory = (function () {
    function FlowRootFactory() {
    }
    FlowRootFactory.create = function (uri, loader) {
        if (loader === undefined) {
            if (typeof (uri) === 'function') {
                loader = TreeLoader_1.default;
            }
            else if (uri.endsWith(".flow")) {
                loader = new GojsLoader_1.default();
            }
            else {
                console.error('FlowRootFactory encountered an unknown file type: ' + uri);
                throw new Error('FlowRootFactory encountered an unknown file type: ' + uri);
            }
        }
        var flowRoot = new FlowRoot_1.default(loader);
        flowRoot.setStartInfoFromUri(uri);
        return flowRoot;
    };
    return FlowRootFactory;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = FlowRootFactory;

},{"./FlowRoot":2,"./loaders/GojsLoader":6,"./loaders/TreeLoader":7}],4:[function(require,module,exports){
"use strict";
var Activity_1 = require('./Activity');
var path = require('path');
var findRoot = require("find-root");
var Procedure = (function () {
    function Procedure(name, uri, local_inits, param_inits) {
        this.name = name;
        this.uri = uri;
        this.activities = {};
        this.local_inits = local_inits;
        this.param_inits = param_inits;
        this.transitions = {};
        this.exceptions = {};
        this.connections = [];
    }
    Procedure.prototype.loadPropertiesFromTreeFunction = function (tree) {
        var treeObj = tree();
        for (var activityId in treeObj) {
            var t = treeObj[activityId];
            if (typeof t === 'function') {
                t = t();
            }
            var activity = new Activity_1.default(t.name, t.id, t.class, t['asset-pack'], t.options);
            activity.transitions = t.transitions;
            activity.exceptions = t.exceptions;
            this.activities[t.id] = activity;
        }
        this.name = treeObj.meta.name;
        this.uri = treeObj.meta.uri;
        this.relativeUri = path.relative(findRoot('.'), this.uri);
        this.local_inits = {};
        this.param_inits = {};
        this.procedureConstructor = tree;
    };
    Procedure.prototype.add_transition = function (transition) {
        var map = transition.value.startsWith('~') ? this.exceptions : this.transitions;
        if (!(transition.frm in map)) {
            map[transition.frm] = [];
        }
        map[transition.frm].push(transition);
        this.connections.push(transition);
    };
    Procedure.prototype.add_activity = function (activity) {
        var transitions = activity.id in this.transitions ? this.transitions[activity.id] : [];
        var exceptions = activity.id in this.exceptions ? this.exceptions[activity.id] : [];
        activity.set_transitions(transitions);
        activity.set_exceptions(exceptions);
        activity.set_procedure(this);
        this.activities[activity.id] = activity;
    };
    Procedure.prototype.getBeginActivity = function () {
        for (var key in this.activities) {
            var activity = this.activities[key];
            if (activity.class === 'Flow.Begin') {
                return activity;
            }
        }
        throw new Error("No Flow.Begin Activity in procedure " + this.name);
    };
    Procedure.prototype.getAsyncBeginActivities = function () {
        var asyncs = [];
        for (var key in this.activities) {
            var activity = this.activities[key];
            if (activity.class === 'Flow.Begin-Parallel') {
                asyncs.push(activity);
            }
        }
        return asyncs;
    };
    return Procedure;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Procedure;

},{"./Activity":1,"find-root":undefined,"path":undefined}],5:[function(require,module,exports){
"use strict";
var Transition = (function () {
    function Transition(frm, to, value) {
        this.frm = frm;
        this.to = to;
        this.value = value;
    }
    return Transition;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Transition;

},{}],6:[function(require,module,exports){
"use strict";
var Activity_1 = require('../Activity');
var Procedure_1 = require('../Procedure');
var Transition_1 = require('../Transition');
var fs = require('fs');
var GojsLoader = (function () {
    function GojsLoader() {
        this.procedures = {};
    }
    GojsLoader.prototype.getStartProcedureOrName = function (uri) {
        var names = uri.split('/');
        var name = names[names.length - 1].replace(/\.[^.]*$/, '');
        return name;
    };
    GojsLoader.prototype.getUri = function (uri) {
        return uri;
    };
    GojsLoader.prototype.getBase = function (uri) {
        return uri.replace(/\/[^/]*$/, '');
    };
    GojsLoader.prototype.loadProcedureRelative = function (currentUri, name) {
        var baseUri = this.getBase(currentUri);
        var uri = baseUri + '/' + name + '.flow';
        return this.loadProcedure(uri);
    };
    GojsLoader.prototype.loadProcedure = function (uri) {
        var name = this.getStartProcedureOrName(uri);
        if (!this.procedures.hasOwnProperty(name)) {
            this.procedures[name] = this.readProcedure(name, uri);
        }
        return this.procedures[name];
    };
    GojsLoader.prototype.readProcedure = function (name, uri) {
        var data = fs.readFileSync(uri, 'utf8');
        var base = JSON.parse(data);
        var procedure = new Procedure_1.default(name, uri, {}, {});
        for (var i = 0; i < base.linkDataArray.length; i++) {
            var goLink = base.linkDataArray[i];
            var transitionValue = goLink.text === undefined ? "" : goLink.text;
            var transition = new Transition_1.default(goLink.from, goLink.to, transitionValue);
            procedure.add_transition(transition);
        }
        for (var i = 0; i < base.nodeDataArray.length; i++) {
            var goNode = base.nodeDataArray[i];
            var activity = new Activity_1.default(goNode.name, goNode.id, goNode.class, goNode['asset-pack'], goNode.options);
            procedure.add_activity(activity);
        }
        return procedure;
    };
    return GojsLoader;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = GojsLoader;

},{"../Activity":1,"../Procedure":4,"../Transition":5,"fs":undefined}],7:[function(require,module,exports){
"use strict";
var Procedure_1 = require('../Procedure');
var TreeLoader = (function () {
    function TreeLoader() {
    }
    TreeLoader.getStartProcedureOrName = function (data) {
        return data;
    };
    TreeLoader.getUri = function (data) {
        return data().meta.uri;
    };
    TreeLoader.loadProcedureRelative = function (uri, procedureObject) {
        var procedure = new Procedure_1.default();
        procedure.loadPropertiesFromTreeFunction(procedureObject);
        return procedure;
    };
    return TreeLoader;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = TreeLoader;

},{"../Procedure":4}],8:[function(require,module,exports){
"use strict";
var GojsLoader_1 = require('./GojsLoader');
var TreeLoader_1 = require('./TreeLoader');
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    GojsLoader: GojsLoader_1.default,
    TreeLoader: TreeLoader_1.default
};

},{"./GojsLoader":6,"./TreeLoader":7}],9:[function(require,module,exports){
"use strict";
var Activity_1 = require('./Activity');
exports.Activity = Activity_1.default;
var FlowRoot_1 = require('./FlowRoot');
exports.FlowRoot = FlowRoot_1.default;
var FlowRootFactory_1 = require('./FlowRootFactory');
exports.FlowRootFactory = FlowRootFactory_1.default;
var Procedure_1 = require('./Procedure');
exports.Procedure = Procedure_1.default;
var Transition_1 = require('./Transition');
exports.Transition = Transition_1.default;
var loaders_1 = require('./loaders');
exports.loaders = loaders_1.default;

},{"./Activity":1,"./FlowRoot":2,"./FlowRootFactory":3,"./Procedure":4,"./Transition":5,"./loaders":8}]},{},[9])(9)
});
