// jibo-state-machine/lib/jibo-state-machine.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboStateMachine = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Transition_1 = require("./Transition");
var ImmediateTransition = (function (_super) {
    __extends(ImmediateTransition, _super);
    function ImmediateTransition(name, destState) {
        var _this = _super.call(this, name, destState) || this;
        _this.onEntry = function () {
            _this.trigger();
        };
        return _this;
    }
    return ImmediateTransition;
}(Transition_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ImmediateTransition;

},{"./Transition":6}],2:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Transition_1 = require("./Transition");
var InternalTransition = (function (_super) {
    __extends(InternalTransition, _super);
    function InternalTransition(name, destState) {
        return _super.call(this, name, destState) || this;
    }
    return InternalTransition;
}(Transition_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = InternalTransition;

},{"./Transition":6}],3:[function(require,module,exports){
"use strict";
var Transition_1 = require("./Transition");
var InternalTransition_1 = require("./InternalTransition");
var TimeoutTransition_1 = require("./TimeoutTransition");
var State = (function () {
    function State(sm, name, transitions) {
        this.sm = sm;
        this.name = name;
        this._isCurrent = false;
        this.transitions = [];
        this._internalTransitions = new Map();
        this._events = new Map();
        this._typedEvents = new Map();
        sm._addState(this);
        if (transitions) {
            this.addTransition(transitions);
        }
    }
    State.prototype.addTransition = function (transition) {
        var _this = this;
        if (Array.isArray(transition)) {
            transition.forEach(function (trans) {
                _this._addTransition(trans);
            });
        }
        else {
            this._addTransition(transition);
        }
        return this;
    };
    State.prototype._addTransition = function (transition) {
        if (transition.getSourceState()) {
            throw new Error("This transition has already been assigned to a source state");
        }
        transition._setSourceState(this);
        this.transitions.push(transition);
        if (transition instanceof InternalTransition_1.default) {
            if (this._internalTransitions.has(transition.getDestinationState())) {
                throw new Error("State already has internal transition to destination" +
                    (" state: " + this._internalTransitions.get(transition.getDestinationState())));
            }
            this._internalTransitions.set(transition.getDestinationState(), transition);
        }
    };
    State.prototype.addEventTransition = function (event, destState) {
        var _this = this;
        var trans = new Transition_1.default('Event: ' + event, destState);
        var eventHandler = function (result) {
            if (_this._isCurrent) {
                trans.trigger(result);
            }
        };
        this.sm.on(event, eventHandler);
        this._events.set(event, eventHandler);
        this.addTransition(trans);
        return this;
    };
    State.prototype.addTypedEventTransition = function (event, destState) {
        var _this = this;
        var trans = new Transition_1.default('TEvent: ' + event.name, destState);
        var eventHandler = function (result) {
            if (_this._isCurrent) {
                trans.trigger(result);
            }
        };
        event.on(eventHandler);
        this._typedEvents.set(event, eventHandler);
        this.addTransition(trans);
        return this;
    };
    State.prototype._enter = function (transition, result) {
        this._isCurrent = true;
        this._lastIncomingResults = result;
        // We evaluate this state's onEntry method
        if (this.onEntry) {
            this.onEntry(transition, result);
        }
        // We evaluate this state's transitions' onEntry methods
        for (var _i = 0, _a = this.transitions; _i < _a.length; _i++) {
            var trans = _a[_i];
            trans._enter();
        }
    };
    State.prototype._exit = function (transition, done) {
        if (this.onExit) {
            this.onExit(transition);
        }
        for (var _i = 0, _a = this.transitions; _i < _a.length; _i++) {
            var trans = _a[_i];
            trans._exit();
        }
        this._isCurrent = false;
        if (this.postExit) {
            this.postExit(done);
        }
        else {
            done();
        }
    };
    State.prototype._stop = function () {
        this.transitions.forEach(function (trans) {
            if (trans instanceof TimeoutTransition_1.default) {
                trans.stop();
            }
        });
        if (this.onStop) {
            this.onStop();
        }
    };
    /**
     * Retrieves the last result that was provided to this state upon entry
     * @returns {Object}
     */
    State.prototype.getLastIncomingResult = function () {
        return this._lastIncomingResults;
    };
    /**
     * Transitions to a destination state. An `InternalTransition` needs to
     * have been added from this state to the destination state
     * Note: If this state is not the current state then an exception is thrown
     * @param {State} state The desired next state
     * @param {Object} result Optional result object to pass to next state
     */
    State.prototype.transitionTo = function (state, result) {
        var trans = this._internalTransitions.get(state);
        if (!trans) {
            throw new Error("State has no internal transition to state " + state);
        }
        if (!this._isCurrent) {
            throw new Error("Can't transition out of state that isn't current. " +
                ("This state: '" + this + "', target state: '" + state + "', ") +
                ("current state: '" + this.sm.current + "'"));
        }
        trans.trigger(result);
    };
    /**
     * Destroys the state. Should be called when state machine is getting discarded
     */
    State.prototype.destroy = function () {
        var _this = this;
        // this._isCurrent = false;
        this._events.forEach(function (handler, eventName) {
            _this.sm.removeListener(eventName, handler);
        });
        this._events.clear();
        this._typedEvents.forEach(function (handler, event) {
            event.removeListener(handler);
        });
        this._typedEvents.clear();
        this.transitions.forEach(function (trans) { return trans.destroy(); });
        // this.transitions = [];
        // this._internalTransitions.clear();
        this._lastIncomingResults = null;
        this.onEntry = null;
        this.onExit = null;
        this.onUpdate = null;
        this.onStop = null;
        this.postExit = null;
    };
    State.prototype.toString = function () {
        return 'State: ' + this.name;
    };
    return State;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = State;

},{"./InternalTransition":2,"./TimeoutTransition":5,"./Transition":6}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var events = require("events");
var fs = require("fs");
var Transition_1 = require("./Transition");
var StateMachine = (function (_super) {
    __extends(StateMachine, _super);
    function StateMachine() {
        var _this = _super.call(this) || this;
        _this.states = new Map();
        _this._inTransition = false;
        _this._loopDetection = new Map();
        _this._traceIndex = 0;
        _this._trace = [];
        return _this;
    }
    /**
     * Destroys the state machine and all of its states.
     * Should be called when state machine is getting discarded
     */
    StateMachine.prototype.destroy = function () {
        this.states.forEach(function (state) {
            state.destroy();
        });
        this.states.clear();
        this._inTransition = false;
        this._loopDetection.clear();
        this._traceIndex = 0;
        this._trace = [];
    };
    /**
     * Returns current state of state machine
     * @returns {State}
     */
    StateMachine.prototype.getCurrentState = function () {
        return this.current;
    };
    /**
     * Sets the initial state of the state machine
     * @param state
     * @returns {StateMachine}
     */
    StateMachine.prototype.setInitial = function (state) {
        this.initial = state;
        return this;
    };
    /**
     * Starts the state machine. This is required to be called once.
     */
    StateMachine.prototype.start = function () {
        if (!this.initial) {
            throw new Error("State machine does not yet have an initial state");
        }
        var initialTransition = new Transition_1.default('INITIAL TRANSITON', this.initial);
        this._addToTrace(initialTransition);
        this.current = this.initial;
        this.current._isCurrent = true;
        this.current._enter(initialTransition);
    };
    StateMachine.prototype._addState = function (state) {
        if (this.states.has(state.name)) {
            throw new Error("State with name '" + state.name + "' already exists");
        }
        if (!this.initial) {
            this.setInitial(state);
        }
        this.states.set(state.name, state);
        return this;
    };
    /**
     * Get state by name
     * @param {string} name
     * @returns {State}
     */
    StateMachine.prototype.getState = function (name) {
        return this.states.get(name);
    };
    StateMachine.prototype._addToTrace = function (transition) {
        if (this._trace.length < StateMachine.TRACE_MAX_LENGTH) {
            this._trace.push(transition);
        }
        else {
            this._trace[this._traceIndex] = transition;
            this._traceIndex = (this._traceIndex + 1) % StateMachine.TRACE_MAX_LENGTH;
        }
    };
    /**
     * Updates the current state, if it has an onUpdate callback.
     * Normal functionality of the StateMachine does not require that you call this, only if your
     * state should be connected to an update loop.
     */
    StateMachine.prototype.update = function () {
        if (this.current && this.current.onUpdate) {
            this.current.onUpdate();
        }
    };
    /**
     * Stops the current state (calls its onStop callback).
     */
    StateMachine.prototype.stop = function () {
        if (this.current) {
            this.current._stop();
            this.current = null;
        }
    };
    /**
     * Returns the trace of transitions leading up to the current one (limited
     * by StateMachine.TRACE_MAX_LENGTH)
     * @returns {Transition[]}
     */
    StateMachine.prototype.getTrace = function () {
        var out = [];
        if (this._trace.length < StateMachine.TRACE_MAX_LENGTH) {
            for (var i = 0; i < this._trace.length; i++) {
                out.push(this._trace[i]);
            }
        }
        else {
            for (var i = 0; i < this._trace.length; i++) {
                var ind = (this._traceIndex + i) % StateMachine.TRACE_MAX_LENGTH;
                out.push(this._trace[ind]);
            }
        }
        return out;
    };
    /**
     * Provides a convenient string representation of getTrace()
     * @returns {string}
     */
    StateMachine.prototype.traceToString = function () {
        var trace = this.getTrace();
        var str = '';
        trace.forEach(function (t) {
            var sName = t.getSourceState() ? t.getSourceState().name : 'none';
            str += "Source: '" + sName + "' -> Transition: '" + t.name + "' " +
                ("-> Destination: '" + t.getDestinationState().name + "' \n");
        });
        return str;
    };
    /**
     * Transitions using a particular transition
     * @param transition
     * @param result
     * @private
     */
    StateMachine.prototype._transitionTo = function (transition, result) {
        var _this = this;
        // if the state machine is stopped, do nothing
        if (!this.current) {
            return;
        }
        // See if this is the first in sequence of continuous transitions
        var isThisFirstTransition = !this._inTransition;
        if (isThisFirstTransition) {
            this._inTransition = true;
        }
        // Here we look for loops
        var destName = transition.getDestinationState().name;
        var loopCounter = this._loopDetection.get(destName);
        if (!loopCounter) {
            this._loopDetection.set(destName, 1);
        }
        else if (loopCounter > StateMachine.MAX_LOOPS) {
            throw new Error("State machine has looped '" + loopCounter + "' times in " +
                ("one transition. Next state: '" + destName + "'"));
        }
        else {
            this._loopDetection.set(destName, loopCounter + 1);
        }
        this._addToTrace(transition);
        if (this.current !== transition.getSourceState()) {
            throw new Error("Can't apply transition '" + transition + "' since the "
                + ("current state is '" + this.current + "'"));
        }
        // Notify current state and transitions of exit
        this.current._exit(transition, function () {
            // Notify next state and transitions of entry
            _this.current = transition.getDestinationState();
            _this.current._enter(transition, result);
            // When leaving the first transition in a sequence, we clear our loop
            // detection mechanism
            if (isThisFirstTransition) {
                _this._inTransition = false;
                _this._loopDetection.clear();
            }
        });
    };
    /**
     * Produces a .dot file from current state graph
     * Writes dotfile to disk at filepath
     * @param {string} filePath
     * @returns {string}
     */
    StateMachine.prototype.toDotFile = function (filePath) {
        var toVisit = [this.initial];
        var visited = new Set();
        var str = 'digraph graphname {\n';
        var _loop_1 = function () {
            var state = toVisit.pop();
            if (visited.has(state)) {
                return "continue";
            }
            this_1.states.forEach(function (state) {
                var name = state.name.replace(/\"/ig, '\\"');
                str += "\"" + name + "\" [style=filled,color=lightgray];\n";
            });
            visited.add(state);
            state.transitions.forEach(function (transition) {
                var name = state.name.replace(/\"/ig, '\\"');
                var targetName = transition.getDestinationState().name.replace(/\"/ig, '\\"');
                var transitionName = transition.name.replace(/\"/ig, '\\"');
                str += "  \"" + name + "\" -> \"" + targetName + "\" [label = \"" + transitionName + "\"];\n";
                if (!visited.has(transition.getDestinationState())) {
                    toVisit.push(transition.getDestinationState());
                }
            });
        };
        var this_1 = this;
        while (toVisit.length) {
            _loop_1();
        }
        str += '}';
        if (filePath) {
            fs.writeFileSync(filePath, str, { encoding: 'utf8' });
        }
        return str;
    };
    return StateMachine;
}(events.EventEmitter));
StateMachine.TRACE_MAX_LENGTH = 10;
StateMachine.MAX_LOOPS = 5;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = StateMachine;

},{"./Transition":6,"events":undefined,"fs":undefined}],5:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Transition_1 = require("./Transition");
var TimeoutTransition = (function (_super) {
    __extends(TimeoutTransition, _super);
    function TimeoutTransition(name, destState, timeMs) {
        var _this = _super.call(this, name, destState) || this;
        _this.onEntry = function () {
            _this._timeoutHandle = setTimeout(function () {
                _this._timeoutHandle = null;
                _this.trigger();
            }, timeMs);
        };
        _this.onExit = function () {
            if (_this._timeoutHandle) {
                clearTimeout(_this._timeoutHandle);
                _this._timeoutHandle = null;
            }
        };
        return _this;
    }
    TimeoutTransition.prototype.stop = function () {
        this.onExit();
    };
    return TimeoutTransition;
}(Transition_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = TimeoutTransition;

},{"./Transition":6}],6:[function(require,module,exports){
"use strict";
var Transition = (function () {
    function Transition(name, _destState) {
        this.name = name;
        this._destState = _destState;
    }
    /**
     * Called to trigger the current transition
     * @param {object} [result] to pass to next state
     */
    Transition.prototype.trigger = function (result) {
        if (this._sourceState) {
            this._sourceState.sm._transitionTo(this, result);
        }
    };
    Transition.prototype._enter = function () {
        if (this.onEntry) {
            this.onEntry();
        }
    };
    Transition.prototype._exit = function () {
        if (this.onExit) {
            this.onExit();
        }
    };
    /**
     * The source state for this transition
     * @returns {State}
     */
    Transition.prototype.getSourceState = function () {
        return this._sourceState;
    };
    /**
     * The destination for this transition
     * @returns {State}
     */
    Transition.prototype.getDestinationState = function () {
        return this._destState;
    };
    /**
     * Sets the source state, should only be called when adding transition to
     * the source state.
     * Can only be called once per transition
     * @param sourceState
     * @private
     */
    Transition.prototype._setSourceState = function (sourceState) {
        if (this._sourceState) {
            throw new Error("Transition '" + this + "' already has source state");
        }
        this._sourceState = sourceState;
    };
    /**
     * Destroys the transition. Should be called when state machine is getting discarded
     */
    Transition.prototype.destroy = function () {
        this.onEntry = null;
        this.onExit = null;
        this._sourceState = null;
    };
    Transition.prototype.toString = function () {
        return this.name;
    };
    return Transition;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Transition;

},{}],7:[function(require,module,exports){
"use strict";
var StateMachine_1 = require("./StateMachine");
var Transition_1 = require("./Transition");
exports.Transition = Transition_1.default;
var State_1 = require("./State");
exports.State = State_1.default;
var TimeoutTransition_1 = require("./TimeoutTransition");
exports.TimeoutTransition = TimeoutTransition_1.default;
var InternalTransition_1 = require("./InternalTransition");
exports.InternalTransition = InternalTransition_1.default;
var ImmediateTransition_1 = require("./ImmediateTransition");
exports.ImmediateTransition = ImmediateTransition_1.default;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = StateMachine_1.default;

},{"./ImmediateTransition":1,"./InternalTransition":2,"./State":3,"./StateMachine":4,"./TimeoutTransition":5,"./Transition":6}]},{},[7])(7)
});
