// @be/jukebox/src/index.ts
/// <reference path="../typings-local/index.d.ts" />

import { BeSkill } from '@be/be-framework';
import jibo = require('jibo');

import MusicLibrary from './models/MusicLibrary';
import MusicView from './views/MusicView';

/**
 * Jukebox: a touchscreen MP3 player. On open it scans the skill's `music/`
 * folder and shows a DOM overlay player. There is no flow / mim / voice grammar
 * -- the whole UI is driven by taps. Swiping down exits back to idle.
 */
class JukeboxSkill extends BeSkill {

  private musicView: MusicView = null;
  private exiting: boolean = false;
  private screenGestureHandler: (gesture: string) => void = null;

  constructor (assetPack?: string) {
    super(assetPack);
  }

  public postInit (done: () => any): void {
    // Nothing to preload; never leave `done` uncalled (Be waits on it at startup).
    done();
  }

  public preload (done: (err?: any) => void): void {
    // jibo 5 required installing the asset pack's speech delegate here; jibo 14
    // does it automatically and removed the API. Guard so we never throw.
    const es: any = (jibo as any).embodied && (jibo as any).embodied.speech;
    if (es && typeof es.installDelegate === 'function') {
      es.installDelegate(this.assetPack);
    }
    done();
  }

  public open (result?: any): void {
    this.exiting = false;
    this.subscribeSwipeDown();

    let tracks = [];
    try {
      tracks = MusicLibrary.scan(this.assetPack);
    } catch (err) {
      console.error('[jukebox] failed to load music library:', err);
    }

    this.musicView = new MusicView(tracks);
  }

  public close (done: () => void): void {
    this.unsubscribeSwipeDown();
    if (this.musicView) {
      this.musicView.cleanup();
      this.musicView = null;
    }
    done();
  }

  // Swiping down on the screen exits the skill (mirrors the recipe skill's
  // behaviour): jibo emits a global screenGesture we listen for and exit to idle.
  protected subscribeSwipeDown (): void {
    try {
      const shared: any = (jibo as any).globalEvents && (jibo as any).globalEvents.shared;
      if (!shared || !shared.screenGesture) { return; }
      this.screenGestureHandler = (gesture: string) => {
        if (String(gesture).toLowerCase() !== 'swipedown' || this.exiting) { return; }
        this.exiting = true;
        console.log('[jukebox] swipe-down detected -> exiting to idle');
        (this as any).exit();
      };
      shared.screenGesture.on(this.screenGestureHandler);
    } catch (err) {
      console.warn('[jukebox] could not subscribe to swipe-down gesture', err);
    }
  }

  protected unsubscribeSwipeDown (): void {
    if (!this.screenGestureHandler) { return; }
    try {
      const shared: any = (jibo as any).globalEvents && (jibo as any).globalEvents.shared;
      if (shared && shared.screenGesture) {
        shared.screenGesture.removeListener(this.screenGestureHandler);
      }
    } catch (err) {
      console.warn('[jukebox] could not unsubscribe swipe-down gesture', err);
    }
    this.screenGestureHandler = null;
  }
}

module.exports = JukeboxSkill;
