// @be/jukebox/src/index.ts
/// <reference path="../typings-local/index.d.ts" />

import { BeSkill } from '@be/be-framework';
import jibo = require('jibo');

import MusicLibrary from './models/MusicLibrary';
import MusicView from './views/MusicView';
import StatusOverlay from './views/StatusOverlay';

/**
 * Jukebox: a touchscreen album player. On open it shows a loading screen,
 * scans the skill's `music/` folder, then shows the cover carousel. Errors
 * stay on the loading screen so the user never only sees black.
 */
class JukeboxSkill extends BeSkill {

  private musicView: MusicView = null;
  private status: StatusOverlay = null;
  private exiting: boolean = false;
  private screenGestureHandler: (gesture: string) => void = null;

  constructor (assetPack?: string) {
    super(assetPack);
  }

  public postInit (done: () => any): void {
    done();
  }

  public preload (done: (err?: any) => void): void {
    const es: any = (jibo as any).embodied && (jibo as any).embodied.speech;
    if (es && typeof es.installDelegate === 'function') {
      es.installDelegate(this.assetPack);
    }
    done();
  }

  public open (result?: any): void {
    this.exiting = false;
    this.subscribeSwipeDown();

    // Always paint something immediately — before scan / UI work.
    try {
      this.status = StatusOverlay.show('Loading Jukebox...\nScanning music folder...');
    } catch (err) {
      console.error('[jukebox] could not show loading screen:', err);
    }

    // Defer the heavy work so the loading frame can paint first.
    const self = this;
    setTimeout(() => {
      self.finishOpen();
    }, 50);
  }

  private finishOpen (): void {
    try {
      if (this.status) {
        this.status.setLoading('Loading Jukebox...\nScanning music folder...');
      }

      const scan = MusicLibrary.scan(this.assetPack);
      console.log('[jukebox] scan result:\n' + (scan.detail || '(no detail)'));

      if (scan.error) {
        if (this.status) {
          this.status.showError(scan.error, scan.detail);
        }
        return;
      }

      if (this.status) {
        this.status.setLoading(
          scan.albums.length
            ? ('Found ' + scan.albums.length + ' album(s).\nOpening player...')
            : 'No albums found yet.\nOpening player...'
        );
      }

      try {
        this.musicView = new MusicView(scan.albums);
      } catch (err) {
        console.error('[jukebox] MusicView failed:', err);
        if (this.status) {
          this.status.showError(
            'Player UI failed to start.',
            (scan.detail ? scan.detail + '\n\n' : '') +
              (err && err.stack ? err.stack : String(err))
          );
        }
        return;
      }

      // Player is up — drop the loading layer.
      if (this.status) {
        this.status.dismiss();
        this.status = null;
      }
    } catch (err) {
      console.error('[jukebox] open failed:', err);
      if (this.status) {
        this.status.showError(
          'Jukebox failed to open.',
          err && err.stack ? err.stack : String(err)
        );
      }
    }
  }

  public close (done: () => void): void {
    this.unsubscribeSwipeDown();
    if (this.musicView) {
      try { this.musicView.cleanup(); } catch (e) { /* no-op */ }
      this.musicView = null;
    }
    if (this.status) {
      try { this.status.dismiss(); } catch (e) { /* no-op */ }
      this.status = null;
    }
    done();
  }

  protected subscribeSwipeDown (): void {
    try {
      const shared: any = (jibo as any).globalEvents && (jibo as any).globalEvents.shared;
      if (!shared || !shared.screenGesture) { return; }
      this.screenGestureHandler = (gesture: string) => {
        if (String(gesture).toLowerCase() !== 'swipedown' || this.exiting) { return; }
        this.exiting = true;
        console.log('[jukebox] swipe-down detected -> exiting to idle');
        (this as any).exit();
      };
      shared.screenGesture.on(this.screenGestureHandler);
    } catch (err) {
      console.warn('[jukebox] could not subscribe to swipe-down gesture', err);
    }
  }

  protected unsubscribeSwipeDown (): void {
    if (!this.screenGestureHandler) { return; }
    try {
      const shared: any = (jibo as any).globalEvents && (jibo as any).globalEvents.shared;
      if (shared && shared.screenGesture) {
        shared.screenGesture.removeListener(this.screenGestureHandler);
      }
    } catch (err) {
      console.warn('[jukebox] could not unsubscribe swipe-down gesture', err);
    }
    this.screenGestureHandler = null;
  }
}

module.exports = JukeboxSkill;
