// @be/jukebox/src/views/MusicView.ts
import { Track } from '../models/MusicLibrary';

/**
 * A self-contained, touchscreen MP3 player rendered as a DOM overlay on top of
 * the WebGL face canvas. It owns a single HTMLAudioElement and a track list;
 * all controls are plain DOM click handlers (a screen tap fires a click), so
 * there is no flow / mim / voice grammar involved.
 */
export default class MusicView {

  private tracks: Track[];
  private index: number = -1;

  private audio: HTMLAudioElement;
  private root: HTMLDivElement;

  private titleEl: HTMLDivElement;
  private listEl: HTMLDivElement;
  private rowEls: HTMLDivElement[] = [];
  private playPauseBtn: HTMLDivElement;
  private currentTimeEl: HTMLSpanElement;
  private totalTimeEl: HTMLSpanElement;
  private progressTrack: HTMLDivElement;
  private progressFill: HTMLDivElement;

  constructor (tracks: Track[]) {
    this.tracks = tracks || [];
    this.audio = new Audio();
    this.audio.preload = 'metadata';
    this.build();
    this.bindAudio();
  }

  /** Stop playback and remove the overlay from the DOM. */
  public cleanup (): void {
    try {
      if (this.audio) {
        this.audio.pause();
        this.audio.src = '';
      }
    } catch (e) { /* no-op */ }
    if (this.root && this.root.parentNode) {
      this.root.parentNode.removeChild(this.root);
    }
    this.root = null;
  }

  // ---- rendering ----------------------------------------------------------

  private build (): void {
    this.root = document.createElement('div');
    this.applyStyle(this.root, {
      position: 'fixed', left: '0', top: '0', width: '1280px', height: '720px',
      zIndex: '99999', background: 'linear-gradient(160deg, #1b1035 0%, #0b0b16 100%)',
      color: '#fff', fontFamily: 'Helvetica, Arial, sans-serif',
      userSelect: 'none', overflow: 'hidden'
    });

    const header = document.createElement('div');
    this.applyStyle(header, {
      position: 'absolute', left: '0', top: '0', width: '100%', height: '96px',
      display: 'flex', alignItems: 'center', paddingLeft: '48px',
      boxSizing: 'border-box', fontSize: '44px', fontWeight: 'bold',
      letterSpacing: '2px'
    });
    header.textContent = 'Jukebox';
    this.root.appendChild(header);

    // Scrollable track list.
    this.listEl = document.createElement('div');
    this.applyStyle(this.listEl, {
      position: 'absolute', left: '48px', top: '112px', right: '48px',
      bottom: '190px', overflowY: 'auto', borderRadius: '16px',
      background: 'rgba(255,255,255,0.04)'
    });
    this.root.appendChild(this.listEl);

    if (this.tracks.length === 0) {
      const empty = document.createElement('div');
      this.applyStyle(empty, {
        padding: '48px', fontSize: '30px', lineHeight: '1.5',
        color: 'rgba(255,255,255,0.75)'
      });
      empty.textContent = 'No music found. Drop .mp3 files into the skill\u2019s music folder and reopen the Jukebox.';
      this.listEl.appendChild(empty);
    } else {
      this.tracks.forEach((track, i) => this.listEl.appendChild(this.buildRow(track, i)));
    }

    this.root.appendChild(this.buildControls());

    const host = document.body || document.documentElement;
    host.appendChild(this.root);
  }

  private buildRow (track: Track, i: number): HTMLDivElement {
    const row = document.createElement('div');
    this.applyStyle(row, {
      display: 'flex', alignItems: 'center', height: '84px',
      padding: '0 28px', boxSizing: 'border-box', fontSize: '30px',
      cursor: 'pointer', borderBottom: '1px solid rgba(255,255,255,0.06)'
    });

    const num = document.createElement('span');
    this.applyStyle(num, {
      width: '54px', color: 'rgba(255,255,255,0.45)', fontSize: '24px', flex: '0 0 auto'
    });
    num.textContent = String(i + 1);
    row.appendChild(num);

    const name = document.createElement('span');
    this.applyStyle(name, {
      flex: '1 1 auto', whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis'
    });
    name.textContent = track.title;
    row.appendChild(name);

    row.addEventListener('click', () => this.selectTrack(i));
    this.rowEls[i] = row;
    return row;
  }

  private buildControls (): HTMLDivElement {
    const bar = document.createElement('div');
    this.applyStyle(bar, {
      position: 'absolute', left: '0', bottom: '0', width: '100%', height: '170px',
      boxSizing: 'border-box', padding: '18px 48px 24px',
      background: 'rgba(0,0,0,0.35)'
    });

    // Now-playing title.
    this.titleEl = document.createElement('div');
    this.applyStyle(this.titleEl, {
      fontSize: '28px', marginBottom: '10px', whiteSpace: 'nowrap',
      overflow: 'hidden', textOverflow: 'ellipsis', color: 'rgba(255,255,255,0.9)'
    });
    this.titleEl.textContent = this.tracks.length ? 'Tap a song to play' : '';
    bar.appendChild(this.titleEl);

    // Progress row: current time / seek bar / total time.
    const progressRow = document.createElement('div');
    this.applyStyle(progressRow, { display: 'flex', alignItems: 'center', marginBottom: '14px' });

    this.currentTimeEl = document.createElement('span');
    this.applyStyle(this.currentTimeEl, { width: '72px', fontSize: '22px', flex: '0 0 auto' });
    this.currentTimeEl.textContent = '0:00';
    progressRow.appendChild(this.currentTimeEl);

    this.progressTrack = document.createElement('div');
    this.applyStyle(this.progressTrack, {
      position: 'relative', flex: '1 1 auto', height: '12px', margin: '0 18px',
      background: 'rgba(255,255,255,0.18)', borderRadius: '6px', cursor: 'pointer'
    });
    this.progressFill = document.createElement('div');
    this.applyStyle(this.progressFill, {
      position: 'absolute', left: '0', top: '0', height: '100%', width: '0%',
      background: '#00d1b2', borderRadius: '6px'
    });
    this.progressTrack.appendChild(this.progressFill);
    this.progressTrack.addEventListener('click', (e) => this.onSeek(e));
    progressRow.appendChild(this.progressTrack);

    this.totalTimeEl = document.createElement('span');
    this.applyStyle(this.totalTimeEl, {
      width: '72px', fontSize: '22px', flex: '0 0 auto', textAlign: 'right'
    });
    this.totalTimeEl.textContent = '0:00';
    progressRow.appendChild(this.totalTimeEl);

    bar.appendChild(progressRow);

    // Transport buttons: prev / play-pause / next.
    const controls = document.createElement('div');
    this.applyStyle(controls, {
      display: 'flex', alignItems: 'center', justifyContent: 'center'
    });

    const prevBtn = this.buildButton('\u23EE', () => this.prev(), false);
    this.playPauseBtn = this.buildButton('\u25B6', () => this.togglePlay(), true);
    const nextBtn = this.buildButton('\u23ED', () => this.next(), false);

    controls.appendChild(prevBtn);
    controls.appendChild(this.playPauseBtn);
    controls.appendChild(nextBtn);
    bar.appendChild(controls);

    return bar;
  }

  private buildButton (glyph: string, onClick: () => void, primary: boolean): HTMLDivElement {
    const btn = document.createElement('div');
    const size = primary ? 84 : 64;
    this.applyStyle(btn, {
      width: size + 'px', height: size + 'px', margin: '0 22px',
      borderRadius: '50%', display: 'flex', alignItems: 'center',
      justifyContent: 'center', cursor: 'pointer',
      fontSize: (primary ? 40 : 30) + 'px',
      background: primary ? '#00d1b2' : 'rgba(255,255,255,0.14)',
      color: primary ? '#04231f' : '#fff'
    });
    btn.textContent = glyph;
    btn.addEventListener('click', onClick);
    return btn;
  }

  // ---- playback -----------------------------------------------------------

  private bindAudio (): void {
    this.audio.addEventListener('timeupdate', () => this.onTimeUpdate());
    this.audio.addEventListener('loadedmetadata', () => this.onTimeUpdate());
    this.audio.addEventListener('ended', () => this.next());
    this.audio.addEventListener('play', () => this.setPlayGlyph(true));
    this.audio.addEventListener('pause', () => this.setPlayGlyph(false));
    this.audio.addEventListener('error', () => {
      console.error('[jukebox] audio error for', this.audio.src,
        this.audio.error && this.audio.error.code);
    });
  }

  private selectTrack (i: number): void {
    if (i < 0 || i >= this.tracks.length) { return; }
    this.index = i;
    const track = this.tracks[i];
    this.titleEl.textContent = track.title;
    this.highlightActive();
    this.audio.src = track.url;
    const p: any = this.audio.play();
    if (p && typeof p.catch === 'function') {
      p.catch((err: any) => console.error('[jukebox] play() rejected:', err && err.name));
    }
  }

  private togglePlay (): void {
    if (this.index === -1) {
      this.selectTrack(0);
      return;
    }
    if (this.audio.paused) {
      const p: any = this.audio.play();
      if (p && typeof p.catch === 'function') {
        p.catch((err: any) => console.error('[jukebox] play() rejected:', err && err.name));
      }
    } else {
      this.audio.pause();
    }
  }

  private next (): void {
    if (!this.tracks.length) { return; }
    this.selectTrack((this.index + 1) % this.tracks.length);
  }

  private prev (): void {
    if (!this.tracks.length) { return; }
    // Restart the current track if we're more than 3s in, else go to previous.
    if (this.audio.currentTime > 3) {
      this.audio.currentTime = 0;
      return;
    }
    const prevIndex = (this.index - 1 + this.tracks.length) % this.tracks.length;
    this.selectTrack(prevIndex);
  }

  private onSeek (e: MouseEvent): void {
    if (!this.audio.duration || isNaN(this.audio.duration)) { return; }
    const rect = this.progressTrack.getBoundingClientRect();
    const ratio = Math.max(0, Math.min(1, (e.clientX - rect.left) / rect.width));
    this.audio.currentTime = ratio * this.audio.duration;
  }

  private onTimeUpdate (): void {
    const dur = this.audio.duration;
    const cur = this.audio.currentTime;
    const ratio = dur && !isNaN(dur) ? (cur / dur) : 0;
    if (this.progressFill) { this.progressFill.style.width = (ratio * 100) + '%'; }
    if (this.currentTimeEl) { this.currentTimeEl.textContent = MusicView.formatTime(cur); }
    if (this.totalTimeEl) { this.totalTimeEl.textContent = MusicView.formatTime(dur); }
  }

  private setPlayGlyph (playing: boolean): void {
    if (this.playPauseBtn) { this.playPauseBtn.textContent = playing ? '\u23F8' : '\u25B6'; }
  }

  private highlightActive (): void {
    this.rowEls.forEach((row, i) => {
      if (!row) { return; }
      row.style.background = i === this.index ? 'rgba(0,209,178,0.22)' : 'transparent';
    });
  }

  private static formatTime (seconds: number): string {
    if (!seconds || isNaN(seconds) || !isFinite(seconds)) { return '0:00'; }
    const m = Math.floor(seconds / 60);
    const s = Math.floor(seconds % 60);
    return m + ':' + (s < 10 ? '0' + s : String(s));
  }

  private applyStyle (el: HTMLElement, style: { [k: string]: string }): void {
    for (const key in style) {
      if (style.hasOwnProperty(key)) {
        (el.style as any)[key] = style[key];
      }
    }
  }
}
