// jibo-kb/lib/jibo-kb.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboKb = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs-extra");
var path = require("path");
var stream = require("stream");
var request = require("@jibo/request");
var url = require("url");
var uuid = require("uuid");
var blobToBuffer = require("blob-to-buffer");
var decorators_1 = require("./decorators");
var log_1 = require("./log");
var log = log_1.default.createChild('Asset');
var Asset = (function () {
    function Asset(filenameOrURL, subtype, ext) {
        if (!filenameOrURL) {
            this._id = uuid.v4();
            this.subtype = subtype || 'asset';
            this.ext = ext;
        }
        else {
            this.setSelfFromFilenameOrURL(filenameOrURL);
        }
    }
    Asset.prototype.setRootDir = function (rootDir) {
        this.rootDir = rootDir;
    };
    Asset.prototype.filename = function () {
        var filename = this._id + "." + this.subtype;
        if (this.ext) {
            filename += "." + this.ext;
        }
        return filename;
    };
    Asset.prototype.fullFilenameOrURL = function () {
        if (!this.rootDir) {
            throw new Error('asset rootDir not set');
        }
        var filename;
        if (this.rootDir.startsWith('http://')) {
            filename = this._url();
        }
        else {
            filename = this.filename();
            filename = path.join(this.rootDir, 'assets', filename);
        }
        return filename;
    };
    Asset.prototype.toString = function () { return this.fullFilenameOrURL(); };
    Asset.prototype.save = function (data, callback) {
        var _this = this;
        if (this.rootDir.startsWith('http://')) {
            if ((data instanceof Buffer) || (data instanceof stream.Stream)) {
                this._saveViaWeb(data, callback);
            }
            else {
                blobToBuffer(data, function (err, buffer) {
                    log.iferr(err, 'blobToBuffer');
                    if (!err) {
                        _this._saveViaWeb(buffer, callback);
                    }
                    else {
                        callback(err);
                    }
                });
            }
        }
        else {
            if (data instanceof Buffer) {
                this._saveBuffer(data, callback);
            }
            else if (data instanceof stream.Stream) {
                this._saveStream(data, callback);
            }
            else {
                blobToBuffer(data, function (err, buffer) {
                    log.iferr(err, 'blobToBuffer');
                    if (!err) {
                        _this._saveBuffer(buffer, callback);
                    }
                    else {
                        callback(err);
                    }
                });
            }
        }
    };
    Asset.prototype.setSelfFromFilenameOrURL = function (filenameOrURL) {
        if (filenameOrURL.startsWith('http://')) {
            this._setSelfFromURL(filenameOrURL);
        }
        else {
            this._setSelfFromFilename(filenameOrURL);
        }
    };
    Asset.prototype.load = function (callback) {
        if (this.rootDir.startsWith('http://')) {
            this._loadBufferViaWeb(callback);
        }
        else {
            var filename_1 = this.fullFilenameOrURL();
            fs.readFile(filename_1, function (err, data) {
                log.iferr(err, 'readFile on', filename_1);
                callback(err, data);
            });
        }
    };
    Asset.prototype.loadStream = function () {
        if (this.rootDir.startsWith('http://')) {
            return this._loadStreamViaWeb();
        }
        else {
            var filename = this.fullFilenameOrURL();
            return fs.createReadStream(filename);
        }
    };
    Asset.prototype.loadBlob = function (callback) {
        this.load(function (err, data) {
            var blob;
            log.iferr(err, 'could not load asset file');
            if (!err) {
                blob = new Blob(data, { type: 'application/octet-binary' });
            }
            callback(err, blob);
        });
    };
    Asset.prototype.remove = function (callback) {
        if (this.rootDir.startsWith('http://')) {
            this._removeViaWeb(callback);
        }
        else {
            var filename_2 = this.fullFilenameOrURL();
            log.info('removing asset file', filename_2);
            fs.unlink(filename_2, function (err) {
                log.iferr(err, 'fs.unlink of', filename_2);
                if (!err) {
                    log.debug('asset file', filename_2, 'removed from disk');
                }
                callback(err);
            });
        }
    };
    Asset.prototype._url = function () {
        if (!this.rootDir) {
            throw new Error('asset rootDir not set');
        }
        var filename = this.filename();
        return this.rootDir + "/asset/" + filename;
    };
    Asset.prototype._saveViaWeb = function (data, callback) {
        var _this = this;
        var url = this._url();
        var opts = {
            body: data,
            headers: { 'Content-Type': 'application/octet-stream' }
        };
        request.post(url, opts, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.post', url);
            callback(err, url);
        });
    };
    Asset.prototype._setSelfFromFilename = function (filename) {
        if (path.isAbsolute(filename)) {
            this.rootDir = path.dirname(path.dirname(filename));
            filename = path.basename(filename);
        }
        var re = new RegExp("^([^.]*)[.]([^.]*)([.]([^.]*))?$");
        var vals = filename.match(re);
        if (!vals) {
            throw new Error('filename did not parse');
        }
        this._id = vals[1];
        this.subtype = vals[2] || '';
        this.ext = vals[4] || '';
    };
    Asset.prototype._setSelfFromURL = function (httpurl) {
        var urlParts = url.parse(httpurl);
        var pathParts = urlParts.pathname.split('/');
        var rootPath = pathParts.slice(0, 3).join('/');
        this.rootDir = "http://" + urlParts.hostname + ":" + urlParts.port + "/" + rootPath;
        var filename = pathParts[pathParts.length - 1];
        this._setSelfFromFilename(filename);
    };
    Asset.prototype._removeViaWeb = function (callback) {
        var _this = this;
        var url = this._url();
        request.del(url, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.delete', url);
            callback(err);
        });
    };
    Asset.prototype._saveBuffer = function (buffer, callback) {
        if (!buffer) {
            callback();
            return;
        }
        var filename = this.fullFilenameOrURL();
        fs.ensureDir(path.dirname(filename), function (err) {
            log.iferr(err, 'ensureDir');
            if (!err) {
                log.info('writing asset file', filename);
                fs.writeFile(filename, buffer, function (err) {
                    log.iferr(err, 'file write error', filename);
                    callback(err, filename);
                });
            }
            else {
                callback(err);
            }
        });
    };
    Asset.prototype._saveStream = function (stream, callback) {
        if (!stream) {
            callback();
            return;
        }
        var filename = this.fullFilenameOrURL();
        fs.ensureDir(path.dirname(filename), function (err) {
            log.iferr(err, 'ensureDir');
            if (!err) {
                log.info('writing asset file', filename);
                var writeStream = fs.createWriteStream(filename);
                stream.pipe(writeStream);
                writeStream.on('finish', function () {
                    callback(null, filename);
                });
                writeStream.on('error', function (err) {
                    log.error('file write error', filename, err);
                });
            }
            else {
                callback(err);
            }
        });
    };
    Asset.prototype._loadBufferViaWeb = function (callback) {
        var _this = this;
        var url = this._url();
        var opts = {
            headers: { 'Content-Type': 'application/octet-stream' },
            encoding: null
        };
        request.get(url, opts, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.get', url);
            callback(err, body);
        });
    };
    Asset.prototype._loadStreamViaWeb = function () {
        var url = this._url();
        var opts = {
            headers: { 'Content-Type': 'application/octet-stream' },
            encoding: null
        };
        return request.get(url, opts);
    };
    Asset.prototype._checkStatusCode = function (err, res, message) {
        if (message === void 0) { message = null; }
        if (!err) {
            if (res.statusCode < 200 || res.statusCode > 299) {
                var description = 'HTTP Error Code ' + res.statusCode;
                if (message) {
                    description += message;
                }
                err = new Error(description);
            }
        }
        return err;
    };
    return Asset;
}());
__decorate([
    decorators_1.promisify
], Asset.prototype, "save", null);
__decorate([
    decorators_1.promisify
], Asset.prototype, "load", null);
__decorate([
    decorators_1.promisify
], Asset.prototype, "loadBlob", null);
__decorate([
    decorators_1.promisify
], Asset.prototype, "remove", null);
exports.default = Asset;

},{"./decorators":12,"./log":13,"@jibo/request":undefined,"blob-to-buffer":undefined,"fs-extra":undefined,"path":undefined,"stream":undefined,"url":undefined,"uuid":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var log_1 = require("./log");
var log = log_1.default.createChild('Cache');
var Cache = (function () {
    function Cache() {
    }
    Cache.prototype.fetch = function (id, quietly) {
        if (quietly === void 0) { quietly = false; }
        var node = undefined;
        if (id in this) {
            node = this[id];
        }
        else {
            if (!quietly) {
                log.warn('fetch id', id, 'not in cache');
            }
        }
        return node;
    };
    Cache.prototype.add = function (node, quietly) {
        if (quietly === void 0) { quietly = false; }
        if (node && node._id) {
            this._add(node._id, node, quietly);
        }
        else {
            log.error('add: can not add invalid object', node);
        }
    };
    Cache.prototype.remove = function (idOrNode, quietly) {
        if (quietly === void 0) { quietly = false; }
        var id = this._toId(idOrNode);
        if (id) {
            if (id in this) {
                delete this[id];
            }
            else {
                if (!quietly) {
                    log.warn('remove: id not in cache', id);
                }
            }
        }
        else {
            log.error('remove: invalid idOrNode', idOrNode);
        }
    };
    Cache.prototype.isPresent = function (idOrNode) {
        var id = this._toId(idOrNode);
        if (!id) {
            log.error('isPresent: invalid idOrNode', idOrNode);
        }
        return (id in this);
    };
    Cache.prototype.interceptLoad = function (id, callback, load) {
        var _this = this;
        var node = this.fetch(id, true);
        if (node) {
            setImmediate(function () {
                callback(null, node);
            });
        }
        else {
            load(id, function (err, loadedNode) {
                if (!err) {
                    if (loadedNode) {
                        _this.add(loadedNode);
                    }
                    else {
                        _this._add(id, null);
                    }
                }
                callback(err, loadedNode);
            });
        }
    };
    Cache.prototype._add = function (_id, value, quietly) {
        if (quietly === void 0) { quietly = false; }
        if (!(_id in this)) {
            this[_id] = value;
        }
        else if (this[_id] === null) {
            this[_id] = value;
        }
        else {
            if (Object.is(value, this[_id])) {
                if (!quietly) {
                    log.warn('add: there was already an entry in the cache for node id', _id);
                }
            }
            else {
                log.warn('add: existing entry for node id', _id, 'and new object does not match it!');
                this[_id] = value;
            }
        }
    };
    Cache.prototype._toId = function (idOrNode) {
        if (typeof idOrNode === 'object') {
            idOrNode = idOrNode._id;
        }
        return idOrNode;
    };
    return Cache;
}());
exports.default = Cache;

},{"./log":13}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Collection = require("nedb");
var log_1 = require("./log");
var log = log_1.default.createChild('Database');
var Database = (function () {
    function Database(filename) {
        this.filename = filename;
    }
    Database.prototype.init = function (callback) {
        var _this = this;
        this.nodes = new Collection({ filename: this.filename });
        this.nodes.loadDatabase(function (err) {
            log.iferr(err, 'Collection#loadDatabase for nodes', _this.filename);
            if (!err) {
                _this.nodes.count({}, function (err, count) {
                    log.iferr(err, 'error counting nodes in', _this.filename);
                    if (!err) {
                        log.info(count, 'KB node' + ((count === 1) ? '' : 's') + ' in', _this.filename);
                    }
                    callback(err, count);
                });
            }
            else {
                callback(err);
            }
        });
    };
    Database.prototype.load = function (id, callback) {
        var _this = this;
        this.nodes.findOne({ _id: id }, function (err, node) {
            log.iferr(err, 'Database#findOne of', id, 'on', _this.filename);
            callback(err, node);
        });
    };
    Database.prototype.loadList = function (ids, callback) {
        var _this = this;
        this.nodes.find({ _id: { $in: ids } }, function (err, nodes) {
            log.iferr(err, 'Database#find of _id $in ', ids, 'on', _this.filename);
            callback(err, nodes);
        });
    };
    Database.prototype.loadNodeOfType = function (nodeType, callback) {
        var _this = this;
        this.nodes.findOne({ type: nodeType }, function (err, node) {
            log.iferr(err, 'Database#findOne of type', nodeType, 'on', _this.filename);
            callback(err, node);
        });
    };
    Database.prototype.loadNodesOfType = function (nodeType, callback) {
        var q = { type: nodeType };
        this.nodes.find(q).exec(function (err, nodes) {
            log.iferr(err, 'Collection#find error');
            callback(err, nodes);
        });
    };
    Database.prototype.save = function (node, callback) {
        var _this = this;
        if (!node._id) {
            throw new Error('node does not have an id');
        }
        this.nodes.update({ _id: node._id }, node, { upsert: true }, function (err, numReplaced) {
            log.iferr(err, 'Collection#update error', _this.filename);
            if (err) {
                node._id = undefined;
            }
            if (callback) {
                callback(err, node);
            }
        });
    };
    Database.prototype.remove = function (idOrNode, callback) {
        var _this = this;
        var id = this._toId(idOrNode);
        this.nodes.remove({ _id: id }, {}, function (err, numRemoved) {
            log.iferr(err, 'Database#remove of', id, 'on', _this.filename);
            if (callback) {
                callback(err, numRemoved);
            }
        });
    };
    Database.prototype._toId = function (idOrNode) {
        if (typeof idOrNode === 'object') {
            idOrNode = idOrNode._id;
        }
        return idOrNode;
    };
    return Database;
}());
exports.default = Database;

},{"./log":13,"nedb":undefined}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var fsExtra = require("fs-extra");
var KnowledgeDatabase_1 = require("./KnowledgeDatabase");
var log_1 = require("./log");
var log = log_1.default.createChild('DatabaseManager');
var DatabaseManager = (function () {
    function DatabaseManager() {
    }
    DatabaseManager.get = function (kbName, callback) {
        var _this = this;
        var database = this.databases[kbName];
        if (database) {
            setImmediate(function () {
                callback(null, database);
            });
        }
        else {
            this.exists(kbName, function (err, exists) {
                if (err || !exists) {
                    callback(new Error('KB slice ' + kbName + ' does not exist, create it first'));
                }
                else {
                    database = new KnowledgeDatabase_1.default(kbName);
                    _this.databases[kbName] = database;
                    database.init(function (err) {
                        log.iferr(err, 'KnowledgeDatabase.init');
                        if (!err) {
                            callback(null, database);
                        }
                        else {
                            callback(err);
                        }
                    });
                }
            });
        }
    };
    DatabaseManager.exists = function (kbName, callback) {
        var dbFilename = KnowledgeDatabase_1.default.getKbFilename(kbName);
        fs.access(dbFilename, function (err) {
            var exists = !err;
            callback(null, exists);
        });
    };
    Object.defineProperty(DatabaseManager.prototype, "databases", {
        get: function () {
            return DatabaseManager.databases;
        },
        enumerable: true,
        configurable: true
    });
    DatabaseManager.prototype.create = function (kbName, callback) {
        if (this.databases[kbName]) {
            setImmediate(function () {
                callback(null, false);
            });
        }
        else {
            var dbDirectory = KnowledgeDatabase_1.default.getKbDirectory(kbName);
            fsExtra.ensureDir(dbDirectory, function (err) {
                if (err) {
                    callback(err);
                }
                else {
                    var dbFilename = KnowledgeDatabase_1.default.getKbFilename(kbName);
                    fs.open(dbFilename, 'ax', function (err, fd) {
                        var created = false;
                        if (err) {
                            if (err && err.code === 'EEXIST') {
                                err = null;
                                created = false;
                            }
                        }
                        else {
                            fs.close(fd);
                            created = true;
                        }
                        callback(err, created);
                    });
                }
            });
        }
    };
    DatabaseManager.prototype.exists = function (kbName, callback) {
        DatabaseManager.exists(kbName, callback);
    };
    DatabaseManager.prototype.get = function (kbName, callback) {
        DatabaseManager.get(kbName, callback);
    };
    DatabaseManager.prototype.release = function (kbName) {
        delete this.databases[kbName];
    };
    return DatabaseManager;
}());
DatabaseManager.databases = {};
exports.DatabaseManager = DatabaseManager;
exports.default = DatabaseManager;

},{"./KnowledgeDatabase":6,"./log":13,"fs":undefined,"fs-extra":undefined}],5:[function(require,module,exports){
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var request = require("@jibo/request");
var querystring = require("querystring");
var decorators_1 = require("./decorators");
var Model_1 = require("./Model");
var KnowledgeDatabase_1 = require("./KnowledgeDatabase");
var Node_1 = require("./Node");
var Asset_1 = require("./Asset");
var DatabaseManager_1 = require("./DatabaseManager");
var LoopModel_1 = require("./LoopModel");
var UserNode_1 = require("./UserNode");
var log_1 = require("./log");
var log = log_1.default.createChild('KnowledgeBase');
var KnowledgeBase = (function () {
    function KnowledgeBase() {
        this.Asset = Asset_1.default;
        this.KnowledgeDatabase = KnowledgeDatabase_1.default;
        this.Model = Model_1.default;
        this.Node = Node_1.default;
        this.databaseManager = new DatabaseManager_1.default();
        this.onInitCallbacks = [];
        this.registerModelClass('jibo/loop', LoopModel_1.default);
        this.registerNodeClass('user', UserNode_1.default, 'jibo/loop');
    }
    KnowledgeBase.prototype.init = function (service, callback) {
        this.httpUrl = 'http://' + service.host + ':' + service.port;
        process.nextTick(callback);
        this.onInitCallbacks.forEach(function (onInitCallback) {
            process.nextTick(onInitCallback);
        });
        this.onInitCallbacks = [];
    };
    KnowledgeBase.prototype.onInit = function (callback) {
        if (!this.httpUrl) {
            this.onInitCallbacks.push(callback);
        }
        else {
            process.nextTick(function () {
                callback(null);
            });
        }
    };
    KnowledgeBase.prototype.initLoop = function () {
        var loop = this.createModel('/jibo/loop');
        this.loop = loop;
    };
    KnowledgeBase.prototype.initMedia = function () {
        return;
    };
    KnowledgeBase.prototype.createSlice = function (sliceName, httpUrl, callback) {
        var _this = this;
        if (typeof httpUrl === 'function' && !callback) {
            callback = httpUrl;
            httpUrl = null;
        }
        httpUrl = httpUrl || this.httpUrl;
        var escapedName = querystring.escape(sliceName);
        var url = this.httpUrl + "/v1/kb/" + escapedName + "/create";
        request.post(url, { json: true }, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.post', url);
            var created = body && body.created;
            if (callback) {
                callback(err, created);
            }
        });
    };
    KnowledgeBase.prototype.existsSlice = function (sliceName, httpUrl, callback) {
        var _this = this;
        if (typeof httpUrl === 'function' && !callback) {
            callback = httpUrl;
            httpUrl = null;
        }
        httpUrl = httpUrl || this.httpUrl;
        var escapedName = querystring.escape(sliceName);
        var url = this.httpUrl + "/v1/kb/" + escapedName + "/exists";
        request.get(url, { json: true }, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.get', url);
            var exists = body && body.exists;
            if (callback) {
                callback(err, exists);
            }
        });
    };
    KnowledgeBase.prototype.createModel = function (kbNames, httpUrl) {
        kbNames = this._toArray(kbNames);
        httpUrl = httpUrl || this.httpUrl;
        if (!httpUrl) {
            throw new Error('Knowledge Base has not been inited');
        }
        var newKBNames = [];
        kbNames.forEach(function (kbName) {
            if (!kbName.startsWith('/')) {
                throw new Error('kb#createModel: kb names must be rooted (must start with /)');
            }
            else {
                kbName = kbName.substr(1);
            }
            newKBNames.push(kbName);
        });
        var ModelClass = this.findModelClass(newKBNames);
        return new ModelClass(newKBNames, httpUrl);
    };
    KnowledgeBase.prototype.registerNodeClass = function (nodeType, classConstructor, kbName) {
        if (kbName === void 0) { kbName = '*'; }
        Node_1.default.registerNodeClass(nodeType, classConstructor, kbName);
    };
    KnowledgeBase.prototype.registerModelClass = function (kbNames, classConstructor) {
        Model_1.default.registerModelClass(kbNames, classConstructor);
    };
    KnowledgeBase.prototype.findNodeClass = function (nodeType, kbName) {
        return Node_1.default.findNodeClass(nodeType, kbName);
    };
    KnowledgeBase.prototype.findModelClass = function (kbNames) {
        return Model_1.default.findModelClass(kbNames);
    };
    KnowledgeBase.prototype.removeSlice = function (sliceName, callback) {
        var _this = this;
        var escapedName = querystring.escape(sliceName);
        var url = this.httpUrl + "/v1/kb/" + escapedName + "/remove/yesiamsure";
        request.del(url, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.del', url);
            if (callback) {
                callback(err);
            }
        });
    };
    KnowledgeBase.prototype.removeAll = function (callback) {
        var _this = this;
        var url = this.httpUrl + "/v1/removeall/yesiamsure";
        request.del(url, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.del', url);
            if (callback) {
                callback(err);
            }
        });
    };
    KnowledgeBase.prototype._checkStatusCode = function (err, res, message) {
        if (message === void 0) { message = null; }
        if (!err) {
            if (res.statusCode < 200 || res.statusCode > 299) {
                var description = 'HTTP Error Code ' + res.statusCode;
                if (message) {
                    description += message;
                }
                err = new Error(description);
            }
        }
        return err;
    };
    KnowledgeBase.prototype._toArray = function (items) {
        if (!Array.isArray(items)) {
            return [items];
        }
        else {
            return items;
        }
    };
    return KnowledgeBase;
}());
__decorate([
    decorators_1.promisify
], KnowledgeBase.prototype, "onInit", null);
__decorate([
    decorators_1.promisify
], KnowledgeBase.prototype, "createSlice", null);
__decorate([
    decorators_1.promisify
], KnowledgeBase.prototype, "existsSlice", null);
__decorate([
    decorators_1.promisify
], KnowledgeBase.prototype, "removeSlice", null);
__decorate([
    decorators_1.promisify
], KnowledgeBase.prototype, "removeAll", null);
exports.default = KnowledgeBase;

},{"./Asset":1,"./DatabaseManager":4,"./KnowledgeDatabase":6,"./LoopModel":7,"./Model":8,"./Node":9,"./UserNode":10,"./decorators":12,"./log":13,"@jibo/request":undefined,"querystring":undefined}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs-extra");
var path = require("path");
var Database_1 = require("./Database");
var Node_1 = require("./Node");
var log_1 = require("./log");
var log = log_1.default.createChild('KnowledgeDatabase');
var reservedNames = ['nodes', 'assets'];
var RunMode = {
    SIMULATOR: "SIMULATOR",
    REMOTELY: "REMOTELY",
    ON_ROBOT: "ON_ROBOT",
    UNIT_TESTS: "UNIT_TESTS"
};
var KnowledgeDatabase = (function () {
    function KnowledgeDatabase(kbName) {
        this.kbName = kbName;
        KnowledgeDatabase._validateKbName(kbName);
        this.dbDirectory = KnowledgeDatabase.getKbDirectory(this.kbName);
        this.dbFilename = KnowledgeDatabase.getKbFilename(this.kbName);
    }
    KnowledgeDatabase.getRootDirectory = function () {
        if (this._onRobot()) {
            return '/opt/jibo/Knowledge';
        }
        else {
            return path.join(process.env.HOME || process.env.USERPROFILE, '.jibo', 'kb');
        }
    };
    KnowledgeDatabase.getKbDirectory = function (kbName) {
        KnowledgeDatabase._validateKbName(kbName);
        var rootDirectory = KnowledgeDatabase.getRootDirectory();
        return path.join(rootDirectory, kbName);
    };
    KnowledgeDatabase.getKbFilename = function (kbName) {
        KnowledgeDatabase._validateKbName(kbName);
        var dbDirectory = KnowledgeDatabase.getKbDirectory(kbName);
        return path.join(dbDirectory, 'nodes');
    };
    KnowledgeDatabase._validateKbName = function (kbName) {
        if (kbName.length > 200) {
            throw new Error('kb name is too long');
        }
        if (/[^a-z0-9\/_-]/.test(kbName)) {
            throw new Error('illegal character in kb name');
        }
        if (kbName.endsWith('/')) {
            throw new Error('kb name ends with a slash');
        }
        var parts = kbName.split('/');
        if (parts.length > 10) {
            throw new Error('kb name has too many sub directories');
        }
        parts.forEach(function (part, n) {
            if (part.length === 0 && n !== 0) {
                throw new Error('kb name component is empty');
            }
            if (part.length > 75) {
                throw new Error('kb name component is too long');
            }
            if (/^[^a-z0-9]/.test(part)) {
                throw new Error('kb names must start with an alphanumeric character');
            }
            if (reservedNames.indexOf(part) >= 0) {
                throw new Error('kb name is reserved word');
            }
        });
    };
    KnowledgeDatabase._onRobot = function () {
        var runMode = process.env.runMode || process.env.RUNMODE;
        if (!runMode && process.platform === 'linux' && process.arch === 'arm') {
            runMode = RunMode.ON_ROBOT;
        }
        return (runMode === RunMode.ON_ROBOT);
    };
    KnowledgeDatabase.prototype.init = function (callback) {
        var _this = this;
        this._setupDirectory(function (err) {
            log.iferr(err, 'KnowledgeDatabase#_setupDirectory');
            if (!err) {
                _this.database = new Database_1.default(_this.dbFilename);
                _this.database.init(function (err, nodeCount) {
                    log.iferr(err, 'Database#init');
                    if (!err) {
                        if (nodeCount === 0) {
                            log.info('creating root node for', _this.kbName);
                            var NodeClass = Node_1.default.findNodeClass('root', _this.kbName);
                            var rootNode = new NodeClass('root');
                            _this.adoptNodeAsOurOwn(rootNode);
                            rootNode.save(function (err) {
                                log.iferr(err, 'error creating initial root node for', _this.kbName);
                                callback(err);
                            });
                        }
                        else {
                            callback(err);
                        }
                    }
                    else {
                        callback(err);
                    }
                });
            }
        });
    };
    KnowledgeDatabase.prototype.load = function (id, callback) {
        var _this = this;
        this.database.load(id, function (err, data) {
            if (!err) {
                var node = null;
                if (data) {
                    node = _this.createNodeFromObject(data);
                }
                callback(null, node);
            }
            else {
                callback(err);
            }
        });
    };
    KnowledgeDatabase.prototype.loadList = function (ids, callback) {
        var _this = this;
        this.database.loadList(ids, function (err, data) {
            if (!err) {
                var nodes = [];
                if (data) {
                    var resultsMap_1 = {};
                    data.forEach(function (node) {
                        if (node) {
                            resultsMap_1[node._id] =
                                _this.createNodeFromObject(node);
                        }
                    });
                    nodes = ids.map(function (id) { return resultsMap_1[id] || null; });
                }
                callback(null, nodes);
            }
            else {
                callback(err);
            }
        });
    };
    KnowledgeDatabase.prototype.loadRoot = function (callback) {
        var _this = this;
        this.database.loadNodeOfType('root', function (err, data) {
            log.iferr(err, 'Database#loadNodeOfType');
            var node;
            if (!err && data) {
                node = _this.createNodeFromObject(data);
            }
            callback(err, node);
        });
    };
    KnowledgeDatabase.prototype.save = function (node, callback) {
        node.setUpdated();
        this.database.save(node, function (err) {
            if (callback) {
                callback(err);
            }
        });
    };
    KnowledgeDatabase.prototype.remove = function (idOrNode, callback) {
        var id = this.toId(idOrNode);
        this.database.remove(id, callback);
    };
    KnowledgeDatabase.prototype.createNodeFromObject = function (object) {
        var nodeType = object.type || 'node';
        var NodeClass = Node_1.default.findNodeClass(nodeType, this.kbName);
        var node = new NodeClass(null, null, object);
        this.adoptNodeAsOurOwn(node);
        return node;
    };
    KnowledgeDatabase.prototype.createNode = function (nodeTypeOrClass, data) {
        var NodeClass;
        var nodeType;
        if (typeof nodeTypeOrClass === 'function') {
            NodeClass = nodeTypeOrClass;
        }
        else {
            nodeType = nodeTypeOrClass;
            NodeClass = Node_1.default.findNodeClass(nodeType, this.kbName);
            if (nodeType === 'root') {
                throw new Error('can not create additional root nodes');
            }
        }
        var node;
        if (nodeType && NodeClass === Node_1.default) {
            node = new NodeClass(nodeType, data);
        }
        else {
            node = new NodeClass(data);
        }
        this.adoptNodeAsOurOwn(node);
        return node;
    };
    KnowledgeDatabase.prototype.adoptNodeAsOurOwn = function (node) {
        node.setKb(this);
    };
    KnowledgeDatabase.prototype.getDirectory = function () {
        return this.dbDirectory;
    };
    KnowledgeDatabase.prototype.toId = function (idOrNode) {
        if (typeof idOrNode === 'object') {
            idOrNode = idOrNode._id;
        }
        return idOrNode;
    };
    KnowledgeDatabase.prototype._setupDirectory = function (callback) {
        fs.ensureDir(this.dbDirectory, callback);
    };
    return KnowledgeDatabase;
}());
exports.default = KnowledgeDatabase;

},{"./Database":3,"./Node":9,"./log":13,"fs-extra":undefined,"path":undefined}],7:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var request = require("@jibo/request");
var decorators_1 = require("./decorators");
var Model_1 = require("./Model");
var log_1 = require("./log");
var log = log_1.default.createChild('LoopModel');
;
var LoopModel = (function (_super) {
    __extends(LoopModel, _super);
    function LoopModel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LoopModel.prototype.loadLoop = function (callback) {
        var _this = this;
        this.loadLoopAll(function (err, loop) {
            if (!err) {
                callback(null, _this._notDeclinedOrRemoved(loop));
            }
            else {
                callback(err);
            }
        });
    };
    LoopModel.prototype.loadLoopInvited = function (callback) {
        this.loadLoop(callback);
    };
    LoopModel.prototype.loadLoopActive = function (callback) {
        this.loadLoop(callback);
    };
    LoopModel.prototype.loadLoopAll = function (callback) {
        var _this = this;
        this.loadRoot(null, function (err, rootNode) {
            log.iferr(err, 'loadRoot');
            if (!err && rootNode) {
                var userIds = rootNode.getEdges('user');
                _this.load(userIds, function (err, nodes) {
                    log.iferr(err, 'load');
                    callback(err, nodes);
                });
            }
            else {
                callback(err);
            }
        });
    };
    LoopModel.prototype.getUserNodeById = function (id, callback) {
        this.loadLoop(function (err, loop) {
            if (err) {
                return callback(err);
            }
            var user = loop.find(function (node) {
                return node.id === id;
            });
            if (user) {
                callback(null, user);
            }
            else {
                callback('Could not find user node for loop id ' + id);
            }
        });
    };
    LoopModel.prototype.getWrittenNameById = function (id, callback) {
        this.getUserNodeById(id, function (error, user) {
            if (error) {
                return callback(error);
            }
            callback(null, user.getWrittenName());
        });
    };
    LoopModel.prototype.getSpokenNameById = function (id, callback) {
        this.getUserNodeById(id, function (error, user) {
            if (error) {
                return callback(error);
            }
            callback(null, user.toString());
        });
    };
    LoopModel.prototype.fetchLoop = function () {
        return this._onlyAccepted(this.fetchLoopAll());
    };
    LoopModel.prototype.fetchLoopInvited = function () {
        return this.fetchLoop();
    };
    LoopModel.prototype.fetchLoopActive = function () {
        return this.fetchLoop();
    };
    LoopModel.prototype.fetchLoopAll = function () {
        return this.fetch(this.fetchRoot().getEdges('user'));
    };
    LoopModel.prototype.setPhoneticName = function (idOrNode, phoneticName, callback) {
        var memberId = (typeof idOrNode === 'object') ? idOrNode._id : idOrNode;
        var url = this.httpUrl + "/v1/loop/updatePhoneticName";
        this.loadRoot(function (err, root) {
            log.iferr(err, 'loadRoot');
            var params = {
                loopId: root.data.id,
                id: memberId,
                phoneticName: phoneticName
            };
            request.post(url, { json: params }, function (err, res, body) {
                log.iferr(err, 'request.post', url);
                if (callback) {
                    callback(err);
                }
            });
        });
    };
    LoopModel.prototype.setEnrollmentFace = function (idOrNode, face, callback) {
        var memberId = (typeof idOrNode === 'object') ? idOrNode._id : idOrNode;
        var params = {
            id: memberId,
            face: !!face
        };
        this.setEnrollment(params, callback);
    };
    LoopModel.prototype.setEnrollmentVoice = function (idOrNode, voice, callback) {
        var memberId = (typeof idOrNode === 'object') ? idOrNode._id : idOrNode;
        var params = {
            id: memberId,
            voice: !!voice
        };
        this.setEnrollment(params, callback);
    };
    LoopModel.prototype.setEnrollment = function (params, callback) {
        var _this = this;
        var url = this.httpUrl + "/v1/loop/enrollment";
        this.loadRoot(function (err, root) {
            log.iferr(err, 'loadRoot');
            params.loopId = root.data.id;
            request.post(url, { json: params }, function (err, res, body) {
                err = _this._checkStatusCode(err, res, body);
                log.iferr(err, 'request.post', url);
                if (callback) {
                    callback(err);
                }
            });
        });
    };
    LoopModel.prototype.suspend = function (callback) {
        var _this = this;
        var url = this.httpUrl + "/v1/loop/suspend";
        this.loadRoot(function (err, root) {
            log.iferr(err, 'loadRoot');
            var params = {
                loopId: root.data.id
            };
            request.post(url, { json: params }, function (err, res, body) {
                err = _this._checkStatusCode(err, res, body);
                log.iferr(err, 'request.post', url);
                if (callback) {
                    callback(err);
                }
            });
        });
    };
    LoopModel.prototype._notDeclinedOrRemoved = function (loop) {
        return loop.filter(function (user) { return user.data.status !== 'declined' && user.data.status !== 'removed'; });
    };
    LoopModel.prototype._onlyAccepted = function (loop) {
        return loop.filter(function (user) { return user.data.status === 'accepted'; });
    };
    LoopModel.prototype._checkStatusCode = function (err, res, message) {
        if (message === void 0) { message = null; }
        if (!err) {
            if (res.statusCode < 200 || res.statusCode > 299) {
                var description = 'HTTP Error Code ' + res.statusCode;
                if (message) {
                    description += message;
                }
                err = new Error(description);
            }
        }
        return err;
    };
    return LoopModel;
}(Model_1.default));
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "loadLoop", null);
__decorate([
    decorators_1.promisify,
    decorators_1.deprecate('loadLoopInvited is no longer useful, use loadLoop() instead')
], LoopModel.prototype, "loadLoopInvited", null);
__decorate([
    decorators_1.promisify,
    decorators_1.deprecate('loadLoopActive is no longer useful, use loadLoop() instead')
], LoopModel.prototype, "loadLoopActive", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "loadLoopAll", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "getUserNodeById", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "getWrittenNameById", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "getSpokenNameById", null);
__decorate([
    decorators_1.deprecate('no longer useful, use fetchLoop() instead')
], LoopModel.prototype, "fetchLoopInvited", null);
__decorate([
    decorators_1.deprecate('no longer useful, use fetchLoop() instead')
], LoopModel.prototype, "fetchLoopActive", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "setPhoneticName", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "setEnrollmentFace", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "setEnrollmentVoice", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "setEnrollment", null);
__decorate([
    decorators_1.promisify
], LoopModel.prototype, "suspend", null);
exports.default = LoopModel;

},{"./Model":8,"./decorators":12,"./log":13,"@jibo/request":undefined}],8:[function(require,module,exports){
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var async = require("async");
var decorators_1 = require("./decorators");
var Cache_1 = require("./Cache");
var DatabaseManager_1 = require("./DatabaseManager");
var WebClient_1 = require("./WebClient");
var log_1 = require("./log");
var log = log_1.default.createChild('Model');
var Model = (function () {
    function Model(kbNames, httpUrl) {
        if (httpUrl === void 0) { httpUrl = null; }
        var _this = this;
        this.kbNames = this._toArray(kbNames);
        this.httpUrl = httpUrl;
        this.pool = [];
        if (this.httpUrl) {
            this.kbNames.forEach(function (kbName) {
                var knowledgeDatabase = new WebClient_1.default(kbName, _this.httpUrl);
                _this.pool.push(knowledgeDatabase);
            });
        }
    }
    Model.registerModelClass = function (kbNames, classConstructor) {
        if (!Array.isArray(kbNames)) {
            kbNames = [kbNames];
        }
        var key = kbNames.join(';');
        var registry = this.modelClassRegistry;
        if (registry[key]) {
            log.warn('registerModelClass: overwriting previous class registered for ', key);
        }
        registry[key] = classConstructor;
    };
    Model.findModelClass = function (kbNames) {
        if (!Array.isArray(kbNames)) {
            kbNames = [kbNames];
        }
        var key = kbNames.join(';');
        var ModelClass = Model;
        var registry = this.modelClassRegistry;
        if (registry[key]) {
            ModelClass = registry[key];
        }
        return ModelClass;
    };
    Model.prototype.init = function (callback) {
        var _this = this;
        if (this.httpUrl) {
            setImmediate(callback);
        }
        else {
            var tasks_1 = [];
            this.kbNames.forEach(function (kbName) {
                tasks_1.push(function (next) {
                    DatabaseManager_1.default.get(kbName, next);
                });
            });
            async.series(tasks_1, function (err, results) {
                log.iferr(err, 'KnowledgeDatabase#init');
                if (!err) {
                    _this.pool = results;
                }
                callback(err);
            });
        }
    };
    Model.prototype.createModel = function (kbNames, httpUrl) {
        if (httpUrl === void 0) { httpUrl = null; }
        kbNames = this._toArray(kbNames);
        httpUrl = httpUrl || this.httpUrl;
        var defaultBaseName = this.kbNames[0];
        var newKBNames = [];
        kbNames.forEach(function (kbName) {
            if (!kbName.startsWith('/')) {
                kbName = '/' + defaultBaseName + '/' + kbName;
            }
            newKBNames.push(kbName);
        });
        var ModelClass = Model.findModelClass(newKBNames);
        return new ModelClass(newKBNames, httpUrl);
    };
    Model.prototype.createNode = function (nodeType, data) {
        var node = this.pool[0].createNode(nodeType, data);
        if (this.cache) {
            this.cache.add(node);
        }
        return node;
    };
    Model.prototype.load = function (ids, callback) {
        if (!Array.isArray(ids)) {
            this._loadOne(ids, callback);
        }
        else {
            this._loadArray(ids, callback);
        }
    };
    Model.prototype.loadList = function (ids, callback) {
        throw new Error('kb.Model.loadList is an unsupported method, use kb.Model.load');
    };
    ;
    Model.prototype.fetchList = function (ids, quietly) {
        if (quietly === void 0) { quietly = false; }
        throw new Error('kb.Model.fetchList is an unsupported method, use kb.Model.fetch');
    };
    ;
    Model.prototype.loadRoot = function (kbName, callback) {
        var _this = this;
        if (typeof kbName === 'function' && !callback) {
            callback = kbName;
            kbName = null;
        }
        var knowledgeDatabase;
        if (!kbName) {
            knowledgeDatabase = this.pool[0];
        }
        else {
            knowledgeDatabase = this._getKnowledgeDatabase(kbName);
        }
        if (!knowledgeDatabase) {
            setImmediate(function () {
                var err = new Error('loadRoot kb slice name ' + kbName + ' not found');
                callback(err);
            });
        }
        else {
            if (this.cache) {
                var rootNode_1 = this.fetchRoot(kbName, true);
                if (rootNode_1) {
                    setImmediate(function () {
                        callback(null, rootNode_1);
                    });
                }
                else {
                    knowledgeDatabase.loadRoot(function (err, loadedRootNode) {
                        if (!err && loadedRootNode) {
                            _this.roots[kbName] = loadedRootNode;
                        }
                        callback(err, loadedRootNode);
                    });
                }
            }
            else {
                knowledgeDatabase.loadRoot(callback);
            }
        }
    };
    Model.prototype.loadLayers = function (node, layers, callback) {
        var action = function (eachNode, callback) { return callback(); };
        this.visitLayers(node, layers, action, callback);
    };
    Model.prototype.saveLayers = function (node, layers, callback) {
        var action = function (eachNode, callback) { return eachNode.save(callback); };
        this.visitLayers(node, layers, action, callback);
    };
    Model.prototype.visitLayers = function (node, layers, action, callback) {
        var _this = this;
        layers = this._toArray(layers);
        var visited = {};
        visited[node._id] = true;
        var visitNodes = function (idList, callback) {
            var tasks = [];
            var queue = [];
            idList.forEach(function (id) {
                if (!(id in visited)) {
                    visited[id] = true;
                    tasks.push(function (next) {
                        _this.load(id, function (err, eachNode) {
                            if (!err && eachNode) {
                                action(eachNode, function (err) {
                                    eachNode.getEdges(layers).forEach(function (everyId) {
                                        if (!(everyId in visited)) {
                                            queue.push(everyId);
                                        }
                                    });
                                    next(err);
                                });
                            }
                            else {
                                next(err);
                            }
                        });
                    });
                }
            });
            async.series(tasks, function (err) {
                if (!err && queue.length) {
                    visitNodes(queue, callback);
                }
                else {
                    callback(err);
                }
            });
        };
        action(node, function (err) {
            log.iferr(err, 'action on initial node failed');
            if (!err) {
                visitNodes(node.getEdges(layers), callback);
            }
            else {
                callback(err);
            }
        });
    };
    Model.prototype.begin = function () {
        var ourConstructor = this.constructor;
        var model = new ourConstructor(this.kbNames, this.httpUrl);
        model.enableCache();
        return model;
    };
    Model.prototype.fetch = function (ids, quietly) {
        var _this = this;
        if (quietly === void 0) { quietly = false; }
        this._assertCache();
        if (!Array.isArray(ids)) {
            return this.cache.fetch(ids, quietly);
        }
        else {
            var list_1 = [];
            var notFoundIds_1 = [];
            ids.forEach(function (id) {
                var node = _this.cache.fetch(id, true);
                if (node || node === null) {
                    list_1.push(node);
                }
                else {
                    list_1.push(undefined);
                    notFoundIds_1.push(id);
                }
            });
            if (notFoundIds_1.length && !quietly) {
                log.warn('fetchList ids were not in cache, skipped', notFoundIds_1.join(', '));
            }
            return list_1;
        }
    };
    Model.prototype.fetchRoot = function (kbName, quietly) {
        if (quietly === void 0) { quietly = false; }
        this._assertCache();
        if (!kbName) {
            kbName = this.pool[0].kbName;
        }
        var root = null;
        if (this.roots[kbName]) {
            root = this.roots[kbName];
        }
        else {
            if (!quietly) {
                log.warn('fetchRoot for name', kbName, 'root not in cache');
            }
        }
        return root;
    };
    Model.prototype.enableCache = function () {
        this.cache = new Cache_1.default();
        this.roots = {};
    };
    Model.prototype._load = function (id, callback) {
        var tasks = [];
        this.pool.forEach(function (knowledgeDatabase) {
            tasks.push(function (node, next) {
                if (!next) {
                    next = node;
                    node = null;
                }
                if (!node) {
                    knowledgeDatabase.load(id, function (err, node) {
                        next(err, node);
                    });
                }
                else {
                    next(null, node);
                }
            });
        });
        async.waterfall(tasks, callback);
    };
    Model.prototype._loadOne = function (id, callback) {
        var _this = this;
        if (this.cache) {
            this.cache.interceptLoad(id, callback, function (missingId, cacheCallback) {
                _this._load(missingId, cacheCallback);
            });
        }
        else {
            this._load(id, callback);
        }
    };
    Model.prototype._loadArray = function (ids, callback) {
        var _this = this;
        var nodes = new Map();
        var remainingIDs = ids.slice(0);
        if (this.cache) {
            var fakeLoad_1 = function (id, callback) {
                return setImmediate(function () { return callback(null, null); });
            };
            ids.forEach(function (id) {
                var addNode = function (err, node) {
                    if (node) {
                        nodes[id] = node;
                        remainingIDs.splice(remainingIDs.indexOf(id), 1);
                    }
                };
                _this.cache.interceptLoad(id, addNode, fakeLoad_1);
            });
        }
        function getNodes(slice, callback) {
            slice.loadList(remainingIDs, callback);
        }
        async.map(this.pool, getNodes, function (err, allSlicesResults) {
            allSlicesResults.forEach(function (nodesForSlice) {
                nodesForSlice.forEach(function (node) {
                    if (node && !nodes[node._id]) {
                        nodes[node._id] = node;
                        if (_this.cache) {
                            _this.cache.add(node);
                        }
                    }
                });
            });
            callback(null, ids.map(function (id) { return nodes[id] || null; }));
        });
    };
    Model.prototype._getKnowledgeDatabase = function (kbName) {
        var knowledgeDatabase;
        for (var n = 0; n < this.pool.length; n++) {
            if (kbName === this.pool[n].kbName) {
                knowledgeDatabase = this.pool[n];
                break;
            }
        }
        return knowledgeDatabase;
    };
    Model.prototype._assertCache = function () {
        if (!this.cache) {
            throw new Error('fetch method called on a Model without a cache');
        }
    };
    Model.prototype._toArray = function (items) {
        if (!Array.isArray(items)) {
            return [items];
        }
        else {
            return items;
        }
    };
    return Model;
}());
Model.modelClassRegistry = {};
__decorate([
    decorators_1.promisify
], Model.prototype, "load", null);
__decorate([
    decorators_1.promisify
], Model.prototype, "loadRoot", null);
__decorate([
    decorators_1.promisify
], Model.prototype, "loadLayers", null);
__decorate([
    decorators_1.promisify
], Model.prototype, "saveLayers", null);
__decorate([
    decorators_1.promisify_4
], Model.prototype, "visitLayers", null);
exports.default = Model;

},{"./Cache":2,"./DatabaseManager":4,"./WebClient":11,"./decorators":12,"./log":13,"async":undefined}],9:[function(require,module,exports){
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash/lodash.min");
var uuid = require("uuid");
var async = require("async");
var decorators_1 = require("./decorators");
var Asset_1 = require("./Asset");
var log_1 = require("./log");
var log = log_1.default.createChild('Node');
var Node = (function () {
    function Node(nodeType, data, cloneFrom) {
        if (cloneFrom) {
            _.extend(this, cloneFrom);
        }
        this._id = this._id || uuid.v4();
        this.data = data || this.data || {};
        this.type = nodeType || this.type || 'node';
        var timestamp = Number(new Date());
        this.created = this.created || timestamp;
        this.updated = this.updated || timestamp;
    }
    Node.registerNodeClass = function (nodeType, classConstructor, kbName) {
        if (kbName === void 0) { kbName = '*'; }
        var registry = Node.nodeClassRegistry;
        if (registry[kbName] && registry[kbName][nodeType]) {
            log.warn('registerNodeClass: overwriting previous class registered for type', nodeType);
        }
        if (!registry[kbName]) {
            registry[kbName] = {};
        }
        registry[kbName][nodeType] = classConstructor;
    };
    Node.findNodeClass = function (nodeType, kbName) {
        var NodeClass = Node;
        var registry = Node.nodeClassRegistry;
        if (registry[kbName] && registry[kbName][nodeType]) {
            NodeClass = registry[kbName][nodeType];
        }
        else if (registry['*'] && registry['*'][nodeType]) {
            NodeClass = registry['*'][nodeType];
        }
        return NodeClass;
    };
    Node.prototype.save = function (callback) {
        this.getKb().save(this, callback);
    };
    Node.prototype.remove = function (callback) {
        this.getKb().remove(this, callback);
    };
    Node.prototype.addEdges = function (idsOrNodes, layer) {
        var _this = this;
        idsOrNodes = this._toArray(idsOrNodes);
        this.edges = this.edges || {};
        idsOrNodes.forEach(function (idOrNode) {
            var _a = _this._resolveIdAndLayer(idOrNode, layer, 'Node#addEdges()'), id = _a.id, l = _a.l;
            _this.edges[l] = _this.edges[l] || [];
            _this.edges[l].push(id);
        });
    };
    Node.prototype.removeEdges = function (idsOrNodes, layer) {
        var _this = this;
        idsOrNodes = this._toArray(idsOrNodes);
        if (this.edges) {
            idsOrNodes.forEach(function (idOrNode) {
                var _a = _this._resolveIdAndLayer(idOrNode, layer, 'Node#removeEdges()'), id = _a.id, l = _a.l;
                if (_this.edges[l]) {
                    _.pull(_this.edges[l], id);
                }
            });
        }
    };
    Node.prototype.clearEdges = function (layers) {
        var _this = this;
        layers = this._toArray(layers);
        if (this.edges) {
            layers.forEach(function (layer) {
                if (_this.edges[layer]) {
                    _this.edges[layer] = [];
                }
            });
        }
    };
    Node.prototype.getEdges = function (layers) {
        var _this = this;
        layers = this._toArray(layers);
        var edges = [];
        layers.forEach(function (layer) {
            var theseEdges = _.get(_this, 'edges.' + layer, []);
            theseEdges.forEach(function (thisEdge) {
                if (edges.indexOf(thisEdge) < 0) {
                    edges.push(thisEdge);
                }
            });
        });
        return edges;
    };
    Node.prototype.getLayers = function () {
        if (this.edges) {
            return Object.keys(this.edges);
        }
        else {
            return [];
        }
    };
    Node.prototype.createAsset = function (subtype, ext) {
        var asset = new Asset_1.default(null, subtype, ext);
        if (this.getKb) {
            var rootDir = this.getKb().getDirectory();
            asset.setRootDir(rootDir);
        }
        else {
            log.warn('createAsset: creating an asset from a node without a kb, rootDir not set');
        }
        this.addAssets(asset);
        return asset;
    };
    Node.prototype.addAssets = function (assets, subtype) {
        var _this = this;
        assets = this._toArray(assets);
        this.assets = this.assets || {};
        assets.forEach(function (asset) {
            var st = subtype || asset.subtype;
            _this.assets[st] = _this.assets[st] || [];
            _this.assets[st].push(asset.filename());
        });
    };
    Node.prototype.getAssets = function (subtype) {
        if (subtype === void 0) { subtype = 'asset'; }
        var assets = [];
        if (this.assets && this.assets[subtype]) {
            var assetFilenames = this.assets[subtype];
            var rootDir_1 = this.getKb().getDirectory();
            assetFilenames.forEach(function (filename) {
                var asset = new Asset_1.default(filename);
                asset.setRootDir(rootDir_1);
                assets.push(asset);
            });
        }
        return assets;
    };
    Node.prototype.getAllAssets = function () {
        var _this = this;
        var assets = [];
        var subtypes = this.getAssetSubtypes();
        subtypes.forEach(function (subtype) {
            assets = assets.concat(_this.getAssets(subtype));
        });
        return assets;
    };
    Node.prototype.getAssetSubtypes = function () {
        var subtypes = _.keys(this.assets);
        return subtypes;
    };
    Node.prototype.removeAsset = function (asset, callback) {
        var st = asset.subtype;
        var assetIndex = this.assets[st].indexOf(asset.filename());
        this.assets[st].splice(assetIndex, 1);
        asset.remove(callback);
    };
    Node.prototype.removeAllAssets = function (callback) {
        var _this = this;
        var subtypes = this.getAssetSubtypes();
        var tasks = [];
        subtypes.forEach(function (subtype) {
            var assets = _this.getAssets(subtype);
            assets.forEach(function (asset) {
                tasks.push(function (next) {
                    _this.removeAsset(asset, next);
                });
            });
        });
        async.series(tasks, callback);
    };
    Node.prototype.setKb = function (knowledgeDatabase) {
        this.getKb = function () { return knowledgeDatabase; };
        return this;
    };
    Node.prototype.setUpdated = function (timestamp) {
        this.updated = timestamp || Number(new Date());
    };
    Node.prototype._resolveIdAndLayer = function (idOrNode, layer, methodName) {
        var id;
        var l;
        if (typeof idOrNode === 'string') {
            id = idOrNode;
            if (!layer) {
                throw Error(methodName + ": must specify layer when supplying ids");
            }
            l = layer;
        }
        else {
            var node = idOrNode;
            id = node._id;
            l = layer || node.type;
        }
        return { id: id, l: l };
    };
    Node.prototype._toArray = function (items) {
        if (!Array.isArray(items)) {
            return [items];
        }
        else {
            return items;
        }
    };
    return Node;
}());
Node.nodeClassRegistry = {};
__decorate([
    decorators_1.promisify
], Node.prototype, "save", null);
__decorate([
    decorators_1.promisify
], Node.prototype, "remove", null);
__decorate([
    decorators_1.promisify
], Node.prototype, "removeAsset", null);
__decorate([
    decorators_1.promisify
], Node.prototype, "removeAllAssets", null);
exports.default = Node;

},{"./Asset":1,"./decorators":12,"./log":13,"async":undefined,"lodash/lodash.min":undefined,"uuid":undefined}],10:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Node_1 = require("./Node");
var UserNode = (function (_super) {
    __extends(UserNode, _super);
    function UserNode() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(UserNode.prototype, "id", {
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserNode.prototype, "firstName", {
        get: function () {
            return this.data.firstName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserNode.prototype, "lastName", {
        get: function () {
            return this.data.lastName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserNode.prototype, "nickName", {
        get: function () {
            return this.data.nickName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserNode.prototype, "gender", {
        get: function () {
            return this.data.gender;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserNode.prototype, "isJibo", {
        get: function () {
            return !this.data.firstName;
        },
        enumerable: true,
        configurable: true
    });
    UserNode.prototype.getWrittenName = function () {
        return this.data.nickName || this.data.firstName || 'Jibo';
    };
    UserNode.prototype.toString = function () {
        return this.data.phoneticName || this.getWrittenName();
    };
    UserNode.prototype.getInitials = function () {
        if (this.isJibo) {
            return 'J';
        }
        if (!this.data.firstName || !this.data.lastName) {
            return '';
        }
        return this.data.firstName.charAt(0).toUpperCase() +
            this.data.lastName.charAt(0).toUpperCase();
    };
    UserNode.prototype.toLog = function () {
        return "[UserNode id:" + this._id + "]";
    };
    return UserNode;
}(Node_1.default));
exports.default = UserNode;

},{"./Node":9}],11:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var async = require("async");
var request = require("@jibo/request");
var querystring = require("querystring");
var KnowledgeDatabase_1 = require("./KnowledgeDatabase");
var log_1 = require("./log");
var log = log_1.default.createChild('WebClient');
var WebClient = (function (_super) {
    __extends(WebClient, _super);
    function WebClient(kbName, httpUrl) {
        var _this = _super.call(this, kbName) || this;
        _this.httpUrl = httpUrl.replace(/\/$/, '');
        return _this;
    }
    WebClient.prototype.init = function (callback) {
        setImmediate(callback);
    };
    WebClient.prototype.load = function (id, callback) {
        var _this = this;
        var url = this._makeUrl("/node/load/" + id);
        request.get(url, { json: true }, function (err, res, data) {
            err = _this._checkStatusCode(err, res, data);
            log.iferr(err, 'request.get', url);
            if (!err) {
                var node = null;
                if (data) {
                    node = _this.createNodeFromObject(data);
                }
                callback(null, node);
            }
            else {
                callback(err);
            }
        });
    };
    WebClient.prototype.loadList = function (ids, callback) {
        var _this = this;
        var url = this._makeUrl("/node/load");
        request.post(url, { json: true, body: ids }, function (err, res, data) {
            err = _this._checkStatusCode(err, res, data);
            log.iferr(err, 'request.post', url);
            if (!err) {
                var nodes = null;
                if (data) {
                    nodes = data.map(function (node) {
                        return node ? _this.createNodeFromObject(node) : null;
                    });
                }
                callback(null, nodes);
            }
            else {
                log.warn('Deprecation warning: jibo-kb will soon drop support for legacy SSM without /node/load POST endpoint.');
                async.map(ids, _this.load.bind(_this), callback);
            }
        });
    };
    WebClient.prototype.loadRoot = function (callback) {
        var _this = this;
        var url = this._makeUrl("/node/loadRoot");
        request.get(url, { json: true }, function (err, res, data) {
            err = _this._checkStatusCode(err, res, data);
            log.iferr(err, 'request.get', url);
            var node;
            if (!err && data) {
                node = _this.createNodeFromObject(data);
            }
            callback(err, node);
        });
    };
    WebClient.prototype.save = function (node, callback) {
        var _this = this;
        var url = this._makeUrl("/node/save");
        request.post(url, { json: node }, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.post', url);
            if (callback) {
                callback(err);
            }
        });
    };
    WebClient.prototype.remove = function (idOrNode, callback) {
        var _this = this;
        var id = this.toId(idOrNode);
        var url = this._makeUrl("/node/remove/" + id);
        request.del(url, function (err, res, body) {
            err = _this._checkStatusCode(err, res, body);
            log.iferr(err, 'request.delete', url);
            callback(err);
        });
    };
    WebClient.prototype.getDirectory = function () {
        return this._makeUrl();
    };
    WebClient.prototype._makeUrl = function (addPath) {
        if (addPath === void 0) { addPath = ''; }
        var escapedName = querystring.escape(this.kbName);
        return this.httpUrl + "/v1/kb/" + escapedName + addPath;
    };
    WebClient.prototype._checkStatusCode = function (err, res, message) {
        if (message === void 0) { message = null; }
        if (!err) {
            if (res.statusCode < 200 || res.statusCode > 299) {
                var description = 'HTTP Error Code ' + res.statusCode;
                if (message) {
                    description += message;
                }
                err = new Error(description);
            }
        }
        return err;
    };
    return WebClient;
}(KnowledgeDatabase_1.default));
exports.default = WebClient;

},{"./KnowledgeDatabase":6,"./log":13,"@jibo/request":undefined,"async":undefined,"querystring":undefined}],12:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function promisify(target, propertyKey, descriptor) {
    var originalMethod = descriptor.value;
    descriptor.value = function () {
        var _this = this;
        var result;
        var args = Array.from(arguments);
        var lastArg = args.slice(-1)[0];
        if (typeof lastArg === 'function') {
            result = originalMethod.apply(this, args);
        }
        else {
            result = new Promise(function (resolve, reject) {
                var callback = function (err, arg) {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(arg);
                    }
                };
                var newArgs = args.concat(callback);
                originalMethod.apply(_this, newArgs);
            });
        }
        return result;
    };
    return descriptor;
}
exports.promisify = promisify;
function promisify_4(target, propertyKey, descriptor) {
    var originalMethod = descriptor.value;
    descriptor.value = function () {
        var result;
        var args = Array.from(arguments);
        var fourthArg = args[4];
        if (typeof fourthArg === 'function') {
            result = originalMethod.apply(this, args);
        }
        else {
            var originalThis_1 = this;
            result = new Promise(function (resolve, reject) {
                var callback = function (err, arg) {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(arg);
                    }
                };
                var newArgs = args.concat(callback);
                originalMethod.apply(originalThis_1, newArgs);
            });
        }
        return result;
    };
    return descriptor;
}
exports.promisify_4 = promisify_4;
function deprecate(msg) {
    msg = msg === undefined ? '' : msg;
    function warn() {
        var stack = new Error().stack;
        stack = stack.split('\n').splice(4).join('\n');
        if (console.groupCollapsed) {
            console.groupCollapsed("%cDeprecation Warning: %c%s", "color:#614108;background:#fffbe6", "font-weight:normal;color:#614108;background:#fffbe6", msg);
            console.warn(stack);
            console.groupEnd();
        }
        else {
            console.warn("Deprecation Warning:", msg);
            console.warn(stack);
        }
    }
    function doDeprecate(target, propertyKey, descriptor) {
        var originalMethod = descriptor.value;
        descriptor.value = function () {
            warn();
            return originalMethod.apply(this, arguments);
        };
        return descriptor;
    }
    return doDeprecate;
}
exports.deprecate = deprecate;

},{}],13:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Log = require("jibo-log");
exports.default = new Log('Jibo.KB');

},{"jibo-log":undefined}],14:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./KnowledgeBase"));
var Asset_1 = require("./Asset");
exports.Asset = Asset_1.default;
var Cache_1 = require("./Cache");
exports.Cache = Cache_1.default;
var Database_1 = require("./Database");
exports.Database = Database_1.default;
var DatabaseManager_1 = require("./DatabaseManager");
exports.DatabaseManager = DatabaseManager_1.default;
var KnowledgeBase_1 = require("./KnowledgeBase");
exports.KnowledgeBase = KnowledgeBase_1.default;
var KnowledgeDatabase_1 = require("./KnowledgeDatabase");
exports.KnowledgeDatabase = KnowledgeDatabase_1.default;
var LoopModel_1 = require("./LoopModel");
exports.LoopModel = LoopModel_1.default;
var Model_1 = require("./Model");
exports.Model = Model_1.default;
var Node_1 = require("./Node");
exports.Node = Node_1.default;
var UserNode_1 = require("./UserNode");
exports.UserNode = UserNode_1.default;
var WebClient_1 = require("./WebClient");
exports.WebClient = WebClient_1.default;

},{"./Asset":1,"./Cache":2,"./Database":3,"./DatabaseManager":4,"./KnowledgeBase":5,"./KnowledgeDatabase":6,"./LoopModel":7,"./Model":8,"./Node":9,"./UserNode":10,"./WebClient":11}]},{},[14])(14)
});
