// jibo-emotion-system/src/emotion/EmotionSpace.ts
/*
 * @fileOverview
 *
 * EmotionSpace
 * ===============
 *
 * Example Usage:
 *
 * ```
 *   // The impact as a result of an event is interpreted by Jibo.
 *   // TODO create these easily interpretable influence types
 *   // TODO a la SKILL_IMPACTS in EmotionDefinition
 *   var impact = new EmotionPoint({
 *     valence: ValenceType.VERY_GOOD,
 *     arousal: ArousalType.BORING,
 *   });
 *
 *   EmotionSpace.update(impact);
 *
 *   var currentEmotion = EmotionSpace.getNearestEmotion();
 *
 *   >> <Emotion:'Contentment'/>
 *
 * ```
 */
import {EmotionTrace} from '../emotion/EmotionTrace';
import {Event, EventContainer} from 'jibo-typed-events';
import {EmotionSystem} from "../EmotionSystem";
import {Utils} from "../common/Utils";
import {Emotion} from "./Emotion";
import {EmotionPoint} from "../common/Types";
import {JIBO_BASELINE, EMOTION_DECAY_RATE} from "./EmotionDefinitions";

import parentLog from '../log';
const log = parentLog.createChild('Space');

export {}

/**
 * @description Events related to Jibo's emotional space.
 * @namespace jibo.emotion.emotionSpace.events
 * @private
 * @internal
 */
export class EmotionSpaceEvents extends EventContainer {

    /**
     * Triggered whenever the current emotion has changed.
     * Emits the new emotional state.
     * @type {jibo.emotion.Emotion}
     * @name jibo.emotion.emotionSpace.events#changeEvent
     */
    public changeEvent = new Event<Emotion>('Emotion has changed.');
    /**
     * Triggers every emotional update.
     * Emits the trace with which it was updated
     * @type {jibo.emotion.EmotionTrace}
     * @name jibo.emotion.emotionSpace.events#updated
     */
    public updated = new Event<EmotionTrace>('Emotion has changed.');
}

/**
 * @description The management and coordination of Jibo's current internal emotional state.
 * @namespace jibo.emotion.emotionSpace
 * @private
 * @internal
 */
export class EmotionSpace {

    /**
     * Events related to changes in jibo's internal emotional state.
     * @type {jibo.emotion.emotionSpace.events}
     * @name jibo.emotion.emotionSpace#events
     */
    public events = new EmotionSpaceEvents();
    /**
     * All registered emotions that the system is aware of.
     * @type {Set<jibo.emotion.Emotion>}
     * @name jibo.emotion.emotionSpace#emotions
     */
    public emotions: Set<Emotion>;
    /**
     * The current emotional position of Jibo, aka his current emotional state.
     * @type {Set<jibo.emotion.EmotionPoint>}
     * @name jibo.emotion.emotionSpace#currentEmotionPoint
     * @private
     */
    private currentEmotionPoint: EmotionPoint;
    /**
     * The last known position of Jibo's emotional state.
     * @type {Set<jibo.emotion.EmotionPoint>}
     * @name jibo.emotion.emotionSpace#lastEmotionPoint
     * @private
     */
    private lastEmotionPoint: EmotionPoint;
    /**
     * The rate by which the emotion will normalize back to a neutral state
     * after impact in any direction.
     * @type {number}
     * @name jibo.emotion.emotionSpace#_decayImpactPerHour
     * @private
     */
    private _decayImpactPerHour: number = 10;
    /**
     * The unique ID of the setInterval timer for updating the decay rate tick.
     * @type {number}
     * @name jibo.emotion.emotionSpace#_decayTimerId
     * @private
     */
    private _decayTimerId = null;

    /**
     * The type of timeout we'll be setting -- either jibo internal or sinon for tests
     * @name jibo.emotion.emotionSpace#setTimeout
     * @private
     */
    private setInterval: (callback: ()=>void, delay: number)=>any;
    /**
     * The type of timeout we'll be setting -- either jibo internal or sinon for tests
     * @name jibo.emotion.emotionSpace#setTimeout
     * @private
     */
    private clearInterval: (handle: any)=>void;
    /**
     * The most recent emotion trace that impacted this space.
     * @name jibo.emotion.emotionSpace#lastEmotionTrace
     * @private
     */
    private lastEmotionTrace: EmotionTrace;

    /** @override */
    constructor(private parent: EmotionSystem) {
        this.emotions = new Set<Emotion>();
        this.currentEmotionPoint = new EmotionPoint(JIBO_BASELINE.emotionValues.valence, JIBO_BASELINE.emotionValues.confidence);
        this.lastEmotionPoint = new EmotionPoint(JIBO_BASELINE.emotionValues.valence, JIBO_BASELINE.emotionValues.confidence);
        this.setInterval = (callback: any, delay: number): any => {
            return this.parent.jibo.timer.setInterval(callback, delay);
        };
        this.clearInterval = (timerId: any) => {
            this.parent.jibo.timer.clearInterval(timerId);
        };
    }

    /**
     * Initialize Emotional Space
     * @param {Array<jibo.emotion.Emotion>} emotions Emotions to initialize Jibo's emotional
     * space with.
     * @method jibo.emotion.emotionSpace#init
     */
    public init(emotions: Emotion[]): void {
        for (let i = 0; i < emotions.length; i++) {
            this.emotions.add(emotions[i]);
        }
        // initialize to baseline!
        this._startDecayToBaseline();
    }

    /**
     * Retrieves the emotion closest to the current state given a threshold.
     * @param {jibo.emotion.Emotion} referenceEmotion An emotion to find the neighbor of.
     * @return {jibo.emotion.Emotion} The emotional state or null if under threshold.
     * @method jibo.emotion.emotionSpace#getNearestEmotion
     */
    public getNearestEmotion(referencePoint: EmotionPoint = this.currentEmotionPoint): Emotion {
        let bestMatch: Emotion = null;
        let bestScore = Infinity;
        this.emotions.forEach((emotion: Emotion) => {
            let currentScore = Utils.getEuclideanDistance(emotion.emotionValues, referencePoint, ['valence', 'confidence']);
            if (currentScore < bestScore) {
                bestMatch = emotion;
                bestScore = currentScore;
            }
        });
        return bestMatch;
    }

    /**
     * Returns the current Emotional State
     * @return {jibo.emotion.EmotionPoint} The current emotional state.
     * @method jibo.emotion.emotionSpace#getCurrentPosition
     */
    public getCurrentPosition(): EmotionPoint {
        return this.currentEmotionPoint;
    }

    /**
     * Get the most recent emotion trace that impacted this space.
     * @return {jibo.emotion.EmotionTrace} The most recent impacting emotional trace.
     * @method jibo.emotion.emotionSpace#getLastTrace
     */
    public getLastTrace(): EmotionTrace{
        return this.lastEmotionTrace;
    }

    /**
     * Cleans up and disconnects all allocated resources.
     * @method jibo.emotion.emotionSpace#dispose
     */
    public dispose(): void {
        this.events.removeAllListeners();
        this.clearInterval(this._decayTimerId);
    }

    /**
     * Updates the current emotional state of Jibo in a P.A.D space.
     * Emits an event to the Expression System if a change in Emotion occurs
     * @param {jibo.emotion.emotionTrace} emotionTrace. The trace of the emotion events that led to this
     * @method jibo.emotion.emotionSpace#update
     */
    public update(emotionTrace: EmotionTrace): void {
        //log.debug("updating emotion", emotionTrace);
        let impact: EmotionPoint = emotionTrace.emotionalImpact;
        this.currentEmotionPoint.valence += impact.valence;
        this.currentEmotionPoint.confidence += impact.confidence;
        //add the old and new positions of the Space to the trace
        emotionTrace.startingPosition = new EmotionPoint(this.lastEmotionPoint.valence, this.lastEmotionPoint.confidence);
        emotionTrace.newPosition = new EmotionPoint(this.currentEmotionPoint.valence, this.currentEmotionPoint.confidence);
        emotionTrace.nearestEmotionAfterImpact = this.getNearestEmotion(this.currentEmotionPoint);
        //log.debug("impact:", emotionTrace);
        const lastFeltEmotion:Emotion = this.getNearestEmotion(this.lastEmotionPoint);

        // TODO: Perform some kind of state transition magic here to see
        // If we should legitimately transition states to prevent erratic
        // emotional state shifting.
        if (lastFeltEmotion.name !== emotionTrace.nearestEmotionAfterImpact.name) {
            log.info(`jibo's emotion changed to ${emotionTrace.nearestEmotionAfterImpact.name}`);
            this.events.changeEvent.emit(emotionTrace.nearestEmotionAfterImpact);
            this.parent.jibo.analytics.track("Emotion Change", {
              emotion_before: lastFeltEmotion.name,
              emotion_after: emotionTrace.nearestEmotionAfterImpact.name
            });
        }

        this.lastEmotionPoint.valence = this.currentEmotionPoint.valence;
        this.lastEmotionPoint.confidence = this.currentEmotionPoint.confidence;
        this.lastEmotionTrace = emotionTrace;
        //log.debug(`updated jibo's emotion to ${this.currentEmotionPoint.valence} ${this.currentEmotionPoint.confidence}`);
        this.events.updated.emit(emotionTrace);
    }

    /**
     * Resets the current emotion state to jibo baseline values
     * @returns {jibo.emotion.emotionTrace} Updated emotionTrace that caused neutral state
     * @method jibo.emotion.emotionSpace#_resetToNeutral
     */
    public _resetToNeutral():EmotionTrace {
        let valenceDiff = JIBO_BASELINE.emotionValues.valence - this.currentEmotionPoint.valence;
        let confidenceDiff = JIBO_BASELINE.emotionValues.confidence - this.currentEmotionPoint.confidence;
        let impact: EmotionPoint = new EmotionPoint(valenceDiff, confidenceDiff);
        const emotionTrace = new EmotionTrace(null, null, impact);
        this.update(emotionTrace);
        return emotionTrace;
    }

    /**
     * Begins the process of decaying Jibo's emotions, which will return his
     * emotions back to a neutral state over time.
     * @method jibo.emotion.emotionSpace#_startToDecayEmotions
     * @private
     */
    private _startDecayToBaseline(): void {
        if (this._decayTimerId) {
            this.clearInterval(this._decayTimerId);
        }
        // decay impact per hour, over minutes in an hour...
        let decayRate = this._decayImpactPerHour / (60);
        // TODO: determine if this deviates from wall-clock time too much
        // e.g. if this interval loop is called v slowly for some reason,
        // do we want jibo's emotional decay to "catch up?"
        this._decayTimerId = this.setInterval(() => {
            const impactUpdatePoint = new EmotionPoint();

            const baselineValence = JIBO_BASELINE.emotionValues.valence;
            const baselineConfidence = JIBO_BASELINE.emotionValues.confidence;
            // See which direction the property value is, and move towards Jibo's Baseline.
            if (this.currentEmotionPoint.valence > baselineValence) {
                if (this.currentEmotionPoint.valence - decayRate < baselineValence) {
                    impactUpdatePoint.valence = baselineValence - this.currentEmotionPoint.valence;
                }
                else {
                    impactUpdatePoint.valence = -decayRate;
                }
            }
            else if (this.currentEmotionPoint.valence < baselineValence) {
                if (this.currentEmotionPoint.valence + decayRate > baselineValence) {
                    impactUpdatePoint.valence =  baselineValence - this.currentEmotionPoint.valence;
                }
                else {
                    impactUpdatePoint.valence = decayRate;
                }
            }
            // See which direction the property value is, and move towards Jibo's Baseline.
            if (this.currentEmotionPoint.confidence > baselineConfidence) {
                if (this.currentEmotionPoint.confidence - decayRate < baselineConfidence) {
                    impactUpdatePoint.confidence =  baselineConfidence - this.currentEmotionPoint.confidence;
                }
                else {
                    impactUpdatePoint.confidence = -decayRate;
                }
            }
            else if (this.currentEmotionPoint.confidence < baselineConfidence) {
                if (this.currentEmotionPoint.confidence + decayRate > baselineConfidence) {
                    impactUpdatePoint.confidence = baselineConfidence - this.currentEmotionPoint.confidence;
                }
                else {
                    impactUpdatePoint.confidence = decayRate;
                }
            }
            const updateTrace = new EmotionTrace(null, null, impactUpdatePoint);
            this.update(updateTrace);

        }, EMOTION_DECAY_RATE * 1000); // Every decay rate interval seconds, tick.
    }
}
