// jibo-emotion-system/src/emotion/EmotionDefinitions.ts
import {EmotionName, ImpactSize, EmotionPoint} from "../common/Types";
import {Emotion} from "./Emotion";

export { }

/**
 * @internal
 */
export const JIBO_BASELINE = new Emotion(EmotionName.BASELINE, 0.45, 0.2);

/**
 * The number of seconds between emotion decays.
 * @internal
 */
export const EMOTION_DECAY_RATE = 10;

/**
 * @internal
 */
export const ImpactValues = new Map<ImpactSize, number>();
ImpactValues.set(ImpactSize.NONE, 0);
ImpactValues.set(ImpactSize.LOW_POS, 0.5);
ImpactValues.set(ImpactSize.MEDIUM_POS, 1);
ImpactValues.set(ImpactSize.HIGH_POS, 2);
ImpactValues.set(ImpactSize.LOW_NEG, -0.5);
ImpactValues.set(ImpactSize.MEDIUM_NEG, -1);
ImpactValues.set(ImpactSize.HIGH_NEG, -2);

/**
 * @internal
 */
export const AppraisalEvents = {
    IMPACT_TRIGGERED: 'IMPACT_TRIGGERED',
};

/**
 * @enum jibo.emotion.DEFAULT_EMOTIONS
 * @prop {number} [valence=0]
 * @prop {number} [confidence=0]
 * @intdocs
 * @internal
 */
export const DEFAULT_EMOTIONS = new Map<EmotionName, EmotionPoint>();
DEFAULT_EMOTIONS.set(EmotionName.PLEASED, {valence: 1, confidence: 0});
DEFAULT_EMOTIONS.set(EmotionName.SAD, {valence: -1, confidence: 0});
DEFAULT_EMOTIONS.set(EmotionName.NEUTRAL, {valence: 0, confidence: 0});
DEFAULT_EMOTIONS.set(EmotionName.CONFIDENT, {valence: 0, confidence: 1});
DEFAULT_EMOTIONS.set(EmotionName.UNCERTAIN, {valence: 0, confidence: -1});
DEFAULT_EMOTIONS.set(EmotionName.PESSIMISTIC, {valence: -1, confidence: 1});
DEFAULT_EMOTIONS.set(EmotionName.FRUSTRATED, {valence: -1, confidence: -1});
DEFAULT_EMOTIONS.set(EmotionName.HOPEFUL, {valence: -1, confidence: 1});
DEFAULT_EMOTIONS.set(EmotionName.JOYFUL, {valence: 1, confidence: 1});

/**
 * @internal
 */
export const EMOTION_TO_ANIM_CATEGORY = new Map<EmotionName, string>();
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.PLEASED, 'happy');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.SAD, 'sad');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.NEUTRAL, 'politereply');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.CONFIDENT, 'proud');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.UNCERTAIN, 'confused');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.PESSIMISTIC, 'disgusted');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.FRUSTRATED, 'frustrated');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.HOPEFUL, 'curious');
EMOTION_TO_ANIM_CATEGORY.set(EmotionName.JOYFUL, 'laughing');
