// jibo-emotion-system/src/emotion/EmotionTrace.ts
import {SensorDataWrapper} from '../sensing/sensors/Sensor';
import {AppraisalRule} from '../appraisal/rules/AppraisalRule';
import {EmotionalExpression} from '../expression/EmotionalExpression';
import {Emotion} from "./Emotion";
import {EmotionPoint} from "../common/Types";

export {}

/**
 * @description Trace of an emotional impact transition.
 * @namespace jibo.emotion.emotionTrace
 * @private
 * @internal
 */
export class EmotionTrace {

    /**
     * A list of the sensors involved in this emotional event.
     * @type {Array<SensorDataWrapper>}
     * @name jibo.emotion.emotionTrace#sensorDataWrappers
     */
    public sensorDataWrappers: SensorDataWrapper<any>[];
    /**
     * The rule whicsh applies in this emotional event.
     * @type {AppraisalRule}
     * @name jibo.emotion.emotionTrace#appraisalRule
     */
    public appraisalRule: AppraisalRule;
    /**
     * The impact on Jibo's emotional state of this emotional event.
     * @type {jibo.emotion.EmotionPoint}
     * @name jibo.emotion.emotionTrace#emotionalImpact
     */
    public emotionalImpact: EmotionPoint;
    /**
     * The starting position of Jibo's emotional state before this emotional event.
     * @type {jibo.emotion.Emotion}
     * @name jibo.emotion.emotionTrace#startingPosition
     */
    public startingPosition: EmotionPoint;
    /**
     * The ending position of Jibo's emotional state after this emotional event.
     * @type {jibo.emotion.Emotion}
     * @name jibo.emotion.emotionTrace#newPosition
     */
    public newPosition: EmotionPoint;
    /**
     * The nearest categorized emotion after this emotional event.
     * @type {jibo.emotion.Emotion}
     * @name jibo.emotion.emotionTrace#nearestEmotionAfterImpact
     */
    public nearestEmotionAfterImpact: Emotion;
    /**
     * The nearest categorized emotion before this emotional event.
     * @type {jibo.emotion.Emotion}
     * @name jibo.emotion.emotionTrace#emotionBeforeImpact
     */
    public emotionBeforeImpact: Emotion;
    /**
     * The expression of the nearest emotion after impact.
     * @type {EmotionalExpression}
     * @name jibo.emotion.emotionTrace#expression
     */
    public expression: EmotionalExpression;

    constructor(sensorDataWrappers: SensorDataWrapper<any>[], appraisalRule: AppraisalRule, emotionalImpact: EmotionPoint) {
        this.sensorDataWrappers = sensorDataWrappers;
        this.appraisalRule = appraisalRule;
        this.emotionalImpact = emotionalImpact;
    }
}
