// jibo-background/lib/debug/services/NotificationsService.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _jiboClientFramework = require('jibo-client-framework');

var _events = require('events');

class NotificationsService extends _events.EventEmitter {
    constructor() {
        super();
    }

    init(service, cb) {
        var _this = this;

        console.log('notifications-service');
        this.httpClient = new _jiboClientFramework.HTTPClient('127.0.0.1', service.port);
        this.wsClient = new _jiboClientFramework.WSClient('ws://' + '127.0.0.1' + ':' + service.port);

        this.wsClient.on('open', function () {
            cb();
        });

        this.wsClient.on('message', function (message) {
            _this.emit(message.eventName, message.event);
        });
    }

    /*
     * Add a basic 'message' notification to the notifications queue for this skill
     */
    create(notification, cb) {
        this.httpClient.sendRequest('POST', '/notifications', JSON.stringify(notification), cb);
    }

    /*
     * Update a notification. Any field provide overrides the existing one, other wise
     * fields are left unchanged.
     */
    update(id, notificationChanges, cb) {
        this.httpClient.sendRequest('PUT', `/notifications/${ id }`, JSON.stringify(notificationChanges), cb);
    }

    /*
     * Get all notification
     */
    getAll(cb) {
        this.httpClient.sendRequest('GET', '/notifications', '{}', cb);
    }

    /*
     * Delete a notification
     */
    delete(id, cb) {
        this.httpClient.sendRequest('DELETE', `/notifications/${ id }`, '{}', cb);
    }

    /*
     * Delete all notifications
     */
    deleteAll(cb) {
        this.httpClient.sendRequest('DELETE', `/notifications`, '{}', cb);
    }
}

exports.default = NotificationsService;
