// jibo-state-machine/lib/jibo-state-machine.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboStateMachine = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Transition_1 = require("./Transition");
var TimeoutTransition_1 = require("./TimeoutTransition");
var log_1 = require("./log");
var log = log_1.default.createChild("State");
var State = (function () {
    function State(sm, name, transitions) {
        this.sm = sm;
        this.name = name;
        this.transitions = [];
        this._isCurrent = false;
        this._entryCounter = 0;
        this._exitCounter = 0;
        this._isExiting = false;
        this._internals = [];
        this._events = new Map();
        this._typedEvents = new Map();
        sm._addState(this);
        if (transitions) {
            this.addTransition(transitions);
        }
    }
    State.prototype.addTransition = function (transition) {
        var _this = this;
        if (Array.isArray(transition)) {
            transition.forEach(function (trans) {
                _this._addTransition(trans);
            });
        }
        else {
            this._addTransition(transition);
        }
        return this;
    };
    State.prototype._addTransition = function (transition) {
        if (transition.getSourceState()) {
            this.sm._rejectCurrentPromise(new Error("This transition has already been assigned to a source state"));
        }
        transition._setSourceState(this);
        this.transitions.push(transition);
    };
    State.prototype.addInternalTransition = function (name, destState) {
        var exists = this._internals.some(function (i) { return (i.name === name && i.destState === destState); });
        if (exists) {
            throw new Error("State already has internal transition of name '" + name + "' to state '" + destState.name + "'");
        }
        var internal = {
            name: name,
            destState: destState,
            transition: new Transition_1.Transition(name, destState)
        };
        this._internals.push(internal);
        this.addTransition(internal.transition);
        return this;
    };
    State.prototype.addEventTransition = function (event, destState) {
        var _this = this;
        if (this._events.has(event)) {
            throw new Error("Duplicate event transition. State: '" + this.name + "', event: '" + event + "'");
        }
        var trans = new Transition_1.Transition('Event: ' + event, destState);
        var handler = function (result) {
            if (_this._isCurrent) {
                trans.trigger(result);
            }
        };
        this._events.set(event, { event: event, handler: handler });
        this.addTransition(trans);
        return this;
    };
    State.prototype.addTypedEventTransition = function (event, destState) {
        var _this = this;
        var trans = new Transition_1.Transition('TEvent: ' + event.name, destState);
        var handler = function (result) {
            if (_this._isCurrent) {
                trans.trigger(result);
            }
        };
        this._typedEvents.set(event, { event: event, handler: handler });
        this.addTransition(trans);
        return this;
    };
    State.prototype.addDoneTransition = function (destState) {
        if (this._doneTransition) {
            this.sm._rejectCurrentPromise(new Error("Done transition for state '" + this.name + "' already registered"));
        }
        this._doneTransition = new Transition_1.Transition('Done', destState);
        this.addTransition(this._doneTransition);
    };
    State.prototype._enter = function (transition, result) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var ret, entryCounterThisTime, retFromPromise, e_1, _i, _a, trans;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this._entryCounter++;
                        this._isCurrent = true;
                        this._lastIncomingResults = result;
                        // Here we subscribe to all events
                        this._events.forEach(function (eventCont) {
                            _this.sm.on(eventCont.event, eventCont.handler);
                        });
                        this._typedEvents.forEach(function (eventCont) {
                            eventCont.event.on(eventCont.handler);
                        });
                        // Here we emit the sm transition event since all the state is appropriate at this time
                        this.sm.stateChanged.emit(transition);
                        if (!this.onEntry) return [3 /*break*/, 9];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 7, , 8]);
                        ret = this.onEntry(transition, result);
                        if (!(ret instanceof Promise)) return [3 /*break*/, 5];
                        if (!this._doneTransition) return [3 /*break*/, 3];
                        entryCounterThisTime = this._entryCounter;
                        return [4 /*yield*/, ret];
                    case 2:
                        retFromPromise = _b.sent();
                        if (this._isCurrent && this._entryCounter === entryCounterThisTime) {
                            this._doneTransition.trigger(retFromPromise);
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        ret.catch(function (error) {
                            _this.sm._rejectCurrentPromise(error);
                        });
                        _b.label = 4;
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        if (this._doneTransition) {
                            this._doneTransition.trigger(ret);
                        }
                        _b.label = 6;
                    case 6: return [3 /*break*/, 8];
                    case 7:
                        e_1 = _b.sent();
                        this.sm._rejectCurrentPromise(e_1);
                        return [3 /*break*/, 8];
                    case 8: return [3 /*break*/, 10];
                    case 9:
                        if (this._doneTransition) {
                            this._doneTransition.trigger(null);
                        }
                        _b.label = 10;
                    case 10:
                        // We evaluate this state's transitions' onEntry methods
                        // There is a chance that we are not current anymore
                        if (this._isCurrent) {
                            for (_i = 0, _a = this.transitions; _i < _a.length; _i++) {
                                trans = _a[_i];
                                trans._enter();
                            }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    State.prototype._exit = function (transition) {
        var _this = this;
        this._isExiting = true;
        // Here we unsubscribe from all events
        this._events.forEach(function (eventCont) {
            _this.sm.removeListener(eventCont.event, eventCont.handler);
        });
        this._typedEvents.forEach(function (eventCont) {
            eventCont.event.removeListener(eventCont.handler);
        });
        var retPromise;
        if (this.onExit) {
            try {
                var ret = this.onExit(transition);
                if (ret instanceof Promise) {
                    retPromise = ret;
                }
            }
            catch (e) {
                this.sm._rejectCurrentPromise(e);
            }
        }
        for (var _i = 0, _a = this.transitions; _i < _a.length; _i++) {
            var trans = _a[_i];
            trans._exit();
        }
        var exitCleanup = function () {
            _this._exitCounter++;
            _this._isExiting = false;
            _this._isCurrent = false;
        };
        if (retPromise) {
            return retPromise
                .then(exitCleanup)
                .catch(function (e) { return _this.sm._rejectCurrentPromise(e); });
        }
        else {
            exitCleanup();
        }
    };
    State.prototype._stop = function () {
        var retPromise;
        this.transitions.forEach(function (trans) {
            if (trans instanceof TimeoutTransition_1.TimeoutTransition) {
                trans.stop();
            }
        });
        // First we call onStop if it is defined
        if (this.onStop) {
            try {
                var ret = this.onStop();
                if (ret instanceof Promise) {
                    retPromise = ret;
                }
            }
            catch (e) {
                this.sm._rejectCurrentPromise(e);
            }
        }
        if (!this._isExiting && this._exitCounter !== this._entryCounter) {
            // We also call exit
            var finalTransition = new Transition_1.Transition('FINAL TRANSITON', null);
            finalTransition._setSourceState(this);
            var exitRet = this._exit(finalTransition);
            if (exitRet instanceof Promise) {
                var exitRetPr_1 = exitRet;
                if (retPromise) {
                    retPromise = retPromise.then(function () { return exitRetPr_1; });
                }
                else {
                    retPromise = exitRetPr_1;
                }
            }
        }
        if (retPromise) {
            return retPromise;
        }
    };
    /**
     * Retrieves the last result that was provided to this state upon entry
     * @returns {Object}
     */
    State.prototype.getLastIncomingResult = function () {
        return this._lastIncomingResults;
    };
    /**
     * Returns whether this is the current state of the state machine
     * @return {boolean}
     */
    State.prototype.isCurrent = function () {
        return this._isCurrent;
    };
    State.prototype.transitionTo = function (arg1, arg2, arg3) {
        var destState;
        var name;
        var result;
        // If we are being given a transition name
        if (typeof arg1 === 'string') {
            name = arg1;
            if (arg2 instanceof State) {
                destState = arg2;
                result = arg3;
            }
            else {
                result = arg2;
            }
        }
        else if (arg1 instanceof State) {
            destState = arg1;
            result = arg2;
        }
        // Here we find the internal transition that matches
        var results = this._internals.filter(function (i) {
            return ((name === undefined) || (name === i.name)) &&
                ((destState === undefined) || (destState === i.destState));
        });
        if (results.length === 0) {
            throw new Error("State '" + this.name + "': no internal transition of name '" + name + "' to state '" + destState.name + "' found");
        }
        else if (results.length > 1) {
            throw new Error("State '" + this.name + "': has more than one matching internal transition: '" + name + "' to state '" + destState.name + "'");
        }
        else {
            var internal = results[0];
            if (!this._isCurrent) {
                var warn = "Attempting internal transition out of a non-current state. " +
                    ("This state: '" + this.name + "', target state: '" + internal.destState.name + "', ") +
                    ("transition name: '" + internal.transition.name + "'. ");
                if (this.sm.current) {
                    warn += "Current state: '" + this.sm.current.name + "'.";
                }
                else {
                    warn += "The state machine is currently not running.";
                }
                log.warn(warn);
            }
            else {
                internal.transition.trigger(result);
            }
        }
    };
    /**
     * Destroys the state. Should be called when state machine is getting discarded
     */
    State.prototype.destroy = function () {
        var _this = this;
        // this._isCurrent = false;
        this._events.forEach(function (eventContainer) {
            _this.sm.removeListener(eventContainer.event, eventContainer.handler);
        });
        this._events.clear();
        this._typedEvents.forEach(function (eventContainer) {
            eventContainer.event.removeListener(eventContainer.handler);
        });
        this._typedEvents.clear();
        this.transitions.forEach(function (trans) { return trans.destroy(); });
        // this.transitions = [];
        // this._internalTransitions.clear();
        this._lastIncomingResults = null;
        this.onEntry = null;
        this.onExit = null;
        this.onUpdate = null;
        this.onStop = null;
    };
    State.prototype.toString = function () {
        return 'State: ' + this.name;
    };
    return State;
}());
exports.State = State;
// For unit tests
State.prototype.log = log;

},{"./TimeoutTransition":3,"./Transition":4,"./log":5}],2:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var events = require("events");
var fs = require("fs");
var jibo_typed_events_1 = require("jibo-typed-events");
var Transition_1 = require("./Transition");
var log_1 = require("./log");
var log = log_1.default.createChild("StateMachine");
var StateMachine = (function (_super) {
    __extends(StateMachine, _super);
    function StateMachine() {
        var _this = _super.call(this) || this;
        _this.states = new Map();
        _this.stateChanged = new jibo_typed_events_1.Event("State transitioned");
        _this._inTransition = false;
        _this._loopDetection = new Map();
        _this._isStopped = false;
        _this._traceIndex = 0;
        _this._trace = [];
        _this._cleanupAll();
        return _this;
    }
    /**
     * Rejects currently running start promise
     * @param {string|Error} error
     * @private
     */
    StateMachine.prototype._rejectCurrentPromise = function (error) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (this._promiseReject) {
                    this._promiseReject(error);
                }
                else {
                    log.warn("Reject being called when SM isn't running, with error: ", error);
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Resoles currently running start promise
     * @param {object} data
     * @private
     */
    StateMachine.prototype._resolveCurrentPromise = function (data) {
        if (this._promiseResolve) {
            this._promiseResolve(data);
        }
        else {
            log.warn("Resolve called when SM isn't running, with data: ", data);
        }
    };
    /**
     * Destroys the state machine and all of its states.
     * Should be called when state machine is getting discarded
     */
    StateMachine.prototype.destroy = function () {
        var _this = this;
        var destroy = function () {
            _this.states.forEach(function (state) {
                state.destroy();
            });
            _this.states.clear();
            _this._cleanupAll();
        };
        if (this._currentRunPromise) {
            return this.stop().then(destroy);
        }
        else {
            return Promise.resolve().then(destroy);
        }
    };
    /**
     * Returns current state of state machine
     * @returns {State}
     */
    StateMachine.prototype.getCurrentState = function () {
        return this.current;
    };
    /**
     * Sets the initial state of the state machine
     * @param state
     * @returns {StateMachine}
     */
    StateMachine.prototype.setInitial = function (state) {
        this.initial = state;
        return this;
    };
    /**
     * Starts the state machine. This is required to be called once.
     * @param {Object} [input] - Optional input data
     * @returns {Promise} resolved when any state throws an error or calls this.exit();
     */
    StateMachine.prototype.start = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var initialTransition;
            return __generator(this, function (_a) {
                if (!this.initial) {
                    return [2 /*return*/, Promise.reject(new Error("State machine does not yet have an initial state"))];
                }
                else if (this._currentRunPromise) {
                    return [2 /*return*/, Promise.reject(new Error("State machine already running, please call stop first"))];
                }
                this._currentRunPromise = new Promise(function (resolve, reject) {
                    _this._promiseResolve = resolve;
                    _this._promiseReject = reject;
                }).then(function (data) {
                    _this._cleanupAfterStart();
                    return data;
                }).catch(function (e) { return __awaiter(_this, void 0, void 0, function () {
                    var stopError_1;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                _a.trys.push([0, 2, 3, 4]);
                                return [4 /*yield*/, this.stop()];
                            case 1:
                                _a.sent();
                                return [3 /*break*/, 4];
                            case 2:
                                stopError_1 = _a.sent();
                                log.error("Error while stopping state machine because of an earlier error");
                                log.error(stopError_1);
                                return [3 /*break*/, 4];
                            case 3:
                                this._cleanupAfterStart();
                                throw e;
                            case 4: return [2 /*return*/];
                        }
                    });
                }); });
                this._cleanupBeforeStart();
                this.current = this.initial;
                this.current._isCurrent = true;
                initialTransition = new Transition_1.Transition('INITIAL TRANSITON', this.initial);
                this._addToTrace(initialTransition);
                this.current._enter(initialTransition, input);
                return [2 /*return*/, this._currentRunPromise];
            });
        });
    };
    /**
     * Informs whether the state machine is currently running
     * @return {boolean} True if the state machine is currently running
     */
    StateMachine.prototype.isRunning = function () {
        return (!!this._currentRunPromise);
    };
    /**
     * Waits until state machine is finished. If it isn't running, it resolves promise immediately
     * @return {Promise<any>}
     */
    StateMachine.prototype.waitUntilFinished = function () {
        if (this._currentRunPromise) {
            return this._currentRunPromise;
        }
        else {
            return Promise.resolve();
        }
    };
    StateMachine.prototype._addState = function (state) {
        if (this.states.has(state.name)) {
            this._rejectCurrentPromise(new Error("State with name '" + state.name + "' already exists"));
            return;
        }
        if (!this.initial) {
            this.setInitial(state);
        }
        this.states.set(state.name, state);
        return this;
    };
    /**
     * Get state by name
     * @param {string} name
     * @returns {State}
     */
    StateMachine.prototype.getState = function (name) {
        return this.states.get(name);
    };
    StateMachine.prototype._addToTrace = function (transition) {
        var te = {
            transition: transition,
            timestamp: Date.now(),
        };
        if (this._trace.length < StateMachine.TRACE_MAX_LENGTH) {
            this._trace.push(te);
        }
        else {
            this._trace[this._traceIndex] = te;
            this._traceIndex = (this._traceIndex + 1) % StateMachine.TRACE_MAX_LENGTH;
        }
    };
    /**
     * Updates the current state, if it has an onUpdate callback.
     * Normal functionality of the StateMachine does not require that you call this, only if your
     * state should be connected to an update loop.
     */
    StateMachine.prototype.update = function () {
        if (this.current && this.current.onUpdate) {
            this.current.onUpdate();
        }
    };
    /**
     * Stops the current state (calls its onStop callback).
     * @returns {Promise} resolved when stop is complete
     */
    StateMachine.prototype.stop = function (data) {
        var _this = this;
        // If we have a current run instance
        if (this._currentRunPromise) {
            this._isStopped = true;
            if (this.current) {
                this._isStopped = true;
                var stopRet_1 = this.current._stop();
                // We wait for the exit promise to complete if it exists
                var exitWait = this._currentStateExitPromise || Promise.resolve();
                if (stopRet_1 instanceof Promise) {
                    return exitWait.then(function () { return stopRet_1; }).then(function () {
                        _this.current = null;
                        _this._resolveCurrentPromise(data);
                    });
                }
                else {
                    return exitWait.then(function () {
                        _this.current = null;
                        _this._resolveCurrentPromise(data);
                    });
                }
            }
            else {
                return Promise.resolve();
            }
        }
        else {
            // Stop being called but SM wasn't running
            return Promise.resolve();
        }
    };
    /**
     * Returns the trace of transitions leading up to the current one (limited
     * by StateMachine.TRACE_MAX_LENGTH)
     * @returns {TraceElement[]}
     */
    StateMachine.prototype.getTrace = function () {
        var out = [];
        if (this._trace.length < StateMachine.TRACE_MAX_LENGTH) {
            for (var i = 0; i < this._trace.length; i++) {
                out.push(this._trace[i]);
            }
        }
        else {
            for (var i = 0; i < this._trace.length; i++) {
                var ind = (this._traceIndex + i) % StateMachine.TRACE_MAX_LENGTH;
                out.push(this._trace[ind]);
            }
        }
        return out;
    };
    /**
     * Provides a convenient string representation of getTrace()
     * @returns {string}
     */
    StateMachine.prototype.traceToString = function () {
        var trace = this.getTrace();
        var str = '';
        trace.forEach(function (t) {
            var sName = t.transition.getSourceState() ? t.transition.getSourceState().name : 'none';
            str += "Source: '" + sName + "' -> Transition: '" + t.transition.name + "' @ " + t.timestamp + " " +
                ("-> Destination: '" + t.transition.getDestinationState().name + "' \n");
        });
        return str;
    };
    /**
     * Transitions using a particular transition
     * @param transition
     * @param result
     * @private
     */
    StateMachine.prototype._transitionTo = function (transition, result) {
        var _this = this;
        if (this._isStopped) {
            return;
        }
        // if the state machine is stopped, do nothing
        if (!this.current) {
            return;
        }
        // See if this is the first in sequence of continuous transitions
        var isThisFirstTransition = !this._inTransition;
        if (isThisFirstTransition) {
            this._inTransition = true;
        }
        // Here we look for loops
        var destName = transition.getDestinationState().name;
        var loopCounter = this._loopDetection.get(destName);
        if (!loopCounter) {
            this._loopDetection.set(destName, 1);
        }
        else if (loopCounter > StateMachine.MAX_LOOPS) {
            return this._rejectCurrentPromise(new Error("State machine has looped '" + loopCounter + "' times in " +
                ("one transition. Next state: '" + destName + "'")));
        }
        else {
            this._loopDetection.set(destName, loopCounter + 1);
        }
        this._addToTrace(transition);
        if (this.current !== transition.getSourceState()) {
            this._rejectCurrentPromise(new Error("Can't apply transition '" + transition + "' since the "
                + ("current state is '" + this.current + "'")));
        }
        // Notify current state and transitions of exit
        var exitDone = function () {
            _this._currentStateExitPromise = null;
            if (!_this._isStopped) {
                // Notify next state and transitions of entry
                _this.current = transition.getDestinationState();
                _this.current._enter(transition, result);
                // When leaving the first transition in a sequence, we clear our loop
                // detection mechanism
                if (isThisFirstTransition) {
                    _this._inTransition = false;
                    _this._loopDetection.clear();
                }
            }
        };
        var exitRet = this.current._exit(transition);
        if (exitRet instanceof Promise) {
            this._currentStateExitPromise = exitRet;
            return exitRet.then(exitDone).catch(function (e) { return _this._rejectCurrentPromise(e); });
        }
        else {
            exitDone();
        }
    };
    /**
     * Produces a .dot file from current state graph
     * Writes dotfile to disk at filepath
     * @param {string} filePath
     * @returns {string}
     */
    StateMachine.prototype.toDotFile = function (filePath) {
        var toVisit = [this.initial];
        var visited = new Set();
        var str = 'digraph graphname {\n';
        var _loop_1 = function () {
            var state = toVisit.pop();
            if (visited.has(state)) {
                return "continue";
            }
            this_1.states.forEach(function (state) {
                var name = state.name.replace(/\"/ig, '\\"');
                str += "\"" + name + "\" [style=filled,color=lightgray];\n";
            });
            visited.add(state);
            state.transitions.forEach(function (transition) {
                var name = state.name.replace(/\"/ig, '\\"');
                var targetName = transition.getDestinationState().name.replace(/\"/ig, '\\"');
                var transitionName = transition.name.replace(/\"/ig, '\\"');
                str += "  \"" + name + "\" -> \"" + targetName + "\" [label = \"" + transitionName + "\"];\n";
                if (!visited.has(transition.getDestinationState())) {
                    toVisit.push(transition.getDestinationState());
                }
            });
        };
        var this_1 = this;
        while (toVisit.length) {
            _loop_1();
        }
        str += '}';
        if (filePath) {
            fs.writeFileSync(filePath, str, { encoding: 'utf8' });
        }
        return str;
    };
    StateMachine.prototype._cleanupAll = function () {
        this._cleanupBeforeStart();
        this._cleanupAfterStart();
    };
    StateMachine.prototype._cleanupAfterStart = function () {
        this._isStopped = false;
        this._currentRunPromise = null;
        this._currentStateExitPromise = null;
        this._promiseResolve = null;
        this._promiseReject = null;
        this._inTransition = false;
    };
    StateMachine.prototype._cleanupBeforeStart = function () {
        this._isStopped = false;
        this._inTransition = false;
        this._traceIndex = 0;
        this._trace = [];
        this._loopDetection.clear();
    };
    return StateMachine;
}(events.EventEmitter));
StateMachine.TRACE_MAX_LENGTH = 10;
StateMachine.MAX_LOOPS = 5;
exports.StateMachine = StateMachine;

},{"./Transition":4,"./log":5,"events":undefined,"fs":undefined,"jibo-typed-events":undefined}],3:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Transition_1 = require("./Transition");
var TimeoutTransition = (function (_super) {
    __extends(TimeoutTransition, _super);
    function TimeoutTransition(name, destState, timeMs) {
        var _this = _super.call(this, name, destState) || this;
        _this.onEntry = function () {
            _this._timeoutHandle = setTimeout(function () {
                _this._timeoutHandle = null;
                _this.trigger();
            }, timeMs);
        };
        _this.onExit = function () {
            if (_this._timeoutHandle) {
                clearTimeout(_this._timeoutHandle);
                _this._timeoutHandle = null;
            }
        };
        return _this;
    }
    TimeoutTransition.prototype.stop = function () {
        this.onExit();
    };
    return TimeoutTransition;
}(Transition_1.Transition));
exports.TimeoutTransition = TimeoutTransition;

},{"./Transition":4}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Transition = (function () {
    function Transition(name, _destState) {
        this.name = name;
        this._destState = _destState;
    }
    /**
     * Called to trigger the current transition
     * @param {object} [result] to pass to next state
     */
    Transition.prototype.trigger = function (result) {
        if (this._sourceState) {
            this._sourceState.sm._transitionTo(this, result);
        }
    };
    Transition.prototype._enter = function () {
        if (this.onEntry) {
            this.onEntry();
        }
    };
    Transition.prototype._exit = function () {
        if (this.onExit) {
            this.onExit();
        }
    };
    /**
     * The source state for this transition
     * @returns {State}
     */
    Transition.prototype.getSourceState = function () {
        return this._sourceState;
    };
    /**
     * The destination for this transition
     * @returns {State}
     */
    Transition.prototype.getDestinationState = function () {
        return this._destState;
    };
    /**
     * Sets the source state, should only be called when adding transition to
     * the source state.
     * Can only be called once per transition
     * @param sourceState
     * @private
     */
    Transition.prototype._setSourceState = function (sourceState) {
        if (this._sourceState) {
            this._sourceState.sm._rejectCurrentPromise(new Error("Transition '" + this + "' already has source state"));
        }
        this._sourceState = sourceState;
    };
    /**
     * Destroys the transition. Should be called when state machine is getting discarded
     */
    Transition.prototype.destroy = function () {
        this.onEntry = null;
        this.onExit = null;
        this._sourceState = null;
    };
    Transition.prototype.toString = function () {
        return this.name;
    };
    return Transition;
}());
exports.Transition = Transition;

},{}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Log = require("jibo-log");
exports.default = new Log('StateMachine');

},{"jibo-log":undefined}],6:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./StateMachine"));
__export(require("./Transition"));
__export(require("./State"));
__export(require("./TimeoutTransition"));

},{"./State":1,"./StateMachine":2,"./TimeoutTransition":3,"./Transition":4}]},{},[6])(6)
});
