// jibo-log/lib/debug/syslog-backend.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _syslogClient = require('syslog-client');

var _syslogClient2 = _interopRequireDefault(_syslogClient);

var _formatHelper = require('./format-helper');

var _formatHelper2 = _interopRequireDefault(_formatHelper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const syslogInfo = {
    port: 514,
    target: "127.0.0.1"
};

class SyslogBackend extends _events2.default {
    // gets original arguments
    constructor() {
        super();
        this.isWriting = false;
        this.writeQueue = [];
        this.client = undefined;
        this.enabled = false;
        this._test = []; // map of what "names" can go to syslog

        if (process.platform === 'linux' && process.arch === 'arm') {
            this.enabled = true;
        } else {
            this.enabled = false; // don't bother on anything other than our linux/arm environment
        }

        // if we are turning on syslogging and there isn't a client, then start the client
        if (this.enabled && !this.client) {
            let options = {
                port: syslogInfo.port,
                transport: _syslogClient2.default.Transport.Udp
            };
            this.client = _syslogClient2.default.createClient(syslogInfo.target, options);
            console.warn("Initializing a syslog client");

            this.client.on("error", function (error) {
                console.error(error);
                if (this.client) {
                    this.client.close();
                    this.client = undefined;
                }
            });
        }

        // if we are turning off syslogging and there currently is a client, then close the client
        if (!this.enabled && this.client) {
            this.client.close();
            this.client = undefined;
        }
    }

    // allow logging with this name to go to syslog
    allow(name) {
        if (name) {
            this._test.push(name);
        }
    }

    processNextInQueue() {
        var _this = this;

        if (this.writeQueue.length === 0) {
            this.isWriting = false;
            return;
        }

        var _writeQueue$shift = this.writeQueue.shift();

        const name = _writeQueue$shift.name;
        const level = _writeQueue$shift.level;
        const args = _writeQueue$shift.args;

        let type = undefined;
        switch (level) {
            case 'log':
            case 'debug':
                {
                    type = "Debug";
                    break;
                }
            case 'warn':
                {
                    type = "Warning";
                    break;
                }
            case 'error':
                {
                    type = "Error";
                    break;
                }
            case 'info':
            default:
                {
                    type = "Informational";
                    break;
                }
        }

        let logOptions = {
            facility: _syslogClient2.default.Facility.Local0,
            severity: _syslogClient2.default.Severity[type]
        };

        let msg = (0, _formatHelper2.default)(args);
        this.isWriting = true;
        this.client.log(`${ name }: [${ level }] ${ msg }`, logOptions, function (error) {
            if (error) {
                console.error(error);
            }
            _this.client.close();
            _this.processNextInQueue();
        });
    }

    write(name, level, args) {
        // EventEmitter
        // only write to syslog if it's enabled, has a client and the name exists in global mapping
        if (this.enabled && this.client && this._test.indexOf(name) >= 0) {
            this.writeQueue.push({ name, level, args });
            if (!this.isWriting) {
                this.processNextInQueue();
            }
        }
    }
}

if (!GLOBAL.syslogBackend) {
    // Export the syslogBackend singleton
    GLOBAL.syslogBackend = new SyslogBackend();
}

exports.default = GLOBAL.syslogBackend;
