// jibo-action-system/lib/jibo-action-system.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboActionSystem = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_cai_utils_1 = require("jibo-cai-utils");
var Types_1 = require("../common/Types");
var log_1 = require("../common/log");
var DEFAULT_ACTION_TIMEOUT = jibo_cai_utils_1.TimeUtils.secondsToMs(30);
var Action = (function () {
    function Action(parent, name, options) {
        if (options === void 0) { options = {}; }
        this.name = name;
        this.options = options;
        this.status = Types_1.ActionStatus.NOT_RUNNING;
        this.parent = parent;
        this.options.driveEffects = this.options.driveEffects || new Map();
        this.options.timeoutMs = (this.options.timeoutMs === undefined) ?
            DEFAULT_ACTION_TIMEOUT :
            this.options.timeoutMs;
        this.log = log_1.log.createChild("Action." + name);
    }
    Action.prototype.cancel = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._cancel(Types_1.ActionResult.CANCELED);
                return [2];
            });
        });
    };
    Action.prototype._cancel = function (result) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this._isCanceling && this._runResolve) {
                    this._isCanceling = true;
                    try {
                        this.cancelInternal();
                    }
                    catch (e) {
                        this.log.error('Error canceling: ', e);
                    }
                    finally {
                        this.status = Types_1.ActionStatus.NOT_RUNNING;
                        this._isCanceling = false;
                        if (this._runResolve) {
                            this._runResolve(result);
                        }
                    }
                }
                return [2];
            });
        });
    };
    Action.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Promise.resolve()];
            });
        });
    };
    Action.prototype.getDriveEffect = function (drive) {
        return this.options.driveEffects.get(drive) || 0;
    };
    Action.prototype.run = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var timeoutHandle, handler, cleanup;
            return __generator(this, function (_a) {
                if (this.status === Types_1.ActionStatus.RUNNING) {
                    return [2, Promise.reject('Action already running')];
                }
                this.status = Types_1.ActionStatus.RUNNING;
                Action.active.push(this);
                this.log.debug("Active actions " + Action.active.length + ": ", Action.active);
                if (this.options.timeoutMs > 0) {
                    handler = function () { return _this._cancel(Types_1.ActionResult.TIMEOUT); };
                    handler.isGlobalTimer = true;
                    timeoutHandle = this.parent.jibo.timer.setTimeout(handler, this.options.timeoutMs);
                }
                cleanup = function () {
                    _this.status = Types_1.ActionStatus.NOT_RUNNING;
                    _this._runResolve = null;
                    if (timeoutHandle) {
                        timeoutHandle.destroy();
                    }
                    var ind = Action.active.indexOf(_this);
                    if (ind >= 0) {
                        Action.active.splice(ind, 1);
                    }
                };
                return [2, new Promise(function (resolve, reject) {
                        _this._runResolve = resolve;
                        _this.runInternal().then(function (result) {
                            _this.status = Types_1.ActionStatus.NOT_RUNNING;
                            _this.log.debug('Completed with result: ', result);
                            if (!_this._isCanceling && _this._runResolve) {
                                _this._runResolve(result);
                            }
                        }).catch(function (e) { return reject(e); });
                    }).catch(function (e) {
                        cleanup();
                        _this.log.error('Error running: ', e);
                        throw e;
                    }).then(function (result) {
                        cleanup();
                        return result;
                    })];
            });
        });
    };
    Action.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Promise.resolve(Types_1.ActionResult.SUCCEEDED)];
            });
        });
    };
    Action.prototype.isEligible = function (state) {
        return true;
    };
    return Action;
}());
Action.active = [];
exports.Action = Action;

},{"../common/Types":26,"../common/log":30,"jibo-cai-utils":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BasicPolicy_1 = require("./policies/BasicPolicy");
var MotivationGoal_1 = require("../goal/goals/MotivationGoal");
var Types_1 = require("../common/Types");
var _ = require("lodash/lodash.min");
var log_1 = require("./log");
var log = log_1.log.createChild('System');
var ActionSystem = (function () {
    function ActionSystem(parent) {
        this.parent = parent;
        this.actions = [];
        this.policy = new BasicPolicy_1.BasicPolicy(this.parent);
    }
    ActionSystem.prototype.addAction = function (action) {
        this.actions.push(action);
    };
    ActionSystem.prototype.chooseAction = function (state, goal, eligibleActions) {
        var suggestedActionProvider = goal.options.action;
        if (suggestedActionProvider) {
            var suggestedAction = suggestedActionProvider();
            if (suggestedAction.isEligible(state)) {
                return suggestedAction;
            }
        }
        var action = this.policy.selectAction(goal, state);
        if (action) {
            return action;
        }
        if (!eligibleActions) {
            eligibleActions = this.actions.filter(function (a) { return a.isEligible(state); });
        }
        if (eligibleActions.length < 1) {
            log.debug("no eligible actions");
        }
        if (goal instanceof MotivationGoal_1.MotivationGoal) {
            var bestScore = 0;
            var bestActions = [];
            var dat = this.parent.motivations.drivesAboveThreshold();
            var _loop_1 = function (a) {
                var score = 0;
                dat.forEach(function (d) {
                    score += a.getDriveEffect(Types_1.DriveName[d.name]) * d.value;
                });
                if (score < bestScore) {
                    bestActions = [a];
                    bestScore = score;
                }
                else if (Math.abs(score - bestScore) < 0.01) {
                    bestActions.push(a);
                }
            };
            for (var _i = 0, eligibleActions_1 = eligibleActions; _i < eligibleActions_1.length; _i++) {
                var a = eligibleActions_1[_i];
                _loop_1(a);
            }
            if (bestScore < 0) {
                return _.sample(bestActions);
            }
        }
    };
    return ActionSystem;
}());
exports.ActionSystem = ActionSystem;

},{"../common/Types":26,"../goal/goals/MotivationGoal":54,"./log":16,"./policies/BasicPolicy":17,"lodash/lodash.min":undefined}],3:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Types_1 = require("../common/Types");
var log_1 = require("./log");
var log = log_1.log.createChild('GoalDrivenAction');
var GoalDrivenAction = (function () {
    function GoalDrivenAction(goal, action, parent) {
        this.goal = goal;
        this.action = action;
        this.parent = parent;
        this.currentRunPromise = null;
    }
    GoalDrivenAction.prototype.run = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var failed;
            return __generator(this, function (_a) {
                failed = function () {
                    log.info("Action: '" + _this.action.name + "' failed'");
                    if (_this.goal.attempts >= _this.goal.options.maxAttempts) {
                        _this.goal.setFinishedStatus(Types_1.GoalFinishedStatus.FAILED);
                        _this.parent.goals.removeGoal(_this.goal);
                    }
                    else {
                        _this.goal.setProgressStatus(Types_1.GoalProgressStatus.ATTEMPTED);
                    }
                };
                log.info("Taking action: '" + this.action.name + "' to achieve goal '" + this.goal.name + "'");
                this.goal.attempts++;
                this.currentRunPromise = this.action.run()
                    .then(function (result) {
                    log.info("Completed action: '" + _this.action.name + "' with result: '" + result + "'");
                    _this.currentRunPromise = null;
                    switch (result) {
                        case Types_1.ActionResult.SUCCEEDED:
                            _this.goal.setFinishedStatus(Types_1.GoalFinishedStatus.SUCCEEDED);
                            _this.parent.motivations.applyMotivationalEffects(_this.action);
                            _this.parent.goals.removeGoal(_this.goal);
                            break;
                        case Types_1.ActionResult.TIMEOUT:
                        case Types_1.ActionResult.CANCELED:
                        case Types_1.ActionResult.FAILED:
                            failed();
                            break;
                        default:
                            log.warn("Action outcome '" + result + "' for action '" + _this.action.name + "' unknown");
                    }
                    return result;
                }).catch(function (error) {
                    log.error("Error running action " + _this.action.name + ": ", error);
                    _this.currentRunPromise = null;
                    failed();
                    return Types_1.ActionResult.FAILED;
                });
                return [2, this.currentRunPromise];
            });
        });
    };
    return GoalDrivenAction;
}());
exports.GoalDrivenAction = GoalDrivenAction;

},{"../common/Types":26,"./log":16}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var Types_1 = require("../../common/Types");
var _ = require("lodash/lodash.min");
var AnimationAction = (function (_super) {
    __extends(AnimationAction, _super);
    function AnimationAction(parent, data, options, name) {
        if (name === void 0) { name = "Animation Action"; }
        var _this = _super.call(this, parent, name, options) || this;
        _this.data = data;
        return _this;
    }
    AnimationAction.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var anim, animRes, animConfig, Orientation, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.data.category) {
                            animRes = this.parent.jibo.animDB.query({
                                category: this.data.category,
                                includeMeta: this.data.includeMeta || [],
                                includeSomeMeta: this.data.includeSomeMeta || [],
                                excludeMeta: this.data.excludeMeta || []
                            });
                            if (!animRes || animRes.matching.length === 0) {
                                this.log.error("Couldn't find animation of category '" + this.data.category + "'");
                                return [2, Types_1.ActionResult.FAILED];
                            }
                            anim = _.sample(animRes.matching);
                        }
                        else if (this.data.name) {
                            anim = this.parent.jibo.animDB.getAnimByName(this.data.name);
                            if (!anim) {
                                this.log.error("No animation of name '" + this.data.name + "' found");
                                return [2, Types_1.ActionResult.FAILED];
                            }
                        }
                        else {
                            this.log.error("Invalid anim data");
                            return [2, Types_1.ActionResult.FAILED];
                        }
                        animConfig = {};
                        Orientation = this.parent.jibo.animDB.Orientation;
                        if (this.data.randomOrientation) {
                            animConfig.orientation = (Math.random() > 0.5) ? Orientation.LEFT : Orientation.RIGHT;
                        }
                        this.log.info("AnimationAction playing animation of name: '" + anim.name + "', sampled from cat: '" + this.data.category + "', meta: '" + this.data.includeMeta + "'");
                        _a = this;
                        return [4, anim.createFromConfig(animConfig)];
                    case 1:
                        _a.instance = _b.sent();
                        return [4, this.instance.play()];
                    case 2:
                        _b.sent();
                        this.instance = null;
                        return [2, Types_1.ActionResult.SUCCEEDED];
                }
            });
        });
    };
    AnimationAction.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (this.instance) {
                    this.instance.stop();
                }
                return [2];
            });
        });
    };
    return AnimationAction;
}(Action_1.Action));
exports.AnimationAction = AnimationAction;

},{"../../common/Types":26,"../Action":1,"lodash/lodash.min":undefined}],5:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var Types_1 = require("../../common/Types");
var BasicLookat = (function (_super) {
    __extends(BasicLookat, _super);
    function BasicLookat(parent, location) {
        var _this = _super.call(this, parent, 'Basic Lookat') || this;
        _this.location = location;
        return _this;
    }
    BasicLookat.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this._handle) return [3, 2];
                        return [4, this._handle.cancel()];
                    case 1:
                        _a.sent();
                        this._handle = null;
                        _a.label = 2;
                    case 2:
                        if (!this._mode) return [3, 4];
                        return [4, this._mode.release()];
                    case 3:
                        _a.sent();
                        this._mode = null;
                        _a.label = 4;
                    case 4: return [2];
                }
            });
        });
    };
    BasicLookat.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var expression, result, _a, pos, _b, e_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        expression = this.parent.jibo.expression;
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 7, , 8]);
                        _a = this;
                        return [4, expression.pushAttentionMode(expression.AttentionMode.COMMAND)];
                    case 2:
                        _a._mode = _c.sent();
                        pos = this.location;
                        _b = this;
                        return [4, expression.acquireTarget({
                                position: {
                                    x: pos.x,
                                    y: pos.y,
                                    z: Math.max(0.7, pos.z)
                                }
                            })];
                    case 3:
                        _b._handle = _c.sent();
                        return [4, this._handle.promise];
                    case 4:
                        _c.sent();
                        if (!this._mode) return [3, 6];
                        return [4, this._mode.release()];
                    case 5:
                        _c.sent();
                        this._mode = null;
                        _c.label = 6;
                    case 6:
                        result = Types_1.ActionResult.SUCCEEDED;
                        return [3, 8];
                    case 7:
                        e_1 = _c.sent();
                        this.log.error("'" + this.name + ": error: '", e_1);
                        result = Types_1.ActionResult.FAILED;
                        return [3, 8];
                    case 8:
                        this._handle = null;
                        this.parent.events.oriented.emit();
                        return [2, result];
                }
            });
        });
    };
    return BasicLookat;
}(Action_1.Action));
exports.BasicLookat = BasicLookat;

},{"../../common/Types":26,"../Action":1}],6:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var Types_1 = require("../../common/Types");
var BeSkillSwitchAction = (function (_super) {
    __extends(BeSkillSwitchAction, _super);
    function BeSkillSwitchAction(parent, goalData) {
        var _this = _super.call(this, parent, 'Be Skill Switch Action') || this;
        _this.goalData = goalData;
        return _this;
    }
    BeSkillSwitchAction.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.parent.currentSkillName = this.goalData.skillName;
                return [2, Types_1.ActionResult.SUCCEEDED];
            });
        });
    };
    return BeSkillSwitchAction;
}(Action_1.Action));
exports.BeSkillSwitchAction = BeSkillSwitchAction;

},{"../../common/Types":26,"../Action":1}],7:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var Types_1 = require("../../common/Types");
var EmitHeadPat = (function (_super) {
    __extends(EmitHeadPat, _super);
    function EmitHeadPat(parent, data) {
        var _this = _super.call(this, parent, "Emit Head Pat Action") || this;
        _this.data = data;
        return _this;
    }
    EmitHeadPat.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.parent.events.headPat.emit(this.data);
                return [2, Types_1.ActionResult.SUCCEEDED];
            });
        });
    };
    return EmitHeadPat;
}(Action_1.Action));
exports.EmitHeadPat = EmitHeadPat;

},{"../../common/Types":26,"../Action":1}],8:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var jibo_state_machine_1 = require("jibo-state-machine");
var Types_1 = require("../../common/Types");
var EventWaiter_1 = require("../../common/EventWaiter");
var classes = require("./LookatClasses");
var jibo_common_types_1 = require("jibo-common-types");
var HJOrient = (function (_super) {
    __extends(HJOrient, _super);
    function HJOrient(parent, location) {
        var _this = _super.call(this, parent, 'HJOrient') || this;
        _this.location = location;
        _this.haveReceivedResult = false;
        _this.sm = new jibo_state_machine_1.StateMachine();
        _this.doesVisibleFaceExist = new classes.PersonVisibleNearReferenceState(_this.sm, _this.parent, classes.AngleCalculatorType.XY);
        _this.attendToHJLocation = new classes.AttendToState(_this.sm, _this.parent, 'Attend to HJ location');
        _this.orientToInitialPerson = new classes.AttendToState(_this.sm, _this.parent, 'Orient to initial face');
        _this.haveReceivedResultState = new classes.HaveReceivedResultState(_this);
        _this.waitForFace = new classes.WaitForSignals(_this);
        _this.orientToFace = new classes.AttendToState(_this.sm, _this.parent, 'Orient to found face');
        _this.waitForResult = new classes.WaitForResults(_this);
        _this.done = new classes.DoneState(_this.sm);
        _this.sm.setInitial(_this.doesVisibleFaceExist);
        _this.doesVisibleFaceExist.installTransitions(_this.orientToInitialPerson, _this.attendToHJLocation);
        _this.orientToInitialPerson.addDoneTransition(_this.done);
        _this.attendToHJLocation.addDoneTransition(_this.haveReceivedResultState);
        _this.haveReceivedResultState.installTransitions(_this.done, _this.waitForFace);
        _this.waitForFace.installTransitions(_this.done, _this.orientToFace);
        _this.orientToFace.addDoneTransition(_this.waitForResult);
        _this.waitForResult.addDoneTransition(_this.done);
        _this.doesVisibleFaceExist.setReferencePosition(location);
        return _this;
    }
    HJOrient.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var pmHandler, e_1, trace;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.parent.jibo.expression.pushAttentionMode(jibo_common_types_1.AttentionMode.COMMAND)];
                    case 1:
                        pmHandler = _a.sent();
                        this.resultWaiters = [
                            ['Skill relaunch', new EventWaiter_1.TypedEventWaiter(null, this.parent.jibo.globalEvents.skillRelaunch).start()],
                            ['HJ Only', new EventWaiter_1.TypedEventWaiter(null, this.parent.jibo.gl.events.hjOnly).start()],
                            ['Global', new EventWaiter_1.TypedEventWaiter(null, this.parent.jibo.globalEvents.global).start()],
                            ['No Global', new EventWaiter_1.TypedEventWaiter(null, this.parent.jibo.gl.events.noGlobalMatch).start()],
                        ];
                        this.haveReceivedResult = false;
                        Promise.race(this.resultWaiters.map(function (e) {
                            return e[1].promise.then(function () { return e; });
                        })).then(function (result) {
                            _this.haveReceivedResult = true;
                            _this.log.info("HJOrient: event waiter '" + result[0] + "'");
                        });
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, 5, 7]);
                        return [4, this.sm.start()];
                    case 3:
                        _a.sent();
                        return [2, Types_1.ActionResult.SUCCEEDED];
                    case 4:
                        e_1 = _a.sent();
                        this.log.error('HJOrient state machine error: ', e_1);
                        return [2, Types_1.ActionResult.FAILED];
                    case 5:
                        this.log.debug(this.name + " SM trace: ", this.sm.traceToString());
                        trace = this.sm.getTrace();
                        this.log.info("HJOrient time to face: ", this.computeTimeToFace(trace));
                        this.log.info("HJOrient had initial face?: ", this.hadInitialFace(trace));
                        return [4, pmHandler.release()];
                    case 6:
                        _a.sent();
                        this.resultWaiters.forEach(function (e) { return e[1].stop(); });
                        return [7];
                    case 7: return [2];
                }
            });
        });
    };
    HJOrient.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.log.warn("'" + this.name + "' cancelled!");
                this.sm.stop();
                return [2];
            });
        });
    };
    HJOrient.prototype.computeTimeToFace = function (trace) {
        if (trace.length < 2) {
            return Number.POSITIVE_INFINITY;
        }
        for (var i = 1; i < trace.length; i++) {
            if (trace[i].transition.getDestinationState() === this.orientToInitialPerson ||
                trace[i].transition.getDestinationState() === this.orientToFace) {
                return trace[i].timestamp - trace[0].timestamp;
            }
        }
        return Number.POSITIVE_INFINITY;
    };
    HJOrient.prototype.hadInitialFace = function (trace) {
        for (var i = 0; i < trace.length; i++) {
            if (trace[i].transition.getDestinationState() === this.orientToInitialPerson) {
                return true;
            }
        }
        return false;
    };
    return HJOrient;
}(Action_1.Action));
exports.HJOrient = HJOrient;

},{"../../common/EventWaiter":23,"../../common/Types":26,"../Action":1,"./LookatClasses":11,"jibo-common-types":undefined,"jibo-state-machine":undefined}],9:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var Types_1 = require("../../common/Types");
var SSWGoal_1 = require("../../goal/goals/SSWGoal");
var Utils_1 = require("../../common/Utils");
var findRoot = require('find-root');
var ROOT = findRoot(__filename);
var SOUNDS_ROOT = ROOT + '/resources/audio/';
var Sound = {
    ONTOUCH: SOUNDS_ROOT + "touch-on.m4a",
    ONEXIT: SOUNDS_ROOT + "touch-off.m4a"
};
var SECONDS_TO_HOLD_TOUCH = 1;
var TOUCH_RESET_TIME = 1000;
var HeadTouch = (function (_super) {
    __extends(HeadTouch, _super);
    function HeadTouch(parent) {
        var _this = _super.call(this, parent, 'HeadTouch') || this;
        _this._canceled = false;
        _this._lastTouchTime = -1;
        return _this;
    }
    HeadTouch.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var promises, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promises = [];
                        return [4, new Promise(function (resolve) {
                                if (Date.now() - _this._lastTouchTime > TOUCH_RESET_TIME) {
                                    promises.push(_this._playSound(Sound.ONTOUCH));
                                }
                                _this._canceled = false;
                                _this.parent.jibo.system.events.touchOff.waitFor(SECONDS_TO_HOLD_TOUCH * 1000)
                                    .then(function (data) {
                                    resolve(Types_1.ActionResult.FAILED);
                                }).catch(function (timeout) {
                                    if (!_this._canceled) {
                                        promises.push(_this._playSound(Sound.ONEXIT));
                                        _this.log.info("quitting jibo via touch");
                                        _this.parent.jibo.analytics.track('Global Stop', { type: 'touch', status: 'done' });
                                        var sswGoal = new SSWGoal_1.SSWGoal(_this.parent, {
                                            skillName: Utils_1.SkillNames.IDLE,
                                            skillOptions: {}
                                        }, "Hold Switch to Idle Goal");
                                        sswGoal.events.finished.on(function (result) {
                                            if (result === Types_1.GoalFinishedStatus.SUCCEEDED) {
                                                resolve(Types_1.ActionResult.SUCCEEDED);
                                            }
                                            else {
                                                resolve(Types_1.ActionResult.FAILED);
                                            }
                                        });
                                        _this.parent.goals.addGoal(sswGoal);
                                    }
                                });
                            })];
                    case 1:
                        result = _a.sent();
                        return [4, Promise.all(promises)];
                    case 2:
                        _a.sent();
                        return [2, result];
                }
            });
        });
    };
    HeadTouch.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._canceled = true;
                return [2];
            });
        });
    };
    HeadTouch.prototype._playSound = function (path) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2, new Promise(function (resolve, reject) {
                        if (!_this.parent.jibo.loader.getCache('global')) {
                            return;
                        }
                        var sound = _this.parent.jibo.loader.cached(path);
                        if (sound) {
                            sound.play();
                        }
                        else {
                            _this.parent.jibo.loader.load({
                                type: 'sound',
                                id: path,
                                cache: 'global',
                                src: path,
                                complete: function (err, sound) {
                                    if (err) {
                                        reject();
                                        return;
                                    }
                                    sound.play();
                                    resolve();
                                }
                            });
                        }
                    })];
            });
        });
    };
    return HeadTouch;
}(Action_1.Action));
exports.HeadTouch = HeadTouch;

},{"../../common/Types":26,"../../common/Utils":28,"../../goal/goals/SSWGoal":55,"../Action":1,"find-root":undefined}],10:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var AnimationAction_1 = require("./AnimationAction");
var Types_1 = require("../../common/Types");
var Utils_1 = require("../../common/Utils");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var IdleAnimationVisual = (function (_super) {
    __extends(IdleAnimationVisual, _super);
    function IdleAnimationVisual(parent) {
        var _this = _super.call(this, parent, {
            category: 'idle',
            includeSomeMeta: ['pong', 'paddleball', 'art'],
            randomOrientation: true
        }, {
            driveEffects: new Map([[Types_1.DriveName.PLAYFUL, -1.0]])
        }, "IdleAnimationVisual") || this;
        _this.parent = parent;
        return _this;
    }
    IdleAnimationVisual.prototype.isEligible = function (state) {
        return (state.skillName === Utils_1.SkillNames.IDLE) &&
            (state.sleep === Types_1.SleepState.AWAKE) &&
            (this.parent.dateProvider.getDate().getTime() - state.skillStartTime.getTime()) > jibo_cai_utils_1.TimeUtils.minutesToMs(5);
    };
    IdleAnimationVisual.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var retVal;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, _super.prototype.runInternal.call(this)];
                    case 1:
                        retVal = _a.sent();
                        if (retVal) {
                            this.parent.jibo.analytics.track('Idle behavior', { idle_behavior_class: "visual" });
                        }
                        return [2, retVal];
                }
            });
        });
    };
    return IdleAnimationVisual;
}(AnimationAction_1.AnimationAction));
exports.IdleAnimationVisual = IdleAnimationVisual;
var IdleAnimationBody = (function (_super) {
    __extends(IdleAnimationBody, _super);
    function IdleAnimationBody(parent) {
        var _this = _super.call(this, parent, {
            category: 'idle',
            includeSomeMeta: ['hip-shake'],
            randomOrientation: true
        }, {
            driveEffects: new Map([[Types_1.DriveName.PLAYFUL, -1.0]])
        }, "IdleAnimationBody") || this;
        _this.parent = parent;
        return _this;
    }
    IdleAnimationBody.prototype.isEligible = function (state) {
        return (state.skillName === Utils_1.SkillNames.IDLE) &&
            (state.sleep === Types_1.SleepState.AWAKE) &&
            (this.parent.dateProvider.getDate().getTime() - state.skillStartTime.getTime()) > jibo_cai_utils_1.TimeUtils.minutesToMs(30);
    };
    IdleAnimationBody.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var retVal;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, _super.prototype.runInternal.call(this)];
                    case 1:
                        retVal = _a.sent();
                        if (retVal) {
                            this.parent.jibo.analytics.track('Idle behavior', { idle_behavior_class: "body" });
                        }
                        return [2, retVal];
                }
            });
        });
    };
    return IdleAnimationBody;
}(AnimationAction_1.AnimationAction));
exports.IdleAnimationBody = IdleAnimationBody;

},{"../../common/Types":26,"../../common/Utils":28,"./AnimationAction":4,"jibo-cai-utils":undefined}],11:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_state_machine_1 = require("jibo-state-machine");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var jibo_common_types_1 = require("jibo-common-types");
var log_1 = require("../log");
var FACE_ANGLE_THRESH = Math.PI / 8;
var AWAIT_TIMEOUT = 3000;
var OWNER_IDENTIFIER = 'AttentionCommand';
var Outcome;
(function (Outcome) {
    Outcome[Outcome["Result"] = 'Result'] = "Result";
    Outcome[Outcome["Face"] = 'Face'] = "Face";
    Outcome[Outcome["Timeout"] = 'Timeout'] = "Timeout";
    Outcome[Outcome["Motion"] = 'Motion'] = "Motion";
})(Outcome = exports.Outcome || (exports.Outcome = {}));
function faceNearReference(position, actionRuntime, angleCalculator) {
    var bestAngle = Number.MAX_SAFE_INTEGER;
    var bestFace = null;
    actionRuntime.jibo.lps.identity.getVisibleFaces().forEach(function (face) {
        if (face.in_fov) {
            var angle = angleCalculator(position, face.position);
            if (angle < bestAngle) {
                bestAngle = angle;
                bestFace = face;
            }
        }
    });
    log_1.log.info("Initial face: ", position, bestFace ? bestFace.id : null);
    if (bestFace && bestAngle < FACE_ANGLE_THRESH) {
        return bestFace;
    }
}
function angleXY(point1, point2) {
    var p1 = { x: point1.x, y: point1.y, z: 0 };
    var p2 = { x: point2.x, y: point2.y, z: 0 };
    return angle3d(p1, p2);
}
function angle3d(point1, point2) {
    var dotProduct = point1.x * point2.x + point1.y * point2.y + point1.z * point2.z;
    var length1 = Math.sqrt(point1.x * point1.x + point1.y * point1.y + point1.z * point1.z);
    var length2 = Math.sqrt(point2.x * point2.x + point2.y * point2.y + point2.z * point2.z);
    if (length1 === 0 || length2 === 0) {
        return Math.PI;
    }
    else {
        return Math.acos(dotProduct / (length1 * length2));
    }
}
var AngleCalculatorType;
(function (AngleCalculatorType) {
    AngleCalculatorType[AngleCalculatorType["XY"] = 0] = "XY";
    AngleCalculatorType[AngleCalculatorType["XYZ"] = 1] = "XYZ";
})(AngleCalculatorType = exports.AngleCalculatorType || (exports.AngleCalculatorType = {}));
;
var YesNoState = (function (_super) {
    __extends(YesNoState, _super);
    function YesNoState(sm, name, transitionNames) {
        if (transitionNames === void 0) { transitionNames = ['Yes', 'No']; }
        var _this = _super.call(this, sm, name) || this;
        _this.transitionNames = transitionNames;
        return _this;
    }
    YesNoState.prototype.installTransitions = function (yesState, noState) {
        this._yesState = yesState;
        this._noState = noState;
        this.addInternalTransition(this.transitionNames[0], this._yesState);
        this.addInternalTransition(this.transitionNames[1], this._noState);
    };
    return YesNoState;
}(jibo_state_machine_1.State));
exports.YesNoState = YesNoState;
var AttendToState = (function (_super) {
    __extends(AttendToState, _super);
    function AttendToState(sm, actionRuntime, name, floorZ) {
        if (floorZ === void 0) { floorZ = true; }
        var _this = _super.call(this, sm, name) || this;
        _this.actionRuntime = actionRuntime;
        _this.floorZ = floorZ;
        _this.log = log_1.log.createChild('AttendToState');
        _this.onEntry = _this._onEntry.bind(_this);
        _this.onStop = function () {
            if (_this._handle) {
                _this._handle.cancel();
                _this._handle = null;
            }
        };
        return _this;
    }
    AttendToState.prototype.overrideLocation = function (data) {
        return data;
    };
    AttendToState.prototype._onEntry = function (transition, data) {
        return __awaiter(this, void 0, void 0, function () {
            var jibo, pos, _a, e_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        jibo = this.actionRuntime.jibo;
                        data = this.overrideLocation(data);
                        if (!data || !((data.entity && data.entity.id) || (data.position && (data.position.x !== undefined)))) {
                            this.log.error('Bad AttendTarget data in AttendToState. SM trace: ', this.sm.traceToString());
                            throw new Error(this.name + ": Bad AttendTarget data " + JSON.stringify(data));
                        }
                        pos = null;
                        if (data.position) {
                            pos = {
                                x: data.position.x,
                                y: data.position.y,
                                z: this.floorZ ? Math.max(-0.1, data.position.z) : data.position.z
                            };
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 4, 5, 6]);
                        _a = this;
                        return [4, jibo.expression.acquireTarget({ position: pos, entity: data.entity })];
                    case 2:
                        _a._handle = _b.sent();
                        return [4, this._handle.promise];
                    case 3: return [2, _b.sent()];
                    case 4:
                        e_1 = _b.sent();
                        throw e_1;
                    case 5:
                        this._handle = null;
                        return [7];
                    case 6: return [2];
                }
            });
        });
    };
    return AttendToState;
}(jibo_state_machine_1.State));
exports.AttendToState = AttendToState;
var FlowState = (function (_super) {
    __extends(FlowState, _super);
    function FlowState(sm, actionRuntime, name, flow) {
        var _this = _super.call(this, sm, name) || this;
        _this.actionRuntime = actionRuntime;
        _this.flow = flow;
        _this.log = log_1.log.createChild('FlowState');
        _this.onEntry = _this._onEntry.bind(_this);
        _this.onStop = function () {
            if (_this._flowHandle) {
                _this._flowHandle.stop();
                _this._flowHandle = null;
            }
        };
        return _this;
    }
    FlowState.prototype._onEntry = function (trans, data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var jibo, result, options_1, e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        jibo = this.actionRuntime.jibo;
                        if (!!this.flow) return [3, 1];
                        this.log.warn("Skipping flow: '" + this.name + "'");
                        return [3, 6];
                    case 1:
                        this.log.info("Playing flow: '" + this.name + "'");
                        result = {};
                        options_1 = {
                            result: result,
                            assetPack: 'jibo/node_modules/jibo-action-system',
                            enableLogging: false,
                        };
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, 5, 6]);
                        return [4, jibo_cai_utils_1.PromiseUtils.promisify(function (h) {
                                _this._flowHandle = jibo.flow.run(_this.flow, options_1, h);
                            })];
                    case 3:
                        _a.sent();
                        return [3, 6];
                    case 4:
                        e_2 = _a.sent();
                        throw e_2;
                    case 5:
                        this._flowHandle = null;
                        return [7];
                    case 6: return [2, data];
                }
            });
        });
    };
    return FlowState;
}(jibo_state_machine_1.State));
exports.FlowState = FlowState;
var DoneState = (function (_super) {
    __extends(DoneState, _super);
    function DoneState(sm, name, data) {
        if (name === void 0) { name = 'Done'; }
        var _this = _super.call(this, sm, name) || this;
        _this.onEntry = function () {
            _this.sm.stop(data);
        };
        return _this;
    }
    return DoneState;
}(jibo_state_machine_1.State));
exports.DoneState = DoneState;
var PersonVisibleNearReferenceState = (function (_super) {
    __extends(PersonVisibleNearReferenceState, _super);
    function PersonVisibleNearReferenceState(sm, actionRuntime, angleCalculator) {
        var _this = _super.call(this, sm, 'Is person visible?') || this;
        _this.actionRuntime = actionRuntime;
        _this.onEntry = function () {
            if (!_this.referencePosition) {
                throw new Error(_this.name + ": Missing reference position");
            }
            var calculator = angleCalculator === AngleCalculatorType.XY ? angleXY : angle3d;
            var face = faceNearReference(_this.referencePosition, actionRuntime, calculator);
            if (face) {
                _this.transitionTo(_this._yesState, face);
            }
            else {
                _this.transitionTo(_this._noState, { position: _this.referencePosition });
            }
        };
        return _this;
    }
    PersonVisibleNearReferenceState.prototype.setReferencePosition = function (referencePosition) {
        this.referencePosition = referencePosition;
    };
    return PersonVisibleNearReferenceState;
}(YesNoState));
exports.PersonVisibleNearReferenceState = PersonVisibleNearReferenceState;
var WaitForResults = (function (_super) {
    __extends(WaitForResults, _super);
    function WaitForResults(hjOrient) {
        var _this = _super.call(this, hjOrient.sm, 'Waiting for results') || this;
        _this.hjOrient = hjOrient;
        _this.onEntry = function () {
            return Promise.race(_this.hjOrient.resultWaiters.map(function (e) { return e[1].promise; }));
        };
        return _this;
    }
    return WaitForResults;
}(jibo_state_machine_1.State));
exports.WaitForResults = WaitForResults;
var WaitForSignals = (function (_super) {
    __extends(WaitForSignals, _super);
    function WaitForSignals(hjOrient) {
        var _this = _super.call(this, hjOrient.sm, 'Waiting for face or signal') || this;
        _this.hjOrient = hjOrient;
        _this.onEntry = _this._onEntry.bind(_this);
        _this.onExit = function () {
            if (_this._attendHandle) {
                _this._attendHandle.cancel();
                _this._attendHandle = null;
            }
        };
        return _this;
    }
    WaitForSignals.prototype.installTransitions = function (rcvResults, faceFound) {
        this._rcvResults = rcvResults;
        this._faceFound = faceFound;
        this.addInternalTransition('Received results', this._rcvResults);
        this.addInternalTransition('Found face', this._faceFound);
    };
    WaitForSignals.prototype._onEntry = function () {
        var _this = this;
        this.hjOrient.log.debug('HJOrient waiting for signals');
        var promises = this.hjOrient.resultWaiters
            .map(function (e) { return e[1].promise.then(function () {
            _this.hjOrient.log.debug("result waiter resolved");
            return { outcome: Outcome.Result };
        }); });
        promises.push(this.waitForFace());
        Promise.race(promises).then(function (outcome) {
            _this.hjOrient.log.debug("hjOrient got outcome");
            if (outcome.outcome === Outcome.Face) {
                _this.transitionTo(_this._faceFound, outcome.data);
            }
            else if (outcome.outcome === Outcome.Result) {
                _this.transitionTo(_this._rcvResults);
            }
            else if (outcome.outcome === Outcome.Timeout) {
                _this.hjOrient.log.warn("Wait for signals timed out.");
                _this.transitionTo(_this._rcvResults);
            }
            else {
                throw new Error("Unknown outcome: '" + outcome + "'");
            }
        });
    };
    WaitForSignals.prototype.waitForFace = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4, this.hjOrient.parent.jibo.expression.awaitFace(AWAIT_TIMEOUT)];
                    case 1:
                        _a._attendHandle = _b.sent();
                        return [2, this._attendHandle.promise.then(function (face) {
                                if (face && face.status === jibo_common_types_1.ResultStatus.SUCCEEDED) {
                                    return { outcome: Outcome.Face, data: face };
                                }
                                else {
                                    return { outcome: Outcome.Timeout };
                                }
                            })];
                }
            });
        });
    };
    return WaitForSignals;
}(jibo_state_machine_1.State));
exports.WaitForSignals = WaitForSignals;
var HaveReceivedResultState = (function (_super) {
    __extends(HaveReceivedResultState, _super);
    function HaveReceivedResultState(hjOrient) {
        var _this = _super.call(this, hjOrient.sm, 'Have received result') || this;
        _this.hjOrient = hjOrient;
        _this.onEntry = function () {
            if (_this.hjOrient.haveReceivedResult) {
                _this.transitionTo(_this._yesState);
            }
            else {
                _this.transitionTo(_this._noState);
            }
        };
        return _this;
    }
    return HaveReceivedResultState;
}(YesNoState));
exports.HaveReceivedResultState = HaveReceivedResultState;
var LIGHTING_QUALITY_THRESHOLD = 0.25;
var LightingGoodState = (function (_super) {
    __extends(LightingGoodState, _super);
    function LightingGoodState(sm, actionRuntime) {
        var _this = _super.call(this, sm, 'Is lighting good?') || this;
        _this.actionRuntime = actionRuntime;
        _this.log = log_1.log.createChild('LightingGoodState');
        _this.onEntry = function () {
            var lightQual = _this.actionRuntime.jibo.lps.lightingQuality;
            _this.log.info("lighting quality: ", lightQual);
            if (lightQual >= LIGHTING_QUALITY_THRESHOLD) {
                _this.transitionTo(_this._yesState);
            }
            else {
                _this.transitionTo(_this._noState);
            }
        };
        return _this;
    }
    return LightingGoodState;
}(YesNoState));
exports.LightingGoodState = LightingGoodState;
var PickNextLocationState = (function (_super) {
    __extends(PickNextLocationState, _super);
    function PickNextLocationState(search) {
        var _this = _super.call(this, search.sm, 'Is another search location available?') || this;
        _this.onEntry = function () {
            if (search.searchCandidates.length) {
                var nextLoc = search.searchCandidates.shift();
                _this.transitionTo(_this._yesState, nextLoc);
            }
            else {
                _this.transitionTo(_this._noState);
            }
        };
        return _this;
    }
    return PickNextLocationState;
}(YesNoState));
exports.PickNextLocationState = PickNextLocationState;
var AwaitFaceState = (function (_super) {
    __extends(AwaitFaceState, _super);
    function AwaitFaceState(parent, name, timeOut) {
        if (timeOut === void 0) { timeOut = -1; }
        var _this = _super.call(this, parent.sm, name) || this;
        _this.parent = parent;
        _this.timeOut = timeOut;
        _this.onEntry = _this._onEntry.bind(_this);
        _this.onExit = function () {
            if (_this._faceHandle) {
                _this._faceHandle.cancel();
                _this._faceHandle = null;
            }
        };
        return _this;
    }
    AwaitFaceState.prototype.installTransitions = function (faceFoundState, timeoutState, motionDetectedState) {
        this._faceFoundState = faceFoundState;
        this._timeoutState = timeoutState;
        this._motionDetectedState = motionDetectedState;
        this.addInternalTransition('Face found', this._faceFoundState);
        this.addInternalTransition('Timeout', this._timeoutState);
        this.addInternalTransition('Motion detected', this._motionDetectedState);
    };
    AwaitFaceState.prototype.createMotionDetectionPromise = function () {
        return new Promise(function (resolve) {
        });
    };
    AwaitFaceState.prototype._onEntry = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var jibo, _a, promises, outcome, error_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        jibo = this.parent.parent.jibo;
                        _a = this;
                        return [4, jibo.expression.awaitFace({ timeout: this.timeOut })];
                    case 1:
                        _a._faceHandle = _b.sent();
                        promises = [
                            this._faceHandle.promise.then(function (face) {
                                if (face && face.status === jibo_common_types_1.ResultStatus.SUCCEEDED) {
                                    return { outcome: Outcome.Face, data: face };
                                }
                                else {
                                    return { outcome: Outcome.Timeout };
                                }
                            }).catch(function (e) {
                                _this.parent.log.error("AwaitFace: ", e);
                            }).then(function (data) {
                                _this._faceHandle = null;
                                return data;
                            }),
                            this.createMotionDetectionPromise()
                        ];
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 4, , 5]);
                        return [4, Promise.race(promises)];
                    case 3:
                        outcome = _b.sent();
                        if (outcome.outcome === Outcome.Face) {
                            this.transitionTo(this._faceFoundState, outcome.data);
                        }
                        else if (outcome.outcome === Outcome.Timeout) {
                            this.transitionTo(this._timeoutState);
                        }
                        else if (outcome.outcome === Outcome.Motion) {
                            this.transitionTo(this._motionDetectedState, outcome.data);
                        }
                        else {
                            this.parent.log.error("Await face, invalid outcome: ", outcome);
                            this.transitionTo(this._timeoutState);
                        }
                        return [3, 5];
                    case 4:
                        error_1 = _b.sent();
                        this.parent.log.error("Await face, error: ", error_1);
                        this.transitionTo(this._timeoutState);
                        return [3, 5];
                    case 5: return [2];
                }
            });
        });
    };
    return AwaitFaceState;
}(jibo_state_machine_1.State));
exports.AwaitFaceState = AwaitFaceState;
var OrientToStandard = (function (_super) {
    __extends(OrientToStandard, _super);
    function OrientToStandard(sm, actionRuntime) {
        var _this = _super.call(this, sm, 'Orient to standard location') || this;
        _this.onEntry = function () {
            return actionRuntime.jibo.expression.centerRobot({
                requestor: OWNER_IDENTIFIER,
                centerGlobally: true
            });
        };
        return _this;
    }
    return OrientToStandard;
}(jibo_state_machine_1.State));
exports.OrientToStandard = OrientToStandard;
var AttendToFixedState = (function (_super) {
    __extends(AttendToFixedState, _super);
    function AttendToFixedState(sm, actionRuntime, name, standard) {
        var _this = _super.call(this, sm, actionRuntime, name) || this;
        _this.standard = standard;
        return _this;
    }
    AttendToFixedState.prototype.overrideLocation = function (data) {
        return this.standard;
    };
    return AttendToFixedState;
}(AttendToState));
exports.AttendToFixedState = AttendToFixedState;

},{"../log":16,"jibo-cai-utils":undefined,"jibo-common-types":undefined,"jibo-state-machine":undefined}],12:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var Types_1 = require("../../common/Types");
var SSWAction = (function (_super) {
    __extends(SSWAction, _super);
    function SSWAction(parent, skillName, skillData) {
        var _this = _super.call(this, parent, "Skill Switch Action: '" + skillName + "'") || this;
        _this.skillName = skillName;
        _this.skillData = skillData;
        return _this;
    }
    SSWAction.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var status_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, new Promise(function (resolve) { return process.nextTick(resolve); })];
                    case 1:
                        _a.sent();
                        if (!!this.parent.skillSwitcher) return [3, 2];
                        this.log.error("No installed skill switcher");
                        return [2, Types_1.ActionResult.FAILED];
                    case 2: return [4, this.parent.skillSwitcher(this.skillName, this.skillData)];
                    case 3:
                        status_1 = _a.sent();
                        switch (status_1) {
                            case Types_1.Status.SUCCEEDED:
                                return [2, Types_1.ActionResult.SUCCEEDED];
                            case Types_1.Status.FAILED:
                                return [2, Types_1.ActionResult.FAILED];
                            default: {
                                this.log.error("Unknown Skill Switch outcome: '" + status_1 + "'");
                                return [2, Types_1.ActionResult.FAILED];
                            }
                        }
                        _a.label = 4;
                    case 4: return [2];
                }
            });
        });
    };
    SSWAction.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Promise.resolve()];
            });
        });
    };
    return SSWAction;
}(Action_1.Action));
exports.SSWAction = SSWAction;

},{"../../common/Types":26,"../Action":1}],13:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Action_1 = require("../Action");
var jibo_state_machine_1 = require("jibo-state-machine");
var Types_1 = require("../../common/Types");
var classes = require("./LookatClasses");
var jibo_common_types_1 = require("jibo-common-types");
var AWAIT_FACE_TIMEOUT = 3;
var Flows = {
    React1: null,
    React2: null,
    React3: null,
    React4: null,
    React5: null,
};
var SearchAction = (function (_super) {
    __extends(SearchAction, _super);
    function SearchAction(parent, initialRefPosition) {
        var _this = _super.call(this, parent, 'Search Action') || this;
        _this.initialRefPosition = initialRefPosition;
        _this.sm = new jibo_state_machine_1.StateMachine();
        _this.searchCandidates = [];
        _this.isPersonVisible = new classes.PersonVisibleNearReferenceState(_this.sm, _this.parent, classes.AngleCalculatorType.XYZ);
        _this.isLightingGood = new classes.LightingGoodState(_this.sm, _this.parent);
        _this.attendToCandLoc = new classes.AttendToState(_this.sm, _this.parent, 'Attend to candidate location', false);
        _this.attendToMotion = new classes.AttendToState(_this.sm, _this.parent, 'Attend to motion');
        _this.reactNoCandidates = new classes.FlowState(_this.sm, _this.parent, 'React to no more search candidates', Flows.React4);
        _this.pickNextSearchCandidate = new classes.PickNextLocationState(_this);
        _this.reactMotionDetected = new classes.FlowState(_this.sm, _this.parent, 'React to motion detected', Flows.React5);
        _this.reactFaceFound = new classes.FlowState(_this.sm, _this.parent, 'React to face found', Flows.React3);
        _this.reactFaceNotFound = new classes.FlowState(_this.sm, _this.parent, 'React to face not found', Flows.React2);
        _this.awaitFace = new classes.AwaitFaceState(_this, 'Await face', AWAIT_FACE_TIMEOUT);
        _this.reactBadLighting = new classes.FlowState(_this.sm, _this.parent, 'React to bad lighting', Flows.React1);
        _this.attendToFace = new classes.AttendToState(_this.sm, _this.parent, 'Attend to face found');
        _this.doneSucceeded = new classes.DoneState(_this.sm, 'Done - Success', Types_1.ActionResult.SUCCEEDED);
        _this.doneFailed = new classes.DoneState(_this.sm, 'Done - Failure', Types_1.ActionResult.FAILED);
        _this.sm.setInitial(_this.isPersonVisible);
        _this.isPersonVisible.installTransitions(_this.attendToFace, _this.pickNextSearchCandidate);
        _this.attendToFace.addDoneTransition(_this.doneSucceeded);
        _this.pickNextSearchCandidate.installTransitions(_this.attendToCandLoc, _this.reactNoCandidates);
        _this.reactNoCandidates.addDoneTransition(_this.doneFailed);
        _this.attendToCandLoc.addDoneTransition(_this.isLightingGood);
        _this.isLightingGood.installTransitions(_this.awaitFace, _this.reactBadLighting);
        _this.reactBadLighting.addDoneTransition(_this.pickNextSearchCandidate);
        _this.awaitFace.installTransitions(_this.reactFaceFound, _this.reactFaceNotFound, _this.reactMotionDetected);
        _this.reactMotionDetected.addDoneTransition(_this.attendToMotion);
        _this.attendToMotion.addDoneTransition(_this.doneSucceeded);
        _this.reactFaceNotFound.addDoneTransition(_this.pickNextSearchCandidate);
        _this.reactFaceFound.addDoneTransition(_this.attendToFace);
        return _this;
    }
    SearchAction.prototype.runInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var pmHandler, refPosition, features, down, res, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.parent.jibo.expression.pushAttentionMode(jibo_common_types_1.AttentionMode.COMMAND)];
                    case 1:
                        pmHandler = _a.sent();
                        refPosition = this.initialRefPosition;
                        if (!refPosition) {
                            features = this.parent.jibo.expression.features;
                            refPosition = features.head.direction.clone().add(features.head.position);
                        }
                        this.isPersonVisible.setReferencePosition(refPosition);
                        this.searchCandidates = [];
                        down = { x: refPosition.x, y: refPosition.y, z: 0.2 };
                        this.searchCandidates.push({ position: down });
                        this.searchCandidates.push({ position: refPosition });
                        this.log.debug("Searching locations: ", this.searchCandidates);
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, 5, 7]);
                        return [4, this.sm.start()];
                    case 3:
                        res = _a.sent();
                        return [2, res ? res : Types_1.ActionResult.FAILED];
                    case 4:
                        e_1 = _a.sent();
                        this.log.error('Search action state machine error: ', e_1);
                        return [2, Types_1.ActionResult.FAILED];
                    case 5:
                        this.log.debug("'" + this.name + "' SM trace: ", this.sm.traceToString());
                        return [4, pmHandler.release()];
                    case 6:
                        _a.sent();
                        return [7];
                    case 7: return [2];
                }
            });
        });
    };
    SearchAction.prototype.cancelInternal = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.log.info("'" + this.name + "' canceled");
                this.sm.stop();
                return [2];
            });
        });
    };
    return SearchAction;
}(Action_1.Action));
exports.SearchAction = SearchAction;

},{"../../common/Types":26,"../Action":1,"./LookatClasses":11,"jibo-common-types":undefined,"jibo-state-machine":undefined}],14:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./SSWAction"));
__export(require("./AnimationAction"));
__export(require("./IdleAnimationActions"));
__export(require("./BasicLookat"));
__export(require("./SearchAction"));
__export(require("./HJOrient"));
__export(require("./BeSkillSwitchAction"));
__export(require("./HeadTouch"));
__export(require("./EmitHeadPat"));

},{"./AnimationAction":4,"./BasicLookat":5,"./BeSkillSwitchAction":6,"./EmitHeadPat":7,"./HJOrient":8,"./HeadTouch":9,"./IdleAnimationActions":10,"./SSWAction":12,"./SearchAction":13}],15:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./Action"));
__export(require("./ActionSystem"));
__export(require("./GoalDrivenAction"));
var actions = require("./actions");
exports.actions = actions;
var policies = require("./policies");
exports.policies = policies;

},{"./Action":1,"./ActionSystem":2,"./GoalDrivenAction":3,"./actions":14,"./policies":18}],16:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var log_1 = require("../common/log");
var log = log_1.log.createChild('Action');
exports.log = log;

},{"../common/log":30}],17:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HJGoal_1 = require("../../goal/goals/HJGoal");
var FindPersonGoal_1 = require("../../goal/goals/FindPersonGoal");
var HJOnlyGoal_1 = require("../../goal/goals/HJOnlyGoal");
var BasicLookat_1 = require("../actions/BasicLookat");
var SearchAction_1 = require("../actions/SearchAction");
var HJOrient_1 = require("../actions/HJOrient");
var Utils_1 = require("../../common/Utils");
var SSWGoal_1 = require("../../goal/goals/SSWGoal");
var SSWAction_1 = require("../actions/SSWAction");
var AnimationAction_1 = require("../actions/AnimationAction");
var AnimationGoal_1 = require("../../goal/goals/AnimationGoal");
var HeadTouchGoal_1 = require("../../goal/goals/HeadTouchGoal");
var HeadTouch_1 = require("../actions/HeadTouch");
var EmitHeadPat_1 = require("../actions/EmitHeadPat");
var BeSSWGoal_1 = require("../../goal/goals/BeSSWGoal");
var BeSkillSwitchAction_1 = require("../../action/actions/BeSkillSwitchAction");
var BasicPolicy = (function () {
    function BasicPolicy(parent) {
        this.parent = parent;
    }
    BasicPolicy.prototype.selectAction = function (goal, state) {
        if (goal instanceof HJGoal_1.HJGoal) {
            var hjGoal = goal;
            var pos = hjGoal.hjData.primary.position;
            if (state.skillName === Utils_1.SkillNames.IDLE) {
                return new HJOrient_1.HJOrient(this.parent, pos);
            }
            else {
                return new BasicLookat_1.BasicLookat(this.parent, pos);
            }
        }
        else if (goal instanceof BeSSWGoal_1.BeSSWGoal) {
            return new BeSkillSwitchAction_1.BeSkillSwitchAction(this.parent, goal.data);
        }
        else if (goal instanceof HJOnlyGoal_1.HJOnlyGoal) {
            var data = {
                NLParse: {
                    skill: Utils_1.SkillNames.GREETINGS,
                    intent: 'heyJibo'
                }
            };
            return new SSWAction_1.SSWAction(this.parent, Utils_1.SkillNames.GREETINGS, data);
        }
        else if (goal instanceof SSWGoal_1.SSWGoal) {
            var sswGoal = goal;
            return new SSWAction_1.SSWAction(this.parent, sswGoal.data.skillName, sswGoal.data.skillOptions);
        }
        else if (goal instanceof AnimationGoal_1.AnimationGoal) {
            var animGoal = goal;
            return new AnimationAction_1.AnimationAction(this.parent, animGoal.data);
        }
        else if (goal instanceof FindPersonGoal_1.FindPersonGoal) {
            return new SearchAction_1.SearchAction(this.parent);
        }
        else if (goal instanceof HeadTouchGoal_1.HeadTouchGoal) {
            var headTouchGoal = goal;
            if (state.skillName === Utils_1.SkillNames.IDLE) {
                return new EmitHeadPat_1.EmitHeadPat(this.parent, headTouchGoal.data);
            }
            else {
                return new HeadTouch_1.HeadTouch(this.parent);
            }
        }
    };
    return BasicPolicy;
}());
exports.BasicPolicy = BasicPolicy;

},{"../../action/actions/BeSkillSwitchAction":6,"../../common/Utils":28,"../../goal/goals/AnimationGoal":48,"../../goal/goals/BeSSWGoal":49,"../../goal/goals/FindPersonGoal":50,"../../goal/goals/HJGoal":51,"../../goal/goals/HJOnlyGoal":52,"../../goal/goals/HeadTouchGoal":53,"../../goal/goals/SSWGoal":55,"../actions/AnimationAction":4,"../actions/BasicLookat":5,"../actions/EmitHeadPat":7,"../actions/HJOrient":8,"../actions/HeadTouch":9,"../actions/SSWAction":12,"../actions/SearchAction":13}],18:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./BasicPolicy"));

},{"./BasicPolicy":17}],19:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ActionRuntime_1 = require("./common/ActionRuntime");
var ActionEvents_1 = require("./common/ActionEvents");
var action_ns = require("./action");
var types = require("./common/Types");
exports.types = types;
var goal_ns = require("./goal");
var goal;
(function (goal) {
    goal.GoalEvents = goal_ns.GoalEvents;
    goal.Goal = goal_ns.Goal;
})(goal = exports.goal || (exports.goal = {}));
var action;
(function (action) {
    action.Action = action_ns.Action;
})(action = exports.action || (exports.action = {}));
function init(options) {
    exports._runtime = new ActionRuntime_1.ActionRuntime();
    exports.events = new ActionEvents_1.PublicActionEvents(exports._runtime.events);
    return exports._runtime.init(options);
}
exports.init = init;
function getState() {
    var state = exports._runtime.createState();
    return {
        timeOfDay: state.timeOfDay,
        sleep: state.sleep,
        time: state.time,
    };
}
exports.getState = getState;
function setCurrentSkill(skillName) {
    exports._runtime.currentSkillName = skillName;
}
exports.setCurrentSkill = setCurrentSkill;
function setSkillSwitchHandler(skillSwitcher) {
    exports._runtime.skillSwitcher = skillSwitcher;
}
exports.setSkillSwitchHandler = setSkillSwitchHandler;
function getActiveGoalActions() {
    return exports._runtime.activeGoals.map(function (entry) {
        return [entry.goal, entry.action];
    });
}
exports.getActiveGoalActions = getActiveGoalActions;
function addSkillSwitchGoal(data) {
    var goal = new goal_ns.goals.SSWGoal(exports._runtime, data);
    process.nextTick(function () { return exports._runtime.goals.addGoal(goal); });
    return goal;
}
exports.addSkillSwitchGoal = addSkillSwitchGoal;
function addBeSkillSwitchGoal(data) {
    var goal = new goal_ns.goals.BeSSWGoal(exports._runtime, data);
    process.nextTick(function () { return exports._runtime.goals.addGoal(goal); });
    return goal;
}
exports.addBeSkillSwitchGoal = addBeSkillSwitchGoal;
function addPlayAnimationGoal(data) {
    var goal = new goal_ns.goals.AnimationGoal(exports._runtime, data);
    process.nextTick(function () { return exports._runtime.goals.addGoal(goal); });
    return goal;
}
exports.addPlayAnimationGoal = addPlayAnimationGoal;
function addFindPersonGoal() {
    var goal = new goal_ns.goals.FindPersonGoal(exports._runtime);
    process.nextTick(function () { return exports._runtime.goals.addGoal(goal); });
    return goal;
}
exports.addFindPersonGoal = addFindPersonGoal;
function getMotivationalDriveValue(drive) {
    return exports._runtime.motivations.drives.get(drive).value;
}
exports.getMotivationalDriveValue = getMotivationalDriveValue;
function applyMotivationalEffect(drive, effect) {
    exports._runtime.motivations.drives.get(drive).applyEffect(effect);
}
exports.applyMotivationalEffect = applyMotivationalEffect;
function configure(configOptions) {
    exports._runtime.configure(configOptions);
}
exports.configure = configure;

},{"./action":15,"./common/ActionEvents":20,"./common/ActionRuntime":21,"./common/Types":26,"./goal":57}],20:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_typed_events_1 = require("jibo-typed-events");
var ActionEvents = (function (_super) {
    __extends(ActionEvents, _super);
    function ActionEvents() {
        var _this = _super.call(this) || this;
        _this.oriented = new jibo_typed_events_1.Event("Jibo has oriented towards target");
        _this.goalAdded = new jibo_typed_events_1.Event('Goal Created');
        _this.headPat = new jibo_typed_events_1.Event("Jibo's head was touched in non-exit");
        _this.skillChange = new jibo_typed_events_1.Event("Skill changed");
        _this.sleepChange = new jibo_typed_events_1.Event("Sleep state changed");
        return _this;
    }
    return ActionEvents;
}(jibo_typed_events_1.EventContainer));
exports.ActionEvents = ActionEvents;
var PublicActionEvents = (function (_super) {
    __extends(PublicActionEvents, _super);
    function PublicActionEvents(events) {
        var _this = _super.call(this) || this;
        _this.events = events;
        _this.headPat = _this.events.headPat;
        _this.sleepChange = _this.events.sleepChange;
        return _this;
    }
    return PublicActionEvents;
}(jibo_typed_events_1.EventContainer));
exports.PublicActionEvents = PublicActionEvents;

},{"jibo-typed-events":undefined}],21:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var GoalSet_1 = require("../goal/GoalSet");
var Memory_1 = require("./Memory");
var motivation_1 = require("../motivation");
var ActionSystem_1 = require("../action/ActionSystem");
var GoalDrivenAction_1 = require("../action/GoalDrivenAction");
var Types_1 = require("../common/Types");
var _1 = require(".");
var goalproviders_1 = require("../goal/goalproviders");
var DateProvider_1 = require("./DateProvider");
var Types_2 = require("./Types");
var IdleAnimationActions_1 = require("../action/actions/IdleAnimationActions");
var TunableDebug_1 = require("./TunableDebug");
var HeadTouchGoalProvider_1 = require("../goal/goalproviders/HeadTouchGoalProvider");
var log_1 = require("./log");
var ActionRuntime = (function () {
    function ActionRuntime() {
        var _this = this;
        this.goals = new GoalSet_1.GoalSet(this);
        this.actions = new ActionSystem_1.ActionSystem(this);
        this.motivations = new motivation_1.MotivationSystem(this);
        this.events = new _1.ActionEvents();
        this.activeGoals = [];
        this.dateProvider = new DateProvider_1.DefaultDateProvider();
        this.memory = new Memory_1.Memory(this);
        this.goalProviders = [];
        this._skillStartTime = this.dateProvider.getDate();
        this._updater = new _1.Updater();
        this.updateAll = function () { return _this._updater.update(); };
    }
    ActionRuntime.prototype.addGoalProvider = function (goalProvider) {
        this.goalProviders.push(goalProvider);
    };
    ActionRuntime.prototype.init = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this._options = options;
                if (options.autoUpdate === undefined) {
                    options.autoUpdate = true;
                }
                this.jibo = options.jibo;
                this.circadianGoalProvider = new goalproviders_1.CircadianGoalProvider(this);
                this.addGoalProvider(this.circadianGoalProvider);
                this.hjGoalProvider = new goalproviders_1.HJGoalProvider(this);
                this.addGoalProvider(this.hjGoalProvider);
                this.addGoalProvider(new goalproviders_1.HJNoMatchGoalProvider(this));
                this.addGoalProvider(new goalproviders_1.HJOnlyGoalProvider(this));
                this.addGoalProvider(new goalproviders_1.ProactiveGreetingGoalProvider(this));
                this.addGoalProvider(new HeadTouchGoalProvider_1.HeadTouchGoalProvider(this));
                this.goalProviders.forEach(function (gp) { return gp.init(); });
                this._updater.addUpdateable(function () { return _this.motivations.update(); }, 1000);
                this._updater.addUpdateable(function () { return _this.update(); }, 1000);
                this._updater.addUpdateable(function () { return _this.circadianGoalProvider.update(); }, 3000);
                this.events.goalAdded.on(function (goal) {
                    _this.handleIncomingGoal(goal);
                });
                this.actions.addAction(new IdleAnimationActions_1.IdleAnimationVisual(this));
                this.actions.addAction(new IdleAnimationActions_1.IdleAnimationBody(this));
                if (this._options.autoUpdate) {
                    this.updateAll.isGlobalTimer = true;
                    this.jibo.timer.on('update', this.updateAll);
                }
                TunableDebug_1.TunableDebug.setup(this);
                return [2];
            });
        });
    };
    ActionRuntime.prototype.dispose = function () {
        if (this._options && this._options.autoUpdate) {
            this.jibo.timer.off('update', this.updateAll);
        }
        this.goalProviders.forEach(function (gp) { return gp.dispose(); });
        this.goalProviders = [];
        this._updater.clear();
    };
    ActionRuntime.prototype.configure = function (configOptions) {
        if (configOptions.hasOwnProperty('orientToHJ')) {
            this.hjGoalProvider.setListening(configOptions.orientToHJ);
        }
    };
    Object.defineProperty(ActionRuntime.prototype, "currentSkillName", {
        get: function () {
            return this._currentSkillName;
        },
        set: function (name) {
            var oldSkillName = this._currentSkillName;
            this._currentSkillName = name;
            if (oldSkillName !== name) {
                this._skillStartTime = this.dateProvider.getDate();
                this.events.skillChange.emit(this._currentSkillName);
            }
        },
        enumerable: true,
        configurable: true
    });
    ActionRuntime.prototype.createState = function () {
        var now = this.dateProvider.getDate();
        return {
            time: now,
            users: this.jibo.lps.identity.getPresentPersons(),
            skillName: this.currentSkillName,
            skillStartTime: this._skillStartTime,
            timeOfDay: Types_2.TimeOfDay.getTimeOfDay(now),
            sleep: this.circadianGoalProvider.getCurrentSleepState()
        };
    };
    ActionRuntime.prototype.update = function () {
        if (!this.activeGoals.length) {
            this.pursueNextGoal();
        }
    };
    ActionRuntime.prototype.handleIncomingGoal = function (goal) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var state, i, entry, shouldBeCanceled, goalNames;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        log_1.log.info("Incoming new goal '" + goal.name + "'");
                        state = this.createState();
                        if (!goal.isEligible(state)) {
                            log_1.log.info("not currently eligible, in queue");
                            return [2];
                        }
                        if (!!this.activeGoals.length) return [3, 1];
                        this.pursueNextGoal();
                        return [3, 5];
                    case 1:
                        for (i = 0; i < this.activeGoals.length; i++) {
                            entry = this.activeGoals[i];
                            if ((entry.goal.options.priority >= goal.options.priority) &&
                                !this.goals.goalParallelism.canBeAchievedInParallel(goal, entry.goal)) {
                                log_1.log.info("incoming goal is of lower priority and can not be performed in parallel with existing goal: '" + entry.goal.name + "', is queued");
                                return [2];
                            }
                        }
                        shouldBeCanceled = this.activeGoals.filter(function (entry) {
                            return ((entry.goal.options.priority < goal.options.priority) &&
                                !_this.goals.goalParallelism.canBeAchievedInParallel(goal, entry.goal));
                        });
                        if (!shouldBeCanceled.length) return [3, 3];
                        goalNames = shouldBeCanceled.map(function (e) { return e.goal.name; }).join(', ');
                        log_1.log.info("Canceling following lower priority goals in favor of incoming one: '" + goalNames + "'");
                        return [4, Promise.all(shouldBeCanceled.map(function (e) { return _this.cancelGoalDrivenAction(e); }))];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [4, this.achieveGoal(state, goal)];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5: return [2];
                }
            });
        });
    };
    ActionRuntime.prototype.pursueNextGoal = function () {
        return __awaiter(this, void 0, void 0, function () {
            var state, goal;
            return __generator(this, function (_a) {
                state = this.createState();
                goal = this.goals.chooseGoal(state);
                if (goal) {
                    log_1.log.debug("Selected goal to pursue '" + goal.name + "'");
                    return [2, this.achieveGoal(state, goal)];
                }
                return [2];
            });
        });
    };
    ActionRuntime.prototype.cancelGoalDrivenAction = function (current) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        current.goal.setFinishedStatus(Types_1.GoalFinishedStatus.CANCELED);
                        return [4, current.action.cancel()];
                    case 1:
                        _a.sent();
                        if (!current.currentRunPromise) return [3, 3];
                        return [4, current.currentRunPromise];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2];
                }
            });
        });
    };
    ActionRuntime.prototype.achieveGoal = function (state, newGoal) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var action, goalDrivenAction_1, postAction_1;
            return __generator(this, function (_a) {
                action = this.actions.chooseAction(state, newGoal);
                if (action) {
                    newGoal.setProgressStatus(Types_1.GoalProgressStatus.SELECTED);
                    goalDrivenAction_1 = new GoalDrivenAction_1.GoalDrivenAction(newGoal, action, this);
                    this.activeGoals.push(goalDrivenAction_1);
                    this.activeGoals.sort(function (a, b) { return a.goal.options.priority - b.goal.options.priority; });
                    postAction_1 = function () {
                        var ind = _this.activeGoals.indexOf(goalDrivenAction_1);
                        if (ind >= 0) {
                            _this.activeGoals.splice(ind, 1);
                        }
                        process.nextTick(function () {
                            _this.pursueNextGoal();
                        });
                    };
                    return [2, goalDrivenAction_1.run()
                            .then(function (result) {
                            postAction_1();
                            return result;
                        }).catch(function (e) {
                            postAction_1();
                            throw e;
                        })];
                }
                else {
                    newGoal.setProgressStatus(Types_1.GoalProgressStatus.INITIAL);
                    return [2, null];
                }
                return [2];
            });
        });
    };
    return ActionRuntime;
}());
exports.ActionRuntime = ActionRuntime;

},{".":29,"../action/ActionSystem":2,"../action/GoalDrivenAction":3,"../action/actions/IdleAnimationActions":10,"../common/Types":26,"../goal/GoalSet":35,"../goal/goalproviders":46,"../goal/goalproviders/HeadTouchGoalProvider":40,"../motivation":63,"./DateProvider":22,"./Memory":24,"./TunableDebug":25,"./Types":26,"./log":30}],22:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DefaultDateProvider = (function () {
    function DefaultDateProvider() {
    }
    DefaultDateProvider.prototype.getDate = function () {
        return new Date();
    };
    return DefaultDateProvider;
}());
exports.DefaultDateProvider = DefaultDateProvider;

},{}],23:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function create(predicate, addListener, removeListener, timeout) {
    if (timeout === void 0) { timeout = -1; }
    var done = false;
    var eventHandler;
    var timeoutHandle = null;
    return {
        promise: new Promise(function (resolve, reject) {
            if (timeout > 0) {
                timeoutHandle = setTimeout(function () {
                    if (!done) {
                        done = true;
                        removeListener(eventHandler);
                        timeoutHandle = null;
                        reject('TIMEOUT');
                    }
                }, timeout);
            }
            eventHandler = function (data) {
                if (!predicate || predicate(data)) {
                    if (!done) {
                        done = true;
                        removeListener(eventHandler);
                        if (timeoutHandle) {
                            clearTimeout(timeoutHandle);
                        }
                        resolve(data);
                    }
                }
            };
            addListener(eventHandler);
        }),
        stop: function () {
            if (!done) {
                done = true;
                removeListener(eventHandler);
                if (timeoutHandle) {
                    clearTimeout(timeoutHandle);
                }
            }
        }
    };
}
var EventWaiter = (function () {
    function EventWaiter(predicate, eventName, eventEmitter, timeout) {
        if (timeout === void 0) { timeout = -1; }
        this.predicate = predicate;
        this.eventName = eventName;
        this.eventEmitter = eventEmitter;
        this.timeout = timeout;
    }
    EventWaiter.prototype.start = function () {
        var _this = this;
        return create(this.predicate, function (handler) { return _this.eventEmitter.on(_this.eventName, handler); }, function (handler) { return _this.eventEmitter.removeListener(_this.eventName, handler); }, this.timeout);
    };
    return EventWaiter;
}());
exports.EventWaiter = EventWaiter;
var TypedEventWaiter = (function () {
    function TypedEventWaiter(predicate, event, timeout) {
        if (timeout === void 0) { timeout = -1; }
        this.predicate = predicate;
        this.event = event;
        this.timeout = timeout;
    }
    TypedEventWaiter.prototype.start = function () {
        var _this = this;
        return create(this.predicate, function (handler) { return _this.event.on(handler); }, function (handler) { return _this.event.removeListener(handler); }, this.timeout);
    };
    return TypedEventWaiter;
}());
exports.TypedEventWaiter = TypedEventWaiter;

},{}],24:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Memory = (function () {
    function Memory(parent) {
        this.parent = parent;
        this.map = new Map();
    }
    Memory.prototype.getHJData = function () {
        return this.get('HJData');
    };
    Memory.prototype.setHJData = function (data) {
        this.set('HJData', data);
    };
    Memory.prototype.set = function (key, data) {
        this.map.set(key, {
            time: this.parent.dateProvider.getDate(),
            data: data
        });
    };
    Memory.prototype.get = function (key) {
        return this.map.get(key);
    };
    return Memory;
}());
exports.Memory = Memory;

},{}],25:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Types_1 = require("./Types");
var FindPersonGoal_1 = require("../goal/goals/FindPersonGoal");
var ProactiveGreetingGoalProvider_1 = require("../goal/goalproviders/ProactiveGreetingGoalProvider");
var log_1 = require("./log");
var log = log_1.log.createChild('Tunable');
var tunable = null;
try {
    tunable = require('jibo-tunable');
}
catch (e) {
}
var FakeDateProvider = (function () {
    function FakeDateProvider(hour, minute) {
        this.hour = hour;
        this.minute = minute;
    }
    FakeDateProvider.prototype.getDate = function () {
        var d = new Date();
        d.setHours(this.hour);
        d.setMinutes(this.minute);
        d.setSeconds(0);
        d.setMilliseconds(0);
        return d;
    };
    return FakeDateProvider;
}());
var WIN = 'Action System';
var TunableDebug = (function () {
    function TunableDebug() {
    }
    TunableDebug.setup = function (runtime) {
        if (tunable) {
            var T = tunable.Tunable;
            T.getButtonField("Log status report", WIN).events.change.on(function () {
                var state = runtime.createState();
                log.info("Action System status:");
                log.info("State: \n" + JSON.stringify(state, null, 4));
                log.info("Active goals/actions (" + runtime.activeGoals.length + ")");
                runtime.activeGoals.forEach(function (g) {
                    log.info(" - goal: '" + g.goal.name + "', action: '" + g.action.name + "'");
                });
            });
            T.getButtonField("Force pro-active greeting", WIN).events.change.on(function () {
                runtime.goalProviders.forEach(function (gp) {
                    if (gp instanceof ProactiveGreetingGoalProvider_1.ProactiveGreetingGoalProvider) {
                        gp.handleVisibleFaceInternal();
                    }
                });
            });
            var originalDateProvider_1 = runtime.dateProvider;
            T.getDropdownField('Fake time', ['disabled', '5:30 am', '7:00 am', '2:00 pm', '10:00 pm', '10:10 pm', '11:00 pm'], 0, WIN).events.change.on(function (data) {
                if (data === 'disabled') {
                    runtime.dateProvider = originalDateProvider_1;
                }
                else {
                    var _a = data.split(' '), time = _a[0], am_pm = _a[1];
                    var _b = time.split(':'), hour = _b[0], minute = _b[1];
                    runtime.dateProvider = new FakeDateProvider((am_pm === 'pm') ? Number(hour) + 12 : Number(hour), Number(minute));
                }
                runtime.circadianGoalProvider.update();
            });
            T.getButtonField("FallingAsleep: fake long time", WIN).events.change.on(function () {
                runtime.circadianGoalProvider.circadianSM.fallingAsleepState.startTime = -1;
            });
            T.getButtonField("Clear current goal / action", WIN).events.change.on(function () {
                if (!runtime.activeGoals.length) {
                    log.warn("No current goal/action to clear");
                }
                else {
                    var name_1 = runtime.activeGoals.map(function (a) { return a.action.name; }).join(', ');
                    log.warn("Canceling current action: " + name_1);
                    runtime.activeGoals.forEach(function (a) { return a.action.cancel(); });
                }
            });
            T.getButtonField("Create search goal", WIN).events.change.on(function () {
                runtime.goals.addGoal(new FindPersonGoal_1.FindPersonGoal(runtime));
            });
            var driveDropdown_1 = T.getDropdownField('Motivational drive', Object.keys(Types_1.DriveName), 0, WIN);
            T.getButtonField("Saturate drive", WIN).events.change.on(function () {
                var drive = runtime.motivations.drives.get(driveDropdown_1.current);
                drive._value = drive.threshold;
            });
        }
    };
    return TunableDebug;
}());
exports.TunableDebug = TunableDebug;

},{"../goal/goalproviders/ProactiveGreetingGoalProvider":41,"../goal/goals/FindPersonGoal":50,"./Types":26,"./log":30,"jibo-tunable":undefined}],26:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EARLY_MORNING_HOUR = 5;
var MID_MORNING_HOUR = 7;
var DAY_HOUR = 11;
var NIGHT_HOUR = 22;
var NIGHT_MINUTE = 15;
var GoalFinishedStatus;
(function (GoalFinishedStatus) {
    GoalFinishedStatus.NOT_FINISHED = 'NOT_FINISHED';
    GoalFinishedStatus.SUCCEEDED = 'SUCCEEDED';
    GoalFinishedStatus.FAILED = 'FAILED';
    GoalFinishedStatus.CANCELED = 'CANCELED';
    GoalFinishedStatus.CULLED = 'CULLED';
})(GoalFinishedStatus = exports.GoalFinishedStatus || (exports.GoalFinishedStatus = {}));
var GoalProgressStatus;
(function (GoalProgressStatus) {
    GoalProgressStatus.INITIAL = 'INITIAL';
    GoalProgressStatus.SELECTED = 'SELECTED';
    GoalProgressStatus.FINISHED = 'FINISHED';
    GoalProgressStatus.ATTEMPTED = 'ATTEMPTED';
})(GoalProgressStatus = exports.GoalProgressStatus || (exports.GoalProgressStatus = {}));
var ActionResult;
(function (ActionResult) {
    ActionResult.SUCCEEDED = 'SUCCEEDED';
    ActionResult.FAILED = 'FAILED';
    ActionResult.CANCELED = 'CANCELED';
    ActionResult.TIMEOUT = 'TIMEOUT';
})(ActionResult = exports.ActionResult || (exports.ActionResult = {}));
var ActionStatus;
(function (ActionStatus) {
    ActionStatus.RUNNING = 'RUNNING';
    ActionStatus.NOT_RUNNING = 'NOT_RUNNING';
})(ActionStatus = exports.ActionStatus || (exports.ActionStatus = {}));
var SleepState;
(function (SleepState) {
    SleepState.ASLEEP = 'ASLEEP';
    SleepState.AWAKE = 'AWAKE';
    SleepState.FALLING_ASLEEP = 'FALLING_ASLEEP';
})(SleepState = exports.SleepState || (exports.SleepState = {}));
var Status;
(function (Status) {
    Status.SUCCEEDED = 'SUCCEEDED';
    Status.FAILED = 'FAILED';
})(Status = exports.Status || (exports.Status = {}));
var TimeOfDay;
(function (TimeOfDay) {
    TimeOfDay.MORNING_EARLY = 'MORNING_EARLY';
    TimeOfDay.MORNING = 'MORNING';
    TimeOfDay.DAY = 'DAY';
    TimeOfDay.NIGHT = 'NIGHT';
    TimeOfDay.NIGHT_EARLY = 'NIGHT_EARLY';
    function getTimeOfDay(time) {
        var hour = time.getHours();
        var min = time.getMinutes();
        if (hour < EARLY_MORNING_HOUR || hour >= NIGHT_HOUR) {
            if (hour === NIGHT_HOUR && min <= NIGHT_MINUTE) {
                return TimeOfDay.NIGHT_EARLY;
            }
            else {
                return TimeOfDay.NIGHT;
            }
        }
        else if (hour < MID_MORNING_HOUR) {
            return TimeOfDay.MORNING_EARLY;
        }
        else if (hour < DAY_HOUR) {
            return TimeOfDay.MORNING;
        }
        else if (hour < NIGHT_HOUR) {
            return TimeOfDay.DAY;
        }
    }
    TimeOfDay.getTimeOfDay = getTimeOfDay;
})(TimeOfDay = exports.TimeOfDay || (exports.TimeOfDay = {}));
var GoalPriorityType;
(function (GoalPriorityType) {
    GoalPriorityType.REGULAR = 'REGULAR';
    GoalPriorityType.SINGLETON_KEEP_FIRST = 'SINGLETON_KEEP_FIRST';
    GoalPriorityType.SINGLETON_KEEP_LAST = 'SINGLETON_KEEP_LAST';
})(GoalPriorityType = exports.GoalPriorityType || (exports.GoalPriorityType = {}));
var DriveName;
(function (DriveName) {
    DriveName.SOCIAL = 'SOCIAL';
    DriveName.PLAYFUL = 'PLAYFUL';
    DriveName.HELPFUL = 'HELPFUL';
})(DriveName = exports.DriveName || (exports.DriveName = {}));

},{}],27:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Updater = (function () {
    function Updater() {
        this.updateables = [];
    }
    Updater.prototype.clear = function () {
        this.updateables = [];
    };
    Updater.prototype.addUpdateable = function (update, intervalMs) {
        if (intervalMs === void 0) { intervalMs = -1; }
        this.updateables.push({
            update: update,
            interval: intervalMs,
            last: -1
        });
    };
    Updater.prototype.update = function () {
        var now = Date.now();
        for (var i = 0; i < this.updateables.length; i++) {
            var el = this.updateables[i];
            if (el.interval === -1 || (now - el.last) > el.interval) {
                el.last = now;
                el.update();
            }
        }
    };
    return Updater;
}());
exports.Updater = Updater;

},{}],28:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkillNames = {
    IDLE: '@be/idle',
    GREETINGS: '@be/greetings',
    SETTINGS: '@be/settings',
    CLOCK: '@be/clock'
};
var SkillUtils = (function () {
    function SkillUtils() {
    }
    SkillUtils.isErrorSkill = function (skillData) {
        return skillData.NLParse &&
            skillData.NLParse.skill === exports.SkillNames.SETTINGS &&
            skillData.NLParse.errorId;
    };
    SkillUtils.isAlarmTimerSkill = function (skillData) {
        return skillData.NLParse &&
            skillData.NLParse.skill === exports.SkillNames.CLOCK &&
            ((skillData.NLParse.domain === 'alarm' && skillData.NLParse.intent === 'finished') ||
                (skillData.NLParse.domain === 'timer' && skillData.NLParse.intent === 'finished'));
    };
    return SkillUtils;
}());
exports.SkillUtils = SkillUtils;
function constrainToBounds(value, min, max) {
    return Math.max(min, Math.min(max, value));
}
exports.constrainToBounds = constrainToBounds;
var DoubleMap = (function () {
    function DoubleMap() {
        this.map = new Map();
    }
    DoubleMap.prototype.set = function (key1, key2, value) {
        var second = this.map.get(key1);
        if (!second) {
            second = new Map();
            this.map.set(key1, second);
        }
        second.set(key2, value);
    };
    DoubleMap.prototype.get = function (key1, key2) {
        var second = this.map.get(key1);
        if (!second) {
            return undefined;
        }
        return second.get(key2);
    };
    return DoubleMap;
}());
exports.DoubleMap = DoubleMap;

},{}],29:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./Types"));
__export(require("./ActionEvents"));
__export(require("./ActionRuntime"));
__export(require("./Utils"));
__export(require("./DateProvider"));
__export(require("./Updater"));

},{"./ActionEvents":20,"./ActionRuntime":21,"./DateProvider":22,"./Types":26,"./Updater":27,"./Utils":28}],30:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Log = require("jibo-log");
var log = new exports.Log('Jibo.Action');
exports.log = log;

},{"jibo-log":undefined}],31:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_typed_events_1 = require("jibo-typed-events");
var GoalPriority_1 = require("./GoalPriority");
var Types_1 = require("../common/Types");
var GoalEvents = (function (_super) {
    __extends(GoalEvents, _super);
    function GoalEvents() {
        var _this = _super.call(this) || this;
        _this.progressChange = new jibo_typed_events_1.Event('Goal progress change');
        _this.finished = new jibo_typed_events_1.Event('Goal finished');
        return _this;
    }
    return GoalEvents;
}(jibo_typed_events_1.EventContainer));
exports.GoalEvents = GoalEvents;
var DEFAULT_GOAL_OPTIONS = {
    priority: GoalPriority_1.GoalPriority.Default,
    priorityType: Types_1.GoalPriorityType.REGULAR,
    maxAttempts: 1
};
var Goal = (function () {
    function Goal(parent, name, options) {
        if (options === void 0) { options = DEFAULT_GOAL_OPTIONS; }
        this.name = name;
        this.events = new GoalEvents();
        this.attempts = 0;
        this._finishedStatus = Types_1.GoalFinishedStatus.NOT_FINISHED;
        this._progressStatus = Types_1.GoalProgressStatus.INITIAL;
        this.parent = parent;
        this.options = Object.assign({}, DEFAULT_GOAL_OPTIONS, options);
        this.created = parent.dateProvider.getDate();
    }
    Goal.prototype.getProgressStatus = function () {
        return this._progressStatus;
    };
    Goal.prototype.setProgressStatus = function (status) {
        if (status !== this._progressStatus) {
            this._progressStatus = status;
            this.events.progressChange.emit(status);
        }
    };
    Goal.prototype.getFinishedStatus = function () {
        return this._finishedStatus;
    };
    Goal.prototype.setFinishedStatus = function (status) {
        var finishedStatusChanged = (status !== this._finishedStatus);
        this._finishedStatus = status;
        if (status !== Types_1.GoalFinishedStatus.NOT_FINISHED) {
            this.setProgressStatus(Types_1.GoalProgressStatus.FINISHED);
        }
        if (finishedStatusChanged) {
            this.events.finished.emit(status);
        }
    };
    Goal.prototype.isEligible = function (state) {
        if (this.options.criteria) {
            var criteria = this.options.criteria;
            if (criteria.startTime && state.time < criteria.startTime) {
                return false;
            }
            if (criteria.endTime && state.time > criteria.endTime) {
                return false;
            }
            if (criteria.currentSkill && state.skillName !== criteria.currentSkill) {
                return false;
            }
            if (criteria.personPresent) {
                if (state.users.map(function (u) { return u.id; }).indexOf(criteria.personPresent) === -1) {
                    return false;
                }
            }
        }
        return true;
    };
    Goal.prototype.canBeCulled = function (state) {
        if (this.options.criteria) {
            var criteria = this.options.criteria;
            return (criteria.endTime && state.time > criteria.endTime);
        }
        return false;
    };
    return Goal;
}());
exports.Goal = Goal;

},{"../common/Types":26,"./GoalPriority":33,"jibo-typed-events":undefined}],32:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Utils_1 = require("../common/Utils");
var GoalParallelism = (function () {
    function GoalParallelism() {
        this.allowedNormal = new Utils_1.DoubleMap();
        this.allowedInverse = new Utils_1.DoubleMap();
    }
    GoalParallelism.prototype.canBeAchievedInParallel = function (goalA, goalB) {
        var parallel = this.allowedNormal.get(goalA.constructor, goalB.constructor);
        if (parallel) {
            if (parallel.goalPredicate) {
                return parallel.goalPredicate(goalA, goalB);
            }
            else {
                return true;
            }
        }
        parallel = this.allowedInverse.get(goalB.constructor, goalA.constructor);
        if (parallel) {
            if (parallel.goalPredicate) {
                return parallel.goalPredicate(goalB, goalA);
            }
            else {
                return true;
            }
        }
        return false;
    };
    GoalParallelism.prototype.registerParallelGoalConstructors = function (goalConstructorA, goalConstructorB, predicate) {
        var goalParallel = {
            constructorA: goalConstructorA,
            constructorB: goalConstructorB,
            goalPredicate: predicate
        };
        this.allowedNormal.set(goalConstructorA, goalConstructorB, goalParallel);
        this.allowedInverse.set(goalConstructorA, goalConstructorB, goalParallel);
    };
    return GoalParallelism;
}());
exports.GoalParallelism = GoalParallelism;

},{"../common/Utils":28}],33:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoalPriority = {
    HeyJibo: 12,
    BeSkillSwitch: 11,
    SkillSwitch: 10,
    HeadTouch: 9,
    Animation: 9,
    FindPerson: 8,
    HJOnly: 7,
    Default: 5,
    Circadian: 4,
    Proactive: 4,
    Motivation: 3,
};

},{}],34:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var log_1 = require("./log");
var GoalProvider = (function () {
    function GoalProvider(parent, name) {
        this.name = name;
        this.parent = parent;
        this.log = log_1.log.createChild("GoalProvider." + name);
    }
    GoalProvider.prototype.provideGoal = function (goal) {
        goal.provider = this;
        this.parent.goals.addGoal(goal);
    };
    return GoalProvider;
}());
exports.GoalProvider = GoalProvider;

},{"./log":58}],35:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Types_1 = require("../common/Types");
var GoalParallelism_1 = require("./GoalParallelism");
var BeSSWGoal_1 = require("./goals/BeSSWGoal");
var SSWGoal_1 = require("./goals/SSWGoal");
var HeadTouchGoal_1 = require("./goals/HeadTouchGoal");
var HJOnlyGoal_1 = require("./goals/HJOnlyGoal");
var log_1 = require("./log");
var log = log_1.log.createChild('GoalSet');
var GoalSet = (function () {
    function GoalSet(parent) {
        this.parent = parent;
        this.goalParallelism = new GoalParallelism_1.GoalParallelism();
        this.goals = [];
        this.goalParallelism.registerParallelGoalConstructors(BeSSWGoal_1.BeSSWGoal, SSWGoal_1.SSWGoal, function (beGoal, sswGoal) {
            return (beGoal.data.skillName === sswGoal.data.skillName);
        });
        this.goalParallelism.registerParallelGoalConstructors(BeSSWGoal_1.BeSSWGoal, HeadTouchGoal_1.HeadTouchGoal);
        this.goalParallelism.registerParallelGoalConstructors(HeadTouchGoal_1.HeadTouchGoal, SSWGoal_1.SSWGoal);
        this.goalParallelism.registerParallelGoalConstructors(BeSSWGoal_1.BeSSWGoal, HJOnlyGoal_1.HJOnlyGoal);
    }
    GoalSet.prototype.addGoal = function (goal) {
        var _this = this;
        log.debug('Adding goal: ', goal.name, ', nr of existing goals: ', this.goals.length);
        var toRemove = [];
        var shouldAdd = true;
        var considerRemovingGoal = function (oldGoal) {
            var active = _this.parent.activeGoals.find(function (current) { return current.goal === oldGoal; });
            if (active) {
                active.action.cancel();
            }
            toRemove.push(oldGoal);
        };
        if (goal.options.priorityType !== Types_1.GoalPriorityType.REGULAR) {
            log.debug('Singleton goal');
            for (var i = 0; i < this.goals.length; i++) {
                var otherGoal = this.goals[i];
                if (otherGoal.constructor === goal.constructor) {
                    log.debug('Found goal of same type');
                    if (goal.options.priority > otherGoal.options.priority) {
                        log.debug('Incoming is higher priority');
                        considerRemovingGoal(otherGoal);
                    }
                    else if (otherGoal.options.priority > goal.options.priority) {
                        log.debug('Incoming is lower priority');
                        shouldAdd = false;
                    }
                    else if (goal.options.priorityType === Types_1.GoalPriorityType.SINGLETON_KEEP_FIRST) {
                        log.debug('Priority type is SINGLETON_KEEP_FIRST');
                        shouldAdd = false;
                    }
                    else if (goal.options.priorityType === Types_1.GoalPriorityType.SINGLETON_KEEP_LAST) {
                        log.debug('Priority type is SINGLETON_KEEP_LAST');
                        considerRemovingGoal(otherGoal);
                    }
                }
            }
        }
        toRemove.forEach(function (g) {
            _this.removeGoal(g);
            g.setFinishedStatus(Types_1.GoalFinishedStatus.CANCELED);
        });
        if (shouldAdd) {
            this.goals.push(goal);
            this.parent.events.goalAdded.emit(goal);
        }
        else {
            goal.setFinishedStatus(Types_1.GoalFinishedStatus.CANCELED);
        }
    };
    GoalSet.prototype.chooseGoal = function (state) {
        if (this.goals.length === 0) {
            return null;
        }
        var eligibleGoals = [];
        var nextGoals = [];
        var canBeCulled = [];
        this.goals.forEach(function (goal) {
            if (goal.canBeCulled(state)) {
                canBeCulled.push(goal);
            }
            else {
                nextGoals.push(goal);
                if (goal.isEligible(state)) {
                    eligibleGoals.push(goal);
                }
            }
        });
        this.goals = nextGoals;
        canBeCulled.forEach(function (g) { return g.setFinishedStatus(Types_1.GoalFinishedStatus.CULLED); });
        if (eligibleGoals.length) {
            eligibleGoals.sort(function (a, b) {
                var diff = b.options.priority - a.options.priority;
                if (diff === 0) {
                    return (Math.random() > 0.5) ? -1 : 1;
                }
                else {
                    return diff;
                }
            });
            return eligibleGoals[0];
        }
        else {
            return null;
        }
    };
    GoalSet.prototype.removeGoal = function (goal) {
        var index = this.goals.indexOf(goal);
        if (index > -1) {
            this.goals.splice(index, 1);
        }
        else {
            log.warn("Trying to remove goal '" + goal.name + "' but didn't find it in goal list");
        }
    };
    return GoalSet;
}());
exports.GoalSet = GoalSet;

},{"../common/Types":26,"./GoalParallelism":32,"./goals/BeSSWGoal":49,"./goals/HJOnlyGoal":52,"./goals/HeadTouchGoal":53,"./goals/SSWGoal":55,"./log":58}],36:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Types_1 = require("../../common/Types");
var Utils_1 = require("../../common/Utils");
var GoalProvider_1 = require("../GoalProvider");
var SSWGoal_1 = require("../goals/SSWGoal");
var CircadianStateMachine_1 = require("./circadian/CircadianStateMachine");
var CircadianGoalProvider = (function (_super) {
    __extends(CircadianGoalProvider, _super);
    function CircadianGoalProvider(parent) {
        var _this = _super.call(this, parent, 'Circadian rhythm goal provider') || this;
        _this.circadianSM = new CircadianStateMachine_1.CircadianStateMachine(_this);
        _this.currentToD = null;
        _this.goToSleepHandlerBind = _this.goToSleepHandler.bind(_this);
        return _this;
    }
    CircadianGoalProvider.prototype.init = function () {
        this.parent.jibo.globalEvents.sleep.on(this.goToSleepHandlerBind);
        this.circadianSM.init();
        var now = this.parent.dateProvider.getDate();
        this.currentToD = Types_1.TimeOfDay.getTimeOfDay(now);
        this.circadianSM.start();
        this.lastSleepState = this.getCurrentSleepState();
    };
    CircadianGoalProvider.prototype.dispose = function () {
        this.parent.jibo.globalEvents.sleep.removeListener(this.goToSleepHandlerBind);
        this.circadianSM.destroy();
    };
    CircadianGoalProvider.prototype.checkSleepStateChange = function () {
        var current = this.getCurrentSleepState();
        if (current !== this.lastSleepState) {
            this.lastSleepState = current;
            this.parent.events.sleepChange.emit(current);
        }
    };
    CircadianGoalProvider.prototype.getCurrentSleepState = function () {
        var current = this.circadianSM.getCurrentState();
        if (current === this.circadianSM.fallingAsleepState) {
            return Types_1.SleepState.FALLING_ASLEEP;
        }
        else if (current === this.circadianSM.asleepState) {
            return Types_1.SleepState.ASLEEP;
        }
        else {
            return Types_1.SleepState.AWAKE;
        }
    };
    CircadianGoalProvider.prototype.goToSleepHandler = function () {
        var state = this.parent.createState();
        if (state.skillName !== Utils_1.SkillNames.IDLE) {
            this.provideGoal(new SSWGoal_1.SSWGoal(this.parent, {
                skillName: Utils_1.SkillNames.IDLE,
                skillOptions: {}
            }));
        }
        this.circadianSM.events.goToSleep.emit();
    };
    CircadianGoalProvider.prototype.update = function () {
        var now = this.parent.dateProvider.getDate();
        var newToD = Types_1.TimeOfDay.getTimeOfDay(now);
        if (newToD !== this.currentToD) {
            this.currentToD = newToD;
            if (this.currentToD === Types_1.TimeOfDay.MORNING ||
                this.currentToD === Types_1.TimeOfDay.DAY) {
                this.circadianSM.events.dayTimeStarts.emit();
            }
            else if (this.currentToD === Types_1.TimeOfDay.NIGHT_EARLY) {
                this.circadianSM.events.earlyNightTimeStarts.emit();
            }
            else if (this.currentToD === Types_1.TimeOfDay.NIGHT) {
                this.circadianSM.events.nightTimeStarts.emit();
            }
        }
        this.circadianSM.update();
    };
    return CircadianGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.CircadianGoalProvider = CircadianGoalProvider;

},{"../../common/Types":26,"../../common/Utils":28,"../GoalProvider":34,"../goals/SSWGoal":55,"./circadian/CircadianStateMachine":43}],37:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GoalProvider_1 = require("../GoalProvider");
var HJGoal_1 = require("../goals/HJGoal");
var HJGoalProvider = (function (_super) {
    __extends(HJGoalProvider, _super);
    function HJGoalProvider(parent) {
        var _this = _super.call(this, parent, 'Hey Jibo Goal Provider') || this;
        _this.listening = false;
        _this.processHJBind = function (data) { return _this.processHJ(data); };
        return _this;
    }
    HJGoalProvider.prototype.init = function () {
        this.setListening(true);
    };
    HJGoalProvider.prototype.dispose = function () {
        this.setListening(false);
    };
    HJGoalProvider.prototype.setListening = function (listen) {
        if (listen) {
            if (this.listening) {
                this.log.warn("setListening(" + listen + ") called when listening is already " + this.listening);
            }
            else {
                this.log.info("enabling listening for HJ event");
                this.parent.jibo.lps.identity.events.hjEvent.on(this.processHJBind);
                this.listening = true;
            }
        }
        else {
            if (this.listening) {
                this.log.info("disabling listening for HJ event");
                this.parent.jibo.lps.identity.events.hjEvent.removeListener(this.processHJBind);
                this.listening = false;
            }
            else {
                this.log.warn("setListening(" + listen + ") called when listening is already " + this.listening);
            }
        }
    };
    HJGoalProvider.prototype.processHJ = function (hjData) {
        this.provideGoal(new HJGoal_1.HJGoal(this.parent, hjData));
        this.parent.memory.setHJData(hjData);
    };
    return HJGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.HJGoalProvider = HJGoalProvider;

},{"../GoalProvider":34,"../goals/HJGoal":51}],38:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GoalProvider_1 = require("../GoalProvider");
var Utils_1 = require("../../common/Utils");
var SSWGoal_1 = require("../goals/SSWGoal");
var HJNoMatchGoalProvider = (function (_super) {
    __extends(HJNoMatchGoalProvider, _super);
    function HJNoMatchGoalProvider(parent) {
        var _this = _super.call(this, parent, 'HJNoMatch goal provider') || this;
        _this.handleHJNoMatchBind = _this.handleHJNoMatch.bind(_this);
        return _this;
    }
    HJNoMatchGoalProvider.prototype.init = function () {
        this.parent.jibo.gl.events.noMatch.on(this.handleHJNoMatchBind);
    };
    HJNoMatchGoalProvider.prototype.dispose = function () {
        this.parent.jibo.gl.events.noMatch.removeListener(this.handleHJNoMatchBind);
    };
    HJNoMatchGoalProvider.prototype.handleHJNoMatch = function (data) {
        if (!data.result.triggeredByHJ) {
            return;
        }
        if (this.parent.currentSkillName === Utils_1.SkillNames.IDLE) {
            this.provideGoal(new SSWGoal_1.SSWGoal(this.parent, {
                skillName: Utils_1.SkillNames.IDLE,
                skillOptions: {
                    intent: 'hjNomatch',
                    text: data.result.Input
                },
                criteria: {
                    currentSkill: Utils_1.SkillNames.IDLE
                }
            }, "HJNoMatch"));
        }
    };
    return HJNoMatchGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.HJNoMatchGoalProvider = HJNoMatchGoalProvider;

},{"../../common/Utils":28,"../GoalProvider":34,"../goals/SSWGoal":55}],39:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GoalProvider_1 = require("../GoalProvider");
var HJOnlyGoal_1 = require("../goals/HJOnlyGoal");
var Utils_1 = require("../../common/Utils");
var HJOnlyGoalProvider = (function (_super) {
    __extends(HJOnlyGoalProvider, _super);
    function HJOnlyGoalProvider(parent) {
        var _this = _super.call(this, parent, 'Hey Jibo Only Goal Provider') || this;
        _this.handleHJOnlyBind = _this.handleHJOnly.bind(_this);
        return _this;
    }
    HJOnlyGoalProvider.prototype.init = function () {
        this.parent.jibo.gl.events.hjOnly.on(this.handleHJOnlyBind);
    };
    HJOnlyGoalProvider.prototype.dispose = function () {
        this.parent.jibo.gl.events.hjOnly.removeListener(this.handleHJOnlyBind);
    };
    HJOnlyGoalProvider.prototype.handleHJOnly = function (data) {
        var goal = new HJOnlyGoal_1.HJOnlyGoal(this.parent);
        goal.options.criteria = {
            currentSkill: Utils_1.SkillNames.IDLE,
        };
        var state = this.parent.createState();
        if (state.skillName === Utils_1.SkillNames.IDLE) {
            this.provideGoal(goal);
        }
    };
    return HJOnlyGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.HJOnlyGoalProvider = HJOnlyGoalProvider;

},{"../../common/Utils":28,"../GoalProvider":34,"../goals/HJOnlyGoal":52}],40:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GoalProvider_1 = require("../GoalProvider");
var HeadTouchGoal_1 = require("../goals/HeadTouchGoal");
var HeadTouchGoalProvider = (function (_super) {
    __extends(HeadTouchGoalProvider, _super);
    function HeadTouchGoalProvider(parent) {
        var _this = _super.call(this, parent, 'Head Touch Goal Provider') || this;
        _this.touchBind = _this.handleHeadTouch.bind(_this);
        return _this;
    }
    HeadTouchGoalProvider.prototype.init = function () {
        this.parent.jibo.globalEvents.touchStop.on(this.touchBind);
    };
    HeadTouchGoalProvider.prototype.dispose = function () {
        this.parent.jibo.globalEvents.touchStop.removeListener(this.touchBind);
    };
    HeadTouchGoalProvider.prototype.handleHeadTouch = function (data) {
        var goal = new HeadTouchGoal_1.HeadTouchGoal(this.parent, data);
        this.provideGoal(goal);
    };
    return HeadTouchGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.HeadTouchGoalProvider = HeadTouchGoalProvider;

},{"../GoalProvider":34,"../goals/HeadTouchGoal":53}],41:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GoalProvider_1 = require("../GoalProvider");
var GoalPriority_1 = require("../GoalPriority");
var Utils_1 = require("../../common/Utils");
var Types_1 = require("../../common/Types");
var SSWGoal_1 = require("../goals/SSWGoal");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var ProactiveGreetingGoalProvider = (function (_super) {
    __extends(ProactiveGreetingGoalProvider, _super);
    function ProactiveGreetingGoalProvider(parent) {
        var _this = _super.call(this, parent, 'Pro-active Greetings goal provider') || this;
        _this.handleVisibleFaceBind = _this.handleVisibleFace.bind(_this);
        _this.handleHJHEardBind = _this.handleHJHeard.bind(_this);
        _this.lastHJHeard = -1;
        return _this;
    }
    ProactiveGreetingGoalProvider.prototype.init = function () {
        this.parent.jibo.lps.identity.events.visibleFaceStarted.on(this.handleVisibleFaceBind);
        this.parent.jibo.gl.events.hjHeard.on(this.handleHJHEardBind);
    };
    ProactiveGreetingGoalProvider.prototype.dispose = function () {
        this.parent.jibo.lps.identity.events.visibleFaceStarted.removeListener(this.handleVisibleFaceBind);
        this.parent.jibo.gl.events.hjHeard.removeListener(this.handleHJHEardBind);
    };
    ProactiveGreetingGoalProvider.prototype.handleHJHeard = function () {
        var now = this.parent.dateProvider.getDate();
        this.lastHJHeard = now.getTime();
    };
    ProactiveGreetingGoalProvider.prototype.handleVisibleFace = function () {
        var state = this.parent.createState();
        var now = this.parent.dateProvider.getDate();
        var nowMs = now.getTime();
        var msInSkill = nowMs - state.skillStartTime.getTime();
        var msSinceHJ = nowMs - this.lastHJHeard;
        if (state.skillName === Utils_1.SkillNames.IDLE &&
            jibo_cai_utils_1.TimeUtils.msToSeconds(msInSkill) > 60 &&
            jibo_cai_utils_1.TimeUtils.msToSeconds(msSinceHJ) > 60 &&
            state.sleep === Types_1.SleepState.AWAKE) {
            this.handleVisibleFaceInternal();
        }
    };
    ProactiveGreetingGoalProvider.prototype.handleVisibleFaceInternal = function () {
        var goal = new SSWGoal_1.SSWGoal(this.parent, {
            skillName: Utils_1.SkillNames.GREETINGS,
            skillOptions: {
                NLParse: {
                    skill: Utils_1.SkillNames.GREETINGS,
                    intent: 'proactiveGreeting'
                }
            },
            criteria: {
                currentSkill: Utils_1.SkillNames.IDLE
            }
        });
        goal.options.priority = GoalPriority_1.GoalPriority.Proactive;
        this.provideGoal(goal);
    };
    return ProactiveGreetingGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.ProactiveGreetingGoalProvider = ProactiveGreetingGoalProvider;

},{"../../common/Types":26,"../../common/Utils":28,"../GoalPriority":33,"../GoalProvider":34,"../goals/SSWGoal":55,"jibo-cai-utils":undefined}],42:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_state_machine_1 = require("jibo-state-machine");
var Types_1 = require("../../../common/Types");
var ChooseTimeState = (function (_super) {
    __extends(ChooseTimeState, _super);
    function ChooseTimeState(parent) {
        var _this = _super.call(this, parent, "Time of day?") || this;
        _this.parent = parent;
        _this.onEntry = function (transition) {
            var tod = _this.parent.parent.currentToD;
            if (tod === Types_1.TimeOfDay.NIGHT_EARLY) {
                _this.transitionTo(_this.fallingAsleepState);
            }
            if (tod === Types_1.TimeOfDay.NIGHT) {
                _this.transitionTo(_this.asleepState);
            }
            else {
                _this.transitionTo(_this.awakeState);
            }
        };
        return _this;
    }
    ChooseTimeState.prototype.installTransitions = function (awakeState, fallingAsleepState, asleepState) {
        this.awakeState = awakeState;
        this.fallingAsleepState = fallingAsleepState;
        this.asleepState = asleepState;
        this.addInternalTransition("Daytime", awakeState);
        this.addInternalTransition("Early Nighttime", fallingAsleepState);
        this.addInternalTransition("Nighttime", asleepState);
    };
    return ChooseTimeState;
}(jibo_state_machine_1.State));
exports.ChooseTimeState = ChooseTimeState;

},{"../../../common/Types":26,"jibo-state-machine":undefined}],43:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_typed_events_1 = require("jibo-typed-events");
var jibo_state_machine_1 = require("jibo-state-machine");
var ChooseTimeState_1 = require("./ChooseTimeState");
var Types_1 = require("../../../common/Types");
var InterruptState_1 = require("./InterruptState");
var log_1 = require("../log");
var SMEvents = (function (_super) {
    __extends(SMEvents, _super);
    function SMEvents() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.earlyNightTimeStarts = new jibo_typed_events_1.Event("Early night time starts");
        _this.nightTimeStarts = new jibo_typed_events_1.Event("Night time starts");
        _this.dayTimeStarts = new jibo_typed_events_1.Event("Day time starts");
        _this.goToSleep = new jibo_typed_events_1.Event("\"Go to sleep\"");
        _this.hjInterruption = new jibo_typed_events_1.Event("HJ Interruption");
        return _this;
    }
    return SMEvents;
}(jibo_typed_events_1.EventContainer));
exports.SMEvents = SMEvents;
var CircadianStateMachine = (function (_super) {
    __extends(CircadianStateMachine, _super);
    function CircadianStateMachine(parent) {
        var _this = _super.call(this) || this;
        _this.parent = parent;
        _this.events = new SMEvents();
        _this.chooseTodState = new ChooseTimeState_1.ChooseTimeState(_this);
        _this.asleepState = new jibo_state_machine_1.State(_this, Types_1.SleepState.ASLEEP);
        _this.fallingAsleepState = new jibo_state_machine_1.State(_this, Types_1.SleepState.FALLING_ASLEEP);
        _this.awakeState = new jibo_state_machine_1.State(_this, Types_1.SleepState.AWAKE);
        _this.interruptSleepState = new InterruptState_1.InterruptState(_this);
        _this.onHjHeard = function () { return _this.events.hjInterruption.emit(); };
        return _this;
    }
    CircadianStateMachine.prototype.init = function () {
        var _this = this;
        var jibo = this.parent.parent.jibo;
        this.setInitial(this.chooseTodState);
        this.awakeState.onEntry = function () { return log_1.log.info("Circadian: entering '" + _this.awakeState.name + "'"); };
        this.fallingAsleepState.onEntry = function () { return log_1.log.info("Circadian: entering '" + _this.fallingAsleepState.name + "'"); };
        this.asleepState.onEntry = function () { return log_1.log.info("Circadian: entering '" + _this.asleepState.name + "'"); };
        this.chooseTodState.installTransitions(this.awakeState, this.fallingAsleepState, this.asleepState);
        [this.awakeState,
            this.fallingAsleepState,
            this.asleepState,
            this.interruptSleepState].forEach(function (s) {
            s.addTypedEventTransition(_this.events.goToSleep, _this.asleepState);
        });
        this.awakeState.addTypedEventTransition(this.events.earlyNightTimeStarts, this.fallingAsleepState);
        this.awakeState.addTypedEventTransition(this.events.nightTimeStarts, this.asleepState);
        this.fallingAsleepState.addTypedEventTransition(this.events.dayTimeStarts, this.awakeState);
        this.fallingAsleepState.addTypedEventTransition(this.events.nightTimeStarts, this.asleepState);
        this.fallingAsleepState.addTypedEventTransition(this.events.hjInterruption, this.interruptSleepState);
        this.asleepState.addTypedEventTransition(this.events.dayTimeStarts, this.awakeState);
        this.asleepState.addTypedEventTransition(this.events.hjInterruption, this.interruptSleepState);
        this.interruptSleepState.installTransitions(this.chooseTodState);
        this.stateChanged.on(function (transition) {
            _this.parent.checkSleepStateChange();
        });
        jibo.gl.events.hjHeard.on(this.onHjHeard);
    };
    CircadianStateMachine.prototype.dispose = function () {
        var jibo = this.parent.parent.jibo;
        jibo.gl.events.hjHeard.removeListener(this.onHjHeard);
        this.destroy();
    };
    return CircadianStateMachine;
}(jibo_state_machine_1.StateMachine));
exports.CircadianStateMachine = CircadianStateMachine;

},{"../../../common/Types":26,"../log":47,"./ChooseTimeState":42,"./InterruptState":44,"jibo-state-machine":undefined,"jibo-typed-events":undefined}],44:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var jibo_state_machine_1 = require("jibo-state-machine");
var INTERRUPTION_TIMEOUT_MIN = 5;
var InterruptState = (function (_super) {
    __extends(InterruptState, _super);
    function InterruptState(parent) {
        var _this = _super.call(this, parent, "Interrupt") || this;
        var jibo = parent.parent.parent.jibo;
        var timeOfLastInterruption = -1;
        var interrupt = function () {
            timeOfLastInterruption = Date.now();
        };
        _this.onEntry = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                timeOfLastInterruption = Date.now();
                jibo.gl.events.hjHeard.on(interrupt);
                jibo.system.events.touchOn.on(interrupt);
                return [2];
            });
        }); };
        _this.onExit = function () {
            jibo.gl.events.hjHeard.removeListener(interrupt);
            jibo.system.events.touchOn.removeListener(interrupt);
        };
        _this.onUpdate = function () {
            var timeSinceInterrupt = Date.now() - timeOfLastInterruption;
            if (timeSinceInterrupt > INTERRUPTION_TIMEOUT_MIN * 60 * 1000) {
                timeOfLastInterruption = Date.now();
                _this.transitionTo(_this.doneState);
            }
        };
        return _this;
    }
    InterruptState.prototype.installTransitions = function (doneState) {
        this.doneState = doneState;
        this.addInternalTransition('Done', doneState);
    };
    return InterruptState;
}(jibo_state_machine_1.State));
exports.InterruptState = InterruptState;

},{"jibo-state-machine":undefined}],45:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./ChooseTimeState"));
__export(require("./CircadianStateMachine"));

},{"./ChooseTimeState":42,"./CircadianStateMachine":43}],46:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./HJGoalProvider"));
__export(require("./HJOnlyGoalProvider"));
__export(require("./ProactiveGreetingGoalProvider"));
__export(require("./HJNoMatchGoalProvider"));
__export(require("./CircadianGoalProvider"));
var circadian = require("./circadian");
exports.circadian = circadian;

},{"./CircadianGoalProvider":36,"./HJGoalProvider":37,"./HJNoMatchGoalProvider":38,"./HJOnlyGoalProvider":39,"./ProactiveGreetingGoalProvider":41,"./circadian":45}],47:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var log_1 = require("../log");
var log = log_1.log.createChild('Provider');
exports.log = log;

},{"../log":58}],48:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var GoalPriority_1 = require("../GoalPriority");
var AnimationGoal = (function (_super) {
    __extends(AnimationGoal, _super);
    function AnimationGoal(parent, data) {
        var _this = _super.call(this, parent, AnimationGoal.getName(data), {
            priority: GoalPriority_1.GoalPriority.Animation,
            criteria: data.criteria,
        }) || this;
        _this.data = data;
        return _this;
    }
    AnimationGoal.getName = function (data) {
        var name = "Play Animation Goal [";
        name += (data.category) ? "cat: " + data.category + "]" : "name: " + data.name + "]";
        return name;
    };
    return AnimationGoal;
}(Goal_1.Goal));
exports.AnimationGoal = AnimationGoal;

},{"../Goal":31,"../GoalPriority":33}],49:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var Types_1 = require("../../common/Types");
var GoalPriority_1 = require("../GoalPriority");
var BeSSWGoal = (function (_super) {
    __extends(BeSSWGoal, _super);
    function BeSSWGoal(parent, data, name) {
        if (name === void 0) { name = "Be Skill Switch Goal: '" + data.skillName + "'"; }
        var _this = _super.call(this, parent, name, {
            priority: GoalPriority_1.GoalPriority.BeSkillSwitch,
            criteria: data.criteria,
            priorityType: Types_1.GoalPriorityType.SINGLETON_KEEP_LAST
        }) || this;
        _this.data = data;
        if (data.beSkillPreferences) {
            if (data.beSkillPreferences.cancelOrientOnStart) {
                _this.options.priority = GoalPriority_1.GoalPriority.HeyJibo + 1;
            }
        }
        return _this;
    }
    return BeSSWGoal;
}(Goal_1.Goal));
exports.BeSSWGoal = BeSSWGoal;

},{"../../common/Types":26,"../Goal":31,"../GoalPriority":33}],50:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var GoalPriority_1 = require("../GoalPriority");
var FindPersonGoal = (function (_super) {
    __extends(FindPersonGoal, _super);
    function FindPersonGoal(parent) {
        return _super.call(this, parent, 'Find person goal', {
            priority: GoalPriority_1.GoalPriority.FindPerson,
        }) || this;
    }
    return FindPersonGoal;
}(Goal_1.Goal));
exports.FindPersonGoal = FindPersonGoal;

},{"../Goal":31,"../GoalPriority":33}],51:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Types_1 = require("../../common/Types");
var Goal_1 = require("../Goal");
var GoalPriority_1 = require("../GoalPriority");
var HJGoal = (function (_super) {
    __extends(HJGoal, _super);
    function HJGoal(parent, hjData) {
        var _this = _super.call(this, parent, 'Hey Jibo', {
            priority: GoalPriority_1.GoalPriority.HeyJibo,
            priorityType: Types_1.GoalPriorityType.SINGLETON_KEEP_LAST
        }) || this;
        _this.hjData = hjData;
        return _this;
    }
    return HJGoal;
}(Goal_1.Goal));
exports.HJGoal = HJGoal;

},{"../../common/Types":26,"../Goal":31,"../GoalPriority":33}],52:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var GoalPriority_1 = require("../GoalPriority");
var HJOnlyGoal = (function (_super) {
    __extends(HJOnlyGoal, _super);
    function HJOnlyGoal(parent) {
        return _super.call(this, parent, 'Hey Jibo Only', {
            priority: GoalPriority_1.GoalPriority.HJOnly,
        }) || this;
    }
    HJOnlyGoal.prototype.canBeCulled = function (state) {
        var seconds = jibo_cai_utils_1.TimeUtils.msToSeconds(this.parent.dateProvider.getDate().getTime() - this.created.getTime());
        return _super.prototype.canBeCulled.call(this, state) || (seconds > 10);
    };
    return HJOnlyGoal;
}(Goal_1.Goal));
exports.HJOnlyGoal = HJOnlyGoal;

},{"../Goal":31,"../GoalPriority":33,"jibo-cai-utils":undefined}],53:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var GoalPriority_1 = require("../GoalPriority");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var HeadTouchGoal = (function (_super) {
    __extends(HeadTouchGoal, _super);
    function HeadTouchGoal(parent, data) {
        var _this = _super.call(this, parent, 'Head Touch', {
            priority: GoalPriority_1.GoalPriority.HeadTouch,
            maxAttempts: 1
        }) || this;
        _this.data = data;
        return _this;
    }
    HeadTouchGoal.prototype.canBeCulled = function (state) {
        var seconds = jibo_cai_utils_1.TimeUtils.msToSeconds(this.parent.dateProvider.getDate().getTime() - this.created.getTime());
        return _super.prototype.canBeCulled.call(this, state) || (seconds > 0.5);
    };
    return HeadTouchGoal;
}(Goal_1.Goal));
exports.HeadTouchGoal = HeadTouchGoal;

},{"../Goal":31,"../GoalPriority":33,"jibo-cai-utils":undefined}],54:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var common_1 = require("../../common");
var GoalPriority_1 = require("../GoalPriority");
var Types_1 = require("../../common/Types");
var MotivationGoal = (function (_super) {
    __extends(MotivationGoal, _super);
    function MotivationGoal(parent) {
        return _super.call(this, parent, 'MotivationGoal', {
            priority: GoalPriority_1.GoalPriority.Motivation,
            priorityType: Types_1.GoalPriorityType.SINGLETON_KEEP_FIRST,
            maxAttempts: 1,
        }) || this;
    }
    MotivationGoal.prototype.isEligible = function (state) {
        if (state.skillName !== common_1.SkillNames.IDLE) {
            return false;
        }
        return this.parent.motivations.drivesAboveThreshold().length > 0;
    };
    MotivationGoal.prototype.canBeCulled = function (state) {
        return !(this.parent.motivations.drivesAboveThreshold().length > 0);
    };
    return MotivationGoal;
}(Goal_1.Goal));
exports.MotivationGoal = MotivationGoal;

},{"../../common":29,"../../common/Types":26,"../Goal":31,"../GoalPriority":33}],55:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Goal_1 = require("../Goal");
var Types_1 = require("../../common/Types");
var GoalPriority_1 = require("../GoalPriority");
var SSWGoal = (function (_super) {
    __extends(SSWGoal, _super);
    function SSWGoal(parent, data, name) {
        if (name === void 0) { name = "Skill Launch: '" + data.skillName + "'"; }
        var _this = _super.call(this, parent, name, {
            priority: GoalPriority_1.GoalPriority.SkillSwitch,
            criteria: data.criteria,
            priorityType: Types_1.GoalPriorityType.SINGLETON_KEEP_LAST
        }) || this;
        _this.data = data;
        return _this;
    }
    return SSWGoal;
}(Goal_1.Goal));
exports.SSWGoal = SSWGoal;

},{"../../common/Types":26,"../Goal":31,"../GoalPriority":33}],56:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./HJGoal"));
__export(require("./SSWGoal"));
__export(require("./BeSSWGoal"));
__export(require("./HJOnlyGoal"));
__export(require("./AnimationGoal"));
__export(require("./FindPersonGoal"));
__export(require("./HeadTouchGoal"));

},{"./AnimationGoal":48,"./BeSSWGoal":49,"./FindPersonGoal":50,"./HJGoal":51,"./HJOnlyGoal":52,"./HeadTouchGoal":53,"./SSWGoal":55}],57:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./Goal"));
__export(require("./GoalSet"));
__export(require("./GoalProvider"));
__export(require("./GoalPriority"));
__export(require("./GoalParallelism"));
var goalproviders = require("./goalproviders");
exports.goalproviders = goalproviders;
var goals = require("./goals");
exports.goals = goals;

},{"./Goal":31,"./GoalParallelism":32,"./GoalPriority":33,"./GoalProvider":34,"./GoalSet":35,"./goalproviders":46,"./goals":56}],58:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var log_1 = require("../common/log");
var log = log_1.log.createChild('Goal');
exports.log = log;

},{"../common/log":30}],59:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Utils_1 = require("../common/Utils");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var Drive = (function () {
    function Drive(name, ratePerMin) {
        if (ratePerMin === void 0) { ratePerMin = 0; }
        this.name = name;
        this.ratePerMin = 0;
        this.threshold = 1;
        this._value = 0;
        if (ratePerMin < 0) {
            console.error("invalid rate for drive " + name + ": " + ratePerMin + ". Setting to 0.");
            ratePerMin = 0;
        }
        this.ratePerMin = ratePerMin;
    }
    Object.defineProperty(Drive.prototype, "value", {
        get: function () {
            return this._value;
        },
        enumerable: true,
        configurable: true
    });
    Drive.prototype.update = function () {
        this.computeDriveIncrement();
    };
    Drive.prototype.applyEffect = function (effect) {
        this._value = Utils_1.constrainToBounds(this.value + effect, 0, 1);
    };
    Drive.prototype.computeDriveIncrement = function () {
        if (!this.lastUpdate) {
            this.lastUpdate = Date.now();
        }
        else {
            var now = Date.now();
            var delta_t = now - this.lastUpdate;
            var delta_d = jibo_cai_utils_1.TimeUtils.msToMinutes(delta_t) * this.ratePerMin;
            this._value = Utils_1.constrainToBounds(this._value + delta_d, 0, 1);
            this.lastUpdate = now;
        }
    };
    return Drive;
}());
exports.Drive = Drive;

},{"../common/Utils":28,"jibo-cai-utils":undefined}],60:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var MotivationGoal_1 = require("../goal/goals/MotivationGoal");
var GoalProvider_1 = require("../goal/GoalProvider");
var Drive_1 = require("./Drive");
var Types_1 = require("../common/Types");
var WakingDrive_1 = require("./drives/WakingDrive");
var log_1 = require("../common/log");
var MotivationGoalProvider = (function (_super) {
    __extends(MotivationGoalProvider, _super);
    function MotivationGoalProvider(parent) {
        return _super.call(this, parent, "Motivation") || this;
    }
    MotivationGoalProvider.prototype.init = function () {
    };
    MotivationGoalProvider.prototype.dispose = function () {
    };
    return MotivationGoalProvider;
}(GoalProvider_1.GoalProvider));
exports.MotivationGoalProvider = MotivationGoalProvider;
var MotivationSystem = (function () {
    function MotivationSystem(parent) {
        this.parent = parent;
        this.drives = new Map();
        this.drives.set(Types_1.DriveName.SOCIAL, new WakingDrive_1.WakingDrive(this.parent, Types_1.DriveName.SOCIAL, 1 / 120));
        this.drives.set(Types_1.DriveName.PLAYFUL, new WakingDrive_1.WakingDrive(this.parent, Types_1.DriveName.PLAYFUL, 1 / 15));
        this.drives.set(Types_1.DriveName.HELPFUL, new Drive_1.Drive(Types_1.DriveName.HELPFUL));
        this.goalProvider = new MotivationGoalProvider(parent);
        this.log = log_1.log.createChild('Motivation');
    }
    MotivationSystem.prototype.update = function () {
        var makeGoal = false;
        this.drives.forEach(function (drive) {
            drive.update();
            if (drive.value >= drive.threshold) {
                makeGoal = true;
            }
        });
        if (makeGoal) {
            this.goalProvider.provideGoal(new MotivationGoal_1.MotivationGoal(this.parent));
        }
    };
    MotivationSystem.prototype.applyMotivationalEffects = function (action) {
        var _this = this;
        action.options.driveEffects.forEach(function (dv, dn) {
            _this.parent.motivations.drives.get(dn).applyEffect(dv);
        });
    };
    MotivationSystem.prototype.drivesAboveThreshold = function () {
        var aboveThreshold = new Array();
        this.drives.forEach(function (d) {
            if (d.value >= d.threshold) {
                aboveThreshold.push(d);
            }
        });
        return aboveThreshold;
    };
    return MotivationSystem;
}());
exports.MotivationSystem = MotivationSystem;

},{"../common/Types":26,"../common/log":30,"../goal/GoalProvider":34,"../goal/goals/MotivationGoal":54,"./Drive":59,"./drives/WakingDrive":61}],61:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Drive_1 = require("../Drive");
var Types_1 = require("../../common/Types");
var WakingDrive = (function (_super) {
    __extends(WakingDrive, _super);
    function WakingDrive(parent, name, ratePerMin) {
        if (ratePerMin === void 0) { ratePerMin = 0; }
        var _this = _super.call(this, name, ratePerMin) || this;
        _this.parent = parent;
        _this._wakingRatePerMin = _this.ratePerMin;
        return _this;
    }
    WakingDrive.prototype.update = function () {
        if (this.parent.circadianGoalProvider.getCurrentSleepState() === Types_1.SleepState.AWAKE) {
            this.ratePerMin = this._wakingRatePerMin;
        }
        else {
            this.ratePerMin = 0;
            this._value = 0;
        }
        _super.prototype.update.call(this);
    };
    return WakingDrive;
}(Drive_1.Drive));
exports.WakingDrive = WakingDrive;

},{"../../common/Types":26,"../Drive":59}],62:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./WakingDrive"));

},{"./WakingDrive":61}],63:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var drives = require("./drives");
exports.drives = drives;
__export(require("./Drive"));
__export(require("./MotivationSystem"));

},{"./Drive":59,"./MotivationSystem":60,"./drives":62}],64:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var action = require("./action");
exports.action = action;
var motivation = require("./motivation");
exports.motivation = motivation;
var common = require("./common");
exports.common = common;
var goal = require("./goal");
exports.goal = goal;
var api = require("./api");
exports.api = api;

},{"./action":15,"./api":19,"./common":29,"./goal":57,"./motivation":63}]},{},[64])(64)
});
