// @be/jukebox/src/models/MusicLibrary.ts
import jibo = require('jibo');

export interface Track {
  title: string;
  url: string;
  file: string;
  format: string;
}

export interface Album {
  /** Folder name under music/ (e.g. "CHASER"). */
  id: string;
  /** Display title (folder name, prettified). */
  title: string;
  /** Cover image URL, or null if none found. */
  coverUrl: string;
  tracks: Track[];
}

export interface ScanResult {
  albums: Album[];
  /** Absolute on-disk music directory that was scanned (if known). */
  dir: string;
  /** Human-readable error for the status screen; null when OK. */
  error: string;
  /** Extra diagnostic text (paths, exception stacks, etc.). */
  detail: string;
}

/** Audio extensions the jukebox lists and tries to play. */
const AUDIO_EXT = /\.(mp3|opus|ogg|oga|flac)$/i;
/** Cover image names looked for inside each album folder (case-insensitive). */
const COVER_NAMES = ['cover.png', 'cover.jpg', 'cover.jpeg', 'folder.png', 'folder.jpg'];

/**
 * Discovers albums (folders) under the skill's `music/` directory at runtime.
 *
 * Expected layout:
 *
 *   music/
 *     CHASER/
 *       cover.png
 *       GIRL HELL 1999.flac
 *     Some EP/
 *       cover.jpg
 *       track-01.mp3
 *
 * Nothing here is bundled: the folder is read from disk every time the skill
 * opens, so a user can add albums simply by dropping folders into `music/`
 * with no rebuild.
 */
export default class MusicLibrary {

  /**
   * Scan music/ and return albums plus diagnostics for the status screen.
   * Never throws — failures are reported in `error` / `detail`.
   */
  public static scan (assetPack?: string): ScanResult {
    const empty: ScanResult = { albums: [], dir: null, error: null, detail: '' };
    try {
      let fs: any;
      let path: any;
      try {
        fs = require('fs');
        path = require('path');
      } catch (err) {
        return {
          albums: [],
          dir: null,
          error: 'Cannot access the filesystem (require("fs") / require("path") failed).',
          detail: MusicLibrary.formatErr(err)
        };
      }

      const dir = MusicLibrary.resolveMusicDir(assetPack);
      if (!dir) {
        return {
          albums: [],
          dir: null,
          error: 'Could not resolve the music/ folder path.',
          detail: 'assetPack=' + String(assetPack) +
            '\nPathUtils.getAssetUri and __dirname fallback both failed.'
        };
      }

      if (!fs.existsSync(dir)) {
        return {
          albums: [],
          dir,
          error: 'music/ folder not found on disk.',
          detail: 'Looked for:\n' + dir +
            '\n\nCreate album folders under that path, e.g.\n' +
            'music/CHASER/cover.png\nmusic/CHASER/song.flac'
        };
      }

      let entries: string[];
      try {
        entries = fs.readdirSync(dir);
      } catch (err) {
        return {
          albums: [],
          dir,
          error: 'Could not read the music/ folder.',
          detail: 'Path: ' + dir + '\n' + MusicLibrary.formatErr(err)
        };
      }

      const albums: Album[] = [];
      const skipped: string[] = [];

      for (let i = 0; i < entries.length; i++) {
        const name = entries[i];
        if (name.charAt(0) === '.') { continue; }
        const albumPath = path.join(dir, name);
        let stat: any;
        try { stat = fs.statSync(albumPath); } catch (e) { continue; }
        if (!stat || !stat.isDirectory()) {
          skipped.push(name + ' (not a folder — albums must be folders)');
          continue;
        }

        try {
          const album = MusicLibrary.scanAlbum(name, albumPath, assetPack, fs, path);
          if (album && album.tracks.length > 0) {
            albums.push(album);
          } else {
            skipped.push(name + ' (no playable audio files)');
          }
        } catch (err) {
          skipped.push(name + ' (error: ' + (err && err.message ? err.message : String(err)) + ')');
        }
      }

      albums.sort((a, b) => a.title.toLowerCase().localeCompare(b.title.toLowerCase()));
      console.log('[jukebox] found', albums.length, 'album(s) in', dir);

      const detailLines = [
        'music dir: ' + dir,
        'assetPack: ' + String(assetPack),
        'albums: ' + albums.length
      ];
      if (skipped.length) {
        detailLines.push('skipped:');
        for (let s = 0; s < skipped.length; s++) {
          detailLines.push('  - ' + skipped[s]);
        }
      }

      // Empty library is not a hard error — the player shows its own empty state.
      // Still return detail so open() can log it.
      return {
        albums,
        dir,
        error: null,
        detail: detailLines.join('\n')
      };
    } catch (err) {
      console.error('[jukebox] failed to scan music folder:', err);
      return {
        albums: [],
        dir: null,
        error: 'Unexpected error while scanning music/.',
        detail: MusicLibrary.formatErr(err)
      };
    }
  }

  protected static formatErr (err: any): string {
    if (!err) { return String(err); }
    const msg = err.message || String(err);
    const stack = err.stack ? '\n' + err.stack : '';
    return msg + stack;
  }

  protected static scanAlbum (
    folderName: string,
    albumPath: string,
    assetPack: string,
    fs: any,
    path: any
  ): Album {
    const files: string[] = fs.readdirSync(albumPath);
    const tracks: Track[] = [];
    let coverFile: string = null;

    const lowerFiles = files.map((f: string) => ({ raw: f, lower: f.toLowerCase() }));

    for (let c = 0; c < COVER_NAMES.length && !coverFile; c++) {
      const want = COVER_NAMES[c];
      for (let f = 0; f < lowerFiles.length; f++) {
        if (lowerFiles[f].lower === want) {
          coverFile = lowerFiles[f].raw;
          break;
        }
      }
    }

    const audioFiles = files
      .filter((name: string) => AUDIO_EXT.test(name))
      .sort((a: string, b: string) => a.toLowerCase().localeCompare(b.toLowerCase()));

    for (let i = 0; i < audioFiles.length; i++) {
      const file = audioFiles[i];
      const rel = folderName + '/' + file;
      tracks.push({
        file: rel,
        title: MusicLibrary.prettifyName(file),
        format: MusicLibrary.formatLabel(file),
        url: MusicLibrary.resolveAssetUrl('music/' + rel, assetPack)
      });
    }

    return {
      id: folderName,
      title: MusicLibrary.prettifyFolder(folderName),
      coverUrl: coverFile
        ? MusicLibrary.resolveAssetUrl('music/' + folderName + '/' + coverFile, assetPack)
        : null,
      tracks
    };
  }

  /**
   * Absolute on-disk path to the `music/` directory. Resolves it the same way
   * jibo resolves view assets (PathUtils.getAssetUri with this asset pack) and
   * strips the `file://` scheme so `fs` can read it.
   */
  protected static resolveMusicDir (assetPack?: string): string {
    try {
      const PathUtils: any = (jibo as any).utils && (jibo as any).utils.PathUtils;
      if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
        const uri = PathUtils.getAssetUri('music', assetPack);
        if (uri) { return MusicLibrary.uriToPath(uri); }
      }
    } catch (err) {
      console.warn('[jukebox] resolveMusicDir via PathUtils failed:', err);
    }
    try {
      const path: any = require('path');
      return path.join(__dirname, '..', '..', 'music');
    } catch (err) {
      return null;
    }
  }

  protected static resolveAssetUrl (relPath: string, assetPack?: string): string {
    try {
      const PathUtils: any = (jibo as any).utils && (jibo as any).utils.PathUtils;
      if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
        const uri = PathUtils.getAssetUri(relPath, assetPack);
        if (uri) { return uri; }
      }
    } catch (err) {
      console.warn('[jukebox] resolveAssetUrl failed for', relPath, err);
    }
    return './' + relPath.split('/').map(encodeURIComponent).join('/');
  }

  protected static uriToPath (uri: string): string {
    let p = String(uri);
    if (p.indexOf('file://') === 0) {
      p = p.replace(/^file:\/\//, '');
    }
    try { p = decodeURIComponent(p); } catch (e) { /* leave as-is */ }
    return p;
  }

  protected static prettifyName (file: string): string {
    return file
      .replace(/\.(mp3|opus|ogg|oga|flac)$/i, '')
      .replace(/_/g, ' ')
      .trim();
  }

  protected static prettifyFolder (name: string): string {
    return String(name).replace(/_/g, ' ').trim();
  }

  protected static formatLabel (file: string): string {
    const match = file.match(/\.([^.]+)$/);
    if (!match) { return ''; }
    const ext = match[1].toLowerCase();
    if (ext === 'oga') { return 'OGG'; }
    return ext.toUpperCase();
  }
}
