// @be/jukebox/src/models/MusicLibrary.ts
import jibo = require('jibo');

export interface Track {
  title: string;
  url: string;
  file: string;
}

/**
 * Discovers the .mp3 files sitting in the skill's `music/` folder at runtime.
 *
 * Nothing here is bundled: the folder is read from disk every time the skill
 * opens, so a user can add songs simply by dropping files into `music/` (and
 * redeploying, or copying straight onto the robot) with no rebuild.
 */
export default class MusicLibrary {

  /**
   * Return the playable tracks found in `music/`, sorted by name. Always returns
   * an array (empty if the folder is missing or unreadable) so callers never
   * have to guard against a throw.
   */
  public static scan (assetPack?: string): Track[] {
    try {
      const fs: any = require('fs');
      const dir = MusicLibrary.resolveMusicDir(assetPack);
      if (!dir || !fs.existsSync(dir)) {
        console.warn('[jukebox] music dir not found:', dir);
        return [];
      }

      const files: string[] = fs.readdirSync(dir)
        .filter((name: string) => /\.mp3$/i.test(name))
        .sort((a: string, b: string) => a.toLowerCase().localeCompare(b.toLowerCase()));

      const tracks = files.map((file: string) => ({
        file,
        title: MusicLibrary.prettifyName(file),
        url: MusicLibrary.resolveTrackUrl(file, assetPack)
      }));

      console.log('[jukebox] found', tracks.length, 'track(s) in', dir);
      return tracks;
    } catch (err) {
      console.error('[jukebox] failed to scan music folder:', err);
      return [];
    }
  }

  /**
   * Absolute on-disk path to the `music/` directory. Resolves it the same way
   * jibo resolves view assets (PathUtils.getAssetUri with this asset pack) and
   * strips the `file://` scheme so `fs` can read it.
   */
  protected static resolveMusicDir (assetPack?: string): string {
    try {
      const PathUtils: any = (jibo as any).utils && (jibo as any).utils.PathUtils;
      if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
        const uri = PathUtils.getAssetUri('music', assetPack);
        if (uri) { return MusicLibrary.uriToPath(uri); }
      }
    } catch (err) {
      console.warn('[jukebox] resolveMusicDir via PathUtils failed:', err);
    }
    // Fallback: relative to this bundle's directory on disk.
    try {
      const path: any = require('path');
      return path.join(__dirname, '..', '..', 'music');
    } catch (err) {
      return null;
    }
  }

  /**
   * URL for the DOM <audio> element's `src`. Uses the asset-pack resolver (which
   * yields a `file://` URL the renderer can load) and falls back to a relative
   * path.
   */
  protected static resolveTrackUrl (file: string, assetPack?: string): string {
    try {
      const PathUtils: any = (jibo as any).utils && (jibo as any).utils.PathUtils;
      if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
        const uri = PathUtils.getAssetUri('music/' + file, assetPack);
        if (uri) { return uri; }
      }
    } catch (err) {
      console.warn('[jukebox] resolveTrackUrl failed for', file, err);
    }
    return './music/' + encodeURIComponent(file);
  }

  protected static uriToPath (uri: string): string {
    let p = String(uri);
    if (p.indexOf('file://') === 0) {
      p = p.replace(/^file:\/\//, '');
    }
    try { p = decodeURIComponent(p); } catch (e) { /* leave as-is */ }
    return p;
  }

  protected static prettifyName (file: string): string {
    return file
      .replace(/\.mp3$/i, '')
      .replace(/_/g, ' ')
      .trim();
  }
}
