// jibo-emotion-system/src/expression/EmotionalExpression.ts
/*
 * @fileOverview
 * EmotionalAction
 * ===============
 * Emotional actions are actions that are performed as a result of Jibo's
 * desire and ability to express himself.
 */
import {EMOTION_TO_ANIM_CATEGORY} from "../emotion/EmotionDefinitions";
import {EmotionSystem} from "../EmotionSystem";
import {Emotion} from "../emotion/Emotion";
import {EmotionName} from "../common/Types";

import parentLog from '../log';
const log = parentLog.createChild('Expression.Express');

export {}

/**
 * @class EmotionalExpression
 * @description A single emotional expression
 * @memberof jibo.emotion.expression
 * @intdocs
 * @internal
 */
export class EmotionalExpression {

    /**
     * The name of the emotion which this expression represents
     * @type {string}
     * @name jibo.emotion.expression.EmotionalExpression#name
     */
    public name: string;
    private parent: EmotionSystem;

    /** @override */
    constructor(name: string, parent: EmotionSystem) {
        this.name = name;
        this.parent = parent;
    }

    /**
     * @param {jibo.emotion.Emotion} emotion The emotion which Jibo will attempt to express.
     * @return {Promise} Resolved when the emotional action is finished being performed.
     * @method jibo.emotion.expression.EmotionalExpression#run
     */
    run(emotion: Emotion): Promise<any> {
        const jibo = this.parent.jibo;
        return new Promise((res, rej) => {
            const animCategory = EMOTION_TO_ANIM_CATEGORY.get(EmotionName[emotion.name]);
            //log.debug(`Jibo queued a ${animCategory} animation!`);
            if (!animCategory) {
                rej(`No animation category with corresponding emotion name ${emotion.name}`);
            } else {
                const state = jibo.action.getState();
                if (state.sleep === jibo.action.types.SleepState.AWAKE) {
                    // We create an animation goal to be played only 
                    // if the current skill is Idle, and gets culled if not selected
                    // within 10 seconds. 
                    const endTime = new Date(state.time.getTime());
                    endTime.setSeconds(endTime.getSeconds() + 10);
                    jibo.action.addPlayAnimationGoal({
                        category: animCategory,
                        criteria: {
                            endTime,
                            currentSkill: '@be/idle',
                        }
                    });
                } else {
                    log.debug(`Not playing animation category '${animCategory}' because we are not awake`);
                }
                res();
            }
        });
    }
}
