// jibo-emotion-system/src/expression/ExpressionSystem.ts
/*
 * @fileOverview
 *
 * ExpressionSystem
 * ================
 * The expression system is where emotion expression is generated and output
 * to their respective outputs.  For example,
 *
 */

import {EmotionalExpression} from './EmotionalExpression';
import {EmotionSpace} from '../emotion/EmotionSpace';
import {EmotionSystem} from "../EmotionSystem";
import {Emotion} from "../emotion/Emotion";
import {EmotionName, RuleNames, AnimationOptions, TouchTypes} from "../common/Types";
import {EmotionTrace} from "../emotion/EmotionTrace";
import {TouchExpressions} from "./ExpressionMaps";

import parentLog from '../log';
const log = parentLog.createChild('Expression');

export {}

/**
 * @description Management and coordination of jibo's expressions.
 * @namespace jibo.emotion.expression
 * @private
 * @internal
 */
export class ExpressionSystem {

    public emotionExpressions: Map<EmotionName, EmotionalExpression>;

    constructor(private parent: EmotionSystem) {
        this.emotionExpressions = new Map<EmotionName, EmotionalExpression>();
    }

    init(emotionExpressions: Map<EmotionName, EmotionalExpression>) {
        this.emotionExpressions = emotionExpressions;
    }

    /**
     * Verify that all emotions in a given emotional space have valid expressions
     * in this expression system instance.
     * @param {jibo.emotion.emotionSpace} emotionSpace emotionSpace to verify against
     * @method jibo.emotion.expression#verifyExpressionsExist
     */
    public verifyExpressionsExist(emotionSpace: EmotionSpace) {
        emotionSpace.emotions.forEach(e => {
            let expression = this.emotionExpressions.get(e.name);
            if (!expression) {
                throw new Error(`No Expression defined for emotion: "${e.name}"`);
            }
        });
    }

    /**
     * Express a reaction to an emotional trace
     * @param {jibo.emotion.EmotionTrace} emotionTrace The emotion to express in expressions.
     * @method jibo.emotion.expression#express
     */
    public express(emotionTrace: EmotionTrace): void {
        //log.debug("expressing trace", emotionTrace);
        const emotionBeforeUpdate:EmotionName = this.parent.getNearestEmotion(emotionTrace.startingPosition).name;
        if (emotionTrace.appraisalRule && emotionTrace.appraisalRule.name === RuleNames.TouchRule) {
            const touchData = emotionTrace.sensorDataWrappers[0].sensorData;
            let snuggleType: string;
            if (touchData.type === TouchTypes.HEAD_LEFT || touchData.type === TouchTypes.HEAD_RIGHT) {
                snuggleType = "Head Pat";
            } else if (touchData.type === TouchTypes.TICKLE) {
                snuggleType = "Tickle";
            }
            if (snuggleType) {
                this.parent.jibo.analytics.track("Snuggle Occurred", { type: snuggleType });
            }
            this.expressAnimation(TouchExpressions.get(touchData.type));
        }
        else if (emotionBeforeUpdate !== this.parent.getNearestEmotion().name){
            this.expressEmotion(emotionTrace.nearestEmotionAfterImpact);
        }
    }

    /**
     * Express a given emotion
     * @param {jibo.emotion.Emotion} emotion The emotion to express in expressions.
     * @return {Promise} Promise that is resolved when Jibo's expression has
     *   finished.
     * @method jibo.emotion.expression#expressEmotion
     */
    public expressEmotion(emotion: Emotion): Promise<any> {
        let expression = this.emotionExpressions.get(emotion.name);
        //log.debug("expressing", expression);
        if (!expression) {
            throw new Error(`No Expression defined for emotion ${emotion.name}.`);
        }
        return expression.run(emotion);
    }

    /**
     * Cleans up and disconnects all allocated resources.
     * @method jibo.emotion.expression#dispose
     */
    public dispose() {
        this.emotionExpressions.clear();
    }

    /**
     * Express the animation based on animation options
     * @method jibo.emotion.expression#expressAnimation
     * @private
     */
    private expressAnimation(options:AnimationOptions) {
        this.parent.jibo.action.addPlayAnimationGoal(options);
    }
}
