// jibo-client-framework/lib/jibo-client-framework.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboClientFramework = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
const events_1 = require('events');
const http = require('http');
/**
 * HTTP Client
 * @class HTTPClient
 * @extends EventEmitter
 * @param {String} host Registry host
 * @param {String} port Registry port
 */
class HTTPClient extends events_1.EventEmitter {
    /**
     * Constructor
     */
    constructor(host, port) {
        super();
        this.host = host;
        this.port = port;
        /**
         * Host
         * @name HTTPClient#host
         * @type {String}
         */
        /**
         * Port
         * @name HTTPClient#port
         * @type {Number}
         */
    }
    /**
     * Send a request.
     * @method HTTPClient#sendRequest
     * @param {String} method
     * @param {String} path
     * @param {String} body
     * @param {Function} callback
     */
    sendRequest(method, path, body, callback) {
        let options = {
            host: this.host,
            port: this.port,
            path: path,
            method: method,
            headers: {
                "Content-Type": "application/json",
                "Content-Length": body.length
            }
        };
        let request = http.request(options, (response) => {
            response.setEncoding('utf8');
            let data = '';
            response.on('data', (chunk) => {
                data += chunk.toString();
            });
            response.on('end', () => {
                if (data.length === 0) {
                    return callback(null);
                }
                callback(null, JSON.parse(data));
            });
        });
        request.on('error', (e) => {
            callback(new Error('problem with request: ' + e.message));
        });
        request.write(body);
        request.end();
    }
    /**
     * Post JSON data.
     * @method HTTPClient#postJSON
     * @param {String} path The API path. Ex. /media/recording/start
     * @param {Object} json JSON payload
     * @param {Function} callback
     */
    postJSON(path, json, callback) {
        this.sendRequest('post', path, JSON.stringify(json), callback);
    }
    /**
     * Get JSON from request.
     * @method HTTPClient#getJSON
     * @param {String} path The API path. Ex. /media/recording/start
     * @param {Object} json JSON payload
     * @param {Function} callback
     */
    getJSON(path, json, callback) {
        this.sendRequest('get', path, JSON.stringify(json), callback);
    }
    get(path, callback) {
        let request = http.get('http://' + this.host + ':' + this.port + path, (response) => {
            let data = '';
            response.on('data', (chunk) => {
                data += chunk.toString();
            });
            response.on('end', () => {
                callback(null, JSON.parse(data));
            });
        });
        request.on('error', (e) => {
            callback(new Error(`Got error: ${e.message}`));
        });
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = HTTPClient;

},{"events":undefined,"http":undefined}],2:[function(require,module,exports){
"use strict";
const HTTPClient_1 = require("./HTTPClient");
/**
 * Registry Client
 * @class RegistryClient
 * @extends HTTPClient
 * @param {String} host Registry host
 * @param {String} port Registry port
 */
class RegistryClient extends HTTPClient_1.default {
    constructor(host, port) {
        super(host, port);
        this.host = host;
        this.port = port;
        if (RegistryClient._instance) {
            throw new Error('Registry Client is a singleton');
        }
        RegistryClient._instance = this;
    }
    static createInstance(host, port) {
        return new RegistryClient(host, port);
    }
    static get instance() {
        return RegistryClient._instance;
    }
    /**
     * Add new record.
     * @method RegistryClient#addNewRecord
     * @param {RegistrationRecord} record
     * @param {Function} callback
     */
    addNewRecord(record, callback) {
        this.sendRequest('PUT', '/registry', JSON.stringify(record), callback);
    }
    /**
     * Edit record.
     * @method RegistryClient#editRecord
     * @param {RegistrationRecord} record
     * @param {Function} callback
     */
    editRecord(record, callback) {
        this.sendRequest('POST', '/registry', JSON.stringify(record), callback);
    }
    /**
     * Delete record.
     * @method RegistryClient#deleteRecord
     * @param {RegistrationRecord} record
     * @param {Function} callback
     */
    deleteRecord(record, callback) {
        this.sendRequest('DELETE', '/registry', JSON.stringify(record), callback);
    }
    /**
     * Get record.
     * @method RegistryClient#getRecords
     * @param {Function} callback
     */
    getRecords(callback) {
        this.get('/registry', (error, records) => {
            callback(error, records.records);
        });
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RegistryClient;

},{"./HTTPClient":1}],3:[function(require,module,exports){
"use strict";
const WebSocket = require('ws-without-native');
const events_1 = require("events");
/**
 * HTTP WebSocket Client
 * @class HTTPWSClient
 * @extends EventEmitter
 * @param {String} host
 */
class HTTPWSClient extends events_1.EventEmitter {
    constructor(host) {
        super();
        this.host = host;
        this.isReconnecting = false;
        this.onStartReconnect = this.onStartReconnect.bind(this);
        this.onError = this.onError.bind(this);
        this.onClose = this.onClose.bind(this);
        this.onMessage = this.onMessage.bind(this);
        this.connect = this.connect.bind(this);
        this.onOpen = this.onOpen.bind(this);
        this.connect();
    }
    /**
     * Send data.
     * @method HTTPWSClient#send
     * @param {String|Object} json Either raw data or data object.
     */
    send(json) {
        if (typeof json !== 'string') {
            try {
                json = JSON.stringify(json);
            }
            catch (e) {
                console.error('JSON.stringify: ', e);
                json = null;
            }
        }
        if (this.socket) {
            this.socket.send(json);
        }
    }
    connect() {
        //console.log('connect');
        this.socket = new WebSocket(this.host);
        this.socket.on('error', this.onError);
        this.socket.on('close', this.onClose);
        this.socket.on('message', this.onMessage);
        this.socket.on('open', this.onOpen);
    }
    /**
     * Called when there is an error to try reconnect
     */
    onStartReconnect() {
        console.log('onStartReconnect');
        this.isReconnecting = true;
        this.socket.removeAllListeners();
        this.socket = null;
        setTimeout(this.connect, 1000);
    }
    onError(err) {
        console.warn('socket error', err);
        this.onStartReconnect();
    }
    onClose(code, message) {
        console.warn('socket closed', code, message);
        this.emit('close');
        this.onStartReconnect();
    }
    onMessage(data) {
        try {
            // TOFIX: Take this out when the platform service
            // fixes on their end; really should throw an error when we receive NaN.
            let command = JSON.parse(data.replace(/\bNaN\b/g, "0"));
            this.emit('message', command);
        }
        catch (e) {
            console.warn('Invalid JSON sent to jibo client on host ' + this.host, data);
            this.emit('error', 'json ' + e, data);
        }
    }
    onOpen() {
        this.emit('open');
        if (this.isReconnecting) {
            this.isReconnecting = false;
            this.emit('reopen');
        }
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = HTTPWSClient;

},{"events":undefined,"ws-without-native":undefined}],4:[function(require,module,exports){
"use strict";
const HTTPClient_1 = require("./HTTPClient");
const WSClient_1 = require("./WSClient");
const RegistryClient_1 = require("./RegistryClient");
class HTTPClient extends HTTPClient_1.default {
    constructor(host, port) {
        super(host, port);
    }
}
exports.HTTPClient = HTTPClient;
class WSClient extends WSClient_1.default {
    constructor(host) {
        super(host);
    }
}
exports.WSClient = WSClient;
class RegistryClient extends RegistryClient_1.default {
    constructor(host, port) {
        super(host, port);
    }
    static get instance() {
        return RegistryClient_1.default.instance;
    }
}
exports.RegistryClient = RegistryClient;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {};

},{"./HTTPClient":1,"./RegistryClient":2,"./WSClient":3}]},{},[4])(4)
});
