// jibo-anim-db/lib/jibo-anim-db.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboAnimDb = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var Utils_1 = require("./Utils");
var Classes_1 = require("./Classes");
var AnimUtils_1 = require("./AnimUtils");
var AnimCollection = (function () {
    function AnimCollection(parent) {
        this.parent = parent;
        this.animations = [];
        this.categories = [];
        this._animationsByName = new Map();
        this._categoriesByName = new Map();
    }
    AnimCollection.prototype.addAnimation = function (animation) {
        var nameLower = animation.name.toLowerCase();
        if (!this._animationsByName.has(nameLower)) {
            this._animationsByName.set(nameLower, animation);
            this.animations.push(animation);
        }
    };
    AnimCollection.prototype.addCategory = function (category) {
        var nameLower = category.name.toLowerCase();
        if (!this._categoriesByName.has(nameLower)) {
            this._categoriesByName.set(nameLower, category);
            this.categories.push(category);
        }
    };
    AnimCollection.prototype.getCategoryByName = function (name) {
        return this._categoriesByName.get(name.toLowerCase());
    };
    AnimCollection.prototype.getAnimByName = function (name) {
        return this._animationsByName.get(name.toLowerCase());
    };
    AnimCollection.prototype.getAnimationNames = function () {
        return Array.from(this._animationsByName.keys());
    };
    AnimCollection.prototype.getAnimationCategories = function () {
        return Array.from(this._categoriesByName.keys());
    };
    AnimCollection.prototype.search = function (searchTerm) {
        var term = searchTerm.toLowerCase();
        var res = new Classes_1.SearchResults();
        this._search(term, res);
        return res;
    };
    AnimCollection.prototype.query = function (query) {
        AnimUtils_1.AnimUtils.insertDefaultAnimQueryParams(query);
        var res = new Classes_1.AnimResults();
        this._query(query, res);
        var matchingNoDupes = [];
        new Set(res.matching).forEach(function (elm) { return matchingNoDupes.push(elm); });
        res.matching = matchingNoDupes;
        var nonMatchingNoDupes = [];
        new Set(res.nonMatching).forEach(function (elm) { return nonMatchingNoDupes.push(elm); });
        res.nonMatching = nonMatchingNoDupes;
        return res;
    };
    AnimCollection.prototype._search = function (term, res) {
        var catByName = this._categoriesByName.get(term);
        if (catByName) {
            res.category = catByName;
        }
        var animByName = this._animationsByName.get(term);
        if (animByName) {
            res.name = animByName;
        }
        this.animations.forEach(function (anim) {
            anim.meta.metaTerms.forEach(function (meta) {
                if (meta === term) {
                    res.meta.push(anim);
                }
            });
        });
    };
    AnimCollection.prototype._query = function (query, res) {
        var _this = this;
        if (query.categories.length) {
            query.categories.forEach(function (catName) {
                var category = _this._categoriesByName.get(catName);
                if (category) {
                    category._query(query, res);
                }
            });
        }
        else {
            Utils_1.Utils.getAnimationsByQuery(this.animations, res, query);
        }
    };
    return AnimCollection;
}());
exports.AnimCollection = AnimCollection;

},{"./AnimUtils":4,"./Classes":7,"./Utils":10}],2:[function(require,module,exports){
"use strict";
var Classes_1 = require("./Classes");
var AnimUtils_1 = require("./AnimUtils");
var AnimDB = (function () {
    function AnimDB(jibo) {
        this.collections = [];
        this.jibo = jibo;
    }
    AnimDB.prototype.push = function (animCollection) {
        this.collections.push(animCollection);
    };
    AnimDB.prototype.pop = function () {
        return this.collections.pop();
    };
    AnimDB.prototype.getAnimByName = function (animName) {
        var anim;
        for (var i = 0; i < this.collections.length; i++) {
            var animDB = this.collections[i];
            var a = animDB.getAnimByName(animName);
            if (a) {
                anim = a;
                break;
            }
        }
        return anim;
    };
    AnimDB.prototype.getAnimationNames = function () {
        var names = [];
        for (var i = 0; i < this.collections.length; i++) {
            names = names.concat(this.collections[i].getAnimationNames());
        }
        return names;
    };
    AnimDB.prototype.getAnimationCategories = function () {
        var names = [];
        for (var i = 0; i < this.collections.length; i++) {
            names = names.concat(this.collections[i].getAnimationCategories());
        }
        return names;
    };
    AnimDB.prototype.search = function (searchTerm) {
        var term = searchTerm.toLowerCase();
        var res = new Classes_1.SearchResults();
        this.collections.forEach(function (animColl) {
            animColl._search(term, res);
        });
        return res;
    };
    AnimDB.prototype.query = function (query) {
        AnimUtils_1.AnimUtils.insertDefaultAnimQueryParams(query);
        var res = new Classes_1.AnimResults();
        this.collections.forEach(function (animColl) {
            animColl._query(query, res);
        });
        var matchingNoDupes = [];
        new Set(res.matching).forEach(function (elm) { return matchingNoDupes.push(elm); });
        res.matching = matchingNoDupes;
        var nonMatchingNoDupes = [];
        new Set(res.nonMatching).forEach(function (elm) { return nonMatchingNoDupes.push(elm); });
        res.nonMatching = nonMatchingNoDupes;
        return res;
    };
    return AnimDB;
}());
exports.AnimDB = AnimDB;

},{"./AnimUtils":4,"./Classes":7}],3:[function(require,module,exports){
"use strict";
var path = require("path");
var fs = require("fs");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var AnimDB_1 = require("./AnimDB");
var AnimCollection_1 = require("./AnimCollection");
var Animation_1 = require("./Animation");
var Category_1 = require("./Category");
var AnimUtils_1 = require("./AnimUtils");
var Utils_1 = require("./Utils");
var Range_1 = require("./Range");
var stringify = require('json-stable-stringify');
var AnimDBParser = (function () {
    function AnimDBParser() {
    }
    AnimDBParser.readAnimDB = function (jibo, animDBPath) {
        var moreAnimDBPaths = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            moreAnimDBPaths[_i - 2] = arguments[_i];
        }
        moreAnimDBPaths.unshift(animDBPath);
        var animDB = new AnimDB_1.AnimDB(jibo);
        var promises = moreAnimDBPaths.map(function (rootDir) {
            return AnimDBParser.readAndAddAnimCollection(animDB, rootDir);
        });
        return Promise.all(promises)
            .then(function () { return animDB; });
    };
    AnimDBParser.readAndAddAnimCollection = function (animDB, animDBPath) {
        var rootDir = path.dirname(animDBPath);
        return jibo_cai_utils_1.FileUtils.readFile(animDBPath)
            .then(function (animStr) {
            var animObjects = JSON.parse(animStr);
            var collection = new AnimCollection_1.AnimCollection(animDB);
            var _loop_1 = function (animName) {
                var animMeta = animObjects[animName];
                animName = animMeta.name.toLowerCase();
                var anim = new Animation_1.Animation(collection, animMeta, rootDir);
                var existingAnim = collection.getAnimByName(anim.name);
                if (existingAnim) {
                    var paths = [existingAnim.meta.path, animMeta.path];
                    throw new Error("Duplicate animation name: '" + animName + "' at paths: " + paths);
                }
                collection.addAnimation(anim);
                anim.meta.categories.forEach(function (categoryName) {
                    var cat = collection.getCategoryByName(categoryName);
                    if (!cat) {
                        cat = new Category_1.Category(categoryName);
                        collection.addCategory(cat);
                    }
                    cat.addAnimation(anim);
                });
            };
            for (var _i = 0, _a = Object.keys(animObjects); _i < _a.length; _i++) {
                var animName = _a[_i];
                _loop_1(animName);
            }
            animDB.push(collection);
        });
    };
    AnimDBParser.createAndWriteAnimMetaData = function (jibo, rootDir, outName) {
        return AnimDBParser.createAnimMetaData(jibo, rootDir)
            .then(function (allAnimData) {
            var outPath = path.join(rootDir, outName + '.json');
            var data = stringify(allAnimData, { space: '  ' });
            return jibo_cai_utils_1.PromiseUtils.promisify(function (h) { return fs.writeFile(outPath, data, 'utf8', h); });
        });
    };
    AnimDBParser.createAnimMetaData = function (jibo, rootDir) {
        var animsDir = path.join(rootDir, 'animations');
        return jibo_cai_utils_1.FileUtils.findAllFilesWithExt(animsDir, 'keys')
            .then(function (filePaths) {
            var goodPaths = filePaths.filter(function (filePath) {
                return !path.basename(filePath).startsWith('_');
            });
            var pr = goodPaths.map(function (data) {
                return AnimDBParser.parseAnimation(jibo, data);
            });
            return Promise.all(pr);
        })
            .then(function (_allAnimData) {
            var allAnimData = _allAnimData;
            var rootDirResolved = path.resolve(rootDir);
            var out = {};
            allAnimData.forEach(function (animData) {
                var firstPart = animData.path.substring(0, rootDirResolved.length);
                if (firstPart !== rootDirResolved) {
                    throw new Error("Animation doesn't have the expected root path: " + [animData.path, rootDirResolved]);
                }
                animData.path = animData.path.substring(rootDirResolved.length + path.sep.length);
                if (out[animData.name]) {
                    var paths = [out[animData.name].path, animData.path];
                    throw new Error("Duplicate animation name: '" + animData.name + "' at paths: " + paths);
                }
                out[animData.name] = animData;
            });
            return out;
        });
    };
    AnimDBParser.parseAnimation = function (jibo, filePath) {
        return AnimDBParser.loadKeys(jibo, filePath)
            .then(function (animData) {
            return AnimDBParser.parseAnimationFromAnim(animData, filePath);
        });
    };
    AnimDBParser.loadKeys = function (jibo, keysPath) {
        return new jibo.rendering.animation.KeysLoader(keysPath).load();
    };
    AnimDBParser.parseAnimationFromAnim = function (animData, filePath) {
        var animdbTag;
        var scale, speed, name;
        try {
            var defaultName = path.basename(filePath, path.extname(filePath));
            animdbTag = animData.animdb || {};
            name = Utils_1.Utils.parseString(animdbTag.name, defaultName);
            scale = [
                Utils_1.Utils.parseNumber(animdbTag.scaleMin, 1),
                Utils_1.Utils.parseNumber(animdbTag.scaleMax, 1)
            ];
            speed = [
                Utils_1.Utils.parseNumber(animdbTag.speedMin, 1),
                Utils_1.Utils.parseNumber(animdbTag.speedMax, 1)
            ];
        }
        catch (error) {
            throw new Error("Error parsing animation at path '" + filePath + "': '" + error + "'");
        }
        if ((speed[0] <= 0) || (speed[0] > speed[1])) {
            throw new Error("Invalid speed bounds '" + JSON.stringify(speed) + "' for animation '" + filePath + "'");
        }
        if (scale[0] > scale[1]) {
            throw new Error("Invalid scale bounds '" + JSON.stringify(scale) + "' for animation '" + filePath + "'");
        }
        var categoryString = animdbTag.categories || '';
        var categories = [];
        categoryString.split(',').forEach(function (category) {
            var catName = category.trim().toLowerCase();
            if (catName.length) {
                categories.push(catName);
            }
        });
        var metaString = animdbTag.meta || '';
        var metaTerms = [];
        metaString.split(',').forEach(function (_metaTerm) {
            var metaName = _metaTerm.trim().toLowerCase();
            if (metaName.length) {
                metaTerms.push(metaName);
            }
        });
        var holdSafeKF = AnimUtils_1.AnimUtils.hasHoldSafe(animData, name);
        var durationRange = [
            speed[0] * animData.duration,
            speed[1] * animData.duration
        ];
        if (holdSafeKF !== -1) {
            durationRange[1] = Range_1.INFINITE_DURATION;
        }
        var animMetadata = {
            name: name,
            scale: scale,
            speed: speed,
            durationRange: durationRange,
            categories: categories,
            metaTerms: metaTerms,
            holdSafeKF: holdSafeKF,
            path: filePath,
            duration: animData.duration,
            startsNeutral: AnimUtils_1.AnimUtils.startsNeutral(animData),
            endsNeutral: AnimUtils_1.AnimUtils.endsNeutral(animData),
            hasAudio: AnimUtils_1.AnimUtils.hasAudio(animData),
            orientation: AnimUtils_1.AnimUtils.calculateOrientation(animData),
            layers: AnimUtils_1.AnimUtils.checkLayers(animData)
        };
        return animMetadata;
    };
    return AnimDBParser;
}());
exports.AnimDBParser = AnimDBParser;

},{"./AnimCollection":1,"./AnimDB":2,"./AnimUtils":4,"./Animation":5,"./Category":6,"./Range":9,"./Utils":10,"fs":undefined,"jibo-cai-utils":undefined,"json-stable-stringify":undefined,"path":undefined}],4:[function(require,module,exports){
"use strict";
var Classes_1 = require("./Classes");
var LayerMocks_1 = require("./animation_formats/LayerMocks");
var BODY_ANGLE_EPSILON = 3;
var HOLD_SAFE_EVENT_NAME = 'HOLD_SAFE';
var AnimUtils = (function () {
    function AnimUtils() {
    }
    AnimUtils.syncAnimLayers = function (animData, layers) {
        for (var layer = animData.layers.length - 1; layer >= 0; layer--) {
            var type = animData.layers[layer].type;
            if (!Classes_1.LayerName[type]) {
                type = type.replace(/\s/g, '');
            }
            if (layers[type] !== undefined && !layers[type]) {
                animData.layers.splice(layer, 1);
            }
        }
        Object.keys(layers).forEach(function (layer) {
            if (!layers[layer]) {
                delete layers[layer];
            }
        });
        var layersToAdd = Object.keys(layers).filter(function (key) { return !AnimUtils.hasLayer(animData, Classes_1.LayerName[key], false); });
        layersToAdd.forEach(function (layerType) { return AnimUtils._addMockLayer(animData, Classes_1.LayerName[layerType].long || Classes_1.LayerName[layerType].short); });
    };
    AnimUtils._addMockLayer = function (animData, type) {
        animData.layers.push(LayerMocks_1.LayerTemplate.generateLayer(type, type));
    };
    AnimUtils.scaleTimeInPlace = function (animData, scale) {
        animData.layers.forEach(function (layer) {
            layer.keyframes.forEach(function (kf) {
                kf.time = Math.round(kf.time * scale);
            });
        });
        animData.duration = Math.round(animData.duration * scale);
    };
    AnimUtils.scaleBodyInPlace = function (animData, scale) {
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            layer.keyframes.forEach(function (kf) {
                kf.value.Head *= scale;
                kf.value.Pelvis *= scale;
                kf.value.Torso *= scale;
            });
        });
    };
    AnimUtils.flipScreenLeftRight = function (animData) {
        var handler = function (layer) {
            layer.keyframes.forEach(function (kf) {
                kf.value.Translate.x *= -1;
            });
        };
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Eye, handler);
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Overlay, handler);
    };
    AnimUtils.startsNeutral = function (animData) {
        var ret = true;
        var earliestFrame = animData.duration;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            if (layer.keyframes.length) {
                var kf = layer.keyframes[0];
                if (kf.time < earliestFrame) {
                    earliestFrame = kf.time;
                    if (Math.abs(kf.value.Pelvis) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Head) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Torso) > BODY_ANGLE_EPSILON) {
                        ret = false;
                    }
                }
            }
        });
        return ret;
    };
    AnimUtils.endsNeutral = function (animData) {
        var ret = true;
        var latestTime = -1;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            if (layer.keyframes.length) {
                var kf = layer.keyframes[layer.keyframes.length - 1];
                if (kf.time > latestTime) {
                    latestTime = kf.time;
                    if (Math.abs(kf.value.Pelvis) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Head) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Torso) > BODY_ANGLE_EPSILON) {
                        ret = false;
                    }
                }
            }
        });
        return ret;
    };
    AnimUtils.checkLayers = function (animData, strict) {
        if (strict === void 0) { strict = true; }
        var layerPresence = {};
        Classes_1.LayerName.Layers.forEach(function (layerName) {
            if (AnimUtils.hasLayer(animData, layerName, strict)) {
                layerPresence[layerName.short] = true;
            }
        });
        return layerPresence;
    };
    AnimUtils.hasAudio = function (animData) {
        return AnimUtils.hasLayer(animData, Classes_1.LayerName.AudioEvent);
    };
    AnimUtils.hasLayer = function (animData, layerName, strict) {
        if (strict === void 0) { strict = true; }
        var layerCheck = [];
        AnimUtils.forEachLayerType(animData, layerName, function (layer) {
            if (strict) {
                layerCheck.push(layer.keyframes.length !== 0);
            }
            else {
                layerCheck.push(true);
            }
        });
        return layerCheck.some(function (hasContent) { return hasContent; });
    };
    AnimUtils.hasHoldSafe = function (animData, name) {
        var kf = -1;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Event, function (layer) {
            layer.keyframes.forEach(function (layerKF) {
                if (layerKF.value.Event.name === HOLD_SAFE_EVENT_NAME) {
                    if (kf !== -1) {
                        throw new Error("Two HOLD_SAFE keyframes defined in animation" +
                            (" '" + name + "' at frames " + kf + " and " + layerKF.time));
                    }
                    kf = layerKF.time;
                }
            });
        });
        return kf;
    };
    AnimUtils.calculateOrientation = function (animData) {
        var headSum = 0;
        var pelvisSum = 0;
        var torsoSum = 0;
        var numFrames = 0;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            layer.keyframes.forEach(function (layerKF) {
                numFrames++;
                headSum += layerKF.value.Head;
                pelvisSum += layerKF.value.Pelvis;
                torsoSum += layerKF.value.Torso;
            });
        });
        headSum /= numFrames;
        pelvisSum /= numFrames;
        torsoSum /= numFrames;
        var bodySum = (torsoSum + pelvisSum) / 2;
        if (Math.abs(bodySum) > 10) {
            return (bodySum > 0) ? Classes_1.Orientation.LEFT : Classes_1.Orientation.RIGHT;
        }
        else {
            return (headSum > 0) ? Classes_1.Orientation.LEFT : Classes_1.Orientation.RIGHT;
        }
    };
    AnimUtils.forEachLayerType = function (animData, layerName, handler) {
        var name = layerName.long || layerName.short;
        var altName = name + ' #';
        animData.layers.forEach(function (layer) {
            if (layer.name === name || layer.name.startsWith(altName)) {
                handler(layer);
            }
        });
    };
    AnimUtils.insertDefaultAnimQueryParams = function (query) {
        if (!query.categories) {
            query.categories = [];
        }
        if (query.category) {
            query.categories.push(query.category);
        }
        for (var i = 0; i < query.categories.length; i++) {
            query.categories[i] = query.categories[i].toLowerCase();
        }
        if (!query.includeMeta) {
            query.includeMeta = [];
        }
        else {
            for (var i = 0; i < query.includeMeta.length; i++) {
                query.includeMeta[i] = query.includeMeta[i].toLowerCase();
            }
        }
        if (!query.excludeMeta) {
            query.excludeMeta = [];
        }
        else {
            for (var i = 0; i < query.excludeMeta.length; i++) {
                query.excludeMeta[i] = query.excludeMeta[i].toLowerCase();
            }
        }
        if (!query.durationError) {
            query.durationError = 0;
        }
        if (!query.duration) {
            query.duration = -1;
        }
        if (typeof query.hasAudio === 'undefined') {
            query.hasAudio = Classes_1.HasAudio.DONT_CARE;
        }
    };
    return AnimUtils;
}());
exports.AnimUtils = AnimUtils;

},{"./Classes":7,"./animation_formats/LayerMocks":12}],5:[function(require,module,exports){
"use strict";
var path = require("path");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var Range_1 = require("./Range");
var AnimUtils_1 = require("./AnimUtils");
var Utils_1 = require("./Utils");
var Playback_1 = require("./Playback");
var Transform = (function () {
    function Transform() {
        this.framesToHold = -1;
        this.speed = 1;
        this.scale = 1;
        this.shouldFlipLeftRight = false;
    }
    return Transform;
}());
exports.Transform = Transform;
var Animation = (function () {
    function Animation(parent, meta, resourceRoot, animCache) {
        this.parent = parent;
        this.meta = meta;
        this.resourceRoot = resourceRoot;
        this.animCache = animCache;
        this.name = meta.name;
        this.durationRange = new Range_1.Range(meta.durationRange[0], meta.durationRange[1]);
    }
    Animation.prototype.loadAnimationObject = function () {
        var _this = this;
        return jibo_cai_utils_1.FileUtils.readFile(path.join(this.resourceRoot, this.meta.path))
            .then(function (animDataStr) {
            var animData;
            try {
                animData = JSON.parse(animDataStr);
            }
            catch (error) {
                throw "Parsing error: '" + error + "' in animation at path: '" + _this.meta.path + "'";
            }
            return animData;
        });
    };
    Animation.prototype.createFromConfig = function (options) {
        var trans = this._getTransformation(options);
        return this._applyTransformation(trans);
    };
    Animation.prototype._applyTransformation = function (trans) {
        var _this = this;
        var animDataPr = (this.animCache) ?
            Promise.resolve(this.animCache) : this.loadAnimationObject();
        return animDataPr.then(function (animData) { return _this._applyTransformationToAnim(animData, trans); });
    };
    Animation.prototype._applyTransformationToAnim = function (animData, trans) {
        var hasFlippedBody = false;
        if (trans.layers) {
            AnimUtils_1.AnimUtils.syncAnimLayers(animData, trans.layers);
        }
        if (trans.scale !== 1) {
            var scale = trans.scale;
            if (trans.shouldFlipLeftRight) {
                scale *= -1;
                hasFlippedBody = true;
            }
            AnimUtils_1.AnimUtils.scaleBodyInPlace(animData, scale);
        }
        if (trans.shouldFlipLeftRight) {
            AnimUtils_1.AnimUtils.flipScreenLeftRight(animData);
            if (!hasFlippedBody) {
                AnimUtils_1.AnimUtils.scaleBodyInPlace(animData, -1);
            }
        }
        var duration = this.meta.duration;
        if (trans.speed !== -1) {
            duration *= trans.speed;
        }
        if (trans.framesToHold !== -1) {
            duration += trans.framesToHold;
        }
        return new Playback_1.Playback(duration, trans.speed, trans.framesToHold, this.resourceRoot, animData, this);
    };
    Animation.prototype._getTransformation = function (options) {
        var trans = new Transform();
        if (options) {
            if (options.orientation) {
                var reqOrientation = options.orientation;
                if (reqOrientation !== this.meta.orientation) {
                    trans.shouldFlipLeftRight = true;
                }
            }
            if (options.speed) {
                trans.speed = Utils_1.Utils.limitToRange(options.speed, this.meta.speed);
            }
            if (options.scale) {
                trans.scale = Utils_1.Utils.limitToRange(options.scale, this.meta.scale);
            }
            if (options.layers) {
                trans.layers = options.layers;
            }
            if (options.duration) {
                if (options.duration < this.meta.durationRange[0]) {
                    trans.speed = this.meta.speed[0];
                }
                else if (options.duration < trans.speed * this.meta.duration) {
                    trans.speed = options.duration / this.meta.duration;
                    trans.speed = Utils_1.Utils.limitToRange(trans.speed, this.meta.speed);
                }
                else if (this.meta.holdSafeKF !== -1) {
                    trans.framesToHold = options.duration - trans.speed * this.meta.duration;
                }
                else if (options.duration > trans.speed * this.meta.duration) {
                    if (typeof options.speed !== 'number') {
                        trans.speed = options.duration / this.meta.duration;
                        trans.speed = Utils_1.Utils.limitToRange(trans.speed, this.meta.speed);
                    }
                }
            }
        }
        return trans;
    };
    return Animation;
}());
exports.Animation = Animation;

},{"./AnimUtils":4,"./Playback":8,"./Range":9,"./Utils":10,"jibo-cai-utils":undefined,"path":undefined}],6:[function(require,module,exports){
"use strict";
var Utils_1 = require("./Utils");
var AnimUtils_1 = require("./AnimUtils");
var Classes_1 = require("./Classes");
var Category = (function () {
    function Category(name) {
        this.name = name;
        this.animations = [];
        this.animationsByName = new Map();
    }
    Category.prototype.addAnimation = function (anim) {
        this.animations.push(anim);
        this.animationsByName.set(anim.name, anim);
    };
    Category.prototype.query = function (query) {
        AnimUtils_1.AnimUtils.insertDefaultAnimQueryParams(query);
        var res = new Classes_1.AnimResults();
        this._query(query, res);
        return res;
    };
    Category.prototype._query = function (query, res) {
        Utils_1.Utils.getAnimationsByQuery(this.animations, res, query);
    };
    return Category;
}());
exports.Category = Category;

},{"./AnimUtils":4,"./Classes":7,"./Utils":10}],7:[function(require,module,exports){
"use strict";
var Orientation;
(function (Orientation) {
    Orientation[Orientation["LEFT"] = 'LEFT'] = "LEFT";
    Orientation[Orientation["RIGHT"] = 'RIGHT'] = "RIGHT";
})(Orientation = exports.Orientation || (exports.Orientation = {}));
var HasAudio;
(function (HasAudio) {
    HasAudio[HasAudio["YES"] = 'YES'] = "YES";
    HasAudio[HasAudio["NO"] = 'NO'] = "NO";
    HasAudio[HasAudio["DONT_CARE"] = 'DONT_CARE'] = "DONT_CARE";
})(HasAudio = exports.HasAudio || (exports.HasAudio = {}));
var AnimResults = (function () {
    function AnimResults() {
        this.matching = [];
        this.nonMatching = [];
    }
    return AnimResults;
}());
exports.AnimResults = AnimResults;
var SearchResults = (function () {
    function SearchResults() {
        this.meta = [];
    }
    return SearchResults;
}());
exports.SearchResults = SearchResults;
var LayerName = (function () {
    function LayerName(short, long) {
        this.short = short;
        this.long = long;
    }
    return LayerName;
}());
LayerName.Body = new LayerName('Body');
LayerName.Eye = new LayerName('Eye');
LayerName.EyeTexture = new LayerName('EyeTexture', 'Eye Texture');
LayerName.Overlay = new LayerName('Overlay');
LayerName.OverlayTexture = new LayerName('OverlayTexture', 'Overlay Texture');
LayerName.AudioEvent = new LayerName('AudioEvent', 'Audio Event');
LayerName.Event = new LayerName('Event');
LayerName.Pixi = new LayerName('Pixi');
LayerName.BackgroundTexture = new LayerName('BackgroundTexture', 'Background Texture');
LayerName.LED = new LayerName('LED');
LayerName.Glow = new LayerName('Glow');
LayerName.Lighting = new LayerName('Lighting');
LayerName.Layers = [
    LayerName.Body,
    LayerName.Eye,
    LayerName.EyeTexture,
    LayerName.Overlay,
    LayerName.OverlayTexture,
    LayerName.AudioEvent,
    LayerName.Event,
    LayerName.Pixi,
    LayerName.BackgroundTexture,
    LayerName.LED,
    LayerName.Glow,
    LayerName.Lighting
];
exports.LayerName = LayerName;
var BuilderLayer;
(function (BuilderLayer) {
    BuilderLayer[BuilderLayer["beat"] = 'beat'] = "beat";
    BuilderLayer[BuilderLayer["default"] = 'default'] = "default";
    BuilderLayer[BuilderLayer["posture"] = 'posture'] = "posture";
})(BuilderLayer = exports.BuilderLayer || (exports.BuilderLayer = {}));

},{}],8:[function(require,module,exports){
"use strict";
var jibo_cai_utils_1 = require("jibo-cai-utils");
var Classes_1 = require("./Classes");
var AnimUtils_1 = require("./AnimUtils");
var FRAMERATE = 30;
var Playback = (function () {
    function Playback(duration, speed, _framesToHold, _resourceRoot, _animData, _animation) {
        this.duration = duration;
        this.speed = speed;
        this._framesToHold = _framesToHold;
        this._resourceRoot = _resourceRoot;
        this._animData = _animData;
        this._animation = _animation;
        this._audioOnly = false;
    }
    Playback.prototype.init = function (options) {
        var jibo = this._animation.parent.parent.jibo;
        if (this._animData.layers.length === 1 && this._animData.layers[0].type === Classes_1.LayerName.AudioEvent.long) {
            AnimUtils_1.AnimUtils._addMockLayer(this._animData, Classes_1.LayerName.Body.short);
            this._audioOnly = true;
        }
        var keysPlaybackOptions = {
            jibo: jibo,
            animData: this._animData,
            rootPath: this._resourceRoot
        };
        keysPlaybackOptions.ownerInformation = (options && options.ownerInformation) ? options.ownerInformation : 'AnimDB';
        keysPlaybackOptions.dofOptions = (options && options.dofOptions) ? options.dofOptions : undefined;
        this.keysPlaybackPr = jibo_cai_utils_1.PlaybackUtils.createKeysPlayback(keysPlaybackOptions);
        return this.keysPlaybackPr;
    };
    Playback.prototype.play = function (options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        var jibo = this._animation.parent.parent.jibo;
        var holdHandler = function (signal, playInst) {
            if (playInst.setPaused) {
                playInst.setPaused(true);
                jibo_cai_utils_1.TimeUtils.wait(_this._framesToHold / FRAMERATE * 1000)
                    .then(function () { return playInst.setPaused(false); });
            }
        };
        var _keysPlaybackPr = this.keysPlaybackPr || this.init(options);
        var builderConfigFn = function (builder) {
            builder.setSpeed(_this.speed);
            builder.on('HOLD_SAFE', holdHandler);
            if (_this._audioOnly) {
                builder.setLayer(Classes_1.BuilderLayer.beat);
            }
            if (options.handleBuilder) {
                options.handleBuilder(builder);
            }
        };
        return _keysPlaybackPr.then(function (keysPlayback) {
            return keysPlayback.play(builderConfigFn, options.handleAnimation, options.screenCenterOverride);
        }).then(function () {
            if (options.forceReturnToIdle) {
                return jibo_cai_utils_1.PromiseUtils.promisify(function (h) {
                    return jibo.animate.centerRobot(jibo.animate.dofs.BODY, null, h);
                });
            }
        }).then(function () { return null; });
    };
    return Playback;
}());
exports.Playback = Playback;

},{"./AnimUtils":4,"./Classes":7,"jibo-cai-utils":undefined}],9:[function(require,module,exports){
"use strict";
exports.INFINITE_DURATION = 1e6;
var Range = (function () {
    function Range(min, max) {
        this.min = min;
        this.max = max;
        if (min > max) {
            throw new Error("Invalid bounds " + this.toString());
        }
    }
    Range.prototype.toString = function () {
        return "[" + this.min + ", " + this.max + "]";
    };
    Range.prototype.limit = function (num) {
        if (num < this.min) {
            return this.min;
        }
        else if (num > this.max) {
            return this.max;
        }
        else {
            return num;
        }
    };
    Range.prototype.isUpperInfinite = function () {
        return (this.max === exports.INFINITE_DURATION);
    };
    Range.prototype.distanceFrom = function (num) {
        if (num < this.min) {
            return this.min - num;
        }
        else if (num > this.max) {
            return num - this.max;
        }
        else {
            return 0;
        }
    };
    return Range;
}());
exports.Range = Range;

},{}],10:[function(require,module,exports){
"use strict";
var Classes_1 = require("./Classes");
var SemverUtils = require('semver-utils');
var Utils = (function () {
    function Utils() {
    }
    Utils.parseString = function (input, defaultVal) {
        if (!input || !(input.length)) {
            return defaultVal;
        }
        return input;
    };
    Utils.parseNumber = function (input, defaultVal) {
        if (typeof input === 'number') {
            return input;
        }
        if (!input || !(input.length)) {
            return defaultVal;
        }
        try {
            var num = parseFloat(input);
            return num;
        }
        catch (e) {
            return defaultVal;
        }
    };
    Utils.parseSemVer = function (input) {
        var parse = SemverUtils.parse(input);
        return {
            semver: parse.semver,
            version: parse.version,
            major: parseInt(parse.major),
            minor: parseInt(parse.minor),
            patch: parseInt(parse.patch),
            release: parse.release
        };
    };
    Utils.getAnimationsByQuery = function (animations, interimResults, query) {
        animations.forEach(function (anim) {
            if (Utils.audioMatch(anim, query.hasAudio) && Utils.metaMatch(anim, query.includeMeta, query.excludeMeta)) {
                Utils.addIfDurationMatch(anim, query.duration, query.durationError, interimResults);
            }
        });
    };
    Utils.audioMatch = function (animation, shouldHaveAudio) {
        return (shouldHaveAudio === Classes_1.HasAudio.DONT_CARE) ||
            (shouldHaveAudio === Classes_1.HasAudio.YES && animation.meta.hasAudio) ||
            (shouldHaveAudio === Classes_1.HasAudio.NO && !animation.meta.hasAudio);
    };
    Utils.metaMatch = function (animation, shouldIncludeMeta, shouldExcludeMeta) {
        if ((shouldIncludeMeta.length > 0) || (shouldExcludeMeta.length > 0)) {
            if ((shouldIncludeMeta.length > 0) && (shouldExcludeMeta.length > 0)) {
                return animation.meta.metaTerms.some(function (meta) { return shouldIncludeMeta.indexOf(meta) !== -1; }) &&
                    animation.meta.metaTerms.every(function (meta) { return shouldExcludeMeta.indexOf(meta) === -1; });
            }
            else if (shouldIncludeMeta.length > 0) {
                return animation.meta.metaTerms.some(function (meta) { return shouldIncludeMeta.indexOf(meta) !== -1; });
            }
            else if (shouldExcludeMeta.length > 0) {
                return animation.meta.metaTerms.every(function (meta) { return shouldExcludeMeta.indexOf(meta) === -1; });
            }
        }
        else {
            return true;
        }
    };
    Utils.addIfDurationMatch = function (animation, duration, durationError, res) {
        if (duration === -1) {
            res.matching.push(animation);
        }
        else {
            var dist = Utils.distanceFromRange(duration, animation.meta.durationRange);
            if (dist === 0) {
                res.matching.push(animation);
            }
            else if (dist <= durationError) {
                res.nonMatching.push(animation);
            }
        }
    };
    Utils.limitToRange = function (num, range) {
        if (num < range[0]) {
            return range[0];
        }
        else if (num > range[1]) {
            return range[1];
        }
        else {
            return num;
        }
    };
    Utils.distanceFromRange = function (num, range) {
        if (num < range[0]) {
            return range[0] - num;
        }
        else if (num > range[1]) {
            return num - range[1];
        }
        else {
            return 0;
        }
    };
    return Utils;
}());
exports.Utils = Utils;

},{"./Classes":7,"semver-utils":undefined}],11:[function(require,module,exports){
"use strict";

},{}],12:[function(require,module,exports){
"use strict";
var generateId = require('jibo-keyframes').generateId;
var LayerTemplate = (function () {
    function LayerTemplate() {
    }
    LayerTemplate.generateLayer = function (name, type) {
        var layer = {};
        layer.id = generateId();
        layer.name = name;
        layer.type = type;
        layer.visible = true;
        layer.locked = false;
        layer.keyframes = [];
        return layer;
    };
    return LayerTemplate;
}());
exports.LayerTemplate = LayerTemplate;

},{"jibo-keyframes":undefined}],13:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./LayerMocks"));
var format_0_6_x = require("./0_6_X");
exports.format_0_6_x = format_0_6_x;

},{"./0_6_X":11,"./LayerMocks":12}],14:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var AnimDBParser_1 = require("./AnimDBParser");
exports.AnimDBParser = AnimDBParser_1.AnimDBParser;
var AnimCollection_1 = require("./AnimCollection");
exports.AnimCollection = AnimCollection_1.AnimCollection;
var Animation_1 = require("./Animation");
exports.Animation = Animation_1.Animation;
var AnimDB_1 = require("./AnimDB");
exports.AnimDB = AnimDB_1.AnimDB;
var Playback_1 = require("./Playback");
exports.Playback = Playback_1.Playback;
__export(require("./animation_formats"));
__export(require("./Classes"));
function init(jibo, animDBPath) {
    var moreAnimDBPaths = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        moreAnimDBPaths[_i - 2] = arguments[_i];
    }
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!animDBPath) return [3 /*break*/, 1];
                    exports._animDB = new AnimDB_1.AnimDB(jibo);
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, AnimDBParser_1.AnimDBParser.readAnimDB.apply(AnimDBParser_1.AnimDBParser, [jibo, animDBPath].concat(moreAnimDBPaths))];
                case 2:
                    exports._animDB = _a.sent();
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
exports.init = init;
function getAnimByName(animName) {
    return exports._animDB.getAnimByName(animName);
}
exports.getAnimByName = getAnimByName;
function getAnimationNames() {
    return exports._animDB.getAnimationNames();
}
exports.getAnimationNames = getAnimationNames;
function getAnimationCategories() {
    return exports._animDB.getAnimationCategories();
}
exports.getAnimationCategories = getAnimationCategories;
function search(searchTerm) {
    return exports._animDB.search(searchTerm);
}
exports.search = search;
function query(query) {
    return exports._animDB.query(query);
}
exports.query = query;
function push(animCollection) {
    exports._animDB.push(animCollection);
}
exports.push = push;
function pop() {
    return exports._animDB.pop();
}
exports.pop = pop;

},{"./AnimCollection":1,"./AnimDB":2,"./AnimDBParser":3,"./Animation":5,"./Classes":7,"./Playback":8,"./animation_formats":13}],15:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./AnimDB"));
__export(require("./Range"));
__export(require("./AnimDBParser"));
__export(require("./AnimUtils"));
__export(require("./AnimCollection"));
__export(require("./Playback"));
__export(require("./Animation"));
__export(require("./Classes"));
__export(require("./AnimUtils"));
__export(require("./Utils"));
var api = require("./api");
exports.api = api;
var animation_formats = require("./animation_formats");
exports.animation_formats = animation_formats;

},{"./AnimCollection":1,"./AnimDB":2,"./AnimDBParser":3,"./AnimUtils":4,"./Animation":5,"./Classes":7,"./Playback":8,"./Range":9,"./Utils":10,"./animation_formats":13,"./api":14}],16:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./main"));

},{"./main":15}]},{},[16])(16)
});
