// jibo-common-types/lib/jibo-common-types.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboCommonTypes = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ResultStatus;
(function (ResultStatus) {
    ResultStatus.SUCCEEDED = "SUCCEEDED";
    ResultStatus.TIMEOUT = "TIMEOUT";
    ResultStatus.INTERRUPTED = "INTERRUPTED";
    ResultStatus.CANCELLED = "CANCELLED";
})(ResultStatus = exports.ResultStatus || (exports.ResultStatus = {}));
var AttentionMode;
(function (AttentionMode) {
    AttentionMode.OFF = "OFF";
    AttentionMode.IDLE = "IDLE";
    AttentionMode.DISENGAGE = "DISENGAGE";
    AttentionMode.ENGAGED = "ENGAGED";
    AttentionMode.SPEAKING = "SPEAKING";
    AttentionMode.FIXATED = "FIXATED";
    AttentionMode.ATTRACTABLE = "ATTRACTABLE";
    AttentionMode.MENU = "MENU";
    AttentionMode.COMMAND = "COMMAND";
})(AttentionMode = exports.AttentionMode || (exports.AttentionMode = {}));
var LEDMode;
(function (LEDMode) {
    LEDMode.OFF = "OFF";
    LEDMode.LISTEN = "LISTEN";
    LEDMode.ERROR = "ERROR";
})(LEDMode = exports.LEDMode || (exports.LEDMode = {}));
var IndexRobotResult;
(function (IndexRobotResult) {
    IndexRobotResult.SUCCEEDED = "SUCCEEDED";
    IndexRobotResult.TIMEOUT = "TIMEOUT";
    IndexRobotResult.FAULT = "FAULT";
})(IndexRobotResult = exports.IndexRobotResult || (exports.IndexRobotResult = {}));
var PlayStatus;
(function (PlayStatus) {
    PlayStatus.OK = "OK";
    PlayStatus.REJECTED = "REJECTED";
})(PlayStatus = exports.PlayStatus || (exports.PlayStatus = {}));
var ExpressionIds;
(function (ExpressionIds) {
    ExpressionIds[ExpressionIds["EXPRESSION"] = 0] = "EXPRESSION";
})(ExpressionIds = exports.ExpressionIds || (exports.ExpressionIds = {}));

},{}],2:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var uuid = require("uuid");
var GlobalCommand;
(function (GlobalCommand) {
    GlobalCommand.SLEEP = 'SLEEP';
    GlobalCommand.HELP = 'HELP';
    GlobalCommand.STOP = 'STOP';
    GlobalCommand.PAUSE = 'PAUSE';
    GlobalCommand.VOLUME = 'VOLUME';
    GlobalCommand.WHATCANIDO = 'WHATCANIDO';
    GlobalCommand.HOLDON = 'HOLDON';
})(GlobalCommand = exports.GlobalCommand || (exports.GlobalCommand = {}));
exports.GlobalCommandMap = new Map();
exports.GlobalCommandMap.set("sleep", GlobalCommand.SLEEP);
exports.GlobalCommandMap.set("help", GlobalCommand.HELP);
exports.GlobalCommandMap.set("stop", GlobalCommand.STOP);
exports.GlobalCommandMap.set("pause", GlobalCommand.PAUSE);
exports.GlobalCommandMap.set("volume", GlobalCommand.VOLUME);
exports.GlobalCommandMap.set("whatCanIDo", GlobalCommand.WHATCANIDO);
exports.GlobalCommandMap.set("holdOn", GlobalCommand.HOLDON);
var NLUResponse = (function () {
    function NLUResponse() {
    }
    return NLUResponse;
}());
exports.NLUResponse = NLUResponse;
var REQ_STATUS;
(function (REQ_STATUS) {
    REQ_STATUS[REQ_STATUS["OK"] = 'OK'] = "OK";
    REQ_STATUS[REQ_STATUS["ERROR"] = 'ERROR'] = "ERROR";
})(REQ_STATUS = exports.REQ_STATUS || (exports.REQ_STATUS = {}));
;
var NLUMessages;
(function (NLUMessages) {
    NLUMessages[NLUMessages["RESET_MEMORY"] = "NLU_EXECUTED_RESET_MEMORY"] = "RESET_MEMORY";
})(NLUMessages = exports.NLUMessages || (exports.NLUMessages = {}));
;
var ListenOptions = (function () {
    function ListenOptions(options) {
        this.hotphrase = options ? options.hotphrase : "";
        this.speech_to_text = options ? options.speech_to_text : false;
        this.parseIncremental = options ? options.parseIncremental : false;
        this.incremental = options ? options.incremental : false;
        this.language = options ? options.language : 'en-US';
        this.verbose = options ? options.verbose : false;
        this.forceGlobal = options ? !!options.forceGlobal : false;
        this.expectedDuration = options ? options.expectedDuration : 'medium';
        this.parseAllASR = options ? options.parseAllASR : false;
        this.metadata = options ? options.metadata : null;
    }
    return ListenOptions;
}());
exports.ListenOptions = ListenOptions;
var ParseResults = (function () {
    function ParseResults(Input, NLParse, launchParse, heuristic_score) {
        this.Input = Input || "";
        this.NLParse = NLParse;
        this.launchParse = launchParse;
        this.heuristic_score = heuristic_score ? heuristic_score : 0;
    }
    return ParseResults;
}());
exports.ParseResults = ParseResults;
var NoMatchResults = (function (_super) {
    __extends(NoMatchResults, _super);
    function NoMatchResults(parse, triggeredByHJ) {
        var _this = _super.call(this, parse.Input, parse.NLParse, parse.launchParse, parse.heuristic_score) || this;
        _this.triggeredByHJ = triggeredByHJ;
        return _this;
    }
    return NoMatchResults;
}(ParseResults));
exports.NoMatchResults = NoMatchResults;
var Utterance = (function () {
    function Utterance(options) {
        this.utterance = options.utterance || "";
        this.score = options.score || 0;
    }
    return Utterance;
}());
exports.Utterance = Utterance;
var ListenResults = (function () {
    function ListenResults(input, parse, otherResults, id) {
        if (!input) {
            input = {
                utterance: "",
                score: 0
            };
        }
        this.result = parse || new ParseResults(input.utterance);
        this.otherResults = otherResults || [];
        this.audioActionId = id || "";
    }
    return ListenResults;
}());
exports.ListenResults = ListenResults;
var PronunciationOptions = (function (_super) {
    __extends(PronunciationOptions, _super);
    function PronunciationOptions(word, listenerId, verbose, collectExample) {
        var _this = _super.call(this) || this;
        _this.word = word || "";
        _this.listenerId = listenerId || uuid.v1();
        _this.verbose = verbose || false;
        _this.collectExample = collectExample || true;
        return _this;
    }
    return PronunciationOptions;
}(ListenOptions));
exports.PronunciationOptions = PronunciationOptions;
var ASRRequestOptions = (function () {
    function ASRRequestOptions(options) {
        if (!options) {
            options = {};
        }
        this.authenticateSpeaker = options.authenticateSpeaker;
        this.bargein = options.bargein;
        this.detectEnd = options.detectEnd;
        this.expectedDuration = options.expectedDruation;
        this.heyJibo = options.heyJibo;
        this.incremental = options.incremental;
        this.metadata = options.metadata;
        this.nbest = options.nbest;
        this.parseFinal = options.parseFinal;
        this.parseIncremental = options.parseIncremental;
        this.precompiledHandle = options.precompiledHandle;
        this.speech_to_text = options.speech_to_text;
        this.hotphrase = options.hotphrase;
        this.language = options.language;
        this.known_speaker_id = options.known_speaker_id;
    }
    return ASRRequestOptions;
}());
exports.ASRRequestOptions = ASRRequestOptions;
var SpeakerIdStatus;
(function (SpeakerIdStatus) {
    SpeakerIdStatus.ACCEPTED = 'ACCEPTED';
    SpeakerIdStatus.NOT_USED = 'NOT_USED';
    SpeakerIdStatus.REJECTED = 'REJECTED';
    SpeakerIdStatus.ID_TI_NOT_AVAILABLE = 'ID_TI_NOT_AVAILABLE';
})(SpeakerIdStatus = exports.SpeakerIdStatus || (exports.SpeakerIdStatus = {}));
var SpeakerIdConfidence;
(function (SpeakerIdConfidence) {
    SpeakerIdConfidence.HIGH = 'HIGH-CONFIDENCE';
    SpeakerIdConfidence.LOW = 'LOW-CONFIDENCE';
})(SpeakerIdConfidence = exports.SpeakerIdConfidence || (exports.SpeakerIdConfidence = {}));
exports.NLU_STATUS = {
    CONNECTED: 'CONNECTED',
    DISCONNECTED: 'DISCONNECTED'
};
exports.ParseStatus = {
    NO_PARSE: 'NO-PARSE',
    GOT_PARSE: 'GOT-PARSE'
};
var SpeakerId = (function () {
    function SpeakerId(speakerId) {
        this.status = "";
        this.speakers = [];
        this.speakers = speakerId ? speakerId.speakers : [new Speaker()];
        this.status = speakerId ? speakerId.status : SpeakerIdStatus.NOT_USED;
        if (speakerId && speakerId.speaker_ids) {
            this.speakers = speakerId.speaker_ids;
            this.status = speakerId.speaker_id_status;
        }
    }
    return SpeakerId;
}());
exports.SpeakerId = SpeakerId;
var Speaker = (function () {
    function Speaker(speaker, score) {
        this.speaker_id = "";
        this.score = 0;
        this.speaker_id = speaker || "";
        this.score = score || 0;
    }
    return Speaker;
}());
exports.Speaker = Speaker;
exports.UTT_STATUS = {
    SAVING_AUDIO: 'SAVING_AUDIO'
};
exports.NoiseLevel = {
    NOISY_AUDIO: 'NOISY_AUDIO'
};
exports.LoudnessLevel = {
    TOO_LOUD_AUDIO: 'TOO_LOUD_AUDIO',
};
exports.AudioQuality = {
    BAD_AUDIO_SIGNAL: 'BAD_AUDIO_SIGNAL'
};
var SpeakerIdRecord = (function () {
    function SpeakerIdRecord(speakerIdRecord) {
        Object.assign(this, speakerIdRecord);
    }
    return SpeakerIdRecord;
}());
exports.SpeakerIdRecord = SpeakerIdRecord;
var SID_Status;
(function (SID_Status) {
    SID_Status[SID_Status["ACCEPTED"] = 'ACCEPTED'] = "ACCEPTED";
    SID_Status[SID_Status["NOT_USED"] = 'NOT-USED'] = "NOT_USED";
    SID_Status[SID_Status["REJECTED"] = 'REJECTED'] = "REJECTED";
    SID_Status[SID_Status["ID_TI_NOT_AVAILABLE"] = 'ID-TI-NOT-AVAILABLE'] = "ID_TI_NOT_AVAILABLE";
})(SID_Status = exports.SID_Status || (exports.SID_Status = {}));
;
var IndividualSpeaker = (function () {
    function IndividualSpeaker(speaker) {
        this.ID = "";
        this.Score = 0;
        this.SID_Confidence = 0;
        this.id = "";
        this.score = 0;
        this.sid_confidence = 0;
        Object.assign(this, speaker);
        this.id = this.id || this.ID;
        this.score = this.score || this.Score;
        this.sid_status = this.sid_status || this.SID_Status;
        this.sid_confidence = this.sid_confidence || this.SID_Confidence;
    }
    return IndividualSpeaker;
}());
exports.IndividualSpeaker = IndividualSpeaker;

},{"uuid":undefined}],3:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var animation_utilities_1 = require("animation-utilities");
var jibo_typed_events_1 = require("jibo-typed-events");
var PositionInSpace = (function () {
    function PositionInSpace(x, y, z) {
        if (x === void 0) { x = 0; }
        if (y === void 0) { y = 0; }
        if (z === void 0) { z = 0; }
        this.x = x;
        this.y = y;
        this.z = z;
    }
    return PositionInSpace;
}());
exports.PositionInSpace = PositionInSpace;
var AudioLPSEntityType;
(function (AudioLPSEntityType) {
    AudioLPSEntityType.ENERGY = 0;
    AudioLPSEntityType.HOT_WORD = 1;
    AudioLPSEntityType.SPEAKER_ID = 2;
})(AudioLPSEntityType = exports.AudioLPSEntityType || (exports.AudioLPSEntityType = {}));
var JiboBodyAxis;
(function (JiboBodyAxis) {
    JiboBodyAxis.PELVIS = 'PELVIS';
    JiboBodyAxis.TORSO = 'TORSO';
    JiboBodyAxis.NECK = 'NECK';
})(JiboBodyAxis = exports.JiboBodyAxis || (exports.JiboBodyAxis = {}));
var IDLabels;
(function (IDLabels) {
    IDLabels.NOT_ENROLLED = 'NOT_ENROLLED';
    IDLabels.ANYBODY = 'ANYBODY';
    IDLabels.IDENTIFIED = 'IDENTIFIED';
    IDLabels.UNKNOWN = 'UNKNOWN';
    IDLabels.NOT_TRAINED = 'NOT_TRAINED';
})(IDLabels = exports.IDLabels || (exports.IDLabels = {}));
var SensoryType;
(function (SensoryType) {
    SensoryType.AUDIO_LOCALIZATION = 'AUDIO_LOCALIZATION';
    SensoryType.MOTION = 'MOTION';
    SensoryType.PERSON = 'PERSON';
    SensoryType.WAKE = 'WAKE';
    SensoryType.VOICE = 'VOICE';
    SensoryType.GROUND_TRUTH = 'GROUND_TRUTH';
    function fromString(name) {
        return SensoryType[name] || SensoryType[name.toUpperCase()];
    }
    SensoryType.fromString = fromString;
})(SensoryType = exports.SensoryType || (exports.SensoryType = {}));
var SensoryRecord = (function () {
    function SensoryRecord(entity, position, type, reportedTS) {
        this.position = null;
        Object.assign(this, entity);
        this.type = type;
        this.raw = entity;
        if (position !== null && position !== undefined) {
            this.position = position;
        }
        this.id = entity.id;
        this.confidence = entity.confidence;
        this.receivedAt = animation_utilities_1.Clock.currentTime();
        this.reportedTS = reportedTS;
    }
    SensoryRecord.prototype.update = function (entity) {
        Object.assign(this, entity);
        this.position = new animation_utilities_1.THREE.Vector3(entity.position.x, entity.position.y, entity.position.z);
        this.lastUpdate = animation_utilities_1.Clock.currentTime();
    };
    return SensoryRecord;
}());
exports.SensoryRecord = SensoryRecord;
var PresenceEvents = (function (_super) {
    __extends(PresenceEvents, _super);
    function PresenceEvents() {
        var _this = _super.call(this) || this;
        _this.presenceStarted = new jibo_typed_events_1.Event("Presence Started");
        _this.presenceEnded = new jibo_typed_events_1.Event("Presence Ended");
        return _this;
    }
    return PresenceEvents;
}(jibo_typed_events_1.EventContainer));
exports.PresenceEvents = PresenceEvents;
var PresenceRecord = (function () {
    function PresenceRecord(presence) {
        this.kind = null;
        this.events = new PresenceEvents();
        Object.assign(this, presence);
        this.lastUpdate = animation_utilities_1.Clock.currentTime();
    }
    PresenceRecord.prototype.update = function (presence) {
        var presenceState = this.present;
        Object.assign(this, presence);
        if (presenceState !== this.present) {
            if (this.present === true) {
                this.events.presenceStarted.emit();
            }
            else {
                this.events.presenceEnded.emit();
            }
        }
        this.lastUpdate = animation_utilities_1.Clock.currentTime();
    };
    return PresenceRecord;
}());
exports.PresenceRecord = PresenceRecord;
var SpeakerIdInfo = (function () {
    function SpeakerIdInfo(info) {
        this.id = info ? info.id : "";
        this.confidence = info ? info.confidence : 0;
    }
    return SpeakerIdInfo;
}());
exports.SpeakerIdInfo = SpeakerIdInfo;
var SpeakerRecord = (function () {
    function SpeakerRecord(speaker) {
        this.idInfo = speaker ? speaker.idInfo : new SpeakerIdInfo();
        this.entity = speaker ? speaker.entity : -1;
        this.location = speaker ? speaker.location : null;
    }
    return SpeakerRecord;
}());
exports.SpeakerRecord = SpeakerRecord;
var RawSpeakerData = (function () {
    function RawSpeakerData() {
    }
    return RawSpeakerData;
}());
exports.RawSpeakerData = RawSpeakerData;

},{"animation-utilities":undefined,"jibo-typed-events":undefined}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BodyState = (function () {
    function BodyState() {
        this.miscState = new MiscState();
        this.touchState = new TouchState();
        this.powerState = new PowerState();
    }
    return BodyState;
}());
exports.BodyState = BodyState;
var MiscState = (function () {
    function MiscState() {
        this.ts = [];
        this.main_board_temp = 0;
        this.cpu_temp = 0;
        this.intake_temp = 0;
        this.exhaust_temp = 0;
        this.neck_motor_temp = 0;
        this.torso_motor_temp = 0;
        this.pelvis_motor_temp = 0;
    }
    return MiscState;
}());
exports.MiscState = MiscState;
var TouchState = (function () {
    function TouchState() {
        this.ts = [];
        this.touched = false;
        this.pad_state = [false, false, false, false, false, false];
        this.changed = [0, 0, 0, 0, 0, 0];
    }
    return TouchState;
}());
exports.TouchState = TouchState;
var PowerState = (function () {
    function PowerState() {
        this.ts = [];
        this.source = "";
        this.vsys = 0;
        this.battery = new BatteryState();
    }
    return PowerState;
}());
exports.PowerState = PowerState;
var BatteryState = (function () {
    function BatteryState() {
        this.id = 0;
        this.cycles = 0;
        this.capacity = 0;
        this.max_capacity = 0;
        this.charge_rate = 0;
        this.temp = 0;
    }
    return BatteryState;
}());
exports.BatteryState = BatteryState;
var InputEnergy = (function () {
    function InputEnergy() {
        this.ts = [null, null];
        this.db_rms = null;
        this.db_high = null;
        this.db_mid = null;
        this.db_low = null;
    }
    return InputEnergy;
}());
exports.InputEnergy = InputEnergy;
var StatusFlags;
(function (StatusFlags) {
    StatusFlags[StatusFlags["INDEXED"] = 0] = "INDEXED";
    StatusFlags[StatusFlags["ENABLED"] = 1] = "ENABLED";
    StatusFlags[StatusFlags["BRAKED"] = 2] = "BRAKED";
    StatusFlags[StatusFlags["MOVING"] = 3] = "MOVING";
    StatusFlags[StatusFlags["STALLED"] = 4] = "STALLED";
    StatusFlags[StatusFlags["TIMEOUT"] = 5] = "TIMEOUT";
    StatusFlags[StatusFlags["FAULT"] = 6] = "FAULT";
})(StatusFlags = exports.StatusFlags || (exports.StatusFlags = {}));
var FaultFlags;
(function (FaultFlags) {
    FaultFlags[FaultFlags["INDEXED"] = 0] = "INDEXED";
    FaultFlags[FaultFlags["POWER"] = 1] = "POWER";
    FaultFlags[FaultFlags["STALL"] = 2] = "STALL";
    FaultFlags[FaultFlags["ENCODER"] = 3] = "ENCODER";
    FaultFlags[FaultFlags["ENCODER_BACKWARDS"] = 4] = "ENCODER_BACKWARDS";
    FaultFlags[FaultFlags["THERMISTOR"] = 5] = "THERMISTOR";
})(FaultFlags = exports.FaultFlags || (exports.FaultFlags = {}));
var LockoutFlags;
(function (LockoutFlags) {
    LockoutFlags[LockoutFlags["NO_BATTERY"] = 0] = "NO_BATTERY";
    LockoutFlags[LockoutFlags["NOT_UPRIGHT"] = 1] = "NOT_UPRIGHT";
    LockoutFlags[LockoutFlags["FALLING"] = 2] = "FALLING";
})(LockoutFlags = exports.LockoutFlags || (exports.LockoutFlags = {}));

},{}],5:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./GlobalListenTypes"));
__export(require("./LPSTypes"));
__export(require("./SystemTypes"));
__export(require("./Expression"));

},{"./Expression":1,"./GlobalListenTypes":2,"./LPSTypes":3,"./SystemTypes":4}]},{},[5])(5)
});
