// jibo-attention-manager/lib/jibo-attention-manager.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboAttentionManager = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Bakery = AnimationUtilities.ui.Bakery;
var ModalityHandler = require("./ModalityHandler");
var TrackedSensory = require("./TrackedSensory");
var SensoryRecord = require("./SensoryRecord");
var slog = AnimationUtilities.slog;
var channel = "ATTENTION";
var ALWakeModalityHandler = function (motionTracker) {
    ModalityHandler.call(this);
    this.motionTracker = motionTracker;
    this.confidenceThreshold = 0;
    this.heightFloor = 0.7;
};
ALWakeModalityHandler.prototype = Object.create(ModalityHandler.prototype);
ALWakeModalityHandler.prototype.constructor = ALWakeModalityHandler;
ALWakeModalityHandler.prototype.isValid = function (sensoryRecord) {
    if (sensoryRecord.raw.type !== 1) {
        slog(channel, "ALWakeModalityHandler: Unexpected packet type: " + JSON.stringify(sensoryRecord.raw));
        return false;
    }
    if (sensoryRecord.position === null) {
        slog(channel, "ALWakeModalityHandler: unexpected wake record format (no position)");
        return false;
    }
    var rejectMovingAudio = Bakery.getBoolean("Reject WAKE during motion", false, "SensoryFilters");
    var rejectMovingAudioRecoveryTime = Bakery.getFloat("Reject WAKE recovery time", 0, 2, 0.7, "SensoryFilters");
    var confidence = sensoryRecord.confidence;
    var confidenceOK = confidence >= Bakery.getFloat("Confidence Threshold (WAKE)", 0, 1, this.confidenceThreshold, "SensoryFilters");
    var motionOK = !rejectMovingAudio || this.motionTracker.timeSinceMoving() >= rejectMovingAudioRecoveryTime;
    return (confidenceOK && motionOK);
};
ALWakeModalityHandler.prototype.accept = function (sensoryRecord) {
    var minHeight = Bakery.getFloat("Localized-Wake Height Floor", -1, 1, this.heightFloor, "SensoryFilters");
    if (sensoryRecord.position.z < minHeight) {
        sensoryRecord.position.z = minHeight;
    }
    var trackedSensory = new TrackedSensory(SensoryRecord.SensoryType.WAKE);
    trackedSensory.incorporateSensorReport(sensoryRecord);
    trackedSensory.wakeAt = sensoryRecord.receivedAt;
    this.tracked.length = 0;
    this.tracked.push(trackedSensory);
};
module.exports = ALWakeModalityHandler;

},{"./ModalityHandler":8,"./SensoryRecord":11,"./TrackedSensory":14,"animation-utilities":undefined}],2:[function(require,module,exports){
"use strict";
var SimpleAnimGenerator = require("./SimpleAnimGenerator");
var slog = require("animation-utilities").slog;
var channel = "ATTENTION";
var builderFromColor = function (animationUtilities, name, r, g, b) {
    var builder = animationUtilities.createAnimationBuilderFromData(SimpleAnimGenerator.animFromColor(name, r, g, b, "lightring"), "./animations", name);
    builder.setTransitionIn(null);
    return builder;
};
var AttentionDebugStateVisualizer = function (animationUtilities, dofArbiter, emChannel) {
    this._animationUtilities = animationUtilities;
    this._dofArbiter = dofArbiter;
    this._emChannel = emChannel;
    this._displayingState = null;
    this._displayFaceTracking = builderFromColor(this._animationUtilities, "Face Track Vis", 0, 1, 0);
    this._displayMotionTracking = builderFromColor(this._animationUtilities, "Motion Track Vis", 0, 0, 1);
    this._displaySoundTracking = builderFromColor(this._animationUtilities, "Sound Track Vis", 1, 0, 1);
    this._displayDisengage = builderFromColor(this._animationUtilities, "Disengage Vis", 1, 1, 0);
    this._displayingNothing = builderFromColor(this._animationUtilities, "No Vis", 0, 0, 0);
};
AttentionDebugStateVisualizer.prototype.handleState = function (state, force) {
    if (this._displayingState !== state || force === true) {
        if (state === "FACE") {
            this._dofArbiter.playAnimation(this._displayFaceTracking, this._emChannel);
        }
        else if (state === "MOTION") {
            this._dofArbiter.playAnimation(this._displayMotionTracking, this._emChannel);
        }
        else if (state === "SOUND") {
            this._dofArbiter.playAnimation(this._displaySoundTracking, this._emChannel);
        }
        else if (state === "DISENGAGE") {
            this._dofArbiter.playAnimation(this._displayDisengage, this._emChannel);
        }
        else if (state === null) {
            this._dofArbiter.playAnimation(this._displayingNothing, this._emChannel);
        }
        else {
            slog(channel, "AttentionDebugStateVisualizer: Unknown debug state:" + state);
        }
        this._displayingState = state;
    }
};
module.exports = AttentionDebugStateVisualizer;

},{"./SimpleAnimGenerator":13,"animation-utilities":undefined}],3:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var slog = AnimationUtilities.slog;
var eventQueue = [];
var AttentionEventDispatcher = {
    queueEvent: function (theFunction, theObject, theArgs) {
        if (theFunction == null) {
            slog.error("AttentionEventDispatcher: Error, null/undefined function queued!\n" + new Error().stack);
        }
        eventQueue.push({ f: theFunction, o: theObject, a: theArgs });
    },
    dispatchQueuedEvents: function () {
        for (var i = 0; i < eventQueue.length; i++) {
            var e = eventQueue[i];
            e.f.apply(e.o, e.a);
        }
        eventQueue.length = 0;
    }
};
module.exports = AttentionEventDispatcher;

},{"animation-utilities":undefined}],4:[function(require,module,exports){
"use strict";
var SimpleAnimGenerator = require("./SimpleAnimGenerator");
var AttentionEyeRecessor = function (animationUtilities, dofArbiter, emChannel) {
    this._animationUtilities = animationUtilities;
    this._dofArbiter = dofArbiter;
    this._emChannel = emChannel;
    this._recessedBuilder = animationUtilities.createAnimationBuilderFromData(SimpleAnimGenerator.animFromAlphaAndScale("Recessed", 0.87, 0.95, animationUtilities.getRobotInfo()));
    this._recessedBuilder.setSpeed(0);
    this._recessedBuilder.getTransitionIn().setLimits(animationUtilities.getRobotInfo().getEyeDOFNames(), 20, 10);
    this._defaultBuilder = animationUtilities.createAnimationBuilderFromData(SimpleAnimGenerator.animFromAlphaAndScale("Default", 1, 1, animationUtilities.getRobotInfo()));
    this._defaultBuilder.getTransitionIn().setLimits(animationUtilities.getRobotInfo().getEyeDOFNames(), 20, 40);
    this._startedListener = this.lookStarted.bind(this);
    this._endedListener = this.lookEnded.bind(this);
    this._lastTarget = null;
    this._currentTarget = null;
    this._recessionThreshold = 0;
    this._isRecessed = false;
};
AttentionEyeRecessor.prototype.setRecessionThreshold = function (threshold) {
    this._recessionThreshold = threshold;
};
AttentionEyeRecessor.prototype.notifyTargets = function (lastTarget, currentTarget) {
    this._lastTarget = lastTarget;
    this._currentTarget = currentTarget;
};
AttentionEyeRecessor.prototype.shouldRecess = function () {
    return (this._lastTarget === null || this._lastTarget.distanceTo(this._currentTarget) > this._recessionThreshold);
};
AttentionEyeRecessor.prototype.attachToBuilder = function (builder) {
    builder.on("STARTED", this._startedListener);
    builder.on("STOPPED", this._endedListener);
    builder.on("CANCELLED", this._endedListener);
    builder.on("TARGET_REACHED", this._endedListener);
};
AttentionEyeRecessor.prototype.lookStarted = function (eventName, instance, payload) {
    if (this.shouldRecess()) {
        this._dofArbiter.playAnimation(this._recessedBuilder, this._emChannel);
        this._isRecessed = true;
    }
};
AttentionEyeRecessor.prototype.lookEnded = function (eventName, instance, payload) {
    this._dofArbiter.playAnimation(this._defaultBuilder, this._emChannel);
    this._isRecessed = false;
};
AttentionEyeRecessor.prototype.getIsRecessed = function () {
    return this._isRecessed;
};
module.exports = AttentionEyeRecessor;

},{"./SimpleAnimGenerator":13}],5:[function(require,module,exports){
"use strict";
var SensoryStore = require("./SensoryStore");
var SensoryRecord = require("./SensoryRecord");
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var Bakery = AnimationUtilities.ui.Bakery;
var JSONBaker = AnimationUtilities.ui.JSONBaker;
var RemoteVisualizerConnection = require("./remote/RemoteVisualizerConnection");
var DataConverter = require("./DataConverter");
var AttentionDebugStateVisualizer = require("./AttentionDebugStateVisualizer");
var THREE = AnimationUtilities.THREE;
var MotorMotionMonitor = require("./MotorMotionMonitor");
var AttentionEyeRecessor = require("./AttentionEyeRecessor");
var findRoot = require("find-root");
var SensoryLogController = require("./log/SensoryLogController");
var slog = AnimationUtilities.slog;
var AttentionEventDispatcher = require("./AttentionEventDispatcher");
var FaceAwaiter = require("./actions/FaceAwaiter");
var TridentSearchIterator = require("./actions/TridentSearchIterator");
var AttendHandle = require("./actions/AttendHandle");
var ResultStatus = require("./actions/ResultStatus");
var LookConfig = require("./lookconfig/LookConfig");
var LookConfigSlow = require("./lookconfig/LookConfigSlow");
var LookConfigShort = require("./lookconfig/LookConfigShort");
var LookConfigDisengage = require("./lookconfig/LookConfigDisengage");
var LookConfigUncommitted = require("./lookconfig/LookConfigUncommitted");
var LookConfigEyeVision = require("./lookconfig/LookConfigEyeVision");
var channel = "ATTENTION";
var EMChannel = "Attention";
var Mode = {
    OFF: "OFF",
    IDLE: "IDLE",
    DISENGAGE: "DISENGAGE",
    ENGAGED: "ENGAGED",
    FIXATED: "FIXATED",
    COMMAND: "COMMAND"
};
var EventType = {
    IGNORED_HEY_JIBO: "IGNORED_HEY_JIBO",
    STARTED_HEY_JIBO: "STARTED_HEY_JIBO",
    FINISHED_HEY_JIBO: "FINISHED_HEY_JIBO"
};
var RulesState = function () {
    this.lookClaimed = false;
};
RulesState.prototype.clear = function () {
    this.lookClaimed = false;
};
var randomPointNearAbsoluteVertical = function (targetPoint, maxAngleAround, minAngleVertical, maxAngleVertical) {
    var angleAround = (Math.random() - 0.5) * 2 * maxAngleAround;
    var verticalAngle = (Math.random() * (maxAngleVertical - minAngleVertical)) + minAngleVertical;
    var newTarget = new THREE.Vector3().copy(targetPoint);
    var vertical = new THREE.Vector3(0, 0, 1);
    newTarget.projectOnPlane(vertical);
    if (newTarget.length() < 0.001) {
        return null;
    }
    newTarget.applyAxisAngle(vertical, angleAround);
    newTarget.setLength(1.5);
    var temp = vertical.cross(newTarget).normalize();
    newTarget.applyAxisAngle(temp, -verticalAngle);
    return newTarget;
};
var randomPointNearRelativeVertical = function (targetPoint, maxAngleAround, maxAngleVertical) {
    var angleAround = (Math.random() - 0.5) * 2 * maxAngleAround;
    var angleVert = (Math.random() - 0.5) * 2 * maxAngleVertical;
    var newTarget = new THREE.Vector3().copy(targetPoint);
    var vertical = new THREE.Vector3(0, 0, 1);
    var horizontal = new THREE.Vector3(0, 0, 1).cross(newTarget);
    if (horizontal.length() < 0.001) {
        return null;
    }
    newTarget.applyAxisAngle(horizontal, angleVert);
    newTarget.applyAxisAngle(vertical, angleAround);
    newTarget.setLength(1.5);
    return newTarget;
};
var LastFullLookConfigFilter = function (lookConfigs, requirePresent) {
    this.lookConfigs = lookConfigs;
    this.requirePresent = requirePresent;
};
LastFullLookConfigFilter.prototype.checkHistory = function (recentAttentionHistory) {
    var anyPresent = false;
    for (var i = 0; i < this.lookConfigs.length; i++) {
        if (this.lookConfigs[i] === recentAttentionHistory.lastFullLookConfig) {
            anyPresent = true;
            break;
        }
    }
    return this.requirePresent === anyPresent;
};
var AttentionRule = function (name, lockTime, isHeyJibo, lookatConfig, minIdleTime, selfResetsIdle, historyFilter, trackDisplayName) {
    this.lockTime = lockTime;
    this.name = name;
    this.firedAt = null;
    this.activeIdleStart = null;
    this.lastTarget = null;
    this.isHeyJibo = isHeyJibo;
    this.lookatConfig = lookatConfig;
    this.trackDisplayName = null;
    if (trackDisplayName != null) {
        this.trackDisplayName = trackDisplayName;
    }
    this.minRequiredIdle = 0;
    this.maxRequiredIdle = 0;
    this.currentRequiredIdle = null;
    if (Array.isArray(minIdleTime)) {
        this.minRequiredIdle = minIdleTime[0];
        this.maxRequiredIdle = minIdleTime[1];
    }
    else {
        this.minRequiredIdle = minIdleTime;
        this.maxRequiredIdle = minIdleTime;
    }
    this.historyFilter = historyFilter;
    this.selfResetsIdle = selfResetsIdle;
    this.updateResult = {
        target: null,
        initLookat: false
    };
    this.enabled = true;
    this.succeededHandler = null;
    this.interruptedHandler = null;
};
AttentionRule.prototype.update = function (time, tracks, rulesState, recentAttentionHistory) {
    var enabled = Bakery.getBoolean(this.name + " enabled", this.enabled, "ATPolicy");
    var lockTime = Bakery.getFloat(this.name + " lockTime", 0, 10, this.lockTime, "ATPolicy");
    var minRequiredIdle = Bakery.getFloat(this.name + " minRequiredIdle", 0, 10, this.minRequiredIdle, "ATPolicy");
    var maxRequiredIdle = Bakery.getFloat(this.name + " maxRequiredIdle", 0, 30, this.maxRequiredIdle, "ATPolicy");
    var currentRequiredIdle = this.currentRequiredIdle;
    if (!enabled || rulesState.lookClaimed) {
        this.reset();
    }
    else {
        if (lockTime > 0 && this.firedAt != null && time.subtract(this.firedAt) < lockTime && this.lastTarget != null) {
            this.updateResult.initLookat = false;
            this.updateResult.target = this.lastTarget;
            return this.updateResult;
        }
        else {
            if (this.activeIdleStart === null) {
                this.activeIdleStart = time;
            }
            if (currentRequiredIdle === null) {
                if (maxRequiredIdle === minRequiredIdle) {
                    currentRequiredIdle = minRequiredIdle;
                }
                else {
                    currentRequiredIdle = Math.random() * (maxRequiredIdle - minRequiredIdle) + minRequiredIdle;
                }
                this.currentRequiredIdle = currentRequiredIdle;
            }
            if (currentRequiredIdle === 0 || time.subtract(this.activeIdleStart) >= currentRequiredIdle) {
                if (this.historyFilter === null || this.historyFilter.checkHistory(recentAttentionHistory) === true) {
                    var t = this.getTarget(time, tracks, recentAttentionHistory);
                    if (t != null) {
                        t = t.clone();
                        this.currentRequiredIdle = null;
                        this.lastTarget = t;
                        this.firedAt = time;
                        if (this.selfResetsIdle) {
                            this.activeIdleStart = null;
                        }
                        this.updateResult.initLookat = true;
                        this.updateResult.target = t;
                        return this.updateResult;
                    }
                }
            }
        }
    }
    return null;
};
AttentionRule.prototype.reset = function () {
    this.lastTarget = null;
    this.firedAt = null;
    this.activeIdleStart = null;
};
AttentionRule.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    return null;
};
AttentionRule.prototype.notifyDidFire = function () {
};
var AttentionRuleHeyJibo = function (name, lockTime, minIdle, lookConfig) {
    AttentionRule.call(this, name, lockTime, true, lookConfig, minIdle, true, null, null);
};
AttentionRuleHeyJibo.prototype = Object.create(AttentionRule.prototype);
AttentionRuleHeyJibo.prototype.constructor = AttentionRuleHeyJibo;
AttentionRuleHeyJibo.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    if (tracks.newHeyJibo !== null) {
        slog(channel, "ATTENTION: Doing hey jibo");
        return tracks.newHeyJibo.position;
    }
    else {
        return null;
    }
};
var AttentionRuleCommandLook = function (name, lockTime, minIdle, lookConfig) {
    var _this = this;
    AttentionRule.call(this, name, lockTime, false, lookConfig, minIdle, true, null, null);
    this._target = null;
    this._attendHandle = null;
    this.succeededHandler = function () {
        _this.resolveCompletion(ResultStatus.SUCCEEDED);
    };
    this.interruptedHandler = function () {
        _this.resolveCompletion(ResultStatus.INTERRUPTED);
    };
    this._promiseResolve = null;
};
AttentionRuleCommandLook.prototype = Object.create(AttentionRule.prototype);
AttentionRuleCommandLook.prototype.constructor = AttentionRuleCommandLook;
AttentionRuleCommandLook.prototype.resolveCompletion = function (status) {
    if (this._attendHandle != null) {
        var result = { status: status };
        var handle = this._attendHandle;
        this._attendHandle = null;
        this._target = null;
        handle.result = result;
        this._promiseResolve(result);
        this._promiseResolve = null;
    }
    else {
        slog(channel, "AttentionRuleCommandLook:Resolved Twice!");
    }
};
AttentionRuleCommandLook.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    var target = null;
    if (this._target !== null) {
        slog(channel, "ATTENTION: Doing command look!");
        target = this._target;
    }
    return target;
};
AttentionRuleCommandLook.prototype.setCommand = function (target) {
    var _this = this;
    if (this._attendHandle !== null) {
        this.resolveCompletion(ResultStatus.INTERRUPTED);
    }
    var attendHandle = new AttendHandle();
    attendHandle.promise = new Promise(function (resolve, reject) {
        _this._promiseResolve = resolve;
    });
    attendHandle.cancel = function () {
        _this.resolveCompletion(ResultStatus.CANCELLED);
    };
    this._target = target;
    this._attendHandle = attendHandle;
    return attendHandle;
};
AttentionRuleCommandLook.prototype.reset = function () {
    AttentionRule.prototype.reset.call(this);
    if (this._target !== null) {
        slog(channel, "AttentionRuleCommandLook reset after command but before activating!  Sending Cancelled");
        this.resolveCompletion(ResultStatus.CANCELLED);
    }
    this._target = null;
};
AttentionRuleCommandLook.prototype.notifyDidFire = function () {
    AttentionRule.prototype.notifyDidFire.call(this);
    this._target = null;
    slog(channel, "ATTENTION: DID command look!");
};
var AttentionRuleLookAway = function (name, lockTime, minIdle, aroundRotation, downRotation, lookConfig) {
    AttentionRule.call(this, name, lockTime, false, lookConfig, minIdle, true, null, "DISENGAGE");
    this._downRotation = downRotation;
    this._aroundRotation = aroundRotation;
};
AttentionRuleLookAway.prototype = Object.create(AttentionRule.prototype);
AttentionRuleLookAway.prototype.constructor = AttentionRuleLookAway;
AttentionRuleLookAway.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    var currentTarget = null;
    slog(channel, "ATTENTION: Doing look away");
    if (tracks.heyJibo !== null && tracks.heyJibo.position !== null) {
        currentTarget = tracks.heyJibo.position;
    }
    else if (recentAttentionHistory.lastBodyTarget !== null) {
        currentTarget = recentAttentionHistory.lastBodyTarget;
    }
    else {
        currentTarget = recentAttentionHistory.initialCenter;
    }
    var around = new THREE.Vector3(0, 0, 1);
    var down = new THREE.Vector3().crossVectors(around, currentTarget).normalize();
    if (down.lengthSq() < 0.001) {
        slog(channel, "ATTENTION: error computing look-away for currentTarget, not doing it");
        return null;
    }
    var sign = Math.random() > 0.5 ? 1 : -1;
    var newTarget = new THREE.Vector3().copy(currentTarget);
    newTarget.applyAxisAngle(down, this._downRotation);
    newTarget.applyAxisAngle(around, this._aroundRotation * sign);
    return newTarget;
};
var AttentionRuleBodySound = function (name, lockTime, minIdle, loudSoundThreshold, lookConfig) {
    AttentionRule.call(this, name, lockTime, false, lookConfig, minIdle, true, null, "SOUND");
    this.loudSoundBodyMoveWeightThreshold = loudSoundThreshold;
};
AttentionRuleBodySound.prototype = Object.create(AttentionRule.prototype);
AttentionRuleBodySound.prototype.constructor = AttentionRuleBodySound;
AttentionRuleBodySound.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    var loudSoundBodyMoveWeightThreshold = Bakery.getFloat(this.name + " loud sound body move threshold", 0, 10, this.loudSoundBodyMoveWeightThreshold, "ATPolicy");
    if (tracks.mostWeightyAudio !== null && tracks.mostWeightyAudio.weight > loudSoundBodyMoveWeightThreshold) {
        slog(channel, "ATTENTION: Doing body sound");
        return tracks.mostWeightyAudio.position;
    }
    else {
        return null;
    }
};
var AttentionRuleBoredomLook = function (name, lockTime, minIdle, boredomReturnToCenterPercent, randomLookHorizontalMax, randomLookMinVerticalAngle, randomLookMaxVerticalAngle, useLastBodyAsCenter, lookatConfig, relativeVertical, historyFilter) {
    var useLookConfig = lookatConfig;
    this.lookatConfigs = null;
    if (Array.isArray(lookatConfig)) {
        useLookConfig = lookatConfig[0];
        this.lookatConfigs = lookatConfig;
    }
    AttentionRule.call(this, name, lockTime, false, useLookConfig, minIdle, true, historyFilter, null);
    this.boredomReturnToHeyJiboPercentage = boredomReturnToCenterPercent;
    this.randomLookAroundHorizontalAngle = randomLookHorizontalMax;
    this.randomLookMinVerticalAngle = randomLookMinVerticalAngle;
    this.randomLookMaxVerticalAngle = randomLookMaxVerticalAngle;
    this.useLastBodyAsCenter = useLastBodyAsCenter;
    this.relativeVertical = relativeVertical;
};
AttentionRuleBoredomLook.prototype = Object.create(AttentionRule.prototype);
AttentionRuleBoredomLook.prototype.constructor = AttentionRuleBoredomLook;
AttentionRuleBoredomLook.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    var boredomReturnToHeyJiboPercentage = Bakery.getFloat(this.name + " return to hey jibo percentage", 0, 1, this.boredomReturnToHeyJiboPercentage, "ATPolicy");
    var maxAngleAround = Bakery.getFloat(this.name + " random look horizontal angle limit", 0, 3, this.randomLookAroundHorizontalAngle, "ATPolicy");
    var minAngleVertical = Bakery.getFloat(this.name + " random look min vertical angle", -1, 1.4, this.randomLookMinVerticalAngle, "ATPolicy");
    var maxAngleVertical = Bakery.getFloat(this.name + " random look max vertical angle", -1, 1.4, this.randomLookMaxVerticalAngle, "ATPolicy");
    var useLastBodyAsCenter = Bakery.getBoolean(this.name + " random look use last body as center", this.useLastBodyAsCenter, "ATPolicy");
    var relativeVertical = Bakery.getBoolean(this.name + " vertical is relative to target", this.relativeVertical, "ATPolicy");
    var center = recentAttentionHistory.initialCenter;
    if (useLastBodyAsCenter) {
        if (recentAttentionHistory.lastBodyTarget !== null) {
            center = recentAttentionHistory.lastBodyTarget;
        }
    }
    else {
        if (tracks.heyJibo !== null && tracks.heyJibo.position !== null) {
            center = tracks.heyJibo.position;
        }
    }
    if (Math.random() <= boredomReturnToHeyJiboPercentage) {
        if (recentAttentionHistory.lastBodyTarget !== null && center.distanceTo(recentAttentionHistory.lastBodyTarget) < 0.001) {
            return null;
        }
        else {
            slog(channel, "ATTENTION: Boredom, returning body to hey jibo via " + this.name + " @" + time);
            return center;
        }
    }
    else {
        slog(channel, "ATTENTION: Boredom, looking randomly " + this.name + " @" + time);
        if (this.lookatConfigs !== null) {
            var select = Math.floor(Math.random() * this.lookatConfigs.length);
            this.lookatConfig = this.lookatConfigs[select];
        }
        if (relativeVertical) {
            return randomPointNearRelativeVertical(center, maxAngleAround, maxAngleVertical);
        }
        else {
            return randomPointNearAbsoluteVertical(center, maxAngleAround, minAngleVertical, maxAngleVertical);
        }
    }
};
var AttentionRuleEyeSound = function (name, lockTime, minIdle, loudSoundThreshold, useFastWeight, lookConfig) {
    AttentionRule.call(this, name, lockTime, false, lookConfig, minIdle, false, null, "SOUND");
    this.loudSoundEyeMoveWeightThreshold = loudSoundThreshold;
    this.useFastWeight = useFastWeight;
};
AttentionRuleEyeSound.prototype = Object.create(AttentionRule.prototype);
AttentionRuleEyeSound.prototype.constructor = AttentionRuleEyeSound;
AttentionRuleEyeSound.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    var loudSoundEyeMoveWeightThreshold = Bakery.getFloat(this.name + " loud sound eye move threshold", 0, 10, this.loudSoundEyeMoveWeightThreshold, "ATPolicy");
    var useFastWeight = Bakery.getBoolean(this.name + " use fast weight", this.useFastWeight, "ATPolicy");
    if (tracks.mostWeightyAudio !== null &&
        (useFastWeight ? tracks.mostWeightyAudio.fastWeight : tracks.mostWeightyAudio.weight) > loudSoundEyeMoveWeightThreshold) {
        slog(channel, "ATTENTION: Doing " + this.name);
        return tracks.mostWeightyAudio.position;
    }
    else {
        return null;
    }
};
var AttentionRuleEyeFace = function (name, lockTime, minIdle, targetVelocityConstant, lookConfig) {
    AttentionRule.call(this, name, lockTime, false, lookConfig, minIdle, false, null, "FACE");
    this.targetVelocityConstant = targetVelocityConstant;
};
AttentionRuleEyeFace.prototype = Object.create(AttentionRule.prototype);
AttentionRuleEyeFace.prototype.constructor = AttentionRuleEyeFace;
AttentionRuleEyeFace.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    if (tracks.trackedFaces !== null && tracks.trackedFaces.length > 0) {
        var targetVelocityConstant = Bakery.getFloat(this.name + " target velocity constant", 0, 4, this.targetVelocityConstant, "ATPolicy");
        var bestTarget = null;
        var minTargetAge = Number.MAX_VALUE;
        for (var i = 0; i < tracks.trackedFaces.length; i++) {
            var track = tracks.trackedFaces[i];
            var targetAge = time.subtract(track.lastIncorporatedReport);
            if (this.lastTarget !== null) {
                targetAge += this.lastTarget.distanceTo(track.position) / targetVelocityConstant;
            }
            if (targetAge < minTargetAge) {
                minTargetAge = targetAge;
                bestTarget = track;
            }
        }
        if (bestTarget) {
            slog(channel, "ATTENTION: Doing " + this.name);
            return bestTarget.position;
        }
        else {
            return null;
        }
    }
    else {
        return null;
    }
};
var AttentionRuleEyeMotion = function (name, lockTime, minIdle, targetVelocityConstant, filterOnTrackAge, minTrackAge, maxTrackAge, lookConfig) {
    AttentionRule.call(this, name, lockTime, false, lookConfig, minIdle, false, null, "MOTION");
    this.targetVelocityConstant = targetVelocityConstant;
    this.filterOnTrackAge = filterOnTrackAge;
    this.minTrackAge = minTrackAge;
    this.maxTrackAge = maxTrackAge;
};
AttentionRuleEyeMotion.prototype = Object.create(AttentionRule.prototype);
AttentionRuleEyeMotion.prototype.constructor = AttentionRuleEyeMotion;
AttentionRuleEyeMotion.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    if (tracks.trackedMotions !== null && tracks.trackedMotions.length > 0) {
        var targetVelocityConstant = Bakery.getFloat(this.name + " target velocity constant", 0, 4, this.targetVelocityConstant, "ATPolicy");
        if (this.filterOnTrackAge) {
            this.minTrackAge = Bakery.getFloat(this.name + " min track age", 0, 15, this.minTrackAge, "ATPolicy");
            this.maxTrackAge = Bakery.getFloat(this.name + " max track age", 0, 30, this.maxTrackAge, "ATPolicy");
        }
        var bestTarget = null;
        var minTargetAge = Number.MAX_VALUE;
        for (var i = 0; i < tracks.trackedMotions.length; i++) {
            var track = tracks.trackedMotions[i];
            if (this.filterOnTrackAge) {
                var trackAge = time.subtract(track.creationTime);
                if (trackAge < this.minTrackAge || trackAge > this.maxTrackAge) {
                    continue;
                }
            }
            var targetAge = time.subtract(track.lastIncorporatedReport);
            if (this.lastTarget !== null) {
                targetAge += this.lastTarget.distanceTo(track.position) / targetVelocityConstant;
            }
            if (targetAge < minTargetAge) {
                minTargetAge = targetAge;
                bestTarget = track;
            }
        }
        if (bestTarget) {
            slog(channel, "ATTENTION: Doing " + this.name);
            return bestTarget.position;
        }
        else {
            return null;
        }
    }
    else {
        return null;
    }
};
var AttentionRuleEyeReturn = function (name, lockTime, minIdle, eyeBodyDeltaPosThreshold, eyeBodyDeltaAngleThreshold, lookatConfig) {
    AttentionRule.call(this, name, lockTime, false, lookatConfig, minIdle, true, null, null);
    this._eyeBodyDeltaPosThreshold = eyeBodyDeltaPosThreshold;
    this._eyeBodyDeltaAngleThreshold = eyeBodyDeltaAngleThreshold;
};
AttentionRuleEyeReturn.prototype = Object.create(AttentionRule.prototype);
AttentionRuleEyeReturn.prototype.constructor = AttentionRuleEyeReturn;
AttentionRuleEyeReturn.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    if (recentAttentionHistory.lastBodyTarget !== null) {
        if (recentAttentionHistory.lastEyeTarget !== null &&
            (this._eyeBodyDeltaPosThreshold === null || recentAttentionHistory.lastEyeTarget.distanceTo(recentAttentionHistory.lastBodyTarget) < this._eyeBodyDeltaPosThreshold) &&
            (this._eyeBodyDeltaAngleThreshold === null || recentAttentionHistory.lastEyeTarget.angleTo(recentAttentionHistory.lastBodyTarget) < this._eyeBodyDeltaAngleThreshold)) {
            return null;
        }
        else {
            slog(channel, "ATTENTION: Doing Eye Return " + this.name);
            return recentAttentionHistory.lastBodyTarget;
        }
    }
};
var AttentionRuleBodyCommit = function (name, lockTime, minIdle, lookatConfig, filter) {
    AttentionRule.call(this, name, lockTime, false, lookatConfig, minIdle, true, filter, null);
};
AttentionRuleBodyCommit.prototype = Object.create(AttentionRule.prototype);
AttentionRuleBodyCommit.prototype.constructor = AttentionRuleBodyCommit;
AttentionRuleBodyCommit.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    if (recentAttentionHistory.lastBodyTarget !== null) {
        slog(channel, "ATTENTION: Doing Body Commit");
        return recentAttentionHistory.lastBodyTarget;
    }
};
var AttentionRuleBlinkOccasionally = function (name, minIdle, blinker) {
    AttentionRule.call(this, name, 0, false, null, minIdle, true, null, null);
    this.blinker = blinker;
};
AttentionRuleBlinkOccasionally.prototype = Object.create(AttentionRule.prototype);
AttentionRuleBlinkOccasionally.prototype.constructor = AttentionRuleBlinkOccasionally;
AttentionRuleBlinkOccasionally.prototype.update = function (time, tracks, rulesState, recentAttentionHistory) {
    var enabled = Bakery.getBoolean(this.name + " enabled", this.enabled, "ATPolicy");
    var minRequiredIdle = Bakery.getFloat(this.name + " minRequiredIdle", 0, 10, this.minRequiredIdle, "ATPolicy");
    var maxRequiredIdle = Bakery.getFloat(this.name + " maxRequiredIdle", 0, 30, this.maxRequiredIdle, "ATPolicy");
    var currentRequiredIdle = this.currentRequiredIdle;
    if (!enabled) {
        this.reset();
    }
    else {
        if (this.activeIdleStart === null) {
            this.activeIdleStart = time;
        }
        if (currentRequiredIdle === null) {
            if (maxRequiredIdle === minRequiredIdle) {
                currentRequiredIdle = minRequiredIdle;
            }
            else {
                currentRequiredIdle = Math.random() * (maxRequiredIdle - minRequiredIdle) + minRequiredIdle;
            }
            this.currentRequiredIdle = currentRequiredIdle;
        }
        if (time.subtract(this.activeIdleStart) > currentRequiredIdle) {
            this.blinker.blink();
            slog(channel, "ATTENTION: Doing Occasional Blink (" + this.currentInterval + ")");
            this.currentRequiredIdle = null;
            this.firedAt = time;
            if (this.selfResetsIdle) {
                this.activeIdleStart = null;
            }
            return null;
        }
    }
    return null;
};
AttentionRuleBlinkOccasionally.prototype.reset = function () {
    AttentionRule.prototype.reset.call(this);
    this.currentInterval = null;
};
var AttentionRuleSwitchMode = function (name, lockTime, minIdle, newMode, attentionManager) {
    AttentionRule.call(this, name, lockTime, false, null, minIdle, true, null, null);
    this.newMode = newMode;
    this.attentionManager = attentionManager;
};
AttentionRuleSwitchMode.prototype = Object.create(AttentionRule.prototype);
AttentionRuleSwitchMode.prototype.constructor = AttentionRuleSwitchMode;
AttentionRuleSwitchMode.prototype.getTarget = function (time, tracks, recentAttentionHistory) {
    var self = this;
    setTimeout(function () {
        self.attentionManager.setMode(self.newMode);
    }, 0);
    return null;
};
var AttentionRuleShiftOccasionally = function (name, minIdleTime, maxIdleTime, minHoldTime, maxHoldTime, animate, postureURLs, centerURL) {
    AttentionRule.call(this, name, 0, false, null, minIdleTime, true, null, null);
    this.maxIdleTime = maxIdleTime;
    this.minHoldTime = minHoldTime;
    this.maxHoldTime = maxHoldTime;
    this.animate = animate;
    this.currentIdleInterval = null;
    this.currentHoldInterval = null;
    this.postureBuilders = [];
    this.centerBuilder = null;
    this.dataLoaded = false;
    var self = this;
    var loadPostureBuilder = function (url) {
        animate.createAnimationBuilder(url, function (builder) {
            builder.setLayer("posture");
            self.postureBuilders.push(builder);
            self.dataLoaded = self.centerBuilder !== null && self.postureBuilders.length === postureURLs.length;
        });
    };
    for (var i = 0; i < postureURLs.length; i++) {
        loadPostureBuilder(postureURLs[i]);
    }
    animate.createAnimationBuilder(centerURL, function (builder) {
        builder.setLayer("posture");
        self.centerBuilder = builder;
        self.dataLoaded = self.centerBuilder !== null && self.postureBuilders.length === postureURLs.length;
    });
};
AttentionRuleShiftOccasionally.prototype = Object.create(AttentionRule.prototype);
AttentionRuleShiftOccasionally.prototype.constructor = AttentionRuleShiftOccasionally;
AttentionRuleShiftOccasionally.prototype.update = function (time, tracks, rulesState, recentAttentionHistory) {
    if (!this.dataLoaded) {
        return null;
    }
    var enabled = Bakery.getBoolean(this.name + " enabled", this.enabled, "ATPolicy");
    var minRequiredIdle = Bakery.getFloat(this.name + " minRequiredIdle", 0, 10, this.minRequiredIdle, "ATPolicy");
    var maxRequiredIdle = Bakery.getFloat(this.name + " maxRequiredIdle", 0, 30, this.maxRequiredIdle, "ATPolicy");
    var minHoldTime = Bakery.getFloat(this.name + " minHoldTime", 0, 10, this.minHoldTime, "ATPolicy");
    var maxHoldTime = Bakery.getFloat(this.name + " maxHoldTime", 0, 30, this.maxHoldTime, "ATPolicy");
    if (!enabled || rulesState.lookClaimed) {
        this.reset();
    }
    else {
        if (this.currentRequiredIdle === null) {
            this.currentRequiredIdle = Math.random() * (maxRequiredIdle - minRequiredIdle) + minRequiredIdle;
        }
        if (this.currentHoldInterval === null) {
            this.currentHoldInterval = Math.random() * (maxHoldTime - minHoldTime) + minHoldTime;
        }
        if (this.activeIdleStart === null) {
            this.activeIdleStart = time;
        }
        if (this.firedAt === null && time.subtract(this.activeIdleStart) > this.currentIdleInterval) {
            var builder = this.postureBuilders[Math.floor(Math.random() * this.postureBuilders.length)];
            builder.play();
            slog(channel, "ATTENTION: Doing Occasional Shift");
            this.firedAt = time;
        }
        else if (this.firedAt !== null && time.subtract(this.firedAt) > this.currentHoldInterval) {
            slog(channel, "ATTENTION: Returning From Occasional Shift");
            this.reset();
        }
    }
    return null;
};
AttentionRuleShiftOccasionally.prototype.reset = function () {
    AttentionRule.prototype.reset.call(this);
    if (!this.dataLoaded) {
        return;
    }
    this.centerBuilder.play();
    this.currentRequiredIdle = null;
    this.currentHoldInterval = null;
};
var install = function (animate, glService, lps, attentionManager, dofArbiter, doRemoteVis, configData) {
    var baker = new JSONBaker(configData);
    Bakery.init(baker);
    var remoteVisualizerConnection = null;
    var sensoryLogController = null;
    if (doRemoteVis) {
        sensoryLogController = new SensoryLogController(attentionManager._sensoryStore);
        remoteVisualizerConnection = new RemoteVisualizerConnection(animate, attentionManager, sensoryLogController, dofArbiter, EMChannel);
    }
    setInterval(function () {
        if (sensoryLogController === null || !sensoryLogController.getBlockLiveData()) {
            attentionManager.getDataConverter().acceptAudioLocalization(lps.lps.audioData);
            attentionManager.getDataConverter().acceptVision(lps.lps.motionData);
        }
        if (sensoryLogController !== null) {
            sensoryLogController.update();
        }
        attentionManager.update();
        if (remoteVisualizerConnection !== null) {
            remoteVisualizerConnection.update();
        }
    }, 33);
    var expressionListener = {
        dofsLost: function (ownerWhoLostDOFs, dofsLost) {
            attentionManager.handleLossOfActiveDOFs(dofsLost);
        },
        dofsGained: function (ownerWhoLostDOFs, dofsLost) {
        },
        dofsAvailable: function (dofsAvailable) {
        }
    };
    dofArbiter.addListener(EMChannel, expressionListener);
};
var AttentionManager = function (jibo, doRemoteVis, configData) {
    var robotInfo = jibo.animate.getRobotInfo();
    var animate = jibo.animate;
    var glService = jibo.gl;
    var lps = jibo.lps;
    var dofArbiter = jibo.dofArbiter;
    this.Mode = Mode;
    this.EventType = EventType;
    this.ResultStatus = ResultStatus;
    this._mode = Mode.OFF;
    this._motionTracker = new MotorMotionMonitor(["bottomSection_r", "middleSection_r", "topSection_r"], animate);
    this._sensoryStore = new SensoryStore(this._motionTracker);
    this._robotInfo = robotInfo;
    this._eventHandlers = {};
    this._animate = null;
    if (animate != null) {
        this._animate = animate;
    }
    this._dofArbiter = dofArbiter;
    this._lastLookatConfig = null;
    this._ignoreHeyJibosAtOrBefore = null;
    this._initialCenter = new THREE.Vector3(1, 0, 0.6);
    this._lookatInstance = null;
    this._currentInstanceIsMissingDOFs = false;
    this._currentInstanceRequiredDOFs = [];
    this._lookResultStatus = [];
    this._dataConverter = new DataConverter(this._sensoryStore, this, true);
    this._debugStateVisualizer = null;
    this._enableDebugStateVisualizer = false;
    this._eyeRecessor = new AttentionEyeRecessor(this._animate, this._dofArbiter, EMChannel);
    this._needToFireOnNewWake = false;
    this._faceAwaiters = [];
    if (this._animate !== null) {
        var attentionDebugStateVisualizer = new AttentionDebugStateVisualizer(this._animate, this._dofArbiter, EMChannel);
        this._debugStateVisualizer = attentionDebugStateVisualizer.handleState.bind(attentionDebugStateVisualizer);
    }
    this._resourceRoot = findRoot(__dirname) + "/res/";
    var lookConfigDefaultEyeOnly = new LookConfig("DefaultEye", robotInfo, true);
    var lookConfigDefault = new LookConfig("Default", robotInfo, false);
    var lookConfigSlow = new LookConfigSlow("Slow", robotInfo, false);
    var lookConfigSlowEyeOnly = new LookConfigSlow("SlowEye", robotInfo, true);
    var lookConfigShort = new LookConfigShort("Short", robotInfo);
    var lookConfigDisengage = new LookConfigDisengage("Disengage", robotInfo);
    var lookConfigUncommitted = new LookConfigUncommitted("Uncommitted", robotInfo);
    var lookConfigVisionFace = new LookConfigEyeVision("EyeVision:Face", robotInfo, "FACE", false, true);
    var lookConfigVisionMotion = new LookConfigEyeVision("EyeVision:Motion", robotInfo, "MOTION", false, true);
    var lookConfigVisionMotionEyeOnly = new LookConfigEyeVision("EyeVision:MotionEye", robotInfo, "MOTION", true, false);
    var requiredDOFsForBlink = animate.dofs.EYE.getDOFs();
    var blinker = {
        blink: function (interrupt) {
            if (dofArbiter.getAvailable(EMChannel, requiredDOFsForBlink, null).length === requiredDOFsForBlink.length) {
                animate.blink(interrupt);
            }
        }
    };
    this._rules = {
        OFF: [],
        IDLE: [
            new AttentionRuleEyeMotion("Idle Eye Recent Motion", 0.0, 0.7, 1.0, true, 1, 4, lookConfigVisionMotionEyeOnly),
            new AttentionRuleEyeFace("Idle Face", 0.0, 0.7, 1.0, lookConfigVisionFace),
            new AttentionRuleEyeMotion("Idle Motion", 0.0, 0.7, 1.0, false, 0, 0, lookConfigVisionMotion),
            new AttentionRuleEyeSound("Idle Eye Sound", 0.1, 2, 0.82, false, lookConfigDefaultEyeOnly),
            new AttentionRuleBoredomLook("Idle Boredom Big", 0.5, [6, 13], 0.3, 1, 0.2, 0.8, false, [lookConfigSlow, lookConfigUncommitted], false, null),
            new AttentionRuleBoredomLook("Idle Boredom Little", 0.5, [0.9, 8], 0.0, 0.08, 0.0, 0.08, true, lookConfigShort, true, new LastFullLookConfigFilter([lookConfigDisengage, lookConfigUncommitted], false)),
            new AttentionRuleEyeReturn("Idle Eye Return", 0.2, 5, 0.001, null, lookConfigSlowEyeOnly),
            new AttentionRuleEyeReturn("Idle Far Eye Return", 0.2, [0.9, 1.3], null, 0.15, lookConfigSlowEyeOnly),
            new AttentionRuleBlinkOccasionally("Idle Blink Occasionally", [3, 17], blinker)
        ],
        ENGAGED: [
            new AttentionRuleHeyJibo("Engaged Hey Jibo", 0.5, 0, lookConfigDefault),
            new AttentionRuleBodySound("Engaged Body Sound", 0.5, 2, 2.63, lookConfigDefault),
            new AttentionRuleEyeMotion("Engaged Eye Recent Motion", 0.0, 0.7, 1.0, true, 1, 4, lookConfigVisionMotionEyeOnly),
            new AttentionRuleEyeFace("Engaged Face", 0.0, 0.7, 1.0, lookConfigVisionFace),
            new AttentionRuleEyeMotion("Engaged Motion", 0.0, 0.7, 1.0, false, 0, 0, lookConfigVisionMotion),
            new AttentionRuleEyeSound("Engaged Eye Sound", 0.1, 2, 0.82, false, lookConfigDefaultEyeOnly),
            new AttentionRuleEyeReturn("Engaged Eye Return", 0.2, 5, 0.001, null, lookConfigSlowEyeOnly),
            new AttentionRuleEyeReturn("Engaged Far Eye Return", 0.2, [0.9, 1.3], null, 0.15, lookConfigSlowEyeOnly)
        ],
        FIXATED: [
            new AttentionRuleHeyJibo("Fixated Hey Jibo", 0.5, 0, lookConfigDefault),
            new AttentionRuleEyeFace("Fixated Face", 0.0, 0.7, 1.0, lookConfigVisionFace),
            new AttentionRuleEyeReturn("Fixated Eye Return", 0.2, 5, 0.001, null, lookConfigSlowEyeOnly),
            new AttentionRuleEyeReturn("Fixated Far Eye Return", 0.2, [0.9, 1.3], null, 0.15, lookConfigSlowEyeOnly)
        ],
        DISENGAGE: [
            new AttentionRuleSwitchMode("Disengage Back To Idle", 0, 4.3, "IDLE", this),
            new AttentionRuleLookAway("Disengaged Look Away", 6, 0, Math.PI / 5, Math.PI / 17, lookConfigDisengage)
        ],
        COMMAND: [
            new AttentionRuleCommandLook("Command Look", 0.5, 0, lookConfigDefault)
        ]
    };
    this._commandRule = this._rules[Mode.COMMAND][0];
    this._rulesState = new RulesState();
    this._attentionCommandHistory = {
        lastEyeTarget: null,
        lastBodyTarget: null,
        lastHeyJiboTarget: null,
        initialCenter: this._initialCenter,
        lastFullLookRule: null,
        lastFullLookConfig: null,
        lastEyeLookRule: null,
        lastEyeLookConfig: null
    };
    this._lookatConfigs = {
        lookConfigDefaultEyeOnly: lookConfigDefaultEyeOnly,
        lookConfigDefault: lookConfigDefault,
        lookConfigSlow: lookConfigSlow,
        lookConfigSlowEyeOnly: lookConfigSlowEyeOnly,
        lookConfigShort: lookConfigShort,
        lookConfigDisengage: lookConfigDisengage,
        lookConfigUncommitted: lookConfigUncommitted,
        lookConfigVisionFace: lookConfigVisionFace,
        lookConfigVisionMotion: lookConfigVisionMotion,
        lookConfigVisionMotionEyeOnly: lookConfigVisionMotionEyeOnly
    };
    this._defaultBuilder = null;
    if (this._animate !== null) {
        var self = this;
        this._animate.createAnimationBuilder(this._resourceRoot + "animations/jibo-default.anim", function (builder) {
            self._defaultBuilder = builder;
        });
    }
    this._lpsConnection = null;
    install(animate, glService, lps, this, dofArbiter, doRemoteVis, configData);
};
AttentionManager.Mode = Mode;
AttentionManager.EventType = EventType;
AttentionManager.ResultStatus = ResultStatus;
AttentionManager.prototype.setEnableEyeColorStateVisualizer = function (enable) {
    if (enable !== this._enableDebugStateVisualizer) {
        this._enableDebugStateVisualizer = enable;
        if (!this._enableDebugStateVisualizer) {
            if (this._debugStateVisualizer != null) {
                this._debugStateVisualizer(null);
            }
        }
    }
};
AttentionManager.prototype.displayTrackState = function (trackState) {
    if (this._enableDebugStateVisualizer) {
        if (this._debugStateVisualizer != null) {
            this._debugStateVisualizer(trackState);
        }
    }
};
AttentionManager.prototype.setDebugVisualizer = function (visualizerFunction) {
    this._debugStateVisualizer = visualizerFunction;
};
AttentionManager.prototype.attendToTarget = function (position, entity) {
    var target;
    if (position === null || position === undefined) {
        target = new THREE.Vector3(entity.position.x, entity.position.y, entity.position.z);
    }
    else if (Array.isArray(position)) {
        target = new THREE.Vector3(position[0], position[1], position[2]);
    }
    else {
        target = new THREE.Vector3(position.x, position.y, position.z);
    }
    var attendHandle = this._commandRule.setCommand(target);
    if (this._mode !== Mode.COMMAND) {
        slog(channel, "Cancelling attendToTarget because we are not in COMMAND mode!");
        attendHandle.cancel();
    }
    return attendHandle;
};
AttentionManager.prototype.awaitFace = function (timeout, maxAngle, fullSearchTime) {
    var currentTime = Clock.currentTime();
    if (maxAngle === undefined || maxAngle === null) {
        maxAngle = 0;
    }
    if (fullSearchTime === undefined || fullSearchTime === null) {
        fullSearchTime = 0;
    }
    var awaiter = new FaceAwaiter(currentTime, timeout, maxAngle, fullSearchTime, this._animate, this._lpsConnection);
    this._faceAwaiters.push(awaiter);
    return awaiter.getHandle();
};
AttentionManager.prototype.getSearchIterator = function () {
    var kinematicFeatures = this._animate.getKinematicFeatures();
    var eyeFeature = kinematicFeatures["Eye"];
    var worldLookTarget = eyeFeature.direction.clone().setLength(2).add(eyeFeature.position);
    slog(channel, "Starting search from (" + worldLookTarget.x + ", " + worldLookTarget.y + ", " + worldLookTarget.z + ")");
    return new TridentSearchIterator(worldLookTarget);
};
AttentionManager.prototype.setMode = function (mode) {
    var r;
    if (mode === Mode.OFF && this._mode !== Mode.OFF) {
        for (var i = 0; i < this._lookResultStatus.length; i++) {
            if (!this._lookResultStatus[i].stopped) {
                this._lookResultStatus[i].instance.stop();
                this._lookResultStatus[i].stopped = true;
            }
        }
        this._lookatInstance = null;
        this._currentInstanceIsMissingDOFs = false;
        this._currentInstanceRequiredDOFs = [];
    }
    if (mode !== this._mode) {
        slog(channel, "Attention Mode changed from " + this._mode + " to " + mode);
        for (r = 0; r < this._rules[this._mode].length; r++) {
            this._rules[this._mode][r].reset();
        }
        for (r = 0; r < this._rules[mode].length; r++) {
            this._rules[mode][r].reset();
        }
    }
    this._mode = mode;
};
AttentionManager.prototype.getMode = function () {
    return this._mode;
};
AttentionManager.prototype.getDataConverter = function () {
    return this._dataConverter;
};
AttentionManager.prototype.notifyWakeHappened = function (sensoryRecord) {
    this._needToFireOnNewWake = true;
};
AttentionManager.prototype.addListener = function (eventType, listener) {
    var handlersForType = this._eventHandlers[eventType];
    if (!handlersForType) {
        handlersForType = [];
        this._eventHandlers[eventType] = handlersForType;
    }
    if (handlersForType.indexOf(listener) === -1) {
        handlersForType.push(listener);
    }
};
AttentionManager.prototype.removeListener = function (eventType, listener) {
    var handlersForType = this._eventHandlers[eventType];
    if (handlersForType) {
        var index = handlersForType.indexOf(listener);
        if (index !== -1) {
            handlersForType.splice(index, 1);
        }
    }
};
var LookResultStatus = function (instance, target, reachedHandler, interruptedHandler, isHeyJibo) {
    this.instance = instance;
    this.target = target;
    this.reachedHandler = reachedHandler;
    this.interruptedHandler = interruptedHandler;
    this.isHeyJibo = isHeyJibo;
    this.stopped = false;
};
var processLookResult = function (lrh, target, interrupted, attentionManager) {
    if (interrupted === false && lrh.reachedHandler !== null) {
        lrh.reachedHandler(lrh.target);
        lrh.reachedHandler = null;
        lrh.interruptedHandler = null;
    }
    if (interrupted === true && lrh.interruptedHandler !== null) {
        lrh.interruptedHandler(lrh.target);
        lrh.reachedHandler = null;
        lrh.interruptedHandler = null;
    }
    if (lrh.isHeyJibo === true) {
        attentionManager.fireEvent(EventType.FINISHED_HEY_JIBO, {
            position: target,
            interrupted: interrupted
        }, false);
        lrh.isHeyJibo = false;
    }
};
AttentionManager.prototype.handleLookFinished = function (instance, target, interrupted) {
    for (var i = 0; i < this._lookResultStatus.length; i++) {
        var lrh = this._lookResultStatus[i];
        if (lrh.instance === instance) {
            processLookResult(lrh, target, interrupted, this);
        }
    }
};
AttentionManager.prototype.handleInstanceFinished = function (instance, interrupted) {
    for (var i = this._lookResultStatus.length - 1; i >= 0; i--) {
        var lrh = this._lookResultStatus[i];
        if (lrh.instance === instance) {
            processLookResult(lrh, null, interrupted, this);
            lrh.stopped = true;
            this._lookResultStatus.splice(i, 1);
        }
    }
};
AttentionManager.prototype.fireEvent = function (eventType, info, useDispatch) {
    var handlersForType = this._eventHandlers[eventType];
    if (handlersForType) {
        for (var i = 0; i < handlersForType.length; i++) {
            if (useDispatch) {
                AttentionEventDispatcher.queueEvent(handlersForType[i], null, [eventType, info]);
            }
            else {
                handlersForType[i](eventType, info);
            }
        }
    }
};
AttentionManager.prototype.handleLossOfActiveDOFs = function (dofsLost) {
    for (var li = 0; li < dofsLost.length; li++) {
        for (var ri = 0; ri < this._currentInstanceRequiredDOFs.length; ri++) {
            if (dofsLost[li] === this._currentInstanceRequiredDOFs[ri]) {
                this._currentInstanceIsMissingDOFs = true;
                return;
            }
        }
    }
};
AttentionManager.prototype.configureNewBuilder = function (lookatBuilder, lookatConfig) {
    var _this = this;
    var dofs = lookatConfig.getDOFs();
    if (this._eyeRecessor != null && lookatConfig.doRecession && Bakery.getBoolean("Allow Any Eye Recession", true, "ATMotion")) {
        this._eyeRecessor.setRecessionThreshold(lookatConfig.recessionThreshold);
        this._eyeRecessor.attachToBuilder(lookatBuilder);
    }
    lookatBuilder.setContinuousMode(true);
    if (dofs !== null) {
        lookatBuilder.setDOFs(dofs);
    }
    lookatConfig.configureBuilder(lookatBuilder);
    lookatBuilder.on(this._animate.LookatEventType.TARGET_REACHED, function (eventName, instance, data) {
        _this.handleLookFinished(instance, data.target, false);
    });
    lookatBuilder.on(this._animate.LookatEventType.TARGET_SUPERSEDED, function (eventName, instance, data) {
        _this.handleLookFinished(instance, data.target, true);
    });
    lookatBuilder.on(this._animate.LookatEventType.STOPPED, function (eventName, instance, data) {
        _this.handleInstanceFinished(instance, data.interrupted);
    });
    lookatBuilder.on(this._animate.LookatEventType.CANCELLED, function (eventName, instance, data) {
        _this.handleInstanceFinished(instance, true);
    });
};
AttentionManager.prototype.setTarget = function (position, isHeyJibo, lookatConfig, lastTarget, lookSucceededListener, lookInterruptedListener, rule) {
    var newLook = this._currentInstanceIsMissingDOFs ||
        lookatConfig !== this._lastLookatConfig ||
        isHeyJibo || Bakery.getBoolean("Every Lookat New", false, "AttentionDebug") ||
        lookSucceededListener !== null || lookInterruptedListener !== null;
    var dofs = lookatConfig.getDOFs();
    var body = !lookatConfig.eyeOnly;
    var torsoEnabled = lookatConfig.torsoEnabled;
    if (this._lookatInstance === null || newLook) {
        var availableDOFs = this._dofArbiter.getAvailable(EMChannel, dofs, null);
        if (availableDOFs.length === dofs.length) {
            var lookatBuilder = lookatConfig.cachedBuilder;
            if (lookatBuilder === null || Bakery.getBoolean("Every Lookat New", false, "AttentionDebug")) {
                lookatBuilder = this._animate.createLookatBuilder();
                this.configureNewBuilder(lookatBuilder, lookatConfig);
                lookatConfig.cachedBuilder = lookatBuilder;
            }
            else {
            }
            if (this._eyeRecessor != null && lookatConfig.doRecession && Bakery.getBoolean("Allow Any Eye Recession", true, "ATMotion")) {
                this._eyeRecessor.notifyTargets(lastTarget, position);
            }
            if (isHeyJibo) {
                this.fireEvent(EventType.STARTED_HEY_JIBO, { position: position }, true);
                if (!this._needToFireOnNewWake) {
                    slog.error("AttentionManager unexpected state: firing on hey-jibo, but _needToFireOnNewWake not set.");
                }
                this._needToFireOnNewWake = false;
            }
            this._lookatInstance = this._dofArbiter.startLookat(lookatBuilder, EMChannel, position, null);
            if (this._lookatInstance !== null) {
                this._lastLookatConfig = lookatConfig;
                this._lookResultStatus.push(new LookResultStatus(this._lookatInstance, position, lookSucceededListener, lookInterruptedListener, isHeyJibo));
                this._currentInstanceIsMissingDOFs = false;
                this._currentInstanceRequiredDOFs = dofs;
                if (body && !torsoEnabled && this._defaultBuilder !== null) {
                    var torsoDOFName = this._robotInfo.getDOFSet("BODY").getDOFs()[1];
                    this._defaultBuilder.setDOFs([torsoDOFName]);
                    this._dofArbiter.playAnimation(this._defaultBuilder, EMChannel, null);
                }
            }
            else {
                slog.warn("Unexpected: Attention Manager got 0 dofs from DOF Arbiter when starting a lookat");
            }
            rule.notifyDidFire();
        }
        else {
            slog(channel, "Attention Manager Not Trying to do lookat, we have no dofs!!");
        }
    }
    else {
        if (this._eyeRecessor != null && this._lastLookatConfig !== null && this._lastLookatConfig.doRecession) {
            if (Bakery.getBoolean("Allow Any Eye Recession", true, "ATMotion")) {
                this._eyeRecessor.notifyTargets(lastTarget, position);
                this._eyeRecessor.lookStarted();
            }
        }
        this._lookatInstance.updateTarget(position);
    }
};
var assembleData = function (sensoryStore, currentTime, timestampOfLastUsedHeyJibo) {
    var i;
    var mostRecentHeyJibo = null;
    var trackedWake = sensoryStore.getTracked(SensoryRecord.SensoryType.WAKE);
    if (trackedWake.length > 0 && trackedWake[0].hasOwnProperty("wakeAt")) {
        mostRecentHeyJibo = trackedWake[0];
    }
    var validNewHeyJibo = null;
    if (mostRecentHeyJibo !== null &&
        currentTime.subtract(mostRecentHeyJibo.wakeAt) < Bakery.getFloat("IDLE Hey-is-New time", 0, 10, 2, "ATPolicy") &&
        (timestampOfLastUsedHeyJibo === null || mostRecentHeyJibo.wakeAt.isGreater(timestampOfLastUsedHeyJibo))) {
        validNewHeyJibo = mostRecentHeyJibo;
    }
    var trackedFaces = [];
    var allTrackedFaces = sensoryStore.getTracked(SensoryRecord.SensoryType.VISION);
    for (i = 0; i < allTrackedFaces.length; i++) {
        if (currentTime.subtract(allTrackedFaces[i].lastIncorporatedReport) < Bakery.getFloat("Face Min Recency", 0, 5, 0.3, "ATPolicy")) {
            trackedFaces.push(allTrackedFaces[i]);
        }
    }
    var trackedMotions = [];
    var allTrackedMotions = sensoryStore.getTracked(SensoryRecord.SensoryType.MOTION);
    for (i = 0; i < allTrackedMotions.length; i++) {
        if (currentTime.subtract(allTrackedMotions[i].lastIncorporatedReport) < Bakery.getFloat("Motion Min Recency", 0, 5, 0.3, "ATPolicy")) {
            trackedMotions.push(allTrackedMotions[i]);
        }
    }
    var mostWeighty = null;
    var trackedAudio = sensoryStore.getTracked(SensoryRecord.SensoryType.AUDIO_LOCALIZATION);
    for (i = 0; i < trackedAudio.length; i++) {
        if (mostWeighty === null || trackedAudio[i].weight > mostWeighty.weight) {
            mostWeighty = trackedAudio[i];
        }
    }
    return { newHeyJibo: validNewHeyJibo, heyJibo: mostRecentHeyJibo, mostWeightyAudio: mostWeighty, trackedFaces: trackedFaces, trackedMotions: trackedMotions };
};
AttentionManager.prototype.update = function () {
    var i;
    var overrideMode = Bakery.getFloat("Override Mode", -1, 5, -1, "AttentionDebug");
    if (overrideMode >= 0) {
        var m = Math.floor(overrideMode);
        var modeName = Object.keys(Mode)[m];
        this.setMode(modeName);
    }
    var state = this._rulesState;
    state.clear();
    var attentionCommandHistory = this._attentionCommandHistory;
    var currentTime = Clock.currentTime();
    this._sensoryStore.update();
    var tracks = assembleData(this._sensoryStore, currentTime, this._ignoreHeyJibosAtOrBefore);
    var rules = this._rules[this._mode];
    for (i = 0; i < rules.length; i++) {
        var rule = rules[i];
        var ruleResult = rule.update(currentTime, tracks, state, attentionCommandHistory);
        if (ruleResult !== null) {
            if (state.lookClaimed) {
                throw new Error("Rule " + rule.name + " claimed look but it is already claimed!");
            }
            state.lookClaimed = true;
            var eyeOnly = rule.lookatConfig.eyeOnly;
            if (ruleResult.initLookat === true) {
                if (rule.isHeyJibo) {
                    slog(channel, "We're initing a lookat that counts as \"hey jibo\" for rule " + rule.name);
                }
                rule.lookatConfig.updateValues();
                this.setTarget(ruleResult.target, rule.isHeyJibo, rule.lookatConfig, attentionCommandHistory.lastEyeTarget, rule.succeededHandler, rule.interruptedHandler, rule);
                attentionCommandHistory.lastEyeTarget = ruleResult.target;
                attentionCommandHistory.lastEyeLookRule = rule;
                attentionCommandHistory.lastEyeLookConfig = rule.lookatConfig;
                if (!eyeOnly) {
                    attentionCommandHistory.lastBodyTarget = ruleResult.target;
                    attentionCommandHistory.lastFullLookRule = rule;
                    attentionCommandHistory.lastFullLookConfig = rule.lookatConfig;
                }
            }
            if (rule.isHeyJibo && ruleResult.initLookat === true) {
                this._ignoreHeyJibosAtOrBefore = currentTime;
                attentionCommandHistory.lastHeyJiboTarget = ruleResult.target;
            }
            this.displayTrackState(rule.trackDisplayName);
        }
    }
    if (this._needToFireOnNewWake) {
        this.fireEvent(EventType.IGNORED_HEY_JIBO, {}, true);
        this._ignoreHeyJibosAtOrBefore = currentTime;
        this._needToFireOnNewWake = false;
    }
    if (this._lookatInstance !== null && this._eyeRecessor !== null && this._eyeRecessor.getIsRecessed()) {
        if (this._lookatInstance.getBuilder().motionLookat != null && this._lookatInstance.getBuilder().motionLookat.getDistanceRemaining(["eyeSubRootBn_t", "eyeSubRootBn_t_2"]) < 0.001) {
            this._eyeRecessor.lookEnded();
        }
    }
    for (i = this._faceAwaiters.length - 1; i >= 0; i--) {
        var keepGoing = this._faceAwaiters[i].update(currentTime, tracks);
        if (!keepGoing) {
            this._faceAwaiters.splice(i, 1);
        }
    }
    AttentionEventDispatcher.dispatchQueuedEvents();
};
module.exports = AttentionManager;

},{"./AttentionDebugStateVisualizer":2,"./AttentionEventDispatcher":3,"./AttentionEyeRecessor":4,"./DataConverter":7,"./MotorMotionMonitor":10,"./SensoryRecord":11,"./SensoryStore":12,"./actions/AttendHandle":16,"./actions/FaceAwaiter":18,"./actions/ResultStatus":19,"./actions/TridentSearchIterator":21,"./log/SensoryLogController":23,"./lookconfig/LookConfig":26,"./lookconfig/LookConfigDisengage":27,"./lookconfig/LookConfigEyeVision":28,"./lookconfig/LookConfigShort":29,"./lookconfig/LookConfigSlow":30,"./lookconfig/LookConfigUncommitted":31,"./remote/RemoteVisualizerConnection":41,"animation-utilities":undefined,"find-root":undefined}],6:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var Bakery = AnimationUtilities.ui.Bakery;
var ModalityHandler = require("./ModalityHandler");
var TrackedSensory = require("./TrackedSensory");
var SensoryRecord = require("./SensoryRecord");
var THREE = AnimationUtilities.THREE;
var AudioModalityHandler = function (motionTracker) {
    ModalityHandler.call(this);
    this.motionTracker = motionTracker;
    this.distanceThreshold = 0.3;
    this.confidenceThreshold = 0.1;
};
AudioModalityHandler.prototype = Object.create(ModalityHandler.prototype);
AudioModalityHandler.prototype.constructor = AudioModalityHandler;
AudioModalityHandler.prototype.isValid = function (sensoryRecord) {
    if (sensoryRecord.raw.type !== 0) {
        console.log("AudioModalityHandler: Unexpected packet type: " + JSON.stringify(sensoryRecord.raw));
        return false;
    }
    var rejectMovingAudio = Bakery.getBoolean("Reject audio during motion", true, "SensoryFilters");
    var rejectMovingAudioRecoveryTime = Bakery.getFloat("Reject moving audio recovery time", 0, 2, 0.7, "SensoryFilters");
    var confidence = sensoryRecord.confidence;
    var accept;
    if (confidence > Bakery.getFloat("Confidence Threshold (Audio)", 0, 1, this.confidenceThreshold, "SensoryFilters")) {
        if (rejectMovingAudio && this.motionTracker.timeSinceMoving() < rejectMovingAudioRecoveryTime) {
            accept = false;
        }
        else {
            accept = true;
        }
    }
    else {
        accept = false;
    }
    return accept;
};
AudioModalityHandler.prototype.accept = function (sensoryRecord) {
    var matching = null;
    var minHeight = Bakery.getFloat("Localized-Audio Height Floor", -1, 1, 0.7, "SensoryFilters");
    if (sensoryRecord.position.z < minHeight) {
        sensoryRecord.position.z = minHeight;
    }
    this.distanceThreshold = Bakery.getFloat("Distance Threshold (Audio)", 0, 1, this.distanceThreshold, "SensoryFilters");
    for (var i = 0; i < this.tracked.length; i++) {
        if (this.tracked[i].position.distanceTo(sensoryRecord.position) < this.distanceThreshold) {
            matching = this.tracked[i];
            break;
        }
    }
    if (matching == null) {
        matching = new TrackedSensory(SensoryRecord.SensoryType.AUDIO_LOCALIZATION);
        matching.weight = 0;
        matching.fastWeight = 0;
        matching.weightHistory = new Array(30);
        matching.weightHistoryIndex = 0;
        this.tracked.push(matching);
    }
    matching.incorporateSensorReport(sensoryRecord);
    var addWeight = sensoryRecord.confidence * sensoryRecord.confidence;
    matching.weight += addWeight;
    matching.fastWeight += addWeight;
    if (matching.weight > 10) {
        matching.weight = 10;
    }
    if (matching.fastWeight > 10) {
        matching.fastWeight = 10;
    }
};
AudioModalityHandler.prototype.update = function () {
    var i;
    var curTime = Clock.currentTime();
    for (i = 0; i < this.tracked.length; i++) {
        if (curTime.subtract(this.tracked[i].lastIncorporatedReport) > Bakery.getFloat("Delay before weight-fall begins", 0, 1, 0.4, "SensoryFilters")) {
            this.tracked[i].weight *= 0.95;
        }
        this.tracked[i].fastWeight *= 0.9;
        this.tracked[i].weightHistory[this.tracked[i].weightHistoryIndex] = this.tracked[i].fastWeight;
        this.tracked[i].weightHistoryIndex = (this.tracked[i].weightHistoryIndex + 1) % this.tracked[i].weightHistory.length;
    }
    var horizontalMerge2 = Math.pow(Bakery.getFloat("HorizontalMerge", 0, 1, 0.2, "SensoryFilters"), 2);
    var iV = new THREE.Vector3(), jV = new THREE.Vector3();
    for (i = 0; i < this.tracked.length; i++) {
        iV.set(this.tracked[i].position.x, this.tracked[i].position.y, 0);
        for (var j = i + 1; j < this.tracked.length; j++) {
            jV.set(this.tracked[j].position.x, this.tracked[j].position.y, 0);
            if (iV.distanceToSquared(jV) < horizontalMerge2) {
                if (this.tracked[j].position.z < this.tracked[i].position.z) {
                    this.tracked[i].weight += this.tracked[j].weight;
                    this.tracked[i].fastWeight += this.tracked[j].fastWeight;
                    this.tracked[j].weight = 0;
                    this.tracked[j].fastWeight = 0;
                }
                else {
                    this.tracked[j].weight += this.tracked[i].weight;
                    this.tracked[j].fastWeight += this.tracked[i].fastWeight;
                    this.tracked[i].weight = 0;
                    this.tracked[i].fastWeight = 0;
                }
            }
        }
    }
    i = 0;
    while (i < this.tracked.length) {
        var last = this.tracked[i].lastIncorporatedReport;
        if (last != null && Clock.currentTime().subtract(last) > 5 && !this.tracked[i].hasOwnProperty("wakeAt")) {
            this.tracked.splice(i, 1);
        }
        else {
            i++;
        }
    }
};
module.exports = AudioModalityHandler;

},{"./ModalityHandler":8,"./SensoryRecord":11,"./TrackedSensory":14,"animation-utilities":undefined}],7:[function(require,module,exports){
"use strict";
var SensoryRecord = require("./SensoryRecord");
var AnimationUtilities = require("animation-utilities");
var THREE = AnimationUtilities.THREE;
var Time = AnimationUtilities.Time;
var Clock = AnimationUtilities.Clock;
var DataConverter = function (sensoryStore, attentionManager, passWakeFromLocalization) {
    if (passWakeFromLocalization === undefined || passWakeFromLocalization === null) {
        passWakeFromLocalization = false;
    }
    this._sensoryStore = sensoryStore;
    this._lastVisionTimestamp = null;
    this._lastAudioLocalizationTimestamp = null;
    this._lastLocalizedWakeTimestamp = null;
    this._attentionManager = attentionManager || null;
    this._passWakeFromLocalization = passWakeFromLocalization;
};
DataConverter.prototype.acceptAudioLocalization = function (data) {
    var thisTS;
    for (var aei = 0; aei < data.entities.length; aei++) {
        var entity = data.entities[aei];
        var sensoryType = null;
        if (entity.type === 0) {
            thisTS = new Time(entity.ts[0], entity.ts[1]);
            if (this._lastAudioLocalizationTimestamp === null || thisTS.isGreater(this._lastAudioLocalizationTimestamp)) {
                sensoryType = SensoryRecord.SensoryType.AUDIO_LOCALIZATION;
                this._lastAudioLocalizationTimestamp = thisTS;
            }
        }
        else if (entity.type === 1 && this._passWakeFromLocalization) {
            thisTS = new Time(entity.ts[0], entity.ts[1]);
            if (this._lastLocalizedWakeTimestamp === null || thisTS.isGreater(this._lastLocalizedWakeTimestamp)) {
                sensoryType = SensoryRecord.SensoryType.WAKE;
                this._lastLocalizedWakeTimestamp = thisTS;
            }
        }
        if (sensoryType !== null) {
            var sr = new SensoryRecord(sensoryType, new THREE.Vector3(entity.position.x, entity.position.y, entity.position.z), entity, Clock.currentTime(), thisTS, entity.id, entity.confidence);
            if (sensoryType === SensoryRecord.SensoryType.WAKE) {
                this._notifyWakeHappened(sr);
            }
            this._sensoryStore.addData(sr);
        }
    }
};
DataConverter.prototype.acceptVision = function (data) {
    var dataTimestamp = new Time(data.ts[0], data.ts[1]);
    if (this._lastVisionTimestamp === null || !this._lastVisionTimestamp.equals(dataTimestamp)) {
        this._lastVisionTimestamp = dataTimestamp;
        var entities = data.entities;
        for (var vei = 0; vei < entities.length; vei++) {
            var sensoryType = SensoryRecord.SensoryType.VISION;
            if (entities[vei].description === "motion") {
                sensoryType = SensoryRecord.SensoryType.MOTION;
            }
            var sr = new SensoryRecord(sensoryType, new THREE.Vector3(entities[vei].position.x, entities[vei].position.y, entities[vei].position.z), entities[vei], Clock.currentTime(), dataTimestamp, entities[vei].id, entities[vei].confidence);
            this._sensoryStore.addData(sr);
        }
    }
};
DataConverter.prototype.acceptWakeWord = function (data) {
    var sr = new SensoryRecord(SensoryRecord.SensoryType.WAKE, null, data, Clock.currentTime(), null, null, null);
    this._notifyWakeHappened(sr);
    this._sensoryStore.addData(sr);
};
DataConverter.prototype._notifyWakeHappened = function (sensoryRecord) {
    if (this._attentionManager !== null) {
        this._attentionManager.notifyWakeHappened(sensoryRecord);
    }
};
module.exports = DataConverter;

},{"./SensoryRecord":11,"animation-utilities":undefined}],8:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var ModalityHandler = function () {
    this.tracked = [];
};
ModalityHandler.prototype.isValid = function (sensoryRecord) {
    return true;
};
ModalityHandler.prototype.accept = function (sensoryRecord) {
};
ModalityHandler.prototype.update = function () {
    var i = 0;
    while (i < this.tracked.length) {
        var last = this.tracked[i].lastIncorporatedReport;
        if (last != null && Clock.currentTime().subtract(last) > 5) {
            this.tracked.splice(i, 1);
        }
        else {
            i++;
        }
    }
};
ModalityHandler.prototype.getTracked = function () {
    return this.tracked;
};
module.exports = ModalityHandler;

},{"animation-utilities":undefined}],9:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Time = AnimationUtilities.Time;
var Bakery = AnimationUtilities.ui.Bakery;
var ModalityHandler = require("./ModalityHandler");
var TrackedSensory = require("./TrackedSensory");
var SensoryRecord = require("./SensoryRecord");
var THREE = AnimationUtilities.THREE;
var MotionModalityHandler = function (motionTracker) {
    ModalityHandler.call(this);
    this.motionTracker = motionTracker;
    this.confidenceThreshold = 0;
    this.distanceThreshold = 0.0;
};
MotionModalityHandler.prototype = Object.create(ModalityHandler.prototype);
MotionModalityHandler.prototype.constructor = MotionModalityHandler;
MotionModalityHandler.prototype.isValid = function (sensoryRecord) {
    var r = sensoryRecord.raw;
    if (sensoryRecord.confidence < Bakery.getFloat("Confidence Threshold (Motion)", 0, 1, this.confidenceThreshold, "SensoryFilters")) {
        return false;
    }
    var rejectMovingData = Bakery.getBoolean("Reject motion data during motion", false, "SensoryFilters");
    var rejectMovingDataRecoveryTime = Bakery.getFloat("Reject moving motion data recovery time", 0, 6, 4, "SensoryFilters");
    if (rejectMovingData && this.motionTracker.timeSinceMoving() < rejectMovingDataRecoveryTime) {
        return false;
    }
    var ray = null;
    var rect = null;
    var trackerUpdateTimestamp = null;
    var detectionTimestamp = null;
    var inFOV = false;
    if (r.parts.length > 0) {
        var part = r.parts[0];
        if (part.value.rays.length > 0) {
            ray = part.value.rays[0];
        }
        if (part.value.trackers.length > 0) {
            var tracker = part.value.trackers[0];
            rect = tracker.rectangle;
            if (tracker.lastUpdate) {
                trackerUpdateTimestamp = Time.createFromTimestamp(tracker.lastUpdate);
            }
            inFOV = tracker.inFOV;
        }
        if (part.value.detections.length > 0) {
            var detection = part.value.detections[0];
            if (detection.timestamp) {
                detectionTimestamp = Time.createFromTimestamp(detection.timestamp);
            }
        }
    }
    var requireDetection = Bakery.getBoolean("Require Motion Detection", false, "SensoryFilters");
    if (requireDetection) {
        if (trackerUpdateTimestamp === null || detectionTimestamp === null) {
            return false;
        }
        else {
            var detectionAge = trackerUpdateTimestamp.subtract(detectionTimestamp);
            if (detectionAge > Bakery.getFloat("Max Motion Detection Age", 0, 15, 6, "SensoryFilters")) {
                return false;
            }
        }
    }
    var requireRay = Bakery.getBoolean("Require Motion Ray", false, "SensoryFilters");
    if (requireRay) {
        if (ray === null) {
            return false;
        }
        else if (Bakery.getBoolean("Use Ray Instead of position (Motion)", false, "SensoryFilters")) {
            var newPos = new THREE.Vector3(ray.origin.x, ray.origin.y, ray.origin.z);
            var d = new THREE.Vector3(ray.dir.x, ray.dir.y, ray.dir.z);
            d.multiplyScalar(2);
            newPos.add(d);
            sensoryRecord.position = newPos;
        }
    }
    var requireFOV = Bakery.getBoolean("Require Motion In FOV", true, "SensoryFilters");
    if (requireFOV && !inFOV) {
        return false;
    }
    var requireRect = Bakery.getBoolean("Require Motion Rect", false, "SensoryFilters");
    if (requireRect) {
        if (rect != null) {
            var height = Math.abs(rect.top - rect.bottom);
            if (height > Bakery.getFloat("Min Height Motion", 0, 500, 100, "SensoryFilters") &&
                height < Bakery.getFloat("Max Height Motion", 0, 1000, 500, "SensoryFilters")) {
                return true;
            }
            else {
                return false;
            }
        }
        else {
            return false;
        }
    }
    else {
        return true;
    }
};
MotionModalityHandler.prototype.accept = function (sensoryRecord) {
    var matching = null;
    var i;
    for (i = 0; i < this.tracked.length; i++) {
        if (this.tracked[i].id === sensoryRecord.id) {
            matching = this.tracked[i];
            break;
        }
    }
    this.distanceThreshold = Bakery.getFloat("Distance Threshold (Motion)", 0, 10, this.distanceThreshold, "SensoryFilters");
    if (matching === null) {
        for (i = 0; i < this.tracked.length; i++) {
            if (this.tracked[i].position.distanceTo(sensoryRecord.position) < this.distanceThreshold) {
                matching = this.tracked[i];
                break;
            }
        }
    }
    if (matching === null) {
        matching = new TrackedSensory(SensoryRecord.SensoryType.MOTION);
        matching.id = sensoryRecord.id;
        this.tracked.push(matching);
    }
    matching.slowFilter = Bakery.getFloat("Slow Filter (Motion)", 0, 1, 1, "SensoryFilters");
    matching.fastFilter = Bakery.getFloat("Fast Filter (Motion)", 0, 1, 1, "SensoryFilters");
    matching.incorporateSensorReport(sensoryRecord);
    matching.id = sensoryRecord.id;
};
module.exports = MotionModalityHandler;

},{"./ModalityHandler":8,"./SensoryRecord":11,"./TrackedSensory":14,"animation-utilities":undefined}],10:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var MotorMotionMonitor = function (monitorDOFs, animationUtilities) {
    this._animationUtilities = animationUtilities;
    this._realPose = new Array(monitorDOFs.length);
    this._lastPose = new Array(monitorDOFs.length);
    this._dofsToIndex = {};
    for (var i = 0; i < monitorDOFs.length; i++) {
        this._dofsToIndex[monitorDOFs[i]] = i;
        this._realPose[i] = 0;
        this._lastPose[i] = 0;
    }
    this._epsilon = 0.02;
    this._moving = false;
    this._lastMotionTime = null;
    if (this._animationUtilities != null) {
        this.registerForMotion();
    }
};
MotorMotionMonitor.prototype.update = function () {
    var moving = false;
    for (var i = 0; i < this._realPose.length; i++) {
        var delta = Math.abs(this._realPose[i] - this._lastPose[i]);
        if (delta > this._epsilon) {
            moving = true;
        }
        this._lastPose[i] = this._realPose[i];
    }
    if (moving) {
        this._lastMotionTime = Clock.currentTime();
    }
    this._moving = moving;
};
MotorMotionMonitor.prototype.isMoving = function () {
    return this._moving;
};
MotorMotionMonitor.prototype.timeSinceMoving = function () {
    if (this._moving) {
        return 0;
    }
    else if (this._lastMotionTime != null) {
        return Clock.currentTime().subtract(this._lastMotionTime);
    }
    else {
        return Number.POSITIVE_INFINITY;
    }
};
MotorMotionMonitor.prototype.registerForMotion = function () {
    var self = this;
    var outputs = this._animationUtilities.timeline.getOutputs();
    var infoListener = function (bodyInfo) {
        self._realPose[self._dofsToIndex[bodyInfo.dofName]] = bodyInfo.observedPosition;
    };
    for (var i = 0; i < outputs.length; i++) {
        if (outputs[i].addInfoListener !== undefined) {
            outputs[i].addInfoListener(infoListener);
        }
    }
};
module.exports = MotorMotionMonitor;

},{"animation-utilities":undefined}],11:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var THREE = AnimationUtilities.THREE;
var Time = AnimationUtilities.Time;
var SensoryType = {
    AUDIO_LOCALIZATION: "AUDIO_LOCALIZATION",
    MOTION: "MOTION",
    VISION: "VISION",
    WAKE: "WAKE"
};
var SensoryRecord = function (type, position, rawData, receivedAt, reportedTS, id, confidence) {
    this.type = type;
    this.position = null;
    if (position != null) {
        this.position = position;
    }
    this.raw = rawData;
    this.receivedAt = receivedAt;
    this.reportedTS = reportedTS;
    this.id = id;
    this.confidence = confidence;
};
SensoryRecord.SensoryType = SensoryType;
SensoryRecord.createFromJSON = function (json) {
    var position = null;
    if (json.position != null) {
        position = new THREE.Vector3(json.position.x, json.position.y, json.position.z);
    }
    var receivedAt = null;
    if (json.receivedAt != null) {
        receivedAt = new Time(json.receivedAt._timestamp[0], json.receivedAt._timestamp[1]);
    }
    var reportedTS = null;
    if (json.reportedTS != null) {
        reportedTS = new Time(json.reportedTS._timestamp[0], json.reportedTS._timestamp[1]);
    }
    return new SensoryRecord(json.type, position, json.raw, receivedAt, reportedTS, json.id, json.confidence);
};
module.exports = SensoryRecord;

},{"animation-utilities":undefined}],12:[function(require,module,exports){
"use strict";
var SensoryRecord = require("./SensoryRecord");
var VisionModalityHandler = require("./VisionModalityHandler");
var MotionModalityHandler = require("./MotionModalityHandler");
var AudioModalityHandler = require("./AudioModalityHandler");
var ALWakeModalityHandler = require("./ALWakeModalityHandler");
var SensoryStore = function (motionTracker) {
    this.sensoryListeners = null;
    this.motionTracker = motionTracker;
    this.modalityHanders = {};
    this.modalityHanders[SensoryRecord.SensoryType.AUDIO_LOCALIZATION] = new AudioModalityHandler(motionTracker);
    this.modalityHanders[SensoryRecord.SensoryType.MOTION] = new MotionModalityHandler(motionTracker);
    this.modalityHanders[SensoryRecord.SensoryType.VISION] = new VisionModalityHandler(motionTracker);
    this.modalityHanders[SensoryRecord.SensoryType.WAKE] = new ALWakeModalityHandler(motionTracker);
};
SensoryStore.prototype.addSensoryListener = function (sensoryListener) {
    if (this.sensoryListeners === null) {
        this.sensoryListeners = [];
    }
    if (this.sensoryListeners.indexOf(sensoryListener) < 0) {
        this.sensoryListeners.push(sensoryListener);
    }
};
SensoryStore.prototype.removeSensoryListener = function (sensoryListener) {
    if (this.sensoryListeners !== null) {
        var index = this.sensoryListeners.indexOf(sensoryListener);
        if (index >= 0) {
            this.sensoryListeners.splice(index, 1);
            if (this.sensoryListeners.length === 0) {
                this.sensoryListeners = null;
            }
        }
    }
};
SensoryStore.prototype.addData = function (sensoryRecord) {
    var handler = this.modalityHanders[sensoryRecord.type];
    var ok = false;
    if (handler != null) {
        ok = handler.isValid(sensoryRecord);
    }
    if (this.sensoryListeners !== null) {
        for (var i = 0; i < this.sensoryListeners.length; i++) {
            this.sensoryListeners[i].rawDataAdded(sensoryRecord, ok);
        }
    }
    if (ok) {
        handler.accept(sensoryRecord);
    }
};
SensoryStore.prototype.update = function () {
    var i;
    if (this.motionTracker != null) {
        this.motionTracker.update();
    }
    var modalityKeys = Object.keys(this.modalityHanders);
    var allTracked = [];
    for (i = 0; i < modalityKeys.length; i++) {
        this.modalityHanders[modalityKeys[i]].update();
        if (this.sensoryListeners !== null) {
            allTracked = allTracked.concat(this.modalityHanders[modalityKeys[i]].getTracked());
        }
    }
    if (this.sensoryListeners != null) {
        for (i = 0; i < this.sensoryListeners.length; i++) {
            this.sensoryListeners[i].notifyCurrentTracked(allTracked);
        }
    }
};
SensoryStore.prototype.getTracked = function (modality) {
    if (modality == null) {
        var allTracked = [];
        var modalityKeys = Object.keys(this.modalityHanders);
        for (var i = 0; i < modalityKeys.length; i++) {
            allTracked = allTracked.concat(this.modalityHanders[modalityKeys[i]].getTracked());
        }
        return allTracked;
    }
    else {
        return this.modalityHanders[modality].getTracked();
    }
};
module.exports = SensoryStore;

},{"./ALWakeModalityHandler":1,"./AudioModalityHandler":6,"./MotionModalityHandler":9,"./SensoryRecord":11,"./VisionModalityHandler":15}],13:[function(require,module,exports){
"use strict";
var SimpleAnimGenerator = {
    channelFromValue: function (dofName, value) {
        return {
            "dofName": dofName,
            "times": [
                0.0
            ],
            "values": [
                value
            ],
            "length": 0.1
        };
    },
    animFromColor: function (name, r, g, b, dofPrefix) {
        dofPrefix = dofPrefix || "eye";
        return {
            "header": {
                "fileType": "DOFAnimation",
                "version": "0.002",
                "creationTime": new Date().getTime()
            },
            "content": {
                "name": name,
                "channels": [
                    SimpleAnimGenerator.channelFromValue(dofPrefix + "_redChannelBn_r", r),
                    SimpleAnimGenerator.channelFromValue(dofPrefix + "_greenChannelBn_r", g),
                    SimpleAnimGenerator.channelFromValue(dofPrefix + "_blueChannelBn_r", b)
                ]
            }
        };
    },
    animFromColorAndScale: function (name, r, g, b, scale, robotInfo) {
        var defaultVals = robotInfo.getDefaultDOFValues();
        var eyeDeformerNames = robotInfo.getDOFSet("EYE_DEFORM").getDOFs();
        var channels = [
            SimpleAnimGenerator.channelFromValue("eye_redChannelBn_r", r),
            SimpleAnimGenerator.channelFromValue("eye_greenChannelBn_r", g),
            SimpleAnimGenerator.channelFromValue("eye_blueChannelBn_r", b)
        ];
        for (var i = 0; i < eyeDeformerNames.length; i++) {
            var dofName = eyeDeformerNames[i];
            channels.push(SimpleAnimGenerator.channelFromValue(dofName, defaultVals[dofName] * scale));
        }
        return {
            "header": {
                "fileType": "DOFAnimation",
                "version": "0.002",
                "creationTime": new Date().getTime()
            },
            "content": {
                "name": name,
                "channels": channels
            }
        };
    },
    animFromAlphaAndScale: function (name, a, scale, robotInfo) {
        var defaultVals = robotInfo.getDefaultDOFValues();
        var eyeDeformerNames = robotInfo.getDOFSet("EYE_DEFORM").getDOFs();
        var channels = [
            SimpleAnimGenerator.channelFromValue("eye_alphaChannelBn_r", a)
        ];
        for (var i = 0; i < eyeDeformerNames.length; i++) {
            var dofName = eyeDeformerNames[i];
            channels.push(SimpleAnimGenerator.channelFromValue(dofName, defaultVals[dofName] * scale));
        }
        return {
            "header": {
                "fileType": "DOFAnimation",
                "version": "0.002",
                "creationTime": new Date().getTime()
            },
            "content": {
                "name": name,
                "channels": channels
            }
        };
    }
};
module.exports = SimpleAnimGenerator;

},{}],14:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var THREE = AnimationUtilities.THREE;
var TrackedSensory = function (type) {
    this.type = type;
    this.position = null;
    this.fastPosition = null;
    this.creationTime = Clock.currentTime();
    this.lastIncorporatedReport = null;
    this.slowFilter = 0.05;
    this.fastFilter = 0.2;
};
TrackedSensory.prototype.incorporateSensorReport = function (sensorReport) {
    if (this.position === null) {
        this.position = new THREE.Vector3().copy(sensorReport.position);
    }
    else {
        this.position.lerp(sensorReport.position, this.slowFilter);
    }
    if (this.fastPosition === null) {
        this.fastPosition = new THREE.Vector3().copy(sensorReport.position);
    }
    else {
        this.fastPosition.lerp(sensorReport.position, this.fastFilter);
    }
    this.lastIncorporatedReport = Clock.currentTime();
};
module.exports = TrackedSensory;

},{"animation-utilities":undefined}],15:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Time = AnimationUtilities.Time;
var Bakery = AnimationUtilities.ui.Bakery;
var ModalityHandler = require("./ModalityHandler");
var TrackedSensory = require("./TrackedSensory");
var SensoryRecord = require("./SensoryRecord");
var THREE = AnimationUtilities.THREE;
var VisionModalityHandler = function (motionTracker) {
    ModalityHandler.call(this);
    this.motionTracker = motionTracker;
    this.confidenceThreshold = 0.3;
    this.distanceThreshold = 0.0;
};
VisionModalityHandler.prototype = Object.create(ModalityHandler.prototype);
VisionModalityHandler.prototype.constructor = VisionModalityHandler;
VisionModalityHandler.prototype.isValid = function (sensoryRecord) {
    var r = sensoryRecord.raw;
    if (sensoryRecord.confidence < Bakery.getFloat("Confidence Threshold (Face)", 0, 1, this.confidenceThreshold, "SensoryFilters")) {
        return false;
    }
    var rejectMovingData = Bakery.getBoolean("Reject face data during motion", false, "SensoryFilters");
    var rejectMovingDataRecoveryTime = Bakery.getFloat("Reject moving face data recovery time", 0, 2, 0.7, "SensoryFilters");
    if (rejectMovingData && this.motionTracker.timeSinceMoving() < rejectMovingDataRecoveryTime) {
        return false;
    }
    var ray = null;
    var rect = null;
    var trackerUpdateTimestamp = null;
    var detectionTimestamp = null;
    var inFOV = false;
    if (r.parts.length > 0) {
        var part = r.parts[0];
        if (part.value.rays.length > 0) {
            ray = part.value.rays[0];
        }
        if (part.value.trackers.length > 0) {
            var tracker = part.value.trackers[0];
            rect = tracker.rectangle;
            if (tracker.lastUpdate) {
                trackerUpdateTimestamp = Time.createFromTimestamp(tracker.lastUpdate);
            }
            inFOV = tracker.inFOV;
        }
        if (part.value.detections.length > 0) {
            var detection = part.value.detections[0];
            if (detection.timestamp) {
                detectionTimestamp = Time.createFromTimestamp(detection.timestamp);
            }
        }
    }
    var requireDetection = Bakery.getBoolean("Require Face Detection", false, "SensoryFilters");
    if (requireDetection) {
        if (trackerUpdateTimestamp === null || detectionTimestamp === null) {
            return false;
        }
        else {
            var detectionAge = trackerUpdateTimestamp.subtract(detectionTimestamp);
            if (detectionAge > Bakery.getFloat("Max Face Detection Age", 0, 30, 15, "SensoryFilters")) {
                return false;
            }
        }
    }
    var requireRay = Bakery.getBoolean("Require Face Ray", false, "SensoryFilters");
    if (requireRay) {
        if (ray === null) {
            return false;
        }
        else if (Bakery.getBoolean("Use Ray Instead of position", false, "SensoryFilters")) {
            var newPos = new THREE.Vector3(ray.origin.x, ray.origin.y, ray.origin.z);
            var d = new THREE.Vector3(ray.dir.x, ray.dir.y, ray.dir.z);
            d.multiplyScalar(2);
            newPos.add(d);
            sensoryRecord.position = newPos;
        }
    }
    var requireFOV = Bakery.getBoolean("Require Face In FOV", true, "SensoryFilters");
    if (requireFOV && !inFOV) {
        return false;
    }
    var requireRect = Bakery.getBoolean("Require Face Rect", false, "SensoryFilters");
    if (requireRect) {
        if (rect != null) {
            var height = Math.abs(rect.top - rect.bottom);
            if (height > Bakery.getFloat("Min Height Face", 0, 500, 10, "SensoryFilters") &&
                height < Bakery.getFloat("Max Height Face", 0, 1000, 500, "SensoryFilters")) {
                return true;
            }
            else {
                return false;
            }
        }
        else {
            return false;
        }
    }
    else {
        return true;
    }
};
VisionModalityHandler.prototype.accept = function (sensoryRecord) {
    var matching = null;
    var i;
    for (i = 0; i < this.tracked.length; i++) {
        if (this.tracked[i].id === sensoryRecord.id) {
            matching = this.tracked[i];
            break;
        }
    }
    this.distanceThreshold = Bakery.getFloat("Distance Threshold (Face)", 0, 1, this.distanceThreshold, "SensoryFilters");
    if (matching === null) {
        for (i = 0; i < this.tracked.length; i++) {
            if (this.tracked[i].position.distanceTo(sensoryRecord.position) < this.distanceThreshold) {
                matching = this.tracked[i];
                break;
            }
        }
    }
    if (matching == null) {
        matching = new TrackedSensory(SensoryRecord.SensoryType.VISION);
        matching.id = sensoryRecord.id;
        this.tracked.push(matching);
    }
    matching.slowFilter = Bakery.getFloat("Slow Filter (Face)", 0, 1, 1, "SensoryFilters");
    matching.fastFilter = Bakery.getFloat("Fast Filter (Face)", 0, 1, 1, "SensoryFilters");
    matching.incorporateSensorReport(sensoryRecord);
    matching.id = sensoryRecord.id;
};
module.exports = VisionModalityHandler;

},{"./ModalityHandler":8,"./SensoryRecord":11,"./TrackedSensory":14,"animation-utilities":undefined}],16:[function(require,module,exports){
"use strict";
var AttendHandle = function () {
    this.promise = null;
    this.result = null;
};
AttendHandle.prototype.cancel = function () {
};
module.exports = AttendHandle;

},{}],17:[function(require,module,exports){
"use strict";
var AwaitFaceHandle = function () {
    this.promise = null;
    this.result = null;
};
AwaitFaceHandle.prototype.cancel = function () {
};
module.exports = AwaitFaceHandle;

},{}],18:[function(require,module,exports){
"use strict";
var AwaitFaceHandle = require("./AwaitFaceHandle");
var ResultStatus = require("./ResultStatus");
var AnimationUtilities = require("animation-utilities");
var slog = AnimationUtilities.slog;
var FaceAwaiter = function (currentTime, timeout, maxAngle, fullSearchTime, animate, lpsConnection) {
    var _this = this;
    this._startTime = currentTime;
    this._timeout = timeout;
    this._maxAngle = maxAngle;
    this._fullSearchTime = fullSearchTime;
    this._animate = animate;
    this._lpsConnection = lpsConnection;
    this._minRecency = 0.1;
    this._timeoutImmediatelyIfDemandFindsNoone = true;
    this._haveDemandedFaces = false;
    this._demandFacesFoundNoFaces = false;
    this._promiseResolve = null;
    this._awaitHandle = new AwaitFaceHandle();
    this._awaitHandle.promise = new Promise(function (resolve, reject) {
        _this._promiseResolve = resolve;
    });
    this._awaitHandle.cancel = function () {
        _this.resolveCompletion(ResultStatus.CANCELLED, null, null, null);
    };
};
FaceAwaiter.prototype.resolveCompletion = function (status, entity, position, angle) {
    if (this._awaitHandle != null) {
        var result = { status: status, entity: entity, position: position, angle: angle };
        var handle = this._awaitHandle;
        this._awaitHandle = null;
        handle.result = result;
        this._promiseResolve(result);
        this._promiseResolve = null;
    }
};
FaceAwaiter.prototype.getHandle = function () {
    return this._awaitHandle;
};
FaceAwaiter.prototype.update = function (currentTime, perceptionState) {
    var _this = this;
    if (this._awaitHandle === null) {
        return false;
    }
    var kinematicFeatures = this._animate.getKinematicFeatures();
    var eyePos = kinematicFeatures["Eye"].position;
    var eyeDir = kinematicFeatures["Eye"].direction;
    var enforceMaxAngle = this._maxAngle !== 0;
    var currentMaxAngle = this._maxAngle;
    if (enforceMaxAngle && this._fullSearchTime > 0) {
        var alpha = Math.max(0, Math.min(1, currentTime.subtract(this._startTime) / this._fullSearchTime));
        currentMaxAngle = this._maxAngle * alpha;
    }
    var faces = perceptionState.trackedFaces;
    var winningFace = null;
    var winningFaceAngle = Number.MAX_VALUE;
    for (var i = 0; i < faces.length; i++) {
        var face = faces[i];
        if (currentTime.subtract(face.lastIncorporatedReport) > this._minRecency) {
            continue;
        }
        var angle = face.position.clone().sub(eyePos).angleTo(eyeDir);
        if (enforceMaxAngle && angle > currentMaxAngle) {
            continue;
        }
        if (angle < winningFaceAngle) {
            winningFace = face;
            winningFaceAngle = angle;
        }
    }
    if (winningFace !== null) {
        this.resolveCompletion(ResultStatus.SUCCEEDED, null, winningFace.position.clone(), winningFaceAngle);
        return false;
    }
    else if (currentTime.subtract(this._startTime) > this._timeout) {
        this.resolveCompletion(ResultStatus.TIMEOUT, null, null, null);
        return false;
    }
    else if (this._timeoutImmediatelyIfDemandFindsNoone && this._demandFacesFoundNoFaces) {
        slog("ATTENTION", "Exiting await-face with timeout at dt " + currentTime.subtract(this._startTime) + " < timeout " + this._timeout + " because demandFaces found no faces");
        this.resolveCompletion(ResultStatus.TIMEOUT, null, null, null);
        return false;
    }
    if (this._timeout > 0.5 && this._haveDemandedFaces === false && this._lpsConnection !== null) {
        this._haveDemandedFaces = true;
        this._lpsConnection.demandFaceDetect(function (numFacesDetected) {
            slog("ATTENTION", "Demand Face Detect found " + numFacesDetected + " faces");
            if (numFacesDetected === 0) {
                _this._demandFacesFoundNoFaces = true;
            }
        });
    }
    return true;
};
module.exports = FaceAwaiter;

},{"./AwaitFaceHandle":17,"./ResultStatus":19,"animation-utilities":undefined}],19:[function(require,module,exports){
"use strict";
var ResultStatus = {
    SUCCEEDED: "SUCCEEDED",
    TIMEOUT: "TIMEOUT",
    INTERRUPTED: "INTERRUPTED",
    CANCELLED: "CANCELLED"
};
module.exports = ResultStatus;

},{}],20:[function(require,module,exports){
"use strict";
var SearchIterator = function () {
};
SearchIterator.prototype.nextTarget = function () {
};
module.exports = SearchIterator;

},{}],21:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var THREE = AnimationUtilities.THREE;
var slog = AnimationUtilities.slog;
var SearchIterator = require("./SearchIterator");
var CHANNEL = "ATTENTION";
var TridentSearchIterator = function (currentTarget) {
    SearchIterator.call(this);
    this._initialTarget = currentTarget;
    this._targetNumber = 0;
};
TridentSearchIterator.prototype = Object.create(SearchIterator.prototype);
TridentSearchIterator.prototype.constructor = TridentSearchIterator;
TridentSearchIterator.prototype.nextTarget = function () {
    var newTarget = this._initialTarget.clone().projectOnPlane(new THREE.Vector3(0, 0, 1));
    newTarget.setLength(2);
    var angle = 0;
    if (this._targetNumber % 4 === 0) {
        angle = Math.PI / 4.8;
    }
    else if (this._targetNumber % 4 === 1) {
        angle = -Math.PI / 4.8;
    }
    else if (this._targetNumber % 4 === 2) {
        angle = -Math.PI;
    }
    else if (this._targetNumber % 4 === 3) {
        angle = 0;
    }
    var rotation = new THREE.Quaternion().setFromAxisAngle(new THREE.Vector3(0, 0, 1), angle);
    newTarget.applyQuaternion(rotation);
    var height = (Math.random() * 0.4) + 0.7;
    newTarget.add(new THREE.Vector3(0, 0, height));
    slog(CHANNEL, "Trident iterator returning target (" + newTarget.x + ", " + newTarget.y + ", " + newTarget.z + "), started (" + this._initialTarget.x + ", " + this._initialTarget.y + ", " + this._initialTarget.z + "), num " + this._targetNumber);
    this._targetNumber++;
    return newTarget;
};
module.exports = TridentSearchIterator;

},{"./SearchIterator":20,"animation-utilities":undefined}],22:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var AttentionLog = {};
AttentionLog.createLog = function (filename) {
    var fs = "fs";
    fs = require(fs);
    var attentionStream = fs.createWriteStream(filename);
    var first = true;
    return function (info) {
        if (first) {
            first = false;
        }
        else {
            attentionStream.write(",\n");
        }
        attentionStream.write(JSON.stringify(info));
    };
};
AttentionLog.loadLog = function (filename, cb) {
    var FileTools = AnimationUtilities.core.FileTools;
    FileTools.loadText(filename, function (error, data) {
        if (error) {
            cb(error, null);
        }
        else {
            var infoArray = null;
            try {
                var infoString = "[" + data + "]";
                infoArray = JSON.parse(infoString);
            }
            catch (e) {
                cb(e, null);
            }
            cb(null, infoArray);
        }
    });
};
module.exports = AttentionLog;

},{"animation-utilities":undefined}],23:[function(require,module,exports){
"use strict";
var SensoryLogger = require("./SensoryLogger");
var SensoryLogDataSource = require("./SensoryLogDataSource");
var SensoryLogController = function (sensoryStore) {
    this._sensoryStore = sensoryStore;
    this._blockLiveData = false;
    this._sensoryLogger = new SensoryLogger(sensoryStore);
    this._sensoryLogDataSource = new SensoryLogDataSource();
    this._respectFilter = false;
    this._isPlaying = false;
};
SensoryLogController.prototype.startLogging = function (filename) {
    this._sensoryLogger.startLogging(filename);
};
SensoryLogController.prototype.stopLogging = function () {
    this._sensoryLogger.stopLogging();
};
SensoryLogController.prototype.startPlayingLog = function (filename) {
    this._sensoryLogDataSource.playLog(filename);
};
SensoryLogController.prototype.stopPlayingLog = function () {
    this._sensoryLogDataSource.stop();
};
SensoryLogController.prototype.setBlockLiveData = function (block) {
    this._blockLiveData = block;
};
SensoryLogController.prototype.getBlockLiveData = function () {
    return this._blockLiveData || this._isPlaying;
};
SensoryLogController.prototype.getLogPlaybackStatus = function () {
    return this._sensoryLogDataSource.getPlayStatus();
};
SensoryLogController.prototype.update = function () {
    var playStatus = this._sensoryLogDataSource.getPlayStatus();
    if (playStatus !== null && playStatus.playing) {
        this._isPlaying = true;
        var newRecords = this._sensoryLogDataSource.getNewRecords(this._respectFilter);
        if (newRecords !== null && newRecords.length > 0) {
            for (var i = 0; i < newRecords.length; i++) {
                this._sensoryStore.addData(newRecords[i]);
            }
        }
    }
    else {
        this._isPlaying = false;
    }
};
module.exports = SensoryLogController;

},{"./SensoryLogDataSource":24,"./SensoryLogger":25}],24:[function(require,module,exports){
"use strict";
var AttentionLog = require("./AttentionLog");
var SensoryRecord = require("../SensoryRecord");
var Clock = require("animation-utilities").Clock;
var SensoryLogDataSource = function () {
    this._logData = null;
    this._logFileName = null;
    this._recordedDataStartTime = null;
    this._recordedDataEndTime = null;
    this._playStartTime = null;
    this._cachedTime = null;
    this._indexForCachedTime = 0;
    this._lastIterativeReadTime = null;
};
SensoryLogDataSource.prototype.resetState = function () {
    this._logData = null;
    this._recordedDataStartTime = null;
    this._recordedDataEndTime = null;
    this._playStartTime = null;
    this._cachedTime = null;
    this._indexForCachedTime = 0;
    this._lastIterativeReadTime = null;
    this._logFileName = null;
};
SensoryLogDataSource.prototype.playLog = function (filename) {
    var self = this;
    var logAcceptor = function (error, logData) {
        self.resetState();
        if (error !== null) {
            console.log("SensoryLogDataSource:Error loading log \"" + filename + "\": " + error);
        }
        else {
            if (logData != null && logData.length > 0) {
                self._logData = [];
                for (var i = 0; i < logData.length; i++) {
                    var sensoryJSON = logData[i];
                    var sr = SensoryRecord.createFromJSON(sensoryJSON.record);
                    self._logData.push({
                        record: sr,
                        passedFilter: sensoryJSON.passedFilter,
                        originalReceiptTime: sr.receivedAt
                    });
                }
                self._recordedDataStartTime = self._logData[0].record.receivedAt;
                self._recordedDataEndTime = self._logData[self._logData.length - 1].record.receivedAt;
                self._playStartTime = Clock.currentTime();
                self._lastIterativeReadTime = self._playStartTime;
                self._logFileName = filename;
                console.log("SensoryLogDataSource:Starting to play a log of length " + self._recordedDataEndTime.subtract(self._recordedDataStartTime));
            }
            else {
                console.log("SensoryLogDataSource:Not playing log (missing or 0 length) " + filename);
            }
        }
    };
    AttentionLog.loadLog(filename, logAcceptor);
};
SensoryLogDataSource.prototype.stop = function () {
    this.resetState();
};
SensoryLogDataSource.prototype.getLogFileName = function () {
    return this._logFileName;
};
SensoryLogDataSource.prototype.getPlayStatus = function () {
    if (this._logFileName !== null) {
        var time = this._lastIterativeReadTime.subtract(this._playStartTime);
        var duration = this._recordedDataEndTime.subtract(this._recordedDataStartTime);
        return {
            time: time,
            duration: duration,
            logFileName: this._logFileName,
            playing: (duration > 0) && time < duration
        };
    }
    else {
        return null;
    }
};
var indexForTimeSearchForward = function (time, startingIndex, records) {
    for (var i = startingIndex; i < records.length; i++) {
        if (records[i].originalReceiptTime.isGreaterOrEqual(time)) {
            return i;
        }
    }
    return -1;
};
var convertTimeBase = function (oldStartTime, newStartTime, inplaceRecords) {
    for (var i = 0; i < inplaceRecords.length; i++) {
        inplaceRecords[i].record.receivedAt = newStartTime.add(inplaceRecords[i].originalReceiptTime.subtract(oldStartTime));
    }
};
SensoryLogDataSource.prototype.getRecordsInTimeRange = function (startTime, endTime) {
    if (this._logData !== null) {
        var startFromIndex = 0;
        if (this._cachedTime !== null && startTime.isGreaterOrEqual(this._cachedTime)) {
            startFromIndex = this._indexForCachedTime;
        }
        else {
            console.log("Need to do full search for time " + startTime + ", is not greater than cached " + this._cachedTime);
        }
        var foundIndex = indexForTimeSearchForward(startTime, startFromIndex, this._logData);
        if (foundIndex !== -1) {
            var matching = [];
            for (var i = foundIndex; i < this._logData.length; i++) {
                if (endTime.isGreater(this._logData[i].originalReceiptTime)) {
                    matching.push(this._logData[i]);
                    this._cachedTime = this._logData[i].originalReceiptTime;
                    this._indexForCachedTime = i;
                }
                else {
                    break;
                }
            }
            return matching;
        }
    }
    return null;
};
SensoryLogDataSource.prototype.getNewRecords = function (respectFilter) {
    if (this._logData !== null) {
        var curTime = Clock.currentTime();
        var deltaToStart = this._lastIterativeReadTime.subtract(this._playStartTime);
        var deltaToEnd = curTime.subtract(this._playStartTime);
        var startTime = this._recordedDataStartTime.add(deltaToStart);
        var endTime = this._recordedDataStartTime.add(deltaToEnd);
        var recordedRecords = this.getRecordsInTimeRange(startTime, endTime);
        convertTimeBase(this._recordedDataStartTime, this._playStartTime, recordedRecords);
        var records = [];
        for (var i = 0; i < recordedRecords.length; i++) {
            if (!respectFilter || recordedRecords[i].passedFilter) {
                records.push(recordedRecords[i].record);
            }
        }
        this._lastIterativeReadTime = curTime;
        return records;
    }
    return null;
};
module.exports = SensoryLogDataSource;

},{"../SensoryRecord":11,"./AttentionLog":22,"animation-utilities":undefined}],25:[function(require,module,exports){
"use strict";
var AttentionLog = require("./AttentionLog");
var SensoryLogger = function (sensoryStore) {
    this._sensoryStore = sensoryStore;
    this._activeLog = null;
    this._logFileName = null;
};
SensoryLogger.prototype.startLogging = function (filename) {
    this._activeLog = AttentionLog.createLog(filename);
    this._logFileName = filename;
    this._sensoryStore.addSensoryListener(this);
};
SensoryLogger.prototype.stopLogging = function () {
    this._sensoryStore.removeSensoryListener(this);
    this._activeLog = null;
    this._logFileName = null;
};
SensoryLogger.prototype.rawDataAdded = function (rawRecord, passedFilter) {
    if (this._activeLog !== null) {
        this._activeLog({ record: rawRecord, passedFilter: passedFilter });
    }
};
SensoryLogger.prototype.getCurrentLogTarget = function () {
    return this._logFileName;
};
SensoryLogger.prototype.notifyCurrentTracked = function (allTrackedData) { };
module.exports = SensoryLogger;

},{"./AttentionLog":22}],26:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Bakery = AnimationUtilities.ui.Bakery;
var computeLimits = function (robotInfo) {
    var eyeHalfWidth = Math.abs(robotInfo.getDefaultDOFValues()["vertexJoint1_t"]);
    var eyeHalfHeight = Math.abs(robotInfo.getDefaultDOFValues()["vertexJoint1_t_2"]);
    var screenWidth = robotInfo.getEyeScreenInfo().getWidth();
    var screenHeight = robotInfo.getEyeScreenInfo().getHeight();
    var xLimitNormal = Math.max(0, screenWidth / 2 - eyeHalfWidth);
    var yLimitNormal = Math.max(0, screenHeight / 2 - eyeHalfHeight);
    return [xLimitNormal, yLimitNormal];
};
var LookConfig = function (name, robotInfo, eyeOnly) {
    this.name = name;
    this.doWindupOvershoot = true;
    this.doSaccadeBlink = true;
    this.doRecession = true;
    this.robotInfo = robotInfo;
    this.recessionThreshold = 0;
    this.eyeOnly = eyeOnly;
    this.torsoEnabled = true;
    var limits = computeLimits(robotInfo);
    this.normalEyeLimitX = limits[0];
    this.normalEyeLimitY = limits[1];
    this.cachedBuilder = null;
    this._allDOFs = robotInfo.getDOFSet("EYE_TRANSLATE").plus("BODY").getDOFs();
    this._eyeOnlyDOFs = robotInfo.getDOFSet("EYE_TRANSLATE").getDOFs();
    this._allWithoutTorsoDOFs = robotInfo.getDOFSet("EYE_TRANSLATE").getDOFs();
    this._allWithoutTorsoDOFs.push(robotInfo.getDOFSet("BODY").getDOFs()[0]);
    this._allWithoutTorsoDOFs.push(robotInfo.getDOFSet("BODY").getDOFs()[2]);
};
LookConfig.prototype.updateValues = function () {
    this.doWindupOvershoot = Bakery.getBoolean("LC:" + this.name + " do windup/overshoot", this.doWindupOvershoot, "ATMotion");
    this.doSaccadeBlink = Bakery.getBoolean("LC:" + this.name + " do saccade blink", this.doSaccadeBlink, "ATMotion");
    this.doRecession = Bakery.getBoolean("LC:" + this.name + " do recession", this.doRecession, "ATMotion");
    this.eyeOnly = Bakery.getBoolean("LC:" + this.name + " eye only", this.eyeOnly, "ATMotion");
    this.torsoEnabled = Bakery.getBoolean("LC:" + this.name + " torso enabled", this.torsoEnabled, "ATMotion");
};
LookConfig.prototype.lookatConfigEyeWindupOvershoot = function (lookatBuilder) {
    var eyeConfig = lookatBuilder.getLookatNodeConfig("Eye");
    eyeConfig.WPTargetDeltaToTriggerNewWindup = Bakery.getFloat("Eye Windup Trigger On Target Delta", 0, 10, 0.2, "ATMotion");
    eyeConfig.WPMaxAllowedTriggerSpeed = Bakery.getFloat("Eye Windup min trigger speed", 0, 1, 0.005, "ATMotion");
    eyeConfig.WPMinAllowedTriggerDistance = Bakery.getFloat("Eye Windup min trigger distance", 0, 1, 0.01, "ATMotion");
    eyeConfig.WPMaxAllowedTriggerDistance = Bakery.getFloat("Eye Windup max trigger distance", 0, 100000, 100000, "ATMotion");
    eyeConfig.WMWindupDistanceRatio = eyeConfig.WPOvershootDistanceRatio = Bakery.getFloat("Eye windup distance ratio", 0, 2, 0.05, "ATMotion");
    eyeConfig.WPWindupMinDistance = eyeConfig.WPOvershootMinDistance = Bakery.getFloat("Eye windup min distance", 0, 0.1, 0.001, "ATMotion");
    eyeConfig.WPWindupMaxDistance = eyeConfig.WPOvershootMaxDistance = Bakery.getFloat("Eye windup max distance", 0, 0.1, 0.002, "ATMotion");
    var eyeXDofConfig = lookatBuilder.getLookatDOFConfig("eyeSubRootBn_t");
    eyeXDofConfig.min = Math.min(0, -this.normalEyeLimitX + eyeConfig.WPWindupMaxDistance);
    eyeXDofConfig.max = Math.max(0, this.normalEyeLimitX - eyeConfig.WPWindupMaxDistance);
    var eyeYDofConfig = lookatBuilder.getLookatDOFConfig("eyeSubRootBn_t_2");
    eyeYDofConfig.min = Math.min(0, -this.normalEyeLimitY + eyeConfig.WPWindupMaxDistance);
    eyeYDofConfig.max = Math.max(0, this.normalEyeLimitY - eyeConfig.WPWindupMaxDistance);
};
LookConfig.prototype.lookatConfigSaccadeBlink = function (lookatBuilder) {
    var eyeConfig = lookatBuilder.getLookatNodeConfig("Eye");
    eyeConfig.BPTriggerDelta = Bakery.getFloat("Saccade Blink trigger distance", 0, 0.1, 0.02, "ATMotion");
    eyeConfig.BPOnlyAtOrAfterWindupPhase = Bakery.getFloat("Saccade Blink wait until windup phase", 0, 4, 4, "ATMotion");
};
LookConfig.prototype.configureBuilder = function (lookatBuilder) {
    if (this.doWindupOvershoot && Bakery.getBoolean("Do Any Eye Windup", true, "ATMotion")) {
        this.lookatConfigEyeWindupOvershoot(lookatBuilder);
    }
    if (this.doSaccadeBlink && Bakery.getBoolean("Do Any Saccade Blink", true, "ATMotion")) {
        this.lookatConfigSaccadeBlink(lookatBuilder);
    }
};
LookConfig.prototype.getDOFs = function () {
    if (this.eyeOnly) {
        return this._eyeOnlyDOFs;
    }
    else if (this.torsoEnabled) {
        return this._allDOFs;
    }
    else {
        return this._allWithoutTorsoDOFs;
    }
};
module.exports = LookConfig;

},{"animation-utilities":undefined}],27:[function(require,module,exports){
"use strict";
var LookConfigSlow = require("./LookConfigSlow");
var LookConfigDisengage = function (name, robotInfo) {
    LookConfigSlow.call(this, name, robotInfo, false);
};
LookConfigDisengage.prototype = Object.create(LookConfigSlow.prototype);
LookConfigDisengage.prototype.constructor = LookConfigDisengage;
LookConfigDisengage.prototype.configureBuilder = function (lookatBuilder) {
    LookConfigSlow.prototype.configureBuilder.call(this, lookatBuilder);
    var baseConfig = lookatBuilder.getLookatNodeConfig("BaseLookatNode");
    baseConfig.TPTDLimitInner = baseConfig.TPTDLimitOuter = 10000;
    baseConfig.TPTDMoveImmediatelyPastOuter = false;
    var topConfig = lookatBuilder.getLookatNodeConfig("TopLookatNode");
    topConfig.TAUndershootDistance = 0.33;
};
module.exports = LookConfigDisengage;

},{"./LookConfigSlow":30}],28:[function(require,module,exports){
"use strict";
var LookConfig = require("./LookConfig");
var AnimationUtilities = require("animation-utilities");
var Bakery = AnimationUtilities.ui.Bakery;
var LookConfigEyeVision = function (name, robotInfo, modalityName, eyeOnly, torsoEnabled) {
    LookConfig.call(this, name, robotInfo, eyeOnly);
    this.modalityName = modalityName;
    this.torsoEnabled = torsoEnabled;
    this.recessionThreshold = 0.3;
};
LookConfigEyeVision.prototype = Object.create(LookConfig.prototype);
LookConfigEyeVision.prototype.constructor = LookConfigEyeVision;
LookConfigEyeVision.prototype.configureBuilder = function (lookatBuilder) {
    LookConfig.prototype.configureBuilder.call(this, lookatBuilder);
    var eyeConfig = lookatBuilder.getLookatNodeConfig("Eye");
    eyeConfig.acceleration = Bakery.getFloat("Eye accel: " + this.modalityName, 0, 5, 1, "ATMotion");
    eyeConfig.TPTDLimitInner = Bakery.getFloat("Eye dead zone: " + this.modalityName, 0, 0.1, 0.002, "ATMotion");
    eyeConfig.TPTDLimitOuter = eyeConfig.TPTDLimitInner;
    eyeConfig.TPTDAccumInner = 0;
    eyeConfig.TPTDAccumOuter = 0;
    eyeConfig.TPTDMoveImmediatelyPastOuter = true;
    eyeConfig.TPTMTDeadZone = eyeConfig.TPTDLimitInner;
    eyeConfig.TPTMTDeadTime = Bakery.getFloat("Eye dead time: " + this.modalityName, 0, 0.5, 0.05, "ATMotion");
    eyeConfig.TPTMTDeadVelocity = Bakery.getFloat("Eye dead velocity: " + this.modalityName, 0, 10, 0.001, "ATMotion");
    if (!this.eyeOnly) {
        var baseConfig = lookatBuilder.getLookatNodeConfig("BaseLookatNode");
        baseConfig.TPTDLimitInner = Math.PI / 6;
        baseConfig.TPTDLimitOuter = Math.PI / 2;
        baseConfig.TPTDAccumInner = 0.2;
        baseConfig.TPTDAccumOuter = 0.5;
        baseConfig.TPTDMoveImmediatelyPastOuter = false;
        var torsoConfig = lookatBuilder.getLookatNodeConfig("TorsoLookatNode");
        torsoConfig.TPTDLimitInner = Math.PI / 18;
        torsoConfig.TPTDLimitOuter = Math.PI / 6;
        torsoConfig.TPTDAccumInner = 0;
        torsoConfig.TPTDAccumOuter = 0;
        torsoConfig.TPTDMoveImmediatelyPastOuter = true;
        torsoConfig.TPTPMoveIfParentMoves = true;
        var topConfig = lookatBuilder.getLookatNodeConfig("TopLookatNode");
        topConfig.TPTDLimitInner = Math.PI / 50;
        topConfig.TPTDLimitOuter = Math.PI / 6;
        topConfig.TPTDAccumInner = 0.15;
        topConfig.TPTDAccumOuter = 1;
        topConfig.TPTDMoveImmediatelyPastOuter = true;
        topConfig.TPTPMoveIfParentMoves = true;
    }
};
module.exports = LookConfigEyeVision;

},{"./LookConfig":26,"animation-utilities":undefined}],29:[function(require,module,exports){
"use strict";
var LookConfig = require("./LookConfig");
var LookConfigShort = function (name, robotInfo) {
    LookConfig.call(this, name, robotInfo, true);
    this.doSaccadeBlink = false;
    this.doRecession = false;
};
LookConfigShort.prototype = Object.create(LookConfig.prototype);
LookConfigShort.prototype.constructor = LookConfigShort;
LookConfigShort.prototype.configureBuilder = function (lookatBuilder) {
    LookConfig.prototype.configureBuilder.call(this, lookatBuilder);
    lookatBuilder.getLookatNodeConfig("BaseLookatNode").acceleration = 5;
    lookatBuilder.getLookatNodeConfig("TorsoLookatNode").acceleration = 4;
    lookatBuilder.getLookatNodeConfig("TopLookatNode").acceleration = 5;
    lookatBuilder.getLookatNodeConfig("Eye").acceleration = 1.7;
};
module.exports = LookConfigShort;

},{"./LookConfig":26}],30:[function(require,module,exports){
"use strict";
var LookConfig = require("./LookConfig");
var LookConfigSlow = function (name, robotInfo, eyeOnly) {
    LookConfig.call(this, name, robotInfo, eyeOnly);
};
LookConfigSlow.prototype = Object.create(LookConfig.prototype);
LookConfigSlow.prototype.constructor = LookConfigSlow;
LookConfigSlow.prototype.configureBuilder = function (lookatBuilder) {
    LookConfig.prototype.configureBuilder.call(this, lookatBuilder);
    lookatBuilder.getLookatNodeConfig("BaseLookatNode").acceleration = 1;
    lookatBuilder.getLookatNodeConfig("TorsoLookatNode").acceleration = 0.8;
    lookatBuilder.getLookatNodeConfig("TopLookatNode").acceleration = 1;
    lookatBuilder.getLookatNodeConfig("Eye").acceleration = 0.33;
};
module.exports = LookConfigSlow;

},{"./LookConfig":26}],31:[function(require,module,exports){
"use strict";
var LookConfigSlow = require("./LookConfigSlow");
var LookConfigUncommitted = function (name, robotInfo) {
    LookConfigSlow.call(this, name, robotInfo, false);
};
LookConfigUncommitted.prototype = Object.create(LookConfigSlow.prototype);
LookConfigUncommitted.prototype.constructor = LookConfigUncommitted;
LookConfigUncommitted.prototype.configureBuilder = function (lookatBuilder) {
    LookConfigSlow.prototype.configureBuilder.call(this, lookatBuilder);
    var baseConfig = lookatBuilder.getLookatNodeConfig("BaseLookatNode");
    baseConfig.TAUndershootDistance = 0.33;
    var torsoConfig = lookatBuilder.getLookatNodeConfig("TorsoLookatNode");
    torsoConfig.TAUndershootDistance = 0.15;
    var topConfig = lookatBuilder.getLookatNodeConfig("TopLookatNode");
    topConfig.TAUndershootDistance = 0.33;
};
module.exports = LookConfigUncommitted;

},{"./LookConfigSlow":30}],32:[function(require,module,exports){
"use strict";
var slog = require("animation-utilities").slog;
var channel = "ATTENTION";
var AttentionModeHandler = function (attentionServer, attentionManager) {
    this._lastReportedMode = null;
    this._attentionManager = attentionManager;
    this._attentionServer = attentionServer;
    attentionServer.registerHandler("ChangeMode", function (type, data) {
        if (data.switchToMode) {
            var newMode = data.switchToMode;
            slog(channel, "Switching to attention mode " + newMode + " based on remote request");
            attentionManager.setMode(newMode);
        }
    });
};
AttentionModeHandler.prototype.update = function () {
    var m = this._attentionManager.getMode();
    if (m !== this._lastReportedMode) {
        this._lastReportedMode = m;
        this._attentionServer.sendData("CurrentMode", { currentMode: m });
    }
};
module.exports = AttentionModeHandler;

},{"animation-utilities":undefined}],33:[function(require,module,exports){
"use strict";
var WebSocketServer = require("websocket-without-native").server;
var http = require("http");
var slog = require("animation-utilities").slog;
var channel = "ATTENTION";
var AttentionServer = function () {
    this.connections = [];
    this.info = {};
    this.info.realPose = {};
    this.handlers = {};
    slog(channel, "creating attention server");
    var port = 9197;
    var server = http.createServer(function (request, response) {
        slog(channel, (new Date()) + " Attention Server is listening on port " + port);
        response.writeHead(404);
        response.end();
    });
    server.listen(port, function () {
        console.log((new Date()) + " Attention Server is listening on port " + port);
    });
    var wsServer = new WebSocketServer({ httpServer: server, autoAcceptConnections: false });
    var self = this;
    wsServer.on('request', function (request) {
        var connection = request.accept(null, request.origin);
        slog(channel, (new Date()) + " AttentionServer: Connection accepted.");
        self.connections.push(connection);
        connection.on('message', function (message) {
            if (message.type === 'utf8') {
                try {
                    var messageData = JSON.parse(message.utf8Data);
                    if (messageData.hasOwnProperty("type")) {
                        self.fireHandlers(messageData.type, messageData.data);
                    }
                }
                catch (e) {
                    slog(channel, "JSON parse failed on incoming message: " + message + " error: " + e);
                }
            }
            else if (message.type === 'binary') {
                slog(channel, 'Ignoring binary message');
            }
        });
        connection.on('close', function (reasonCode, description) {
            slog(channel, (new Date()) + " AttentionServer: Peer disconnected.");
            if (self.connections.indexOf(connection) > -1) {
                self.connections.splice(self.connections.indexOf(connection), 1);
            }
        });
    });
};
AttentionServer.prototype.hasClients = function () {
    return this.connections.length > 0;
};
AttentionServer.prototype.sendData = function (type, data) {
    var packet = { type: type, data: data };
    for (var i = 0; i < this.connections.length; i++) {
        var connection = this.connections[i];
        connection.send(JSON.stringify(packet));
    }
};
AttentionServer.prototype.fireHandlers = function (type, data) {
    if (this.handlers[type] != null) {
        for (var i = 0; i < this.handlers[type].length; i++) {
            this.handlers[type][i](type, data);
        }
    }
};
AttentionServer.prototype.registerHandler = function (type, handler) {
    if (this.handlers[type] == null) {
        this.handlers[type] = [];
    }
    this.handlers[type].push(handler);
};
AttentionServer.prototype.removeHandler = function (type, handler) {
    if (this.handlers[type] != null) {
        var index = this.handlers[type].indexOf(handler);
        if (index >= 0) {
            this.handlers[type].splice(index, 1);
        }
        else {
            slog(channel, "Error, asked to remove AS handler (" + type + "), but it was not registered");
        }
        if (this.handlers[type].length === 0) {
            delete (this.handlers[type]);
        }
    }
    else {
        slog(channel, "Error, asked to remove AS handler (" + type + "), but it was not registered");
    }
};
module.exports = AttentionServer;

},{"animation-utilities":undefined,"http":undefined,"websocket-without-native":undefined}],34:[function(require,module,exports){
"use strict";
var BodyStreamer = function (attentionServer, animate) {
    this.realPose = null;
    this.attentionServer = attentionServer;
    this.animate = animate;
    var self = this;
    var outputs = this.animate.timeline.getOutputs();
    for (var i = 0; i < outputs.length; i++) {
        if (outputs[i].addInfoListener !== undefined) {
            outputs[i].addInfoListener(function (bodyInfo) {
                if (self.realPose === null) {
                    self.realPose = {};
                }
                self.realPose[bodyInfo.dofName] = bodyInfo.observedPosition;
            });
        }
    }
};
BodyStreamer.prototype.update = function () {
    if (this.realPose !== null) {
        this.attentionServer.sendData("RealPose", this.realPose);
    }
};
module.exports = BodyStreamer;

},{}],35:[function(require,module,exports){
"use strict";
var DoAnimHandler = function (attentionServer, animate) {
    attentionServer.registerHandler("PlayAnimation", function (type, data) {
        animate.createAnimationBuilder(data.filename, function (builder) {
            if (data.layer != null) {
                builder.setLayer(data.layer);
            }
            if (data.dofs != null) {
                builder.setDOFs(data.dofs);
            }
            if (data.speed != null) {
                builder.setSpeed(data.speed);
            }
            builder.play();
        });
    });
};
module.exports = DoAnimHandler;

},{}],36:[function(require,module,exports){
"use strict";
var AttentionDebugStateVisualizer = require("../AttentionDebugStateVisualizer");
var DoColorHandler = function (attentionServer, animate) {
    var attentionStateVisualizer = new AttentionDebugStateVisualizer(animate);
    attentionServer.registerHandler("DoDebugState", function (type, data) {
        attentionStateVisualizer.handleState(data.state, data.force);
    });
};
module.exports = DoColorHandler;

},{"../AttentionDebugStateVisualizer":2}],37:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Bakery = AnimationUtilities.ui.Bakery;
var DoLookHandler = function (attentionServer, animate, lookatConfigs, attentionManager, dofArbiter, emChannel) {
    var lookatInstance = null;
    attentionServer.registerHandler("DoLook", function (type, data) {
        if (lookatInstance === null || data.newLook) {
            var lookatBuilder = animate.createLookatBuilder();
            lookatBuilder.setContinuousMode(true);
            if (data.dofs) {
                lookatBuilder.setDOFs(data.dofs);
            }
            if (data.disengage) {
                lookatConfigs.lookConfigDisengage.configureBuilder(lookatBuilder);
            }
            else {
                lookatConfigs.lookConfigDefault.configureBuilder(lookatBuilder);
            }
            if (data.innerEyeDistanceX) {
                console.log("Permanently modifying inner eye X to " + data.innerEyeDistanceX);
                lookatBuilder.constructor.LookatDOFGeometryConfig.EyeLeftRight.InternalDistance = data.innerEyeDistanceX;
            }
            if (data.innerEyeDistanceY) {
                console.log("Permanently modifying inner eye Y to " + data.innerEyeDistanceY);
                lookatBuilder.constructor.LookatDOFGeometryConfig.EyeUpDown.InternalDistance = data.innerEyeDistanceY;
            }
            if (data.eyeColorDebug != null) {
                attentionManager.setEnableEyeColorStateVisualizer(data.eyeColorDebug);
            }
            if (Bakery.getBoolean("Do Eye Recession", true, "ATMotion")) {
                attentionManager._eyeRecessor.attachToBuilder(lookatBuilder);
            }
            lookatInstance = dofArbiter.startLookat(lookatBuilder, emChannel, data.target);
        }
        else {
            lookatInstance.updateTarget(data.target);
        }
    });
};
module.exports = DoLookHandler;

},{"animation-utilities":undefined}],38:[function(require,module,exports){
"use strict";
var SensoryRecord = require("../SensoryRecord");
var ParsedSensoryStreamer = function (attentionServer) {
    this._attentionServer = attentionServer;
    this._sendLightRecords = true;
};
ParsedSensoryStreamer.prototype.rawDataAdded = function (rawRecord, passedFilter) {
    if (this._attentionServer.hasClients()) {
        var recordToSend = rawRecord;
        if (this._sendLightRecords) {
            recordToSend = new SensoryRecord(null, null, null, null, null);
            var keys = Object.keys(rawRecord);
            for (var i = 0; i < keys.length; i++) {
                var k = keys[i];
                if (k !== "raw") {
                    recordToSend[k] = rawRecord[k];
                }
            }
        }
        this._attentionServer.sendData("RawDataAdded", { record: recordToSend, passed: passedFilter });
    }
};
ParsedSensoryStreamer.prototype.notifyCurrentTracked = function (allTrackedData) {
    this._attentionServer.sendData("TrackedData", allTrackedData);
};
module.exports = ParsedSensoryStreamer;

},{"../SensoryRecord":11}],39:[function(require,module,exports){
"use strict";
var AttentionManager = require("../AttentionManager");
var slog = require("animation-utilities").slog;
var channel = "ATTENTION";
var RemoteAttentionManager = function (robotInfo, sensoryVisualizer, attentionSocket) {
    AttentionManager.call(this, robotInfo, sensoryVisualizer);
    this._attentionSocket = attentionSocket;
    var self = this;
    attentionSocket.registerHandler("AudioLocation", function (type, data) {
        self.getDataConverter().acceptAudioLocalization(data);
    });
    attentionSocket.registerHandler("VisionData", function (type, entities) {
        self.getDataConverter().acceptVision(entities);
    });
    attentionSocket.registerHandler("HeyJibo", function (type, data) {
        self.getDataConverter().acceptWakeWord(data);
    });
    attentionSocket.registerHandler("RealPose", function (type, data) {
        var poseKeys = Object.keys(data);
        var tracker = self._motionTracker;
        for (var p = 0; p < poseKeys.length; p++) {
            var dofName = poseKeys[p];
            var index = tracker._dofsToIndex[dofName];
            tracker._realPose[index] = data[dofName];
        }
    });
};
RemoteAttentionManager.prototype = Object.create(AttentionManager.prototype);
RemoteAttentionManager.prototype.constructor = RemoteAttentionManager;
RemoteAttentionManager.prototype.displayTrackState = function (trackState) {
    var data = { state: trackState };
    this._attentionSocket.sendData("DoDebugState", data);
};
RemoteAttentionManager.prototype.setTarget = function (position, body, isHeyJibo, slow) {
    var lookCommand = {};
    lookCommand.target = position.toArray();
    lookCommand.newLook = body !== this._lastBody;
    if (lookCommand.newLook) {
        slog(channel, "Requesting **NEW** lookat!");
    }
    if (!body) {
        lookCommand.dofs = this._robotInfo.getDOFSet("EYE_ROOT").getDOFs();
    }
    this._lastBody = body;
    this._attentionSocket.sendData("DoLook", lookCommand);
};
module.exports = RemoteAttentionManager;

},{"../AttentionManager":5,"animation-utilities":undefined}],40:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var RemoteLogController = function (attentionServer, logController) {
    attentionServer.registerHandler("LogControl", function (type, data) {
        var command = data.command;
        if (command === "StartRecording") {
            console.log("RemoteLogController attempting to start record to log file " + data.filename);
            logController.startLogging(data.filename);
        }
        else if (command === "StopRecording") {
            console.log("RemoteLogController attempting to stop recording");
            logController.stopLogging();
        }
        else if (command === "StartPlayback") {
            console.log("RemoteLogController attempting to start playing log file " + data.filename);
            logController.startPlayingLog(data.filename);
        }
        else if (command === "StopPlayback") {
            console.log("RemoteLogController attempting to stop playing log");
            logController.stopPlayingLog();
        }
        else if (command === "SetBlockLive") {
            var blockLive = data.blocklive === true;
            console.log("RemoteLogController attempting to set blocklive to " + blockLive);
            logController.setBlockLiveData(blockLive);
        }
    });
    this._logController = logController;
    this._attentionServer = attentionServer;
    this._lastStatusTime = null;
};
RemoteLogController.prototype.update = function () {
    var curTime = Clock.currentTime();
    if (this._lastStatusTime === null || curTime.subtract(this._lastStatusTime) > 0.5) {
        var status = this._logController.getLogPlaybackStatus();
        var blockLive = this._logController.getBlockLiveData();
        this._attentionServer.sendData("LogStatus", { playStatus: status, blockLive: blockLive });
        this._lastStatusTime = curTime;
    }
};
module.exports = RemoteLogController;

},{"animation-utilities":undefined}],41:[function(require,module,exports){
"use strict";
var AttentionServer = require("./AttentionServer");
var BodyStreamer = require("./BodyStreamer");
var ParsedSensoryStreamer = require("./ParsedSensoryStreamer");
var DoLookHandler = require("./DoLookHandler");
var DoColorHandler = require("./DoColorHandler");
var DoAnimHandler = require("./DoAnimHandler");
var AttentionModeHandler = require("./AttentionModeHandler");
var RemoteLogController = require("./RemoteLogController");
var RemoteVisualizerConnection = function (animate, attentionManager, logController, dofArbiter, emChannel) {
    this._updateables = [];
    this._attentionServer = new AttentionServer();
    this._bodyStreamer = new BodyStreamer(this._attentionServer, animate);
    this._updateables.push(this._bodyStreamer);
    this._parsedSensoryStreamer = new ParsedSensoryStreamer(this._attentionServer);
    attentionManager._sensoryStore.addSensoryListener(this._parsedSensoryStreamer);
    this._remoteLogController = new RemoteLogController(this._attentionServer, logController);
    this._updateables.push(this._remoteLogController);
    this._doLookHandler = new DoLookHandler(this._attentionServer, animate, attentionManager._lookatConfigs, attentionManager, dofArbiter, emChannel);
    this._doColorHandler = new DoColorHandler(this._attentionServer, animate);
    this._doAnimHandler = new DoAnimHandler(this._attentionServer, animate);
    this._attentionModeHandler = null;
    if (attentionManager != null) {
        this._attentionModeHandler = new AttentionModeHandler(this._attentionServer, attentionManager);
        this._updateables.push(this._attentionModeHandler);
    }
};
RemoteVisualizerConnection.prototype.update = function () {
    for (var u = 0; u < this._updateables.length; u++) {
        this._updateables[u].update();
    }
};
module.exports = RemoteVisualizerConnection;

},{"./AttentionModeHandler":32,"./AttentionServer":33,"./BodyStreamer":34,"./DoAnimHandler":35,"./DoColorHandler":36,"./DoLookHandler":37,"./ParsedSensoryStreamer":38,"./RemoteLogController":40}],42:[function(require,module,exports){
"use strict";
var AnimationUtilities = require("animation-utilities");
var Clock = AnimationUtilities.Clock;
var THREE = AnimationUtilities.THREE;
var Bakery = AnimationUtilities.ui.Bakery;
var AnchoredTargetVisualizer = AnimationUtilities.graphics.AnchoredTargetVisualizer;
var GLTextOverlayPool = AnimationUtilities.graphics.GLTextOverlayPool;
var SensoryRecord = require("../SensoryRecord");
var GLLinePool = AnimationUtilities.graphics.GLLinePool;
var groundPlaneNormal = new THREE.Vector3(0, 0, 1);
var ATObject = function (scene) {
    this.vis = new AnchoredTargetVisualizer(groundPlaneNormal);
    this.vis.installRendererIntoScene(scene);
    if (Bakery.getBoolean("Draw Text", false, "DataVis")) {
        this.textVis = new GLTextOverlayPool(scene);
    }
    else {
        this.textVis = null;
    }
    this.scene = scene;
    this.text = null;
    this.position = null;
};
ATObject.prototype.setPosition = function (position) {
    this.position = position;
    this.vis.setPosition(position);
    this.lastUpdatedTime = Clock.currentTime();
    return this;
};
ATObject.prototype.showText = function (text) {
    this.text = text;
};
ATObject.prototype.release = function () {
    this.vis.removeRendererFromScene(this.scene);
    if (this.textVis !== null) {
        this.textVis.removeFromScene(this.scene);
    }
};
ATObject.prototype.update = function (cullTime) {
    var age = Math.min(1, Math.max(0, Clock.currentTime().subtract(this.lastUpdatedTime) / cullTime));
    var recency = Math.pow(1 - age, 2);
    this.vis.setBrightness(recency);
    this.vis.setLineWidth(Math.max(0.5, recency * 10));
    if (this.text != null && this.textVis !== null) {
        this.textVis.returnAllLeased();
        this.textVis.lease3D(this.text, this.position.x, this.position.y, this.position.z);
    }
};
var ATObjectStore = function () {
    this.atObjects = [];
    this.cullTime = 5;
};
ATObjectStore.prototype.noteObject = function (atObject) {
    this.atObjects.push(atObject);
};
ATObjectStore.prototype.update = function () {
    this.cullOld();
    for (var i = 0; i < this.atObjects.length; i++) {
        this.atObjects[i].update(this.cullTime);
    }
};
ATObjectStore.prototype.cullOld = function () {
    var curTime = Clock.currentTime();
    var i = 0;
    while (i < this.atObjects.length) {
        if (curTime.subtract(this.atObjects[i].lastUpdatedTime) > this.cullTime) {
            this.atObjects[i].release();
            this.atObjects.splice(i, 1);
        }
        else {
            i++;
        }
    }
};
var SensoryVisualizer = function (scene) {
    this.scene = scene;
    this.objectStore = new ATObjectStore();
    this.globalSpeechPrinter = new GLTextOverlayPool(this.scene);
    this.trackedDataPrinter = new GLTextOverlayPool(this.scene);
    this.trackedDataLines = new GLLinePool(this.scene, 100);
    this.trackedDataLines.setLineWidth(3);
    this.trackedAnchors = [];
};
SensoryVisualizer.prototype.showRaw = function (sensoryRecord, passedFilter) {
    var obj;
    var drawRawVision = Bakery.getBoolean("Show Raw Vision", false, "DataVis");
    var drawRawMotion = Bakery.getBoolean("Show Raw Motion", false, "DataVis");
    var drawRawAudio = Bakery.getBoolean("Show Raw Audio", false, "DataVis");
    var drawRawWake = Bakery.getBoolean("Show Raw Wake", true, "DataVis");
    var drawOnlyFiltered = Bakery.getBoolean("Draw only Filtered", true, "DataVis");
    if (passedFilter || !drawOnlyFiltered) {
        if (sensoryRecord.type === SensoryRecord.SensoryType.AUDIO_LOCALIZATION && drawRawAudio) {
            obj = new ATObject(this.scene).setPosition(new THREE.Vector3(sensoryRecord.position.x, sensoryRecord.position.y, sensoryRecord.position.z));
            obj.vis.setBaseColor(new THREE.Color(1, 0, 0.2));
            obj.showText("AUDIO[" + sensoryRecord.id + "]<br>Confidence:" + sensoryRecord.confidence);
            this.objectStore.noteObject(obj);
        }
        else if (sensoryRecord.type === SensoryRecord.SensoryType.VISION && drawRawVision) {
            obj = new ATObject(this.scene).setPosition(new THREE.Vector3(sensoryRecord.position.x, sensoryRecord.position.y, sensoryRecord.position.z));
            obj.vis.setBaseColor(new THREE.Color(0, 1, 0.2));
            obj.showText("VIDEO[" + sensoryRecord.id + "]<br>Confidence:" + sensoryRecord.confidence);
            this.objectStore.noteObject(obj);
        }
        else if (sensoryRecord.type === SensoryRecord.SensoryType.MOTION && drawRawMotion) {
            obj = new ATObject(this.scene).setPosition(new THREE.Vector3(sensoryRecord.position.x, sensoryRecord.position.y, sensoryRecord.position.z));
            obj.vis.setBaseColor(new THREE.Color(0, 0.2, 1));
            obj.showText("MOTION[" + sensoryRecord.id + "]<br>Confidence:" + sensoryRecord.confidence);
            this.objectStore.noteObject(obj);
        }
        else if (sensoryRecord.type === SensoryRecord.SensoryType.WAKE && drawRawWake) {
            var l1 = this.globalSpeechPrinter.lease2D("HEY JIBO", 20, 20, "#ff0000", 50);
            var self = this;
            setTimeout(function () {
                self.globalSpeechPrinter.returnLeased(l1);
            }, 3500);
        }
    }
};
SensoryVisualizer.prototype.rawDataAdded = SensoryVisualizer.prototype.showRaw;
SensoryVisualizer.prototype.drawTracked = function (trackedSensory) {
    while (trackedSensory.length > this.trackedAnchors.length) {
        var additional = new AnchoredTargetVisualizer(groundPlaneNormal);
        additional.installRendererIntoScene(this.scene);
        additional.setLineWidth(4);
        this.trackedAnchors.push(additional);
    }
    this.trackedDataPrinter.returnAllLeased();
    this.trackedDataLines.returnAllLeased();
    var used = 0;
    var drawText = Bakery.getBoolean("Draw Text", false, "DataVis");
    if (Bakery.getBoolean("Show Tracked Data", true, "DataVis")) {
        while (used < trackedSensory.length) {
            this.trackedAnchors[used].setPosition(trackedSensory[used].position);
            if (drawText) {
                this.trackedDataPrinter.lease3D("WIGGLY", trackedSensory[used].fastPosition.x, trackedSensory[used].fastPosition.y, trackedSensory[used].fastPosition.z);
            }
            if (trackedSensory[used].type === SensoryRecord.SensoryType.AUDIO_LOCALIZATION) {
                this.trackedAnchors[used].setBaseColor(new THREE.Color(0.7, 0.7, 0.7));
                if (drawText) {
                    this.trackedDataPrinter.lease3D("<br>Weight:" + trackedSensory[used].weight.toFixed(2) + "<br>FastWeight:" + trackedSensory[used].fastWeight.toFixed(2), trackedSensory[used].fastPosition.x, trackedSensory[used].fastPosition.y, trackedSensory[used].fastPosition.z);
                }
                this.trackedDataLines.leaseLine(trackedSensory[used].position, new THREE.Vector3().copy(trackedSensory[used].position).add(new THREE.Vector3(0, 0, trackedSensory[used].weight)), new THREE.Color(0, 1, 0));
                var w = 0;
                for (var j = 0; j < trackedSensory[used].weightHistory.length; j++) {
                    if (trackedSensory[used].weightHistory[j] > w) {
                        w = trackedSensory[used].weightHistory[j];
                    }
                }
                this.trackedDataLines.leaseLine(new THREE.Vector3().copy(trackedSensory[used].position).add(new THREE.Vector3(0.01, 0.01, 0)), new THREE.Vector3().copy(trackedSensory[used].position).add(new THREE.Vector3(0, 0, w)).add(new THREE.Vector3(0.01, 0.01, 0)), new THREE.Color(0, 0, 1));
                if (trackedSensory[used].hasOwnProperty("wakeAt")) {
                    this.trackedDataPrinter.lease3D("<br><BR>HEYY!:", trackedSensory[used].fastPosition.x, trackedSensory[used].fastPosition.y, trackedSensory[used].fastPosition.z, "#ff0000");
                }
            }
            else if (trackedSensory[used].type === SensoryRecord.SensoryType.VISION) {
                this.trackedAnchors[used].setBaseColor(new THREE.Color(0.7, 0.7, 0));
            }
            else if (trackedSensory[used].type === SensoryRecord.SensoryType.MOTION) {
                this.trackedAnchors[used].setBaseColor(new THREE.Color(0.7, 0, 0.7));
            }
            used++;
        }
    }
    while (used < this.trackedAnchors.length) {
        this.trackedAnchors[used].setPosition(null);
        used++;
    }
};
SensoryVisualizer.prototype.notifyCurrentTracked = SensoryVisualizer.prototype.drawTracked;
SensoryVisualizer.prototype.update = function () {
    this.objectStore.cullTime = Bakery.getFloat("Visualizer Raw Data Display Time", 0, 20, this.objectStore.cullTime, "DataVis");
    this.objectStore.update();
};
module.exports = SensoryVisualizer;

},{"../SensoryRecord":11,"animation-utilities":undefined}],43:[function(require,module,exports){
"use strict";
var slog = require("animation-utilities").slog;
var channel = "ATTENTION";
var SocketClient = function (address) {
    this.connected = false;
    this.handlers = {};
    var self = this;
    var messageHandler = function (event) {
        self.connected = true;
        var packet = JSON.parse(event.data);
        if (packet.hasOwnProperty("type")) {
            self.fireHandlers(packet.type, packet.data);
        }
    };
    var closeHandler = function (closeEvent) {
        slog(channel, "Socket Closed (code:" + closeEvent.code + ", reason:" + closeEvent.reason + ", " + (closeEvent.clean ? "cleanly" : "uncleanly") + ")");
        setTimeout(createSocket, 3000);
    };
    var errorHandler = function (errorEvent) {
    };
    var createSocket = function () {
        if (self.attentionSocket) {
            self.attentionSocket.onmessage = null;
            self.attentionSocket.onclose = null;
            self.attentionSocket.onerror = null;
        }
        self.attentionSocket = new WebSocket(address);
        self.attentionSocket.onmessage = messageHandler;
        self.attentionSocket.onclose = closeHandler;
        self.attentionSocket.onerror = errorHandler;
    };
    createSocket();
};
SocketClient.prototype.sendData = function (type, data) {
    var packet = { type: type, data: data };
    this.attentionSocket.send(JSON.stringify(packet));
};
SocketClient.prototype.fireHandlers = function (type, data) {
    if (this.handlers[type] != null) {
        for (var i = 0; i < this.handlers[type].length; i++) {
            this.handlers[type][i](type, data);
        }
    }
};
SocketClient.prototype.registerHandler = function (type, handler) {
    if (this.handlers[type] == null) {
        this.handlers[type] = [];
    }
    this.handlers[type].push(handler);
};
SocketClient.prototype.removeHandler = function (type, handler) {
    if (this.handlers[type] != null) {
        var index = this.handlers[type].indexOf(handler);
        if (index >= 0) {
            this.handlers[type].splice(index, 1);
        }
        else {
            slog(channel, "Error, asked to remove SC handler (" + type + "), but it was not registered");
        }
        if (this.handlers[type].length === 0) {
            delete (this.handlers[type]);
        }
    }
    else {
        slog(channel, "Error, asked to remove SC handler (" + type + "), but it was not registered");
    }
};
SocketClient.prototype.hasConnected = function () {
    return this.connected;
};
module.exports = SocketClient;

},{"animation-utilities":undefined}],44:[function(require,module,exports){
"use strict";
var findRoot = require("find-root");
if (global._attentionutilities_singleton) {
    module.exports = global._attentionutilities_singleton;
}
else {
    module.exports.AttentionManager = require("./AttentionManager");
    module.exports.sensory = {
        DataConverter: require("./DataConverter"),
        SensoryRecord: require("./SensoryRecord")
    };
    module.exports.remote = {
        RemoteAttentionManager: require("./remote/RemoteAttentionManager"),
        SocketClient: require("./remote/SocketClient"),
        SensoryVisualizer: require("./remote/SensoryVisualizer"),
        RemoteVisualizerConnection: require("./remote/RemoteVisualizerConnection")
    };
    var configRoot = findRoot(__dirname) + "/config/";
    module.exports.config = {
        "mostly-face": require(configRoot + "mostly-face"),
        "mostly-motion": require(configRoot + "mostly-motion")
    };
    global._attentionutilities_singleton = module.exports;
}

},{"./AttentionManager":5,"./DataConverter":7,"./SensoryRecord":11,"./remote/RemoteAttentionManager":39,"./remote/RemoteVisualizerConnection":41,"./remote/SensoryVisualizer":42,"./remote/SocketClient":43,"find-root":undefined}]},{},[44])(44)
});
