// jibo-loader/lib/jibo-loader.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboLoader = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var CacheItem = (function () {
    function CacheItem(request, content) {
        this.request = request;
        this.content = content;
    }
    CacheItem.prototype.destroy = function () {
        if (this.content.destroy) {
            this.content.destroy();
        }
        if (this.content.tagName === 'IMG') {
            this.content.src = '';
        }
        this.content = null;
        this.request = null;
    };
    return CacheItem;
}());
exports.CacheItem = CacheItem;
var AssetCache = (function () {
    function AssetCache() {
        this._cache = null;
        this._cachesMap = {};
        this.addCache(AssetCache.DEFAULT_CACHE);
    }
    AssetCache.prototype.addCache = function (id) {
        if (this._cachesMap[id]) {
            throw new Error("AssetCache: cache alread exists : '" + id + "'");
        }
        this._cachesMap[id] = {};
        this.activeCache = id;
    };
    AssetCache.prototype.getCaches = function () {
        return this._cachesMap;
    };
    Object.defineProperty(AssetCache.prototype, "activeCache", {
        get: function () {
            return this._cacheId;
        },
        set: function (id) {
            var cache = this._cachesMap[id];
            if (!cache) {
                throw new Error("AssetCache: unable to find cache: '" + id + "'");
            }
            this._cache = cache;
            this._cacheId = id;
        },
        enumerable: true,
        configurable: true
    });
    AssetCache.prototype.deleteCache = function (id, select) {
        if (select === void 0) { select = AssetCache.DEFAULT_CACHE; }
        var cache = this._cachesMap[id];
        if (!cache) {
            console.warn("AssetCache: unable to remove cache: '" + id + "'");
        }
        else {
            var active = this._cacheId;
            this.empty(id);
            delete this._cachesMap[id];
            if (select && id === active) {
                this.activeCache = select;
            }
        }
    };
    AssetCache.prototype.read = function (id, cacheId) {
        var cache = this.getCache(cacheId);
        if (!cache) {
            throw new Error("AssetCache: unable to find cache: '" + cacheId + "'");
        }
        return !!cache[id] ? cache[id].content : null;
    };
    AssetCache.prototype.write = function (id, request, content, cacheId) {
        var cache = this.getCache(cacheId);
        var cacheItem = cache[id];
        if (cacheItem) {
            if (cacheItem.request === request) {
                return;
            }
            else {
                throw new Error("AssetCache: cannot load an asset with a different resource path but same id as an asset already in cache."
                    + (" Attempted asset using id: '" + id + "' request: '" + request + "' cache: '" + cacheId + "'"));
            }
        }
        cache[id] = new CacheItem(request, content);
    };
    AssetCache.prototype.getCache = function (cacheId) {
        return cacheId ? this._cachesMap[cacheId] : this._cache;
    };
    AssetCache.prototype.delete = function (asset, cacheId) {
        var cache = this.getCache(cacheId);
        var id = typeof asset === 'string' ? asset : asset.id;
        if (!id) {
            return;
        }
        var item = cache[id];
        if (item) {
            item.destroy();
            delete cache[id];
        }
    };
    AssetCache.prototype.empty = function (cacheId) {
        var cache = this.getCache(cacheId);
        for (var id in cache) {
            this.delete(id);
        }
    };
    AssetCache.prototype.destroy = function () {
        for (var id in this._cachesMap) {
            this.deleteCache(id, null);
        }
        this._cache = null;
    };
    return AssetCache;
}());
AssetCache.DEFAULT_CACHE = 'default';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AssetCache;

},{}],2:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var events_1 = require("events");
var Task_1 = require("./tasks/Task");
var AssetUtils_1 = require("./AssetUtils");
var AssetLoad = (function (_super) {
    __extends(AssetLoad, _super);
    function AssetLoad(manager) {
        var _this = _super.call(this) || this;
        _this.manager = manager;
        _this.id = AssetLoad.ID++;
        _this.mode = MAP_MODE;
        _this.maxLoads = 0;
        _this.cacheAll = false;
        _this.remoteAll = false;
        _this.timeoutAll = 0;
        _this.cache = [];
        _this.tasks = [];
        _this.results = null;
        _this.running = false;
        _this.numLoaded = 0;
        _this.numLoading = 0;
        _this.total = 0;
        return _this;
    }
    AssetLoad.prototype.toString = function () {
        return "[AssetLoad (index: " + this.id + ")]";
    };
    AssetLoad.prototype.setup = function (assets, options) {
        this.maxLoads = options.maxLoads;
        this.cacheAll = options.cacheAll;
        this.remoteAll = options.remoteAll;
        this.timeoutAll = options.timeoutAll;
        this.mode = this.addTasks(assets);
        this.results = this.getAssetsContainer(this.mode);
        if (options.autoStart) {
            this.start();
        }
    };
    AssetLoad.prototype.cancel = function () {
        this.manager.cancel(this);
    };
    AssetLoad.prototype.start = function () {
        this.emit('progress', 0);
        this.running = true;
        this.nextTask();
    };
    AssetLoad.prototype.reset = function () {
        this.removeAllListeners('complete');
        this.removeAllListeners('progress');
        this.removeAllListeners('taskDone');
        this.tasks.forEach(function (task) {
            task.status = Task_1.default.FINISHED;
            task.destroy();
        });
        this.total = 0;
        this.numLoaded = 0;
        this.numLoading = 0;
        this.mode = MAP_MODE;
        this.tasks.length = 0;
        this.results = null;
        this.maxLoads = 0;
        this.cacheAll = false;
        this.remoteAll = false;
        this.timeoutAll = 0;
        this.running = false;
        this.cache = [];
    };
    AssetLoad.prototype.addTasks = function (assets) {
        var asset;
        var mode = MAP_MODE;
        assets = this.applyDefaults(assets);
        var isSingle = this.getTaskByAsset(assets);
        if (isSingle) {
            this.addTask(assets);
            return SINGLE_MODE;
        }
        else {
            var task = void 0;
            if (Array.isArray(assets)) {
                for (var i = 0; i < assets.length; i++) {
                    asset = this.applyDefaults(assets[i]);
                    task = this.addTask(asset);
                    if (!task.id) {
                        mode = LIST_MODE;
                    }
                }
            }
            else if (AssetUtils_1.default.isPlain(assets)) {
                for (var id in assets) {
                    asset = this.applyDefaults(assets[id]);
                    asset.id = id;
                    task = this.addTask(asset);
                }
            }
            else {
                console.error("Asset type unsupported", asset);
            }
        }
        return mode;
    };
    AssetLoad.prototype.applyDefaults = function (asset) {
        if (isString(asset)) {
            return {
                src: asset
            };
        }
        else if (isFunction(asset)) {
            return {
                async: asset
            };
        }
        return asset;
    };
    AssetLoad.prototype.addTask = function (asset) {
        var TaskClass = this.getTaskByAsset(asset);
        var task;
        if (TaskClass) {
            if (asset.cache === undefined && this.cacheAll) {
                asset.cache = this.cacheAll;
            }
            if (asset.remote === undefined && this.remoteAll) {
                asset.remote = true;
            }
            if (asset.timeout === undefined && this.timeoutAll) {
                asset.timeout = this.timeoutAll;
            }
            task = new TaskClass(this.manager, asset);
            this.tasks.push(task);
            ++this.total;
        }
        else {
            console.error("Unable to find a task definition for asset", asset);
        }
        return task;
    };
    AssetLoad.prototype.getTaskByAsset = function (asset) {
        var TaskClass;
        var taskDefs = this.manager.taskDefs;
        for (var i = 0, len = taskDefs.length; i < len; i++) {
            TaskClass = taskDefs[i];
            if (TaskClass.test(asset)) {
                return TaskClass;
            }
        }
        return null;
    };
    AssetLoad.prototype.nextTask = function () {
        var task;
        for (var i = 0; i < this.tasks.length; i++) {
            if (this.tasks[i].status === Task_1.default.WAITING) {
                task = this.tasks[i];
                break;
            }
        }
        if (!task) {
            return;
        }
        task.status = Task_1.default.RUNNING;
        this.numLoading++;
        var cacheId;
        if (typeof task.cache === "string") {
            cacheId = task.cache;
        }
        var result = this.manager.cache.read(task.id, cacheId);
        if (result) {
            this.taskDone(task, null, result);
        }
        else {
            task.start(this.taskDone.bind(this, task));
        }
        if (this.maxLoads === 0 || this.numLoading < this.maxLoads) {
            this.nextTask();
        }
    };
    AssetLoad.prototype.taskDone = function (task, err, result) {
        var _this = this;
        if (result === void 0) { result = null; }
        if (!this.running) {
            return;
        }
        var index = this.tasks.indexOf(task);
        if (index === -1) {
            return;
        }
        this.tasks.splice(index, 1);
        var assets = [];
        if (result) {
            switch (this.mode) {
                case SINGLE_MODE:
                    this.results = result;
                    break;
                case LIST_MODE:
                    this.results.push(result);
                    break;
                case MAP_MODE:
                    this.results[task.id] = result;
                    break;
            }
            if (task.cache) {
                this.cache.push({
                    id: task.id,
                    request: task.src || '',
                    result: result,
                    cache: task.cache
                });
            }
        }
        if (task.complete) {
            task.complete(err, result, task.original, assets);
        }
        if (!err) {
            this.emit('taskDone', result, task.original, assets);
        }
        task.destroy();
        var mode = this.addTasks(assets);
        this.emit('progress', ++this.numLoaded / this.total);
        --this.numLoading;
        if (this.mode === MAP_MODE && mode !== this.mode) {
            console.error("Load assets require IDs to return mapped results", assets);
            return;
        }
        if (err) {
            this.emit('complete', err, null);
            return;
        }
        if (this.tasks.length) {
            this.nextTask();
        }
        else {
            this.cache.forEach(function (item) {
                var cacheId;
                if (typeof item.cache === "string") {
                    cacheId = item.cache;
                }
                _this.manager.cache.write(item.id, item.request, item.result, cacheId);
            });
            this.cache.length = 0;
            this.emit('complete', err, this.results);
        }
    };
    AssetLoad.prototype.getAssetsContainer = function (mode) {
        switch (mode) {
            case SINGLE_MODE:
                return null;
            case LIST_MODE:
                return [];
            case MAP_MODE:
                return {};
        }
    };
    AssetLoad.prototype.destroy = function () {
        this.removeAllListeners();
        this.reset();
        this.tasks = null;
        this.manager = null;
    };
    return AssetLoad;
}(events_1.EventEmitter));
var SINGLE_MODE = 0;
var MAP_MODE = 1;
var LIST_MODE = 2;
function isString(obj) {
    return typeof obj === "string";
}
function isFunction(obj) {
    return typeof obj === "function";
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AssetLoad;

},{"./AssetUtils":4,"./tasks/Task":20,"events":undefined}],3:[function(require,module,exports){
"use strict";
var AssetCache_1 = require("./AssetCache");
var Task_1 = require("./tasks/Task");
var AssetLoad_1 = require("./AssetLoad");
var RemoteLoader_1 = require("./loaders/RemoteLoader");
var LocalLoader_1 = require("./loaders/LocalLoader");
var jibo_plugins_1 = require("jibo-plugins");
var AssetManager = (function () {
    function AssetManager() {
        this.loads = [];
        this.loadPool = [];
        this.taskDefs = [];
        this.cache = new AssetCache_1.default();
        this.remoteLoader = new RemoteLoader_1.default();
        this.localLoader = new LocalLoader_1.default();
        this.maxDefaultLoads = 4;
    }
    AssetManager.prototype.register = function (TaskClass, priority) {
        if (priority === void 0) { priority = 0; }
        TaskClass.priority = priority;
        if (!(TaskClass.prototype instanceof Task_1.default)) {
            console.error("Registering task much extend Task", TaskClass);
        }
        else if (!TaskClass.test) {
            console.error("Registering task much have test method");
        }
        this.taskDefs.push(TaskClass);
        this.taskDefs.sort(function (a, b) {
            return b.priority - a.priority;
        });
    };
    AssetManager.prototype.load = function (assets, options) {
        options = Object.assign({
            complete: null,
            progress: null,
            taskDone: null,
            cacheAll: false,
            maxLoads: this.maxDefaultLoads,
            remoteAll: false,
            timeoutAll: 0,
            autoStart: true
        }, options);
        var startAll = options.startAll;
        if (startAll !== undefined) {
            options.maxLoads = startAll ? 0 : 1;
        }
        var load = this._getLoad();
        this.loads.push(load);
        options.complete = this._onLoaded.bind(this, options.complete, load);
        load.once('complete', options.complete);
        if (options.taskDone) {
            load.on('taskDone', options.taskDone);
        }
        load.setup(assets, options);
        return load;
    };
    AssetManager.prototype.simpleLoad = function (uri, callback, options) {
        if (options === void 0) { options = {}; }
        var currentLoader;
        if (options.remote || jibo_plugins_1.PathUtils.isURL(uri)) {
            currentLoader = this.remoteLoader;
        }
        else {
            currentLoader = this.localLoader;
        }
        currentLoader.load(uri, callback, options);
    };
    AssetManager.prototype.prepare = function (uri, remote) {
        if (remote === void 0) { remote = false; }
        var currentLoader;
        if (remote || jibo_plugins_1.PathUtils.isURL(uri)) {
            currentLoader = this.remoteLoader;
        }
        else {
            currentLoader = this.localLoader;
        }
        return currentLoader.prepare(uri);
    };
    AssetManager.prototype.cancelAll = function () {
        for (var i = this.loads.length - 1; i >= 0; i--) {
            this.cancel(this.loads[i]);
        }
    };
    AssetManager.prototype.destroy = function () {
        this.remoteLoader = null;
        this.localLoader = null;
        this.cache.destroy();
        this.cache = null;
        this.loadPool = null;
        this.loads = null;
        this.taskDefs = null;
    };
    AssetManager.prototype.cancel = function (load) {
        var i = this.loads.indexOf(load);
        if (i >= 0) {
            this.loads.splice(i, 1);
            load.reset();
            this.loadPool.push(load);
        }
    };
    AssetManager.prototype._getLoad = function () {
        if (this.loadPool.length > 0) {
            return this.loadPool.pop();
        }
        return new AssetLoad_1.default(this);
    };
    AssetManager.prototype._onLoaded = function (complete, load, err, results) {
        var index = this.loads.indexOf(load);
        if (index > -1) {
            this.loads.splice(index, 1);
        }
        if (complete) {
            complete(err, results);
        }
        this.cancel(load);
    };
    return AssetManager;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AssetManager;

},{"./AssetCache":1,"./AssetLoad":2,"./loaders/LocalLoader":9,"./loaders/RemoteLoader":10,"./tasks/Task":20,"jibo-plugins":undefined}],4:[function(require,module,exports){
"use strict";
var AssetUtils = (function () {
    function AssetUtils() {
    }
    AssetUtils.isPlain = function (obj) {
        if (typeof obj === 'object' && obj !== null) {
            var proto = Object.getPrototypeOf(obj);
            return proto === Object.prototype || proto === null;
        }
        return false;
    };
    return AssetUtils;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AssetUtils;

},{}],5:[function(require,module,exports){
"use strict";
var AssetManager_1 = require("./AssetManager");
var AssetUtils_1 = require("./AssetUtils");
var Task_1 = require("./tasks/Task");
var LoadTask_1 = require("./tasks/LoadTask");
var ListTask_1 = require("./tasks/ListTask");
var FunctionTask_1 = require("./tasks/FunctionTask");
var LoaderError_1 = require("./errors/LoaderError");
var AssetCache_1 = require("./AssetCache");
var LoaderPlugin = (function () {
    function LoaderPlugin() {
        this.assetManager = new AssetManager_1.default();
        this.AssetUtils = AssetUtils_1.default;
        this.Task = Task_1.default;
        this.LoadTask = LoadTask_1.default;
        this.ListTask = ListTask_1.default;
        this.FunctionTask = FunctionTask_1.default;
        this.LoaderError = LoaderError_1.default;
        this.register(LoadTask_1.default, 0)
            .register(ListTask_1.default, 5)
            .register(FunctionTask_1.default, 10);
    }
    LoaderPlugin.prototype.init = function (done) {
        done();
    };
    LoaderPlugin.prototype.register = function (TaskDefinition, priority) {
        if (priority === void 0) { priority = 0; }
        this.assetManager.register(TaskDefinition, priority);
        return this;
    };
    LoaderPlugin.prototype.load = function (source, complete, progress, cache, data, remote, timeout, format) {
        if (remote === void 0) { remote = false; }
        if (timeout === void 0) { timeout = 0; }
        var options;
        if (typeof source === "string") {
            source = {
                src: source,
                progress: progress || null,
                complete: complete || null,
                cache: !!cache,
                remote: remote,
                timeout: timeout,
                format: format,
                data: data || null
            };
        }
        else {
            options = complete;
            if (typeof complete === "function") {
                options = {
                    complete: complete
                };
            }
        }
        return this.assetManager.load(source, options);
    };
    Object.defineProperty(LoaderPlugin.prototype, "basePath", {
        get: function () {
            return this.assetManager.localLoader.basePath;
        },
        set: function (basePath) {
            this.assetManager.localLoader.basePath = basePath;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LoaderPlugin.prototype, "baseUrl", {
        get: function () {
            return this.assetManager.remoteLoader.baseUrl;
        },
        set: function (baseUrl) {
            this.assetManager.remoteLoader.baseUrl = baseUrl;
        },
        enumerable: true,
        configurable: true
    });
    LoaderPlugin.prototype.unload = function (assets) {
        if (typeof assets === "string") {
            assets = Array.prototype.slice.call(arguments);
        }
        for (var i = 0; i < assets.length; i++) {
            this.assetManager.cache.delete(assets[i]);
        }
        return this;
    };
    LoaderPlugin.prototype.unloadFrom = function (assets, cacheId) {
        if (Array.isArray(assets)) {
            for (var i = 0; i < assets.length; i++) {
                this.assetManager.cache.delete(assets[i], cacheId);
            }
            return this;
        }
        this.assetManager.cache.delete(assets, cacheId);
        return this;
    };
    LoaderPlugin.prototype.unloadAll = function (cacheId) {
        this.assetManager.cache.empty(cacheId);
        return this;
    };
    LoaderPlugin.prototype.cancelAll = function () {
        this.assetManager.cancelAll();
        return this;
    };
    LoaderPlugin.prototype.addCache = function (cacheId) {
        this.assetManager.cache.addCache(cacheId);
        return this;
    };
    Object.defineProperty(LoaderPlugin.prototype, "activeCache", {
        get: function () {
            return this.assetManager.cache.activeCache;
        },
        set: function (cacheId) {
            this.assetManager.cache.activeCache = cacheId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LoaderPlugin.prototype, "maxDefaultLoads", {
        get: function () {
            return this.assetManager.maxDefaultLoads;
        },
        set: function (num) {
            this.assetManager.maxDefaultLoads = num;
        },
        enumerable: true,
        configurable: true
    });
    LoaderPlugin.prototype.defaultCache = function () {
        this.assetManager.cache.activeCache = AssetCache_1.default.DEFAULT_CACHE;
        return this;
    };
    LoaderPlugin.prototype.deleteCache = function (cacheId) {
        this.assetManager.cache.deleteCache(cacheId);
        return this;
    };
    LoaderPlugin.prototype.getCache = function (cacheId) {
        return this.assetManager.cache.getCache(cacheId) || null;
    };
    LoaderPlugin.prototype.getCaches = function () {
        return this.assetManager.cache.getCaches();
    };
    LoaderPlugin.prototype.cached = function (id, cacheId) {
        return this.assetManager.cache.read(id, cacheId);
    };
    return LoaderPlugin;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LoaderPlugin;

},{"./AssetCache":1,"./AssetManager":3,"./AssetUtils":4,"./errors/LoaderError":6,"./tasks/FunctionTask":17,"./tasks/ListTask":18,"./tasks/LoadTask":19,"./tasks/Task":20}],6:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var LoaderError = (function (_super) {
    __extends(LoaderError, _super);
    function LoaderError(originalError, request, message) {
        var _this = _super.call(this, message) || this;
        _this.originalError = originalError;
        _this.request = request;
        Object.setPrototypeOf(_this, LoaderError.prototype);
        _this.name = 'LoaderError';
        _this.message = message || originalError.message || 'load error';
        _this.stack = new Error().stack;
        return _this;
    }
    LoaderError.prototype.toString = function () {
        return this.name + ': ' + this.message;
    };
    return LoaderError;
}(Error));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LoaderError;

},{}],7:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./LoaderError"));
var LoaderError_1 = require("./LoaderError");
exports.LoaderError = LoaderError_1.default;

},{"./LoaderError":6}],8:[function(require,module,exports){
"use strict";
var LoaderError_1 = require("../errors/LoaderError");
var path = require("path");
var FILETYPE;
(function (FILETYPE) {
    FILETYPE.IMAGE = 'image';
    FILETYPE.BINARY = 'binary';
    FILETYPE.JSON = 'json';
    FILETYPE.XML = 'xml';
    FILETYPE.CSS = 'css';
    FILETYPE.JAVASCRIPT = 'js';
    FILETYPE.SVG = 'svg';
    FILETYPE.TEXT = 'text';
    FILETYPE.HTML = 'html';
})(FILETYPE = exports.FILETYPE || (exports.FILETYPE = {}));
var AbstractLoader = (function () {
    function AbstractLoader() {
    }
    AbstractLoader.prototype.prepare = function (url) {
        return url;
    };
    AbstractLoader.prototype.load = function (uri, callback, options) {
        var type = this.getFileType(uri);
        if (options && options.format) {
            type = options.format;
        }
        this.internalLoad(type, uri, callback, options);
    };
    AbstractLoader.prototype.internalLoad = function (type, uri, callback, options) {
    };
    AbstractLoader.prototype.createError = function (err, uri, message) {
        return new LoaderError_1.default(typeof err === 'string' ? new Error(err) : err, uri, message);
    };
    AbstractLoader.prototype.internalDone = function (type, uri, data, callback) {
        switch (type) {
            case FILETYPE.SVG:
                this.parseDOM('image/svg+xml', uri, data, callback);
                break;
            case FILETYPE.XML:
                this.parseDOM('application/xml', uri, data, callback);
                break;
            case FILETYPE.HTML:
                this.parseDOM('text/html', uri, data, callback);
                break;
            case FILETYPE.CSS:
                this.parseStyle(data, callback);
                break;
            case FILETYPE.JSON:
                this.parseJSON(uri, data, callback);
                break;
            default:
                callback(null, data);
                break;
        }
    };
    AbstractLoader.prototype.parseDOM = function (parseType, uri, data, callback) {
        try {
            var parser = new DOMParser();
            var object = parser.parseFromString(data, parseType);
            callback(null, object);
        }
        catch (e) {
            callback(this.createError(e, uri));
        }
    };
    AbstractLoader.prototype.parseStyle = function (data, callback) {
        var style = document.createElement('style');
        style.type = 'text/css';
        style.innerHTML = data;
        callback(null, style);
    };
    AbstractLoader.prototype.parseJSON = function (uri, data, callback) {
        try {
            data = JSON.parse(data);
        }
        catch (e) {
            return callback(this.createError(e, uri, 'Failed to parse JSON'));
        }
        callback(null, data);
    };
    AbstractLoader.prototype.getFileType = function (uri) {
        var extension = path.extname(uri).toLowerCase();
        switch (extension) {
            case '.jpeg':
            case '.jpg':
            case '.gif':
            case '.png':
            case '.webp':
            case '.bmp':
                return FILETYPE.IMAGE;
            case '.json':
            case '.keys':
            case '.bt':
                return FILETYPE.JSON;
            case '.xml':
                return FILETYPE.XML;
            case '.css':
                return FILETYPE.CSS;
            case '.js':
                return FILETYPE.JAVASCRIPT;
            case '.svg':
                return FILETYPE.SVG;
            case '.xml':
                return FILETYPE.XML;
            case '.html':
            case '.htm':
            case '.xhtml':
                return FILETYPE.HTML;
            default:
                return FILETYPE.TEXT;
        }
    };
    return AbstractLoader;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AbstractLoader;

},{"../errors/LoaderError":6,"path":undefined}],9:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var jibo_plugins_1 = require("jibo-plugins");
var AbstractLoader_1 = require("./AbstractLoader");
var path = require("path");
var fs = require("fs");
var MIMETYPE;
(function (MIMETYPE) {
    MIMETYPE.jpg = 'image/jpeg';
    MIMETYPE.jpeg = 'image/jpeg';
    MIMETYPE.gif = 'image/gif';
    MIMETYPE.png = 'image/png';
    MIMETYPE.bmp = 'image/bmp';
    MIMETYPE.webp = 'image/webp';
})(MIMETYPE || (MIMETYPE = {}));
var LocalLoader = (function (_super) {
    __extends(LocalLoader, _super);
    function LocalLoader() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.basePath = '';
        return _this;
    }
    LocalLoader.prototype.prepare = function (uri) {
        if (!path.isAbsolute(uri) && !/^(https?|file)\:\/\//.test(uri)) {
            uri = jibo_plugins_1.PathUtils.getAssetUri(uri, undefined, this.basePath);
            uri = path.resolve(uri);
        }
        return uri;
    };
    LocalLoader.prototype.internalLoad = function (type, uri, callback, options) {
        var _this = this;
        if (!path.isAbsolute(uri)) {
            uri = jibo_plugins_1.PathUtils.getAssetUri(uri, undefined, this.basePath);
            uri = path.resolve(uri);
        }
        fs.readFile(uri, this.getEncodingType(type), function (err, data) {
            if (err) {
                return callback(_this.createError(err, uri, "Unable to read " + type + " file at " + uri));
            }
            switch (type) {
                case AbstractLoader_1.FILETYPE.IMAGE:
                    var mimeType = _this.getImageMimeType(uri);
                    var result = new Image();
                    result.src = "data:" + mimeType + ";base64," + data;
                    callback(null, result);
                    break;
                case AbstractLoader_1.FILETYPE.JAVASCRIPT:
                    try {
                        var result_1 = require(uri);
                        callback(null, result_1);
                    }
                    catch (e) {
                        callback(_this.createError(e, uri, 'Unable to read JavaScript file'));
                    }
                    break;
                default:
                    _this.internalDone(type, uri, data, callback);
                    break;
            }
        });
    };
    LocalLoader.prototype.getEncodingType = function (type) {
        switch (type) {
            case AbstractLoader_1.FILETYPE.IMAGE:
                return 'base64';
            case AbstractLoader_1.FILETYPE.JSON:
            case AbstractLoader_1.FILETYPE.CSS:
            case AbstractLoader_1.FILETYPE.XML:
            case AbstractLoader_1.FILETYPE.SVG:
            case AbstractLoader_1.FILETYPE.HTML:
            case AbstractLoader_1.FILETYPE.TEXT:
            default:
                return 'utf8';
        }
    };
    LocalLoader.prototype.getImageMimeType = function (uri) {
        var extension = path.extname(uri).toLowerCase();
        if (extension) {
            extension = extension.substr(1);
        }
        return MIMETYPE[extension] || null;
    };
    return LocalLoader;
}(AbstractLoader_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LocalLoader;

},{"./AbstractLoader":8,"fs":undefined,"jibo-plugins":undefined,"path":undefined}],10:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var AbstractLoader_1 = require("./AbstractLoader");
var ImageResource_1 = require("./remote/ImageResource");
var TextResource_1 = require("./remote/TextResource");
var BinaryResource_1 = require("./remote/BinaryResource");
var RemoteLoader = (function (_super) {
    __extends(RemoteLoader, _super);
    function RemoteLoader() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.baseUrl = '';
        return _this;
    }
    RemoteLoader.prototype.prepare = function (url) {
        if (this.baseUrl.length
            && this.baseUrl.lastIndexOf('/') !== this.baseUrl.length - 1
            && url.charAt(0) !== '/') {
            return this.baseUrl + '/' + url;
        }
        return this.baseUrl + url;
    };
    RemoteLoader.prototype.internalLoad = function (type, url, callback, options) {
        var _this = this;
        options = options || {};
        var resource;
        if (type === AbstractLoader_1.FILETYPE.IMAGE) {
            resource = new ImageResource_1.default(type, url, callback, options);
        }
        else if (type === AbstractLoader_1.FILETYPE.BINARY) {
            resource = new BinaryResource_1.default(type, url, callback, options);
        }
        else {
            resource = new TextResource_1.default(type, url, callback, options);
        }
        resource.onComplete = function (err, resource) {
            if (err) {
                return resource.callback(_this.createError(err, url));
            }
            if (resource.type === AbstractLoader_1.FILETYPE.JAVASCRIPT) {
                var script = document.createElement('script');
                script.innerHTML = String(resource.result);
                script.dataset['url'] = resource.url;
                resource.callback(null, script);
            }
            else if (resource.type === AbstractLoader_1.FILETYPE.IMAGE) {
                resource.callback(null, resource.result);
            }
            else {
                _this.internalDone(resource.type, resource.url, resource.result, resource.callback);
            }
        };
        resource.start();
    };
    return RemoteLoader;
}(AbstractLoader_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RemoteLoader;

},{"./AbstractLoader":8,"./remote/BinaryResource":12,"./remote/ImageResource":13,"./remote/TextResource":15}],11:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./LocalLoader"));
__export(require("./RemoteLoader"));
__export(require("./AbstractLoader"));
__export(require("./remote"));
var LocalLoader_1 = require("./LocalLoader");
exports.LocalLoader = LocalLoader_1.default;
var RemoteLoader_1 = require("./RemoteLoader");
exports.RemoteLoader = RemoteLoader_1.default;
var AbstractLoader_1 = require("./AbstractLoader");
exports.AbstractLoader = AbstractLoader_1.default;

},{"./AbstractLoader":8,"./LocalLoader":9,"./RemoteLoader":10,"./remote":16}],12:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var http = require("http");
var Resource_1 = require("./Resource");
var BinaryResource = (function (_super) {
    __extends(BinaryResource, _super);
    function BinaryResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BinaryResource.prototype.start = function () {
        var _this = this;
        http.get(this.url, function (res) {
            var data = [];
            res.on('data', function (chunk) {
                data.push(chunk);
            }).on('end', function () {
                _this.result = Buffer.concat(data);
                _this.complete();
            }).on('error', function (e) {
                _this.complete(e);
            });
        });
    };
    return BinaryResource;
}(Resource_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BinaryResource;

},{"./Resource":14,"http":undefined}],13:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Resource_1 = require("./Resource");
var ImageResource = (function (_super) {
    __extends(ImageResource, _super);
    function ImageResource(type, url, callback, options) {
        var _this = _super.call(this, type, url, callback, options) || this;
        _this.onError = _this.onError.bind(_this);
        _this.onLoad = _this.onLoad.bind(_this);
        _this.onTimeout = _this.onTimeout.bind(_this);
        var result = _this.result = new Image();
        options = options || {};
        if (options.crossOrigin) {
            result.crossOrigin = options.crossOrigin;
        }
        result.addEventListener('error', _this.onError, false);
        result.addEventListener('load', _this.onLoad, false);
        if (options.timeout) {
            _this.timeoutId = setTimeout(_this.onTimeout, options.timeout);
        }
        return _this;
    }
    ImageResource.prototype.destroy = function () {
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
        if (this.result) {
            this.result.removeEventListener('error', this.onError);
            this.result.removeEventListener('load', this.onLoad);
        }
        this.onError = null;
        this.onLoad = null;
        this.onTimeout = null;
        this.result = null;
        _super.prototype.destroy.call(this);
    };
    ImageResource.prototype.start = function () {
        this.result.src = this.url;
    };
    ImageResource.prototype.onError = function () {
        this.complete('Failed to load element using <IMG>');
    };
    ImageResource.prototype.onLoad = function () {
        this.complete();
    };
    return ImageResource;
}(Resource_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ImageResource;

},{"./Resource":14}],14:[function(require,module,exports){
"use strict";
var Resource = (function () {
    function Resource(type, url, callback, options) {
        this.type = type;
        this.url = url;
        this.callback = callback;
        this.options = options;
        this.options = options || {};
    }
    Resource.prototype.start = function () {
    };
    Resource.prototype.complete = function (err) {
        this.onComplete(err, this);
        this.destroy();
    };
    Resource.prototype.onTimeout = function () {
        this.complete('Request timed out.');
    };
    Resource.prototype.destroy = function () {
        this.onComplete = null;
        this.callback = null;
        this.result = null;
        this.options = null;
        this.url = null;
        this.type = null;
    };
    return Resource;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Resource;

},{}],15:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Resource_1 = require("./Resource");
var STATUS_NONE = 0;
var STATUS_OK = 200;
var STATUS_EMPTY = 204;
var TextResource = (function (_super) {
    __extends(TextResource, _super);
    function TextResource(type, url, callback, options) {
        var _this = _super.call(this, type, url, callback, options) || this;
        _this.onError = _this.onError.bind(_this);
        _this.onAbort = _this.onAbort.bind(_this);
        _this.onLoad = _this.onLoad.bind(_this);
        _this.onTimeout = _this.onTimeout.bind(_this);
        var request = new XMLHttpRequest();
        request.responseType = 'text';
        options = options || {};
        if (options.timeout) {
            request.timeout = options.timeout;
            request.addEventListener('timeout', _this.onTimeout, false);
        }
        request.addEventListener('error', _this.onError, false);
        request.addEventListener('abort', _this.onAbort, false);
        request.addEventListener('load', _this.onLoad, false);
        _this.request = request;
        return _this;
    }
    TextResource.prototype.start = function () {
        this.request.open('GET', this.url, true);
        this.request.send();
    };
    TextResource.prototype.destroy = function () {
        _super.prototype.destroy.call(this);
        if (this.request) {
            this.request.removeEventListener('error', this.onError);
            this.request.removeEventListener('abort', this.onAbort);
            this.request.removeEventListener('load', this.onLoad);
            this.request.removeEventListener('timeout', this.onTimeout);
        }
        this.onError = null;
        this.onAbort = null;
        this.onLoad = null;
        this.onTimeout = null;
        this.request = null;
    };
    TextResource.prototype.onLoad = function () {
        var request = this.request;
        var status = request.status;
        if (status === STATUS_OK || status === STATUS_EMPTY || (status === STATUS_NONE && request.responseText.length > 0)) {
            this.result = request.response || request.responseText;
        }
        else {
            return this.complete('[' + request.status + ']' + request.statusText + ':' + request.responseURL);
        }
        this.complete();
    };
    TextResource.prototype.onError = function () {
        this.complete(this.reqType() + ' Request failed. '
            + 'Status: ' + this.request.status + ', text: "' + this.request.statusText + '"');
    };
    TextResource.prototype.onAbort = function () {
        this.complete(this.reqType() + ' Request was aborted by the user.');
    };
    TextResource.prototype.reqType = function () {
        return this.request.toString().replace('object ', '');
    };
    return TextResource;
}(Resource_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = TextResource;

},{"./Resource":14}],16:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./ImageResource"));
__export(require("./TextResource"));
__export(require("./Resource"));
__export(require("./BinaryResource"));
var ImageResource_1 = require("./ImageResource");
exports.ImageResource = ImageResource_1.default;
var TextResource_1 = require("./TextResource");
exports.TextResource = TextResource_1.default;
var Resource_1 = require("./Resource");
exports.Resource = Resource_1.default;
var BinaryResource_1 = require("./BinaryResource");
exports.BinaryResource = BinaryResource_1.default;

},{"./BinaryResource":12,"./ImageResource":13,"./Resource":14,"./TextResource":15}],17:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Task_1 = require("./Task");
var FunctionTask = (function (_super) {
    __extends(FunctionTask, _super);
    function FunctionTask(manager, asset) {
        var _this = _super.call(this, manager, asset) || this;
        _this.async = asset.async;
        return _this;
    }
    FunctionTask.test = function (asset) {
        return !!asset.async;
    };
    FunctionTask.prototype.start = function (callback) {
        this.async(callback);
    };
    FunctionTask.prototype.destroy = function () {
        _super.prototype.destroy.call(this);
        this.async = null;
    };
    return FunctionTask;
}(Task_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = FunctionTask;

},{"./Task":20}],18:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Task_1 = require("./Task");
var AssetUtils_1 = require("../AssetUtils");
var ListTask = (function (_super) {
    __extends(ListTask, _super);
    function ListTask(manager, asset) {
        var _this = _super.call(this, manager, asset) || this;
        _this.assets = asset.assets;
        _this.cacheAll = asset.cacheAll;
        _this.remoteAll = asset.remoteAll;
        _this.progress = asset.progress;
        return _this;
    }
    ListTask.test = function (asset) {
        return !!asset.assets && (Array.isArray(asset.assets) || AssetUtils_1.default.isPlain(asset.assets));
    };
    ListTask.prototype.start = function (callback) {
        this.load(this.assets, {
            complete: callback,
            progress: this.progress,
            cacheAll: this.cacheAll,
            remoteAll: this.remoteAll
        });
    };
    ListTask.prototype.destroy = function () {
        _super.prototype.destroy.call(this);
        this.assets = null;
    };
    return ListTask;
}(Task_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ListTask;

},{"../AssetUtils":4,"./Task":20}],19:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Task_1 = require("./Task");
var LoadTask = (function (_super) {
    __extends(LoadTask, _super);
    function LoadTask(manager, asset) {
        var _this = _super.call(this, manager, asset, asset.src) || this;
        _this.src = _this.prepare(asset.src);
        return _this;
    }
    LoadTask.test = function (asset) {
        return !!asset.src;
    };
    LoadTask.prototype.start = function (callback) {
        this.simpleLoad(this.src, callback, {
            remote: this.remote,
            timeout: this.timeout,
            format: this.format
        });
    };
    return LoadTask;
}(Task_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LoadTask;

},{"./Task":20}],20:[function(require,module,exports){
"use strict";
var path = require("path");
var Task = (function () {
    function Task(manager, asset, fallbackId) {
        this.manager = manager;
        this.status = Task.WAITING;
        this.complete = asset.complete || null;
        this.cache = asset.cache === undefined ? false : asset.cache;
        this.id = asset.id || null;
        this.remote = asset.remote === undefined ? false : asset.remote;
        this.timeout = asset.timeout || 0;
        this.format = asset.format || null;
        this.original = asset;
        if (this.cache && !this.id) {
            if (fallbackId && typeof fallbackId === "string") {
                asset.id = this.id = path.parse(fallbackId).name;
            }
            if (!this.id) {
                console.error("Caching an asset requires an id, none set", asset);
                this.cache = false;
            }
        }
    }
    Task.prototype.start = function (callback) {
    };
    Task.prototype.load = function (source, options) {
        if (typeof options === "function") {
            options = { complete: options };
        }
        return this.manager.load(source, options);
    };
    Task.prototype.simpleLoad = function (url, complete, options) {
        this.manager.simpleLoad(url, complete, options || {
            remote: this.remote,
            timeout: this.timeout,
            format: this.format
        });
    };
    Task.prototype.prepare = function (uri) {
        return this.manager.prepare(uri, this.remote);
    };
    Task.prototype.destroy = function () {
        this.manager = null;
        this.status = Task.FINISHED;
        this.id = null;
        this.complete = null;
        this.original = null;
    };
    return Task;
}());
Task.WAITING = 0;
Task.RUNNING = 1;
Task.FINISHED = 2;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Task;

},{"path":undefined}],21:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./FunctionTask"));
__export(require("./ListTask"));
__export(require("./LoadTask"));
__export(require("./Task"));
var FunctionTask_1 = require("./FunctionTask");
exports.FunctionTask = FunctionTask_1.default;
var ListTask_1 = require("./ListTask");
exports.ListTask = ListTask_1.default;
var LoadTask_1 = require("./LoadTask");
exports.LoadTask = LoadTask_1.default;
var Task_1 = require("./Task");
exports.Task = Task_1.default;

},{"./FunctionTask":17,"./ListTask":18,"./LoadTask":19,"./Task":20}],22:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./tasks"));
__export(require("./errors"));
__export(require("./loaders"));
__export(require("./AssetCache"));
__export(require("./AssetLoad"));
__export(require("./AssetManager"));
__export(require("./AssetUtils"));
__export(require("./LoaderPlugin"));
var AssetCache_1 = require("./AssetCache");
exports.AssetCache = AssetCache_1.default;
var AssetLoad_1 = require("./AssetLoad");
exports.AssetLoad = AssetLoad_1.default;
var AssetManager_1 = require("./AssetManager");
exports.AssetManager = AssetManager_1.default;
var AssetUtils_1 = require("./AssetUtils");
exports.AssetUtils = AssetUtils_1.default;
var LoaderPlugin_1 = require("./LoaderPlugin");
exports.LoaderPlugin = LoaderPlugin_1.default;

},{"./AssetCache":1,"./AssetLoad":2,"./AssetManager":3,"./AssetUtils":4,"./LoaderPlugin":5,"./errors":7,"./loaders":11,"./tasks":21}]},{},[22])(22)
});
