// @be/recipe/src/views/RecipeSelectionView.ts
import jibo = require('jibo');
import Recipe from '../entities/Recipe';
import FavoriteBtn from '../components/buttons/FavoriteBtn';
import RecipeMenuItem from '../components/buttons/RecipeMenuItem';
import ActionItem from '../entities/ActionItem';
import AssetItem from '../entities/AssetItem';
import ItemData from '../entities/ItemData';
import ViewItemData from '../entities/ViewItemData';
import ListData from '../entities/ListData';
import { stateModel } from '../models/StateModel';

export default class RecipeSelectionView extends jibo.rendering.gui.components.Element {

  public static TYPE: string = 'RecipeSelectionView';

  public index: number;
  public favoriteBtn: FavoriteBtn;

  private view: any;
  private list: jibo.rendering.gui.components.List;
  private activeRecipe: Recipe;
  private recipes: Recipe[];

  constructor () {
    super();
  }

  public init (recipes: Recipe[]): void {
    this.recipes = recipes.slice(); // clone reference
    this.index = 0;
    this.activeRecipe = recipes[0];

    let listData: any = this.getListData();
    let componentConfigsList = this.getComponentConfigsList();
    componentConfigsList.push(listData);

    let viewData = new ViewItemData('RecipeList', 'View', componentConfigsList);

    this.view = jibo.face.views.createView('recipeSelection', viewData, true, view => {
      this.list = this.view.getComponentById('recipeList') as jibo.rendering.gui.components.List;
      this.favoriteBtn = this.view.getComponentById('favorite_btn') as FavoriteBtn;
      this.setEventHandlers();
      this.updateDisplay();
      this.emit('ready');
    });
  }

  public cleanup (done: () => any): void {
    jibo.face.views.changeView({ removeAll: true }, done);
  }

  public moveToIndex (index: number): void {
    if (index < 0 || index === this.recipes.length) { return; }
    let dir = index > this.index; // true = forward, false = backward
    for (let i = 0, l = Math.abs(index - this.index); i < l; i++) {
      this.list.changePage(dir);
    }
  }

  protected getComponentConfigsList (): ItemData[] {
    return [
      new ItemData('favorite_btn', 'FavoriteBtn',
        new AssetItem('icon_favorite', '@be/recipe://timelines/FavoriteButton/FavoriteButton.js', 'timeline'),
        { x: 84, y: 603 }, new ActionItem('event', { event: 'press', choice: 'favorite' })),
      new ItemData('ingredients_btn', 'Button',
        new AssetItem('icon_ingredient', '@be/recipe://resources/buttons/ingredients_btn.png', 'texture'),
        { x: 212, y: 603 }, new ActionItem('event', { event: 'press', choice: 'ingredients' })),
      new ItemData('make_btn', 'Button',
        new AssetItem('icon_makeit', '@be/recipe://resources/buttons/makeIt_btn.png', 'texture'),
        { x: 552, y: 603 }, new ActionItem('event', { event: 'press', choice: 'make' })),
      new ItemData('playvideo_btn', 'Button',
        new AssetItem('icon_playvideo', '@be/recipe://resources/buttons/playVideo_btn.png', 'texture'),
        { x: 893, y: 603 }, new ActionItem('event', { event: 'press', choice: 'playVideo' })),
    ];
  }

  protected getListData (): ListData {
    let list = new ListData('recipeList', 'List');
    let data = list.data;
    data.componentConfigs = this.recipes.map((recipe, i) => {
      return {
        id: `recipe${i}`,
        assets: [{
          id: 'recipe',
          src: '@be/recipe://timelines/RecipeMenuItem/recipeMenuItem.js',
          type: 'timeline',
          index: i,
          data: recipe
        }]
      };
    });
    data.elementBuffer = 40;
    data.position = { x: 0, y: 298 };
    data.elementDimensions = { x: 1112, y: 512 };
    data.defaultElement.type = RecipeMenuItem.TYPE;
    list.data = data;
    return list;
  }

  public updateDisplay (): void {
    this.favoriteBtn.updateDisplay(stateModel.isFavoriteRecipe(this.activeRecipe.recipeID));
  }

  public setEventHandlers () {
    this.view = jibo.face.views.currentView;
    this.view.on('paged', event => {
      let direction = this.index > this.list.pageIndex ? 'previous' : 'next';
      this.index = this.list.pageIndex;
      this.activeRecipe = this.recipes[this.index];
      this.updateDisplay();
      this.emit('nav', direction);
    });

    this.view.on('press', event => {
      switch (event.choice) {
        case 'make':
          this.emit('make');
          break;
        case 'ingredients':
          this.emit('ingredients');
          break;
        case 'favorite':
          this.emit('favorite');
          break;
        case 'playVideo':
          this.emit('video');
          break;
      }
    });
  }

}
