// @be/recipe/src/views/TakePhotoView.ts
import jibo = require('jibo');
import eases = require('eases');
import Recipe from '../entities/Recipe';
import ButtonUtil from '../utils/ButtonUtil';
import { EventEmitter } from 'events';

export default class TakePhotoView extends EventEmitter {

  public static TYPE: string = 'TakePhotoView';

  public componentType: any;
  public mc: any;
  public heartBtn: any;
  public closeBtn: any;
  public imageContainer: PIXI.Container;
  public retakeBtn: any;
  public saveBtn: any;
  public photoImg: any;
  public thumbnails: { [id: string]: [number, number] };
  private recipe: Recipe;

  constructor () {
    super();
  }

  public init (recipe: Recipe): void {
    this.recipe = recipe;

    jibo.face.views.addView('resources/views/TakePhoto.json', (view) => {
      this.componentType = view.getComponentById('takePhotoClip');
      this.mc = this.componentType.movieClip;
      this.closeBtn = this.mc.closeBtn;
      this.retakeBtn = this.mc.retakeBtn;
      this.saveBtn = this.mc.saveBtn;
      this.photoImg = this.mc.photoImg;

      this.imageContainer = new PIXI.Container();
      this.mc.addChildAt(this.imageContainer, 1);

      jibo.face.tween.play(this.closeBtn, { to:{x:1160}, from:{x:1300}, ease:'circInOut',delay:200, duration:350});
      jibo.face.tween.play(this.retakeBtn, { to:{y:600}, from:{y:950}, ease:'circInOut', delay:200, duration:350});
      jibo.face.tween.play(this.saveBtn, { to:{y:600}, from:{y:950}, ease:'circInOut', delay:300, duration:350});

      this.thumbnails = {
        "80x80": [80, 80],
        "300x300": [300, 300]
      };

      this.setupButtons([ this.retakeBtn, this.closeBtn, this.saveBtn ]);

      this.setEventHandlers();
      this.emit('ready');
    }, jibo.face.views.UP, jibo.face.views.OUT);
  }

  public cleanup (done: () => any): void {
    jibo.face.views.removeView(done);
  }

  protected setEventHandlers (): void {
    ButtonUtil.once(this.closeBtn, () => {
      this.emit('close');
    });

    ButtonUtil.once(this.saveBtn, () => {
      this.emit('take');
    });

    ButtonUtil.once(this.retakeBtn, () => {
      this.emit('retake');
    });
  }

  protected showError (error): void {
    let errorMessage = "";
    for(var n in error) {
      errorMessage+= n + " : " + error[n];
    }
    console.log("Error:"+ errorMessage)
    this.emit("imageError", error);
  }

  public retakePhoto (): void {
    this.imageContainer.removeChildren();
  }

  public takePhoto (): void {
    jibo.lps.takePhoto(jibo.lps.PhotoRes.LARGE, true, jibo.lps.CameraID.LEFT, jibo.lps.PhotoType.FULL, (error, imageUrl) => {
      if (error != null) { this.showError(error); return; }
      jibo.loader.load({ src: imageUrl, format: 'binary' }, (error, buffer) => {
        if (error != null) { this.showError(error); return; }
        jibo.media.storePhoto(buffer, this.thumbnails, (error, result) => {
          if (error != null) { this.showError(error); return; }
          jibo.media.getPhoto(result.thumbnails['330x330'], (err, pixiTexture) => {
            if (err != null) { this.showError(err); return; }
            const s = new PIXI.Sprite();
            s.texture = pixiTexture;
            this.imageContainer.addChild(s);
            this.emit("imageCaptured");
          })
        })
      })
    })
  }

  protected setupButtons (buttons: any[]): void {
    for (let i = 0, l = buttons.length; i < l; i++) {
      let btn = buttons[i];
      btn.interactive = btn.buttonMode = true;
    }
  }

}
