// @be/recipe/src/views/VideoView.ts
import jibo = require('jibo');
import eases = require('eases');
import TimeUtil from '../utils/TimeUtil';
import ButtonUtil from '../utils/ButtonUtil';
import Recipe from '../entities/Recipe';
import { EventEmitter } from 'events';

const NORMAL = 'normal';
const STEP = 'step';

export default class VideoView extends EventEmitter {

  public static TYPE: string = 'VideoView';

  public isPlaying: boolean = false;
  public isComplete: boolean = false;

  private recipe: Recipe;
  private timestamps: number[];
  private componentType: any;
  private mc: any;
  private timeline: any;
  private library: any;
  private libRefs: any;
  private videoSource: HTMLVideoElement;
  private videoSprite: HTMLDivElement;
  private playBtn: any;
  private closeBtn: any;
  private volumeBtn: any;
  private bar: any;
  private currentTimeTxt: PIXI.Text;
  private totalTimeTxt: PIXI.Text;
  private view: any;
  private track: any;
  private texture: PIXI.Texture;
  private markerContainer: PIXI.Container;
  private videoruntime: any;

  private duration: number;
  private offset: number;
  private progressBar: any;
  private scales: any;
  private videoSrc: string;

  private face: Element;
  private canvas: any;

  private static MODE: string = NORMAL;
  private static BAR_WIDTH: number = 820;
  private static BAR_Y: number = 620;

  constructor () {
    super();
    this.face = document.querySelector('#face');
    this.canvas = document.querySelector('canvas');
  }

  public init (recipe: Recipe, truncated: boolean, step?: number, videoSrc?: string) {
    VideoView.MODE = NORMAL;
    this.recipe = recipe;
    this.videoSrc = videoSrc || (recipe.video && recipe.video.url);
    this.timestamps = recipe.directions.map(direction => direction.videoTimestamp);
    this.isComplete = this.isPlaying = false;

    this.view = jibo.face.views.addView('resources/views/PlayerUI.json', view => {
      this.componentType = view.getComponentById('playerUIClip');
      this.mc = this.componentType.movieClip;
      this.markerContainer = new PIXI.Container();
      this.mc.addChild(this.markerContainer);
      this.timeline = (<any>this.componentType).timeline;
      this.library = this.timeline.library;
      this.libRefs = this.library.library;

      this.playBtn = this.mc.playBtn;
      this.closeBtn = this.mc.closeBtn;
      this.volumeBtn = this.mc.volumeBtn;
      this.bar = this.mc.bar;
      this.track = this.mc.track;
      this.setupButtons();

      this.videoruntime = this.mc.videoruntime;
      this.currentTimeTxt = this.videoruntime.currentTime_txt;
      this.totalTimeTxt = this.videoruntime.totalTime_txt;
      this.bar.scale.x = 0;

      jibo.face.tween.play(this.playBtn, { to:{x:80}, from:{x:-100}, delay:200, duration:350});
      jibo.face.tween.play(this.videoruntime, { to:{x:1060}, from:{x:1300}, delay:300, duration:350});
      jibo.face.tween.play(this.bar, { to:{y:630}, from:{y:900}, delay:300, duration:350});
      jibo.face.tween.play(this.track, { to:{y:630}, from:{y:900}, delay:300, duration:350});

      let videoReady = this.createDOMVideo();
      if (truncated && !isNaN(step)) {
        videoReady.then(() => {
          this.setupMarkers(step);
          this.stepTo(step);
          this.emit('ready');
        });
      } else {
        videoReady.then(() => {
          this.setupMarkers();
          this.emit('ready');
        });
      }
    }, jibo.face.views.UP, jibo.face.views.OUT);
  }

  public cleanup (done: () => any): void {
    this.closeVideo();
    jibo.face.views.removeView(done);
  }

  public closeVideo (): void {
    this.isPlaying = false;
    this.videoSource.pause();
    this.dispose();
  }

  public playVideo (): void {
    if (this.isComplete) {
      this.setTime(this.offset);
      this.isComplete = false;
    }
    this.isPlaying = true;
    const p: any = this.videoSource.play();
    if (p && typeof p.catch === 'function') {
      p.catch(err => console.error('[recipe] video play() rejected:', err && err.name, err && err.message));
    }
    this.playBtn.gotoAndStop('pause');
  }

  public pauseVideo (): void {
    this.isPlaying = false;
    this.videoSource.pause();
    this.playBtn.gotoAndStop('play');
  }

  public toggleMute (): void {
    if (this.videoSource.volume <= 0) {
      this.volumeBtn.gotoAndStop('volume');
      this.videoSource.volume = 1;
    } else {
      this.volumeBtn.gotoAndStop('volumeOff');
      this.videoSource.volume = 0;
    }
  }

  public stepTo (step: number): void {
    this.setTime(this.timestamps[step - 1]);
  }

  protected showReplay (): void {
    this.playBtn.gotoAndStop('refresh');
    this.isPlaying = false;
    this.isComplete = true;
  }

  protected setTime (timestamp: number): void {
    // check if timestamp is in S or MS format, divide accordingly as currentTime is expecting S
    this.videoSource.currentTime = timestamp / (timestamp >= 1000 ? 1000 : 1);
  }

  protected dispose (): void {
    if (this.face && this.videoSprite) {
      this.face.removeChild(this.videoSprite);
    }
  }

  protected setupButtons (): void {
    this.closeBtn.gotoAndStop('close');
    this.playBtn.gotoAndStop('pause');
    this.volumeBtn.gotoAndStop('volume');

    this.closeBtn.buttonMode = this.closeBtn.interactive = true;
    this.playBtn.buttonMode = this.playBtn.interactive = true;
    this.volumeBtn.buttonMode = this.volumeBtn.interactive = true;
    this.track.interactive = this.track.buttonMode = true;

    ButtonUtil.once(this.closeBtn, () => {
      this.emit('close');
    });

    ButtonUtil.once(this.playBtn, () => {
      this.emit('play-pause');
    });

    ButtonUtil.once(this.volumeBtn, () => {
      this.emit('volume');
    });

    ButtonUtil.once(this.track, event => {
      this.handleMouseMove(event);
      this.track.on('mousemove', this.handleMouseMove.bind(this));
    }, event => {
      this.track.off('mousemove');
    });
  }

  protected handleMouseMove (event): void {
    let pt = event.data.getLocalPosition(this.track);
    let val = pt.x / this.track.width;
    if (VideoView.MODE === NORMAL) {
      this.setTime(this.videoSource.duration * val);
    } else {
      this.setTime(((this.duration / 1000) * val) + this.offset);
    }
  }

  protected createDOMVideo (): Promise<any>  {
    return new Promise((resolve, reject) => {
      this.videoSource = document.createElement('video');
      this.videoSource.id = 'video-element';
      // jibo 14 needs an explicit load; some hosts also block autoplay with audio.
      this.videoSource.setAttribute('playsinline', '');
      this.videoSource.src = this.videoSrc || this.recipe.video.url;
      console.log('[recipe] video src =', this.videoSource.src);

      this.videoSprite = document.createElement('div');
      this.videoSprite.style.opacity = '1';
      this.videoSprite.style.position = 'absolute';
      this.videoSprite.appendChild(this.videoSource);

      this.canvas.style.position = 'absolute';
      this.face.insertBefore(this.videoSprite, this.canvas);

      // this.videoSource.addEventListener('readyState', event => {});
      // this.videoSource.addEventListener('progress', event => {});
      this.videoSource.addEventListener('loadeddata', event => {
        this.playVideo();
        resolve();
      });
      this.videoSource.addEventListener('ended', event => {
        this.showReplay();
      });
      this.videoSource.addEventListener('error', () => {
        const err: any = this.videoSource.error;
        console.error('[recipe] video error', err && err.code, err && err.message, 'src=', this.videoSource.src);
        // Resolve the readiness promise so the flow does not stall waiting on a
        // video that will never load; the control UI stays usable.
        resolve();
      });
      // Some renderers require an explicit load() before src takes effect.
      try { this.videoSource.load(); } catch (e) { /* no-op */ }
      this.videoSource.addEventListener('timeupdate', () => {
        let scale, currentTime, duration, time;

        if (VideoView.MODE === NORMAL) {
          scale = this.videoSource.currentTime / this.videoSource.duration;
          this.updateMarkers(scale);
          currentTime = this.videoSource.currentTime;
          duration = this.videoSource.duration;
        } else {
          scale = (this.videoSource.currentTime - this.offset) / (this.duration / 1000);
          if (scale >= 1) {
            this.pauseVideo();
            this.showReplay();
          }
          currentTime = this.videoSource.currentTime - this.offset;
          duration = this.duration / 1000;
        }

        this.bar.scale.x = scale;
        time = TimeUtil.secondsToTime(currentTime);
        this.currentTimeTxt.text = `${time.m}:${time.s}`;
        time = TimeUtil.secondsToTime(duration);
        this.totalTimeTxt.text = `${time.m}:${time.s}`;
      });
    });
  }

  protected setupMarkers (step?: number): void {
    let Marker = this.libRefs.Marker;
    let directions = this.recipe.directions;
    let len = directions.length;
    let totalTime = this.videoSource.duration * 1000;
    let startTime, endTime, startIndex, endIndex;

    this.markerContainer.removeChildren();

    if (isNaN(step)) {
      VideoView.MODE = NORMAL;
      startTime = startIndex = 0;
      endIndex = len - 1;
      this.duration = endTime = totalTime;
      this.offset = 0;
    } else {
      VideoView.MODE = STEP;
      step = Math.max(1, Math.min(this.timestamps.length, step));
      startIndex = step - 1;
      endIndex = step;
      startTime = this.timestamps[startIndex];
      endTime = this.timestamps[endIndex] || totalTime;
      this.duration = endTime - startTime;
      this.offset = startTime / 1000;
    }

    const baseX = this.bar.x;
    const baseY = VideoView.BAR_Y;
    const width = VideoView.BAR_WIDTH;
    let delay = 250;

    for (let i = startIndex; i <= endIndex; i++) {
      let ordinal = i + 1;
      let direction = directions[i];
      let timestamp = direction ? direction.videoTimestamp : totalTime;
      let marker = new Marker();
      this.markerContainer.addChild(marker);
      // 'F' marks 'finish' of timeline
      let str = String(ordinal);
      if (VideoView.MODE === STEP && i === len) { str = 'F'; }
      marker.marker_txt.text = str;
      marker.interactive = marker.buttonMode = true;
      marker.alpha = 0;
      marker.y = VideoView.BAR_Y;
      marker.timestamp = timestamp;
      let posX = marker.posX = (timestamp / 1000) / (totalTime / 1000);
      if (timestamp === startTime) { marker.x = baseX; }
      else if (timestamp === endTime) { marker.x = baseX + width; }
      else { marker.x = baseX + (width * posX); }

      ButtonUtil.once(marker, event => {
        this.setTime(event.target.timestamp);
      });

      // don't stagger if there's only 2 indicies
      let even = i % 2 === 0;
      let paddingY = baseY + (startIndex + 1 === endIndex ? 20 : (even ? 60 : -20));
      delay += ordinal * (even ? 25 : endIndex / 2);

      jibo.face.tween.play(marker, {
        to: { alpha: 1, y: paddingY },
        from: { alpha: 0, y: paddingY + (even ? 20 : -20) },
        ease: 'circInOut',
        delay
      }, () => {
        if (i === endIndex) {
          jibo.face.tween.play(this.closeBtn, { to: { x: 1120 }, from: { x: 1300 }, ease: 'circInOut', delay: 200, duration: 450 });
          jibo.face.tween.play(this.volumeBtn, { to: { x: 1120 }, from: { x: 1300 }, ease: 'circInOut', delay: 300, duration: 450 });
        }
      });
    }
  }

  protected updateMarkers (scale: number): void {
    this.markerContainer.children.forEach((marker: any) => marker.gotoAndStop(scale >= marker.posX ? 1 : 0));
  }

  /*protected createVideo(): Promise<any> {
    return new Promise((resolve, reject) => {

      this.texture = PIXI.Texture.fromVideo(this.recipe.video.url);
      this.videoSprite =  new PIXI.Sprite(this.texture);

      this.videoSprite.width = 1280;
      this.videoSprite.height = 720;
      this.isPlaying = true;
      jibo.face.stage.addChildAt(this.videoSprite, 0);

      let baseTexture = this.texture.baseTexture;

      this.videoSource = <any>this.texture.baseTexture.source;

      this.videoSource.addEventListener('readyState', event => {
        console.log('readyState', event);
      });
      this.videoSource.addEventListener('progress', event => {
        //console.log('progress', event.target);
        //this.progressBar.scale.x = event.loaded / event.total;
      })
      this.videoSource.addEventListener('loadeddata', event => {
        console.log('loadeddata', event);
        this.videoSource.pause();
        resolve();
      });
      this.videoSource.addEventListener('ended', event => {
        //SD:Do we want it to close or just stop?
        //SD:Turn this into replay
      });
      this.videoSource.addEventListener('error', event => {
        console.log('error', event);
      });
      this.videoSource.addEventListener('timeupdate', event => {

        let scale = (VideoView.MODE === NORMAL) ? this.videoSource.currentTime / this.videoSource.duration : (this.videoSource.currentTime - this.offset) / (this.timeInbetween/1000);
        this.bar.scale.x = scale;
        if (VideoView.MODE === 'normal') { this.updateMarkers(scale); }
        let currentTime = (VideoView.MODE === NORMAL) ? this.videoSource.currentTime : this.videoSource.currentTime - this.offset;
        let time = TimeUtil.secondsToTime(currentTime);
        this.currentTimeTxt.text = `${time.m}:${time.s}`;
        let duration = (VideoView.MODE === NORMAL) ? this.videoSource.duration : this.timeInbetween/1000;
        time = TimeUtil.secondsToTime(duration);
        //this.videoSource.pause();
        //console.log(VideoView.MODE, this.timeInbetween, this.videoSource.currentTime, this.offset);
        this.totalTimeTxt.text = `${time.m}:${time.s}`;
        if (VideoView.MODE === STEP && this.bar.scale.x >= 1) {
          this.pauseVideo();
        }
      });
    });
  }*/

}
