// @be/recipe/src/components/buttons/StepMenuItem.ts
import jibo = require('jibo');

export default class StepMenuItem extends jibo.rendering.gui.components.Button {

  public stepNumber_txt;
  public direction_txt;
  public mc;
  protected data;
  
  static createFromConfig (config: any): StepMenuItem {
    let btn: StepMenuItem = new StepMenuItem();
    btn.assignConfig(config);
    return btn;
  }

  public static TYPE: string = 'StepMenuItem';

  constructor () {
    super();
  }

  public setupDisplay (assets: any): void {
    super.setupDisplay(assets);
    this.mc = (<any>this.display.children[0]);

    let { data, index } = <any>this.assetDescriptors[0];
    this.data = data;
    this.setup();
  }

  protected setup (): void {
    const offsetX = 20;
    const offsetY = 0;

    this.direction_txt = this.getText();
    this.direction_txt.x = offsetX;
    this.direction_txt.y = offsetY;
    this.direction_txt.text = this.data.displayValue;

    this.mc.addChildAt(this.direction_txt, 0);
  }

  protected getText() {
    const fontFamily = 'Proxima Nova Light';
    const align = 'left';
    const fontSize = 75;
    const fill = 0xFFFFFF;

    let txt = new PIXI.Text('', { fontFamily, fontSize, align, fill });
    txt.style.wordWrap = true;
    txt.style.breakWords = true;
    txt.style.wordWrapWidth = 1120;
    return txt;
  }
}
