// jibo-log/lib/jibo-log.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboLog = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var fsExtra = require("fs-extra");
var osHomedir = require("os-homedir");
var path = require("path");
var types = require("./types");
var defaultConfig_1 = require("./defaultConfig");
var LogOutputs_1 = require("./LogOutputs");
var types_1 = require("./types");
var Level_1 = require("./types/Level");
var utils_1 = require("./utils");
var PromiseUtils_1 = require("./utils/PromiseUtils");
var output_handlers_1 = require("./output-handlers");
var NAMESPACE_DELIMITER = '.';
var prify = PromiseUtils_1.PromiseUtils.promisify;
/**
 * Log class - centralized logging facility
 * @class Log
 * @param {string} [namespace=''] The namespace to log to
 * @param {Log} [parent] The log parent (prepended to the namespace)
 */
var Log = (function () {
    function Log(namespace, parent) {
        if (namespace === void 0) { namespace = ''; }
        this.namespace =
            parent
                ? parent.namespace + NAMESPACE_DELIMITER + namespace
                : namespace;
    }
    Object.defineProperty(Log, "outputs", {
        /**
         * The level of each output channel of the global logger
         * @name Log.outputs
         * @type {LogOutputs}
         */
        get: function () {
            return Log.global.outputs;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Load the default logging configuration statically
     * @method Log.initialize
     */
    Log.initialize = function () {
        if (!Log._initialized) {
            Log._initialized = true;
            Log._loadConfig(defaultConfig_1.default);
        }
    };
    /**
     * Set the logging configuration from input (e.g. parsed from JSON file)
     * @method Log.loadConfig
     * @param {LoggingConfig} loggingConfig The configuration to load
     */
    // Load local configuration (on- or off-robot) after processes load their
    // own configuration, so that local configuration takes precedence
    Log.loadConfig = function (loggingConfig) {
        return __awaiter(this, void 0, void 0, function () {
            var configPath_1, _a, _b, _c, _d, err_1, oldConfigPath_1, configPath_2, oldConfigExists, configExists, err_2, _e, _f, _g, _h, err_3;
            return __generator(this, function (_j) {
                switch (_j.label) {
                    case 0:
                        Log._loadConfig(loggingConfig);
                        if (!!Log._skipLocalConfig) return [3 /*break*/, 19];
                        if (!Log._onRobot) return [3 /*break*/, 6];
                        configPath_1 = '/var/jibo/t.logging.json';
                        return [4 /*yield*/, prify(function (h) { return fs.exists(configPath_1, h); }, false)];
                    case 1:
                        if (!_j.sent()) return [3 /*break*/, 5];
                        _j.label = 2;
                    case 2:
                        _j.trys.push([2, 4, , 5]);
                        _b = (_a = Log)._loadConfig;
                        _d = (_c = JSON).parse;
                        return [4 /*yield*/, prify(function (h) { return fs.readFile(configPath_1, 'utf-8', h); })];
                    case 3:
                        _b.apply(_a, [_d.apply(_c, [_j.sent()])]);
                        Log._jiboLog.debug("Loaded local config from " + configPath_1);
                        return [3 /*break*/, 5];
                    case 4:
                        err_1 = _j.sent();
                        Log._jiboLog.warn("Could not load logging config file " + configPath_1, err_1);
                        return [3 /*break*/, 5];
                    case 5: return [3 /*break*/, 19];
                    case 6:
                        oldConfigPath_1 = path.resolve(osHomedir(), '.jibo', 'logging.json');
                        configPath_2 = path.resolve(osHomedir(), '.jibo', 't.logging.json');
                        return [4 /*yield*/, prify(function (h) { return fs.exists(oldConfigPath_1, h); }, false)];
                    case 7:
                        oldConfigExists = _j.sent();
                        return [4 /*yield*/, prify(function (h) { return fs.exists(configPath_2, h); }, false)];
                    case 8:
                        configExists = _j.sent();
                        if (!oldConfigExists) return [3 /*break*/, 12];
                        if (!configExists) return [3 /*break*/, 9];
                        Log._jiboLog.warn('Found old logging.json file, ignoring; please remove in favor of t.logging.json');
                        return [3 /*break*/, 12];
                    case 9:
                        _j.trys.push([9, 11, , 12]);
                        return [4 /*yield*/, prify(function (h) {
                                return fs.rename(oldConfigPath_1, configPath_2, h);
                            })];
                    case 10:
                        _j.sent();
                        configExists = true;
                        Log._jiboLog.info('Renamed logging.json to t.logging.json');
                        return [3 /*break*/, 12];
                    case 11:
                        err_2 = _j.sent();
                        Log._jiboLog.warn('Could not rename logging.json to t.logging.json');
                        return [3 /*break*/, 12];
                    case 12:
                        if (!configExists) return [3 /*break*/, 17];
                        _j.label = 13;
                    case 13:
                        _j.trys.push([13, 15, , 16]);
                        _f = (_e = Log)._loadConfig;
                        _h = (_g = JSON).parse;
                        return [4 /*yield*/, prify(function (h) {
                                return fs.readFile(configPath_2, 'utf-8', h);
                            })];
                    case 14:
                        _f.apply(_e, [_h.apply(_g, [_j.sent()])]);
                        Log._jiboLog.debug("Loaded local config from " + configPath_2);
                        return [3 /*break*/, 16];
                    case 15:
                        err_3 = _j.sent();
                        Log._jiboLog.warn("Could not load logging config file " + configPath_2, err_3);
                        return [3 /*break*/, 16];
                    case 16: return [3 /*break*/, 19];
                    case 17: return [4 /*yield*/, Log.saveConfig()];
                    case 18:
                        _j.sent();
                        _j.label = 19;
                    case 19: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Generate LoggingConfig for output (e.g. to stringify to JSON file)
     * @method Log.getConfig
     * @return {LoggingConfig}
     */
    Log.getConfig = function () {
        var logUncaughtExceptions = Log._globalErrorHandler.logUncaughtExceptions;
        var logUnhandledRejections = Log._globalErrorHandler.logUnhandledRejections;
        var stackTraceLimit = Error.stackTraceLimit;
        // Don't save out the skipLocalConfig value - it's just for tests
        var outputs = {
            console: this._outputHandlers.console.config,
            syslog: this._outputHandlers.syslog.config,
            file: this._outputHandlers.file.config
        };
        var namespaces = {};
        for (var namespace in Log._outputPerNamespace) {
            namespaces[namespace] = Log._outputPerNamespace[namespace];
        }
        return {
            logUncaughtExceptions: logUncaughtExceptions,
            logUnhandledRejections: logUnhandledRejections,
            stackTraceLimit: stackTraceLimit,
            outputs: outputs,
            namespaces: namespaces
        };
    };
    /**
     * Save the current log configuration file to a local t.logging.json file
     * On-robot: /var/jibo/t.logging.json
     * Off-robot: ~/.jibo/t.logging.json
     * @method Log.saveConfig
     */
    Log.saveConfig = function () {
        return __awaiter(this, void 0, void 0, function () {
            var configPath, err_4, stringConfig_1, err_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        configPath = Log._onRobot
                            ? '/var/jibo/t.logging.json'
                            : path.resolve(osHomedir(), '.jibo', 't.logging.json');
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prify(function (h) { return fsExtra.ensureDir(path.dirname(configPath), h); })];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        err_4 = _a.sent();
                        Log._jiboLog.warn("Error creating directory for " + configPath, err_4);
                        return [2 /*return*/];
                    case 4:
                        _a.trys.push([4, 6, , 7]);
                        stringConfig_1 = JSON.stringify(Log.getConfig(), null, 4);
                        return [4 /*yield*/, prify(function (h) { return fs.writeFile(configPath, stringConfig_1, h); })];
                    case 5:
                        _a.sent();
                        return [3 /*break*/, 7];
                    case 6:
                        err_5 = _a.sent();
                        Log._jiboLog.warn("Error writing local config to " + configPath, err_5);
                        return [2 /*return*/];
                    case 7:
                        Log._jiboLog.info("Wrote local config to " + configPath);
                        return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(Log, "_onRobot", {
        get: function () {
            // Check environment variables for run mode
            var runMode = process.env.runMode || process.env.RUNMODE;
            // Default to ON_ROBOT under arm/linux if not set
            if (!runMode && process.platform === 'linux' && process.arch === 'arm') {
                runMode = 'ON_ROBOT';
            }
            return runMode === 'ON_ROBOT';
        },
        enumerable: true,
        configurable: true
    });
    Log._loadConfig = function (loggingConfig) {
        if (loggingConfig.logUncaughtExceptions) {
            Log._globalErrorHandler.logUncaughtExceptions =
                loggingConfig.logUncaughtExceptions;
        }
        if (loggingConfig.logUnhandledRejections) {
            Log._globalErrorHandler.logUnhandledRejections =
                loggingConfig.logUnhandledRejections;
        }
        if (loggingConfig.stackTraceLimit) {
            Error.stackTraceLimit = loggingConfig.stackTraceLimit;
        }
        if (loggingConfig.skipLocalConfig) {
            Log._skipLocalConfig = loggingConfig.skipLocalConfig;
        }
        if (loggingConfig.outputs) {
            var outputsIn = loggingConfig.outputs;
            for (var output in outputsIn) {
                this._outputHandlers[output].config = outputsIn[output];
            }
        }
        if (loggingConfig.namespaces) {
            var outputPerNamespaceIn = loggingConfig.namespaces;
            for (var namespace in outputPerNamespaceIn) {
                var levelPerOutput = Log._outputPerNamespace[namespace] = {};
                var levelPerOutputIn = outputPerNamespaceIn[namespace];
                for (var output in levelPerOutputIn) {
                    var levelIn = levelPerOutputIn[output];
                    var level = levelIn;
                    levelPerOutput[output] = level;
                }
            }
        }
    };
    // Chop off the last part of the namespace, including its delimiter
    Log._parentNamespace = function (namespace) {
        var lastSlash = namespace.lastIndexOf(NAMESPACE_DELIMITER);
        return lastSlash > -1
            ? namespace.slice(0, lastSlash)
            : '';
    };
    Object.defineProperty(Log, "logUncaughtExceptions", {
        get: function () {
            return Log._globalErrorHandler.logUncaughtExceptions;
        },
        /**
         * Whether or not to log globally uncaught exceptions
         * @name Log.logUncaughtExceptions
         * @type {boolean}
         */
        set: function (log) {
            Log._globalErrorHandler.logUncaughtExceptions = log;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Log, "logUnhandledRejections", {
        get: function () {
            return Log._globalErrorHandler.logUnhandledRejections;
        },
        /**
         * Whether or not to log globally unhandled promise rejections
         * @name Log.logUnhandledRejections
         * @type {boolean}
         */
        set: function (log) {
            Log._globalErrorHandler.logUnhandledRejections = log;
        },
        enumerable: true,
        configurable: true
    });
    // Check to see if the first level is greater than or equal to the second.
    Log._levelGreaterOrEqual = function (levelOne, levelTwo) {
        return Level_1.levelOrder.indexOf(levelOne) >= Level_1.levelOrder.indexOf(levelTwo);
    };
    Object.defineProperty(Log.prototype, "global", {
        /**
         * A reference to the top level logger instance with namespace ''
         * @name Log#global
         * @type {Log}
         */
        get: function () { return Log.global; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Log.prototype, "outputs", {
        /**
         * Configuration for the output of this logger
         * @name Log#outputs
         * @type {LogOutputs}
         */
        get: function () {
            var namespace = this.namespace;
            while (!Log._outputPerNamespace[namespace]) {
                namespace = Log._parentNamespace(namespace);
            }
            return new LogOutputs_1.default(new Log(namespace), this.namespace);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Creates a new Log instance with namespace `this.namespace + '/' + subNamespace`
     * @method Log#createChild
     * @param {string} subNamespace New namespace which will be appended to the parent namespace
     * @return {Log}
     */
    Log.prototype.createChild = function (subNamespace) {
        return new Log(subNamespace, this);
    };
    /**
     * Set the minimum logging level for a given output for this namespace
     * This is generally not invoked directly, but by LogOutputs by setting
     * one of the properties of the outputs property of a log instance
     * e.g. `log.outputs.console = 'debug';`
     * @method Log#setLevelForOutput
     * @param {Output} output The output for which to set the level
     * @param {Level} level The level to which to set the output
     */
    Log.prototype.setLevelForOutput = function (output, level) {
        var namespaceConfig = Log._outputPerNamespace[this.namespace];
        if (!namespaceConfig) {
            namespaceConfig = Log._outputPerNamespace[this.namespace] = {};
        }
        namespaceConfig[output.toString()] = level;
    };
    /**
     * Check minimum logging level for each output channel for this log's
     * namespace, and log the message if this level is debug or lower
     * @method Log#debug
     * @param {any[]} ...args The arguments to log
     * @return {Log} Instance of log for chaining
     */
    Log.prototype.debug = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this._log.apply(this, [types_1.Level.debug].concat(args));
    };
    /**
     * Check minimum logging level for each output channel for this log's
     * namespace, and log the message if this level is info or lower
     * @method Log#info
     * @param {any[]} ...args The arguments to log
     * @return {Log} Instance of log for chaining
     */
    Log.prototype.info = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this._log.apply(this, [types_1.Level.info].concat(args));
    };
    /**
     * Check minimum logging level for each output channel for this log's
     * namespace, and log the message if this level is warn or lower
     * @method Log#warn
     * @param {any[]} ...args The arguments to log
     * @return {Log} Instance of log for chaining
     */
    Log.prototype.warn = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this._log.apply(this, [types_1.Level.warn].concat(args));
    };
    /**
     * Check minimum logging level for each output channel for this log's
     * namespace, and log the message if this level isn't none
     * @method Log#error
     * @param {any[]} ...args The arguments to log
     * @return {Log} Instance of log for chaining
     */
    Log.prototype.error = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this._log.apply(this, [types_1.Level.error].concat(args));
    };
    /**
     * A synonym for Log#debug
     * @method Log#log
     * @param {any[]} ...args The arguments to log
     * @return {Log} Instance of log for chaining
     */
    Log.prototype.log = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.debug.apply(this, args);
        return this;
    };
    /**
     * Log the error if it is set
     * @method Log#iferr
     * @param {Error|String} err The error which is checked and logged if truthy
     * @param {any[]} [...args] Additional arguments to log along with err
     * @return {Log} Instance of log for chaining
     */
    Log.prototype.iferr = function (err) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (err) {
            args.push(err); // add the error to the end
            this.error.apply(// add the error to the end
            this, args);
        }
        return this;
    };
    /**
     * Get the minimum logging level for the given output for this namespace.
     * If a level isn't set for the namespace, it checks its parent namespace
     * and so on until it finds one, or uses the global namespace.
     * @method Log#getLevelForOutput
     * @param {Output} output The output channel for which to get the level
     */
    Log.prototype.getLevelForOutput = function (output) {
        var namespace = this.namespace;
        var config = Log._outputPerNamespace[namespace];
        while (!(config && config[output.toString()])) {
            if (!namespace || namespace.length < 1) {
                break;
            }
            namespace = Log._parentNamespace(namespace);
            config = Log._outputPerNamespace[namespace];
        }
        return config && config[output.toString()]
            ? config[output.toString()]
            : types_1.Level.none;
    };
    // Check minimum logging level for each output channel for this log's
    // namespace, and log the message if the passed-in level is greater than or
    // equal to the minimum logging level for the output
    Log.prototype._log = function (level) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        for (var output in this._outputConfig) {
            if (Log._levelGreaterOrEqual(level, this.getLevelForOutput(types_1.Output[output]))) {
                (_a = Log._outputHandlers[output]).write.apply(_a, [Log.processName,
                    this.namespace,
                    level].concat(args));
            }
        }
        return this;
        var _a;
    };
    Object.defineProperty(Log.prototype, "_outputConfig", {
        // Get the minimum logging level for each output channel for this namespace
        // Note that getLevelForOutput walks up the namespace hierarchy if needed.
        get: function () {
            var outputConfig = {};
            for (var _i = 0, _a = Log._outputs; _i < _a.length; _i++) {
                var output = _a[_i];
                var level = this.getLevelForOutput(output);
                if (level && level !== types_1.Level.none) {
                    outputConfig[output.toString()] = level;
                }
            }
            return outputConfig;
        },
        enumerable: true,
        configurable: true
    });
    return Log;
}());
/**
 * Current version of the library.
 * @name Log.version
 * @type {string}
 */
Log.version = '3.2.3';
/**
 * The global logger, with empty namespace
 */
Log.global = new Log();
/**
 * Name of the currently running process, sent to syslog in the ProcessName
 * header; should be set by each process during its initialization
 * @name processName {string}
 */
Log.processName = "unknown";
// Local logger for use only from this class
Log._jiboLog = new Log('Jibo.Log');
// Whether or not the default logging config has been loaded
Log._initialized = false;
// Whether or not to skip loading local logging configuration (for tests)
Log._skipLocalConfig = false;
// Global logging configuration, usually loaded from JSON
// Holds the minimum logging levels per output, per namespace
Log._outputPerNamespace = {};
// A list of the output levels, for enumerating
Log._outputs = utils_1.enumerate(types_1.Output);
// The handlers for each output channel, referenced by their name
Log._outputHandlers = output_handlers_1.default;
// A special logger just for unhandled exceptions
Log._unhandledExceptionLogger = new Log('Unhandled');
Log._uncaughtMap = {};
// Global error handler, for globally uncaught exceptions and globally
// unhandled rejections
Log._globalErrorHandler = new utils_1.GlobalErrorHandler(function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    var message = args[0];
    if (message instanceof Error) {
        message = message.message;
    }
    var firstLog = false;
    if (Log._uncaughtMap[message] === undefined) {
        firstLog = true;
        Log._uncaughtMap[message] = Date.now();
    }
    // //only log every 5 minutes if happening at 30fps
    if (Date.now() - Log._uncaughtMap[message] > 60000 || firstLog) {
        (_a = Log._unhandledExceptionLogger).error.apply(_a, args);
        Log._uncaughtMap[message] = Date.now();
    }
    var _a;
});
// Write to all outputs except console
// private _writeAllNoConsole(level: Level, ...args: any[]): void {
//     for(let output in this._outputConfig) {
//         if(Output[output] !== Output.console
//             && Log._levelGreaterOrEqual(
//                 level,
//                 this.getLevelForOutput(Output[output]))
//         ) {
//             Log._outputHandlers[output].write(
//                 Log.processName,
//                 this.namespace,
//                 level,
//                 ...args
//             );
//         }
//     }
// }
// These are here for testing purposes only
// tslint:disable
Log.LogOutputs = LogOutputs_1.default;
Log.types = types;
Log.Level = types_1.Level;
Log.Output = types_1.Output;
Log.ConsoleHandler = output_handlers_1.ConsoleHandler;
Log.SyslogHandler = output_handlers_1.SyslogHandler;
Log.FileHandler = output_handlers_1.FileHandler;
exports.default = Log;

},{"./LogOutputs":2,"./defaultConfig":4,"./output-handlers":10,"./types":13,"./types/Level":11,"./utils":18,"./utils/PromiseUtils":14,"fs":undefined,"fs-extra":undefined,"os-homedir":undefined,"path":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Log_1 = require("./Log");
var types_1 = require("./types");
/**
 * LogOutputs class - logging level for each output channel
 * @class LogOutputs
 * @param {Log} [_log] The log on which this is the outputs property
 * @param {string} [_namespace?] Defaults to the log's namespace, but overridable
 */
var LogOutputs = (function () {
    function LogOutputs(_log, _namespace) {
        this._log = _log;
        this._namespace = _namespace;
        if (!this._log) {
            throw new Error('log is a required parameter');
        }
    }
    LogOutputs.prototype.createIfNeeded = function () {
        if (this._namespace && this._namespace !== this._log.namespace) {
            this._log = new Log_1.default(this._namespace);
        }
    };
    Object.defineProperty(LogOutputs.prototype, "console", {
        get: function () {
            return this._log.getLevelForOutput(types_1.Output.console);
        },
        /**
         * The logging level for the console output channel
         * @name LogOutputs#console
         * @type {Level}
         */
        set: function (level) {
            this.createIfNeeded();
            this._log.setLevelForOutput(types_1.Output.console, level);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LogOutputs.prototype, "syslog", {
        get: function () {
            return this._log.getLevelForOutput(types_1.Output.syslog);
        },
        /**
         * The logging level for the syslog output channel
         * @name LogOutputs#syslog
         * @type {Level}
         */
        set: function (level) {
            this.createIfNeeded();
            this._log.setLevelForOutput(types_1.Output.syslog, level);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LogOutputs.prototype, "file", {
        get: function () {
            return this._log.getLevelForOutput(types_1.Output.file);
        },
        /**
         * The logging level for the file output channel
         * @name LogOutputs#file
         * @type {Level}
         */
        set: function (level) {
            this.createIfNeeded();
            this._log.setLevelForOutput(types_1.Output.file, level);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LogOutputs.prototype, "all", {
        /**
         * Set the logging level for all output channels at once
         * @name LogOutputs#all
         * @type {Level}
         */
        set: function (level) {
            this.console = level;
            this.syslog = level;
            this.file = level;
        },
        enumerable: true,
        configurable: true
    });
    return LogOutputs;
}());
exports.default = LogOutputs;

},{"./Log":1,"./types":13}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
global.Singletons = global.Singletons || {};
var Singletons = (function () {
    function Singletons() {
    }
    Singletons.enforce = function (candidate) {
        var singleton = candidate;
        var singletonName = candidate.name;
        if (!global.Singletons[singletonName]) {
            global.Singletons[singletonName] = singleton;
        }
        else {
            singleton = global.Singletons[singletonName];
            // let's only warn if the versions don't match
            // console.warn(
            // 'Singletons: using existing singleton',
            // singletonName
            // );
            // optionally check if versions match (and warn if not)
            if (candidate.version) {
                var candidateVersion = candidate.version;
                var singletonVersion = singleton.version;
                if (typeof candidate.version === 'function') {
                    candidateVersion = candidate.version();
                }
                if (typeof singleton.version === 'function') {
                    singletonVersion = singleton.version();
                }
                if (candidateVersion !== singletonVersion) {
                    console.log(singletonName, 'singleton: existing version', singletonVersion, 'does not equal candidate version', candidateVersion);
                }
            }
        }
        return singleton;
    };
    return Singletons;
}());
exports.default = Singletons;

},{}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var types = require("./types");
var Level = types.Level;
var FileHandler_1 = require("./output-handlers/FileHandler");
exports.default = {
    logUncaughtExceptions: true,
    logUnhandledRejections: true,
    stackTraceLimit: 30,
    skipLocalConfig: false,
    outputs: {
        console: {
            outputFileAndLine: false
        },
        syslog: {
            port: 514,
            target: "127.0.0.1",
            outputFileAndLine: false
        },
        file: {
            filename: '',
            mode: FileHandler_1.FileWriteMode.overwrite
        }
    },
    namespaces: {
        '': {
            console: Level.info,
            syslog: Level.info,
            file: Level.none
        }
    }
};

},{"./output-handlers/FileHandler":7,"./types":13}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Log_1 = require("./Log");
var types_1 = require("./types");
function warn(msg) {
    var stack = new Error().stack;
    stack = stack.split('\n').splice(3).join('\n');
    if (console.groupCollapsed) {
        console.groupCollapsed("%cDeprecation Warning: %c%s", "color:#614108;background:#fffbe6", "font-weight:normal;color:#614108;background:#fffbe6", msg);
        console.warn(stack);
        console.groupEnd();
    }
    else {
        console.warn("Deprecation Warning:", msg);
        console.warn(stack);
    }
}
var LogPrototype = Log_1.default.prototype;
/**
 * Log to syslog
 * @method Log#toSyslog
 * @deprecated since 3.0.0
 * @see LogOutputs#syslog
 */
LogPrototype.toSyslog = function () {
    warn('toSyslog is deprecated; please use JSON configuration');
    this.outputs.syslog = types_1.Level.debug;
};
/**
 * Log to syslog
 * @method Log#toFile
 * @deprecated since 3.0.0
 * @see LogOutputs#file
 */
LogPrototype.toFile = function (filename) {
    warn('toFile is deprecated, please use JSON configuration');
    this.outputs.file = types_1.Level.debug;
};
/**
 * Set minimum logging level
 * @method Log#toFile
 * @deprecated since 3.0.0
 * @see LogOutputs#all
 */
Object.defineProperty(LogPrototype, 'minLevel', {
    set: function (level) {
        // TODO: Figure out how to properly add a property setter to an any prototype
        warn('minLevel is deprecated, please use JSON configuration');
        this.outputs.all = level;
    }
});
/**
 * Allow handling uncaught and unhandled exceptions.
 * @method Log#enableUncaughtAndUnhandled
 * @deprecated since 1.1.0
 * @see Log.logUncaughtExceptions
 * @see Log.logUnhandledRejections
 */
LogPrototype.enableUncaughtAndUnhandled = function () {
    warn('enableUncaughtAndUnhandled is deprecated; please use JSON configuration');
    this.handleAll = true;
};
/**
 * Allow handling uncaught and unhandled exceptions.
 * @method Log#disableUncaughtAndUnhandled
 * @deprecated since 1.1.0
 * @see Log.logUncaughtExceptions
 * @see Log.logUnhandledRejections
 */
LogPrototype.disableUncaughtAndUnhandled = function () {
    warn('enableUncaughtAndUnhandled is deprecated; please use JSON configuration');
    this.handleAll = false;
};
/**
 * Set the default logging directory.
 * @method Log#setDefaultLogDir
 * @deprecated since 1.1.0
 * @see FileConfig#filename
 */
LogPrototype.setDefaultLogDir = function (dir, callback) {
    warn('setDefaultLogDir is deprecated; please use JSON configuration');
    this.defaultDir = dir;
    if (callback) {
        setImmediate(callback);
    }
};
/**
 * Allow for debug messages.
 * @method Log#enableDebug
 * @deprecated since 1.1.0
 * @see LogOutputs#all
 */
LogPrototype.enableDebug = function (name, level) {
    if (level === void 0) { level = types_1.Level.debug; }
    warn('enableDebug is deprecated, please use JSON configuration');
    this.outputs.all = level;
};

},{"./Log":1,"./types":13}],6:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var OutputHandlerBase_1 = require("./OutputHandlerBase");
var types_1 = require("../types");
/**
 * ConsoleHandler class - logging to global.console
 * @class ConsoleHandler
 * @param {any} [_console = global.console] You may pass a alternative console
 */
var ConsoleHandler = (function (_super) {
    __extends(ConsoleHandler, _super);
    function ConsoleHandler(_console) {
        if (_console === void 0) { _console = global.console; }
        var _this = _super.call(this) || this;
        _this._console = _console;
        return _this;
    }
    /**
     * Write to the console
     * @method ConsoleHandler#write
     * @param {string} processName The name of the running process
     * @param {string} namespace The Log namespace
     * @param {Level} level The level of logging
     * @param {any[]} ...args The arguments to log
     */
    ConsoleHandler.prototype.write = function (processName, namespace, level) {
        var args = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            args[_i - 3] = arguments[_i];
        }
        if (this._config && this._config.outputFileAndLine) {
            args.push(new Error().stack.split('\n')[4]);
        }
        switch (level) {
            case types_1.Level.debug:
                (_a = this._console).log.apply(_a, [namespace].concat(args));
                break;
            case types_1.Level.info:
                (_b = this._console).info.apply(_b, [namespace].concat(args));
                break;
            case types_1.Level.warn:
                (_c = this._console).warn.apply(_c, [namespace].concat(args));
                break;
            case types_1.Level.error:
                (_d = this._console).error.apply(_d, [namespace].concat(args));
                break;
        }
        var _a, _b, _c, _d;
    };
    Object.defineProperty(ConsoleHandler.prototype, "config", {
        get: function () {
            return this._config;
        },
        /**
         * The console output configuration for this handler
         * @name ConsoleHandler#config
         * @type {ConsoleConfig}
         */
        set: function (config) {
            this._config = config;
        },
        enumerable: true,
        configurable: true
    });
    return ConsoleHandler;
}(OutputHandlerBase_1.default));
exports.default = ConsoleHandler;

},{"../types":13,"./OutputHandlerBase":8}],7:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var OutputHandlerBase_1 = require("./OutputHandlerBase");
var FileWriteMode;
(function (FileWriteMode) {
    FileWriteMode.overwrite = 'overwrite';
    FileWriteMode.append = 'append';
})(FileWriteMode || (FileWriteMode = {}));
exports.FileWriteMode = FileWriteMode;
/**
 * FileHandler class - logging to filesystem
 * @class FileHandler
 */
var FileHandler = (function (_super) {
    __extends(FileHandler, _super);
    function FileHandler() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Write to the file
     * @method FileHandler#write
     * @param {string} processName The name of the running process
     * @param {string} namespace The Log namespace
     * @param {Level} level The level of logging
     * @param {any[]} ...args The arguments to log
     */
    FileHandler.prototype.write = function (processName, namespace, level) {
        var args = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            args[_i - 3] = arguments[_i];
        }
        return;
    };
    Object.defineProperty(FileHandler.prototype, "config", {
        get: function () {
            return this._config;
        },
        /**
         * The filesystem output configuration for this handler
         * @name FileHandler#config
         * @type {FileConfig}
         */
        set: function (config) {
            this._config = config;
        },
        enumerable: true,
        configurable: true
    });
    return FileHandler;
}(OutputHandlerBase_1.default));
exports.default = FileHandler;

},{"./OutputHandlerBase":8}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OutputHandlerBase = (function () {
    function OutputHandlerBase() {
    }
    return OutputHandlerBase;
}());
exports.default = OutputHandlerBase;

},{}],9:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var OutputHandlerBase_1 = require("./OutputHandlerBase");
var syslog = require("syslog-rfc5424-client");
var types_1 = require("../types");
var utils_1 = require("../utils");
/**
 * SyslogHandler class - logging to syslog (/var/log/messages)
 * Only enabled on linux/arm (e.g. Jibo robot)
 * @class SyslogHandler
 */
var SyslogHandler = (function (_super) {
    __extends(SyslogHandler, _super);
    function SyslogHandler() {
        var _this = _super.call(this) || this;
        // don't bother on anything other than our linux/arm environment
        _this._enabled = process.platform === 'linux' && process.arch === 'arm';
        return _this;
    }
    /**
     * Write to the log
     * @method SyslogHandler#write
     * @param {string} processName The name of the running process
     * @param {string} namespace The Log namespace
     * @param {Level} level The level of logging
     * @param {any[]} ...args The arguments to log
     */
    SyslogHandler.prototype.write = function (processName, namespace, level) {
        var args = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            args[_i - 3] = arguments[_i];
        }
        if (this._config && this._config.outputFileAndLine) {
            args.push("\n\t" + new Error().stack.split('\n')[4]);
        }
        // only write to syslog if it's enabled and has a client
        if (this._enabled && this._client) {
            var type = level === types_1.Level.debug ? "Debug" :
                level === types_1.Level.warn ? "Warning" :
                    level === types_1.Level.error ? "Error" :
                        "Informational";
            var logOptions = {
                facility: syslog.Facility.Local0,
                severity: syslog.Severity[type]
            };
            var message = processName + " " + process.pid + " - [] T." + namespace + ": " + utils_1.formatOutput(args);
            this._client.log(message, logOptions, function (error) {
                if (error) {
                    console.error(error);
                }
            });
        }
    };
    Object.defineProperty(SyslogHandler.prototype, "config", {
        get: function () {
            return this._config;
        },
        /**
         * The syslog output configuration for this handler
         * @name SyslogHandler#config
         * @type {SyslogConfig}
         */
        set: function (config) {
            if (!this._config) {
                this._config = {};
            }
            if (config.port) {
                this._config.port = config.port;
            }
            if (config.target) {
                this._config.target = config.target;
            }
            if (config.outputFileAndLine) {
                this._config.outputFileAndLine = config.outputFileAndLine;
            }
            this._createClient();
        },
        enumerable: true,
        configurable: true
    });
    // Create a syslog client on the configured port and IP/hostname
    SyslogHandler.prototype._createClient = function () {
        if (this._client) {
            this._client.close();
            this._client = undefined;
        }
        if (this._enabled) {
            var options = {
                port: this._config.port,
                transport: syslog.Transport.Udp
            };
            this._client = syslog.createClient(this._config.target, options);
            console.info("Initializing a syslog client");
            this._client.on("error", function (error) {
                console.error(error);
                if (this._client) {
                    this._client.close();
                    this._client = undefined;
                }
            });
        }
    };
    return SyslogHandler;
}(OutputHandlerBase_1.default));
exports.default = SyslogHandler;

},{"../types":13,"../utils":18,"./OutputHandlerBase":8,"syslog-rfc5424-client":undefined}],10:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ConsoleHandler_1 = require("./ConsoleHandler");
exports.ConsoleHandler = ConsoleHandler_1.default;
var ConsoleHandler_2 = require("./ConsoleHandler");
exports.ConsoleConfig = ConsoleHandler_2.default;
var SyslogHandler_1 = require("./SyslogHandler");
exports.SyslogHandler = SyslogHandler_1.default;
var SyslogHandler_2 = require("./SyslogHandler");
exports.SyslogConfig = SyslogHandler_2.default;
var FileHandler_1 = require("./FileHandler");
exports.FileHandler = FileHandler_1.default;
exports.FileWriteMode = FileHandler_1.FileWriteMode;
var types_1 = require("../types");
var FileHandler_2 = require("./FileHandler");
exports.FileConfig = FileHandler_2.default;
var outputHandlers = (_a = {},
    _a[types_1.Output.console] = new ConsoleHandler_1.default(),
    _a[types_1.Output.syslog] = new SyslogHandler_1.default(),
    _a[types_1.Output.file] = new FileHandler_1.default(),
    _a);
exports.default = outputHandlers;
var _a;

},{"../types":13,"./ConsoleHandler":6,"./FileHandler":7,"./SyslogHandler":9}],11:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Level;
(function (Level) {
    Level.none = 'none';
    Level.debug = 'debug';
    Level.info = 'info';
    Level.warn = 'warn';
    Level.error = 'error';
})(Level || (Level = {}));
exports.default = Level;
exports.levelOrder = [
    Level.debug,
    Level.info,
    Level.warn,
    Level.error
];

},{}],12:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Output;
(function (Output) {
    Output.console = 'console';
    Output.syslog = 'syslog';
    Output.file = 'file';
})(Output || (Output = {}));
exports.default = Output;

},{}],13:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Level_1 = require("./Level");
exports.Level = Level_1.default;
var Output_1 = require("./Output");
exports.Output = Output_1.default;

},{"./Level":11,"./Output":12}],14:[function(require,module,exports){
"use strict";
/**
 * @fileOverview
 *
 * Created on 5/12/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var PromiseUtils = (function () {
    function PromiseUtils() {
    }
    /**
     * Returns promise that succeeds when the first input promise succeeds
     * If all input promises fail, then the returned promise fails.
     * @param {Promise<T>[]} promises
     * @returns {Promise<T>}
     */
    PromiseUtils.firstToSucceed = function (promises) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (res, rej) {
                        var errors = [];
                        var errorCount = 0;
                        var _loop_1 = function (i) {
                            promises[i].then(res).catch(function (e) {
                                errors[i] = e;
                                errorCount++;
                                if (errorCount === promises.length) {
                                    rej(errors);
                                }
                            });
                        };
                        for (var i = 0; i < promises.length; i++) {
                            _loop_1(i);
                        }
                    })];
            });
        });
    };
    PromiseUtils.promisify = function (func, firstParamError) {
        if (firstParamError === void 0) { firstParamError = true; }
        return new Promise(function (resolve, reject) {
            func(function (error, result) {
                if (firstParamError) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(result);
                    }
                }
                else {
                    resolve(error);
                }
            });
        });
    };
    return PromiseUtils;
}());
exports.PromiseUtils = PromiseUtils;

},{}],15:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Return a list of elements from an enum, for enumerating over them
 * e.g. `for(let output in Output) { ... }`
 * @function enumerate
 * @param {any} enumType an Enum type
 * @return {Enum[]} List of elements
 */
function enumerate(enumType) {
    return Object.keys(enumType)
        .filter(function (key) { return typeof (key) === 'string'; })
        .map(function (output) { return enumType[output]; });
}
exports.enumerate = enumerate;

},{}],16:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util = require("util");
var format_error_1 = require("format-error");
// Format the passed object for string output
var formatErrorOptions = {
    noColor: true
};
var inspectOptions = {
    depth: 3,
    maxArrayLength: 10
};
function formatError(error) {
    return formatObject(format_error_1.format(error, formatErrorOptions));
}
exports.formatError = formatError;
;
function formatObject(obj) {
    return util.inspect(obj, inspectOptions);
}
exports.formatObject = formatObject;
;
/**
 * Concatenate the passed args as a single string for logging (i.e. syslog)
 * @function formatOutput
 * @param args {any[]} Arguments to Concatenate
 * @return {string} Concatenated args.
 */
function formatOutput(args) {
    var text = args.map(function (item) {
        return typeof item === 'string' ? item :
            item instanceof Error ? formatError(item) :
                formatObject(item);
    }).join(' ').trim();
    return text;
}
exports.formatOutput = formatOutput;

},{"format-error":undefined,"util":undefined}],17:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
;
/**
 * Utility class to handle globally-uncaught exceptions and globally-unhandled
 * promise rejections
 * @class GlobalErrorHandler
 * @param  {string|Error} handleError Callback where errors are sent
 */
var GlobalErrorHandler = (function () {
    function GlobalErrorHandler(handleError) {
        this._handleError = handleError;
    }
    // Log uncaught exceptions caught by process
    GlobalErrorHandler.prototype._uncaughtExceptionProcessLogger = function (error) {
        this._handleError(error);
    };
    // Log uncaught exceptions caught by window
    GlobalErrorHandler.prototype._uncaughtExceptionWindowLogger = function (message, filename, lineno, colno, error) {
        this._handleError(message + ", source: " + filename + ", line: " + lineno + ", column: " + colno, error);
    };
    Object.defineProperty(GlobalErrorHandler.prototype, "logUncaughtExceptions", {
        get: function () {
            return this._loggingUncaughtExceptions;
        },
        /**
         * Whether or not to log globally uncaught exceptions
         * @name GlobalErrorHandler.logUncaughtExceptions
         * @type {boolean}
         */
        set: function (log) {
            if (log && !this._loggingUncaughtExceptions) {
                if (typeof window !== 'undefined') {
                    window.onerror = this._uncaughtExceptionWindowLogger.bind(this);
                }
                else {
                    process.on('uncaughtException', this._uncaughtExceptionProcessLogger.bind(this));
                }
            }
            else if (!log && this._loggingUncaughtExceptions) {
                if (typeof window !== 'undefined') {
                    window.onerror = null;
                }
                else {
                    process.removeListener('uncaughtException', this._uncaughtExceptionProcessLogger);
                }
            }
            this._loggingUncaughtExceptions = log;
        },
        enumerable: true,
        configurable: true
    });
    // Log unhandled promise rejections caught by process
    GlobalErrorHandler.prototype._unhandledRejectionProcessLogger = function (error) {
        this._handleError(error);
    };
    // In electron, unhandledrejection event is on window, not process, and
    // the callback signature is different
    GlobalErrorHandler.prototype._unhandledRejectionWindowLogger = function (event) {
        this._handleError(event.reason);
    };
    Object.defineProperty(GlobalErrorHandler.prototype, "logUnhandledRejections", {
        get: function () {
            return this._loggingUnhandledRejections;
        },
        /**
         * Whether or not to log globally unhandled promise rejections
         * @name GlobalErrorHandler.logUnhandledRejections
         * @type {boolean}
         */
        set: function (log) {
            if (log && !this._loggingUnhandledRejections) {
                if (typeof window !== 'undefined') {
                    window.onunhandledrejection =
                        this._unhandledRejectionWindowLogger.bind(this);
                }
                else {
                    process.on('unhandledRejection', this._unhandledRejectionProcessLogger.bind(this));
                }
            }
            else if (!log && this._loggingUnhandledRejections) {
                if (typeof window !== 'undefined') {
                    window.onunhandledrejection = null;
                }
                else {
                    process.removeListener('unhandledRejection', this._unhandledRejectionProcessLogger);
                }
            }
            this._loggingUnhandledRejections = log;
        },
        enumerable: true,
        configurable: true
    });
    return GlobalErrorHandler;
}());
exports.GlobalErrorHandler = GlobalErrorHandler;

},{}],18:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var enums_1 = require("./enums");
exports.enumerate = enums_1.enumerate;
var formatting_1 = require("./formatting");
exports.formatError = formatting_1.formatError;
exports.formatObject = formatting_1.formatObject;
exports.formatOutput = formatting_1.formatOutput;
var global_errors_1 = require("./global-errors");
exports.GlobalErrorHandler = global_errors_1.GlobalErrorHandler;

},{"./enums":15,"./formatting":16,"./global-errors":17}],19:[function(require,module,exports){
"use strict";
var Log_1 = require("./Log");
var Singletons_1 = require("./Singletons");
require("./deprecation");
var LogSingleton = Singletons_1.default.enforce(Log_1.default);
LogSingleton.initialize();
module.exports = LogSingleton;

},{"./Log":1,"./Singletons":3,"./deprecation":5}]},{},[19])(19)
});
