// jibo-anim-db/lib/jibo-anim-db.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboAnimDb = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Utils_1 = require("./Utils");
var Classes_1 = require("./Classes");
var AnimUtils_1 = require("./AnimUtils");
var AnimCollection = (function () {
    function AnimCollection(parent) {
        this.parent = parent;
        this.animations = [];
        this.categories = [];
        this._animationsByName = new Map();
        this._categoriesByName = new Map();
    }
    AnimCollection.prototype.addAnimation = function (animation) {
        var nameLower = animation.name.toLowerCase();
        if (!this._animationsByName.has(nameLower)) {
            this._animationsByName.set(nameLower, animation);
            this.animations.push(animation);
        }
    };
    AnimCollection.prototype.addCategory = function (category) {
        var nameLower = category.name.toLowerCase();
        if (!this._categoriesByName.has(nameLower)) {
            this._categoriesByName.set(nameLower, category);
            this.categories.push(category);
        }
    };
    AnimCollection.prototype.getCategoryByName = function (name) {
        return this._categoriesByName.get(name.toLowerCase());
    };
    AnimCollection.prototype.getAnimByName = function (name) {
        return this._animationsByName.get(name.toLowerCase());
    };
    AnimCollection.prototype.getAnimationNames = function () {
        return Array.from(this._animationsByName.keys());
    };
    AnimCollection.prototype.getAnimationCategories = function () {
        return Array.from(this._categoriesByName.keys());
    };
    AnimCollection.prototype.search = function (searchTerm) {
        var term = searchTerm.toLowerCase();
        var res = new Classes_1.SearchResults();
        this._search(term, res);
        return res;
    };
    AnimCollection.prototype.query = function (query) {
        AnimUtils_1.AnimUtils.insertDefaultAnimQueryParams(query);
        var res = new Classes_1.AnimResults();
        this._query(query, res);
        var matchingNoDupes = [];
        new Set(res.matching).forEach(function (elm) { return matchingNoDupes.push(elm); });
        res.matching = matchingNoDupes;
        var nonMatchingNoDupes = [];
        new Set(res.nonMatching).forEach(function (elm) { return nonMatchingNoDupes.push(elm); });
        res.nonMatching = nonMatchingNoDupes;
        return res;
    };
    AnimCollection.prototype._search = function (term, res) {
        var catByName = this._categoriesByName.get(term);
        if (catByName) {
            res.category = catByName;
        }
        var animByName = this._animationsByName.get(term);
        if (animByName) {
            res.name = animByName;
        }
        this.animations.forEach(function (anim) {
            anim.meta.metaTerms.forEach(function (meta) {
                if (meta === term) {
                    res.meta.push(anim);
                }
            });
        });
    };
    AnimCollection.prototype._query = function (query, res) {
        var _this = this;
        if (query.categories.length) {
            query.categories.forEach(function (catName) {
                var category = _this._categoriesByName.get(catName);
                if (category) {
                    category._query(query, res);
                }
            });
        }
        else {
            Utils_1.Utils.getAnimationsByQuery(this.animations, res, query);
        }
    };
    return AnimCollection;
}());
exports.AnimCollection = AnimCollection;

},{"./AnimUtils":4,"./Classes":7,"./Utils":10}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Classes_1 = require("./Classes");
var AnimUtils_1 = require("./AnimUtils");
var AnimDB = (function () {
    function AnimDB(jibo) {
        this.collections = [];
        this.jibo = jibo;
    }
    AnimDB.prototype.push = function (animCollection) {
        this.collections.push(animCollection);
    };
    AnimDB.prototype.pop = function () {
        return this.collections.pop();
    };
    AnimDB.prototype.getAnimByName = function (animName) {
        var anim;
        for (var i = 0; i < this.collections.length; i++) {
            var animDB = this.collections[i];
            var a = animDB.getAnimByName(animName);
            if (a) {
                anim = a;
                break;
            }
        }
        return anim;
    };
    AnimDB.prototype.getAnimationNames = function () {
        var names = [];
        for (var i = 0; i < this.collections.length; i++) {
            names = names.concat(this.collections[i].getAnimationNames());
        }
        return names;
    };
    AnimDB.prototype.getAnimationCategories = function () {
        var names = [];
        for (var i = 0; i < this.collections.length; i++) {
            names = names.concat(this.collections[i].getAnimationCategories());
        }
        return names;
    };
    AnimDB.prototype.search = function (searchTerm) {
        var term = searchTerm.toLowerCase();
        var res = new Classes_1.SearchResults();
        this.collections.forEach(function (animColl) {
            animColl._search(term, res);
        });
        return res;
    };
    AnimDB.prototype.query = function (query) {
        AnimUtils_1.AnimUtils.insertDefaultAnimQueryParams(query);
        var res = new Classes_1.AnimResults();
        this.collections.forEach(function (animColl) {
            animColl._query(query, res);
        });
        var matchingNoDupes = [];
        new Set(res.matching).forEach(function (elm) { return matchingNoDupes.push(elm); });
        res.matching = matchingNoDupes;
        var nonMatchingNoDupes = [];
        new Set(res.nonMatching).forEach(function (elm) { return nonMatchingNoDupes.push(elm); });
        res.nonMatching = nonMatchingNoDupes;
        return res;
    };
    return AnimDB;
}());
exports.AnimDB = AnimDB;

},{"./AnimUtils":4,"./Classes":7}],3:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var fs = require("fs");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var AnimDB_1 = require("./AnimDB");
var AnimCollection_1 = require("./AnimCollection");
var Animation_1 = require("./Animation");
var Category_1 = require("./Category");
var AnimUtils_1 = require("./AnimUtils");
var Utils_1 = require("./Utils");
var Classes_1 = require("./Classes");
var Range_1 = require("./Range");
var stringify = require('json-stable-stringify');
var AnimDBParser = (function () {
    function AnimDBParser() {
    }
    AnimDBParser.readAnimDB = function (jibo, animDBPath) {
        var moreAnimDBPaths = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            moreAnimDBPaths[_i - 2] = arguments[_i];
        }
        moreAnimDBPaths.unshift(animDBPath);
        var animDB = new AnimDB_1.AnimDB(jibo);
        var promises = moreAnimDBPaths.map(function (rootDir) {
            return AnimDBParser.readAndAddAnimCollection(animDB, rootDir);
        });
        return Promise.all(promises)
            .then(function () { return animDB; });
    };
    AnimDBParser.readAndAddAnimCollection = function (animDB, animDBPath) {
        var rootDir = path.dirname(animDBPath);
        return jibo_cai_utils_1.FileUtils.readFile(animDBPath)
            .then(function (animStr) {
            var animMetaMap = JSON.parse(animStr);
            AnimDBParser.addAnimCollection(animDB, rootDir, animMetaMap);
        });
    };
    AnimDBParser.addAnimCollection = function (animDB, rootDir, animMetaMap) {
        var collection = new AnimCollection_1.AnimCollection(animDB);
        AnimDBParser.addToAnimCollection(collection, rootDir, animMetaMap);
        animDB.push(collection);
    };
    AnimDBParser.addToAnimCollection = function (collection, rootDir, animMetadataCollection) {
        var kvAnimMetaArray;
        if (animMetadataCollection instanceof Array) {
            kvAnimMetaArray = animMetadataCollection;
        }
        else {
            kvAnimMetaArray = [];
            Object.keys(animMetadataCollection).forEach(function (key) {
                kvAnimMetaArray.push([key, animMetadataCollection[key]]);
            });
        }
        var _loop_1 = function (animName, animMeta) {
            if (!animMeta.name) {
                throw new Error("Missing anim name: " + animName);
            }
            animName = animMeta.name.toLowerCase();
            var anim = new Animation_1.Animation(collection, animMeta, rootDir);
            var existingAnim = collection.getAnimByName(anim.name);
            if (existingAnim) {
                var paths = [existingAnim.meta.path, animMeta.path];
                throw new Error("Duplicate animation name: '" + animName + "' at paths: " + paths);
            }
            collection.addAnimation(anim);
            anim.meta.categories.forEach(function (categoryName) {
                var cat = collection.getCategoryByName(categoryName);
                if (!cat) {
                    cat = new Category_1.Category(categoryName);
                    collection.addCategory(cat);
                }
                cat.addAnimation(anim);
            });
        };
        for (var _i = 0, kvAnimMetaArray_1 = kvAnimMetaArray; _i < kvAnimMetaArray_1.length; _i++) {
            var _a = kvAnimMetaArray_1[_i], animName = _a[0], animMeta = _a[1];
            _loop_1(animName, animMeta);
        }
    };
    AnimDBParser.createAndWriteAnimMetaData = function (jibo, rootDir, outName) {
        return AnimDBParser.createAnimMetaData(jibo, rootDir)
            .then(function (allAnimData) {
            var outPath = path.join(rootDir, outName + '.json');
            var outData = stringify(allAnimData, { space: '  ' });
            return jibo_cai_utils_1.PromiseUtils.promisify(function (h) { return fs.writeFile(outPath, outData, 'utf8', h); });
        });
    };
    AnimDBParser.createAnimMetaData = function (jibo, rootDir) {
        var animsDir = path.join(rootDir, 'animations');
        return jibo_cai_utils_1.FileUtils.findAllFilesWithExt(animsDir, 'keys')
            .then(function (filePaths) {
            var goodPaths = filePaths.filter(function (filePath) {
                var relFilePath = path.relative(animsDir, filePath);
                var pathElements = relFilePath.split(path.sep);
                return !pathElements.some(function (el) { return el.startsWith('_'); });
            });
            return AnimDBParser._createAnimMetaData(jibo, rootDir, goodPaths);
        });
    };
    AnimDBParser._createAnimMetaData = function (jibo, rootDir, paths) {
        return __awaiter(this, void 0, void 0, function () {
            var cachedAnimDB, collection, completeMetadata, unparsed, parsed, _loop_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        cachedAnimDB = jibo.animDB._animDB;
                        jibo.animDB._animDB = new AnimDB_1.AnimDB(jibo);
                        collection = new AnimCollection_1.AnimCollection(jibo.animDB._animDB);
                        jibo.animDB.push(collection);
                        completeMetadata = {};
                        unparsed = [];
                        parsed = [];
                        _loop_2 = function () {
                            var unparsedCount, _i, paths_1, path_1, meta, err_1, rootDirResolved, kvAnimMetaArray, tempAnimMetaMap;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        unparsedCount = paths.length;
                                        _i = 0, paths_1 = paths;
                                        _a.label = 1;
                                    case 1:
                                        if (!(_i < paths_1.length)) return [3, 6];
                                        path_1 = paths_1[_i];
                                        _a.label = 2;
                                    case 2:
                                        _a.trys.push([2, 4, , 5]);
                                        return [4, AnimDBParser.parseAnimation(jibo, path_1, rootDir)];
                                    case 3:
                                        meta = _a.sent();
                                        parsed.push(meta);
                                        return [3, 5];
                                    case 4:
                                        err_1 = _a.sent();
                                        if (err_1 instanceof ReferenceError) {
                                            unparsed.push(path_1);
                                        }
                                        else {
                                            throw err_1;
                                        }
                                        return [3, 5];
                                    case 5:
                                        _i++;
                                        return [3, 1];
                                    case 6:
                                        if (unparsedCount === unparsed.length) {
                                            throw new Error("Could not resolve " + unparsedCount + " animations: " + unparsed);
                                        }
                                        else {
                                            paths = unparsed;
                                        }
                                        rootDirResolved = path.resolve(rootDir);
                                        kvAnimMetaArray = [];
                                        parsed.forEach(function (animData) {
                                            var firstPart = animData.path.substring(0, rootDirResolved.length);
                                            if (firstPart !== rootDirResolved) {
                                                throw new Error("Animation doesn't have the expected root path: " + [animData.path, rootDirResolved]);
                                            }
                                            animData.path = animData.path.substring(rootDirResolved.length + path.sep.length);
                                            kvAnimMetaArray.push([animData.name, animData]);
                                        });
                                        AnimDBParser.addToAnimCollection(collection, rootDir, kvAnimMetaArray);
                                        tempAnimMetaMap = {};
                                        kvAnimMetaArray.forEach(function (_a) {
                                            var key = _a[0], value = _a[1];
                                            tempAnimMetaMap[key] = value;
                                        });
                                        Object.assign(completeMetadata, tempAnimMetaMap);
                                        parsed = [];
                                        unparsed = [];
                                        return [2];
                                }
                            });
                        };
                        _a.label = 1;
                    case 1:
                        if (!(paths.length > 0)) return [3, 3];
                        return [5, _loop_2()];
                    case 2:
                        _a.sent();
                        return [3, 1];
                    case 3:
                        jibo.animDB._animDB = cachedAnimDB;
                        return [2, completeMetadata];
                }
            });
        });
    };
    AnimDBParser.parseAnimation = function (jibo, filePath, rootDir) {
        return __awaiter(this, void 0, void 0, function () {
            var animData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, AnimDBParser.loadKeys(jibo, filePath, rootDir)];
                    case 1:
                        animData = _a.sent();
                        return [2, AnimDBParser.parseAnimationFromAnim(animData, filePath)];
                }
            });
        });
    };
    AnimDBParser.loadKeys = function (jibo, keysPath, rootDir) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, new jibo.rendering.animation.KeysLoader(keysPath, rootDir).load()];
            });
        });
    };
    AnimDBParser.parseAnimationFromAnim = function (animData, filePath) {
        var animdbTag;
        var scale, speed, name;
        try {
            var defaultName = filePath ? path.basename(filePath, path.extname(filePath)) : 'UNKNOWN_ANIM';
            animdbTag = animData.animdb || {};
            name = Utils_1.Utils.parseString(animdbTag.name, defaultName);
            scale = [
                Utils_1.Utils.parseNumber(animdbTag.scaleMin, 1),
                Utils_1.Utils.parseNumber(animdbTag.scaleMax, 1)
            ];
            speed = [
                Utils_1.Utils.parseNumber(animdbTag.speedMin, 1),
                Utils_1.Utils.parseNumber(animdbTag.speedMax, 1)
            ];
        }
        catch (error) {
            throw new Error("Error parsing animation at path '" + filePath + "': '" + error + "'");
        }
        if ((speed[0] <= 0) || (speed[0] > speed[1])) {
            throw new Error("Invalid speed bounds '" + JSON.stringify(speed) + "' for animation '" + filePath + "'");
        }
        if (scale[0] > scale[1]) {
            throw new Error("Invalid scale bounds '" + JSON.stringify(scale) + "' for animation '" + filePath + "'");
        }
        var categoryString = animdbTag.categories || '';
        var categories = [];
        categoryString.split(',').forEach(function (category) {
            var catName = category.trim().toLowerCase();
            if (catName.length) {
                categories.push(catName);
            }
        });
        var metaString = animdbTag.meta || '';
        var metaTerms = [];
        metaString.split(',').forEach(function (_metaTerm) {
            var metaName = _metaTerm.trim().toLowerCase();
            if (metaName.length) {
                metaTerms.push(metaName);
            }
        });
        var holdSafeKF = AnimUtils_1.AnimUtils.hasHoldSafe(animData, name);
        var durationRange = [
            speed[0] * animData.duration,
            speed[1] * animData.duration
        ];
        if (holdSafeKF !== -1) {
            durationRange[1] = Range_1.INFINITE_DURATION;
        }
        var layers = AnimUtils_1.AnimUtils.checkLayers(animData);
        if (metaTerms.indexOf(AnimUtils_1.AUDIO_ONLY_META) !== -1) {
            Object.keys(layers).forEach(function (layer) {
                if (layers[layer] && (layer !== Classes_1.LayerName.AudioEvent.short)) {
                    layers[layer] = false;
                }
            });
        }
        var animMetadata = {
            name: name,
            scale: scale,
            speed: speed,
            durationRange: durationRange,
            categories: categories,
            metaTerms: metaTerms,
            holdSafeKF: holdSafeKF,
            path: filePath,
            duration: animData.duration,
            startsNeutral: AnimUtils_1.AnimUtils.startsNeutral(animData),
            endsNeutral: AnimUtils_1.AnimUtils.endsNeutral(animData),
            hasAudio: AnimUtils_1.AnimUtils.hasAudio(animData),
            orientation: AnimUtils_1.AnimUtils.calculateOrientation(animData),
            layers: layers
        };
        return animMetadata;
    };
    return AnimDBParser;
}());
exports.AnimDBParser = AnimDBParser;

},{"./AnimCollection":1,"./AnimDB":2,"./AnimUtils":4,"./Animation":5,"./Category":6,"./Classes":7,"./Range":9,"./Utils":10,"fs":undefined,"jibo-cai-utils":undefined,"json-stable-stringify":undefined,"path":undefined}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Classes_1 = require("./Classes");
var LayerMocks_1 = require("./animation_formats/LayerMocks");
var BODY_ANGLE_EPSILON = 3;
var HOLD_SAFE_EVENT_NAME = 'HOLD_SAFE';
exports.AUDIO_ONLY_META = "audio-only";
var AnimUtils = (function () {
    function AnimUtils() {
    }
    AnimUtils._addMockLayer = function (animData, type) {
        animData.layers.push(LayerMocks_1.LayerTemplate.generateLayer(type, type));
    };
    AnimUtils.scaleTimeInPlace = function (animData, scale) {
        animData.layers.forEach(function (layer) {
            layer.keyframes.forEach(function (kf) {
                kf.time = Math.round(kf.time * scale);
            });
        });
        animData.duration = Math.round(animData.duration * scale);
    };
    AnimUtils.scaleBodyInPlace = function (animData, scale) {
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            layer.keyframes.forEach(function (kf) {
                kf.value.Head *= scale;
                kf.value.Pelvis *= scale;
                kf.value.Torso *= scale;
            });
        });
    };
    AnimUtils.flipScreenLeftRight = function (animData) {
        var handler = function (layer) {
            layer.keyframes.forEach(function (kf) {
                kf.value.Translate.x *= -1;
            });
        };
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Eye, handler);
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Overlay, handler);
    };
    AnimUtils.startsNeutral = function (animData) {
        var ret = true;
        var earliestFrame = animData.duration;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            if (layer.keyframes.length) {
                var kf = layer.keyframes[0];
                if (kf.time < earliestFrame) {
                    earliestFrame = kf.time;
                    if (Math.abs(kf.value.Pelvis) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Head) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Torso) > BODY_ANGLE_EPSILON) {
                        ret = false;
                    }
                }
            }
        });
        return ret;
    };
    AnimUtils.endsNeutral = function (animData) {
        var ret = true;
        var latestTime = -1;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            if (layer.keyframes.length) {
                var kf = layer.keyframes[layer.keyframes.length - 1];
                if (kf.time > latestTime) {
                    latestTime = kf.time;
                    if (Math.abs(kf.value.Pelvis) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Head) > BODY_ANGLE_EPSILON ||
                        Math.abs(kf.value.Torso) > BODY_ANGLE_EPSILON) {
                        ret = false;
                    }
                }
            }
        });
        return ret;
    };
    AnimUtils.checkLayers = function (animData, strict) {
        if (strict === void 0) { strict = true; }
        var layerPresence = {};
        Classes_1.LayerName.Layers.forEach(function (layerName) {
            if (AnimUtils.hasLayer(animData, layerName, strict)) {
                layerPresence[layerName.short] = true;
            }
        });
        return layerPresence;
    };
    AnimUtils.hasAudio = function (animData) {
        return AnimUtils.hasLayer(animData, Classes_1.LayerName.AudioEvent);
    };
    AnimUtils.hasLayer = function (animData, layerName, strict) {
        if (strict === void 0) { strict = true; }
        var layerCheck = [];
        AnimUtils.forEachLayerType(animData, layerName, function (layer) {
            if (strict) {
                layerCheck.push(layer.keyframes.length !== 0);
            }
            else {
                layerCheck.push(true);
            }
        });
        return layerCheck.some(function (hasContent) { return hasContent; });
    };
    AnimUtils.hasHoldSafe = function (animData, name) {
        var kf = -1;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Event, function (layer) {
            layer.keyframes.forEach(function (layerKF) {
                if (layerKF.value.Event.name === HOLD_SAFE_EVENT_NAME) {
                    if (kf !== -1) {
                        throw new Error("Two HOLD_SAFE keyframes defined in animation" +
                            (" '" + name + "' at frames " + kf + " and " + layerKF.time));
                    }
                    kf = layerKF.time;
                }
            });
        });
        return kf;
    };
    AnimUtils.calculateOrientation = function (animData) {
        var headSum = 0;
        var pelvisSum = 0;
        var torsoSum = 0;
        var numFrames = 0;
        AnimUtils.forEachLayerType(animData, Classes_1.LayerName.Body, function (layer) {
            layer.keyframes.forEach(function (layerKF) {
                numFrames++;
                headSum += layerKF.value.Head;
                pelvisSum += layerKF.value.Pelvis;
                torsoSum += layerKF.value.Torso;
            });
        });
        headSum /= numFrames;
        pelvisSum /= numFrames;
        torsoSum /= numFrames;
        var bodySum = (torsoSum + pelvisSum) / 2;
        if (Math.abs(bodySum) > 10) {
            return (bodySum > 0) ? Classes_1.Orientation.LEFT : Classes_1.Orientation.RIGHT;
        }
        else {
            return (headSum > 0) ? Classes_1.Orientation.LEFT : Classes_1.Orientation.RIGHT;
        }
    };
    AnimUtils.forEachLayerType = function (animData, layerTypeName, handler) {
        var name = layerTypeName.long || layerTypeName.short;
        animData.layers.forEach(function (layer) {
            if (layer.type === name) {
                handler(layer);
            }
        });
    };
    AnimUtils.insertDefaultAnimQueryParams = function (query) {
        if (!query.categories) {
            query.categories = [];
        }
        if (query.category) {
            query.categories.push(query.category);
        }
        for (var i = 0; i < query.categories.length; i++) {
            query.categories[i] = query.categories[i].toLowerCase();
        }
        query.includeCat = query.includeCat || [];
        query.includeSomeCat = query.includeSomeCat || [];
        query.excludeCat = query.excludeCat || [];
        query.includeMeta = query.includeMeta || [];
        query.includeSomeMeta = query.includeSomeMeta || [];
        query.excludeMeta = query.excludeMeta || [];
        var filters = [query.includeCat, query.includeSomeCat, query.excludeCat, query.includeMeta, query.includeSomeMeta, query.excludeMeta];
        filters.forEach(function (array) {
            for (var i = 0; i < array.length; i++) {
                array[i] = array[i].toLowerCase();
            }
        });
        if (!query.durationError) {
            query.durationError = 0;
        }
        if (!query.duration) {
            query.duration = -1;
        }
        if (typeof query.hasAudio === 'undefined') {
            query.hasAudio = Classes_1.HasAudio.DONT_CARE;
        }
    };
    return AnimUtils;
}());
exports.AnimUtils = AnimUtils;

},{"./Classes":7,"./animation_formats/LayerMocks":12}],5:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Range_1 = require("./Range");
var Utils_1 = require("./Utils");
var Playback_1 = require("./Playback");
var Classes_1 = require("./Classes");
var log_1 = require("./log");
var log = log_1.default.createChild("Animation");
var Animation = (function () {
    function Animation(parent, meta, resourceRoot, animData) {
        this.parent = parent;
        this.meta = meta;
        this.resourceRoot = resourceRoot;
        this.animData = animData;
        this.animDataCache = {};
        this.name = meta.name;
        this.durationRange = new Range_1.Range(meta.durationRange[0], meta.durationRange[1]);
    }
    Animation.prototype.createFromConfig = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var transformation, playback;
            return __generator(this, function (_a) {
                log.debug("Configuring Animation for playback with options: ", options);
                transformation = this._getTransformation(options);
                playback = new Playback_1.Playback(transformation, options.cache, this);
                return [2, playback.init(options)];
            });
        });
    };
    Animation.prototype._getTransformation = function (options) {
        var trans = new Classes_1.Transform();
        if (options) {
            if (options.orientation) {
                var reqOrientation = options.orientation;
                if (reqOrientation !== this.meta.orientation) {
                    trans.shouldFlipLeftRight = true;
                }
            }
            if (options.speed) {
                trans.speed = Utils_1.Utils.limitToRange(options.speed, this.meta.speed);
            }
            if (options.exaggerate) {
                trans.exaggerate = Utils_1.Utils.limitToRange(options.exaggerate, this.meta.scale);
            }
            if (options.duration) {
                if (options.duration < this.meta.durationRange[0]) {
                    trans.speed = this.meta.speed[0];
                }
                else if (options.duration < trans.speed * this.meta.duration) {
                    trans.speed = options.duration / this.meta.duration;
                    trans.speed = Utils_1.Utils.limitToRange(trans.speed, this.meta.speed);
                }
                else if (this.meta.holdSafeKF !== -1) {
                    trans.framesToHold = options.duration - trans.speed * this.meta.duration;
                }
                else if (options.duration > trans.speed * this.meta.duration) {
                    if (typeof options.speed !== 'number') {
                        trans.speed = options.duration / this.meta.duration;
                        trans.speed = Utils_1.Utils.limitToRange(trans.speed, this.meta.speed);
                    }
                }
            }
        }
        return trans;
    };
    return Animation;
}());
exports.Animation = Animation;

},{"./Classes":7,"./Playback":8,"./Range":9,"./Utils":10,"./log":15}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Utils_1 = require("./Utils");
var AnimUtils_1 = require("./AnimUtils");
var Classes_1 = require("./Classes");
var Category = (function () {
    function Category(name) {
        this.name = name;
        this.animations = [];
        this.animationsByName = new Map();
    }
    Category.prototype.addAnimation = function (anim) {
        this.animations.push(anim);
        this.animationsByName.set(anim.name, anim);
    };
    Category.prototype.query = function (query) {
        AnimUtils_1.AnimUtils.insertDefaultAnimQueryParams(query);
        var res = new Classes_1.AnimResults();
        this._query(query, res);
        return res;
    };
    Category.prototype._query = function (query, res) {
        Utils_1.Utils.getAnimationsByQuery(this.animations, res, query);
    };
    return Category;
}());
exports.Category = Category;

},{"./AnimUtils":4,"./Classes":7,"./Utils":10}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Transform = (function () {
    function Transform() {
        this.framesToHold = -1;
        this.speed = 1;
        this.exaggerate = 1;
        this.shouldFlipLeftRight = false;
    }
    return Transform;
}());
exports.Transform = Transform;
var Orientation;
(function (Orientation) {
    Orientation.LEFT = "LEFT";
    Orientation.RIGHT = "RIGHT";
})(Orientation = exports.Orientation || (exports.Orientation = {}));
var HasAudio;
(function (HasAudio) {
    HasAudio.YES = "YES";
    HasAudio.NO = "NO";
    HasAudio.DONT_CARE = "DONT_CARE";
})(HasAudio = exports.HasAudio || (exports.HasAudio = {}));
var AnimResults = (function () {
    function AnimResults() {
        this.matching = [];
        this.nonMatching = [];
    }
    return AnimResults;
}());
exports.AnimResults = AnimResults;
var SearchResults = (function () {
    function SearchResults() {
        this.meta = [];
    }
    return SearchResults;
}());
exports.SearchResults = SearchResults;
var LayerName = (function () {
    function LayerName(short, long) {
        this.short = short;
        this.long = long;
    }
    return LayerName;
}());
LayerName.Body = new LayerName('Body');
LayerName.Eye = new LayerName('Eye');
LayerName.EyeTexture = new LayerName('EyeTexture', 'Eye Texture');
LayerName.Overlay = new LayerName('Overlay');
LayerName.OverlayTexture = new LayerName('OverlayTexture', 'Overlay Texture');
LayerName.AudioEvent = new LayerName('AudioEvent', 'Audio Event');
LayerName.Event = new LayerName('Event');
LayerName.Pixi = new LayerName('Pixi');
LayerName.BackgroundTexture = new LayerName('BackgroundTexture', 'Background Texture');
LayerName.LED = new LayerName('LED');
LayerName.Layers = [
    LayerName.Body,
    LayerName.Eye,
    LayerName.EyeTexture,
    LayerName.Overlay,
    LayerName.OverlayTexture,
    LayerName.AudioEvent,
    LayerName.Event,
    LayerName.Pixi,
    LayerName.BackgroundTexture,
    LayerName.LED
];
exports.LayerName = LayerName;
var BuilderLayer;
(function (BuilderLayer) {
    BuilderLayer.BEAT = "beat";
    BuilderLayer.DEFAULT = "default";
    BuilderLayer.POSTURE = "posture";
})(BuilderLayer = exports.BuilderLayer || (exports.BuilderLayer = {}));

},{}],8:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var events = require("events");
var jibo_cai_utils_1 = require("jibo-cai-utils");
var Classes_1 = require("./Classes");
var AnimUtils_1 = require("./AnimUtils");
var jibo_expression_client_1 = require("jibo-expression-client");
var log_1 = require("./log");
var log = log_1.default.createChild("Playback");
var FRAMERATE = 30;
var defaultOptions = {
    ownerInformation: 'AnimDB',
    disableSetFaceAnim: false,
    screenCenterOverride: true
};
var Playback = (function (_super) {
    __extends(Playback, _super);
    function Playback(transform, cacheName, animation) {
        var _this = _super.call(this) || this;
        _this.transform = transform;
        _this.cacheName = cacheName;
        _this.animation = animation;
        _this.holdSafeDuration = 0;
        return _this;
    }
    Playback.prototype.init = function (config) {
        if (config === void 0) { config = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var useData, filePath, flipMult_1, flipAndExagMult_1, dofs_1;
            return __generator(this, function (_a) {
                log.debug("Initializing playback  " + this.animation.name + "  with config: ", config);
                useData = false;
                if (!this.animation.meta.path) {
                    filePath = this.animation.resourceRoot;
                    useData = true;
                }
                else {
                    filePath = path.join(this.animation.resourceRoot, this.animation.meta.path);
                }
                if (this.animation.meta.metaTerms.indexOf(AnimUtils_1.AUDIO_ONLY_META) !== -1) {
                    config.layer = Classes_1.BuilderLayer.BEAT;
                }
                if (this.transform.shouldFlipLeftRight || this.transform.exaggerate) {
                    flipMult_1 = this.transform.shouldFlipLeftRight ? -1 : 1;
                    flipAndExagMult_1 = flipMult_1 * (this.transform.exaggerate || 1);
                    if ((flipMult_1 !== 1) || (flipAndExagMult_1 !== 1)) {
                        dofs_1 = {
                            bottomSection_r: 1,
                            middleSection_r: 1,
                            topSection_r: 1,
                            eyeSubRootBn_t: 1
                        };
                        Object.keys(dofs_1).forEach(function (dof) {
                            if (dof === "eyeSubRootBn_t_2") {
                                dofs_1[dof] *= flipMult_1;
                            }
                            else {
                                dofs_1[dof] *= flipAndExagMult_1;
                            }
                        });
                        config.scale = dofs_1;
                    }
                }
                if (this.transform.framesToHold !== -1) {
                    this.holdSafeDuration = ((this.transform.framesToHold / FRAMERATE) * 1000);
                }
                if (this.transform.speed !== 1) {
                    config.speed = this.transform.speed;
                }
                this.duration = this.animation.meta.duration;
                if (this.transform.speed !== -1) {
                    this.duration *= this.transform.speed;
                }
                if (this.transform.framesToHold !== -1) {
                    this.duration += this.transform.framesToHold;
                }
                this.config = config;
                this.playbackPr = this.loadAnimation(filePath, config, useData, this.cacheName)
                    .then(function (keysAnimation) {
                    _this.keysAnimation = keysAnimation;
                    _this.instance = keysAnimation.instance;
                    log.debug("Initialized playback  " + _this.animation.name);
                    return _this;
                });
                return [2, this.playbackPr];
            });
        });
    };
    Playback.prototype.play = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var jibo, playOptions, state;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        jibo = this.animation.parent.parent.jibo;
                        playOptions = Object.assign({}, defaultOptions, options);
                        this.options = playOptions;
                        return [4, this._play(playOptions)];
                    case 1:
                        state = _a.sent();
                        if (!(state === jibo_expression_client_1.AnimationState.REJECTED)) return [3, 2];
                        log.warn("Playback  " + this.animation.name + "  was rejected! (See DOFArbiter debug for more details)");
                        return [3, 4];
                    case 2:
                        if (!playOptions.forceReturnToIdle) return [3, 4];
                        return [4, jibo.expression.centerRobot({
                                requestor: playOptions.ownerInformation,
                                dofs: jibo.expression.dofs.BODY,
                                centerGlobally: false
                            })];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [2, state];
                }
            });
        });
    };
    Playback.prototype.stop = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.keysAnimation) return [3, 4];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4, this.keysAnimation.instance.stop()];
                    case 2:
                        _a.sent();
                        return [3, 4];
                    case 3:
                        error_1 = _a.sent();
                        log.warn("Could not stop instance that no longer exists: ", error_1);
                        return [3, 4];
                    case 4: return [2];
                }
            });
        });
    };
    Playback.prototype._play = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var jibo, dofsInScreen, dofsInAnim, dofsNotInAnim;
            return __generator(this, function (_a) {
                jibo = this.animation.parent.parent.jibo;
                dofsInScreen = jibo.expression.dofs.SCREEN;
                dofsInAnim = this.keysAnimation.instance.dofs;
                dofsNotInAnim = dofsInScreen.minus(dofsInAnim);
                this.hasScreenDOFs = (dofsNotInAnim.getDOFs().length !== dofsInScreen.getDOFs().length);
                if (this.hasScreenDOFs && !options.disableSetFaceAnim) {
                    jibo.face.eye.animation = this.keysAnimation;
                }
                return [2, new Promise(function (resolve, reject) {
                        var hasBeenStopped = false;
                        var stopHandler = function (state) {
                            hasBeenStopped = true;
                            log.debug("Playback  " + _this.animation.name + "  was stopped with state  " + state);
                            _this._cleanup();
                            resolve(state);
                        };
                        var holdSafeHandler = function () { return __awaiter(_this, void 0, void 0, function () {
                            var error_2;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        _a.trys.push([0, 6, , 7]);
                                        if (!this.keysAnimation) return [3, 3];
                                        return [4, this.keysAnimation.instance.pause()];
                                    case 1:
                                        _a.sent();
                                        return [4, jibo_cai_utils_1.TimeUtils.wait(this.holdSafeDuration)];
                                    case 2:
                                        _a.sent();
                                        _a.label = 3;
                                    case 3:
                                        if (!this.keysAnimation) return [3, 5];
                                        return [4, this.keysAnimation.instance.resume()];
                                    case 4:
                                        _a.sent();
                                        _a.label = 5;
                                    case 5: return [3, 7];
                                    case 6:
                                        error_2 = _a.sent();
                                        log.warn("Error in hold-safe logic for animation  " + this.animation.name + ": ", error_2);
                                        return [3, 7];
                                    case 7: return [2];
                                }
                            });
                        }); };
                        _this.keysAnimation.instance.events.stopped.on(function () { return stopHandler(jibo_expression_client_1.AnimationState.STOPPED); });
                        _this.keysAnimation.instance.events.cancelled.on(function () { return stopHandler(jibo_expression_client_1.AnimationState.CANCELLED); });
                        _this.keysAnimation.instance.events.rejected.on(function () { return stopHandler(jibo_expression_client_1.AnimationState.REJECTED); });
                        if (_this.holdSafeDuration > 0) {
                            _this.keysAnimation.instance.events.holdSafe.on(holdSafeHandler);
                        }
                        if ((options.screenCenterOverride === false) && _this.hasScreenDOFs) {
                            jibo.expression.centerRobot({
                                requestor: options.ownerInformation,
                                centerGlobally: false,
                                dofs: jibo.expression.dofs.SCREEN
                            }).then(function () {
                                if (!hasBeenStopped) {
                                    log.debug("Playback  " + _this.animation.name + "  playing with these options: ", options);
                                    return _this.keysAnimation.instance.play(options.ownerInformation);
                                }
                            }).catch(reject);
                        }
                        else {
                            log.debug("Playback  " + _this.animation.name + "  playing with these options: ", options);
                            _this.keysAnimation.instance.play(options.ownerInformation)
                                .catch(reject);
                        }
                    })];
            });
        });
    };
    Playback.prototype._cleanup = function () {
        if (this.keysAnimation) {
            var jibo = this.animation.parent.parent.jibo;
            if (this.hasScreenDOFs && !this.options.disableSetFaceAnim) {
                jibo.face.eye.animation = null;
            }
            if (this.keysAnimation.isCopy) {
                this.keysAnimation.destroy();
                this.keysAnimation = null;
                this.instance = null;
            }
            if (this.cacheName) {
                if (!this.options.overrideUnload) {
                    jibo.loader.unload(this.aliasInCache);
                }
            }
            else if (!this.options.overrideDestroy && this.keysAnimation) {
                this.keysAnimation.destroy();
            }
        }
        this.keysAnimation = null;
        this.instance = null;
    };
    Playback.prototype.loadKeysData = function (filePath) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var jibo;
            return __generator(this, function (_a) {
                jibo = this.animation.parent.parent.jibo;
                return [2, new Promise(function (resolve, reject) {
                        jibo.loader.load({
                            id: filePath,
                            data: _this.animation.animData,
                            src: filePath,
                            root: _this.animation.resourceRoot,
                            type: 'keys-data',
                            complete: function (error, keysObject) {
                                if (error) {
                                    reject(error);
                                }
                                else {
                                    resolve(keysObject);
                                }
                            }
                        });
                    })];
            });
        });
    };
    Playback.prototype.loadAnimation = function (filePath, config, useData, cache) {
        if (useData === void 0) { useData = false; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var jibo;
            return __generator(this, function (_a) {
                jibo = this.animation.parent.parent.jibo;
                this.aliasInCache = filePath;
                return [2, new Promise(function (resolve, reject) {
                        if (cache && !useData) {
                            var animation = jibo.loader.cached(filePath, cache);
                            if (animation) {
                                animation = animation.getCleanCopy(config);
                                return resolve(animation);
                            }
                        }
                        jibo.loader.load({
                            id: filePath,
                            cache: useData ? false : cache,
                            options: config,
                            data: useData ? _this.animation.animData : null,
                            src: filePath,
                            upload: true,
                            root: _this.animation.resourceRoot,
                            type: 'keys',
                            complete: function (error, keysAnimation) {
                                if (error) {
                                    reject(error);
                                }
                                else {
                                    resolve(keysAnimation);
                                }
                            }
                        });
                    })];
            });
        });
    };
    return Playback;
}(events.EventEmitter));
exports.Playback = Playback;

},{"./AnimUtils":4,"./Classes":7,"./log":15,"events":undefined,"jibo-cai-utils":undefined,"jibo-expression-client":undefined,"path":undefined}],9:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INFINITE_DURATION = 1e6;
var Range = (function () {
    function Range(min, max) {
        this.min = min;
        this.max = max;
        if (min > max) {
            throw new Error("Invalid bounds " + this.toString());
        }
    }
    Range.prototype.toString = function () {
        return "[" + this.min + ", " + this.max + "]";
    };
    Range.prototype.limit = function (num) {
        if (num < this.min) {
            return this.min;
        }
        else if (num > this.max) {
            return this.max;
        }
        else {
            return num;
        }
    };
    Range.prototype.isUpperInfinite = function () {
        return (this.max === exports.INFINITE_DURATION);
    };
    Range.prototype.distanceFrom = function (num) {
        if (num < this.min) {
            return this.min - num;
        }
        else if (num > this.max) {
            return num - this.max;
        }
        else {
            return 0;
        }
    };
    return Range;
}());
exports.Range = Range;

},{}],10:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Classes_1 = require("./Classes");
var SemverUtils = require('semver-utils');
var Utils = (function () {
    function Utils() {
    }
    Utils.parseString = function (input, defaultVal) {
        if (!input || !(input.length)) {
            return defaultVal;
        }
        return input;
    };
    Utils.parseNumber = function (input, defaultVal) {
        if (typeof input === 'number') {
            return input;
        }
        if (!input || !(input.length)) {
            return defaultVal;
        }
        try {
            var num = parseFloat(input);
            return num;
        }
        catch (e) {
            return defaultVal;
        }
    };
    Utils.parseSemVer = function (input) {
        var parse = SemverUtils.parse(input);
        return {
            semver: parse.semver,
            version: parse.version,
            major: parseInt(parse.major),
            minor: parseInt(parse.minor),
            patch: parseInt(parse.patch),
            release: parse.release
        };
    };
    Utils.getAnimationsByQuery = function (animations, interimResults, query) {
        animations.forEach(function (anim) {
            if (Utils.audioMatch(anim, query.hasAudio) &&
                Utils.catMatch(anim, query.includeCat, query.includeSomeCat, query.excludeCat) &&
                Utils.metaMatch(anim, query.includeMeta, query.includeSomeMeta, query.excludeMeta)) {
                Utils.addIfDurationMatch(anim, query.duration, query.durationError, interimResults);
            }
        });
    };
    Utils.audioMatch = function (animation, shouldHaveAudio) {
        return (shouldHaveAudio === Classes_1.HasAudio.DONT_CARE) ||
            (shouldHaveAudio === Classes_1.HasAudio.YES && animation.meta.hasAudio) ||
            (shouldHaveAudio === Classes_1.HasAudio.NO && !animation.meta.hasAudio);
    };
    Utils.catMatch = function (animation, allOfCat, someOfCat, noneOfCat) {
        var categories = animation.meta.categories;
        var allOfMatch = (allOfCat.length > 0) ? allOfCat.every(function (cat) { return categories.indexOf(cat) !== -1; }) : true;
        var someOfMatch = (someOfCat.length > 0) ? someOfCat.some(function (cat) { return categories.indexOf(cat) !== -1; }) : true;
        var noneOfMatch = (noneOfCat.length > 0) ? noneOfCat.every(function (cat) { return categories.indexOf(cat) === -1; }) : true;
        return (allOfMatch && someOfMatch && noneOfMatch);
    };
    Utils.metaMatch = function (animation, allOfMeta, someOfMeta, noneOfMeta) {
        var metas = animation.meta.metaTerms;
        var allOfMatch = (allOfMeta.length > 0) ? allOfMeta.every(function (meta) { return metas.indexOf(meta) !== -1; }) : true;
        var someOfMatch = (someOfMeta.length > 0) ? someOfMeta.some(function (meta) { return metas.indexOf(meta) !== -1; }) : true;
        var noneOfMatch = (noneOfMeta.length > 0) ? noneOfMeta.every(function (meta) { return metas.indexOf(meta) === -1; }) : true;
        return (allOfMatch && someOfMatch && noneOfMatch);
    };
    Utils.addIfDurationMatch = function (animation, duration, durationError, res) {
        if (duration === -1) {
            res.matching.push(animation);
        }
        else {
            var dist = Utils.distanceFromRange(duration, animation.meta.durationRange);
            if (dist === 0) {
                res.matching.push(animation);
            }
            else if (dist <= durationError) {
                res.nonMatching.push(animation);
            }
        }
    };
    Utils.limitToRange = function (num, range) {
        if (num < range[0]) {
            return range[0];
        }
        else if (num > range[1]) {
            return range[1];
        }
        else {
            return num;
        }
    };
    Utils.distanceFromRange = function (num, range) {
        if (num < range[0]) {
            return range[0] - num;
        }
        else if (num > range[1]) {
            return num - range[1];
        }
        else {
            return 0;
        }
    };
    return Utils;
}());
exports.Utils = Utils;

},{"./Classes":7,"semver-utils":undefined}],11:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

},{}],12:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var generateId = require('jibo-keyframes').generateId;
var LayerTemplate = (function () {
    function LayerTemplate() {
    }
    LayerTemplate.generateLayer = function (name, type) {
        var layer = {};
        layer.id = generateId();
        layer.name = name;
        layer.type = type;
        layer.visible = true;
        layer.locked = false;
        layer.keyframes = [];
        return layer;
    };
    return LayerTemplate;
}());
exports.LayerTemplate = LayerTemplate;

},{"jibo-keyframes":undefined}],13:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./LayerMocks"));
var format_0_6_x = require("./0_6_X");
exports.format_0_6_x = format_0_6_x;

},{"./0_6_X":11,"./LayerMocks":12}],14:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var AnimDBParser_1 = require("./AnimDBParser");
exports.AnimDBParser = AnimDBParser_1.AnimDBParser;
var AnimCollection_1 = require("./AnimCollection");
exports.AnimCollection = AnimCollection_1.AnimCollection;
var Animation_1 = require("./Animation");
exports.Animation = Animation_1.Animation;
var AnimDB_1 = require("./AnimDB");
exports.AnimDB = AnimDB_1.AnimDB;
var Playback_1 = require("./Playback");
exports.Playback = Playback_1.Playback;
__export(require("./animation_formats"));
__export(require("./Classes"));
function init(jibo, animDBPath) {
    var moreAnimDBPaths = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        moreAnimDBPaths[_i - 2] = arguments[_i];
    }
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!animDBPath) return [3, 1];
                    exports._animDB = new AnimDB_1.AnimDB(jibo);
                    return [3, 3];
                case 1: return [4, AnimDBParser_1.AnimDBParser.readAnimDB.apply(AnimDBParser_1.AnimDBParser, [jibo, animDBPath].concat(moreAnimDBPaths))];
                case 2:
                    exports._animDB = _a.sent();
                    _a.label = 3;
                case 3: return [2];
            }
        });
    });
}
exports.init = init;
function isInitialized() {
    return !!exports._animDB;
}
exports.isInitialized = isInitialized;
function getAnimByName(animName) {
    return exports._animDB.getAnimByName(animName);
}
exports.getAnimByName = getAnimByName;
function getAnimationNames() {
    return exports._animDB.getAnimationNames();
}
exports.getAnimationNames = getAnimationNames;
function getAnimationCategories() {
    return exports._animDB.getAnimationCategories();
}
exports.getAnimationCategories = getAnimationCategories;
function search(searchTerm) {
    return exports._animDB.search(searchTerm);
}
exports.search = search;
function query(query) {
    return exports._animDB.query(query);
}
exports.query = query;
function push(animCollection) {
    exports._animDB.push(animCollection);
}
exports.push = push;
function pop() {
    return exports._animDB.pop();
}
exports.pop = pop;

},{"./AnimCollection":1,"./AnimDB":2,"./AnimDBParser":3,"./Animation":5,"./Classes":7,"./Playback":8,"./animation_formats":13}],15:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Log = require("jibo-log");
exports.default = new Log('Jibo.AnimDB');

},{"jibo-log":undefined}],16:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./AnimDB"));
__export(require("./Range"));
__export(require("./AnimDBParser"));
__export(require("./AnimUtils"));
__export(require("./AnimCollection"));
__export(require("./Playback"));
__export(require("./Animation"));
__export(require("./Classes"));
__export(require("./AnimUtils"));
__export(require("./Utils"));
var api = require("./api");
exports.api = api;
var animation_formats = require("./animation_formats");
exports.animation_formats = animation_formats;

},{"./AnimCollection":1,"./AnimDB":2,"./AnimDBParser":3,"./AnimUtils":4,"./Animation":5,"./Classes":7,"./Playback":8,"./Range":9,"./Utils":10,"./animation_formats":13,"./api":14}],17:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./main"));

},{"./main":16}]},{},[17])(17)
});
