// jibo-log/lib/jibo-log.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboLog = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs-extra");
var path = require("path");
var events_1 = require("events");
var utils_1 = require("./utils");
var loggingToFiles = {}; // index of all files being logged to
var FileBackend = (function (_super) {
    __extends(FileBackend, _super);
    function FileBackend(filename, directory) {
        var _this = _super.call(this) || this;
        _this._filename = filename;
        _this._directory = directory;
        _this._file = undefined;
        if (!_this._filename) {
            var timestamp = new Date().toISOString();
            if (process.platform === 'win32') {
                // replace all ':'s for '_'s as windows file paths cannot have that character in them
                timestamp = timestamp.replace(/:/g, '_');
            }
            _this._filename = 'skill.' + timestamp + '.log';
        }
        if (!_this._directory && !path.isAbsolute(_this._filename)) {
            // see if we are running on robot
            if (process.platform === 'linux' && process.arch === 'arm' && fs.existsSync('/var/jibo')) {
                _this._directory = '/var/log/jibo';
            }
            else {
                _this._directory = path.join(process.env.HOME || process.env.USERPROFILE, '.jibo', 'logs');
            }
        }
        return _this;
    }
    FileBackend.prototype.init = function (callback) {
        var _this = this;
        this.createDirectory(function () {
            var fullFilename = path.normalize(_this._filename);
            if (!path.isAbsolute(fullFilename)) {
                fullFilename = path.join(_this._directory, fullFilename);
            }
            fullFilename = path.resolve(fullFilename);
            if (!loggingToFiles[fullFilename]) {
                loggingToFiles[fullFilename] = _this;
                console.log('jibo.log: logging to', fullFilename);
                _this._file = fs.createWriteStream(fullFilename, { 'flags': 'a' }); // append
            }
            else {
                console.log('jibo.log: already logging to', fullFilename);
            }
            if (callback) {
                callback();
            }
        });
    };
    FileBackend.prototype.write = function (name, level, args) {
        var timestamp = new Date().toISOString();
        var text = utils_1.FormatHelper(args);
        var message = timestamp + ' '
            + (name ? name + ' ' : '')
            + (level ? level + ' ' : '')
            + text;
        this._file.write(message + '\n');
    };
    FileBackend.prototype.createDirectory = function (callback) {
        var _this = this;
        if (this._directory && !path.isAbsolute(this._filename)) {
            fs.ensureDir(this._directory, function (err) {
                if (err) {
                    console.error('jibo.log: error creating directory', _this._directory, err);
                }
                callback();
            });
        }
        else {
            setImmediate(callback);
        }
    };
    return FileBackend;
}(events_1.EventEmitter));
exports.default = FileBackend;
//how we used to do it:
//import Transform from '../node_modules/minilog/lib/common/transform';
// let pipe = Minilog.pipe(new FormatWithTimestamp()).pipe(file);
// // a plain formatter that adds a timestamp in the first column
// // also expands out stack traces on error objects
// function FormatWithTimestamp() {}
// Transform.mixin(FormatWithTimestamp);
// FormatWithTimestamp.prototype.write = 
//     function(name, level, args) {
//         let timestamp = new Date().toISOString();
//         let text = FormatHelper(args);
//         this.emit('item',
//                   timestamp + ' '
//                   + (name ? name + ' ' : '')
//                   + (level ? level + ' ' : '')
//                   + text
//                   + '\n');
//     };

},{"./utils":6,"events":undefined,"fs-extra":undefined,"path":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var Minilog = require("minilog"); // the one true minilog instance
var utils_1 = require("./utils");
var utils_2 = require("./utils");
var FileBackend_1 = require("./FileBackend");
var SyslogBackend_1 = require("./SyslogBackend");
/**
 * JiboLog class - centralized logging facility
 * @class JiboLog
 * @param {String} [name='[unknown module]'] The log name.
 */
var JiboLog = (function () {
    function JiboLog(name) {
        if (name === void 0) { name = '[unknown module]'; }
        this.name = name;
        this.enabled = true;
        this.handleAll = false; // off by default now
        this.minLevel = utils_2.Level.info;
        this._minilog = Minilog(this._name);
    }
    /**
     * General log message, alias for `debug`.
     * @method JiboLog#log
     * @param {*} ..args Arguments for logging
     * @return {JiboLog} Instance of log for chaining.
     */
    JiboLog.prototype.log = function () {
        return this.debug.apply(this, arguments);
    };
    /**
     * Debug log message, lowest level message.
     * @method JiboLog#debug
     * @param {*} ..args Arguments for logging
     * @return {JiboLog} Instance of log for chaining.
     */
    JiboLog.prototype.debug = function () {
        if (!this._enabled || !this._gt(utils_2.Level.debug)) {
            return this;
        }
        this._minilog.debug.apply(Minilog, arguments);
        return this;
    };
    /**
     * Info log message.
     * @method JiboLog#info
     * @param {*} ..args Arguments for logging
     * @return {JiboLog} Instance of log for chaining.
     */
    JiboLog.prototype.info = function () {
        if (!this._enabled || !this._gt(utils_2.Level.info)) {
            return this;
        }
        this._minilog.info.apply(Minilog, arguments);
        return this;
    };
    /**
     * Warrning log message.
     * @method JiboLog#warn
     * @param {*} ..args Arguments for logging
     * @return {JiboLog} Instance of log for chaining.
     */
    JiboLog.prototype.warn = function () {
        if (!this._enabled || !this._gt(utils_2.Level.warn)) {
            return this;
        }
        this._minilog.warn.apply(Minilog, arguments);
        return this;
    };
    /**
     * Error log message.
     * @method JiboLog#error
     * @param {*} ..args Arguments for logging
     * @return {JiboLog} Instance of log for chaining.
     */
    JiboLog.prototype.error = function () {
        if (!this._enabled || !this._gt(utils_2.Level.error)) {
            return this;
        }
        this._minilog.error.apply(Minilog, arguments);
        return this;
    };
    /**
     * Log the error if it is set.
     * @method JiboLog#iferr
     * @param {Error|String} err
     * @param {*} [...args] Additional arguments to log.
     * @return {JiboLog} Instance of log for chaining.
     */
    JiboLog.prototype.iferr = function (err) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (err) {
            args.push(err); // put the error object on the end
            this._minilog.error.apply(Minilog, args);
        }
        return this;
    };
    Object.defineProperty(JiboLog.prototype, "name", {
        /**
         * Current name of this log.
         * @name JiboLog#name
         * @type {String}
         */
        get: function () {
            return this._name;
        },
        set: function (name) {
            if (name && path.isAbsolute(name)) {
                name = path.basename(name);
            }
            this._name = name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(JiboLog.prototype, "minLevel", {
        get: function () {
            return this._minLevel;
        },
        /**
         * The minimum level to display, either 'debug', 'info', 'warn', or 'error'.
         * @name JiboLog#minLevel
         * @type {String}
         * @default 'info'
         */
        set: function (level) {
            this._minLevel = level;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Log to a file.
     * @method JiboLog#toFile
     * @param {String} [filename] The name of the file to log to.
     * @param {Function} callback Callback when complete.
     */
    JiboLog.prototype.toFile = function (filename, callback) {
        if (process.platform === 'linux' && process.arch === 'arm') {
            if (callback) {
                console.warn("Disabled 'toFile' on robot.");
                callback("Disabled 'toFile' on robot.");
            }
            return;
        }
        // note: we might be logging to multiple files (so can't use this.fileBackend)
        var fileBackend = new FileBackend_1.default(filename, this._defaultDir);
        fileBackend.init(function () {
            Minilog.pipe(fileBackend);
            if (callback) {
                callback();
            }
        });
    };
    Object.defineProperty(JiboLog.prototype, "defaultDir", {
        get: function () {
            return this._defaultDir;
        },
        /**
         * Set the default logging directory.
         * @name JiboLog#defaultDir
         * @type {String}
         */
        set: function (directory) {
            this._defaultDir = directory;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Enable the syslog.
     * @method JiboLog#toSyslog
     */
    JiboLog.prototype.toSyslog = function () {
        if (this._name) {
            JiboLog.syslogBackend.allow(this._name);
        }
    };
    Object.defineProperty(JiboLog.prototype, "enabled", {
        get: function () {
            return this._enabled;
        },
        /**
         * Logs are enabled.
         * @name JiboLog#enabled
         * @type {Boolean}
         * @default true
         */
        set: function (enabled) {
            this._enabled = enabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(JiboLog.prototype, "handleAll", {
        get: function () {
            return this._handleAll;
        },
        /**
         * Handle all uncaught and unhandled exceptions.
         * @name JiboLog#handleAll
         * @type {Boolean}
         * @default false
         */
        set: function (handleAll) {
            if (handleAll) {
                utils_1.handleUncaughtAndUnhandled();
            }
            else {
                utils_1.stopHandlingUncaughtAndUnhandled();
            }
            this._handleAll = handleAll;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Check to see if the current level is greater than the minLevel.
     * @method JiboLog#_gt
     * @private
     * @return {Boolean} If this level is save to log
     */
    JiboLog.prototype._gt = function (level) {
        var levels = [
            utils_2.Level.debug,
            utils_2.Level.info,
            utils_2.Level.warn,
            utils_2.Level.error
        ];
        return levels.indexOf(level) >= levels.indexOf(this._minLevel);
    };
    return JiboLog;
}());
/**
 * Current version of the library.
 * @name JiboLog.version
 * @type String
 */
JiboLog.version = '2.4.10';
// Minilog.enable();  // this is what you would normally do, it sets up the defaults
// instead we are going to tweak things a bit
// reference http://mixu.net/minilog/filter.html
if (Minilog.defaultFormatter) {
    Minilog
        .pipe(Minilog.suggest) // suppress messages < 'info' (no debug)
        .pipe(Minilog.defaultFormatter)
        .pipe(Minilog.defaultBackend);
}
else {
    Minilog
        .pipe(Minilog.suggest) // suppress messages < 'info' (no debug)
        .pipe(Minilog.defaultBackend);
}
if (!JiboLog.syslogBackend) {
    // create the syslogBackend singleton
    JiboLog.syslogBackend = new SyslogBackend_1.default();
    // always say we will try to pipe to syslogBackend; 
    // the filter is done in the write method
    Minilog.pipe(JiboLog.syslogBackend);
}
exports.default = JiboLog;

},{"./FileBackend":1,"./SyslogBackend":4,"./utils":6,"minilog":undefined,"path":undefined}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
global.JiboSingletons = global.JiboSingletons || {};
var JiboSingletons = (function () {
    function JiboSingletons() {
    }
    JiboSingletons.enforceSingleton = function (candidate) {
        var singleton = candidate;
        var singletonName = candidate.constructor.name;
        if (!global.JiboSingletons[singletonName]) {
            global.JiboSingletons[singletonName] = singleton;
        }
        else {
            singleton = global.JiboSingletons[singletonName];
            if (singletonName === 'Function') {
                singletonName = 'jibo-log';
            }
            // let's only warn if the versions don't match
            // console.warn('JiboSingletons: using existing Jibo singleton', singletonName);
            // optionally check if versions match (and warn if not)
            if (candidate.version) {
                var candidateVersion = candidate.version;
                var singletonVersion = singleton.version;
                if (typeof candidate.version === 'function') {
                    candidateVersion = candidate.version();
                }
                if (typeof singleton.version === 'function') {
                    singletonVersion = singleton.version();
                }
                if (candidateVersion !== singletonVersion) {
                    console.log(singletonName, 'singleton: existing version', singletonVersion, 'does not equal candidate version', candidateVersion);
                }
            }
        }
        return singleton;
    };
    return JiboSingletons;
}());
exports.default = JiboSingletons;

},{}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var syslog = require("syslog-client");
var utils_1 = require("./utils");
var syslogInfo = {
    port: 514,
    target: "127.0.0.1"
};
var SyslogBackend = (function (_super) {
    __extends(SyslogBackend, _super);
    function SyslogBackend() {
        var _this = _super.call(this) || this;
        _this._isWriting = false;
        _this._writeQueue = [];
        _this._client = undefined;
        _this._enabled = false;
        _this._test = []; // map of what "names" can go to syslog
        if (process.platform === 'linux' && process.arch === 'arm') {
            _this._enabled = true;
        }
        else {
            _this._enabled = false; // don't bother on anything other than our linux/arm environment
        }
        // if we are turning on syslogging and there isn't a client, then start the client
        if (_this._enabled && !_this._client) {
            var options = {
                port: syslogInfo.port,
                transport: syslog.Transport.Udp
            };
            _this._client = syslog.createClient(syslogInfo.target, options);
            console.warn("Initializing a syslog client");
            _this._client.on("error", function (error) {
                console.error(error);
                if (this._client) {
                    this._client.close();
                    this._client = undefined;
                }
            });
        }
        // if we are turning off syslogging and there currently is a client, then close the client
        if (!_this._enabled && _this._client) {
            _this._client.close();
            _this._client = undefined;
        }
        return _this;
    }
    // allow logging with this name to go to syslog
    SyslogBackend.prototype.allow = function (name) {
        if (name) {
            this._test.push(name);
        }
    };
    SyslogBackend.prototype.processNextInQueue = function () {
        var _this = this;
        if (this._writeQueue.length === 0) {
            this._isWriting = false;
            return;
        }
        var _a = (this._writeQueue.shift()), name = _a.name, level = _a.level, args = _a.args;
        var type = undefined;
        switch (level) {
            case 'log':
            case 'debug': {
                type = "Debug";
                break;
            }
            case 'warn': {
                type = "Warning";
                break;
            }
            case 'error': {
                type = "Error";
                break;
            }
            case 'info':
            default: {
                type = "Informational";
                break;
            }
        }
        var logOptions = {
            facility: syslog.Facility.Local0,
            severity: syslog.Severity[type]
        };
        var msg = utils_1.FormatHelper(args);
        this._isWriting = true;
        this._client.log(name + ": [" + level + "] " + msg, logOptions, function (error) {
            if (error) {
                console.error(error);
            }
            // SDK-529 this .close() was apparently working around a bug in syslog-client,
            // which has now been fixed by this PR:
            // https://github.com/paulgrove/node-syslog-client/pull/7/files
            // which got merged into syslog-client starting in v1.0.10
            // the queueing logic can probably go away too, but i'm leaving it for now -jon
            //this._client.close();
            _this.processNextInQueue();
        });
    };
    SyslogBackend.prototype.write = function (name, level, args) {
        // only write to syslog if it's enabled, has a client and the name exists in global mapping
        if (this._enabled && this._client && (this._test.indexOf(name) >= 0)) {
            this._writeQueue.push({ name: name, level: level, args: args });
            if (!this._isWriting) {
                this.processNextInQueue();
            }
        }
    };
    return SyslogBackend;
}(events_1.EventEmitter));
exports.default = SyslogBackend;

},{"./utils":6,"events":undefined,"syslog-client":undefined}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var JiboLog_1 = require("./JiboLog");
var utils_1 = require("./utils");
function warn(msg) {
    var stack = new Error().stack;
    stack = stack.split('\n').splice(3).join('\n');
    if (console.groupCollapsed) {
        console.groupCollapsed("%cDeprecation Warning: %c%s", "color:#614108;background:#fffbe6", "font-weight:normal;color:#614108;background:#fffbe6", msg);
        console.warn(stack);
        console.groupEnd();
    }
    else {
        console.warn("Deprecation Warning:", msg);
        console.warn(stack);
    }
}
var JiboLogPrototype = JiboLog_1.default.prototype;
/**
 * Allow handling uncaught and unhandled exceptions.
 * @method JiboLog#enableUncaughtAndUnhandled
 * @deprecated since 1.1.0
 * @see JiboLog#handleAll
 */
JiboLogPrototype.enableUncaughtAndUnhandled = function () {
    warn('enableUncaughtAndUnhandled is deprecated, please use handleAll, e.g., log.handleAll = true');
    this.handleAll = true;
};
/**
 * Allow handling uncaught and unhandled exceptions.
 * @method JiboLog#disableUncaughtAndUnhandled
 * @deprecated since 1.1.0
 * @see JiboLog#handleAll
 */
JiboLogPrototype.disableUncaughtAndUnhandled = function () {
    warn('disableUncaughtAndUnhandled is deprecated, please use handleAll, e.g., log.handleAll = false');
    this.handleAll = false;
};
/**
 * Set the default logging directory.
 * @method JiboLog#setDefaultLogDir
 * @deprecated since 1.1.0
 * @see JiboLog#defaultDir
 */
JiboLogPrototype.setDefaultLogDir = function (dir, callback) {
    warn('setDefaultLogDir is deprecated, please use setDefaultDir, e.g., log.setDefaultDir(dir, callback)');
    this.defaultDir = dir;
    if (callback) {
        setImmediate(callback);
    }
};
/**
 * Allow for debug messages.
 * @method JiboLog#enableDebug
 * @deprecated since 1.1.0
 * @see JiboLog#minLevel
 */
JiboLogPrototype.enableDebug = function (name, level) {
    if (level === void 0) { level = utils_1.Level.debug; }
    warn('enableDebug is deprecated, please use allow, e.g., log.minLevel = \'debug\'');
    this.minLevel = level;
};

},{"./JiboLog":2,"./utils":6}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var formatError = require("format-error");
//import stackTrace from 'stack-trace';
var util = require("util");
var Minilog = require("minilog");
var Level;
(function (Level) {
    Level[Level["debug"] = 'debug'] = "debug";
    Level[Level["info"] = 'info'] = "info";
    Level[Level["warn"] = 'warn'] = "warn";
    Level[Level["error"] = 'error'] = "error";
})(Level = exports.Level || (exports.Level = {}));
;
function FormatHelper(args) {
    var formatObj = function (obj) {
        var text = util.inspect(obj);
        if (obj && obj instanceof Error && obj.stack) {
            //text += '\n' + util.inspect(stackTrace.parse(obj));
            text += '\n' + formatError.format(obj);
        }
        return text;
    };
    var text = args.map(function (item) {
        return (typeof item === 'string' ? item : formatObj(item));
    }).join(' ').trim();
    return text;
}
exports.FormatHelper = FormatHelper;
function logUncaught(minilog, err) {
    minilog.error('Uncaught', err);
    var msg = FormatHelper([err]); // hopefully add useful stack traces to uncaught exceptions
    minilog.error(msg);
    // // flush or close logging files? FIXME
    // Minilog.end()  // or:
    // Object.keys(loggingToFiles).forEach( (filename) => {
    //     let pipe = loggingToFiles[filename];
    //     //Minilog.unpipe(pipe);  // hoping unhooking pipe will flush file (nope!)
    //     //Minilog.pipe(pipe);  // hoping i can then reattach pipe
    // });
    throw err; // rethrow the error, which amazingly doesn't cause a loop
}
function logUnhandled(minilog, reason, p) {
    minilog.log('Unhandled Rejection at: Promise ', p, ' reason: ', reason);
    if (reason instanceof Error) {
        minilog.log(reason.stack);
    }
    // we should probably do all the process exit logic that logUncaughtException does FIXME
}
var handlersInstalledOnce = false;
var uncaughtCallback;
var unhandledCallback;
function handleUncaughtAndUnhandled() {
    if (!handlersInstalledOnce && (typeof process === 'object')) {
        var minilog_1 = Minilog('[uncaught]');
        // log uncaughtExceptions
        uncaughtCallback = function (err) { logUncaught(minilog_1, err); };
        unhandledCallback = function (reason, p) { logUnhandled(minilog_1, reason, p); };
        process.on('uncaughtException', uncaughtCallback);
        // log any unhandled, otherwise silent, promise exceptions too
        process.on('unhandledRejection', unhandledCallback);
        handlersInstalledOnce = true;
    }
}
exports.handleUncaughtAndUnhandled = handleUncaughtAndUnhandled;
function stopHandlingUncaughtAndUnhandled() {
    if (handlersInstalledOnce) {
        process.removeListener('uncaughtException', uncaughtCallback);
        process.removeListener('unhandledRejection', unhandledCallback);
        handlersInstalledOnce = false;
    }
}
exports.stopHandlingUncaughtAndUnhandled = stopHandlingUncaughtAndUnhandled;

},{"format-error":undefined,"minilog":undefined,"util":undefined}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var JiboLog_1 = require("./JiboLog");
require("./deprecation");
/**
 * Centralized logging facility.
 * @module jibo-log
 * @see JiboLog
 * @example
 * import * as JiboLog from 'jibo-log';
 * const log = new JiboLog(module.filename);
 * log.info('hello!');
 */
// we want to insure there is only ever one instance of the JiboLog module
// because of the file scope variables and especially the Minilog instance
var JiboSingletons_1 = require("./JiboSingletons");
exports.default = JiboSingletons_1.default.enforceSingleton(JiboLog_1.default);

},{"./JiboLog":2,"./JiboSingletons":3,"./deprecation":5}]},{},[7])(7)
});
