// jibo-dev/lib/jibo-dev.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboDev = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
/**
 * @fileOverview
 * Asset Nodes extend the Node class to include information necessary to play animations and sounds
 * alongside TTS
 *
 * Created on 5/11/16.
 * @author Tom Donahue <tom.donahue@jibo.com>
 *
 * -- E. Woudenberg trimmed this down for use in jibo-dev by removing the Jibo reference (which would be circular).
 */
Object.defineProperty(exports, "__esModule", { value: true });
const jibo_node_xml_1 = require("jibo-node-xml");
/**
 * @description Enum of types of Asset nodes
 */
var AssetNodeType;
(function (AssetNodeType) {
    AssetNodeType["ES"] = "ES";
    AssetNodeType["ANIM"] = "ANIM";
    AssetNodeType["SSA"] = "SSA";
    AssetNodeType["SFX"] = "SFX";
    AssetNodeType["AUDIO"] = "AUDIO";
})(AssetNodeType = exports.AssetNodeType || (exports.AssetNodeType = {}));
/**
 * @description Enum of types of Asset requests
 */
var RequestType;
(function (RequestType) {
    RequestType["NAME"] = "NAME";
    RequestType["PATH"] = "PATH";
    RequestType["QUERY"] = "QUERY";
    RequestType["UNKNOWN"] = "UNKNOWN";
})(RequestType = exports.RequestType || (exports.RequestType = {}));
exports.MetaLayerMap = {
    body: 'Body',
    screen: 'Screen',
    audio: 'Audio'
};
const ESAliasToFilter = {
    ANIM: '!ssa-only, !sfx-only',
    SSA: 'ssa-only',
    SFX: 'sfx-only'
};
const ALL_LAYERS = 'all';
class AssetNode extends jibo_node_xml_1.default {
    /**
     * Takes Node (XML) attributes and creates a RequestInfo object from them
     *
     * @param {Map<string, AttType>} att - Map of attributes to their data
     * @returns {RequestInfo}
     */
    static getRequestInfo(att) {
        let info = {
            type: RequestType.UNKNOWN,
            description: RequestType.UNKNOWN
        };
        if (att.has('name')) {
            info.type = RequestType.NAME;
            info.description = `${info.type}[ ${att.get('name')} ]`;
        }
        else if (att.has('cat') || att.has('cats') || att.has('filter') || att.has('filters') || att.has('meta') || att.has('metas')) {
            info.type = RequestType.QUERY;
            const cat = `CAT[ ${att.get('cat') || att.get('cats')} ]`;
            const meta = `META[ ${att.get('filter') || att.get('filters') || att.get('meta') || att.get('metas')} ]`;
            info.description = `${info.type}[ ${cat} ${meta} ]`;
        }
        else if (att.has('path')) {
            info.type = RequestType.PATH;
            info.description = `${info.type}[ ${att.get('path')} ]`;
        }
        return info;
    }
    /**
     * Takes a comma separated string of filtering categories and modifies the provided
     * categories list with the categories we want filtered at runtime
     *
     * @param {AttType} requestedCategoriesIntent - comma separated string of requested Layers
     * @param {AttrFilter} categoryFilter - categories we want to be filtered at runtime (if any)
     */
    static addCategories(reqestedCategoriesIntent, categoryFilter) {
        AssetNode.addFilterAttributes(reqestedCategoriesIntent, categoryFilter);
    }
    /**
     * Takes a comma separated string of filtering attributes and modifies the provided
     * MetaFilter object with the list of attributes we want filtered at runtime (may contain dupes)
     *
     * @param {string} filterAttr - comma separated string
     * @param {MetaFilter} metaFilter - meta terms we want filtered at runtime (if any)
     */
    static addFilterAttributes(filterAttr, metaFilter) {
        if (filterAttr) {
            // Split on commas or semi-colons, as long as they aren't inside parens
            const terms = filterAttr.split(/[;,](?![^(]*\))/g);
            terms.forEach(term => {
                const [op, attrs] = AssetNode.parseTerm(term);
                if (attrs.length > 0) {
                    if (!op || op === '&') {
                        metaFilter.allOf.push(...attrs);
                    }
                    else if (op === '?') {
                        metaFilter.someOf.push(...attrs);
                    }
                    else if (op === '!') {
                        metaFilter.noneOf.push(...attrs);
                    }
                    else if (op === '+') {
                        metaFilter.addAll.push(...attrs);
                    }
                }
            });
        }
    }
    /**
     * Parses a term within a query and returns the constituent parts
     *
     * @param {string} term - Individual part of a larger query
     * @returns {[string, string[]]} - Parsed term containing OP and ATTRS
     */
    static parseTerm(term) {
        const trimmed = term.trim().toLowerCase();
        let op;
        let attrs = [];
        // Check if term is of the form "OP(1, 2, 3)"
        const groupedMatches = /^([!\?\+&]?)\(([\w\s-,;]+)\)/.exec(trimmed);
        if (groupedMatches) {
            op = groupedMatches[1]; // Pull out the matched operator (&, ?, !, +)
            const subterms = groupedMatches[2].split(','); // Pull out the term list
            attrs.push(...subterms.map(subterm => subterm.trim()));
        }
        else {
            // Otherwise check if term is of the form "OP1"
            const indMatches = /^([!\?\+&]?)([\w-]+)/.exec(trimmed);
            if (indMatches) {
                op = indMatches[1]; // Pull out the matched operator (&, ?, !, +)
                attrs.push(indMatches[2]); // Pull out the term
            }
            else {
                console.warn('ESML Asset Query contains un-parseable elements. No or unexpected results likely.');
            }
        }
        return [op, attrs];
    }
    /**
     * Takes a comma separated string of filtering requested meta terms and modifies the provided
     * MetaFilter object with the unified set of meta we want filtered at runtime
     *
     * @param {string} reqFiltered - comma separated string of requested Meta
     * @param {MetaFilter} metaFilter - meta terms we want filtered at runtime (if any)
     */
    static addMeta(reqFiltered, metaFilter) {
        AssetNode.addFilterAttributes(reqFiltered, metaFilter);
        if (metaFilter.allOf.length > 0) {
            metaFilter.allOf = Array.from(new Set(metaFilter.allOf));
        }
        if (metaFilter.someOf.length > 0) {
            metaFilter.someOf = Array.from(new Set(metaFilter.someOf));
        }
        if (metaFilter.noneOf.length > 0) {
            metaFilter.noneOf = Array.from(new Set(metaFilter.noneOf));
        }
    }
    /**
     * Takes a comma separated string of filtering requested layers terms and modifies the provided
     * two MetaLayerPresence objects describing which sets of layers we will add/subtract at runtime
     *
     * @param {AttType} requestedLayersIntent - comma separated string of requested Layers
     * @param {MetaLayerPresence} layers - layers we want to be filtered at runtime (if any)
     * @param {MetaLayerPresence} layersToAdd - layers we want to be added at runtime (if any)
     */
    static addLayers(reqestedLayersIntent, layers, layersToAdd) {
        let layerFilter = {
            allOf: [],
            someOf: [],
            noneOf: [],
            addAll: []
        };
        AssetNode.addFilterAttributes(reqestedLayersIntent, layerFilter);
        if (layerFilter.addAll.length > 0) {
            if (layerFilter.allOf.length > 0 || layerFilter.noneOf.length > 0) {
                throw new Error('Cannot both add layers and filter inclusively or exclusively at the same time.' +
                    ' Modify layers filter to be one approach or the other.');
            }
            if (layerFilter.addAll.length === 1 && layerFilter.addAll[0] === ALL_LAYERS) {
                layers.Body = true;
                layers.Screen = true;
                layers.Audio = true;
                layersToAdd.Body = true;
                layersToAdd.Screen = true;
                layersToAdd.Audio = true;
            }
            else {
                layerFilter.addAll.forEach(layer => {
                    const metaLayerName = exports.MetaLayerMap[layer];
                    if (metaLayerName === undefined) {
                        throw new Error(`Layer ${layer} is not a valid layers filter.`);
                    }
                    else {
                        layers[metaLayerName] = true;
                        layersToAdd[metaLayerName] = true;
                    }
                });
            }
        }
        else {
            if (layerFilter.allOf.length > 0 && layerFilter.noneOf.length > 0) {
                throw new Error('Cannot filter layers in both an inclusive and exclusive manner at the same time.' +
                    ' Modify layers filter to be one or the other.');
            }
            if (layerFilter.allOf.length > 0) {
                layers.Body = false;
                layers.Screen = false;
                layers.Audio = false;
                layerFilter.allOf.forEach(layer => {
                    const metaLayerName = exports.MetaLayerMap[layer];
                    if (metaLayerName === undefined) {
                        throw new Error(`Layer ${layer} is not a valid layers filter.`);
                    }
                    else {
                        layers[metaLayerName] = true;
                    }
                });
            }
            else {
                layerFilter.noneOf.forEach(layer => {
                    const metaLayerName = exports.MetaLayerMap[layer];
                    if (metaLayerName === undefined) {
                        throw new Error(`Layer ${layer} is not a valid layers filter.`);
                    }
                    else {
                        layers[metaLayerName] = false;
                    }
                });
            }
        }
    }
    constructor(type, att) {
        super(type);
        let defaultFilterIntent;
        if (type !== AssetNodeType.AUDIO) {
            this.type = AssetNodeType.ES;
            defaultFilterIntent = ESAliasToFilter[type];
            this.metaFilter = {
                allOf: [],
                someOf: [],
                noneOf: [],
                addAll: []
            };
            this.categoryFilter = {
                allOf: [],
                someOf: [],
                noneOf: [],
                addAll: []
            };
            AssetNode.addFilterAttributes(defaultFilterIntent, this.metaFilter);
        }
        if (att) {
            this.att = att;
            this.requestInfo = AssetNode.getRequestInfo(att);
            // Handle any categories
            const requestedCategoryIntent = this.att.get('cat') || this.att.get('cats');
            if (requestedCategoryIntent) {
                AssetNode.addCategories(requestedCategoryIntent, this.categoryFilter);
            }
            // Handle any metas
            const requestedFilterIntent = this.att.get('filter') || this.att.get('filters') || this.att.get('meta') || this.att.get('metas');
            if (requestedFilterIntent) {
                AssetNode.addMeta(requestedFilterIntent, this.metaFilter);
            }
            // Handle any layers
            this.layers = {};
            this.layersToAdd = {};
            const reqLayers = this.att.get('layers') || this.att.get('layer');
            if (reqLayers) {
                AssetNode.addLayers(reqLayers, this.layers, this.layersToAdd);
            }
        }
    }
}
exports.AssetNode = AssetNode;

},{"jibo-node-xml":undefined}],2:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const cli_1 = require("./cli");
const vm = require("vm");
const events_1 = require("events");
const glob = require("glob");
const jibo_node_xml_1 = require("jibo-node-xml");
const jibo_anim_db_1 = require("jibo-anim-db");
const jibo_flow_core_1 = require("jibo-flow-core");
const request = require("request");
const AssetNode_1 = require("./AssetNode");
const jibo_data_utils_1 = require("jibo-data-utils");
const resolve = require('resolve');
class AnimRefChecker extends events_1.EventEmitter {
    constructor(args) {
        super();
        this.args = args;
        this.refs = [];
        this.errors = [];
        this.animDBPresent = false;
        this.noManifest = args['no-manifest'];
        this.noReport = args['no-report'];
        this.manifest = args.manifest;
        this.skillDir = args.skill;
        this.animdb = args.animdb;
        this.verbose = args.verbose;
    }
    /**
     * Command-line tool for validating a skill against the Animation DB.
     * @method AnimRefChecker.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    /**
     * Command-line tool entrypoint for validating the AnimDB references in a skill.
     * @method AnimRefChecker.exec
     * @param {any} args Options for controlling the checker.
     * @return {EventEmitter} Emits 'lintError', 'end' and 'log' event
     */
    static exec(args) {
        const refChecker = new AnimRefChecker(args);
        refChecker.run();
        return refChecker;
    }
    /**
     * Main entry point for execution. This loads the required files and proceeds through all the checks.
     * @method AnimRefChecker.run
     */
    run() {
        // Init animation DB
        this.init().then(() => {
            this.animDBPresent = true;
        }).catch((err) => {
            if (err.path && err.path.endsWith('node_modules/jibo-anim-db-animations/animdb.json')) {
                console.log('Skill has no animation database -- animDB references will not be checked.');
            }
            else {
                console.log('Failed to initialize animation database -- animDB references will not be checked', err);
            }
        }).then(() => {
            this.doChecks();
        });
    }
    doChecks() {
        if (this.manifest) {
            this.loadRefsFromManifest(() => this.runChecks());
        }
        else {
            this.loadRefsFromSkill(() => this.runChecks());
        }
    }
    /**
     * Write discovered errors to the given file.
     * @method AnimRefChecker#writeReportToFile
     * @param {string} filename
     */
    writeReportToFile(filename, title) {
        const report = this.getErrorReport();
        fs.writeFileSync(filename, `AnimDB Reference Report for ${title}\n` + report);
    }
    /**
     * Format all the discovered errors into a report.
     * @method AnimRefChecker#getErrorReport
     * @returns {string} Report of all discovered errors.
     */
    getErrorReport() {
        let report = '';
        let location = '';
        // Group all the errors in the same file/activity together
        this.errors.sort((a, b) => a.location.localeCompare(b.location));
        for (const error of this.errors) {
            if (error.location !== location) {
                report += `\nErrors found in '${error.location}':\n`;
                location = error.location;
            }
            report += `${error}\n`;
        }
        return report;
    }
    // Convert possible asset-pack style reference into full path
    resolveAssetPath(src, middle) {
        let parts = src.split('://');
        middle = middle || '';
        if (parts.length > 1) {
            let dir = this.resolve(parts[0]);
            return path.resolve(dir, middle, parts[1]);
        }
        else {
            return path.resolve(this.skillDir, middle, src);
        }
    }
    runChecks() {
        if (!this.noReport) {
            this.checkRefs();
        }
        if (this.skillDir && !this.noManifest) {
            this.createManifest();
        }
        const report = this.getErrorReport();
        if (report) {
            this.emit('lintError', report);
        }
        this.emit('end');
    }
    resolve(wanted) {
        let dir;
        try {
            dir = resolve.sync(wanted, { basedir: this.skillDir });
            dir = path.dirname(dir);
        }
        catch (e) {
            console.log(`Could not resolve module named ${wanted}`);
        }
        return dir;
    }
    // Load the optional animdb if provided, otherwise the skill's installed animation DB.
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            let animDBpath = this.animdb;
            if (!animDBpath) {
                let adaPath = this.resolve('jibo-anim-db-animations');
                animDBpath = adaPath && path.resolve(adaPath, 'animdb.json');
            }
            return jibo_anim_db_1.api.init(null, animDBpath);
        });
    }
    // Create Maps (needed by AssetNode) from the attr object stored in the JSON file.
    fixupManifest(refs) {
        for (let ref of refs) {
            ref.att = new Map();
            for (let key in ref.attr || {}) {
                ref.att.set(key, ref.attr[key]);
            }
        }
    }
    loadRefsFromManifest(cb) {
        if (this.manifest.startsWith('http://') || this.manifest.startsWith('https://')) {
            const url4raw = this.manifest; // .replace('/blob/', '/raw/'); -- won't work unless you supply the user's credentials
            request(url4raw, (error, response, body) => {
                if (error) {
                    return this.emit('end', `Failure loading HTML file ${url4raw}`);
                }
                try {
                    this.refs = JSON.parse(body);
                    this.fixupManifest(this.refs);
                }
                catch (e) {
                    return this.emit('end', `Could not parse contents of ${url4raw} -- did you give a URL for the raw file?`);
                }
                cb();
            });
        }
        else {
            fs.readFile(this.manifest, 'utf8', (error, data) => {
                if (error) {
                    return this.emit('end', `${error}`);
                }
                this.refs = JSON.parse(data);
                this.fixupManifest(this.refs);
                cb();
            });
        }
    }
    loadRefsFromSkill(cb) {
        for (const checker of [
            new FlowChecker(this, this.skillDir),
            new BtChecker(this, this.skillDir),
            new MimChecker(this, this.skillDir)
        ]) {
            checker.extractRefsFromFiles();
            this.refs = this.refs.concat(checker.getReferences());
        }
        cb();
    }
    checkRefs() {
        for (const ref of this.refs) {
            this.checkRef(ref);
        }
    }
    // Check a ref against the DB
    checkRef(ref) {
        let fail = false;
        switch (ref.type) {
            // Handle Markup-based references
            case 'SSA':
            case 'SFX':
            case 'ES':
            case 'ANIM':
                if (!this.animDBPresent) {
                    break;
                }
                if (ref.attr.name) {
                    const anim = jibo_anim_db_1.api.getAnimByName(ref.attr.name);
                    if (!anim) {
                        this.addError(ref.location, `Animation DB found no animation for TTS markup <${ref.type} name="${ref.attr.name}"/>`, 'BAD-ANIM-NAME');
                        fail = true;
                    }
                }
                else {
                    const animQuery = formQuery(ref);
                    const result = jibo_anim_db_1.api.query(animQuery);
                    if (result.matching.length === 0) {
                        this.addError(ref.location, `Animation DB found no matches for TTS markup <${ref.type} ${this.xmlify(ref.attr)}/> query: ${JSON.stringify(animQuery)}`, 'ANIM-QUERY-FAIL');
                        fail = true;
                    }
                }
                break;
            case 'animRefByName':
                if (!this.animDBPresent) {
                    break;
                }
                const anim = jibo_anim_db_1.api.getAnimByName(ref.name);
                if (!anim) {
                    this.addError(ref.location, `Animation name "${ref.name}" not found in animation DB`, 'BAD-ANIM-NAME');
                    fail = true;
                }
                break;
            case 'animRefByQuery':
                if (!this.animDBPresent) {
                    break;
                }
                const querytmp = JSON.parse(JSON.stringify(ref.params));
                const result = jibo_anim_db_1.api.query(querytmp);
                if (result.matching.length === 0) {
                    this.addError(ref.location, `Animation DB found no matches for query params "${JSON.stringify(ref.params)}.`, 'ANIM-QUERY-FAIL');
                    fail = true;
                }
                break;
            case 'animRefByPath':
                let keysfile = this.resolveAssetPath(ref.path, 'animations');
                if (!fs.existsSync(keysfile)) {
                    this.addError(ref.location, `No animation found at ${keysfile}`, 'BAD-ANIM-PATH');
                    fail = true;
                }
                break;
        }
        return fail;
    }
    xmlify(attr) {
        let xattr = '';
        for (const key in attr) {
            xattr += `${key}="${attr[key]}" `;
        }
        return xattr;
    }
    createManifest() {
        const filename = path.join(this.skillDir, 'animdb-manifest.json');
        fs.writeFileSync(filename, JSON.stringify(this.refs, null, '  '), 'utf8');
    }
    /**
     * Accumulate an error referring to an activity.
     * @method AnimRefChecker#addProcedureError
     * @private
     * @param {Activity} activity
     * @param {string} message
     * @param {string} code
     */
    addError(location, message, code) {
        this.errors.push(new Error(location, message, code));
    }
}
exports.AnimRefChecker = AnimRefChecker;
var animSelector;
(function (animSelector) {
    animSelector[animSelector["path"] = 1] = "path";
    animSelector[animSelector["pathFunction"] = 2] = "pathFunction";
    animSelector[animSelector["name"] = 3] = "name";
    animSelector[animSelector["query"] = 4] = "query";
})(animSelector || (animSelector = {}));
class BehaviorChecker {
    constructor(refChecker, dir) {
        this.refChecker = refChecker;
        this.dir = path.resolve(dir);
        this.errors = [];
        this.refs = [];
        this.verbose = refChecker.verbose;
    }
    getReferences() {
        return this.refs;
    }
    extractRefsFromOptions(uri, activities) {
        for (const id in activities) {
            const activity = activities[id];
            // TODO: location to be used (later) in preflight to allow the user to navigate directly to the error.
            let localPath = uri.substr(this.dir.length + 1);
            const location = `${localPath} - ${id}`;
            switch (activity.class) {
                case 'PlayAnimation':
                    switch (activity.options && activity.options.animSelector) {
                        case animSelector.name:
                            this.refs.push({ type: 'animRefByName', location, name: activity.options.animName });
                            break;
                        case animSelector.query:
                            this.refs.push({
                                type: 'animRefByQuery',
                                location,
                                params: this.getQueryParams(activity.options.queryParams)
                            });
                            break;
                        case animSelector.path:
                            this.refs.push({
                                type: 'animRefByPath',
                                location,
                                path: activity.options.animPath
                            });
                    }
                    break;
                case 'TextToSpeech':
                    const refs = getAnimReferencesFromXml(this.refChecker, location, activity.options.words);
                    this.refs = this.refs.concat(refs);
                    break;
            }
        }
    }
    // Parse the queryParams entry:
    // "queryParams": "() => ({
    //         category: 'emoji',
    //         categories: ['happy']
    //         includeMeta: ['sleep'],
    //         duration: 3,
    //         hasAudio: 'NO'
    //     });"
    getQueryParams(queryParams) {
        const ref = {};
        for (const line of queryParams.split('\n')) {
            const parts = line.split(':');
            if (parts.length === 2) {
                const key = parts[0].trim();
                let value = parts[1].trim();
                if (value.slice(-1) === ',') {
                    value = value.slice(0, -1);
                }
                if (key === 'category' || key === 'hasAudio') {
                    ref[key] = value.slice(1, -1); // handle quoted string types
                }
                if (key === 'includeMeta' ||
                    key === 'excludeMeta' ||
                    key === 'includeSomeMeta' ||
                    key === 'categories') {
                    ref[key] = JSON.parse(value.replace(/'/g, '"')); // handle array types
                }
                if (key === 'duration' || key === 'durationError') {
                    ref[key] = JSON.parse(value); // handle number types
                }
            }
        }
        return ref;
    }
}
class FlowChecker extends BehaviorChecker {
    extractRefsFromFiles() {
        const files = glob.sync(path.join(this.dir, 'src/*' + '*/*.flow'));
        if (this.verbose && !files.length) {
            this.refChecker.emit('log', `No Flows found in ${this.dir}/src`);
        }
        for (const filename of files) {
            this.extractRefsFromFile(filename);
        }
    }
    extractRefsFromFile(uri) {
        const loader = new jibo_flow_core_1.loaders.GojsLoader();
        const procedure = loader.loadProcedure(uri);
        this.extractRefsFromOptions(uri, procedure.activities);
    }
}
class BtChecker extends BehaviorChecker {
    extractRefsFromFiles() {
        //                                            vvv  <- Hack to fix jsdocs error.
        const files = glob.sync(path.join(this.dir, 'src/*' + '*/*.bt'));
        if (this.verbose && !files.length) {
            this.refChecker.emit('log', `No Behavior Trees found in ${this.dir}/src`);
        }
        for (const filename of files) {
            this.extractRefsFromFile(filename);
        }
    }
    extractRefsFromFile(uri) {
        let bt;
        try {
            bt = JSON.parse(fs.readFileSync(uri, 'utf8'));
        }
        catch (e) {
            this.refChecker.addError(uri, `BT file '${uri}' could not be loaded and parsed`, 'BT-LOAD');
            return;
        }
        this.extractRefsFromOptions(uri, bt);
    }
}
class MimChecker {
    constructor(refChecker, dir) {
        this.refChecker = refChecker;
        this.dir = path.resolve(dir);
        this.errors = [];
        this.refs = [];
        this.verbose = refChecker.verbose;
        this.variableTypes = null;
    }
    getReferences() {
        return this.refs;
    }
    extractRefsFromFiles() {
        this.setupTypes();
        const files = glob.sync(path.join(this.dir, 'mims/*' + '*/*.mim'));
        if (this.verbose && !files.length) {
            this.refChecker.emit('log', `No mims found in ${this.dir}/mims`);
        }
        for (const filename of files) {
            this.extractRefs(filename);
        }
    }
    // Extract the animation DB references from all the mim prompts and accumulate them in this.refs
    // Also check validity of prompts and conditions
    extractRefs(mimPath) {
        let mim;
        try {
            mim = JSON.parse(fs.readFileSync(mimPath, 'utf8'));
        }
        catch (e) {
            this.refChecker.addError(mimPath, `references Mim file '${mimPath}' with bad JSON`, 'JSON-PARSE');
            return;
        }
        let localPath = mimPath.substr(this.dir.length + 1);
        for (const prompt of mim.prompts) {
            const location = `${localPath} - prompt ${prompt.prompt_id}`;
            this.refs = this.refs.concat(getAnimReferencesFromXml(this.refChecker, location, prompt.prompt));
        }
        let sandbox;
        let filename;
        if (this.variableTypes) {
            filename = path.basename(mimPath);
            sandbox = {};
            this.addVariableToSandbox("_now", { "type": "datetime" }, sandbox); // _now is in every sandbox
            if ('global' in this.variableTypes) {
                for (let variable in this.variableTypes.global) {
                    this.addVariableToSandbox(variable, this.variableTypes.global[variable], sandbox);
                }
            }
            if (filename in this.variableTypes) {
                for (let variable in this.variableTypes[filename]) {
                    this.addVariableToSandbox(variable, this.variableTypes[filename][variable], sandbox);
                }
            }
            sandbox = vm.createContext(sandbox);
            for (const prompt of mim.prompts) {
                this.checkPromptValidity(filename, prompt, sandbox);
            }
        }
    }
    setupTypes() {
        const files = glob.sync(path.join(this.dir, 'mim-types.json'));
        if (!files.length) {
            this.refChecker.emit('log', `No mim-types.json found in ${this.dir}`);
            return;
        }
        for (const filename of files) {
            this.addTypesToVariableTypes(filename);
        }
    }
    addTypesToVariableTypes(filename) {
        if (this.variableTypes === null) {
            this.variableTypes = {};
        }
        let types;
        try {
            types = JSON.parse(fs.readFileSync(filename, 'utf8'));
        }
        catch (e) {
            this.refChecker.addError(filename, `references file '${filename}' with bad JSON`, 'JSON-PARSE');
            return;
        }
        for (let file in types) {
            this.variableTypes[file] = types[file];
        }
    }
    addVariableToSandbox(variable, typeInfo, sandbox) {
        if (typeInfo.type === 'string') {
            sandbox[variable] = 'foo';
        }
        else if (typeInfo.type === 'number') {
            sandbox[variable] = 42;
        }
        else if (typeInfo.type === 'boolean') {
            sandbox[variable] = true;
        }
        else if (typeInfo.type === 'datetime') {
            sandbox[variable] = new jibo_data_utils_1.DateTime();
        }
        else if (typeInfo.type === 'location') {
            sandbox[variable] = jibo_data_utils_1.Location.jiboHome;
        }
    }
    checkPromptValidity(filename, prompt, sandbox) {
        if (this.variableTypes === null) {
            return true; // don't check types if there are no types to check
        }
        return this.conditionIsWellFormed(filename, prompt, sandbox) &&
            this.promptTextIsWellFormed(filename, prompt, sandbox);
    }
    conditionIsWellFormed(filename, prompt, sandbox) {
        if (!prompt.condition) {
            return true;
        }
        //zero or negative weights should be interpreted as 'never pick this prompt'
        if (prompt.weight && prompt.weight <= 0) {
            return false;
        }
        //remove the leading '${' and trailing '}' from the condition string, because
        //we don't want those - older mims may have been created with them present
        let condition = prompt.condition.replace(/^\$\{/, "").replace(/\}$/, "");
        //do a safe eval on the condition
        try {
            return !!vm.runInContext(condition, sandbox);
        }
        catch (e) {
            this.refChecker.addError(`error evaluating prompt condition in ${filename}: ${prompt.condition} - error: ${e.message}`);
            return false;
        }
    }
    promptTextIsWellFormed(filename, prompt, sandbox) {
        if (!prompt.prompt) {
            return true;
        }
        if (prompt.weight && prompt.weight <= 0) {
            return false;
        }
        try {
            return !!vm.runInContext("`" + prompt.prompt + "`", sandbox);
        }
        catch (e) {
            this.refChecker.addError(`error evaluating prompt formation in ${filename}: ${prompt.prompt} - error: ${e.message}`);
            return false;
        }
    }
}
function getAnimReferencesFromXml(refChecker, location, prompt) {
    const xml = `<root>${prompt}</root>`;
    let node;
    try {
        node = jibo_node_xml_1.Parser.parseXML(xml);
    }
    catch (e) {
        const code = 'SSML-PARSE';
        let message = `${e} - ${prompt}`;
        refChecker.addError(location, message, code);
    }
    let refs = [];
    if (node) {
        refs = getAnimReferencesFromNodes(refChecker, location, [node]);
        // debugging
        // if (refs.length !== 0) {
        //     console.log(prompt);
        //     for (let i of refs) {
        //         console.log(formQuery(i));
        //     }
        // }
    }
    return refs;
}
function getAnimReferencesFromNodes(refChecker, location, nodes) {
    const template = /\$\{[^\$\{]*\}/g; // Regex looking for ${}
    let refs = [];
    let resolvable = true;
    for (const elem of nodes) {
        let type = `${elem.type}`.toUpperCase();
        if (['ANIM', 'TTS', 'SSA', 'SFX', 'ES'].indexOf(type) !== -1) {
            // Place both the Map and a normal JS object with the same keys and values in the reference.
            let attr = {};
            for (let key of elem.att.keys()) {
                const val = elem.att.get(key);
                if (template.test(key) || template.test(val)) {
                    resolvable = false; // Mark this ESML node as unresolvable
                }
                attr[key] = val;
            }
            if (resolvable) {
                let query = { type, location, attr, att: elem.att };
                refs.push(query);
            }
            else {
                refChecker.emit('warn', `ESML contains templates, unable to resolve: ${location}`);
            }
        }
        refs = refs.concat(getAnimReferencesFromNodes(refChecker, location, elem.children));
    }
    return refs;
}
// Handle category and meta tag references
function formQuery(ref) {
    let node = new AssetNode_1.AssetNode(ref.type, ref.att);
    let query = {
        categories: node.categoryFilter.allOf,
        includeMeta: node.metaFilter.allOf,
        includeSomeMeta: node.metaFilter.someOf,
        excludeMeta: node.metaFilter.noneOf
    };
    return query;
}
/**
 * Hold one preflight error.
 * @class Error
 * @private
 */
class Error {
    constructor(location, message, code) {
        this.location = location;
        this.message = message;
        this.code = code;
    }
    toString() {
        return `   ${this.code} ${this.message}`;
    }
}
exports.default = AnimRefChecker;

},{"./AssetNode":1,"./cli":3,"events":undefined,"fs":undefined,"glob":undefined,"jibo-anim-db":undefined,"jibo-data-utils":undefined,"jibo-flow-core":undefined,"jibo-node-xml":undefined,"path":undefined,"request":undefined,"resolve":undefined,"vm":undefined}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const minimist = require("minimist");
const chalk = require("chalk");
const anim_ref_checker_1 = require("./anim-ref-checker");
function usage() {
    console.log(`usage: node anim-ref-cheker.js [options]
  options:
    -s --check-skill /path/to/skill/folder (side effect: creates animdb-manifest.json in skill's top level directory)
    -m --check-manifest /path/to/animdb-manifest.json (can also be a URL that returns the **RAW** JSON text.)
    -a --use-animdb /path/to/animdb.json
    --no-manifest ; don't generate an animdb-manifest.json file
    --no-report ; don't generate a reference check report
    --verbose ; report on files that have no errors.

You must provide either:
     1) --check-skill or
     2) --check-manifest and --use-animdb

The 3 modes of use:

1) Check a skill's animdb references against the skill's npm-installed jibo-anim-db-animations (writes manifest in skill folder)
   $ anim-ref-check --check-skill /path/to/skill

2) Check a skill's animdb references against an arbitrary animDB (writes manifest in skill folder)
   $ anim-ref-check --check-skill /path/to/skill --use-animdb /path/to/animdb.json

3) Check the references in a manifest against an arbitrary animDB
   $ anim-ref-check --check-manifest /path/to/skill/animdb-manifest.json --use-animdb /path/to/animdb.json`);
    process.exit(0);
}
function cli() {
    const argv = process.argv.slice(2);
    if (argv.length === 0) {
        usage();
    }
    const args = minimist(argv, {
        string: ['skill', 'animdb', 'manifest'],
        boolean: ['verbose', 'no-manifest', 'no-report'],
        alias: {
            'check-skill': 'skill',
            'check-manifest': 'manifest',
            'use-animdb': 'animdb',
            s: 'skill',
            a: 'animdb',
            m: 'manifest',
            v: 'verbose',
            M: 'no-manifest',
            R: 'no-report',
        },
        default: {
            verbose: false,
            'no-manifest': false,
            'no-report': false,
        }
    });
    if (!!args.skill === !!args.manifest) {
        console.log('Error: a skill or manifest must be specified, but not both.');
        usage();
    }
    const result = anim_ref_checker_1.AnimRefChecker.exec(args);
    result.on('end', (err) => {
        if (err) {
            console.log(err);
        }
        process.exit(0);
    });
    result.on('log', (message) => {
        if (args.verbose) {
            console.log(chalk.green(message));
        }
    });
    result.on('warn', (message) => {
        if (args.verbose) {
            console.log(chalk.yellow(message));
        }
    });
    result.on('lintError', (err) => {
        console.error(chalk.red(err));
    });
}
exports.default = cli;

},{"./anim-ref-checker":2,"chalk":undefined,"minimist":undefined}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Behaviorify_1 = require("../bt-flow/Behaviorify");
const Flowify_1 = require("../bt-flow/Flowify");
const path = require("path");
const through = require("through2");
const fs = require("fs");
// Keep a local cache of all the files we process
// so we don't need to rerun everytime
const browserifyCache = {};
//also keep track of the last modified time, so that we can rerun files that change when watching
const changeTime = {};
/**
 * @description Transform for Browserify to compile `.bt` and `.flow` files into source.
 * ```
 * var browserify = require('browserify');
 * browserify().transform('jibo-dev/browserify-transform');
 * ```
 * ```
 * {
 *     "browserify": {
 *         "transform": [
 *             "jibo-dev/browserify-transform"
 *         ]
 *     }
 * }
 * ```
 * @class BrowserifyTransform
 * @param filename {string} `.bt` or `.flow` file to compile.
 */
function BrowserifyTransform(filename) {
    const chunks = [];
    function transform(chunk, encoding, callback) {
        chunks.push(chunk);
        callback();
    }
    function flush(callback) {
        const stream = this;
        // Don't process bt/flow files more than once
        const time = fs.statSync(filename).mtime.getTime();
        if (browserifyCache[filename] && time === changeTime[filename]) {
            stream.push(browserifyCache[filename]);
            callback();
        }
        else {
            let result;
            const source = Buffer.concat(chunks).toString();
            switch (path.extname(filename)) {
                case '.bt': {
                    result = Behaviorify_1.default.transform(filename, source);
                    break;
                }
                case '.flow': {
                    result = Flowify_1.default.transform(filename, source);
                    break;
                }
                default: {
                    result = source;
                    break;
                }
            }
            browserifyCache[filename] = result;
            changeTime[filename] = time;
            stream.push(result);
            callback();
        }
    }
    return through(transform, flush);
}
exports.default = BrowserifyTransform;

},{"../bt-flow/Behaviorify":5,"../bt-flow/Flowify":6,"fs":undefined,"path":undefined,"through2":undefined}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// External dependencies
const path = require("path");
const fs = require("fs");
const escodegen = require("escodegen");
const esprima = require("esprima");
const findRoot = require("find-root");
const estraverse = require("estraverse");
const chalk = require("chalk");
const jibo_esexpressions_1 = require("jibo-esexpressions");
const cli_1 = require("./cli");
const core_1 = require("./core");
const versions_1 = require("./versions");
/**
 * Converts a raw JSON data structure into a JavaScript module.
 * ```
 * import {Behaviorify} = require('jibo-dev');
 * const bt = new Behaviorify('/path/to/custom.bt');
 * console.log(bt.exec());
 * ```
 * @class Behaviorify
 * @param {String} filename The file name if the first argument is an Object.
 * @param {Object} [data] The `.bt` or `.flow` file data or path to file.
 */
class Behaviorify {
    constructor(filename, data) {
        /**
         * Path to the filename.
         * @name Behaviorify#filename
         * @type {String}
         */
        this.filename = path.resolve(filename);
        // First argument can be a file path
        if (!data) {
            data = JSON.parse(fs.readFileSync(this.filename, 'utf8'));
        }
        // Get the version of the meta field, add backward compatibility
        // for bt files that were created before versioning
        data.meta = data.meta || { version: 0 };
        /**
         * JSON data for a .flow or .bt file
         * @name Behaviorify#data
         * @type {Object}
         */
        this.data = data;
        // Update old versions of the tree
        this.update();
    }
    /**
     * Browserify transform
     * @method Behaviorify.transform
     * @private
     * @param {String} filename File name.
     * @param {String} source Buffer from source file.
     * @return {String} Output JavaScript module buffer.
     */
    static transform(filename, source) {
        const data = JSON.parse(source);
        //substitute the behaviorPaths
        const keys = Object.keys(data);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            if (data[key].class === 'Subtree') {
                const behaviorPath = data[key].options.behaviorPath;
                data[key].options.behaviorPath = [
                    "() => {",
                    "    return require('" + behaviorPath + "');",
                    "}"
                ];
            }
        }
        const result = new Behaviorify(filename, data);
        return result.exec();
    }
    /**
     * Run behaviorify for the commandline.
     * @method Behaviorify.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    /**
     * The core classes.
     * @method Behaviorify.core
     * @return {Map} Collection of core behavior and decorators.
     */
    static get core() {
        return core_1.default;
    }
    /**
     * Update the raw tree with deprecations.
     * @method Behaviorify#update
     */
    update() {
        // Massage the JSON tree to provide forward compatibility
        while (this.data.meta.version < Behaviorify.CURRENT_VERSION) {
            versions_1.default["version" + this.data.meta.version](this.data);
            this.data.meta.version++;
        }
    }
    /**
     * Convert the object.
     * @method Behaviorify#exec
     * @param {Object} [options] - Options.
     * @param {Boolean} [options.noRequirification] if true, the subflow/subtree references are left untouched.
     * @return {String} Resulting buffer JavaScript file.
     */
    exec(options) {
        // Create a new template everytime
        let tree = esprima.parse(this.readFile('templates/tree.js'));
        let treeObject = new jibo_esexpressions_1.ObjectExpression(tree.body[tree.body.length - 1]
            .expression
            .right
            .body
            .body[0]
            .argument
            .properties);
        let isRequirifying = !options || !options.noRequirification;
        let ids = Object.keys(this.data);
        for (let i = 0; i < ids.length; i++) {
            let id = ids[i];
            // Ignore the meta information
            if (id === "meta") {
                let meta = this.data[id];
                let metaExpression = new jibo_esexpressions_1.ObjectExpression();
                Object.keys(meta).forEach((key) => {
                    metaExpression.addKeyValue(key, jibo_esexpressions_1.LiteralExpression(meta[key]));
                });
                treeObject.addKeyValue('meta', metaExpression.expression);
                continue;
            }
            let nodeDef = this.data[id];
            let hasFunction = false;
            let hasSelf = false;
            let objectExpression = new jibo_esexpressions_1.ObjectExpression();
            objectExpression.addKeyValue('id', jibo_esexpressions_1.LiteralExpression(id));
            if (nodeDef.name) {
                objectExpression.addKeyValue('name', jibo_esexpressions_1.LiteralExpression(nodeDef.name));
            }
            if (nodeDef.parent) {
                objectExpression.addKeyValue('parent', jibo_esexpressions_1.LiteralExpression(nodeDef.parent));
            }
            if (nodeDef['asset-pack']) {
                objectExpression.addKeyValue('asset-pack', jibo_esexpressions_1.LiteralExpression(nodeDef['asset-pack']));
            }
            if (nodeDef.transitions) {
                let arrayExpression = new jibo_esexpressions_1.ArrayExpression();
                for (let k = 0; k < nodeDef.transitions.length; k++) {
                    let transition = nodeDef.transitions[k];
                    let transitionExpression = new jibo_esexpressions_1.ObjectExpression();
                    Object.keys(transition).forEach((key) => {
                        transitionExpression.addKeyValue(key, jibo_esexpressions_1.LiteralExpression(transition[key]));
                    });
                    arrayExpression.push(transitionExpression.expression);
                }
                objectExpression.addKeyValue('transitions', arrayExpression.expression);
            }
            if (nodeDef.exceptions) {
                let arrayExpression = new jibo_esexpressions_1.ArrayExpression();
                for (let k = 0; k < nodeDef.exceptions.length; k++) {
                    let exceptions = nodeDef.exceptions[k];
                    let exceptionsExpression = new jibo_esexpressions_1.ObjectExpression();
                    Object.keys(exceptions).forEach((key) => {
                        exceptionsExpression.addKeyValue(key, jibo_esexpressions_1.LiteralExpression(exceptions[key]));
                    });
                    arrayExpression.push(exceptionsExpression.expression);
                }
                objectExpression.addKeyValue('exceptions', arrayExpression.expression);
            }
            objectExpression.addKeyValue('class', jibo_esexpressions_1.LiteralExpression(nodeDef.class));
            if (nodeDef.children) {
                let arrayExpression = new jibo_esexpressions_1.ArrayExpression();
                for (let k = 0; k < nodeDef.children.length; k++) {
                    arrayExpression.push(jibo_esexpressions_1.LiteralExpression(nodeDef.children[k]));
                }
                objectExpression.addKeyValue('children', arrayExpression.expression);
            }
            if (nodeDef.skipped) {
                objectExpression.addKeyValue('skipped', jibo_esexpressions_1.LiteralExpression(nodeDef.skipped));
            }
            if (nodeDef.decorators) {
                let arrayExpression = new jibo_esexpressions_1.ArrayExpression();
                for (let k = 0; k < nodeDef.decorators.length; k++) {
                    arrayExpression.push(jibo_esexpressions_1.LiteralExpression(nodeDef.decorators[k]));
                }
                objectExpression.addKeyValue('decorators', arrayExpression.expression);
            }
            if (nodeDef.options) {
                let optionsExpression = new jibo_esexpressions_1.ObjectExpression();
                for (let k in nodeDef.options) {
                    let value = nodeDef.options[k];
                    if (Array.isArray(value)) {
                        value = nodeDef.options[k] = value.join("\n");
                    }
                    if (typeof value === 'string' && (value.indexOf('function') === 0 || value.indexOf('=>') >= 0)) {
                        hasFunction = true;
                        try {
                            let code = value;
                            let parsed = esprima.parse(code);
                            estraverse.traverse(parsed, {
                                enter: (node, parent) => {
                                    if (node.name && node.name === 'self' &&
                                        parent.object && parent.object.name === 'self') {
                                        hasSelf = true;
                                    }
                                }
                            });
                            if (parsed.body[0].type === 'ExpressionStatement') {
                                parsed = parsed.body[0].expression;
                            }
                            else {
                                parsed = parsed.body[0];
                            }
                            optionsExpression.addKeyValue(k, parsed);
                        }
                        catch (e) {
                            const localFilename = this.filename.replace(process.cwd() + '/', '');
                            console.error(chalk.underline(localFilename));
                            console.error(chalk.red.bold('  error'), e.message);
                            console.error(chalk.gray('  ' + value.replace(/\t/g, '  ').replace(/\n/g, '\n  ')));
                            console.error();
                        }
                    }
                    else if (isRequirifying && nodeDef.class === 'Subtree' && k === 'behaviorPath') {
                        optionsExpression.addKeyValue(k, jibo_esexpressions_1.ArrowFunctionReturnRequire(value));
                    }
                    else if (isRequirifying && nodeDef.class === 'Flow.Subflow' && k === 'subflowId') {
                        optionsExpression.addKeyValue(k, jibo_esexpressions_1.ArrowFunctionReturnRequire(value));
                    }
                    else if (isRequirifying && nodeDef.class === 'Flow.Subtree' && k === 'behaviorPath') {
                        optionsExpression.addKeyValue(k, jibo_esexpressions_1.ArrowFunctionReturnRequire(value));
                    }
                    else {
                        optionsExpression.addKeyValue(k, jibo_esexpressions_1.LiteralExpression(value));
                    }
                }
                objectExpression.addKeyValue('options', optionsExpression.expression);
            }
            if (hasFunction) {
                treeObject.addKeyValue(id, jibo_esexpressions_1.FunctionExpression(objectExpression.expression, hasSelf));
            }
            else {
                treeObject.addKeyValue(id, objectExpression.expression);
            }
        }
        return escodegen.generate(tree);
    }
    /**
     * Clear and don't use after this.
     * @method Behaviorify#dispose
     * @private
     */
    dispose() {
        this.data = null;
        this.filename = null;
    }
    /**
     * Simple utility to read a file.
     * @method Behaviorify#readFile
     * @private
     * @param {String} file The local file path.
     * @return {String} Contents of the file.
     */
    readFile(file) {
        const rootPath = findRoot(__dirname);
        return fs.readFileSync(path.resolve(rootPath, file), 'utf8');
    }
}
/**
 * Current version of the behavior tree format
 * @name Behaviorify.CURRENT_VERSION
 * @type {Number}
 */
Behaviorify.CURRENT_VERSION = 1;
exports.default = Behaviorify;

},{"./cli":7,"./core":8,"./versions":13,"chalk":undefined,"escodegen":undefined,"esprima":undefined,"estraverse":undefined,"find-root":undefined,"fs":undefined,"jibo-esexpressions":undefined,"path":undefined}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GojsBareBonesLoader_1 = require("./flow/gojs-loader/GojsBareBonesLoader");
const Behaviorify_1 = require("./Behaviorify");
const chalk = require("chalk");
const path = require("path");
/**
 * API to convert `.flow` files into JavaScript.
 * ```
 * import {Flowify} = require('jibo-dev');
 * const flow = new Flowify('/path/to/custom.flow');
 * console.log(flow.exec());
 * ```
 * @class Flowify
 * @param {String} filename Path and name of the file to compile.
 * @param {Object} [data] The parsed JSON file to load.
 */
class Flowify extends Behaviorify_1.default {
    /**
     * Browserify transform.
     * @method Flowify.transform
     * @private
     * @param {String} filename File name.
     * @param {String} source Buffer from source file.
     * @return {String} Output JavaScript module buffer.
     */
    static transform(filename, source) {
        const data = JSON.parse(source);
        const nodeDataArray = data.nodeDataArray;
        for (let i = 0; i < nodeDataArray.length; i++) {
            let node = nodeDataArray[i];
            switch (node.class) {
                case 'Flow.Subflow':
                    {
                        node.options.subflowId = [
                            "() => {",
                            "    return require('" + node.options.subflowId + "');",
                            "}"
                        ];
                        break;
                    }
                case 'Subtree':
                    {
                        node.options.behaviorPath = [
                            "() => {",
                            "    return require('" + node.options.behaviorPath + "');",
                            "}"
                        ];
                        break;
                    }
            }
        }
        const result = new Flowify(filename, data);
        return result.exec();
    }
    constructor(filename, data) {
        if (path.extname(filename) !== '.flow') {
            console.error(chalk.red('Flowify cannot handle %s files, only .flow files.'));
            return;
        }
        let procedure;
        if (data) {
            procedure = GojsBareBonesLoader_1.default.loadParsedProcedure(filename, data);
        }
        else {
            procedure = GojsBareBonesLoader_1.default.loadProcedure(filename);
        }
        super(filename, procedure.toBehaviorifyable());
    }
}
exports.default = Flowify;

},{"./Behaviorify":5,"./flow/gojs-loader/GojsBareBonesLoader":12,"chalk":undefined,"path":undefined}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const minimist = require("minimist");
const mkdirp = require("mkdirp");
const glob = require("glob");
const Behaviorify_1 = require("./Behaviorify");
const Flowify_1 = require("./Flowify");
const chalk = require("chalk");
function cli() {
    const args = minimist(process.argv.slice(2), {
        string: ['input', 'output'],
        boolean: ['verbose', 'update'],
        alias: {
            v: 'verbose',
            u: 'update',
            i: 'input',
            o: 'output'
        },
        default: {
            verbose: false,
            update: false
        }
    });
    let files = args._;
    if (args.input) {
        files = glob.sync(path.join(args.input, '**/*.{bt,flow}'));
    }
    if (!files.length) {
        console.error(chalk.red("Must provide files to process"));
        process.exit(0);
    }
    //get the path relative to the behaviors root
    files.forEach((file, i) => {
        files[i] = path.relative(args.input, file);
    });
    // Convert a file
    if (args.update && files.length > 1) {
        files.forEach(function (filename) {
            const uri = path.resolve(filename);
            const data = JSON.parse(fs.readFileSync(uri, 'utf8'));
            const format = new Behaviorify_1.default(filename, data);
            fs.writeFileSync(uri, JSON.stringify(format.data, null, "  "));
            if (args.verbose) {
                console.log(chalk.green(" -> Updated %s"), filename);
            }
        });
        process.exit(0);
    }
    else if (files.length) {
        // Make directory in case
        if (args.output) {
            mkdirp.sync(path.resolve(args.output));
        }
        files.forEach(function (filename) {
            let parts = filename.split('.');
            let extension = '.' + parts[parts.length - 1];
            if (extension !== '.bt' && extension !== '.flow') {
                console.error(chalk.red("Invalid: not a *.bt or *.flow file: %s"), filename);
                return;
            }
            const uri = path.resolve(args.input, filename);
            const data = JSON.parse(fs.readFileSync(uri, 'utf8'));
            // Make sure the file is valid
            try {
                fs.statSync(uri);
            }
            catch (e) {
                console.error(chalk.red("File does not exist '%s'", filename));
                return;
            }
            let format;
            switch (extension) {
                case '.flow':
                    format = new Flowify_1.default(filename, data);
                    break;
                case '.bt':
                    format = new Behaviorify_1.default(filename, data);
                    break;
            }
            if (args.output) {
                let outputName;
                // Rebase to the input folder
                if (args.input) {
                    outputName = filename.replace(RegExp('\\' + extension + '$'), '.js')
                        .replace(args.input + '/', '');
                }
                else {
                    outputName = path.basename(filename, extension) + '.js';
                }
                const outputFile = path.join(args.output, outputName);
                const outputDir = path.dirname(outputFile);
                // Make the output directory
                mkdirp.sync(outputDir);
                fs.writeFileSync(outputFile, format.exec());
                if (args.verbose) {
                    console.log(chalk.green("Saved '%s' to '%s'"), filename, outputName);
                }
            }
            else {
                console.log(format.exec());
            }
        });
        process.exit(0);
    }
    console.error(chalk.red("Invalid command."));
    process.exit(1);
}
exports.default = cli;

},{"./Behaviorify":5,"./Flowify":6,"chalk":undefined,"fs":undefined,"glob":undefined,"minimist":undefined,"mkdirp":undefined,"path":undefined}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    // Behaviors and the arguments order
    Blink: [],
    ExecuteScript: ['exec'],
    ExecuteScriptAsync: ['exec'],
    Listen: ['getOptions', 'rule', 'onResult'],
    ListenEmbedded: ['rule', 'onResult'],
    ListenJs: ['getOptions', 'getRule', 'onResult'],
    LookAt: ['getTarget', 'isContinuous', 'config'],
    Null: [],
    Parallel: ['succeedOnOne', 'children'],
    PlayAnimation: ['animPath', 'config'],
    PlayAudio: ['audioPath'],
    Random: ['children'],
    ReadBarcode: ['onBarcode'],
    Sequence: ['children'],
    Subtree: ['behaviorPath', 'getNotepad', 'onResult'],
    SubtreeJs: ['getFileName', 'getNotepad', 'onResult'],
    Switch: ['children'],
    TakePhoto: ['resolution', 'onPhoto', 'noDistortion'],
    TextToSpeech: ['words', 'onWord'],
    TextToSpeechJs: ['getWords', 'onWord'],
    TimeoutJs: ['getTime'],
    // Decorators
    Case: ['conditional'],
    FailOnCondition: ['init', 'conditional'],
    StartOnAnimEvent: ['eventName', 'onReceived'],
    StartOnCondition: ['init', 'conditional'],
    StartOnEvent: ['eventName', 'onEvent'],
    SucceedOnCondition: ['init', 'conditional'],
    SucceedOnEmbedded: ['rule', 'onResult'],
    SucceedOnEvent: ['eventName', 'onEvent'],
    SucceedOnListen: ['getOptions', 'rule', 'onResult'],
    SucceedOnListenJs: ['getOptions', 'getRule', 'onResult'],
    TimeoutFail: ['timeout'],
    TimeoutSucceed: ['timeout'],
    TimeoutSucceedJs: ['getTime'],
    WhileCondition: ['init', 'conditional']
};

},{}],9:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Each Activity corresponds to a single node in the state machine.
class Activity {
    constructor(name, id, className, assetPack, options) {
        this.name = name;
        this.id = id;
        this.class = className;
        this['asset-pack'] = assetPack; // right now we need this in behaviorify.
        this.assetPack = assetPack; // in the future we'll use this in behaviorify.
        this.options = options;
    }
    set_transitions(transitions) {
        this.transitions = transitions;
    }
    set_exceptions(exceptions) {
        this.exceptions = exceptions;
    }
}
exports.default = Activity;

},{}],10:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Each Procedure corresponds to one page of the Flow document, which corresponds to one .flow file.
class Procedure {
    constructor(name, uri, local_inits, param_inits) {
        this.name = name;
        this.uri = uri;
        this.activities = {};
        this.local_inits = local_inits;
        this.param_inits = param_inits;
        this.transitions = {};
        this.exceptions = {};
    }
    // NB. you must add all the transitions for the procedure first, then add the activities.
    add_transition(transition) {
        let map = transition.value[0] === '~' ? this.exceptions : this.transitions;
        if (!(transition.frm in map)) {
            map[transition.frm] = [];
        }
        map[transition.frm].push(transition);
    }
    add_activity(activity) {
        let transitions = activity.id in this.transitions ? this.transitions[activity.id] : [];
        let exceptions = activity.id in this.exceptions ? this.exceptions[activity.id] : [];
        activity.set_transitions(transitions);
        activity.set_exceptions(exceptions);
        this.activities[activity.id] = activity;
    }
    toBehaviorifyable() {
        let input;
        input = { meta: { version: 1, name: this.name, uri: this.uri } };
        for (let key in this.activities) {
            let activity = this.activities[key];
            input[activity.id] = activity;
        }
        return input;
    }
}
exports.default = Procedure;

},{}],11:[function(require,module,exports){
"use strict";
// Each Transition corresponds to an edge in the state machine.
Object.defineProperty(exports, "__esModule", { value: true });
class Transition {
    constructor(frm, to, value) {
        this.frm = frm;
        this.to = to;
        this.value = value;
    }
}
exports.default = Transition;

},{}],12:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Activity_1 = require("../flow-objects/Activity");
const Procedure_1 = require("../flow-objects/Procedure");
const Transition_1 = require("../flow-objects/Transition");
const fs = require('fs');
/**
 * A minimal Flow File loader used by Behaviorify for parsing a .flow file's JSON into
 * the Flow primary objects (Procedures, Activities and Transitions).
 */
class GojsBareBonesLoader {
    static getName(uri) {
        // Remove directory path
        let names = uri.split('/');
        // Remove extension
        let name = names[names.length - 1].replace(/\.[^.]*$/, '');
        return name;
    }
    static getBase(uri) {
        // Remove filename part from directory path.
        return uri.replace(/\/[^/]*$/, '');
    }
    static loadProcedureRelative(currentUri, name) {
        let baseUri = GojsBareBonesLoader.getBase(currentUri);
        let uri = baseUri + '/' + name + '.flow';
        return GojsBareBonesLoader.loadProcedure(uri);
    }
    static loadProcedure(uri) {
        let data = fs.readFileSync(uri, 'utf8');
        let base = JSON.parse(data);
        return GojsBareBonesLoader.loadParsedProcedure(uri, base);
    }
    static loadParsedProcedure(uri, base) {
        let name = GojsBareBonesLoader.getName(uri);
        let procedure = new Procedure_1.default(name, uri, {}, {});
        for (let i = 0; i < base.linkDataArray.length; i++) {
            let goLink = base.linkDataArray[i];
            let transitionValue = goLink.text === undefined ? "" : goLink.text;
            let transition = new Transition_1.default(goLink.from, goLink.to, transitionValue);
            procedure.add_transition(transition);
        }
        for (let i = 0; i < base.nodeDataArray.length; i++) {
            let goNode = base.nodeDataArray[i];
            let activity = new Activity_1.default(goNode.name, goNode.id, goNode.class, goNode['asset-pack'], goNode.options);
            procedure.add_activity(activity);
        }
        return procedure;
    }
}
exports.default = GojsBareBonesLoader;

},{"../flow-objects/Activity":9,"../flow-objects/Procedure":10,"../flow-objects/Transition":11,"fs":undefined}],13:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const version0_1 = require("./version0");
exports.default = {
    version0: version0_1.default
};

},{"./version0":14}],14:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const glob = require("glob");
const path = require("path");
const chalk = require("chalk");
const core_1 = require("../core");
let cache = {};
// The arguments were converted from an array to an object map
function default_1(rawTree) {
    /**
     * Get the schemas from the current working directory
     * to help process old versions of behavior files
     * @function folderSchema
     * @private
     */
    function folderSchema(dir) {
        let result = {};
        // Combine all the schemes into a single file
        let schemas = {};
        let regex = dir + '/schema?(s)/' + "*.json";
        let files = cache[regex] || glob.sync(regex);
        if (!cache[regex]) {
            cache[regex] = files;
        }
        if (!files.length) {
            return result;
        }
        files.forEach(function (file) {
            let schema = cache[file] || require(file);
            // Upgrade the scheme to version 1!
            // each schema entry should have an array of options
            // and not meta/args fields
            if (schema.meta === undefined) {
                for (let name in schema) {
                    let definition = schema[name];
                    let options = [];
                    if (definition.meta) {
                        definition.meta.forEach(function (field, i) {
                            field.type = definition.args[i];
                            options.push(field);
                        });
                        delete definition.meta;
                        delete definition.args;
                    }
                    definition.options = options;
                }
                schema.meta = { version: 1 };
            }
            if (!cache[file]) {
                cache[file] = schema;
            }
            Object.assign(schemas, schema);
        });
        // Look through the schemas and map the class name
        // with the list of arguments
        for (let id in schemas) {
            if (id === 'meta') {
                continue;
            } // ignore meta
            let definition = schemas[id];
            if (definition.options) {
                let args = [];
                definition.options.forEach(function (key) {
                    args.push(key.field);
                });
                result[id] = args;
            }
        }
        return result;
    }
    // Collection of behaviors by namespace
    let Definitions = {
        core: core_1.default,
        project: folderSchema(process.cwd())
    };
    /**
     * Search through the node modules and find any asset packs
     * @function findAssetPacks
     * @private
     */
    function findAssetPacks() {
        let result = {};
        let regex = process.cwd() + '/node_modules/*/package.json';
        let files = cache[regex] || glob.sync(regex);
        if (!cache[regex]) {
            cache[regex] = files;
        }
        if (!files.length) {
            return result;
        }
        files.forEach(function (file) {
            let data = require(file);
            if (data.jibo) {
                let dir = path.dirname(file);
                Object.assign(Definitions.project, folderSchema(dir));
            }
        });
    }
    /**
     * Utility to convert arguments array into an object
     * @function toObject
     * @private
     * @param {array} values Collection of argument values
     * @param {array<string>} fields Names of fields
     * @return {object}
     */
    function toObject(values, fields) {
        let obj = {};
        for (let i = 0; i < values.length; i++) {
            obj[fields[i]] = values[i];
        }
        return obj;
    }
    // Add the asset packs into the definitions
    findAssetPacks();
    for (let id in rawTree) {
        if (id === "meta") {
            continue;
        }
        let node = rawTree[id];
        // Handle the core modules
        if (Definitions.core[node.class]) {
            node.options = toObject(node.args, Definitions.core[node.class]);
        }
        else if (Definitions.project[node.class]) {
            node.options = toObject(node.args, Definitions.project[node.class]);
        }
        else {
            const message = "Warning: Cannot upgrade arguments in class '%s'. Please update Schema and resave .bt file.";
            console.error(chalk.yellow(message), node.class);
            node.options = {};
        }
        delete node.args;
    }
    return rawTree;
}
exports.default = default_1;

},{"../core":8,"chalk":undefined,"glob":undefined,"path":undefined}],15:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const rimraf = require("rimraf");
const hash = require("object-hash");
const cli_1 = require("./cli");
const events_1 = require("events");
const glob = require("glob");
const jibo_anim_db_1 = require("jibo-anim-db");
const { KeyframesCompositer, keysObjToAnimObject } = require('jibo-keyframes');
const jibo_plugins_1 = require("jibo-plugins");
/**
 * Statically compile a skill's keys files into anim files.
 * ```
 * import {KeysCompiler} from 'jibo-dev';
 * ```
 * @class KeysCompiler
 * @param {string} projectRoot Path to the skill's project-level directory.
 */
class KeysCompiler extends events_1.EventEmitter {
    constructor(args) {
        super();
        this.args = args;
        this.skillDir = args.skill;
        this.files = Array.isArray(args.file) ? args.file : (args.file ? [args.file] : undefined);
        this.animdb = args.animdb;
        this.skipUnderscores = args['skip-underscores'];
    }
    /**
     * Command-line tool for compiling a skill's keys files into anim files.
     * @method KeysCompiler.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    /**
     * Command-line tool entrypoint
     * @method KeysCompiler.exec
     * @param {any} args Options for controlling the compiler.
     * @return {EventEmitter} Emits 'lintError', 'end' and 'log' event
     */
    static exec(args) {
        const compiler = new KeysCompiler(args);
        compiler.run();
        return compiler;
    }
    /**
     * Main entry point for execution. This loads the keys files and compiles them.
     * @method KeysCompiler.run
     */
    run() {
        this.init().then(() => {
            if (!this.files) {
                this.removeOldFiles();
            }
            this.compileKeysFiles(() => this.emit('end'));
        }).catch((err) => {
            if (`${err}`.endsWith('node_modules/jibo-anim-db-animations/animdb.json')) {
                console.log('Skill has no animation database -- any animation references will cause the program to fail.');
            }
            else {
                console.log('Failed to initialize', err);
            }
            this.emit('end');
        });
    }
    // Load the optional animdb if provided, otherwise the skill's installed animation DB.
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            const animDBpath = this.animdb || path.join(this.skillDir, 'node_modules/jibo-anim-db-animations/animdb.json');
            return jibo_anim_db_1.api.init(null, animDBpath);
        });
    }
    removeOldFiles() {
        for (let ext of ['anim', 'assets']) {
            const keysPath = path.resolve(this.skillDir, 'animations/*' + `*/*.${ext}`);
            rimraf.sync(keysPath);
        }
    }
    compileKeysFiles(cb) {
        //                             The "+" below is a hack -- v -- to keep JSDocs from complaining.
        let keysPath = path.resolve(this.skillDir, 'animations/*' + '*/*.keys');
        const files = this.files || glob.sync(keysPath);
        if (!files.length) {
            this.emit('log', `No keys found in ${this.skillDir}/animations`);
        }
        for (const filename of files) {
            if (this.skipUnderscores && filename.indexOf('/_') !== -1) {
                continue;
            }
            let compositer = new KeyframesCompositer(this.skillDir);
            compositer.setNoWarn();
            let animName = filename.replace('.keys', '.anim');
            if (fs.existsSync(animName)) {
                fs.unlinkSync(animName);
            }
            let keyframes = compositer.compositeKeysfile(filename);
            if (compositer.errors.length !== 0) {
                console.error(`FAILED: ${filename} ${compositer.errors.join(', ')}`);
                continue;
            }
            let anim = keysObjToAnimObject(keyframes, filename);
            fs.writeFileSync(animName, JSON.stringify(anim, null, 4), 'utf8');
            let assets = this.getAssetsFromKeys(keyframes, '', this.skillDir);
            fs.writeFileSync(animName.replace(/anim$/, 'assets'), JSON.stringify(assets, null, 4), 'utf8');
        }
        cb();
    }
    getAssetsFromKeys(data, assetPack, root) {
        let assets = {};
        data.layers.forEach((layer) => {
            layer.keyframes.forEach((keyframe) => {
                let obj;
                const AudioEvent = keyframe.value.AudioEvent;
                const Pixi = keyframe.value.Pixi;
                const Texture = keyframe.value.Texture;
                if (AudioEvent) {
                    obj = {
                        id: AudioEvent.file,
                        type: 'sound',
                        src: AudioEvent.file
                    };
                }
                else if (Pixi) {
                    obj = {
                        id: Pixi,
                        type: 'timeline',
                        //upload: this.upload,
                        src: Pixi
                    };
                }
                else if (jibo_plugins_1.PathUtils.isImage(Texture)) {
                    const src = Texture;
                    obj = {
                        id: src,
                        type: 'texture',
                        src: src
                    };
                }
                if (obj) {
                    assets[obj.id] = obj;
                }
            });
        });
        const assetList = [];
        for (let assetId in assets) {
            assetList.push(assets[assetId]);
        }
        // Add animdb object hash metadata so that the .assets file changes in the case that its
        // containing .keys file only changes in its animdb field (which isn't otherwise picked up)
        assetList.push({
            type: 'metadata',
            animdb: data.animdb ? hash(data.animdb) : hash({})
        });
        return assetList;
    }
}
exports.KeysCompiler = KeysCompiler;
exports.default = KeysCompiler;

},{"./cli":16,"events":undefined,"fs":undefined,"glob":undefined,"jibo-anim-db":undefined,"jibo-keyframes":undefined,"jibo-plugins":undefined,"object-hash":undefined,"path":undefined,"rimraf":undefined}],16:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const minimist = require("minimist");
const chalk = require("chalk");
const KeysCompiler_1 = require("./KeysCompiler");
function usage() {
    console.log(`
usage: node bin/keys-compiler.js [options] /path/to/skill

This program compiles all the .keys files within a skill into .anim and .assets files.

  options:
    -a --animdb /path/to/animdb.json ; default is to use the skill's node_modules/jibo-anim-db-animations DB 
    -u --skip-underscores ; skip files and dirs with leading _ (underscore).
    --file <file to process> ; instead of using the given directory, process the given files.
`);
    process.exit(0);
}
function cli() {
    const argv = process.argv.slice(2);
    if (argv.length === 0) {
        usage();
    }
    const args = minimist(argv, {
        string: ['animdb', 'file'],
        boolean: ['verbose', 'skip-underscores'],
        alias: {
            u: 'skip-underscores',
            a: 'animdb',
            v: 'verbose'
        },
        default: {
            verbose: false
        }
    });
    args.skill = args.skill || args._[0];
    if (!args.skill) {
        console.log(chalk.red('Error: a skill directory must be specified.'));
        usage();
    }
    const result = KeysCompiler_1.KeysCompiler.exec(args);
    result.on('end', (err) => {
        if (err) {
            console.log(err);
        }
        process.exit(0);
    });
    result.on('log', (message) => {
        if (args.verbose) {
            console.log(chalk.green(message));
        }
    });
    result.on('lintError', (err) => {
        console.error(chalk.red(err));
    });
}
exports.default = cli;

},{"./KeysCompiler":15,"chalk":undefined,"minimist":undefined}],17:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const gulp = require("gulp");
const path = require("path");
const tasks = require("./tasks");
const prettyTime = require("pretty-hrtime");
const minimist = require("minimist");
/**
 * Provide the build process for skills.
 * @class JiboDev
 */
class JiboDev {
    /**
     * Command line invokation of the jibo-dev build tools
     * @method JiboDev.cli
     */
    static cli() {
        const task = process.argv[2] || 'build';
        if (!task || !tasks[task]) {
            console.log(chalk.red('Must provide a valid task (build, clean, watch)'));
            return process.exit(1);
        }
        const dev = new JiboDev();
        dev.run(task)
            .then(() => {
            process.exit(0);
        })
            .catch(err => {
            console.log(chalk.red(err));
            process.exit(1);
        });
    }
    /**
     * @constructor
     * @param {String} [dirname] The directory name to run tasks
     * @param {JiboDev~JiboDevConfig} [config] Optional configuration
     */
    constructor(dirname, config) {
        /**
         * Path to the directory of skill
         * @name JiboDev#dirname
         * @type {JiboDev~JiboDevConfig}
         */
        this.dirname = dirname || process.cwd();
        /**
         * The package.json of the skill.
         * @name JiboDev#packageInfo
         * @type {Object}
         */
        const packageInfo = require(path.resolve(this.dirname, 'package.json'));
        // Check for the command
        if (!config) {
            config = packageInfo.config;
        }
        /**
         * The configuration
         * @name JiboDev#config
         * @type {JiboDev~JiboDevConfig}
         */
        this.config = Object.assign({
            name: packageInfo.name,
            standalone: false,
            app: 'src/index.ts',
            bundle: 'index.js',
            map: 'index.js.map',
            tslint: 'tslint.json',
            watch: 'src',
            rules: 'rules',
            rulesSrc: path.join('src', 'rules'),
            flowsSrc: path.join('src', 'flows'),
            clean: []
        }, config || {});
        // Support clean as a glob
        if (typeof this.config.clean === "string") {
            this.config.clean = [this.config.clean];
        }
        /**
         * Commandline arguments
         * @name JiboDev#args
         * @type {Object}
         */
        this.args = minimist(process.argv.slice(2), {
            boolean: [
                'production',
                'verbose'
            ],
            alias: {
                prod: 'production'
            },
            default: {
                production: false,
                verbose: false
            }
        });
        // Manually add the built-in tasks first
        for (const name in tasks) {
            tasks[name](gulp, this);
        }
        gulp.on('task_start', function (e) {
            console.log('Starting', '\'' + chalk.cyan(e.task) + '\'...');
        });
        gulp.on('task_stop', function (e) {
            const time = prettyTime(e.hrDuration);
            console.log('Finished', '\'' + chalk.cyan(e.task) + '\'', 'after', chalk.magenta(time));
        });
        gulp.on('task_err', function (e) {
            const time = prettyTime(e.hrDuration);
            console.log('\'' + chalk.cyan(e.task) + '\'' + ' ' +
                chalk.red('errored after'), chalk.magenta(time));
        });
        gulp.on('task_not_found', function (err) {
            console.log(chalk.red('Task \'' + err.task + '\' is not in your gulpfile'));
            console.log('Please check the documentation for proper gulpfile formatting');
            process.exit(1);
        });
    }
    /**
     * Run as ask command.
     * @method JiboDev#run
     * @param {String} [task='build'] Name of the task: build, clean, watch
     * @return Promise<void> The promise when complete.
     */
    run(task) {
        return new Promise((resolve, reject) => {
            gulp.start(task, (err) => {
                // When watching do nothing
                if (task === 'watch') {
                    if (err) {
                        console.log(chalk.red(err.message));
                    }
                    return;
                }
                // Return result of the task
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
}
exports.default = JiboDev;

},{"./tasks":23,"chalk":undefined,"gulp":undefined,"minimist":undefined,"path":undefined,"pretty-hrtime":undefined}],18:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const browserify = require("browserify");
const gulp = require("gulp");
const gulpif = require("gulp-if");
const buffer = require("vinyl-buffer");
const source = require("vinyl-source-stream");
const composer = require("gulp-uglify/composer");
const uglifyes = require("uglify-es");
const tsify = require("tsify-preprocess");
const mapstraction = require("mapstraction");
const watchify = require("watchify");
const BrowserifyTransform_1 = require("../../browserify/BrowserifyTransform");
function default_1(options, args, watch) {
    return function (done) {
        const prod = args.production;
        const bundler = browserify({
            entries: options.app,
            bundleExternal: false,
            debug: !prod,
            builtins: false,
            extensions: ['.bt', '.flow'],
            cache: {},
            packageCache: {},
            standalone: options.standalone ? options.name : undefined
        });
        // TypeScript options
        bundler.plugin(tsify, {
            removeComments: true,
            target: 'es6',
            preprocess: {
                DEBUG: !prod,
                RELEASE: !!prod,
                PARTNER: !!process.env.PARTNER,
                INTERNAL: !process.env.PARTNER,
            }
        });
        // Install the source-map extractor
        // when build in debug mode
        if (!prod) {
            bundler.plugin(mapstraction, {
                _: [options.map]
            });
        }
        // Install the watch
        if (watch) {
            //as watch will check against many files, we will create our RegExp once
            const regexp = new RegExp('^' + path.resolve(options.watch));
            bundler.plugin(watchify, {
                ignoreWatch: [function (path) {
                        //only watch things in the watch folder, because that's what the option is there for
                        return !regexp.test(path);
                    }],
                poll: true
            });
        }
        // Add the transform for bt and flow files
        bundler.transform(BrowserifyTransform_1.default);
        // See https://www.npmjs.com/package/gulp-uglify#using-a-different-uglifyjs
        const uglify = composer(uglifyes, console);
        // Setup for watching
        function rebundle() {
            const bundle = bundler.bundle()
                .on('error', done)
                .pipe(source(options.bundle))
                .pipe(buffer())
                .pipe(gulpif(prod, uglify()))
                .pipe(gulp.dest('.'));
            if (!watch) {
                bundle.on('end', done);
            }
            return bundle;
        }
        bundler.on('update', rebundle);
        bundler.on('log', console.log.bind(console));
        rebundle();
    };
}
exports.default = default_1;

},{"../../browserify/BrowserifyTransform":4,"browserify":undefined,"gulp":undefined,"gulp-if":undefined,"gulp-uglify/composer":undefined,"mapstraction":undefined,"path":undefined,"tsify-preprocess":undefined,"uglify-es":undefined,"vinyl-buffer":undefined,"vinyl-source-stream":undefined,"watchify":undefined}],19:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const anim_ref_checker_1 = require("../../anim-ref/anim-ref-checker");
const chalk = require("chalk");
function default_1(gulp, dev) {
    gulp.task('anims', function (done) {
        const result = anim_ref_checker_1.default.exec({ skill: '.', verbose: dev.args.verbose });
        result.on('log', (message) => {
            console.log(chalk.green(message));
        });
        result.on('warn', (message) => {
            console.log(chalk.yellow(message));
        });
        result.on('lintError', (err) => {
            console.error(chalk.red(err));
        });
        result.on('end', (err) => {
            done();
        });
    });
}
exports.default = default_1;

},{"../../anim-ref/anim-ref-checker":2,"chalk":undefined}],20:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bundle_1 = require("../plugins/bundle");
function default_1(gulp, dev) {
    gulp.task('build', ['clean', 'lint', 'flows', 'anims'], bundle_1.default(dev.config, dev.args));
}
exports.default = default_1;

},{"../plugins/bundle":18}],21:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rimraf = require("rimraf");
function default_1(gulp, dev) {
    gulp.task('clean', function (done) {
        let deletePromises = [];
        deletePromises.push(new Promise((resolve) => {
            rimraf(dev.config.map, { disableGlob: true }, () => {
                resolve();
            });
        }));
        deletePromises.push(new Promise((resolve) => {
            rimraf(dev.config.bundle, { disableGlob: true }, () => {
                resolve();
            });
        }));
        deletePromises.push(new Promise((resolve) => {
            rimraf(dev.config.rules, { disableGlob: true }, () => {
                resolve();
            });
        }));
        for (let extraFileToDelete of dev.config.clean) {
            deletePromises.push(new Promise((resolve) => {
                rimraf(extraFileToDelete, { disableGlob: true }, () => {
                    resolve();
                });
            }));
        }
        return Promise.all(deletePromises);
    });
}
exports.default = default_1;

},{"rimraf":undefined}],22:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FlowLint_1 = require("../../flows/FlowLint");
const chalk = require("chalk");
function default_1(gulp, dev) {
    gulp.task('flows', ['rules'], function (done) {
        const result = FlowLint_1.default.exec(dev.config.flowsSrc, dev.args.verbose);
        result.on('log', (message) => {
            console.log(chalk.green(message));
        });
        result.on('lintError', (err) => {
            console.error(chalk.red(err));
        });
        result.on('end', (err) => {
            done();
        });
    });
}
exports.default = default_1;

},{"../../flows/FlowLint":28,"chalk":undefined}],23:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const build_1 = require("./build");
exports.build = build_1.default;
const clean_1 = require("./clean");
exports.clean = clean_1.default;
const lint_1 = require("./lint");
exports.lint = lint_1.default;
const watch_1 = require("./watch");
exports.watch = watch_1.default;
const watch_build_1 = require("./watch-build");
exports.watchBuild = watch_build_1.default;
const flows_1 = require("./flows");
exports.flows = flows_1.default;
const rules_1 = require("./rules");
exports.rules = rules_1.default;
const anims_1 = require("./anims");
exports.anims = anims_1.default;

},{"./anims":19,"./build":20,"./clean":21,"./flows":22,"./lint":24,"./rules":25,"./watch":27,"./watch-build":26}],24:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const gulp_tslint_1 = require("gulp-tslint");
const path = require("path");
function default_1(gulp, dev) {
    gulp.task('lint', function () {
        return gulp.src(dev.config.app)
            .pipe(gulp_tslint_1.default({
            configuration: path.resolve(dev.dirname, dev.config.tslint)
        }))
            .pipe(gulp_tslint_1.default.report());
    });
}
exports.default = default_1;

},{"gulp-tslint":undefined,"path":undefined}],25:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rulify_1 = require("../../rules/Rulify");
const chalk = require("chalk");
function default_1(gulp, dev) {
    gulp.task('rules', ['clean'], function (done) {
        const result = Rulify_1.default.exec(dev.config.rulesSrc, dev.config.rules, dev.args.verbose);
        result.on('end', (err) => {
            done();
        });
        result.on('fileError', (err) => {
            console.log(chalk.red(err));
        });
        result.on('compileError', (filename, err) => {
            console.error(chalk.underline(filename));
            console.error(chalk.red(err));
        });
        result.on('log', (message) => {
            console.log(chalk.green(message));
        });
    });
}
exports.default = default_1;

},{"../../rules/Rulify":34,"chalk":undefined}],26:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bundle_1 = require("../plugins/bundle");
function default_1(gulp, dev) {
    gulp.task('watch-build', bundle_1.default(dev.config, dev.args, true));
}
exports.default = default_1;

},{"../plugins/bundle":18}],27:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function default_1(gulp, dev) {
    gulp.task('watch', ['watch-build'], function () {
        gulp.watch(dev.config.watch, ['lint']);
        gulp.watch(dev.config.flowsSrc + "**/*.flow", ['flows']);
        gulp.watch(dev.config.rulesSrc + "**/*.rule", ['rules']);
    });
}
exports.default = default_1;

},{}],28:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const cli_1 = require("./cli");
const events_1 = require("events");
const jibo_flow_core_1 = require("jibo-flow-core");
const glob = require("glob");
/**
 * Statically check a `.flow` file and its dependencies for inconsistencies and missing resources.
 * ```
 * import {FlowLint} from 'jibo-dev';
 * var linter = new FlowLint(__dirname, '/full/path/to/file.flow');
 * linter.run();
 * console.log(linter.getErrorReport());
 * ```
 * @class FlowLint
 * @param {string} projectRoot Path to the skill's project-level directory.
 * @param {string} uri Path to the `.flow` file.
 */
class FlowLint {
    constructor(projectRoot, uri) {
        this.projectRoot = projectRoot;
        this.uri = uri;
        this.procedures = {};
        this.errors = [];
        this.flowRoot = jibo_flow_core_1.FlowRootFactory.create(this.uri);
    }
    /**
     * Command-line tool for validating a directory or single flow file.
     * @method FlowLint.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    /**
     * Command-line tool for validating a directory or single flow file.
     * @method FlowLint.exec
     * @param {String} dir Directory containing flow files
     * @param {Boolean} [verbose=false] If we should load
     * @return {EventEmitter} Emits 'lintError', 'end' and 'log' event
     */
    static exec(dir, verbose) {
        const result = new events_1.EventEmitter();
        process.nextTick(() => {
            const files = glob.sync(path.join(dir, '**/*.flow'));
            if (!files.length) {
                result.emit('end', 'Must provide files to process');
                return;
            }
            files.forEach(function (filename) {
                const uri = path.resolve(filename);
                const linter = new FlowLint(process.cwd(), uri);
                linter.run();
                const report = linter.getErrorReport();
                if (report) {
                    result.emit('lintError', report);
                }
                else if (verbose) {
                    result.emit('log', 'Flow is clean ' + filename);
                }
            });
            result.emit('end');
        });
        return result;
    }
    /**
     * Run the preflight check.
     * @method FlowLint#run
     */
    run() {
        this.loadProcedures();
        this.checkProcedures();
    }
    /**
     * Write discovered errors to the given file.
     * @method FlowLint#writeReportToFile
     * @param {string} filename
     */
    writeReportToFile(filename) {
        let report = this.getErrorReport();
        fs.writeFileSync(filename, `Preflight Report for ${this.uri}\n` + report);
    }
    /**
     * Format all the discovered errors into a report.
     * @method FlowLint#getErrorReport
     * @returns {string} Report of all discovered errors.
     */
    getErrorReport() {
        let report = '';
        let currentProcedure = '';
        for (let error of this.getSortedErrors()) {
            if (error.procedure.name !== currentProcedure) {
                report += `\nErrors found in procedure '${error.procedure.name}':\n`;
                currentProcedure = error.procedure.name;
            }
            if (error.isActivity) {
                let name = error.activity.name ? `'${error.activity.name}' ` : '';
                report += `   ${error.code} ${error.activity.class} ${name}${error.message}\n`;
            }
            else {
                report += `   ${error.code} ${error.message}\n`;
            }
        }
        return report;
    }
    /**
     * Recursively load all the procedures reachable from the preflighted procedure.
     * @method FlowLint#loadProcedures
     * @private
     */
    loadProcedures() {
        let startProcedure = this.flowRoot.getStartProcedure();
        this.loadProcedure(startProcedure);
    }
    /**
     * Load a procedure and all its dependents.
     * @method FlowLint#loadProcedure
     * @private
     * @param {Procedure} procedure
     */
    loadProcedure(procedure) {
        if (procedure.uri in this.procedures) {
            return;
        }
        this.procedures[procedure.uri] = procedure;
        this.loadSubflows(procedure);
        this.loadReturnValues(procedure);
    }
    /**
     * Recursively load all the procedures reachable from the preflighted procedure.
     * @method FlowLint#loadProcedures
     * @private
     * @param {Procedure} procedure
     */
    loadSubflows(procedure) {
        for (let activityId in procedure.activities) {
            let activity = procedure.activities[activityId];
            if (activity.class === 'Flow.Subflow') {
                try {
                    let subflow = this.flowRoot.loadProcedure(procedure.uri, activity.options.subflowId);
                    // Stash the subflow procedure in the activity for convenience at check time.
                    activity.subflow = subflow;
                    this.loadProcedure(subflow);
                }
                catch (e) {
                    this.addActivityError(activity, `references a non-existent flow file '${activity.options.subflowId}'`, 'MISSING SUBFLOW');
                    activity.subflow = undefined;
                }
            }
        }
    }
    /**
     * Gather up all the different transition values that a procedure returns to its parent via Flow.End.
     * @method FlowLint#loadReturnValues
     * @private
     * @param {Procedure} procedure
     */
    loadReturnValues(procedure) {
        procedure.validReturns = [];
        for (let activityId in procedure.activities) {
            let activity = procedure.activities[activityId];
            if (activity.class === 'Flow.End') {
                let transition = undefined;
                // Actually try and execute the getTransition() method of the Flow.End shape.
                try {
                    let getTransition = activity.options.getTransition;
                    if (typeof getTransition === typeof []) {
                        getTransition = getTransition.join(' ');
                    }
                    getTransition = `return (${getTransition});`;
                    transition = (new Function(getTransition))()();
                }
                catch (e) {
                    // Skip anything we can't evaluate statically.
                    continue;
                }
                procedure.validReturns.push(transition || activity.name || '');
            }
        }
    }
    /**
     * Check all the procedures found during loading.
     * @method FlowLint#checkProcedures
     * @private
     */
    checkProcedures() {
        for (let procedure of this.getSortedProcedures()) {
            this.checkProcedure(procedure);
        }
    }
    /**
     * Get a sorted list of all the procedures that have been loaded.
     * @method FlowLint#getSortedProcedures
     * @private
     */
    getSortedProcedures() {
        let procedures = [];
        for (let procedureId in this.procedures) {
            procedures.push(this.procedures[procedureId]);
        }
        procedures.sort((a, b) => {
            return a.name.localeCompare(b.name);
        });
        return procedures;
    }
    /**
     * Check a procedure for consistency and missing resources.
     * @method FlowLint#checkProcedure
     * @private
     * @param {Procedure} procedure
     */
    checkProcedure(procedure) {
        this.checkTransitionCounts(procedure);
        this.checkTransitionValues(procedure);
        this.checkStartValidity(procedure);
        this.checkMimReferences(procedure);
    }
    /**
     * Ensure that correct number of transitions are used for each type of activity.
     * @method FlowLint#checkTransitionCounts
     * @private
     * @param {Procedure} procedure
     */
    checkTransitionCounts(procedure) {
        // check each activity's outbound transitions
        for (let activityId in procedure.activities) {
            let activity = procedure.activities[activityId];
            let transitions = activity.transitions.length + activity.exceptions.length;
            if (transitions === 0
                && activity.class !== 'Flow.Throw'
                && activity.class !== 'Flow.Comment'
                && activity.class !== 'Flow.End') {
                // TODO figure out if we need this:
                //this.addActivityError(activity, 'has no outbound transitions', 'NO OUTBOUND');
            }
            if (transitions !== 1 &&
                (activity.class === 'Flow.Begin'
                    || activity.class === 'Flow.Parallel-Begin'
                    || activity.class === 'Flow.Catch')) {
                this.addActivityError(activity, 'has more than one transition', 'MULTIPLE OUTBOUND');
            }
        }
        // check that each activity's outbound transition values are unique
        let outbounds = {};
        for (let transition of procedure.connections) {
            if (!(transition.frm in outbounds)) {
                outbounds[transition.frm] = {};
            }
            if (transition.value in outbounds[transition.frm]) {
                let activity = procedure.activities[transition.frm];
                this.addActivityError(activity, `has duplicate outbound transitions values: "${transition.value}"`, 'DUP OUTBOUND');
            }
            outbounds[transition.frm][transition.value] = true;
        }
    }
    /**
     * Ensure that values returned from an activity (currently just Subflows) will match one of its outbound transitions.
     * @method FlowLint#checkTransitionValues
     * @private
     * @param {Procedure} procedure
     */
    checkTransitionValues(procedure) {
        for (let activityId in procedure.activities) {
            let activity = procedure.activities[activityId];
            if (activity.class === 'Flow.Subflow' && activity.subflow) {
                for (let retval of activity.subflow.validReturns) {
                    let valid = false;
                    for (let transition of activity.transitions) {
                        if (retval === transition.value || transition.value === '') {
                            // We have found a transition that will match the return value.
                            valid = true;
                            break;
                        }
                    }
                    if (!valid) {
                        this.addActivityError(activity, `has no transition matching the value '${retval}' being returned from ${activity.options.subflowId}`, 'NO MATCHING TRANSITION');
                    }
                }
            }
        }
    }
    /**
     * Ensure that the procedure has the correct number of Flow.Begin activities.
     * @method FlowLint#checkStartValidity
     * @private
     * @param {Procedure} procedure
     */
    checkStartValidity(procedure) {
        let beginCount = 0;
        for (let activityId in procedure.activities) {
            let activity = procedure.activities[activityId];
            if (activity.class === 'Flow.Begin') {
                beginCount++;
            }
        }
        if (beginCount < 1) {
            this.addProcedureError(procedure, "Procedure has no Flow.Begin activity", 'MISSING BEGIN');
        }
        if (beginCount > 1) {
            this.addProcedureError(procedure, "Procedure has more than one Flow.Begin activity", 'MULTIPLE BEGIN');
        }
    }
    /**
     * Ensure that the resources referenced by each mim in the procedure are present.
     * @method FlowLint#checkMimReferences
     * @private
     * @param {Procedure} procedure
     */
    checkMimReferences(procedure) {
        for (let activityId in procedure.activities) {
            let activity = procedure.activities[activityId];
            if (activity.class.startsWith('Mim')) {
                this.checkMim(activity);
            }
        }
    }
    /**
     * Ensure that a mim activity's referenced files are present (mim file, rule file, animation files).
     * @method FlowLint#checkMim
     * @private
     * @param {Activity} activity
     */
    checkMim(activity) {
        // Here we skip mim-checking on activities that don't have a static mim path:
        if (!activity.options.mimPath || activity.options.getMimPath) {
            return;
        }
        let mimPath = path.resolve(this.projectRoot, activity.options.mimPath);
        if (!fs.existsSync(mimPath) || !fs.statSync(mimPath).isFile()) {
            this.addActivityError(activity, `references non-existent Mim file '${mimPath}'`, 'MISSING MIM');
            return;
        }
        let mim;
        try {
            mim = JSON.parse(fs.readFileSync(mimPath, 'utf8'));
        }
        catch (e) {
            this.addActivityError(activity, `references Mim file '${mimPath}' with bad JSON`, 'BAD MIM');
            return;
        }
        // Poor-mans xml parse: Simply look for "path=" attributes in the string and report if they exist as files.
        let takenIds = {};
        for (let prompt of mim.prompts) {
            let match, re = /path=['"]([^'"]*)['"]/gi; // The (inner match) is the filename.
            while (match = re.exec(prompt.prompt)) {
                let animFile = match[1];
                let animPath = path.resolve(this.projectRoot, 'animations', animFile);
                if (!fs.existsSync(animPath)) {
                    this.addActivityError(activity, `prompt animation '${animFile}' in '${activity.options.mimPath}' does not exist`, 'MISSING ANIMATION');
                }
            }
            //also, check to see that every prompt has a unique id
            if (!prompt.prompt_id) {
                this.addActivityError(activity, `A prompt in '${activity.options.mimPath}' does not have an id`, 'MISSING PROMPT ID');
            }
            else if (takenIds[prompt.prompt_id]) {
                this.addActivityError(activity, `Prompt id '${prompt.prompt_id}' in '${activity.options.mimPath}' is not unique`, 'DUPLICATE PROMPT ID');
            }
            else {
                takenIds[prompt.prompt_id] = true;
            }
        }
    }
    /**
     * Accumulate an error referring to a procedure.
     * @method FlowLint#addProcedureError
     * @private
     * @param {Procedure} procedure
     * @param {string} message
     * @param {string} code
     */
    addProcedureError(procedure, message, code) {
        this.errors.push(new Error(procedure, message, code));
    }
    /**
     * Accumulate an error referring to an activity.
     * @method FlowLint#addProcedureError
     * @private
     * @param {Activity} activity
     * @param {string} message
     * @param {string} code
     */
    addActivityError(activity, message, code) {
        this.errors.push(new Error(activity, message, code));
    }
    /**
     * Sort the errors by procedure name, error code, activity type, and activity name.
     * @method FlowLint#getSortedErrors
     * @private
     * @returns {Error[]}
     */
    getSortedErrors() {
        this.errors.sort((a, b) => {
            if (a.procedure.name === b.procedure.name) {
                if (a.isActivity === b.isActivity) {
                    // Sort errors within a procedure by Error Code, Class and Name
                    let akey = a.code + a.activity.class + a.activityName;
                    let bkey = b.code + b.activity.class + b.activityName;
                    return akey.localeCompare(bkey);
                }
                // sort procedure errors before activity errors.
                return +a.isActivity - +b.isActivity;
            }
            return a.procedure.name.localeCompare(b.procedure.name);
        });
        return this.errors;
    }
}
exports.FlowLint = FlowLint;
/**
 * Hold one preflight error.
 * @class Error
 * @private
 */
class Error {
    constructor(activityOrProcedure, message, code) {
        if (activityOrProcedure.procedure !== undefined) {
            this.isActivity = true;
            this.procedure = activityOrProcedure.procedure;
            this.activityName = activityOrProcedure.name;
            this.activity = activityOrProcedure;
        }
        else {
            this.isActivity = false;
            this.procedure = activityOrProcedure;
            this.activityName = '';
            this.activity = null;
        }
        this.message = message;
        this.code = code;
    }
}
exports.default = FlowLint;

},{"./cli":29,"events":undefined,"fs":undefined,"glob":undefined,"jibo-flow-core":undefined,"path":undefined}],29:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const minimist = require("minimist");
const chalk = require("chalk");
const FlowLint_1 = require("./FlowLint");
function cli() {
    const args = minimist(process.argv.slice(2), {
        string: ['dir'],
        boolean: ['verbose'],
        alias: {
            v: 'verbose',
            d: 'dir',
        },
        default: {
            verbose: false,
        }
    });
    const result = FlowLint_1.default.exec(args.dir, args.verbose);
    result.on('end', (err) => {
        if (err) {
            console.log(err);
        }
        process.exit(0);
    });
    result.on('log', (message) => {
        console.log(chalk.green(message));
    });
    result.on('lintError', (err) => {
        console.error(chalk.red(err));
    });
}
exports.default = cli;

},{"./FlowLint":28,"chalk":undefined,"minimist":undefined}],30:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const minimist = require("minimist");
const chalk = require("chalk");
const keys_ref_checker_1 = require("./keys-ref-checker");
function usage() {
    console.log(`
usage: node keys-ref-checker.js [options] /path/to/skill

This program checks all the .keys files within a skill to determine if any of the assets it references are missing.

  options:
    -a --animdb /path/to/animdb.json ; default is to use the skill's node_modules/jibo-anim-db-animations DB;
    -u --skip-underscores ; ignore all directories with a leading underscore .
    -v --verbose ; print details as files are checked.
    -F --fix-nonunique-layers ; find any duplicate layerIDs across the animations and create a new unique ID for them.
    --no-report ; don't generate a reference check report.
`);
    process.exit(0);
}
function cli() {
    const argv = process.argv.slice(2);
    if (argv.length === 0) {
        usage();
    }
    const args = minimist(argv, {
        string: ['skill', 'animdb'],
        boolean: ['verbose', 'no-report', 'skip-underscores', 'fix-nonunique-layers'],
        alias: {
            'check-skill': 'skill',
            s: 'skill',
            a: 'animdb',
            v: 'verbose',
            u: 'skip-underscores',
            F: 'fix-nonunique-layers',
            R: 'no-report'
        },
        default: {
            verbose: false,
            'no-report': false,
            'fix-nonunique-layers': false,
            'skip-underscores': false
        }
    });
    args.skill = args.skill || args._[0];
    if (!args.skill) {
        console.log(chalk.red('Error: a skill directory must be specified.'));
        usage();
    }
    const result = keys_ref_checker_1.KeysRefChecker.exec(args);
    console.log('Checking Keys references...');
    result.on('end', (err) => {
        if (err) {
            console.log(err);
        }
        console.log('Done!');
        process.exit(0);
    });
    result.on('status', (message) => {
        console.log(message);
    });
    result.on('log', (message) => {
        if (args.verbose) {
            console.log(chalk.green(message));
        }
    });
    result.on('lintError', (err) => {
        console.error(chalk.red(err));
    });
}
exports.default = cli;

},{"./keys-ref-checker":31,"chalk":undefined,"minimist":undefined}],31:[function(require,module,exports){
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const cli_1 = require("./cli");
const events_1 = require("events");
const glob = require("glob");
const jibo_anim_db_1 = require("jibo-anim-db");
const chalk = require("chalk");
const uuid = require("uuid");
const JiboKeyframeInfo = require('jibo-keyframes').JiboKeyframeInfo;
const resolve = require('resolve');
/**
 * Statically check the validity of various asset references in all the keys files in a skill's animation folder.
 * ```
 * import {KeysRefChecker} from 'jibo-dev';
 * ```
 * @class KeysRefCheckerer
 * @param {string} projectRoot Path to the skill's project-level directory.
 */
class KeysRefChecker extends events_1.EventEmitter {
    constructor(args) {
        super();
        this.args = args;
        this.errors = [];
        this.noReport = args['no-report'];
        this.skillDir = args.skill;
        this.animdb = args.animdb;
        this.skipUnderscores = args['skip-underscores'];
        this.verbose = args.verbose;
        this.fixIds = args['fix-nonunique-layers'];
    }
    /**
     * Command-line tool for validating the asset references in a skill's keys files.
     * @method KeysRefChecker.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    /**
     * Command-line tool entrypoint
     * @method KeysRefChecker.exec
     * @param {any} args Options for controlling the checker.
     * @return {EventEmitter} Emits 'lintError', 'end' and 'log' event
     */
    static exec(args) {
        const refChecker = new KeysRefChecker(args);
        refChecker.run();
        return refChecker;
    }
    /**
     * Main entry point for execution. This loads the required files and proceeds through all the checks.
     * @method KeysRefChecker.run
     */
    run() {
        // Init animation DB
        this.init().catch((err) => {
            if (`${err}`.indexOf('jibo-anim-db-animations') !== -1) {
                console.log(chalk.red('Skill has no animation database -- animDB references will fail when checked.'));
            }
            else {
                console.log(chalk.red('Failed to initialize animation database -- animDB references will fail when checked'), err);
            }
        }).then(() => {
            this.loadRefsFromSkill(() => this.runChecks());
        });
    }
    /**
     * Write discovered errors to the given file.
     * @method KeysRefChecker#writeReportToFile
     * @param {string} filename
     */
    writeReportToFile(filename, title) {
        const report = this.getErrorReport();
        fs.writeFileSync(filename, `AnimDB Reference Report for ${title}\n` + report);
    }
    /**
     * Format all the discovered errors into a report.
     * @method KeysRefChecker#getErrorReport
     * @returns {string} Report of all discovered errors.
     */
    getErrorReport() {
        let lines = [];
        for (const error of this.errors) {
            let line = `${error.code}: ${error.message} in ${error.location}`;
            lines.push(line);
        }
        lines.sort();
        lines.push(''); // force all lines to print
        let report = '';
        let count = 0;
        let prior = lines[0];
        // Here we print each line with a tally in front of duplicated lines.
        for (const line of lines) {
            if (line !== prior) {
                report += `${count === 1 ? ' ' : count} ${prior}\n`;
                prior = line;
                count = 0;
            }
            count++;
        }
        return report;
    }
    getFancyErrorReport() {
        let report = '';
        let location = '';
        // Group all the errors in the same file/activity together
        this.errors.sort((a, b) => a.location.localeCompare(b.location));
        for (const error of this.errors) {
            if (error.location !== location) {
                report += `\nErrors found in '${error.location}':\n`;
                location = error.location;
            }
            report += `${error}\n`;
        }
        return report;
    }
    /**
     * Accumulate an error referring to an activity.
     * @method KeysRefChecker#addProcedureError
     * @private
     * @param {Activity} activity
     * @param {string} message
     * @param {string} code
     */
    addError(location, message, code) {
        this.errors.push(new Error(location, message, code));
    }
    resolve(wanted) {
        const dir = resolve.sync(wanted, { basedir: this.skillDir });
        return path.dirname(dir);
    }
    runChecks() {
        const report = this.getErrorReport();
        if (report) {
            this.emit('lintError', report);
        }
        this.emit('end');
    }
    // Load the optional animdb if provided, otherwise the skill's installed animation DB.
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            let animDBpath = this.animdb;
            if (!animDBpath) {
                let adaPath = this.resolve('jibo-anim-db-animations');
                animDBpath = adaPath && path.resolve(adaPath, 'animdb.json');
            }
            return jibo_anim_db_1.api.init(null, animDBpath);
        });
    }
    loadRefsFromSkill(cb) {
        const checker = new KeysChecker(this, this.skillDir);
        checker.checkSkillRefs();
        this.emit('status', `Checked ${checker.checkedFiles} files`);
        if (checker.fixedFiles) {
            this.emit('status', `Fixed ${checker.fixedFiles} files`);
        }
        cb();
    }
}
exports.KeysRefChecker = KeysRefChecker;
class KeysChecker {
    constructor(refChecker, dir) {
        this.refChecker = refChecker;
        this.checkedFiles = 0;
        this.fixedFiles = 0;
        this.extracted = {};
        this.layerIds = {};
        this.skillDir = path.resolve(dir);
        this.errors = [];
    }
    /**
     * Check all the keys files in the animations folder of the skill.
     */
    checkSkillRefs() {
        let keysPath = path.resolve(this.skillDir, 'animations/*' + '*/*.keys');
        const files = glob.sync(keysPath);
        if (!files.length) {
            this.refChecker.emit('log', `No keys found in ${this.skillDir}/animations`);
        }
        for (const filename of files) {
            this.checkFileRef(filename);
        }
    }
    /**
     * Check all the layers of the given keys file to ensure all the referenced assets are present.
     * @param filename
     */
    checkFileRef(filename) {
        if (filename in this.extracted || (this.refChecker.skipUnderscores && filename.indexOf('/_') !== -1)) {
            return;
        }
        this.extracted[filename] = true;
        if (this.refChecker.verbose) {
            this.refChecker.emit('log', `Checking ${filename}`);
        }
        let keys;
        try {
            keys = JSON.parse(fs.readFileSync(filename, 'utf8'));
        }
        catch (e) {
            this.refChecker.addError(filename, `references Keys file '${filename}' with bad JSON`, 'JSON-PARSE');
            return;
        }
        // Bring the keys file up to the latest spec.
        JiboKeyframeInfo.onLoad(keys);
        this.refChecker.emit('checkFileRefs', filename);
        this.uniqifyLayerIds(filename, keys);
        this.checkedFiles++;
        let checked = {};
        for (let layer of keys.layers) {
            switch (layer.type) {
                case 'Overlay Texture':
                case 'Eye Texture':
                    for (let keyframe of layer.keyframes) {
                        let src = keyframe.value.Texture;
                        if (src === '') {
                            // blank texture paths are legal.
                            continue;
                        }
                        let apath = this.resolveAssetPath(src);
                        this.checkPath(checked, { apath, filename, type: layer.type, src });
                    }
                    break;
                case 'Audio Event':
                    for (let keyframe of layer.keyframes) {
                        let src = keyframe.value.AudioEvent.file;
                        let apath = this.resolveAssetPath(src, 'audio');
                        this.checkPath(checked, { apath, filename, type: layer.type, src });
                    }
                    break;
                case 'Pixi':
                    for (let keyframe of layer.keyframes) {
                        let src = keyframe.value.Pixi;
                        let apath = this.resolveAssetPath(src, 'timelines');
                        if (this.checkPath(checked, { apath, filename, type: layer.type, src })) {
                            let refs = this.getPixiReferences(apath);
                            for (let ref of refs) {
                                let apath = this.resolveAssetPath(ref, 'timelines');
                                this.checkPath(checked, { apath, filename, type: 'Pixi-ref', src: ref });
                                if (ref.endsWith('.json')) {
                                    let jref = this.getJsonReference(apath);
                                    if (jref) {
                                        let aref = path.resolve(path.dirname(apath), jref);
                                        this.checkPath(checked, { apath: aref, filename, type: 'Pixi-image-ref', src: jref });
                                    }
                                }
                            }
                        }
                    }
                    break;
                case 'Reference':
                    for (let keyframe of layer.keyframes) {
                        let apath = this.getReferencedPath(filename, keyframe);
                        if (apath) {
                            let mode = keyframe.value.ReferenceMode.value;
                            if (this.checkPath(checked, {
                                apath,
                                filename,
                                type: 'Reference',
                                mode,
                                src: keyframe.value.KeysFileReference[mode]
                            })) {
                                this.checkFileRef(apath);
                            }
                        }
                    }
                    break;
            }
        }
    }
    /**
     * Convert the anim reference in the given keys file, at the given keyframe, into an absolute file path.
     */
    getReferencedPath(filename, keyframe) {
        let mode = keyframe.value.ReferenceMode.value;
        let src = keyframe.value.KeysFileReference[mode];
        let apath;
        if (mode === 'name') {
            let anim;
            try {
                anim = jibo_anim_db_1.api.getAnimByName(src);
            }
            catch (e) { }
            if (anim) {
                apath = path.resolve(anim.resourceRoot, anim.meta.path);
            }
            else {
                this.refChecker.addError(filename, `Animation name "${src}" not found in animation DB`, 'MISSING-ANIM');
            }
        }
        else {
            apath = path.resolve(path.dirname(filename), src);
        }
        return apath;
    }
    /**
     * Check for the existence of a non-blank path (only once) and report an error if it's missing.
     */
    checkPath(checked, info) {
        let apath = info.apath;
        if (apath.indexOf('/_') !== -1 && this.refChecker.skipUnderscores === true) {
            this.refChecker.addError(info.filename, `${apath}`, 'UNDERSCORE REF');
        }
        if (!(apath in checked)) {
            let success = true;
            if (apath && (!fs.existsSync(apath) || !fs.statSync(apath).isFile())) {
                success = false;
                this.refChecker.addError(info.filename, `'${info.src} (full: ${apath})'`, `missing ${info.type}`.toUpperCase());
            }
            checked[apath] = success;
            this.refChecker.emit('checkPath', apath, success);
        }
        return checked[apath];
    }
    /*
     * Fish out references to json or png assets within Javascript files on lines that look like this:
     *       "emoji_gas_atlas_1": "images/emoji_gas_atlas_1.json"
     */
    getPixiReferences(filename) {
        this.refChecker.emit('checkPixiRefs', filename);
        let refs = [];
        let data;
        try {
            data = fs.readFileSync(filename, 'utf8');
        }
        catch (e) {
            console.log('failed to open file', filename);
            return refs;
        }
        for (let line of data.split('\n')) {
            line = line.trim();
            if (line.indexOf('"images/') !== -1) {
                let ref = line.replace(/^.*"(images\/.*?)".*$/, '$1');
                refs.push(ref);
            }
        }
        return refs;
    }
    // Get the (possibly non-existent) reference to an image in a Pixi JSON file.
    getJsonReference(uri) {
        this.refChecker.emit('checkJsonRefs', uri);
        let body;
        try {
            body = JSON.parse(fs.readFileSync(uri, 'utf8'));
        }
        catch (e) {
            this.refChecker.addError(uri, `Pixi JSON file '${uri}' could not be loaded and parsed`, 'JSON-PARSE');
        }
        let ref = body && body.meta && body.meta.image;
        return ref;
    }
    // Convert possible asset-pack style reference into full path
    resolveAssetPath(src, middle) {
        let parts = src.split('://');
        middle = middle || '';
        if (parts.length > 1) {
            let dir = this.refChecker.resolve(parts[0]);
            let final = path.resolve(dir, middle, parts[1]);
            return final;
        }
        else {
            return path.resolve(this.skillDir, middle, src);
        }
    }
    uniqifyLayerIds(filename, keys) {
        if (this.refChecker.fixIds) {
            let dirty = false;
            for (let layer of keys.layers) {
                if (layer.type === 'Reference' && layer.id in this.layerIds) {
                    dirty = true;
                    layer.id = uuid.v4().replace(/-/g, "");
                }
                this.layerIds[layer.id] = true;
            }
            if (dirty) {
                this.fixedFiles++;
                fs.writeFileSync(filename, JSON.stringify(keys, null, '    '), 'utf8');
                this.refChecker.emit('log', `Making unique layer IDs in ${filename}`);
            }
        }
    }
}
/**
 * Hold one preflight error.
 * @class Error
 * @private
 */
class Error {
    constructor(location, message, code) {
        this.location = location;
        this.message = message;
        this.code = code;
    }
    toString() {
        return `   ${this.code} ${this.message}`;
    }
}
exports.default = KeysRefChecker;

},{"./cli":30,"chalk":undefined,"events":undefined,"fs":undefined,"glob":undefined,"jibo-anim-db":undefined,"jibo-keyframes":undefined,"path":undefined,"resolve":undefined,"uuid":undefined}],32:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const events_1 = require("events");
const glob = require("glob");
const cli_1 = require("./cli");
const CATEGORY_HAS_INDEX = {
    'Entry-Core': false,
    Errors: true,
    Response: false
};
const SUBCATEGORY_ABBREVIATION = {
    Q: 'QN',
    AN: 'AN',
    NM: 'NM',
    NI: 'NI',
    HoldReturn: 'HR',
    Verbose: 'VB',
    Truncated: 'TC',
    Thanks: 'TX'
};
/**
 * Statically check a `.flow` file and its dependencies for inconsistencies and missing resources.
 * ```
 * import {MimFixer} from 'jibo-dev';
 * var fixer = new MimFixer(path.join(__dirname, 'mims'));
 * linter.run();
 * console.log(linter.getErrorReport());
 * ```
 * @class MimFixer
 */
class MimFixer extends events_1.EventEmitter {
    /**
     * Command-line tool for validating a directory or single flow file.
     * @method FlowLint.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    constructor() {
        super();
    }
    /**
     * Processes a folder of .mim files or a single file.
     * @method MimFixer#run
     * @param {string} target Path to the skill's folder to fix mims found within, or path to a specific .mim file to run on.
     * @param {boolean} [verbose=false] If extended information should be logged.
     */
    run(target, verbose) {
        if (!target) {
            this.emit('error', 'Must supply a target file or folder');
            this.emit('end');
            return;
        }
        this.target = target;
        this.verbose = !!verbose;
        let getFileNames;
        if (target.endsWith('.mim')) {
            getFileNames = Promise.resolve([target]);
        }
        else {
            getFileNames = new Promise((resolve, reject) => {
                glob(path.join(target, '**/*.mim'), (err, matches) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(matches);
                });
            });
        }
        getFileNames.then((filenames) => {
            if (!filenames.length) {
                this.emit('end', 'No .mim files found');
            }
            //go run all the mims in sequence
            let handleMims = this.processMim(filenames.shift());
            filenames.forEach((file) => {
                handleMims = handleMims.then(() => {
                    return this.processMim(file);
                });
            });
            //return the long promise chain we just set up
            return handleMims;
        }).catch((err) => {
            this.emit('error', 'Fatal error when processing .mim files: ' + err.message || err);
        }).then(() => {
            this.emit('end');
        });
    }
    processMim(filename) {
        return new Promise((resolve, reject) => {
            fs.readFile(filename, 'utf8', (err, text) => {
                if (err) {
                    this.emit('error', `Error reading file ${filename}: ` + err.message);
                    return resolve();
                }
                let json;
                try {
                    json = JSON.parse(text);
                }
                catch (err) {
                    this.emit('error', `Error parsing file ${filename}: ` + err.message);
                    return resolve();
                }
                if (!json.prompts || !json.prompts.length) {
                    this.emit('error', `File ${filename} has no prompts!`);
                    return resolve();
                }
                const mimName = path.basename(filename, '.mim');
                const promptCountByType = {};
                const prompts = json.prompts;
                for (let i = 0, len = prompts.length; i < len; ++i) {
                    const prompt = prompts[i];
                    let promptType = prompt.prompt_sub_category;
                    const hasIndex = CATEGORY_HAS_INDEX[prompt.prompt_category];
                    if (hasIndex) {
                        promptType += '_' + prompt.index;
                    }
                    if (!promptCountByType[promptType]) {
                        promptCountByType[promptType] = 0;
                    }
                    const promptNum = ++promptCountByType[promptType];
                    prompt.prompt_id = `${mimName}_${SUBCATEGORY_ABBREVIATION[prompt.prompt_sub_category]}_${hasIndex ? prompt.index + '_' : ''}${promptNum < 10 ? '0' + promptNum : promptNum}`;
                }
                fs.writeFile(filename, JSON.stringify(json, null, '\t'), (err) => {
                    if (err) {
                        this.emit('error', `Error writing file ${filename}: ` + err.message);
                        return resolve();
                    }
                    if (this.verbose) {
                        this.emit('log', `Processed ${filename}`);
                    }
                    resolve();
                });
            });
        });
    }
}
exports.default = MimFixer;

},{"./cli":33,"events":undefined,"fs":undefined,"glob":undefined,"path":undefined}],33:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const minimist = require("minimist");
const chalk = require("chalk");
const MimFixer_1 = require("./MimFixer");
function cli() {
    const args = minimist(process.argv.slice(2), {
        string: ['target'],
        boolean: ['verbose'],
        alias: {
            v: 'verbose',
            t: 'target',
        },
        default: {
            verbose: false,
        }
    });
    const fixer = new MimFixer_1.default();
    fixer.on('end', (err) => {
        if (err) {
            console.log(err);
        }
        process.exit(0);
    });
    fixer.on('log', (message) => {
        console.log(chalk.green(message));
    });
    fixer.on('error', (err) => {
        console.error(chalk.red(err));
    });
    fixer.run(args.target, args.verbose);
}
exports.default = cli;

},{"./MimFixer":32,"chalk":undefined,"minimist":undefined}],34:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const findRoot = require("find-root");
const path = require("path");
const fs = require("fs");
const cli_1 = require("./cli");
const mkdirp = require("mkdirp");
const glob = require("glob");
const events_1 = require("events");
const normalize = require("normalize-path");
const crypto = require("crypto");
const os = require("os");
const moment = require("moment");
let handleId = 1;
/**
 * Convert a `.rule` file to a `.fst` file.
 * ```
 * import {Rulify} from 'jibo-dev';
 * const fstPath = Rulify.sync('/path/to/file.rule');
 * console.log("FST file saved to: ", fstPath);
 * ```
 * @class Rulify
 */
class Rulify {
    /**
     * Synchronous version of rulify.
     * @method Rulify.sync
     * @param {String} src Either the contents of the rule or the path to the source (.rule or .fst)
     * @param {String} output Name of file to write, will default to src with `.fst` extension.
     * @return {String} The output file saved.
     */
    static sync(src, output) {
        let srcIsFst = /\.fst$/.test(src);
        const data = (/\.rule$/.test(src) || srcIsFst) ?
            fs.readFileSync(src, 'utf8') : src;
        output = output || Rulify.defaultOutput(src);
        if (srcIsFst) {
            // We are merely re-outputting the source .fst file, maybe to a new location
            fs.writeFileSync(output, data);
        }
        else {
            // compile .rule data
            fs.writeFileSync(output, Rulify.toBuffer(data));
        }
        return output;
    }
    /**
     * Asynchronous version of rulify.
     * @method Rulify.async
     * @param {String} src Either the contents of the rule or the path to the source (.rule or .fst)
     * @param {String} [output] Name of file to write, will default to src with `.fst` extension.
     * @param {Rulify~Callback} callback Called when the conversion completes.
     */
    static async(src, output, callback) {
        if (typeof output === "function") {
            callback = output;
            output = null;
        }
        let srcIsFst = /\.fst$/.test(src);
        if (srcIsFst || /\.rule$/.test(src)) {
            output = output || Rulify.defaultOutput(src);
            fs.readFile(src, 'utf8', function (err, data) {
                if (err) {
                    return callback(err);
                }
                if (srcIsFst) {
                    // No need to compile .fst, just send straight out
                    fs.writeFile(output, data, function (e) {
                        callback(e, output);
                    });
                }
                else {
                    Rulify.fromString(data, output, callback);
                }
            });
        }
        else {
            Rulify.fromString(src, output, callback);
        }
    }
    /**
     * Execute the commandline interface
     * @method Rulify.cli
     * @private
     */
    static cli() {
        cli_1.default();
    }
    /**
     * Get the reference the NLU parser.
     * @method Rulify.nlu
     * @return {NLU} The nlu parser.
     */
    static get nlu() {
        if (!Rulify._nlu) {
            let nlu;
            const parserPath = process.versions.electron ? 'parser-electron' : 'parser-node';
            const root = path.join(findRoot(__dirname));
            const nodePath = path.join(root, parserPath, 'build', 'Release', 'jsjibonlu.node');
            const dataDirectory = path.join(root, parserPath, 'deps', 'data');
            try {
                nlu = require(nodePath);
                nlu.set_data_dir(dataDirectory);
            }
            catch (e) {
                console.error(e);
                nlu = undefined;
            }
            Rulify._nlu = nlu;
        }
        return Rulify._nlu;
    }
    /**
     * Get the reference the NLU parser.
     * @method Rulify.exec
     * @param {Array<String>} files List of files to load
     * @param {String} input Input directory
     * @param {String} output Output directory
     * @param {Boolean} [verbose=false] Extra logging
     * @return {EventEmitter} Emitter dispatches 'end', 'fileError', 'compileError', and 'log'
     */
    static exec(input, output, verbose) {
        const result = new events_1.EventEmitter();
        process.nextTick(() => {
            const files = glob.sync(path.join(input, '**/*.rule'));
            if (!files.length) {
                result.emit('end', 'Must provide files to process');
                return;
            }
            let outFiles = [];
            let errors = this.processRulesFiles(files, outFiles, input, output, true);
            if (errors) {
                result.emit(errors.errorType, errors.errorValue, errors.err);
            }
            else {
                if (verbose) {
                    for (let i = 0; i < outFiles.length; i++) {
                        result.emit('log', 'Saved \'' + files[i] + '\' to \'' + outFiles[i] + '\'');
                    }
                }
                result.emit('end');
            }
        });
        return result;
    }
    /**
     * Compiles a skill's launch rules into a single .fst file. The launch rules are listed in
     * the "rulesToBuild" array of the skill's package.json. The name of the .fst file to output
     * to is found in "launchRule" (it should default to something if launchRule is not found or
     * is an array -- to do)
     *
     * @method Rulify.compileLaunchRules
     * @param {String} packagePath input directory, containing package.json
     * @param {Boolean} [verbose=false] Extra logging
     * @param {Boolean} [forceCompile=false] Force compile even if cache exists
     * @return {EventEmitter} Emitter dispatches 'end', 'fileError', 'compileError', and 'log'
     */
    static compileLaunchRules(packagePath, verbose, forceCompile) {
        packagePath = packagePath || process.cwd();
        packagePath = path.resolve(packagePath);
        const result = new events_1.EventEmitter();
        // Track how long compile takes
        let startTime = moment();
        let endTime, elapsedTime;
        // Check if compiled rules cache exists
        const cacheMapFile = path.join(os.tmpdir(), 'fstCache.json');
        const cacheFile = path.join(os.tmpdir(), 'fstCache');
        let cacheMap = {};
        // If metadata file doesn't exist or is empty, need to create it
        if (fs.existsSync(cacheMapFile) && fs.statSync(cacheMapFile).size) {
            // Also need to protect against corrupted metadata
            try {
                cacheMap = JSON.parse(fs.readFileSync(cacheMapFile, 'utf8'));
            }
            catch (err) {
                console.log("Bad rules cache metadata, recreating");
                fs.writeFileSync(cacheMapFile, "");
                cacheMap = { 'digest': null };
            }
        }
        else {
            // if not, then create one
            fs.writeFileSync(cacheMapFile, "");
            cacheMap = { 'digest': null };
        }
        // Derive package object from package.json
        let pkg;
        try {
            pkg = require(path.join(packagePath, 'package.json'));
        }
        catch (e) {
            result.emit('end', 'Couldn\'t process package.json');
            return result;
        }
        // Where should final file go?
        let outputPath = 'launchRule.fst'; // default name, unless another specified
        if (typeof pkg.jibo.launchRule === "string") {
            outputPath = pkg.jibo.launchRule;
        }
        outputPath = path.resolve(packagePath, outputPath);
        process.nextTick(() => {
            // Get set of paths to rules from package.json
            const rules = pkg.jibo.rulesToBuild;
            if (!rules.length) {
                result.emit('end', 'Must provide files to process');
                return;
            }
            let digestUnion = '';
            for (let rule of rules) {
                const uri = path.resolve(packagePath, rule);
                let hash = crypto.createHash('sha256');
                hash.update(fs.readFileSync(uri, 'utf8'));
                digestUnion += hash.digest('hex');
            }
            if (!forceCompile && (digestUnion === cacheMap['digest'] && fs.existsSync(cacheFile))) {
                console.log("\n----------------");
                console.log("Using cached launch rules FST");
                console.log("----------------\n");
                let cachedData = fs.readFileSync(cacheFile);
                fs.writeFileSync(outputPath, cachedData);
                endTime = moment();
                elapsedTime = moment.duration(endTime.diff(startTime));
            }
            else {
                console.warn("\n----------------");
                console.warn("Cached launch rules FST unavailable or outdated.");
                console.warn("It may take a minute or so to recompile");
                console.warn("----------------\n");
                cacheMap['digest'] = digestUnion;
                fs.writeFileSync(cacheMapFile, JSON.stringify(cacheMap));
                let outUris = [];
                let errors = this.processRulesFiles(rules, outUris, packagePath, packagePath, false);
                if (errors) {
                    result.emit(errors.errorType, errors.errorValue, errors.err);
                }
                else {
                    // We have successfully compiled each .rule file to a .fst file in memory
                    if (verbose) {
                        for (let i = 0; i < outUris.length; i++) {
                            result.emit('log', 'Saved \'' + rules[i] + '\' to \'' + outUris[i] + '\'');
                        }
                    }
                    let unionHandle = null;
                    try {
                        // Compile all the .fst files in memory into a master .fst file
                        unionHandle = Rulify.compileUnion(outUris);
                    }
                    catch (err) {
                        result.emit('compileError', 'unionize error', err.message);
                    }
                    if (unionHandle) {
                        // Output master .fst file to disk
                        let fst = Rulify.nlu.read_fst_from_uri(unionHandle);
                        fst.save_fst(outputPath);
                        // Create cached fst file
                        let cachedData = fs.readFileSync(outputPath);
                        fs.writeFileSync(cacheFile, cachedData);
                        if (verbose) {
                            result.emit('log', 'Generated union file \'' + outputPath + '\'');
                        }
                    }
                    //endTime = Date.now();
                    //console.log("Cache copy took: " + new Date().setMilliseconds(endTime-startTime).toString());
                    endTime = moment();
                    elapsedTime = moment.duration(endTime.diff(startTime));
                }
            }
            if (elapsedTime.asSeconds() < 60) {
                console.log("Compile took: " + elapsedTime.asSeconds() + " seconds");
            }
            else {
                console.log("Compile took: " + elapsedTime.asMinutes() + " minutes");
            }
            result.emit('end');
        });
        return result;
    }
    /**
     * Given a set of .rule files, compile each of them in memory into a .fst file.
     * When this is done, the method will return an array of URIs which contain either
     * the paths to the .fst files on disk or handles to their instances in memory.
     *
     * @param  {string[]}     files        Array of paths to rule files, which are relative
     * @param  {string[]}     outUris      Array that will receive set of outputted URIs
     * @param  {string}       inputFolder  Base input folder
     * @param  {string}       outputFolder Base output folder (relevant only when outputting
     *                                     .fst files to disk)
     * @param  {boolean}      outputToDisk If true, .fst files will be outputted to disk;
     *                                     otherwise they exist in memory, referred to by
     *                                     their handles.
     * @return {RulifyErrors}              error structure, if any errors. Otherwise, return
     *                                     null
     */
    static processRulesFiles(files, outUris, inputFolder, outputFolder, outputToDisk) {
        // Make directory in case
        mkdirp.sync(path.resolve(outputFolder));
        files.forEach(function (filename) {
            // if the input file a .fst, not a .rule
            let isFstFile = (filename.indexOf('.fst') !== -1);
            if (outputToDisk) {
                const uri = path.resolve(filename);
                // Make sure the file is valid
                try {
                    fs.statSync(uri);
                }
                catch (e) {
                    return {
                        errorType: 'fileError',
                        errorValue: "File does not exist '" + filename + "'",
                        err: null
                    };
                }
                // If it's a .fst file, it would be redundant to compile it
                let outputName;
                // Rebase to the input folder
                // (It's fine if the input file is already a .fst)
                if (inputFolder) {
                    outputName = filename.replace(/\.rule$/, '.fst')
                        .replace(normalize(inputFolder) + '/', '');
                }
                else {
                    outputName = path.basename(filename).replace(/\.rule$/, '.fst');
                }
                const outputFile = path.resolve(outputFolder, outputName);
                const outputDir = path.dirname(outputFile);
                // Make the output directory
                mkdirp.sync(outputDir);
                // We compile the .rule file into a .fst file on disk
                try {
                    Rulify.sync(uri, outputFile);
                    outUris.push(outputFile);
                }
                catch (err) {
                    // Handle compile errors
                    return {
                        errorType: 'compileError',
                        errorValue: filename,
                        err: err
                    };
                }
            }
            else {
                let outUri = null;
                try {
                    let inFile = path.resolve(inputFolder, filename);
                    // We compile the .rule file into memory and hang onto the handle for later
                    outUri = `handle:${++handleId}`;
                    if (isFstFile) {
                        Rulify.nlu.load_fst_from_fst_file(inFile, outUri);
                    }
                    else {
                        const data = fs.readFileSync(inFile, 'utf8');
                        Rulify.nlu.compile_fst_from_text(data, outUri);
                    }
                }
                catch (err) {
                    // Handle compile errors
                    return {
                        errorType: 'compileError',
                        errorValue: filename,
                        err: err
                    };
                }
                outUris.push(outUri);
            }
        });
        return null;
    }
    /**
     * Convert file to .fst
     * @method Rulify.defaultOutput
     * @private
     * @param {String} file Convert
     * @return {String} File .fst path
     */
    static defaultOutput(file) {
        return file.replace(/\.rule$/, '.fst');
    }
    /**
     * Write the buffer from data string
     * @function fromString
     * @private
     * @param {String} data contents of .rule file
     * @param {String} output Path to file to save
     * @param {Rulify~Callback} callback Calls when write completes.
     */
    static fromString(data, output, callback) {
        try {
            const buffer = Rulify.toBuffer(data);
            fs.writeFile(output, buffer, function (err) {
                callback(err, output);
            });
        }
        catch (err) {
            callback(err, output);
        }
    }
    /**
     * Create a buffer from data
     * @function toBuffer
     * @private
     * @param {String} data Contents of .rule file
     */
    static toBuffer(data) {
        try {
            const fst = Rulify.nlu.compile_fst_from_text(data, `handle:${++handleId}`);
            return fst.to_buffer();
        }
        catch (e) {
            let error = JSON.parse(e.message);
            let message = `${error.exception}\n`;
            message += `${error.info.msg}\n`;
            if (error.info.line) {
                message += `line: ${error.info.line} `;
                if (error.info.column) {
                    message += `column: ${error.info.column} `;
                }
            }
            throw new Error(message);
        }
    }
    /**
     * Give a set of uris to compiled .fst files (either the paths of the files on
     * disk or handles to their in-memory instances), compile a master .fst file.
     * @param  {string[]} uris Array of uris
     */
    static compileUnion(uris) {
        try {
            let unionHandle = `handle:${++handleId}`;
            Rulify.nlu.union_fst(uris, unionHandle);
            return unionHandle;
        }
        catch (e) {
            let error = JSON.parse(e.message);
            let message = `${error.exception}\n`;
            message += `${error.info.msg}\n`;
            if (error.info.line) {
                message += `line: ${error.info.line} `;
                if (error.info.column) {
                    message += `column: ${error.info.column} `;
                }
            }
            throw new Error(message);
        }
    }
}
exports.default = Rulify;

},{"./cli":35,"crypto":undefined,"events":undefined,"find-root":undefined,"fs":undefined,"glob":undefined,"mkdirp":undefined,"moment":undefined,"normalize-path":undefined,"os":undefined,"path":undefined}],35:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const minimist = require("minimist");
const chalk = require("chalk");
const Rulify_1 = require("./Rulify");
function cli() {
    const args = minimist(process.argv.slice(2), {
        alias: {
            "v": "verbose",
            "i": "input",
            "o": "output",
            "f": "force"
        },
        boolean: ['update', 'verbose', 'force', 'launchRules'],
        string: ['output', 'input']
    });
    let result;
    if (args.launchRules) {
        result = Rulify_1.default.compileLaunchRules(args.input, args.verbose, args.force);
    }
    else {
        result = Rulify_1.default.exec(args.input, args.output, args.verbose);
    }
    result.on('end', (err) => {
        if (err) {
            console.log(err);
        }
        process.exit(0);
    });
    result.on('fileError', (err) => {
        console.log(chalk.red(err));
    });
    result.on('compileError', (filename, err) => {
        console.error(chalk.underline(filename));
        console.error(chalk.red(err));
    });
    result.on('log', (message) => {
        console.log(chalk.green(message));
    });
}
exports.default = cli;

},{"./Rulify":34,"chalk":undefined,"minimist":undefined}],36:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rulify_1 = require("./rules/Rulify");
exports.Rulify = Rulify_1.default;
const Flowify_1 = require("./bt-flow/Flowify");
exports.Flowify = Flowify_1.default;
const FlowLint_1 = require("./flows/FlowLint");
exports.FlowLint = FlowLint_1.default;
const GojsBareBonesLoader_1 = require("./bt-flow/flow/gojs-loader/GojsBareBonesLoader");
exports.GojsBareBonesLoader = GojsBareBonesLoader_1.default;
const anim_ref_checker_1 = require("./anim-ref/anim-ref-checker");
exports.AnimRefChecker = anim_ref_checker_1.AnimRefChecker;
const keys_ref_checker_1 = require("./keys-ref/keys-ref-checker");
exports.KeysRefChecker = keys_ref_checker_1.KeysRefChecker;
const KeysCompiler_1 = require("./compile-keys/KeysCompiler");
exports.KeysCompiler = KeysCompiler_1.KeysCompiler;
const Behaviorify_1 = require("./bt-flow/Behaviorify");
exports.Behaviorify = Behaviorify_1.default;
const BrowserifyTransform_1 = require("./browserify/BrowserifyTransform");
exports.BrowserifyTransform = BrowserifyTransform_1.default;
const JiboDev_1 = require("./dev/JiboDev");
exports.JiboDev = JiboDev_1.default;
const MimFixer_1 = require("./mim-fixer/MimFixer");
exports.MimFixer = MimFixer_1.default;

},{"./anim-ref/anim-ref-checker":2,"./browserify/BrowserifyTransform":4,"./bt-flow/Behaviorify":5,"./bt-flow/Flowify":6,"./bt-flow/flow/gojs-loader/GojsBareBonesLoader":12,"./compile-keys/KeysCompiler":15,"./dev/JiboDev":17,"./flows/FlowLint":28,"./keys-ref/keys-ref-checker":31,"./mim-fixer/MimFixer":32,"./rules/Rulify":34}]},{},[36])(36)
});
