// jibo-keyframes/lib/jibo-keyframes.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboKeyframes = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
/**
 * All meta data related to a jibo .keys file. This is passed to the keyframes.runtime functions
 * to make it aware of specific layers and their properties/info.
 */
var semver = require("semver");
var BodyLayer_1 = require("./layers/BodyLayer");
var EyeLayer_1 = require("./layers/EyeLayer");
var OverlayLayer_1 = require("./layers/OverlayLayer");
var EyeTextureLayer_1 = require("./layers/EyeTextureLayer");
var OverlayTextureLayer_1 = require("./layers/OverlayTextureLayer");
var BackgroundTextureLayer_1 = require("./layers/BackgroundTextureLayer");
var LEDLayer_1 = require("./layers/LEDLayer");
var PixiLayer_1 = require("./layers/PixiLayer");
var EventLayer_1 = require("./layers/EventLayer");
var AudioLayer_1 = require("./layers/AudioLayer");
var ReferenceLayer_1 = require("./layers/ReferenceLayer");
var GlowLayer_1 = require("./layers/GlowLayer");
var LightingLayer_1 = require("./layers/LightingLayer");
var IDUtils_1 = require("./utils/IDUtils");
var Conversion_1 = require("./utils/Conversion");
var Colors_1 = require("./utils/Colors");
var JiboKeyframeInfo = (function () {
    function JiboKeyframeInfo() {
    }
    /**
     * Implement this to do any backward compatibility related touchups to old files
     */
    JiboKeyframeInfo.onLoad = function (keyframes) {
        // TODO: Figure out why the version isn't always in files
        keyframes.version = keyframes.version || "0.3.0";
        // Layers didn't have an id at 0.2.0, add some for old files
        if (keyframes.version === "0.2.0") {
            keyframes.layers.forEach(function (layer) {
                layer.id = IDUtils_1.default.create();
            });
        }
        // 0.4.0 changed the coordinate system of the eye/overlay from
        // meters to pixel coordinates
        if (semver.lt(keyframes.version, "0.4.0")) {
            // Find each eye and overlay layer and then convert each property
            // that represents a translation from meters to pixel coordinates
            keyframes.layers.forEach(function (layer) {
                if (layer.type === "Eye" || layer.type === "Overlay") {
                    layer.keyframes.forEach(function (key) {
                        for (var prop in key.value) {
                            // Everything but Scale and Rotate are x/y pairs
                            // representing 2D translations
                            if (prop !== "Scale" &&
                                prop !== "Rotate") {
                                key.value[prop].x = Conversion_1.default.toPixelsX(key.value[prop].x);
                                key.value[prop].y = Conversion_1.default.toPixelsY(key.value[prop].y);
                            }
                        }
                    });
                }
            });
        }
        if (semver.lte(keyframes.version, '0.4.0')) {
            keyframes.layers.forEach(function (layer) {
                layer.keyframes.forEach(function (key) {
                    for (var prop in key.value) {
                        if (prop === "Tween" && key.value.Tween.value === 'body') {
                            key.value.Tween = {
                                display: 'cubic in out',
                                label: 'Cubic Ease In Out',
                                value: 'cubicInOut'
                            };
                        }
                    }
                });
            });
        }
        // 0.5.0 changed colors to use rgb instead of hsv for blending and internal representation
        if (semver.lt(keyframes.version, "0.5.0")) {
            keyframes.layers.forEach(function (layer) {
                if (layer.type === "Background Texture" || layer.type === "Eye Texture" ||
                    layer.type === "Overlay Texture" || layer.type === "LED") {
                    // Find each Color property and convert it from hsv space to rgb
                    layer.keyframes.forEach(function (key) {
                        key.value.Color = Colors_1.default.hsvToRgb(key.value.Color.h, key.value.Color.s, key.value.Color.v);
                    });
                }
            });
        }
        // 0.6.0 eye and overlay coordinates were off vertically causing squishing. This was
        // because the conversion between vertical meters and pixels was off. We fixed the conversion
        // and now we have to touch up coordinates from old files so that animations created that
        // compensated for the squish would be broken by this fix
        if (semver.lt(keyframes.version, "0.6.0")) {
            keyframes.layers.forEach(function (layer) {
                if (layer.type === "Overlay" || layer.type === "Eye") {
                    // Find each Color property and convert it from hsv space to rgb
                    layer.keyframes.forEach(function (key) {
                        for (var field in key.value) {
                            if (field !== 'Scale' &&
                                typeof key.value[field].y !== 'undefined') {
                                key.value[field].y = key.value[field].y * 0.06922346 / 0.073142126;
                            }
                        }
                    });
                }
            });
        }
        if (semver.lt(keyframes.version, "0.7.0")) {
            keyframes.layers.forEach(function (layer) {
                if (layer.type === "Overlay" || layer.type === "Eye") {
                    // Find each Color property and convert it from hsv space to rgb
                    layer.keyframes.forEach(function (key) {
                        key.value.Alpha = 1.0;
                        key.value.Visible = true;
                    });
                }
                else if (layer.type === "Pixi") {
                    layer.keyframes.forEach(function (key) {
                        console.log("Upgrade to attach and effect");
                        key.value["Attach To Eye"] = false;
                        key.value["Apply Eye Lighting"] = false;
                    });
                }
            });
        }
        // Deprecate the Apply Eye Lighting and add a frame offset
        if (semver.lt(keyframes.version, "0.9.0")) {
            keyframes.layers.forEach(function (layer) {
                if (layer.type === "Pixi") {
                    layer.keyframes.forEach(function (key) {
                        console.log("Add frame offset, deprecated lighting");
                        key.value["Frame Offset"] = 0;
                        delete key.value["Apply Eye Lighting"];
                    });
                }
            });
        }
        // 0.9 Added reference layers (No upgrade needed as it only adds a new layer type)
        // 0.10.0 Deprecate the Apply Eye Lighting and add a frame offset
        if (semver.lt(keyframes.version, "0.10.0")) {
            keyframes.layers.forEach(function (layer) {
                if (layer.type === "Glow") {
                    layer.keyframes.forEach(function (key) {
                        console.log("Deprecated halo speed on Glow");
                        delete key.value.HaloSpeed;
                    });
                }
            });
        }
        // 0.11.0: Reference layer upgrade
        if (semver.lt(keyframes.version, "0.11.0")) {
            keyframes.layers.forEach(function (layer) {
                // Upgrade reference layers to have "hide before" and "hide after" properties set to false
                if (layer.type === "Reference") {
                    layer.keyframes.forEach(function (key) {
                        console.log("Add Hide-Before Hide-After to reference layer");
                        key.value["Hide Referenced Layers Before Use"] = true;
                        key.value["Hide Referenced Layers After Use"] = true;
                    });
                }
                // Upgrade other layers to have their HideLayer properties set to false
                if (layer.type === 'Overlay Texture'
                    || layer.type === 'Overlay'
                    || layer.type === 'Eye Texture'
                    || layer.type === 'Eye') {
                    layer.keyframes.forEach(function (key) {
                        key.value["HideLayer"] = false;
                    });
                }
            });
        }
        // 0.12.0: Remove 'Hide Layers Before Use' field in Reference layer.
        if (semver.lt(keyframes.version, "0.12.0")) {
            keyframes.layers.forEach(function (layer) {
                // Upgrade reference layers to have "hide before" and "hide after" properties set to false
                if (layer.type === "Reference") {
                    layer.keyframes.forEach(function (key) {
                        console.log("Remove Hide-Before and Hide-After from reference layer");
                        delete key.value["Hide Referenced Layers Before Use"];
                        key.value["HoldFinalPose"] = !key.value["Hide Referenced Layers After Use"];
                        delete key.value["Hide Referenced Layers After Use"];
                    });
                }
            });
        }
        // 0.13.0: Remove 'Hide Layer' field.
        if (semver.lt(keyframes.version, "0.13.0")) {
            keyframes.layers.forEach(function (layer) {
                // Downgrade various layers to remove their HideLayer property
                if (layer.type === 'Overlay Texture'
                    || layer.type === 'Overlay'
                    || layer.type === 'Eye Texture'
                    || layer.type === 'Eye') {
                    layer.keyframes.forEach(function (key) {
                        delete key.value["HideLayer"];
                    });
                }
            });
        }
        // Make version is set to the latest
        keyframes.version = JiboKeyframeInfo.defaults.version;
    };
    return JiboKeyframeInfo;
}());
// Layer types that jibo supports. For efficient access, layerTypes indexes
// each Layer Types constructor by it's name (ie. "Body": Body)
JiboKeyframeInfo.layerTypes = {
    Body: BodyLayer_1.default,
    Eye: EyeLayer_1.default,
    "Eye Texture": EyeTextureLayer_1.default,
    Overlay: OverlayLayer_1.default,
    "Overlay Texture": OverlayTextureLayer_1.default,
    "Background Texture": BackgroundTextureLayer_1.default,
    LED: LEDLayer_1.default,
    Event: EventLayer_1.default,
    "Audio Event": AudioLayer_1.default,
    Reference: ReferenceLayer_1.default,
    Glow: GlowLayer_1.default,
    Lighting: LightingLayer_1.default
};
JiboKeyframeInfo.layerTypes.Pixi = PixiLayer_1.default;
JiboKeyframeInfo.defaults = {
    version: "0.13.0",
    framerate: 30,
    duration: 30,
    scale: 1,
    // Layers that newly created Jibo Keyframe files have
    layers: [
        {
            name: "Eye",
            type: "Eye"
        },
        {
            name: "Body",
            type: "Body"
        },
        {
            name: "Overlay",
            type: "Overlay",
            visible: true,
            locked: false,
            keyframes: [
                {
                    "value": {
                        "Rotate": 0,
                        "Scale": {
                            "x": 1,
                            "y": 1
                        },
                        "Translate": {
                            "x": 0,
                            "y": 0
                        },
                        "Visible": false,
                        "Alpha": 1,
                        "Tween": {
                            "display": "sine in out",
                            "label": "Sine Ease In Out",
                            "value": "sineInOut"
                        },
                        "Vertex 1": {
                            "x": -EyeLayer_1.default.eyeRadiusX,
                            "y": -EyeLayer_1.default.eyeRadiusY
                        },
                        "Vertex 2": {
                            "x": 0,
                            "y": -EyeLayer_1.default.eyeRadiusY
                        },
                        "Vertex 3": {
                            "x": EyeLayer_1.default.eyeRadiusX,
                            "y": -EyeLayer_1.default.eyeRadiusY
                        },
                        "Vertex 4": {
                            "x": -EyeLayer_1.default.eyeRadiusX,
                            "y": 0
                        },
                        "Vertex 5": {
                            "x": 0,
                            "y": 0
                        },
                        "Vertex 6": {
                            "x": EyeLayer_1.default.eyeRadiusX,
                            "y": 0
                        },
                        "Vertex 7": {
                            "x": -EyeLayer_1.default.eyeRadiusX,
                            "y": EyeLayer_1.default.eyeRadiusY
                        },
                        "Vertex 8": {
                            "x": 0,
                            "y": EyeLayer_1.default.eyeRadiusY
                        },
                        "Vertex 9": {
                            "x": EyeLayer_1.default.eyeRadiusX,
                            "y": EyeLayer_1.default.eyeRadiusY
                        }
                    },
                    "time": 0
                }
            ]
        }
    ]
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = JiboKeyframeInfo;

},{"./layers/AudioLayer":6,"./layers/BackgroundTextureLayer":7,"./layers/BodyLayer":8,"./layers/EventLayer":9,"./layers/EyeLayer":11,"./layers/EyeTextureLayer":12,"./layers/GlowLayer":13,"./layers/LEDLayer":14,"./layers/LightingLayer":15,"./layers/OverlayLayer":16,"./layers/OverlayTextureLayer":17,"./layers/PixiLayer":18,"./layers/ReferenceLayer":19,"./utils/Colors":22,"./utils/Conversion":23,"./utils/IDUtils":24,"semver":undefined}],2:[function(require,module,exports){
"use strict";
/**
 * Class responsible for
 *   - saving and loading files
 *   - manipulating the animation and notifying views of changes
 *   - computing the values of animation at any given point in time
 */
var fs = require("fs");
var _ = require("lodash");
var Tweening_1 = require("./utils/Tweening");
var IDUtils_1 = require("./utils/IDUtils");
var Search_1 = require("./utils/Search");
var BlendOperations_1 = require("./utils/BlendOperations");
var Runtime = (function () {
    function Runtime() {
    }
    /**
     * Create a new keyframes object.
     *
     * @param keyframeInfo {Object} Meta data that includes supported layer types,
     *                              default values, and etc
     * @returns {Object} The newly created keyframes object with default layers created
     *
     */
    Runtime.create = function (keyframeInfo) {
        var keyframes = {
            version: keyframeInfo.defaults.version,
            framerate: keyframeInfo.defaults.framerate,
            duration: keyframeInfo.defaults.duration,
            scale: keyframeInfo.defaults.scale,
            layers: []
        };
        keyframeInfo.defaults.layers.forEach(function (defaultLayer) {
            keyframes.layers.push({
                id: IDUtils_1.default.create(),
                name: defaultLayer.name,
                type: defaultLayer.type,
                visible: true,
                locked: false,
                keyframes: _.cloneDeep(defaultLayer.keyframes || [])
            });
        });
        return keyframes;
    };
    /**
     * Load an animation file into this instance
     *
     * @param uri {String} Filepath to load the keyframes file from
     * @param keyframeInfo {Object} Meta data that includes supported layer types,
     *                              default values, and etc
     * @returns {Object} Keyframes file content
     *
     * TODO: Write validation code at some point to catch issues like a file that merged incorrectly,
     * was corrupted, or is just the wrong file format/type. Ensure invariants such as:
     *   - That keys are time ordered
     *   - That key values are valid
     *   - That all required fields exist and no extra ones do
     */
    Runtime.load = function (uri, keyframeInfo) {
        var keyframes = JSON.parse(fs.readFileSync(uri, 'utf8'));
        if (keyframeInfo.onLoad) {
            keyframeInfo.onLoad(keyframes);
        }
        return keyframes;
    };
    /**
     * Save the keyframes to disk
     *
     * @param uri {String} Filepath to save to
     * @param keyframes {Object} Keyframe object to serialize to disk
     */
    Runtime.save = function (uri, keyframes) {
        // Save the current animation state to disk with four spaces for indentation
        fs.writeFileSync(uri, JSON.stringify(keyframes, null, '    '), 'utf8');
    };
    Runtime._isEventLayer = function (layer, keyframeInfo) {
        if (layer) {
            var layerClass = keyframeInfo.layerTypes[layer.type];
            return layerClass.isEvent();
        }
        return false;
    };
    /**
     * Get the dof values generated from all the DOF layers at a point in time
     * @param keyframes {Object} Keyframes data we'll be evaluating the layers of
     * @param keyframeInfo {Object} Meta data that includes supported layer types,
     *                              default values, and etc
     * @param timeInSeconds {number} The time to compute the final dof values at
     * @param overrides {Object} If provided each field in this object includes a set of properties
     *                           to override the usually calculated interpolated keyframed properties.
     *                           This is useful in tools for interactively changing
     *                           the properties of a layer at a point in time without
     *                           committing a new keyframe as a user tries out different jibo poses
     * @return {Object} An object who's fields are each dof and it's value after all
     *                  interpolation and layer blending
     */
    Runtime.evaluateAllDOFLayers = function (keyframes, keyframeInfo, timeInSeconds, overrides) {
        var accumulatedProps = Runtime.evaluateAllLayersFiltered(keyframes, keyframeInfo, timeInSeconds, function (layer) {
            return !Runtime._isEventLayer(layer, keyframeInfo); // only care about non-event layers (DOF)
        }, false, overrides);
        // Now that all props have been accumulated, pass them to
        // the layer types generateDofs() function to get our final dof values
        var dofs = {};
        for (var layerType in accumulatedProps) {
            var layerClass = keyframeInfo.layerTypes[layerType];
            var generatedDofs = layerClass.generateDofs(accumulatedProps[layerType]);
            // Copy These
            // TODO: Eventually multiple types of layers will be generating dof values
            // so blending operations will need to occur here as well.
            for (var dofName in generatedDofs) {
                dofs[dofName] = generatedDofs[dofName];
            }
        }
        return dofs;
    };
    /**
     * Detect whether the layer was originally from a referenced layer.
     * (Every referenced layer's ID is prefixed with the parent (i.e. referencing)
     * layer's ID and delimited with "-" as part of the compositing operation).
     * @param {*} layer
     * @returns {boolean}
     */
    Runtime.isReferencedLayer = function (layer) {
        return layer.id.indexOf('-') > 0;
    };
    /**
     * Detect whether there were any Eye layers in the referenced layers.
     * @param {Keyframes} keyframes
     * @returns {boolean}
     */
    Runtime.hasReferencedEyeLayers = function (keyframes) {
        for (var i = 0; i < keyframes.layers.length; i++) {
            var layer = keyframes.layers[i];
            if (this.isReferencedLayer(layer) && layer.type === 'Eye') {
                return true;
            }
        }
        return false;
    };
    /**
     * Evaluate and Blend the layers within keyframes (filtering some out).
     * Handles Blink Layer Mode:
     * Any eye layers found in referenced layers put us in so-called "blink layer" mode, which operates as follows:
     * The eye layers at top level are used to blend all the referenced eye and [any] referenced overlay layers.
     * The overlay layers at top level are ignored.
     * All other layers are blended normally.
     * @param {Keyframes} keyframes
     * @param keyframeInfo
     * @param {number} timeInSeconds
     * @param {Callback} includeFilter
     * @param provideDefaults
     * @param overrides
     * @returns {*}
     */
    Runtime.evaluateAllLayersFiltered = function (keyframes, keyframeInfo, timeInSeconds, includeFilter, provideDefaults, overrides) {
        var _this = this;
        // If we are not in Blink Layer mode, perform the standard layer evaluation for all layers.
        if (!this.hasReferencedEyeLayers(keyframes)) {
            return this.evaluateLayersFiltered(keyframes, keyframeInfo, timeInSeconds, includeFilter, provideDefaults, overrides);
        }
        // "Blink layer mode" combines the Eye and Overlay in the Top and Reference layers in a particular way.
        var topFilter = function (layer) {
            return includeFilter(layer) && !_this.isReferencedLayer(layer) && layer.type === 'Eye';
        };
        var refFilter = function (layer) {
            return includeFilter(layer) && _this.isReferencedLayer(layer) && (layer.type === 'Eye' || layer.type === 'Overlay');
        };
        var otherFilter = function (layer) {
            return includeFilter(layer) && layer.type !== 'Eye' && layer.type !== 'Overlay';
        };
        // Gather up the Eye and Overlay properties from Ref layers.
        var ref = this.evaluateLayersFiltered(keyframes, keyframeInfo, timeInSeconds, refFilter, overrides);
        // If no layers in the reference are currently active, we are not in blink mode.
        if (Object.keys(ref).length === 0) {
            return this.evaluateLayersFiltered(keyframes, keyframeInfo, timeInSeconds, includeFilter, provideDefaults, overrides);
        }
        // Do the standard evaluation of all the non-Eye/Overlay layers.
        var accumulatedProps = this.evaluateLayersFiltered(keyframes, keyframeInfo, timeInSeconds, otherFilter, overrides);
        // Gather up the Eye properties from the top layer.
        var top = this.evaluateLayersFiltered(keyframes, keyframeInfo, timeInSeconds, topFilter, overrides);
        var eyeLayerInfo = keyframeInfo.layerTypes['Eye'].getInfo();
        // Then form the final Eye and Overlay props by combining the referenced layers with the top Eye layer.
        for (var _i = 0, _a = ['Eye', 'Overlay']; _i < _a.length; _i++) {
            var refLayerType = _a[_i];
            if (ref[refLayerType]) {
                // Initialize this layer's properties with the referenced layer values.
                accumulatedProps[refLayerType] = ref[refLayerType];
                // However, if we have a top level Eye layer, we instead make this layer's properties by
                // blending each referenced layer property with the top level Eye layer's properties.
                if (top['Eye']) {
                    for (var prop in eyeLayerInfo.properties) {
                        var propInfo = eyeLayerInfo.properties[prop];
                        var blendOp = BlendOperations_1.default.normal(propInfo.blendOperation, propInfo.type);
                        accumulatedProps[refLayerType][prop] = blendOp(top['Eye'][prop], ref[refLayerType][prop], propInfo.defaultValue);
                    }
                }
            }
        }
        return accumulatedProps;
    };
    // Evaluate all layers by tweening within each layer and then blending the tweened values together.
    Runtime.evaluateLayersFiltered = function (keyframes, keyframeInfo, timeInSeconds, includeFilter, provideDefaults, overrides) {
        // This will hold each layers blended props as they accumulate
        var accumulatedProps = {};
        // Blend layers from bottom to top.
        for (var i = keyframes.layers.length - 1; i >= 0; i--) {
            var layer = keyframes.layers[i];
            // Locked, invisible, or filtered layers don't have any affect on the final props/dofs
            if (!includeFilter(layer) || !layer.visible) {
                continue;
            }
            var override = overrides === undefined ? undefined : overrides[layer.id], props = Runtime.evaluateLayer(layer, keyframes, keyframeInfo, timeInSeconds, override);
            // Ignore the entire layer if is currently marked hidden.
            if (!props) {
                continue;
            }
            var layerType = layer.type;
            // Is this the first instance of a particular layer type
            // If so initialize the accumulated props to this layer's computed values
            if (typeof accumulatedProps[layerType] === "undefined") {
                accumulatedProps[layerType] = props;
            }
            else {
                var layerProps = accumulatedProps[layerType], layerInfo = keyframeInfo.layerTypes[layerType].getInfo();
                // Blend each property based on it's type/blendOperation
                for (var propName in props) {
                    var propInfo = layerInfo.properties[propName];
                    if (!('blendOperation' in propInfo)) {
                        console.log("blendoperation not found in " + Object.keys(propInfo) + " for layerType " + layerType);
                        continue;
                    }
                    layerProps[propName] = BlendOperations_1.default.normal(propInfo.blendOperation, propInfo.type)(layerProps[propName], props[propName], propInfo.defaultValue);
                }
            }
        }
        return accumulatedProps;
    };
    /*
    *   Used by the editor's animation model.
    *   Turn our absoluteProps into ones relative to all the layers of our type that don't include the excluded one.
    * */
    Runtime.computeRelativePropValues = function (excludedLayer, keyframes, keyframeInfo, timeInSeconds, absoluteProps, overrides) {
        var props = Runtime.evaluateAllLayersFiltered(keyframes, keyframeInfo, timeInSeconds, function (layer) {
            if (layer.type !== excludedLayer.type || layer.id === excludedLayer.id) {
                return false;
            }
            else {
                return true;
            }
        }, true, overrides);
        // Return absolute props as is if there were no other layers of it's type
        if (typeof props[excludedLayer.type] === "undefined") {
            return absoluteProps;
        }
        var layerInfo = keyframeInfo.layerTypes[excludedLayer.type].getInfo(), relativeProps = {};
        for (var propName in absoluteProps) {
            var propInfo = layerInfo.properties[propName];
            relativeProps[propName] = BlendOperations_1.default.inverse(propInfo.blendOperation, propInfo.type)(props[excludedLayer.type][propName], absoluteProps[propName], propInfo.defaultValue);
        }
        return relativeProps;
    };
    /**
     * Get the property values generated from a single layer at a point in time
     *
     * @param layer {Object} The layer that we need to evaluate
     * @param keyframes {Object} Keyframes data that this layer is a part of
     * @param keyframeInfo {Object} Meta data that includes supported layer types,
     *                              default values, and etc
     * @param timeInSeconds {number} The time to compute the properties values at
     * @param override {Object}  If provided this object includes a set of properties
     *                           to override the usually calculated keyframed properties.
     *                           This is useful in tools for interactively changing
     *                           the properties of a layer at a point in time without
     *                           committing a new keyframe as a user tries out different jibo poses
     * @return {Object} An object who's fields are each dof and it's value after all
     *                  interpolation and layer blending. Alternatively, return undefined
     *                  if there is no valid layer at the given time.
     */
    Runtime.evaluateLayer = function (layer, keyframes, keyframeInfo, timeInSeconds, override) {
        if (Runtime._isEventLayer(layer, keyframeInfo)) {
            // do not interpolate for event types
            return Runtime._evaluateEventLayer(layer, keyframes, keyframeInfo, timeInSeconds);
        }
        var props = {};
        if (override) {
            for (var propName in override) {
                props[propName] = override[propName];
            }
            return props;
        }
        var layerTypeInfo = keyframeInfo.layerTypes[layer.type].getInfo();
        var defaults = {};
        for (var propName in layerTypeInfo.properties) {
            defaults[propName] = _.cloneDeep(layerTypeInfo.properties[propName].defaultValue);
        }
        // If there are no keyframes, return the defaults
        if (layer.keyframes.length === 0) {
            if (layer.type === 'Overlay') {
                // Except in the case of Overlay -- if there are no keyframes in the
                // overlay, pretend there is no overlay layer.
                return;
            }
            return defaults;
        }
        // There are keyframes, calculate the properties at the supplied time
        // Calculate the keyframe index this time is equivalent to.
        // NOTE: We don't floor it, we just want to have a number in frame counts
        // instead of in seconds so that we can compare times and interpolate between
        // times
        if (keyframes.scale === undefined) {
            keyframes.scale = keyframeInfo.defaults.scale;
        }
        var currentTime = timeInSeconds * keyframes.framerate;
        // Ignore layers that are not valid before the current time.
        if (layer.validFrom !== undefined && currentTime < layer.validFrom) {
            return;
        }
        // Hold layers at a specific timepoint if specified.
        if (layer.holdAfter !== undefined && currentTime >= layer.holdAfter) {
            currentTime = layer.holdAfter;
        }
        // Ignore layers that are not valid after the current time.
        if (layer.validUpto !== undefined && currentTime >= layer.validUpto) {
            return;
        }
        // Return defaults if we're before the beginning.
        if (currentTime < layer.keyframes[0].time) {
            if (layer.type === 'Overlay') {
                // Except in the case of Overlay -- if there are no keyframes before the
                // overlay, pretend there is no overlay layer.
                return;
            }
            return defaults;
        }
        // Is the current time past the time of the last keyframe? If so grab that
        // keyframe's value.
        if (currentTime >= layer.keyframes[layer.keyframes.length - 1].time) {
            return _.cloneDeep(layer.keyframes[layer.keyframes.length - 1].value);
        }
        // The current time is between keyframes.
        // Scan from the last to first keyframe, stop on the first keyframe
        // further back or equal in time to us.
        var _a = Search_1.default.keyframeSearch(layer.keyframes, currentTime), start = _a.start, end = _a.end;
        start = layer.keyframes[start];
        end = layer.keyframes[end];
        // Now that we have our two closest keyframes, interpolate between their values
        var timeStart = start.time, timeEnd = end.time, timeNormalized = (currentTime - timeStart) / (timeEnd - timeStart);
        var easingFunction = 'cubicInOut';
        if (start.value.Tween) {
            easingFunction = start.value.Tween.value;
        }
        for (var propName in start.value) {
            var tweenType = layerTypeInfo.properties[propName].type;
            if (typeof tweenType !== 'string') {
                tweenType = tweenType.name;
            }
            props[propName] = Tweening_1.default[tweenType](start.value[propName], end.value[propName], timeNormalized, easingFunction);
        }
        return props;
    };
    Runtime._evaluateEventLayer = function (layer, keyframes, keyframeInfo, timeInSeconds) {
        var keyframeTime = timeInSeconds * keyframes.framerate;
        // if we have a keyframe at the given keyframeTime, then return it;
        // otherwise, return the default values
        var frame = Search_1.default.keyframeSearchAbsolute(layer.keyframes, keyframeTime);
        if (frame !== undefined) {
            return _.cloneDeep(layer.keyframes[frame].value);
        }
        var layerTypeInfo = keyframeInfo.layerTypes[layer.type].getInfo();
        var props = {};
        for (var propName in layerTypeInfo.properties) {
            props[propName] = _.cloneDeep(layerTypeInfo.properties[propName].defaultValue);
        }
        return props;
    };
    /**
     * Get the event values (payloads) generated from all the event layers at a point in time
     * @param keyframes {Object} Keyframes data we'll be evaluating the layers of
     * @param keyframeInfo {Object} Meta data that includes supported layer types,
     *                              default values, and etc
     * @param timeInSeconds {number} The time to check
     *
     * @return {array} An array of all accumulated event layers and their payloads.
     */
    Runtime.evaluateAllEventLayers = function (keyframes, keyframeInfo, timeInSeconds) {
        var events = [];
        // Find all our event layers
        var layersLen = keyframes.layers.length;
        for (var i = 0; i < layersLen; i++) {
            var layer = keyframes.layers[i];
            // if not visible or not an event layer, move along
            if (!Runtime._isEventLayer(layer, keyframeInfo) || !layer.visible) {
                continue;
            }
            var props = Runtime.evaluateLayer(layer, keyframes, keyframeInfo, timeInSeconds);
            if (props) {
                var layerClass = keyframeInfo.layerTypes[layer.type];
                var generatedEvent = layerClass.generateEvent(props, i);
                var eventType = (typeof generatedEvent);
                if ((eventType === 'object') && (layerClass.isValid(generatedEvent))) {
                    events.push(generatedEvent);
                }
            }
        }
        return events;
    };
    return Runtime;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Runtime;

},{"./utils/BlendOperations":20,"./utils/IDUtils":24,"./utils/Search":26,"./utils/Tweening":27,"fs":undefined,"lodash":undefined}],3:[function(require,module,exports){
"use strict";
var BaseLayer = (function () {
    function BaseLayer() {
    }
    /**
     * Return type information about the layer, such as:
     *   - Properties it supports and their:
     *      - type
     *      - default values
     *      - min/max where applicable
     *      - the operation (for DOF) that the layering system uses to blend with other layers:
     *        - additive for translations, rotations, and color type properties
     *        - multiplicative for scaling
     *        - last one wins for textures
     *   - values (DOF, events, etc.) the layer generates
     * Each layer needs to implement this.
     */
    BaseLayer.getInfo = function () {
        throw "Implement static Layer.getInfo() to supply property meta data";
    };
    BaseLayer.isEvent = function () {
        throw "Implement static Layer.isEvent()";
    };
    return BaseLayer;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BaseLayer;

},{}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var BaseLayer_1 = require("./BaseLayer");
var DOFBaseLayer = (function (_super) {
    __extends(DOFBaseLayer, _super);
    function DOFBaseLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Given a set of properties that a layer supports return the dof values that
     * that this layer generates. The properties will have already been interpolated
     * from the keyframes of the layer and blended with other layers. Only generation
     * logic needs to be implemented.
     *
     * @param props {Object}   An object with each field having the name of each layer
     *                         property and it's interpolated/blended value
     *
     * @returns {Object}    An object with each field having the name of each generated
     *                      dof and it's value. These will then be handed to final blending
     *                      logic that blends all dofs generated from all layer types
     *                      according to their blend operation.
     */
    DOFBaseLayer.generateDofs = function () {
        throw "Implement static Layer.generateDofs()";
    };
    DOFBaseLayer.isEvent = function () {
        return false;
    };
    return DOFBaseLayer;
}(BaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = DOFBaseLayer;

},{"./BaseLayer":3}],5:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var BaseLayer_1 = require("./BaseLayer");
var EventBaseLayer = (function (_super) {
    __extends(EventBaseLayer, _super);
    function EventBaseLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Given a set of properties that a layer supports return the event value (payload) that
     * that this layer generates.
     *
     * @param props {Object}   An object with each field having the name of each layer
     *                         property and its event payloads if applicable
     *
     * @returns {Object}    A JSON object with the payload information for that layer
     */
    EventBaseLayer.generateEvent = function () {
        throw "Implement static Layer.generateEvent()";
    };
    EventBaseLayer.isEvent = function () {
        return true;
    };
    EventBaseLayer.isValid = function () {
        throw "Implement static Layer.isValid(generedatedEvent)";
    };
    return EventBaseLayer;
}(BaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EventBaseLayer;

},{"./BaseLayer":3}],6:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var EventBaseLayer_1 = require("../bases/EventBaseLayer");
var AudioInfo = {
    layerType: "Audio Event",
    properties: {
        AudioEvent: {
            type: "audio",
            blendOperation: 'lastOneWins',
            defaultValue: {
                file: ""
            }
        }
    }
};
/*
* Audio layer with specific payload data.
* */
var AudioLayer = (function (_super) {
    __extends(AudioLayer, _super);
    function AudioLayer() {
        return _super.call(this) || this;
    }
    AudioLayer.getInfo = function () {
        return AudioInfo;
    };
    AudioLayer.generateEvent = function (props) {
        var event = {
            name: "play-audio",
            payload: {
                "file": props.AudioEvent.file
            }
        };
        return event;
    };
    AudioLayer.isValid = function (generatedEvent) {
        return ((generatedEvent.name === "play-audio") &&
            (generatedEvent.payload.file.length > 0));
    };
    return AudioLayer;
}(EventBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AudioLayer;

},{"../bases/EventBaseLayer":5}],7:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var BackgroundTextureInfo = {
    layerType: "Background Texture",
    properties: {
        Texture: {
            type: "texture",
            defaultValue: "",
            blendOperation: "lastOneWins"
        },
        Color: {
            type: "rgb",
            defaultValue: {
                r: 255,
                g: 255,
                b: 255
            },
            blendOperation: "add"
        }
    },
    dofs: {
        screenBGTextureInfixBn_r: true,
        screenBG_redChannelBn_r: true,
        screenBG_greenChannelBn_r: true,
        screenBG_blueChannelBn_r0: true
    }
};
var BackgroundTextureLayer = (function (_super) {
    __extends(BackgroundTextureLayer, _super);
    function BackgroundTextureLayer() {
        return _super.call(this) || this;
    }
    BackgroundTextureLayer.getInfo = function () {
        return BackgroundTextureInfo;
    };
    BackgroundTextureLayer.generateDofs = function (props) {
        return {
            screenBGTextureInfixBn_r: props.Texture,
            screenBG_redChannelBn_r: props.Color.r / 255.0,
            screenBG_greenChannelBn_r: props.Color.g / 255.0,
            screenBG_blueChannelBn_r: props.Color.b / 255.0
        };
    };
    return BackgroundTextureLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BackgroundTextureLayer;

},{"../bases/DOFBaseLayer":4}],8:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var values = [
    {
        display: 'linear',
        label: 'Linear Tween',
        value: 'linear'
    },
    {
        display: 'sine in',
        label: 'Sine Ease In',
        value: 'sineIn'
    },
    {
        display: 'sine out',
        label: 'Sine Ease Out',
        value: 'sineOut'
    },
    {
        display: 'sine in out',
        label: 'Sine Ease In Out',
        value: 'sineInOut'
    },
    {
        display: 'cubic in',
        label: 'Cubic Ease In',
        value: 'cubicIn'
    },
    {
        display: 'cubic out',
        label: 'Cubic Ease Out',
        value: 'cubicOut'
    },
    {
        display: 'cubic in out',
        label: 'Cubic Ease In Out',
        value: 'cubicInOut'
    }
];
var Tween = {
    type: {
        name: 'enum',
        values: values
    },
    defaultValue: values[3],
    blendOperation: 'none'
};
var RotationType = {
    type: "float",
    defaultValue: 0.0,
    blendOperation: "add"
};
var BodyInfo = {
    layerType: "Body",
    properties: {
        Head: RotationType,
        Torso: RotationType,
        Pelvis: RotationType,
        Tween: Tween
    },
    dofs: {
        topSection_r: true,
        middleSection_r: true,
        bottomSection_r: true
    }
};
var BodyLayer = (function (_super) {
    __extends(BodyLayer, _super);
    function BodyLayer() {
        return _super.call(this) || this;
    }
    BodyLayer.getInfo = function () {
        return BodyInfo;
    };
    BodyLayer.generateDofs = function (props) {
        return {
            topSection_r: props.Head / 180 * Math.PI,
            middleSection_r: props.Torso / 180 * Math.PI,
            bottomSection_r: props.Pelvis / 180 * Math.PI
        };
    };
    return BodyLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BodyLayer;

},{"../bases/DOFBaseLayer":4}],9:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var EventBaseLayer_1 = require("../bases/EventBaseLayer");
var EventInfo = {
    layerType: "Event",
    properties: {
        Event: {
            type: "event",
            blendOperation: 'lastOneWins',
            defaultValue: {
                name: "",
                payload: ""
            }
        }
    }
};
/*
* Generic event layer with arbitrary payload data.
* */
var EventLayer = (function (_super) {
    __extends(EventLayer, _super);
    function EventLayer() {
        return _super.call(this) || this;
    }
    EventLayer.getInfo = function () {
        return EventInfo;
    };
    EventLayer.generateEvent = function (props) {
        var event = {
            name: ((props.Event.name.length === 0) ? "" : props.Event.name),
            payload: {}
        };
        // if it's a valid JSON, return the parsed object; otherwise, return
        // whatever the payload was
        if (props.Event.payload.length > 0) {
            try {
                event.payload = JSON.parse(props.Event.payload);
            }
            catch (error) {
                event.payload = props.Event.payload;
            }
        }
        return event;
    };
    EventLayer.isValid = function (generatedEvent) {
        // only need valid name
        return (generatedEvent.name.length !== 0);
    };
    return EventLayer;
}(EventBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EventLayer;

},{"../bases/EventBaseLayer":5}],10:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var _ = require("lodash");
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var Conversion_1 = require("../utils/Conversion");
// Eye/Overlays
var centerX = 0;
var centerY = 0;
var eyeRadiusX = 300;
var eyeRadiusY = 300;
var values = [
    {
        display: 'linear',
        label: 'Linear Tween',
        value: 'linear'
    },
    {
        display: 'sine in',
        label: 'Sine Ease In',
        value: 'sineIn'
    },
    {
        display: 'sine out',
        label: 'Sine Ease Out',
        value: 'sineOut'
    },
    {
        display: 'sine in out',
        label: 'Sine Ease In Out',
        value: 'sineInOut'
    },
    {
        display: 'cubic in',
        label: 'Cubic Ease In',
        value: 'cubicIn'
    },
    {
        display: 'cubic out',
        label: 'Cubic Ease Out',
        value: 'cubicOut'
    },
    {
        display: 'cubic in out',
        label: 'Cubic Ease In Out',
        value: 'cubicInOut'
    }
];
var Tween = {
    type: {
        name: 'enum',
        values: values
    },
    defaultValue: values[3],
    blendOperation: 'none'
};
var VertexType = function (xRelativeToCenter, yRelativeToCenter) {
    return {
        type: "vector2",
        defaultValue: {
            x: xRelativeToCenter,
            y: yRelativeToCenter
        },
        blendOperation: "add"
    };
};
var Info = {
    properties: {
        Rotate: {
            type: "float",
            defaultValue: 0.0,
            blendOperation: "add"
        },
        Scale: {
            type: "vector2",
            defaultValue: {
                x: 1,
                y: 1
            },
            blendOperation: "multiply"
        },
        Translate: {
            type: "vector2",
            defaultValue: {
                x: centerX,
                y: centerY
            },
            blendOperation: "add"
        },
        Visible: {
            type: "bool",
            defaultValue: true,
            blendOperation: "lastOneWins"
        },
        Alpha: {
            type: "float",
            defaultValue: 1.0,
            blendOperation: "multiply"
        },
        Tween: Tween,
        "Vertex 1": VertexType(-eyeRadiusX, -eyeRadiusY),
        "Vertex 2": VertexType(0, -eyeRadiusY),
        "Vertex 3": VertexType(eyeRadiusX, -eyeRadiusY),
        "Vertex 4": VertexType(-eyeRadiusX, 0),
        "Vertex 5": VertexType(0, 0),
        "Vertex 6": VertexType(eyeRadiusX, 0),
        "Vertex 7": VertexType(-eyeRadiusX, eyeRadiusY),
        "Vertex 8": VertexType(0, eyeRadiusY),
        "Vertex 9": VertexType(eyeRadiusX, eyeRadiusY)
    }
};
var EyeAndOverlayLayer = (function (_super) {
    __extends(EyeAndOverlayLayer, _super);
    function EyeAndOverlayLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(EyeAndOverlayLayer, "eyeRadiusY", {
        get: function () {
            return eyeRadiusY;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EyeAndOverlayLayer, "eyeRadiusX", {
        get: function () {
            return eyeRadiusX;
        },
        enumerable: true,
        configurable: true
    });
    EyeAndOverlayLayer._generateDofs = function (props, nameStem) {
        return _a = {},
            _a[nameStem + "vertexJoint1_t"] = Conversion_1.default.toMetersX(props['Vertex 1'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint1_t_2"] = Conversion_1.default.toMetersY(props['Vertex 1'].y) * props.Scale.y,
            _a[nameStem + "vertexJoint2_t"] = Conversion_1.default.toMetersX(props['Vertex 2'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint2_t_2"] = Conversion_1.default.toMetersY(props['Vertex 2'].y) * props.Scale.y,
            _a[nameStem + "vertexJoint3_t"] = Conversion_1.default.toMetersX(props['Vertex 3'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint3_t_2"] = Conversion_1.default.toMetersY(props['Vertex 3'].y) * props.Scale.y,
            //
            _a[nameStem + "vertexJoint4_t"] = Conversion_1.default.toMetersX(props['Vertex 4'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint4_t_2"] = Conversion_1.default.toMetersY(props['Vertex 4'].y) * props.Scale.y,
            _a[nameStem + "vertexJoint5_t"] = Conversion_1.default.toMetersX(props['Vertex 5'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint5_t_2"] = Conversion_1.default.toMetersY(props['Vertex 5'].y) * props.Scale.y,
            _a[nameStem + "vertexJoint6_t"] = Conversion_1.default.toMetersX(props['Vertex 6'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint6_t_2"] = Conversion_1.default.toMetersY(props['Vertex 6'].y) * props.Scale.y,
            //
            _a[nameStem + "vertexJoint7_t"] = Conversion_1.default.toMetersX(props['Vertex 7'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint7_t_2"] = Conversion_1.default.toMetersY(props['Vertex 7'].y) * props.Scale.y,
            _a[nameStem + "vertexJoint8_t"] = Conversion_1.default.toMetersX(props['Vertex 8'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint8_t_2"] = Conversion_1.default.toMetersY(props['Vertex 8'].y) * props.Scale.y,
            _a[nameStem + "vertexJoint9_t"] = Conversion_1.default.toMetersX(props['Vertex 9'].x) * props.Scale.x,
            _a[nameStem + "vertexJoint9_t_2"] = Conversion_1.default.toMetersY(props['Vertex 9'].y) * props.Scale.y,
            _a;
        var _a;
    };
    EyeAndOverlayLayer.overrideProperties = function (newProps) {
        Object.assign(Info.properties, newProps);
    };
    EyeAndOverlayLayer.extend = function (input) {
        var result = _.merge({}, Info, input);
        return result;
    };
    return EyeAndOverlayLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EyeAndOverlayLayer;

},{"../bases/DOFBaseLayer":4,"../utils/Conversion":23,"lodash":undefined}],11:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var EyeAndOverlayLayer_1 = require("./EyeAndOverlayLayer");
var Conversion_1 = require("../utils/Conversion");
var EyeInfo = EyeAndOverlayLayer_1.default.extend({
    layerType: "Eye",
    dofs: {
        eyeSubRootBn_r: true,
        eyeSubRootBn_t: true,
        eyeSubRootBn_t_2: true,
        vertexJoint1_t: true,
        vertexJoint1_t_2: true,
        vertexJoint2_t: true,
        vertexJoint2_t_2: true,
        vertexJoint3_t: true,
        vertexJoint3_t_2: true,
        vertexJoint4_t: true,
        vertexJoint4_t_2: true,
        vertexJoint5_t: true,
        vertexJoint5_t_2: true,
        vertexJoint6_t: true,
        vertexJoint6_t_2: true,
        vertexJoint7_t: true,
        vertexJoint7_t_2: true,
        vertexJoint8_t: true,
        vertexJoint8_t_2: true,
        vertexJoint9_t: true,
        eyeVisibilityBn_r: true,
        eye_alphaChannelBn_r: true
    }
});
var EyeLayer = (function (_super) {
    __extends(EyeLayer, _super);
    function EyeLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EyeLayer.getInfo = function () {
        return EyeInfo;
    };
    Object.defineProperty(EyeLayer, "eyeRadiusX", {
        get: function () {
            return EyeAndOverlayLayer_1.default.eyeRadiusX;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EyeLayer, "eyeRadiusY", {
        get: function () {
            return EyeAndOverlayLayer_1.default.eyeRadiusY;
        },
        enumerable: true,
        configurable: true
    });
    EyeLayer.generateDofs = function (props) {
        var dofs = EyeAndOverlayLayer_1.default._generateDofs(props, "");
        dofs.eyeSubRootBn_r = props.Rotate / 180 * Math.PI;
        dofs.eyeSubRootBn_t = Conversion_1.default.toMetersX(props.Translate.x);
        dofs.eyeSubRootBn_t_2 = Conversion_1.default.toMetersY(props.Translate.y);
        dofs.eyeVisibilityBn_r = props.Visible === undefined ? 1 : (props.Visible ? 1 : 0);
        dofs.eye_alphaChannelBn_r = props.Alpha === undefined ? 1 : props.Alpha;
        return dofs;
    };
    return EyeLayer;
}(EyeAndOverlayLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EyeLayer;

},{"../utils/Conversion":23,"./EyeAndOverlayLayer":10}],12:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var EyeTextureInfo = {
    layerType: "Eye Texture",
    properties: {
        Texture: {
            type: "texture",
            defaultValue: "",
            blendOperation: "lastOneWins"
        },
        Color: {
            type: "rgb",
            defaultValue: {
                r: 255,
                g: 255,
                b: 255
            },
            blendOperation: "add"
        }
    },
    dofs: {
        eyeTextureInfixBn_r: true,
        eye_redChannelBn_r: true,
        eye_greenChannelBn_r: true,
        eye_blueChannelBn_r: true
    }
};
var EyeTextureLayer = (function (_super) {
    __extends(EyeTextureLayer, _super);
    function EyeTextureLayer() {
        return _super.call(this) || this;
    }
    EyeTextureLayer.getInfo = function () {
        return EyeTextureInfo;
    };
    EyeTextureLayer.generateDofs = function (props) {
        return {
            eyeTextureInfixBn_r: props.Texture,
            eye_redChannelBn_r: props.Color.r / 255.0,
            eye_greenChannelBn_r: props.Color.g / 255.0,
            eye_blueChannelBn_r: props.Color.b / 255.0
        };
    };
    return EyeTextureLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EyeTextureLayer;

},{"../bases/DOFBaseLayer":4}],13:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var GlowInfo = {
    layerType: "Glow",
    properties: {
        Enabled: {
            type: "bool",
            defaultValue: true,
            blendOperation: "lastOneWins"
        },
        Amount: {
            type: "float",
            defaultValue: 0.0,
            blendOperation: "add"
        },
        Blur: {
            type: "float",
            defaultValue: 15.0,
            blendOperation: "add"
        },
        HaloMaxDistance: {
            type: "float",
            defaultValue: 80,
            blendOperation: "add"
        },
        Halo: {
            type: "float",
            defaultValue: 0,
            blendOperation: "add"
        },
        Color: {
            type: "rgb",
            defaultValue: {
                r: 125,
                g: 249,
                b: 255
            },
            blendOperation: "add"
        }
    }
};
var GlowLayer = (function (_super) {
    __extends(GlowLayer, _super);
    function GlowLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GlowLayer.generateDofs = function (props) {
        return {
            glow_amount: !props.Enabled ? 0 : props.Amount,
            glow_blur: props.Blur,
            glow_halo: props.Halo,
            glow_haloMaxDistance: props.HaloMaxDistance,
            glow_redChannel: props.Color.r,
            glow_greenChannel: props.Color.g,
            glow_blueChannel: props.Color.b
        };
    };
    GlowLayer.getInfo = function () {
        return GlowInfo;
    };
    return GlowLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = GlowLayer;

},{"../bases/DOFBaseLayer":4}],14:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var LEDInfo = {
    layerType: "LED",
    properties: {
        Color: {
            type: "rgb",
            defaultValue: {
                r: 0,
                g: 0,
                b: 0
            },
            blendOperation: "add"
        }
    },
    dofs: {
        lightring_redChannelBn_r: true,
        lightring_greenChannelBn_r: true,
        lightring_blueChannelBn_r: true
    }
};
var LEDLayer = (function (_super) {
    __extends(LEDLayer, _super);
    function LEDLayer() {
        return _super.call(this) || this;
    }
    LEDLayer.getInfo = function () {
        return LEDInfo;
    };
    LEDLayer.generateDofs = function (props) {
        return {
            lightring_redChannelBn_r: props.Color.r / 255.0,
            lightring_greenChannelBn_r: props.Color.g / 255.0,
            lightring_blueChannelBn_r: props.Color.b / 255.0
        };
    };
    return LEDLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LEDLayer;

},{"../bases/DOFBaseLayer":4}],15:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var LightingInfo = {
    layerType: "Lighting",
    properties: {
        Enabled: {
            type: "bool",
            defaultValue: true,
            blendOperation: "lastOneWins"
        },
        Amount: {
            type: "float",
            defaultValue: 1,
            blendOperation: "add"
        },
        Brightness: {
            type: "float",
            defaultValue: 0.0,
            blendOperation: "add"
        },
        Gradient: {
            type: "float",
            defaultValue: 0.5,
            blendOperation: "add"
        },
        Radius: {
            type: "float",
            defaultValue: 300,
            blendOperation: "add"
        },
        Light: {
            type: "rgb",
            defaultValue: {
                r: 255,
                g: 255,
                b: 255
            },
            blendOperation: "add"
        },
        Mid: {
            type: "rgb",
            defaultValue: {
                r: 192,
                g: 186,
                b: 182
            },
            blendOperation: "add"
        },
        Dark: {
            type: "rgb",
            defaultValue: {
                r: 90,
                g: 85,
                b: 82
            },
            blendOperation: "add"
        }
    }
};
var LightingLayer = (function (_super) {
    __extends(LightingLayer, _super);
    function LightingLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LightingLayer.generateDofs = function (props) {
        return {
            lighting_amount: !props.Enabled ? 0 : props.Amount,
            lighting_brightness: props.Brightness,
            lighting_gradient: props.Gradient,
            lighting_radius: props.Radius,
            lighting_light_redChannel: props.Light.r,
            lighting_light_greenChannel: props.Light.g,
            lighting_light_blueChannel: props.Light.b,
            lighting_mid_redChannel: props.Mid.r,
            lighting_mid_greenChannel: props.Mid.g,
            lighting_mid_blueChannel: props.Mid.b,
            lighting_dark_redChannel: props.Dark.r,
            lighting_dark_greenChannel: props.Dark.g,
            lighting_dark_blueChannel: props.Dark.b
        };
    };
    LightingLayer.getInfo = function () {
        return LightingInfo;
    };
    return LightingLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LightingLayer;

},{"../bases/DOFBaseLayer":4}],16:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var EyeAndOverlayLayer_1 = require("./EyeAndOverlayLayer");
var Conversion_1 = require("../utils/Conversion");
var OverlayInfo = EyeAndOverlayLayer_1.default.extend({
    layerType: "Overlay",
    properties: {
        Visible: {
            defaultValue: false
        }
    },
    dofs: {
        overlay_textureSubRootBn_r: true,
        overlay_textureSubRootBn_t: true,
        overlay_textureSubRootBn_t_2: true,
        overlay_vertexJoint1_t: true,
        overlay_vertexJoint1_t_2: true,
        overlay_vertexJoint2_t: true,
        overlay_vertexJoint2_t_2: true,
        overlay_vertexJoint3_t: true,
        overlay_vertexJoint3_t_2: true,
        overlay_vertexJoint4_t: true,
        overlay_vertexJoint4_t_2: true,
        overlay_vertexJoint5_t: true,
        overlay_vertexJoint5_t_2: true,
        overlay_vertexJoint6_t: true,
        overlay_vertexJoint6_t_2: true,
        overlay_vertexJoint7_t: true,
        overlay_vertexJoint7_t_2: true,
        overlay_vertexJoint8_t: true,
        overlay_vertexJoint8_t_2: true,
        overlay_vertexJoint9_t: true,
        overlay_alphaChannelBn_r: true,
        overlayVisibilityBn_r: true
    }
});
var OverlayLayer = (function (_super) {
    __extends(OverlayLayer, _super);
    function OverlayLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OverlayLayer.getInfo = function () {
        return OverlayInfo;
    };
    OverlayLayer.generateDofs = function (props) {
        var dofs = EyeAndOverlayLayer_1.default._generateDofs(props, "overlay_");
        dofs.overlay_textureSubRootBn_r = props.Rotate / 180 * Math.PI;
        dofs.overlay_textureSubRootBn_t = Conversion_1.default.toMetersX(props.Translate.x);
        dofs.overlay_textureSubRootBn_t_2 = Conversion_1.default.toMetersY(props.Translate.y);
        dofs.overlayVisibilityBn_r = props.Visible === undefined ? 1 : (props.Visible ? 1 : 0);
        dofs.overlay_alphaChannelBn_r = props.Alpha === undefined ? 1 : props.Alpha;
        return dofs;
    };
    return OverlayLayer;
}(EyeAndOverlayLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = OverlayLayer;

},{"../utils/Conversion":23,"./EyeAndOverlayLayer":10}],17:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var OverlayTextureInfo = {
    layerType: "Overlay Texture",
    properties: {
        Texture: {
            type: "texture",
            defaultValue: "",
            blendOperation: "lastOneWins"
        },
        Color: {
            type: "rgb",
            defaultValue: {
                r: 255,
                g: 255,
                b: 255
            },
            blendOperation: "add"
        }
    },
    dofs: {
        overlayTextureInfixBn_r: true,
        overlay_redChannelBn_r: true,
        overlay_greenChannelBn_r: true,
        overlay_blueChannelBn_r: true
    }
};
var OverlayTextureLayer = (function (_super) {
    __extends(OverlayTextureLayer, _super);
    function OverlayTextureLayer() {
        return _super.call(this) || this;
    }
    OverlayTextureLayer.getInfo = function () {
        return OverlayTextureInfo;
    };
    OverlayTextureLayer.generateDofs = function (props) {
        return {
            overlayTextureInfixBn_r: props.Texture,
            overlay_redChannelBn_r: props.Color.r / 255.0,
            overlay_greenChannelBn_r: props.Color.g / 255.0,
            overlay_blueChannelBn_r: props.Color.b / 255.0
        };
    };
    return OverlayTextureLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = OverlayTextureLayer;

},{"../bases/DOFBaseLayer":4}],18:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var EventBaseLayer_1 = require("../bases/EventBaseLayer");
var PixiInfo = {
    layerType: "Pixi",
    properties: {
        Pixi: {
            type: "pixi",
            defaultValue: "",
            blendOperation: "singleton"
        },
        "Attach To Eye": {
            type: "bool",
            defaultValue: false,
            blendOperation: "lastOneWins"
        },
        "Frame Offset": {
            type: "offset",
            defaultValue: "",
            blendOperation: "lastOneWins"
        }
    },
    dofs: {}
};
var PixiLayer = (function (_super) {
    __extends(PixiLayer, _super);
    function PixiLayer() {
        return _super.call(this) || this;
    }
    PixiLayer.getInfo = function () {
        return PixiInfo;
    };
    PixiLayer.generateEvent = function (props, layerNum) {
        return {
            name: "play-pixi",
            payload: {
                file: props.Pixi,
                layerNum: layerNum,
                attach: props["Attach To Eye"],
                offset: props["Frame Offset"],
                effect: false // deprecated
            }
        };
    };
    PixiLayer.isValid = function (generatedEvent) {
        return ((generatedEvent.name === "play-pixi") &&
            (generatedEvent.payload.file.length > 0));
    };
    return PixiLayer;
}(EventBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PixiLayer;

},{"../bases/EventBaseLayer":5}],19:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOFBaseLayer_1 = require("../bases/DOFBaseLayer");
var ReferenceInfo = {
    layerType: "Reference",
    properties: {
        "KeysFileReference": {
            type: "keysref",
            defaultValue: {
                file: ""
            }
        },
        "HoldFinalPose": {
            label: 'Hold Final Pose',
            type: "bool",
            defaultValue: true,
            blendOperation: "lastOneWins"
        }
    }
};
var ReferenceLayer = (function (_super) {
    __extends(ReferenceLayer, _super);
    function ReferenceLayer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ReferenceLayer.generateDofs = function () {
        return {};
    };
    ReferenceLayer.getInfo = function () {
        return ReferenceInfo;
    };
    return ReferenceLayer;
}(DOFBaseLayer_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ReferenceLayer;

},{"../bases/DOFBaseLayer":4}],20:[function(require,module,exports){
/**
 * Operations per property type for blending layers of properties together
 */
"use strict";
var operations = {
    lastOneWins: {
        texture: function (accumulated, newValue) {
            return newValue;
        },
        bool: function (accumulated, newValue) {
            return newValue;
        },
        event: function (accumulated, newValue) {
            return newValue;
        },
        audio: function (accumulated, newValue) {
            return newValue;
        }
    },
    add: {
        float: function (accumulated, newValue, defaultValue) {
            return accumulated + (newValue - defaultValue);
        },
        vector2: function (accumulated, newValue, defaultValue) {
            return {
                x: accumulated.x + (newValue.x - defaultValue.x),
                y: accumulated.y + (newValue.y - defaultValue.y)
            };
        },
        rgb: function (accumulated, newValue, defaultValue) {
            return {
                r: accumulated.r + (newValue.r - defaultValue.r),
                g: accumulated.g + (newValue.g - defaultValue.g),
                b: accumulated.b + (newValue.b - defaultValue.b)
            };
        }
    },
    multiply: {
        float: function (accumulated, newValue) {
            return accumulated * newValue;
        },
        vector2: function (accumulated, newValue) {
            return {
                x: accumulated.x * newValue.x,
                y: accumulated.y * newValue.y
            };
        },
        rgb: function (accumulated, newValue) {
            return {
                r: accumulated.r * newValue.r,
                g: accumulated.g * newValue.g,
                b: accumulated.b * newValue.b
            };
        }
    },
    singleton: {
        video: function (accumulated, newValue) {
            return newValue;
        }
    },
    inverse: {
        lastOneWins: {
            texture: function (accumulated) {
                return accumulated;
            },
            bool: function (accumulated) {
                return accumulated;
            }
        },
        singleton: {
            video: function (accumulated) {
                return accumulated;
            }
        },
        add: {
            float: function (accumulated, newValue, defaultValue) {
                return (newValue + defaultValue) - accumulated;
            },
            vector2: function (accumulated, newValue, defaultValue) {
                return {
                    x: (newValue.x + defaultValue.x) - accumulated.x,
                    y: (newValue.y + defaultValue.y) - accumulated.y
                };
            },
            rgb: function (accumulated, newValue, defaultValue) {
                return {
                    r: (newValue.r + defaultValue.r) - accumulated.r,
                    g: (newValue.g + defaultValue.g) - accumulated.g,
                    b: (newValue.b + defaultValue.b) - accumulated.b
                };
            }
        },
        multiply: {
            float: function (accumulated, newValue) {
                return accumulated / newValue;
            },
            vector2: function (accumulated, newValue) {
                return {
                    x: newValue.x / accumulated.x,
                    y: newValue.y / accumulated.y
                };
            },
            rgb: function (accumulated, newValue) {
                return {
                    r: newValue.r / accumulated.r,
                    g: newValue.g / accumulated.g,
                    b: newValue.b / accumulated.b
                };
            }
        }
    }
};
var BlendOperations = (function () {
    function BlendOperations() {
    }
    BlendOperations.inverse = function (operation, type) {
        // Operation none for any type is the same as lastOneWins for textures
        if (operation === 'none') {
            return operations.inverse.lastOneWins.texture;
        }
        else {
            return operations.inverse[operation][type];
        }
    };
    BlendOperations.normal = function (operation, type) {
        // Operation none for any type is the same as lastOneWins for textures
        if (operation === 'none') {
            return operations.lastOneWins.texture;
        }
        else {
            if (!(operation in operations)) {
                console.log("operation " + operation + " not found in " + Object.keys(operations));
            }
            else {
                return operations[operation][type];
            }
        }
    };
    return BlendOperations;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BlendOperations;

},{}],21:[function(require,module,exports){
/**
 * This file implements meta data associated with .keys and .anim files and conversions between them
 *  - mapping the internal animationutilities channel names to more user friendly
 *    names for display in the jibo animation tool and for developers who
 *    will be manipulating animation tool files themselves
 *  - information about converting between values in each file format. For example
 *    converting between radians and degrees for rotation angles
 */
"use strict";
// Animation robot DOFs and info (currently only includes layer type and display name;
// keeping as object in case we want other data here.
var dofInfo = {
    // BODY
    "topSection_r": {
        type: "Body",
        name: "Head"
    },
    "middleSection_r": {
        type: "Body",
        name: "Torso"
    },
    "bottomSection_r": {
        type: "Body",
        name: "Pelvis"
    },
    // EYE
    "eyeSubRootBn_r": {
        type: "Eye",
        name: "Rotate"
    },
    "eyeSubRootBn_t": {
        type: "Eye",
        name: "Translate"
    },
    "eyeSubRootBn_t_2": {
        type: "Eye",
        name: "Translate"
    },
    "vertexJoint1_t": {
        type: "Eye"
    },
    "vertexJoint1_t_2": {
        type: "Eye"
    },
    "vertexJoint2_t": {
        type: "Eye"
    },
    "vertexJoint2_t_2": {
        type: "Eye"
    },
    "vertexJoint3_t": {
        type: "Eye"
    },
    "vertexJoint3_t_2": {
        type: "Eye"
    },
    "vertexJoint4_t": {
        type: "Eye"
    },
    "vertexJoint4_t_2`": {
        type: "Eye"
    },
    "vertexJoint5_t": {
        type: "Eye"
    },
    "vertexJoint5_t_2": {
        type: "Eye"
    },
    "vertexJoint6_t": {
        type: "Eye"
    },
    "vertexJoint6_t_2": {
        type: "Eye"
    },
    "vertexJoint7_t": {
        type: "Eye"
    },
    "vertexJoint7_t_2": {
        type: "Eye"
    },
    "vertexJoint8_t": {
        type: "Eye"
    },
    "vertexJoint8_t_2": {
        type: "Eye"
    },
    "vertexJoint9_t": {
        type: "Eye"
    },
    "vertexJoint9_t_2": {
        type: "Eye"
    },
    // EYE TEXTURE
    "eyeTextureInfixBn_r": {
        type: "Eye Texture"
    },
    "eye_redChannelBn_r": {
        type: "Eye Texture"
    },
    "eye_greenChannelBn_r": {
        type: "Eye Texture"
    },
    "eye_blueChannelBn_r": {
        type: "Eye Texture"
    },
    // OVERLAY
    "overlay_textureSubRootBn_r": {
        type: "Overlay",
        name: "Rotate"
    },
    "overlay_textureSubRootBn_t": {
        type: "Overlay",
        name: "Translate"
    },
    "overlay_textureSubRootBn_t_2": {
        type: "Overlay",
        name: "Translate"
    },
    "overlay_vertexJoint1_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint1_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint2_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint2_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint3_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint3_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint4_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint4_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint5_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint5_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint6_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint6_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint7_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint7_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint8_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint8_t_2": {
        type: "Overlay"
    },
    "overlay_vertexJoint9_t": {
        type: "Overlay"
    },
    "overlay_vertexJoint9_t_2": {
        type: "Overlay"
    },
    // OVERLAY TEXURE
    "overlayTextureInfixBn_r": {
        type: "Overlay Texture"
    },
    "overlay_redChannelBn_r": {
        type: "Overlay Texture"
    },
    overlay_greenChannelBn_r: {
        type: "Overlay Texture"
    },
    "overlay_blueChannelBn_r": {
        type: "Overlay Texture"
    },
    // BACKGROUND TEXTURE
    "screenBGTextureInfixBn_r": {
        type: "Background Texture"
    },
    "screenBG_redChannelBn_r": {
        type: "Background Texture"
    },
    "screenBG_greenChannelBn_r": {
        type: "Background Texture"
    },
    "screenBG_blueChannelBn_r": {
        type: "Background Texture"
    },
    // LED
    "lightring_redChannelBn_r": {
        type: "LED"
    },
    "lightring_greenChannelBn_r": {
        type: "LED"
    },
    "lightring_blueChannelBn_r": {
        type: "LED"
    }
};
var Channels = (function () {
    function Channels() {
    }
    Channels.dofNameToDisplayType = function (dofName) {
        return dofInfo[dofName].type;
    };
    Channels.dofNameToDisplayName = function (dofName) {
        var displayName = dofInfo[dofName].name;
        return (displayName !== undefined) ? displayName : Channels.dofNameToDisplayType(dofName);
    };
    Channels.displayNameToDof = function (displayName) {
        for (var dofName in dofInfo) {
            if (dofInfo[dofName].name === displayName) {
                return dofName;
            }
        }
        return;
    };
    Channels.radiansToDegrees = function (radians) {
        return radians / Math.PI * 180;
    };
    Channels.degreesToRadians = function (degree) {
        return degree / 180 * Math.PI;
    };
    Channels.framesToSeconds = function (frame, frameRate) {
        return frame / frameRate;
    };
    Channels.secondsToFrames = function (timeInSeconds, totalTimeInSeconds, totalFrames) {
        return Math.round(timeInSeconds / totalTimeInSeconds * (totalFrames));
    };
    Channels.toTextureURL = function (jiboConfig, displayValue) {
        return this.baseTextureURL + this.getPaddedValueString(displayValue) + ".png";
    };
    Channels.toBGTextureURL = function (jiboConfig, displayValue) {
        return this.baseBGURL + this.getPaddedValueString(displayValue) + ".png";
    };
    return Channels;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Channels;

},{}],22:[function(require,module,exports){
"use strict";
var hsv = require("color-space/hsv");
// function clampColorComponent(v){
//     return Math.floor(Math.max(0, Math.min(255, v)));
// }
/**
 * hsv value range is:
 *  h = [0,360],
 *  s = [0,100],
 *  v = [0,100]
 *
 * Uses range/values mapped out in this interface:
 * http://www.colorpicker.com/
 */
var Colors = (function () {
    function Colors() {
    }
    Colors.hsvToRgb = function (h, s, v) {
        var rgb = hsv.rgb([h, s, v]);
        return {
            r: rgb[0],
            g: rgb[1],
            b: rgb[2]
        };
    };
    return Colors;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Colors;

},{"color-space/hsv":undefined}],23:[function(require,module,exports){
"use strict";
var 
// These numbers come from here:
// https://github.jibo.com/sdk/geometry-config/blob/9206df71556c925876362abbd3c382a1e2923fea/P1.0/jibo.jscene#L10-L11
// They are computed based on the CAD model that we have in Maya and is used for exporting our 3D visualizers
// meshes
WIDTH_METERS = 0.13002148, HEIGHT_METERS = 0.073142126, WIDTH_PIXELS = 1280, HEIGHT_PIXELS = 720;
var Conversion = (function () {
    function Conversion() {
    }
    Conversion.toMetersX = function (xInPixels) {
        return (xInPixels) / WIDTH_PIXELS * WIDTH_METERS;
    };
    Conversion.toMetersY = function (yInPixels) {
        return (yInPixels) / HEIGHT_PIXELS * HEIGHT_METERS * -1;
    };
    Conversion.toPixelsX = function (xInMeters) {
        return xInMeters / WIDTH_METERS * WIDTH_PIXELS;
    };
    Conversion.toPixelsY = function (yInMeters) {
        return yInMeters / HEIGHT_METERS * HEIGHT_PIXELS * -1;
    };
    return Conversion;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Conversion;

},{}],24:[function(require,module,exports){
"use strict";
var uuid = require("uuid");
var IDUtils = (function () {
    function IDUtils() {
    }
    IDUtils.create = function () {
        // I dislike the dashes, so I strip them out. This doesn't not affect
        // the chance of collisions over including them
        return uuid.v4().replace(/-/g, "");
    };
    return IDUtils;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = IDUtils;

},{"uuid":undefined}],25:[function(require,module,exports){
// Initialize the data with a default set of values for the necessary channels to have jibo animating his top, middle, and bottom body rotations.
// This includes some eye and overlay texture related default values in order to make sure Jibo's eye is showing centered on his screen
"use strict";
var fs = require("fs");
var path = require("path");
var JiboKeyframeInfo_1 = require("../JiboKeyframeInfo");
var Channels_1 = require("./Channels");
var Runtime_1 = require("../Runtime");
var animData = {
    "header": {
        "fileType": "Animation",
        "version": "0.001",
        "creationTime": 1418336756906
    },
    "content": {
        "name": "AnythingElse_01.dae",
        "channels": [],
        "events": []
    }
};
var KeysUtils = (function () {
    function KeysUtils() {
    }
    KeysUtils.keysToAnimObjectAsync = function (keysPath, cb) {
        // Load .keys file asynchronously
        fs.readFile(keysPath, 'utf8', function (err, data) {
            if (err) {
                console.log("\nError parsing \"" + keysPath + "\". " + err + "\n");
                cb(null);
            }
            else {
                var keysData = JSON.parse(data);
                JiboKeyframeInfo_1.default.onLoad(keysData);
                // use the base name of the uri for the anim data "name"
                var baseName = path.basename(keysPath);
                var animObj = KeysUtils.keysObjToAnimObject(keysData, baseName);
                cb(animObj);
            }
        });
    };
    KeysUtils.keysToAnim = function (keysPath, animPath) {
        console.log(keysPath);
        var animData = KeysUtils.keysToAnimObject(keysPath);
        try {
            fs.writeFileSync(animPath, JSON.stringify(animData, null, '    '), 'utf8');
        }
        catch (error) {
            console.log("\nERROROROROROR\n");
            process.exit(1);
        }
    };
    KeysUtils.keysToAnimObject = function (keysPath) {
        var keysData;
        try {
            // Load .keys file
            keysData = JSON.parse(fs.readFileSync(keysPath, 'utf8'));
        }
        catch (error) {
            console.log("\nError parsing \"" + keysPath + "\". " + error + "\n");
            process.exit(1);
        }
        // use the base name of the uri for the anim data "name"
        var baseName = path.basename(keysPath);
        return KeysUtils.keysObjToAnimObject(keysData, baseName);
    };
    KeysUtils.keysObjToAnimObject = function (keysData, name) {
        JiboKeyframeInfo_1.default.onLoad(keysData);
        /*
         * Fill out "header" fields that need to be changed. Should look like so
         *
         *	"header": {
         *     "fileType": "DOFAnimation",
         *      "version": "0.001",
         *      "creationTime": 1418336756906
         *  },
         */
        animData.header.creationTime = (new Date()).getTime();
        animData.header.version = keysData.version;
        /*
         * Fill out "content" fields that need to be changed. Should look like so
         *
         *	"content": {
         *      "name": "example.keys",
     *          "channels": [
                    {
                        "name": "Top",
                        "type": "number",
                        "keyframes": [
                            {
                                "time": 0,
                                "rotation": 0
                            },
                            ...
                        ]
                    },
                    ...
                ],
                "events": [
                    {
                        "time": 0,
                        "payload": {object}
                    },
                    ...
                ]
         *  },
         */
        // reset data!
        animData.content.name = name;
        animData.content.channels = [];
        animData.content.events = [];
        var animChannels = {};
        // For each frame calculate all dof and event values and save them to file
        for (var frame = 0; frame < keysData.duration; frame++) {
            var timeInSeconds = Channels_1.default.framesToSeconds(frame, keysData.framerate);
            var dofs = Runtime_1.default.evaluateAllDOFLayers(keysData, JiboKeyframeInfo_1.default, timeInSeconds);
            var events = Runtime_1.default.evaluateAllEventLayers(keysData, JiboKeyframeInfo_1.default, timeInSeconds);
            for (var dofName in dofs) {
                if (typeof animChannels[dofName] === "undefined") {
                    animChannels[dofName] = {
                        dofName: dofName,
                        length: keysData.duration / keysData.framerate,
                        times: [timeInSeconds],
                        values: [dofs[dofName]]
                    };
                    animData.content.channels.push(animChannels[dofName]);
                }
                else {
                    animChannels[dofName].times.push(timeInSeconds);
                    animChannels[dofName].values.push(dofs[dofName]);
                }
            }
            for (var i = 0; i < events.length; i++) {
                var animEvent = {
                    time: timeInSeconds,
                    eventName: events[i].name,
                    payload: events[i].payload
                };
                animData.content.events.push(animEvent);
            }
        }
        return animData;
    };
    return KeysUtils;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = KeysUtils;

},{"../JiboKeyframeInfo":1,"../Runtime":2,"./Channels":21,"fs":undefined,"path":undefined}],26:[function(require,module,exports){
/**
 * keyframes entry point.
 */
"use strict";
/**
 * This assumes that some higher level logic has already determined that the currentTime is between the first
 * and last keyframe's times.
 * That is, do not call this with a currentTime < first-keyframe-time or currentTime >= last-keyframe-time.
 * Returns: start-keyframe-idx and end-keyframe-idx where: start-keyframe-time <= currentTime < end-keyframe-time
 */
var Search = (function () {
    function Search() {
    }
    Search.keyframeSearch = function (keyframes, currentTime) {
        var lo = 0, hi = keyframes.length - 2, mid /*,
        steps = 0*/;
        while (lo <= hi) {
            mid = Math.floor((lo + hi) / 2);
            //steps++;
            //console.log(`mid: ${mid}, lo: ${lo}, hi: ${hi} start: ${keyframes[mid].time}, end: ${keyframes[mid+1].time}`)
            if (currentTime >= keyframes[mid + 1].time) {
                //console.log('moving up');
                lo = mid + 1;
            }
            else if (currentTime < keyframes[mid].time) {
                //console.log('moving down');
                hi = mid - 1;
            }
            else {
                //console.log(`steps: ${steps}`);
                return {
                    start: mid,
                    end: mid + 1
                };
            }
        }
    };
    Search.keyframeSearchAbsolute = function (keyframes, keyframeTime) {
        if (keyframes.length === 0) {
            return undefined;
        }
        var lo = 0;
        var hi = keyframes.length - 1;
        var mid;
        while (lo <= hi) {
            mid = Math.floor((lo + hi) / 2);
            var midTime = keyframes[mid].time;
            if (keyframeTime === midTime) {
                return mid; // found it
            }
            else if (keyframeTime > midTime) {
                // if our time is greater, then move up
                lo = mid + 1;
            }
            else {
                // if our time is less, then move down
                hi = mid - 1;
            }
        }
        // not found
        return undefined;
    };
    return Search;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Search;

},{}],27:[function(require,module,exports){
/**
 * A set of functions for blending between a start and end keyframe for a
 * variety of value types. The name of the function matches the value type
 * it knows how to blend
 */
"use strict";
/**
 * These are the set of easing functions our animation system supports. They
 * are used by the animation system to create variations of speed, acceleration,
 * overshooting, and undershooting of blending values.
 *
 * The `t` variable passed to each easing function has the value range: [0,1]. The
 * functions each return a float that should be used to create the usual weighted
 * blend of two values to be tweened. Example
 *
 * tweenFloats(start, end, t) {
 *   let f = easing.linear(t);
 *   return start*(1-f) + end*f;
 * }
 */
var _ = require("lodash");
var easesFunctions = require("eases");
/**
 * For now we support all of the easing functions in 'eases' module.
 * There's some good stuff here: http://www.rich-harris.co.uk/ramjet/
 */
var easing = _.clone(easesFunctions);
/**
 * "snap" always returns a weight fully weighted to the last keyframe
 */
easing.snap = function () { return 0; };
var Tweening = (function () {
    function Tweening() {
    }
    /**
     * getEasingFunc - Resolve which easing function is desired
     *
     * @param  {function|string} easingFuncOrString function or name of a built-in easing function that calculates the blending weights
     * @return {function} the resolved easing function
     */
    Tweening.getEasingFunc = function (easingFuncOrString) {
        if (typeof easingFuncOrString === "function") {
            return easingFuncOrString;
        }
        else {
            return easing[easingFuncOrString];
        }
    };
    /**
     * @param {number} start - Start value
     * @param {number} end - End Value
     * @param {number} time - time between [0,1] indicating where in a blend we are
     * @param {function|string} easing function or name of a built-in easing function that calculates the blending weights
     */
    Tweening.float = function (start, end, time, easing) {
        var f = Tweening.getEasingFunc(easing)(time);
        return start * (1 - f) + end * f;
    };
    /**
     * @param {object} start - Start value
     * @param {object} end - End Value
     * @param {number} time - time between [0,1] indicating where in a blend we are
     * @param {function|string} easing function or name of a built-in easing function that calculates the blending weights
     */
    Tweening.vector2 = function (start, end, time, easing) {
        var f = Tweening.getEasingFunc(easing)(time);
        return {
            x: start.x * (1 - f) + end.x * f,
            y: start.y * (1 - f) + end.y * f
        };
    };
    /**
     * @param {object} start - Start value
     * @param {object} end - End Value
     * @param {number} time - time between [0,1] indicating where in a blend we are
     * @param {function|string} easing function or name of a built-in easing function that calculates the blending weights
     */
    Tweening.rgb = function (start, end, time, easing) {
        var f = Tweening.getEasingFunc(easing)(time);
        return {
            r: start.r * (1 - f) + end.r * f,
            g: start.g * (1 - f) + end.g * f,
            b: start.b * (1 - f) + end.b * f
        };
    };
    /**
     * @param {number} start - Start value
     * @param {number} end - End Value
     * @param {number} time - time between [0,1] indicating where in a blend we are
     * @param {function|string} easing function or name of a built-in easing function that calculates the blending weights
     */
    Tweening.color = function (start, end, time /*, easing*/) {
        var f = Tweening.getEasingFunc("linear")(time);
        return start * (1 - f) + end * f;
    };
    Tweening.texture = function (start /*, end, time, easing*/) {
        return start;
    };
    Tweening.video = function (start /*, end, time, easing*/) {
        return start;
    };
    Tweening.bool = function (start) {
        return start;
    };
    Tweening.enum = function (start /*, end, time, easing*/) {
        return start;
    };
    Tweening.pixi = function (start) {
        return start;
    };
    Tweening.keysref = function (start) {
        return start;
    };
    return Tweening;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Tweening;

},{"eases":undefined,"lodash":undefined}],28:[function(require,module,exports){
"use strict";
/**
 * keyframes entry point.
 */
var KeysUtils_1 = require("./utils/KeysUtils");
var Channels_1 = require("./utils/Channels");
var IDUtils_1 = require("./utils/IDUtils");
var Conversion_1 = require("./utils/Conversion");
var Search_1 = require("./utils/Search");
var Tweening_1 = require("./utils/Tweening");
var JiboKeyframeInfo_1 = require("./JiboKeyframeInfo");
var Runtime_1 = require("./Runtime");
module.exports = {
    JiboKeyframeInfo: JiboKeyframeInfo_1.default,
    keysToAnimObject: KeysUtils_1.default.keysToAnimObject,
    keysToAnimObjectAsync: KeysUtils_1.default.keysToAnimObjectAsync,
    keysObjToAnimObject: KeysUtils_1.default.keysObjToAnimObject,
    keysToAnim: KeysUtils_1.default.keysToAnim,
    search: Search_1.default,
    tweening: Tweening_1.default,
    runtime: Runtime_1.default,
    channels: Channels_1.default,
    generateId: IDUtils_1.default.create,
    conversion: Conversion_1.default
};

},{"./JiboKeyframeInfo":1,"./Runtime":2,"./utils/Channels":21,"./utils/Conversion":23,"./utils/IDUtils":24,"./utils/KeysUtils":25,"./utils/Search":26,"./utils/Tweening":27}]},{},[28])(28)
});
