// jibo-common-types/lib/jibo-common-types.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboCommonTypes = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var uuid = require("uuid");
var GlobalCommand;
(function (GlobalCommand) {
    GlobalCommand[GlobalCommand["SLEEP"] = 'sleep'] = "SLEEP";
    GlobalCommand[GlobalCommand["HELP"] = 'help'] = "HELP";
    GlobalCommand[GlobalCommand["STOP"] = 'stop'] = "STOP";
    GlobalCommand[GlobalCommand["PAUSE"] = 'pause'] = "PAUSE";
    GlobalCommand[GlobalCommand["VOLUME"] = 'volume'] = "VOLUME";
    GlobalCommand[GlobalCommand["WHATCANIDO"] = 'whatCanIDo'] = "WHATCANIDO";
})(GlobalCommand = exports.GlobalCommand || (exports.GlobalCommand = {}));
;
var NLUResponse = (function () {
    function NLUResponse() {
    }
    return NLUResponse;
}());
exports.NLUResponse = NLUResponse;
var REQ_STATUS;
(function (REQ_STATUS) {
    REQ_STATUS[REQ_STATUS["OK"] = 'OK'] = "OK";
    REQ_STATUS[REQ_STATUS["ERROR"] = 'ERROR'] = "ERROR";
})(REQ_STATUS = exports.REQ_STATUS || (exports.REQ_STATUS = {}));
;
var NLUMessages;
(function (NLUMessages) {
    NLUMessages[NLUMessages["RESET_MEMORY"] = "NLU_EXECUTED_RESET_MEMORY"] = "RESET_MEMORY";
})(NLUMessages = exports.NLUMessages || (exports.NLUMessages = {}));
;
var ListenOptions = (function () {
    function ListenOptions(options) {
        this.hotphrase = options ? options.hotphrase : "";
        this.speech_to_text = options ? options.speech_to_text : false;
        this.parseIncremental = options ? options.parseIncremental : false;
        this.incremental = options ? options.incremental : false;
        this.language = options ? options.language : 'en-US';
        this.verbose = options ? options.verbose : false;
        this.forceGlobal = options ? !!options.forceGlobal : false;
        this.expectedDuration = options ? options.expectedDuration : 'medium';
        this.parseAllASR = options ? options.parseAllASR : false;
        this.metadata = options ? options.metadata : null;
    }
    return ListenOptions;
}());
exports.ListenOptions = ListenOptions;
var ParseResults = (function () {
    function ParseResults(Input, NLParse, launchParse, heuristic_score) {
        this.Input = Input || "";
        this.NLParse = NLParse;
        this.launchParse = launchParse;
        this.heuristic_score = heuristic_score ? heuristic_score : 0;
    }
    return ParseResults;
}());
exports.ParseResults = ParseResults;
var Utterance = (function () {
    function Utterance(options) {
        this.utterance = options.utterance || "";
        this.score = options.score || 0;
    }
    return Utterance;
}());
exports.Utterance = Utterance;
var ListenResults = (function () {
    function ListenResults(input, parse, otherResults) {
        if (!input) {
            input = {
                utterance: "",
                score: 0
            };
        }
        this.result = parse || new ParseResults(input.utterance);
        this.otherResults = otherResults || [];
    }
    return ListenResults;
}());
exports.ListenResults = ListenResults;
var PronunciationOptions = (function (_super) {
    __extends(PronunciationOptions, _super);
    function PronunciationOptions(word, listenerId, verbose, collectExample) {
        var _this = _super.call(this) || this;
        _this.word = word || "";
        _this.listenerId = listenerId || uuid.v1();
        _this.verbose = verbose || false;
        _this.collectExample = collectExample || true;
        return _this;
    }
    return PronunciationOptions;
}(ListenOptions));
exports.PronunciationOptions = PronunciationOptions;
var ASRRequestOptions = (function () {
    function ASRRequestOptions(options) {
        if (!options) {
            options = {};
        }
        this.authenticateSpeaker = options.authenticateSpeaker;
        this.bargein = options.bargein;
        this.detectEnd = options.detectEnd;
        this.expectedDuration = options.expectedDruation;
        this.heyJibo = options.heyJibo;
        this.incremental = options.incremental;
        this.metadata = options.metadata;
        this.nbest = options.nbest;
        this.parseFinal = options.parseFinal;
        this.parseIncremental = options.parseIncremental;
        this.precompiledHandle = options.precompiledHandle;
        this.speech_to_text = options.speech_to_text;
        this.hotphrase = options.hotphrase;
        this.language = options.language;
        this.known_speaker_id = options.known_speaker_id;
    }
    return ASRRequestOptions;
}());
exports.ASRRequestOptions = ASRRequestOptions;
exports.SpeakerIdStatus = {
    ACCEPTED: 'ACCEPTED',
    NOT_USED: 'NOT-USED',
    REJECTED: 'REJECTED',
    ID_TI_NOT_AVAILABLE: 'ID-TI-NOT-AVAILABLE'
};
exports.NLU_STATUS = {
    CONNECTED: 'CONNECTED',
    DISCONNECTED: 'DISCONNECTED'
};
exports.ParseStatus = {
    NO_PARSE: 'NO-PARSE',
    GOT_PARSE: 'GOT-PARSE'
};
var SpeakerId = (function () {
    function SpeakerId(speakerId) {
        this.status = "";
        this.speakers = [];
        this.speakers = speakerId ? speakerId.speakers : [new Speaker()];
        this.status = speakerId ? speakerId.status : exports.SpeakerIdStatus.NOT_USED;
        if (speakerId && speakerId.speaker_ids) {
            this.speakers = speakerId.speaker_ids;
            this.status = speakerId.speaker_id_status;
        }
    }
    return SpeakerId;
}());
exports.SpeakerId = SpeakerId;
var Speaker = (function () {
    function Speaker(speaker, score) {
        this.speaker_id = "";
        this.score = 0;
        this.speaker_id = speaker || "";
        this.score = score || 0;
    }
    return Speaker;
}());
exports.Speaker = Speaker;
exports.UTT_STATUS = {
    SAVING_AUDIO: 'SAVING_AUDIO'
};
exports.NoiseLevel = {
    NOISY_AUDIO: 'NOISY_AUDIO'
};
exports.LoudnessLevel = {
    TOO_LOUD_AUDIO: 'TOO_LOUD_AUDIO',
};
exports.AudioQuality = {
    BAD_AUDIO_SIGNAL: 'BAD_AUDIO_SIGNAL'
};
exports.SpeakerConfidence = {
    HIGH_CONFIDENCE: 'HIGH_CONFIDENCE',
    LOW_CONFIDENCE: 'LOW_CONFIDENCE'
};
var SpeakerIdRecord = (function () {
    function SpeakerIdRecord(speakerIdRecord) {
        Object.assign(this, speakerIdRecord);
    }
    return SpeakerIdRecord;
}());
exports.SpeakerIdRecord = SpeakerIdRecord;
var SID_Status;
(function (SID_Status) {
    SID_Status[SID_Status["ACCEPTED"] = 'ACCEPTED'] = "ACCEPTED";
    SID_Status[SID_Status["NOT_USED"] = 'NOT-USED'] = "NOT_USED";
    SID_Status[SID_Status["REJECTED"] = 'REJECTED'] = "REJECTED";
    SID_Status[SID_Status["ID_TI_NOT_AVAILABLE"] = 'ID-TI-NOT-AVAILABLE'] = "ID_TI_NOT_AVAILABLE";
})(SID_Status = exports.SID_Status || (exports.SID_Status = {}));
;
var IndividualSpeaker = (function () {
    function IndividualSpeaker(speaker) {
        this.ID = "";
        this.Score = 0;
        this.SID_Confidence = 0;
        this.id = "";
        this.score = 0;
        this.sid_confidence = 0;
        Object.assign(this, speaker);
        this.id = this.id || this.ID;
        this.score = this.score || this.Score;
        this.sid_status = this.sid_status || this.SID_Status;
        this.sid_confidence = this.sid_confidence || this.SID_Confidence;
    }
    return IndividualSpeaker;
}());
exports.IndividualSpeaker = IndividualSpeaker;

},{"uuid":undefined}],2:[function(require,module,exports){
"use strict";
var LPSVisualEntityTypes;
(function (LPSVisualEntityTypes) {
    LPSVisualEntityTypes[LPSVisualEntityTypes["PERSON"] = "person"] = "PERSON";
    LPSVisualEntityTypes[LPSVisualEntityTypes["MOTION"] = "motion"] = "MOTION";
    LPSVisualEntityTypes[LPSVisualEntityTypes["WAKE"] = "wake"] = "WAKE";
})(LPSVisualEntityTypes = exports.LPSVisualEntityTypes || (exports.LPSVisualEntityTypes = {}));
;

},{}],3:[function(require,module,exports){
"use strict";
var BodyState = (function () {
    function BodyState() {
        this.miscState = new MiscState();
        this.touchState = new TouchState();
        this.powerState = new PowerState();
    }
    return BodyState;
}());
exports.BodyState = BodyState;
var MiscState = (function () {
    function MiscState() {
        this.ts = [];
        this.main_board_temp = 0;
        this.cpu_temp = 0;
        this.intake_temp = 0;
        this.exhaust_temp = 0;
        this.neck_motor_temp = 0;
        this.torso_motor_temp = 0;
        this.pelvis_motor_temp = 0;
    }
    return MiscState;
}());
exports.MiscState = MiscState;
var TouchState = (function () {
    function TouchState() {
        this.ts = [];
        this.touched = false;
        this.pad_state = [false, false, false, false, false, false];
        this.changed = [0, 0, 0, 0, 0, 0];
    }
    return TouchState;
}());
exports.TouchState = TouchState;
var PowerState = (function () {
    function PowerState() {
        this.ts = [];
        this.source = "";
        this.vsys = 0;
        this.battery = new BatteryState();
    }
    return PowerState;
}());
exports.PowerState = PowerState;
var BatteryState = (function () {
    function BatteryState() {
        this.id = 0;
        this.cycles = 0;
        this.capacity = 0;
        this.max_capacity = 0;
        this.charge_rate = 0;
        this.temp = 0;
    }
    return BatteryState;
}());
exports.BatteryState = BatteryState;
var InputEnergy = (function () {
    function InputEnergy() {
        this.ts = [null, null];
        this.db_rms = null;
        this.db_high = null;
        this.db_mid = null;
        this.db_low = null;
    }
    return InputEnergy;
}());
exports.InputEnergy = InputEnergy;

},{}],4:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./GlobalListenTypes"));
__export(require("./LPSTypes"));
__export(require("./SystemTypes"));

},{"./GlobalListenTypes":1,"./LPSTypes":2,"./SystemTypes":3}]},{},[4])(4)
});
