// @jibo/parser-download/lib/parser-download.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.parserDownload = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const minimist = require("minimist");
const parserDownload_1 = require("./parserDownload");
const path = require("path");
function default_1() {
    let args = minimist(process.argv.slice(2), {
        boolean: ['force32', 'rename', 'verbose'],
        string: ['version', 'url', 'type', 'dir', 'temp'],
        alias: {
            f: 'force32',
            r: 'rename',
            v: 'version',
            u: 'url',
            t: 'type',
            d: 'dir',
            c: 'temp',
            b: 'verbose'
        }
    });
    let packageInfo;
    try {
        packageInfo = require(path.join(process.cwd(), 'package'));
    }
    catch (e) {
    }
    if (packageInfo && packageInfo.parser) {
        if (typeof packageInfo.parser === "string") {
            packageInfo.parser = {
                version: packageInfo.parser
            };
        }
        args = Object.assign({}, packageInfo.parser, args);
    }
    args.logger = function (message) {
        console.log(chalk_1.default.gray('[parser-download]'), message);
    };
    const download = new parserDownload_1.default(args);
    download.start()
        .catch((err) => {
        console.log(chalk_1.default.red("ERROR: " + err));
        process.exit(1);
    });
}
exports.default = default_1;

},{"./parserDownload":2,"chalk":undefined,"minimist":undefined,"path":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs-extra");
const path = require("path");
const nugget = require("nugget");
const extractZip = require("extract-zip");
const envPaths = require("env-paths");
const uuid = require("uuid");
class ParserDownload {
    constructor(optionsOrVersion) {
        let options;
        if (typeof optionsOrVersion === "string") {
            options = { version: optionsOrVersion };
        }
        else {
            options = optionsOrVersion;
        }
        options = Object.assign({}, {
            force32: false,
            rename: false,
            verbose: false,
            version: '',
            type: '',
            logger: console.log.bind(console),
            dir: 'parser',
            url: 'https://github.com/jiborobot/jibo-parser/releases/download/'
        }, options || {});
        this.options = options;
    }
    start() {
        return new Promise((resolve, reject) => {
            const options = this.options;
            if (!options.version) {
                return reject('Must specify a \"version\" to download');
            }
            if (options.type === 'auto') {
                options.type = process.versions.modules;
            }
            let arch = process.arch;
            if (process.platform === 'win32' && options.force32) {
                arch = 'ia32';
            }
            this.name = 'jibo-nlu-js';
            this.archiveName = this.name + '-v' +
                options.version + '-' +
                process.platform + '-' +
                arch +
                (options.type ? '-' + options.type : '') +
                '.zip';
            this.url = options.url + "v" + options.version + "/" + this.archiveName;
            this.cacheDir = envPaths(this.name).cache;
            this.tempDir = path.join(this.cacheDir, uuid());
            this.outputDir = path.resolve(options.dir);
            this.versionPath = path.join(this.outputDir, this.name + '.txt');
            this.archivePath = path.join(this.cacheDir, this.archiveName);
            if (fs.existsSync(this.versionPath) && fs.readFileSync(this.versionPath, 'utf8') === this.archiveName) {
                this.log("Current target matches downloaded. Skipping.");
                return resolve();
            }
            if (!fs.existsSync(this.cacheDir)) {
                this.log("Creating global cache directory " + this.cacheDir);
                fs.mkdirsSync(this.cacheDir);
            }
            const downloadAndExtract = () => {
                this.download()
                    .then(() => {
                    this.extract()
                        .then(resolve)
                        .catch(reject);
                })
                    .catch(reject);
            };
            if (fs.existsSync(this.archivePath)) {
                this.log("Using existing download from cache.");
                this.extract()
                    .then(resolve)
                    .catch((err) => {
                    this.log('Error extracting existing download; re-downloading...');
                    fs.removeSync(this.archivePath);
                    downloadAndExtract();
                });
            }
            else {
                downloadAndExtract();
            }
        });
    }
    log(message) {
        if (this.options.verbose) {
            this.options.logger(message);
        }
    }
    extract() {
        return new Promise((resolve, reject) => {
            this.log("Extracting archive..." + this.archiveName);
            try {
                fs.removeSync(this.outputDir);
                fs.removeSync(this.tempDir);
            }
            catch (e) {
            }
            this.log("Creating new target folder: " + this.outputDir);
            fs.mkdirsSync(this.outputDir);
            fs.mkdirsSync(this.tempDir);
            extractZip(this.archivePath, { dir: this.tempDir }, (err) => {
                if (err) {
                    return reject(err);
                }
                fs.copySync(path.join(this.tempDir, this.name), this.outputDir);
                fs.writeFileSync(this.versionPath, this.archiveName, { encoding: 'utf8' });
                fs.removeSync(this.tempDir);
                this.log("Done with extract.");
                if (this.options.rename) {
                    this.rename();
                }
                resolve();
            });
        });
    }
    download() {
        const options = {
            target: this.archiveName,
            dir: this.cacheDir,
            quiet: !('clearLine' in process.stdout),
            verbose: true,
            strictSSL: true,
            resume: true
        };
        return new Promise((resolve, reject) => {
            this.log("Downloading from URL " + this.url);
            nugget(this.url, options, (err) => {
                if (err) {
                    return reject(err);
                }
                this.log("Download complete.");
                resolve();
            });
        });
    }
    rename() {
        const parserRoot = path.join(this.outputDir, 'build', 'Release');
        this.log("Renaming 'jsjibonlu.node' to 'jsjibonlu.jibo'");
        fs.renameSync(path.join(parserRoot, 'jsjibonlu.node'), path.join(parserRoot, 'jsjibonlu.jibo'));
    }
}
exports.default = ParserDownload;

},{"env-paths":undefined,"extract-zip":undefined,"fs-extra":undefined,"nugget":undefined,"path":undefined,"uuid":undefined}],3:[function(require,module,exports){
"use strict";
const cli_1 = require("./cli");
const parserDownload_1 = require("./parserDownload");
function parserDownload(optionsOrVersion, callback) {
    const download = new parserDownload_1.default(optionsOrVersion);
    download.start()
        .then(() => {
        callback();
    })
        .catch(callback);
    return download;
}
parserDownload.cli = cli_1.default;
module.exports = parserDownload;

},{"./cli":1,"./parserDownload":2}]},{},[3])(3)
});
