// @be/jukebox/index.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.bejukebox = f()}})(function(){var define,module,exports;return (function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const be_framework_1 = require("@be/be-framework");
const jibo = require("jibo");
const MusicLibrary_1 = require("./models/MusicLibrary");
const MusicView_1 = require("./views/MusicView");
class JukeboxSkill extends be_framework_1.BeSkill {
    constructor(assetPack) {
        super(assetPack);
        this.musicView = null;
        this.exiting = false;
        this.screenGestureHandler = null;
    }
    postInit(done) {
        done();
    }
    preload(done) {
        const es = jibo.embodied && jibo.embodied.speech;
        if (es && typeof es.installDelegate === 'function') {
            es.installDelegate(this.assetPack);
        }
        done();
    }
    open(result) {
        this.exiting = false;
        this.subscribeSwipeDown();
        let tracks = [];
        try {
            tracks = MusicLibrary_1.default.scan(this.assetPack);
        }
        catch (err) {
            console.error('[jukebox] failed to load music library:', err);
        }
        this.musicView = new MusicView_1.default(tracks);
    }
    close(done) {
        this.unsubscribeSwipeDown();
        if (this.musicView) {
            this.musicView.cleanup();
            this.musicView = null;
        }
        done();
    }
    subscribeSwipeDown() {
        try {
            const shared = jibo.globalEvents && jibo.globalEvents.shared;
            if (!shared || !shared.screenGesture) {
                return;
            }
            this.screenGestureHandler = (gesture) => {
                if (String(gesture).toLowerCase() !== 'swipedown' || this.exiting) {
                    return;
                }
                this.exiting = true;
                console.log('[jukebox] swipe-down detected -> exiting to idle');
                this.exit();
            };
            shared.screenGesture.on(this.screenGestureHandler);
        }
        catch (err) {
            console.warn('[jukebox] could not subscribe to swipe-down gesture', err);
        }
    }
    unsubscribeSwipeDown() {
        if (!this.screenGestureHandler) {
            return;
        }
        try {
            const shared = jibo.globalEvents && jibo.globalEvents.shared;
            if (shared && shared.screenGesture) {
                shared.screenGesture.removeListener(this.screenGestureHandler);
            }
        }
        catch (err) {
            console.warn('[jukebox] could not unsubscribe swipe-down gesture', err);
        }
        this.screenGestureHandler = null;
    }
}
module.exports = JukeboxSkill;

},{"./models/MusicLibrary":2,"./views/MusicView":3,"@be/be-framework":undefined,"jibo":undefined}],2:[function(require,module,exports){
(function (__dirname){(function (){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jibo = require("jibo");
class MusicLibrary {
    static scan(assetPack) {
        try {
            const fs = require('fs');
            const dir = MusicLibrary.resolveMusicDir(assetPack);
            if (!dir || !fs.existsSync(dir)) {
                console.warn('[jukebox] music dir not found:', dir);
                return [];
            }
            const files = fs.readdirSync(dir)
                .filter((name) => /\.mp3$/i.test(name))
                .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
            const tracks = files.map((file) => ({
                file,
                title: MusicLibrary.prettifyName(file),
                url: MusicLibrary.resolveTrackUrl(file, assetPack)
            }));
            console.log('[jukebox] found', tracks.length, 'track(s) in', dir);
            return tracks;
        }
        catch (err) {
            console.error('[jukebox] failed to scan music folder:', err);
            return [];
        }
    }
    static resolveMusicDir(assetPack) {
        try {
            const PathUtils = jibo.utils && jibo.utils.PathUtils;
            if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
                const uri = PathUtils.getAssetUri('music', assetPack);
                if (uri) {
                    return MusicLibrary.uriToPath(uri);
                }
            }
        }
        catch (err) {
            console.warn('[jukebox] resolveMusicDir via PathUtils failed:', err);
        }
        try {
            const path = require('path');
            return path.join(__dirname, '..', '..', 'music');
        }
        catch (err) {
            return null;
        }
    }
    static resolveTrackUrl(file, assetPack) {
        try {
            const PathUtils = jibo.utils && jibo.utils.PathUtils;
            if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
                const uri = PathUtils.getAssetUri('music/' + file, assetPack);
                if (uri) {
                    return uri;
                }
            }
        }
        catch (err) {
            console.warn('[jukebox] resolveTrackUrl failed for', file, err);
        }
        return './music/' + encodeURIComponent(file);
    }
    static uriToPath(uri) {
        let p = String(uri);
        if (p.indexOf('file://') === 0) {
            p = p.replace(/^file:\/\//, '');
        }
        try {
            p = decodeURIComponent(p);
        }
        catch (e) { }
        return p;
    }
    static prettifyName(file) {
        return file
            .replace(/\.mp3$/i, '')
            .replace(/_/g, ' ')
            .trim();
    }
}
exports.default = MusicLibrary;

}).call(this)}).call(this,"/src/models")

},{"fs":undefined,"jibo":undefined,"path":undefined}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MusicView {
    constructor(tracks) {
        this.index = -1;
        this.rowEls = [];
        this.tracks = tracks || [];
        this.audio = new Audio();
        this.audio.preload = 'metadata';
        this.build();
        this.bindAudio();
    }
    cleanup() {
        try {
            if (this.audio) {
                this.audio.pause();
                this.audio.src = '';
            }
        }
        catch (e) { }
        if (this.root && this.root.parentNode) {
            this.root.parentNode.removeChild(this.root);
        }
        this.root = null;
    }
    build() {
        this.root = document.createElement('div');
        this.applyStyle(this.root, {
            position: 'fixed', left: '0', top: '0', width: '1280px', height: '720px',
            zIndex: '99999', background: 'linear-gradient(160deg, #1b1035 0%, #0b0b16 100%)',
            color: '#fff', fontFamily: 'Helvetica, Arial, sans-serif',
            userSelect: 'none', overflow: 'hidden'
        });
        const header = document.createElement('div');
        this.applyStyle(header, {
            position: 'absolute', left: '0', top: '0', width: '100%', height: '96px',
            display: 'flex', alignItems: 'center', paddingLeft: '48px',
            boxSizing: 'border-box', fontSize: '44px', fontWeight: 'bold',
            letterSpacing: '2px'
        });
        header.textContent = 'Jukebox';
        this.root.appendChild(header);
        this.listEl = document.createElement('div');
        this.applyStyle(this.listEl, {
            position: 'absolute', left: '48px', top: '112px', right: '48px',
            bottom: '190px', overflowY: 'auto', borderRadius: '16px',
            background: 'rgba(255,255,255,0.04)'
        });
        this.root.appendChild(this.listEl);
        if (this.tracks.length === 0) {
            const empty = document.createElement('div');
            this.applyStyle(empty, {
                padding: '48px', fontSize: '30px', lineHeight: '1.5',
                color: 'rgba(255,255,255,0.75)'
            });
            empty.textContent = 'No music found. Drop .mp3 files into the skill\u2019s music folder and reopen the Jukebox.';
            this.listEl.appendChild(empty);
        }
        else {
            this.tracks.forEach((track, i) => this.listEl.appendChild(this.buildRow(track, i)));
        }
        this.root.appendChild(this.buildControls());
        const host = document.body || document.documentElement;
        host.appendChild(this.root);
    }
    buildRow(track, i) {
        const row = document.createElement('div');
        this.applyStyle(row, {
            display: 'flex', alignItems: 'center', height: '84px',
            padding: '0 28px', boxSizing: 'border-box', fontSize: '30px',
            cursor: 'pointer', borderBottom: '1px solid rgba(255,255,255,0.06)'
        });
        const num = document.createElement('span');
        this.applyStyle(num, {
            width: '54px', color: 'rgba(255,255,255,0.45)', fontSize: '24px', flex: '0 0 auto'
        });
        num.textContent = String(i + 1);
        row.appendChild(num);
        const name = document.createElement('span');
        this.applyStyle(name, {
            flex: '1 1 auto', whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis'
        });
        name.textContent = track.title;
        row.appendChild(name);
        row.addEventListener('click', () => this.selectTrack(i));
        this.rowEls[i] = row;
        return row;
    }
    buildControls() {
        const bar = document.createElement('div');
        this.applyStyle(bar, {
            position: 'absolute', left: '0', bottom: '0', width: '100%', height: '170px',
            boxSizing: 'border-box', padding: '18px 48px 24px',
            background: 'rgba(0,0,0,0.35)'
        });
        this.titleEl = document.createElement('div');
        this.applyStyle(this.titleEl, {
            fontSize: '28px', marginBottom: '10px', whiteSpace: 'nowrap',
            overflow: 'hidden', textOverflow: 'ellipsis', color: 'rgba(255,255,255,0.9)'
        });
        this.titleEl.textContent = this.tracks.length ? 'Tap a song to play' : '';
        bar.appendChild(this.titleEl);
        const progressRow = document.createElement('div');
        this.applyStyle(progressRow, { display: 'flex', alignItems: 'center', marginBottom: '14px' });
        this.currentTimeEl = document.createElement('span');
        this.applyStyle(this.currentTimeEl, { width: '72px', fontSize: '22px', flex: '0 0 auto' });
        this.currentTimeEl.textContent = '0:00';
        progressRow.appendChild(this.currentTimeEl);
        this.progressTrack = document.createElement('div');
        this.applyStyle(this.progressTrack, {
            position: 'relative', flex: '1 1 auto', height: '12px', margin: '0 18px',
            background: 'rgba(255,255,255,0.18)', borderRadius: '6px', cursor: 'pointer'
        });
        this.progressFill = document.createElement('div');
        this.applyStyle(this.progressFill, {
            position: 'absolute', left: '0', top: '0', height: '100%', width: '0%',
            background: '#00d1b2', borderRadius: '6px'
        });
        this.progressTrack.appendChild(this.progressFill);
        this.progressTrack.addEventListener('click', (e) => this.onSeek(e));
        progressRow.appendChild(this.progressTrack);
        this.totalTimeEl = document.createElement('span');
        this.applyStyle(this.totalTimeEl, {
            width: '72px', fontSize: '22px', flex: '0 0 auto', textAlign: 'right'
        });
        this.totalTimeEl.textContent = '0:00';
        progressRow.appendChild(this.totalTimeEl);
        bar.appendChild(progressRow);
        const controls = document.createElement('div');
        this.applyStyle(controls, {
            display: 'flex', alignItems: 'center', justifyContent: 'center'
        });
        const prevBtn = this.buildButton('\u23EE', () => this.prev(), false);
        this.playPauseBtn = this.buildButton('\u25B6', () => this.togglePlay(), true);
        const nextBtn = this.buildButton('\u23ED', () => this.next(), false);
        controls.appendChild(prevBtn);
        controls.appendChild(this.playPauseBtn);
        controls.appendChild(nextBtn);
        bar.appendChild(controls);
        return bar;
    }
    buildButton(glyph, onClick, primary) {
        const btn = document.createElement('div');
        const size = primary ? 84 : 64;
        this.applyStyle(btn, {
            width: size + 'px', height: size + 'px', margin: '0 22px',
            borderRadius: '50%', display: 'flex', alignItems: 'center',
            justifyContent: 'center', cursor: 'pointer',
            fontSize: (primary ? 40 : 30) + 'px',
            background: primary ? '#00d1b2' : 'rgba(255,255,255,0.14)',
            color: primary ? '#04231f' : '#fff'
        });
        btn.textContent = glyph;
        btn.addEventListener('click', onClick);
        return btn;
    }
    bindAudio() {
        this.audio.addEventListener('timeupdate', () => this.onTimeUpdate());
        this.audio.addEventListener('loadedmetadata', () => this.onTimeUpdate());
        this.audio.addEventListener('ended', () => this.next());
        this.audio.addEventListener('play', () => this.setPlayGlyph(true));
        this.audio.addEventListener('pause', () => this.setPlayGlyph(false));
        this.audio.addEventListener('error', () => {
            console.error('[jukebox] audio error for', this.audio.src, this.audio.error && this.audio.error.code);
        });
    }
    selectTrack(i) {
        if (i < 0 || i >= this.tracks.length) {
            return;
        }
        this.index = i;
        const track = this.tracks[i];
        this.titleEl.textContent = track.title;
        this.highlightActive();
        this.audio.src = track.url;
        const p = this.audio.play();
        if (p && typeof p.catch === 'function') {
            p.catch((err) => console.error('[jukebox] play() rejected:', err && err.name));
        }
    }
    togglePlay() {
        if (this.index === -1) {
            this.selectTrack(0);
            return;
        }
        if (this.audio.paused) {
            const p = this.audio.play();
            if (p && typeof p.catch === 'function') {
                p.catch((err) => console.error('[jukebox] play() rejected:', err && err.name));
            }
        }
        else {
            this.audio.pause();
        }
    }
    next() {
        if (!this.tracks.length) {
            return;
        }
        this.selectTrack((this.index + 1) % this.tracks.length);
    }
    prev() {
        if (!this.tracks.length) {
            return;
        }
        if (this.audio.currentTime > 3) {
            this.audio.currentTime = 0;
            return;
        }
        const prevIndex = (this.index - 1 + this.tracks.length) % this.tracks.length;
        this.selectTrack(prevIndex);
    }
    onSeek(e) {
        if (!this.audio.duration || isNaN(this.audio.duration)) {
            return;
        }
        const rect = this.progressTrack.getBoundingClientRect();
        const ratio = Math.max(0, Math.min(1, (e.clientX - rect.left) / rect.width));
        this.audio.currentTime = ratio * this.audio.duration;
    }
    onTimeUpdate() {
        const dur = this.audio.duration;
        const cur = this.audio.currentTime;
        const ratio = dur && !isNaN(dur) ? (cur / dur) : 0;
        if (this.progressFill) {
            this.progressFill.style.width = (ratio * 100) + '%';
        }
        if (this.currentTimeEl) {
            this.currentTimeEl.textContent = MusicView.formatTime(cur);
        }
        if (this.totalTimeEl) {
            this.totalTimeEl.textContent = MusicView.formatTime(dur);
        }
    }
    setPlayGlyph(playing) {
        if (this.playPauseBtn) {
            this.playPauseBtn.textContent = playing ? '\u23F8' : '\u25B6';
        }
    }
    highlightActive() {
        this.rowEls.forEach((row, i) => {
            if (!row) {
                return;
            }
            row.style.background = i === this.index ? 'rgba(0,209,178,0.22)' : 'transparent';
        });
    }
    static formatTime(seconds) {
        if (!seconds || isNaN(seconds) || !isFinite(seconds)) {
            return '0:00';
        }
        const m = Math.floor(seconds / 60);
        const s = Math.floor(seconds % 60);
        return m + ':' + (s < 10 ? '0' + s : String(s));
    }
    applyStyle(el, style) {
        for (const key in style) {
            if (style.hasOwnProperty(key)) {
                el.style[key] = style[key];
            }
        }
    }
}
exports.default = MusicView;

},{}]},{},[1])(1)
});
