// jibo-dof-arbiter/lib/jibo-dof-arbiter.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboDofArbiter = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
var animation_utilities_1 = require("animation-utilities");
var OwnerInfo_1 = require("./OwnerInfo");
var DOFArbiterPriorityPolicy_1 = require("./DOFArbiterPriorityPolicy");
var DOFEventDispatcher_1 = require("./DOFEventDispatcher");
var DOFArbiter = (function () {
    function DOFArbiter() {
        this.debug = false;
        this.directRequester = "Direct";
        this.mainAnimationLayer = "default";
        this.graceExpiryPeriodS = -0.001;
        this.defaultOptions = {
            allOrNothing: true
        };
        this.animationUtilities = null;
    }
    DOFArbiter.prototype.init = function (animationUtilities, priorityConfig) {
        if (this.animationUtilities === null) {
            this.animationUtilities = animationUtilities;
            this.dofOwners = new Map();
            this.instanceToDOF = new Map();
            this.builderToOwner = new Map();
            this.ownershipListeners = new Map();
            this.ownershipListenersInOrder = [];
            this.eventDispatcher = DOFEventDispatcher_1.default.getInstance();
            if (priorityConfig === null || priorityConfig === undefined) {
                priorityConfig = {
                    priorityForDirectUsers: 5,
                    priorityForUnknownLabels: 2,
                    priorityEntries: [{ owner: "Attention", priority: 1 }]
                };
            }
            this.policy = new DOFArbiterPriorityPolicy_1.DOFArbiterPriorityPolicy(priorityConfig, this.directRequester);
            var dofNames = this.animationUtilities.getRobotInfo().getDOFNames();
            for (var i = 0; i < dofNames.length; i++) {
                var dofName = dofNames[i];
                this.dofOwners.set(dofName, new OwnerInfo_1.OwnershipInformation(dofName));
            }
            setInterval(this.update.bind(this), 100);
            animationUtilities.addGlobalAnimationListener(this.globalAnimationEventHandler.bind(this));
            animationUtilities.addGlobalLookatListener(this.globalLookatEventHandler.bind(this));
        }
        else {
            console.log("DOFArbiter Warning: initialized multiple times!");
        }
    };
    DOFArbiter.prototype.update = function () {
        var curTime = animation_utilities_1.Clock.currentTime();
        var iter = this.dofOwners.values();
        var nextVal;
        var newlyFreedDOFs = [];
        while (!(nextVal = iter.next()).done) {
            var info = nextVal.value;
            if (info.ownershipStatus === OwnerInfo_1.OwnershipStatusFlag.TIMED_RELEASE) {
                if (curTime.subtract(info.releasedAt) > this.graceExpiryPeriodS) {
                    if (this.debug) {
                        console.log("TIMED_RELEASE for " + info.dof + "(owner " + info.owner + ") expiring, releasing");
                    }
                    info.ownershipStatus = OwnerInfo_1.OwnershipStatusFlag.AVAILABLE;
                    info.owner = null;
                    info.ownerInstance = null;
                    info.releasedAt = null;
                    newlyFreedDOFs.push(info.dof);
                }
            }
        }
        if (newlyFreedDOFs.length > 0) {
            for (var i = this.ownershipListenersInOrder.length - 1; i >= 0; i--) {
                var owner = this.ownershipListenersInOrder[i];
                var listeners = this.ownershipListeners.get(owner);
                if (listeners !== undefined) {
                    for (var j = 0; j < listeners.length; j++) {
                        listeners[j].dofsAvailable(newlyFreedDOFs);
                    }
                }
            }
        }
        this.eventDispatcher.dispatchQueuedEvents();
    };
    DOFArbiter.prototype.playAnimation = function (builder, requester, options) {
        if (builder.layer !== "default") {
            return builder.play();
        }
        if (options === null || options === undefined) {
            options = this.defaultOptions;
        }
        var desiredDOFs = builder.getDOFs();
        var dofsToUse = this.policy.acquire(requester, desiredDOFs, this.dofOwners, options);
        if (this.debug) {
            console.log("DOFArbiter: anim requester:" + requester + " is going to be allowed (" + dofsToUse + ") of (" + desiredDOFs + ")");
        }
        builder.setDOFs(dofsToUse);
        this.builderToOwner.set(builder, requester);
        var instance = builder.play();
        this.builderToOwner.delete(builder);
        if (instance !== null) {
        }
        else {
            console.log("DOFArbiter: Error, Got null anim instance!");
        }
        builder.setDOFs(desiredDOFs);
        this.eventDispatcher.dispatchQueuedEvents();
        if (dofsToUse.length > 0) {
            return instance;
        }
        else {
            return null;
        }
    };
    DOFArbiter.prototype.startLookat = function (builder, requester, target, options) {
        if (options === null || options === undefined) {
            options = this.defaultOptions;
        }
        var desiredDOFs = builder.getDOFs();
        var dofsToUse = this.policy.acquire(requester, desiredDOFs, this.dofOwners, options);
        if (this.debug) {
            console.log("DOFArbiter: lookat requester:" + requester + " is going to be allowed (" + dofsToUse + ") of (" + desiredDOFs + ")");
        }
        builder.setDOFs(dofsToUse);
        this.builderToOwner.set(builder, requester);
        var instance = builder.startLookat(target);
        this.builderToOwner.delete(builder);
        if (instance !== null) {
        }
        else {
            console.log("DOFArbiter: Error, Got null lookat instance!");
        }
        builder.setDOFs(desiredDOFs);
        this.eventDispatcher.dispatchQueuedEvents();
        if (dofsToUse.length > 0) {
            return instance;
        }
        else {
            return null;
        }
    };
    DOFArbiter.prototype.attemptToClaimForInstance = function (requester, instance, desiredDOFs, options) {
        if (options === null || options === undefined) {
            options = this.defaultOptions;
        }
        var dofsToUse = this.policy.acquire(requester, desiredDOFs, this.dofOwners, options);
        this.markInUseByInstance(requester, instance, dofsToUse);
        return dofsToUse;
    };
    DOFArbiter.prototype.getAvailable = function (requester, desiredDOFs, options) {
        if (options === null || options === undefined) {
            options = this.defaultOptions;
        }
        var availableDOFs = this.policy.acquire(requester, desiredDOFs, this.dofOwners, options);
        return availableDOFs;
    };
    DOFArbiter.prototype.getDofsInUse = function (requester) {
        var dofsInUse = [];
        var iter = this.dofOwners.values();
        var nextVal;
        while (!(nextVal = iter.next()).done) {
            var info = nextVal.value;
            if (info.owner === requester) {
                dofsInUse.push(info.dof);
            }
        }
        return dofsInUse;
    };
    DOFArbiter.prototype.addListener = function (forEventsForOwner, listener) {
        var listenersForOwner = this.ownershipListeners.get(forEventsForOwner);
        if (listenersForOwner === undefined) {
            listenersForOwner = [];
            this.ownershipListeners.set(forEventsForOwner, listenersForOwner);
        }
        if (listenersForOwner.indexOf(listener) < 0) {
            listenersForOwner.push(listener);
            this.ownershipListenersInOrder = this.policy.ownerPriorityOrder(Array.from(this.ownershipListeners.keys()));
        }
    };
    DOFArbiter.prototype.removeListener = function (forEventsForOwner, listener) {
        var listenersForOwner = this.ownershipListeners.get(forEventsForOwner);
        if (listenersForOwner !== undefined) {
            var index = listenersForOwner.indexOf(listener);
            if (index >= 0) {
                listenersForOwner.splice(index, 1);
                if (listenersForOwner.length === 0) {
                    this.ownershipListeners.delete(forEventsForOwner);
                }
                this.ownershipListenersInOrder = this.policy.ownerPriorityOrder(Array.from(this.ownershipListeners.keys()));
            }
        }
    };
    DOFArbiter.prototype.notifyListenersOfDOFLoss = function (dofLosses) {
        var ownersLosingDOFs = Object.keys(dofLosses);
        for (var i = 0; i < ownersLosingDOFs.length; i++) {
            var owner = ownersLosingDOFs[i];
            var listenersForOwner = this.ownershipListeners.get(owner);
            if (listenersForOwner !== undefined) {
                var dofsLost = dofLosses[owner];
                for (var j = 0; j < listenersForOwner.length; j++) {
                    listenersForOwner[j].dofsLost(owner, dofsLost);
                }
            }
        }
    };
    DOFArbiter.prototype.notifyListenerOfDOFGain = function (ownerGaining, dofsBeingGained) {
        var listenersForOwner = this.ownershipListeners.get(ownerGaining);
        if (listenersForOwner !== undefined) {
            for (var i = 0; i < listenersForOwner.length; i++) {
                listenersForOwner[i].dofsGained(ownerGaining, dofsBeingGained);
            }
        }
    };
    DOFArbiter.prototype.markInUseByInstance = function (requester, instance, dofsToUse) {
        var dofLosses = {};
        var dofsGained = [];
        this.instanceToDOF.set(instance, dofsToUse);
        for (var i = 0; i < dofsToUse.length; i++) {
            var dofToUse = dofsToUse[i];
            var info = this.dofOwners.get(dofToUse);
            if (info.owner !== requester) {
                dofsGained.push(dofToUse);
                if (info.owner !== null) {
                    if (dofLosses[info.owner] === undefined) {
                        dofLosses[info.owner] = [];
                    }
                    dofLosses[info.owner].push(dofToUse);
                }
            }
            info.owner = requester;
            info.ownerInstance = instance;
            info.ownershipStatus = OwnerInfo_1.OwnershipStatusFlag.ACTIVE_AUTO;
            info.releasedAt = null;
        }
        return { dofLosses: dofLosses, dofGains: { owner: requester, dofs: dofsGained } };
    };
    DOFArbiter.prototype.markAsUsedByTransient = function (requester, dofsToUse) {
        var dofLosses = {};
        var dofsGained = [];
        var curTime = animation_utilities_1.Clock.currentTime();
        for (var i = 0; i < dofsToUse.length; i++) {
            var dofToUse = dofsToUse[i];
            var info = this.dofOwners.get(dofToUse);
            if (info.owner !== requester) {
                dofsGained.push(dofToUse);
                if (info.owner !== null) {
                    if (dofLosses[info.owner] === undefined) {
                        dofLosses[info.owner] = [];
                    }
                    dofLosses[info.owner].push(dofToUse);
                }
            }
            info.owner = requester;
            info.ownershipStatus = OwnerInfo_1.OwnershipStatusFlag.TIMED_RELEASE;
            info.ownerInstance = null;
            info.releasedAt = curTime;
        }
        return { dofLosses: dofLosses, dofGains: { owner: requester, dofs: dofsGained } };
    };
    DOFArbiter.prototype.globalAnimationEventHandler = function (eventName, instance, payload) {
        if (payload.layer === this.mainAnimationLayer || eventName !== "ADDED") {
            this.globalEventHandler(eventName, instance, payload);
        }
    };
    ;
    DOFArbiter.prototype.globalLookatEventHandler = function (eventName, instance, payload) {
        this.globalEventHandler(eventName, instance, payload);
    };
    ;
    DOFArbiter.prototype.globalEventHandler = function (eventName, instance, payload) {
        if (eventName === "ADDED") {
            var dofs = payload.dofs;
            var requester = this.directRequester;
            var dofChanges = void 0;
            var directCall = void 0;
            if (payload['instant']) {
                directCall = true;
                dofChanges = this.markAsUsedByTransient(requester, dofs);
                if (this.debug) {
                    console.log("DOFArbiter: Global ADDED:, marking in-use TRANSIENTLY by " + (directCall ? "DIRECT" : "DA-Managed") + " requester:" + requester + ", description:" + payload["instant"] + " (" + dofs + ")");
                }
            }
            else {
                var assignedOwner = this.builderToOwner.get(instance.getBuilder());
                if (assignedOwner !== undefined) {
                    directCall = false;
                    requester = assignedOwner;
                }
                else {
                    directCall = true;
                }
                dofChanges = this.markInUseByInstance(requester, instance, dofs);
                if (this.debug) {
                    console.log("DOFArbiter: Global ADDED:, marking in-use by " + (directCall ? "DIRECT" : "DA-Managed") + " requester:" + requester + ", instance:" + (instance !== null ? instance.getName() : "null") + " (" + dofs + ")");
                }
            }
            if (directCall) {
                this.notifyListenersOfDOFLoss(dofChanges.dofLosses);
                this.notifyListenerOfDOFGain(dofChanges.dofGains.owner, dofChanges.dofGains.dofs);
            }
            else {
                this.eventDispatcher.queueEvent(this.notifyListenersOfDOFLoss, this, [dofChanges.dofLosses]);
                this.eventDispatcher.queueEvent(this.notifyListenerOfDOFGain, this, [dofChanges.dofGains.owner, dofChanges.dofGains.dofs]);
            }
        }
        else if (eventName === "CANCELLED" || eventName === "STOPPED") {
            var relatedDOFs = this.instanceToDOF.get(instance);
            var didAnyRelease = false;
            if (relatedDOFs !== undefined && relatedDOFs !== null) {
                if (this.debug) {
                    console.log("DOFArbiter: Clearing ownership (" + eventName + ") of all dofs for instance:" + (instance !== null ? instance.getName() : "null") + ", at most " + relatedDOFs.length);
                }
                for (var i = 0; i < relatedDOFs.length; i++) {
                    var relatedDOF = relatedDOFs[i];
                    var info = this.dofOwners.get(relatedDOF);
                    if (info.ownerInstance === instance) {
                        if (this.debug) {
                            console.log("DOFArbiter: \tClearing ACTIVE_AUTO status of dof " + relatedDOF + " to TIMED_RELEASE, owned by " + info.owner);
                        }
                        if (info.ownershipStatus === OwnerInfo_1.OwnershipStatusFlag.ACTIVE_AUTO) {
                            info.ownershipStatus = OwnerInfo_1.OwnershipStatusFlag.TIMED_RELEASE;
                            info.releasedAt = animation_utilities_1.Clock.currentTime();
                            didAnyRelease = true;
                        }
                    }
                    else {
                        if (this.debug) {
                            console.log("DOFArbiter: \tNot modifying status of dof " + relatedDOF + " as it is no longer associated with this instance");
                        }
                    }
                }
                this.instanceToDOF.delete(instance);
            }
            else {
                if (this.debug) {
                    console.log("DOFArbiter: Error, got " + eventName + " for instance " + (instance !== null ? instance.getName() : "null") + " which is unregistered");
                }
            }
            if (didAnyRelease) {
                this.update();
            }
        }
    };
    return DOFArbiter;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = DOFArbiter;

},{"./DOFArbiterPriorityPolicy":2,"./DOFEventDispatcher":3,"./OwnerInfo":4,"animation-utilities":undefined}],2:[function(require,module,exports){
"use strict";
var DOFArbiterPriorityPolicy = (function () {
    function DOFArbiterPriorityPolicy(config, directCommandsLabel) {
        this.priorityMap = new Map();
        this.debug = false;
        this.debugPrintFailedGrab = false;
        this.unknownRequesterIDPriority = config.priorityForUnknownLabels;
        this.priorityMap = new Map();
        for (var i = 0; i < config.priorityEntries.length; i++) {
            this.priorityMap.set(config.priorityEntries[i].owner, config.priorityEntries[i].priority);
        }
        this.priorityMap.set(directCommandsLabel, config.priorityForDirectUsers);
    }
    DOFArbiterPriorityPolicy.prototype.acquire = function (requester, dofs, dofOwners, options) {
        var allowedDOFs = [];
        var newPriority = this.priorityMap.get(requester);
        if (this.debug) {
            console.log("DOFArbiterPriorityPolicy: requester " + requester + " has priority " + newPriority + (newPriority === undefined ? "(=" + this.unknownRequesterIDPriority + ")" : ""));
        }
        if (newPriority === undefined) {
            newPriority = this.unknownRequesterIDPriority;
        }
        for (var i = 0; i < dofs.length; i++) {
            var currentOwnerInfo = dofOwners.get(dofs[i]);
            if (this.debug) {
                console.log("DOFArbiterPriorityPolicy: current owner for " + dofs[i] + " is " + currentOwnerInfo.owner + ", priority:" + (currentOwnerInfo.owner === null ? "N/A" : this.priorityMap.get(currentOwnerInfo.owner)));
            }
            var currentPriority = void 0;
            if (currentOwnerInfo.owner === requester ||
                currentOwnerInfo.owner === null ||
                (currentPriority = this.priorityMap.get(currentOwnerInfo.owner)) < newPriority ||
                (currentPriority === undefined && this.unknownRequesterIDPriority < newPriority)) {
                allowedDOFs.push(dofs[i]);
            }
            else {
                if (this.debug || this.debugPrintFailedGrab) {
                    console.log("DOFArbiterPriorityPolicy: requester " + requester + " denied dof " + dofs[i] + ", owned by " + currentOwnerInfo.owner + "(" + (currentOwnerInfo.ownerInstance && currentOwnerInfo.ownerInstance.getName ? currentOwnerInfo.ownerInstance.getName() : "") + ") with priority " + currentPriority + (currentPriority === undefined ? "(=" + this.unknownRequesterIDPriority + ")" : ""));
                }
                if (options.allOrNothing) {
                    return [];
                }
            }
        }
        return allowedDOFs;
    };
    DOFArbiterPriorityPolicy.prototype.ownerPriorityOrder = function (owners) {
        var ownersWithPriority = [];
        for (var i = 0; i < owners.length; i++) {
            var owner = owners[i];
            var ownerPriority = this.priorityMap.get(owner);
            if (ownerPriority === undefined) {
                ownerPriority = this.unknownRequesterIDPriority;
            }
            ownersWithPriority.push({ owner: owner, priority: ownerPriority });
        }
        ownersWithPriority.sort(function (a, b) { return a.priority - b.priority; });
        var ownersInOrder = [];
        for (var i = 0; i < ownersWithPriority.length; i++) {
            ownersInOrder.push(ownersWithPriority[i].owner);
        }
        return ownersInOrder;
    };
    return DOFArbiterPriorityPolicy;
}());
exports.DOFArbiterPriorityPolicy = DOFArbiterPriorityPolicy;

},{}],3:[function(require,module,exports){
"use strict";
var _instance = null;
var _locked = false;
var DOFEventDispatcher = (function () {
    function DOFEventDispatcher() {
        this.eventQueue = [];
        if (_locked) {
            throw Error("Use 'getInstance()'");
        }
    }
    DOFEventDispatcher.getInstance = function () {
        return _instance;
    };
    DOFEventDispatcher.prototype.queueEvent = function (theFunction, theObject, theArgs) {
        if (theFunction === null || theFunction === undefined) {
            console.log("Error, null/undefined function queued in DOFEventDispatcher!\n" + new Error().stack);
        }
        this.eventQueue.push({ f: theFunction, o: theObject, a: theArgs });
    };
    DOFEventDispatcher.prototype.dispatchQueuedEvents = function () {
        for (var i = 0; i < this.eventQueue.length; i++) {
            var e = this.eventQueue[i];
            e.f.apply(e.o, e.a);
        }
        this.eventQueue.length = 0;
    };
    return DOFEventDispatcher;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = DOFEventDispatcher;
_instance = new DOFEventDispatcher();
_locked = true;

},{}],4:[function(require,module,exports){
"use strict";
var OwnershipStatusFlag;
(function (OwnershipStatusFlag) {
    OwnershipStatusFlag[OwnershipStatusFlag["AVAILABLE"] = 0] = "AVAILABLE";
    OwnershipStatusFlag[OwnershipStatusFlag["ACTIVE_AUTO"] = 1] = "ACTIVE_AUTO";
    OwnershipStatusFlag[OwnershipStatusFlag["ACTIVE_LOCKED"] = 2] = "ACTIVE_LOCKED";
    OwnershipStatusFlag[OwnershipStatusFlag["TIMED_RELEASE"] = 3] = "TIMED_RELEASE";
})(OwnershipStatusFlag = exports.OwnershipStatusFlag || (exports.OwnershipStatusFlag = {}));
var OwnershipInformation = (function () {
    function OwnershipInformation(dof) {
        this.dof = dof;
        this.owner = null;
        this.ownershipStatus = OwnershipStatusFlag.AVAILABLE;
        this.releasedAt = null;
    }
    return OwnershipInformation;
}());
exports.OwnershipInformation = OwnershipInformation;

},{}],5:[function(require,module,exports){
"use strict";
var DOFArbiter_1 = require("./DOFArbiter");
exports.DOFArbiter = DOFArbiter_1.default;

},{"./DOFArbiter":1}],6:[function(require,module,exports){
var DOFArbiter = require("./main");
module.exports = DOFArbiter;

},{"./main":5}]},{},[6])(6)
});
