// @be/recipe/src/index.ts
/// <reference path="../typings-local/index.d.ts" />

/*import pixi = require('pixi.js')
const oldRenderer = PIXI.WebGLRenderer;

(pixi as any).WebGLRenderer = function(width, height, options) {
  options.transparent = true;
  oldRenderer.call(this, width, height, options);
};
(pixi as any).WebGLRenderer.prototype = oldRenderer.prototype;;*/

import { BeSkill } from '@be/be-framework';
import jibo = require('jibo');

import { kbModel } from './models/KBModel';
import { userModel } from './models/UserModel';
import { recipeModel } from './models/RecipeModel';
import { stateModel } from './models/StateModel';

import FavoriteBtn from './components/buttons/FavoriteBtn';
import RecipeMenuItem from './components/buttons/RecipeMenuItem';
import StepMenuItem from './components/buttons/StepMenuItem';

import FollowUpView from './views/FollowUpView';
import MealView from './views/MealView';
import TakePhotoView from './views/TakePhotoView';
import RecipeSelectionView from './views/RecipeSelectionView';
import StepView from './views/StepView';
import VideoView from './views/VideoView';
import IngredientsView from './views/IngredientsView';
import OptionsView from './views/OptionsView';

import StringUtils from './utils/StringUtils';
import NavigationUtil from './utils/NavigationUtil';
import Helper from './utils/Helper';

const flows: any = {
  main: require('./flows/Main'),
  test: require('./flows/Test')
};

class RecipeSkill extends BeSkill {

  public stepView: StepView = null;
  public videoView: VideoView = null;
  public followUpView: FollowUpView = null;
  public takePhotoView: TakePhotoView = null;
  public ingredientsView: IngredientsView = null;
  public recipeSelectionView: RecipeSelectionView = null;
  public mealView: MealView = null;
  public optionsView: OptionsView = null;

  private overlay: HTMLDivElement;
  private flow: any;
  private blackboard: any;
  private guiCallback: (interruptPath: string) => void;
  // Touch-first safety net: a tap can fire guiCallback before the flow reaches
  // its "Apply Listeners" node (which calls setGUICallback with the real `done`).
  // Buffer the latest interrupt path so the tap is never silently dropped; it is
  // flushed as soon as the flow arms its listener.
  private pendingGuiPath: string = null;

  constructor (assetPack?: string) {
    super(assetPack);
    this.guiCallback = (path) => {
      console.log('[recipe] guiCallback buffered (listener not armed):', path);
      this.pendingGuiPath = path;
    };
    this.registerClasses();
    // The standalone skill declared #overlay in its own index.html. Inside Be the
    // skill shares Be's host document, so create the overlay element if missing.
    this.overlay = document.querySelector('#overlay') as HTMLDivElement;
    if (!this.overlay) {
      this.overlay = document.createElement('div');
      this.overlay.id = 'overlay';
      this.overlay.style.position = 'absolute';
      this.overlay.style.display = 'none';
      const overlayImg = document.createElement('img');
      overlayImg.id = 'overlayImg';
      overlayImg.src = 'resources/images/EdgeFade.png';
      this.overlay.appendChild(overlayImg);
      if (document.body) {
        document.body.appendChild(this.overlay);
      }
    }
  }

  public postInit (done: () => any): void {
    // Never leave `done` uncalled: Be waits on this task at startup, and jibo 14
    // API drift (e.g. removed kb.loop name lookups) can reject the chain. Log and
    // continue so Be finishes initializing and the skill can still open.
    kbModel.init()
      .then(() => userModel.init())
      .then(() => stateModel.init())
      .then(() => done())
      .catch((err) => {
        console.error('[recipe] postInit error (continuing):', err);
        done();
      });
  }

  public preload (done: (err?: any) => void): void {
    // jibo 14 removed embodied.speech.installDelegate (it pre-registered the asset
    // pack's speech data in jibo 5; jibo 14 does this automatically). Calling it
    // unguarded threw synchronously and made Be abort the skill switch before open().
    const es: any = (jibo as any).embodied && (jibo as any).embodied.speech;
    if (es && typeof es.installDelegate === 'function') {
      es.installDelegate(this.assetPack);
    }
    done();
  }

  public open (result?: any): void {
    // jibo 14 removed kb.loop.loadLoopAll; guard so open() does not throw when
    // the API is unavailable (this call only pre-warms name-recognition loops).
    if (jibo.kb && jibo.kb.loop && typeof jibo.kb.loop.loadLoopAll === 'function') {
      jibo.kb.loop.loadLoopAll((err, loop) => {
        console.log('loop:', err, loop);
      });
    }
    stateModel.initUser(userModel.id)
      .then(() => stateModel.loadState())
      .then(() => {
        let activeRecipe = stateModel.get('active-recipe');
        if (activeRecipe !== null) {
          activeRecipe = recipeModel.getRecipeById(activeRecipe);
        }
        this.blackboard = {
          skill: this,
          userModel,
          recipeModel,
          stateModel,
          Helper,
          StringUtils,
          languageData: require('../data/en-us/Strings.json'),
          navigation: new NavigationUtil(),
          activeRecipe,
          activeStep: stateModel.get('active-step'),
          activeIndex: 0,
          filteredRecipes: [],
          guiPayload: null
        };
        const options = { assetPack: this.assetPack, blackboard: this.blackboard, enableLogging: false };
        this.flow = jibo.flow.run(flows.main, options, () => {
          this.exit();
        });
      });
  }

  public close (done: () => void): void {
    this.flow.destroy();
    this.flow = null;
    this.blackboard = null;
    done();
  }

  public setGUICallback (callback: (interruptPath?: string) => any): void {
    const wrapped = (path: string) => {
      console.log('[recipe] guiCallback -> flow:', path);
      // After delivering one tap, revert to buffering so a stray second tap
      // (before the flow re-arms via its ~ResetGUI cycle) is not sent to an
      // already-consumed continuation.
      this.guiCallback = (p) => {
        console.log('[recipe] guiCallback buffered (post-delivery):', p);
        this.pendingGuiPath = p;
      };
      callback(path);
    };
    this.guiCallback = wrapped;
    // Flush a tap that arrived before this listener was armed.
    if (this.pendingGuiPath !== null) {
      const path = this.pendingGuiPath;
      this.pendingGuiPath = null;
      wrapped(path);
    }
  }

  public clearFilteredRecipes (): void {
    this.blackboard.filteredRecipes = [];
  }

  public updateState (): void {
    let b = this.blackboard;
    b.activeStep = stateModel.get('active-step');
    b.activeRecipe = stateModel.get('active-recipe');
  }

  public cleanup (done: () => any): void {
    stateModel.refreshState().then(() => {
      this.updateState();
      jibo.face.views.changeView({ removeAll: true }, done);
    });
  }

  public showRecipeSelection (done: () => any): void {
    this.showOverlay();
    if (this.recipeSelectionView === null) {
      this.recipeSelectionView = new RecipeSelectionView();
      this.recipeSelectionView.init(this.blackboard.filteredRecipes);
    } else {
      this.recipeSelectionView.init(this.blackboard.filteredRecipes);
      this.recipeSelectionView.on('ready', () => {
        this.setSelectionNav(this.blackboard.activeIndex);
        done();
      });
      return;
    }

    this.recipeSelectionView.on('ready', () => {
      this.setSelectionNav(this.blackboard.activeIndex);
      done();
    });
    this.recipeSelectionView.on('nav', direction => {
      this.blackboard.activeRecipe = this.blackboard.filteredRecipes[
        this.blackboard.activeIndex = this.recipeSelectionView.index
      ];
      this.blackboard.guiPayload = { direction };
      this.guiCallback('nav');
    });
    this.recipeSelectionView.on('favorite', () => this.guiCallback('favorite'));
    this.recipeSelectionView.on('make', () => this.guiCallback('make'));
    this.recipeSelectionView.on('ingredients', () => this.guiCallback('ingredients'));
    this.recipeSelectionView.on('video', () => this.guiCallback('video'));
  }

  public showPreStep (done: () => any): void {
    this.hideOverlay();
    if (this.mealView === null) {
      this.mealView = new MealView();
      this.mealView.init();
    } else {
      this.mealView.init();
      this.mealView.on('ready', done);
      return;
    }

    this.mealView.on('ready', done);
  }

  public showSteps (done: () => any): void {
    this.hideOverlay();
    if (this.stepView === null) {
      this.stepView = new StepView();
      this.stepView.init(this.blackboard.activeRecipe);
    } else {
      this.stepView.init(this.blackboard.activeRecipe);
      this.stepView.on('ready', () => {
        this.stepView.stepTo(this.blackboard.activeStep || 1);
        done();
      });
      return;
    }

    this.stepView.on('ready', () => {
      this.stepView.stepTo(this.blackboard.activeStep || 1);
      done();
    });
    this.stepView.on('video', () => this.guiCallback('video'));
    this.stepView.on('nav', direction => {
      this.blackboard.guiPayload = { direction };
      this.guiCallback('nav');
    });
  }

  public showIngredients (done: () => any): void {
    this.hideOverlay();
    if (this.ingredientsView === null) {
      this.ingredientsView = new IngredientsView();
      this.ingredientsView.init(this.blackboard.activeRecipe);
    } else {
      this.ingredientsView.init(this.blackboard.activeRecipe);
      this.ingredientsView.on('ready', done);
      return;
    }
    this.ingredientsView.on('ready', done);
    this.ingredientsView.on('close', () => this.guiCallback('close'));
    this.ingredientsView.on('make', () => this.guiCallback('make'));
  }

  public showVideo (done: () => any): void {
    this.hideOverlay();
    let truncatedTimeline = this.blackboard.navigation.getPreviousView() === 'guided-steps';
    // The recipe data stores a relative video path (e.g. "./video/foo.mp4"). A raw
    // DOM <video src> resolves against Be's host document, not this asset pack, so
    // resolve it to the asset pack the same way jibo resolves view assets.
    const videoSrc = this.resolveAssetUri(this.blackboard.activeRecipe.video.url);
    if (this.videoView === null) {
      this.videoView = new VideoView();
      this.videoView.init(this.blackboard.activeRecipe, truncatedTimeline, this.blackboard.activeStep, videoSrc);
    } else {
      this.videoView.init(this.blackboard.activeRecipe, truncatedTimeline, this.blackboard.activeStep, videoSrc);
      this.videoView.on('ready', done);
      return;
    }

    this.videoView.on('ready', done);
    this.videoView.on('close', () => this.guiCallback('close'));
    this.videoView.on('volume', () => this.guiCallback('volume'));
    this.videoView.on('play-pause', () => this.guiCallback('play-pause'));
  }

  public showTakePhoto (done: () => any): void {
    this.hideOverlay();
    if (this.takePhotoView === null) {
      this.takePhotoView = new TakePhotoView();
      this.takePhotoView.init(this.blackboard.activeRecipe);
    } else {
      this.takePhotoView.init(this.blackboard.activeRecipe);
      this.takePhotoView.on('ready', done);
      return;
    }

    this.takePhotoView.on('ready', done);
    this.takePhotoView.on('take', () => this.guiCallback('take'));
    this.takePhotoView.on('retake', () => this.guiCallback('retake'));
    this.takePhotoView.on('close', () => this.guiCallback('close'));
  }

  public showFollowUp (done: () => any): void {
    this.hideOverlay();
    if (this.followUpView === null) {
      this.followUpView = new FollowUpView(this.blackboard.activeRecipe);
      this.followUpView.init(this.blackboard.activeRecipe);
    } else {
      this.followUpView.init(this.blackboard.activeRecipe);
      this.followUpView.on('ready', done);
      return;
    }

    this.followUpView.on('ready', done);
    this.followUpView.on('favorite', () => this.guiCallback('favorite'));
    this.followUpView.on('close', () => this.guiCallback('close'));
  }

  public showOptions (done: () => any): void {
    this.hideOverlay();
    if (this.optionsView === null) {
      this.optionsView = new OptionsView();
      this.optionsView.init();
    } else {
      this.optionsView.init();
      this.optionsView.on('ready', done);
      return;
    }

    this.optionsView.on('ready', done);
    this.optionsView.on('close', () => this.guiCallback('close'));
    this.optionsView.on('help', () => this.guiCallback('help'));
    this.optionsView.on('choice', choice => {
      this.blackboard.guiPayload = { choice };
      this.guiCallback('choice');
    });
  }

  // Resolve a relative asset-pack path (e.g. "./video/foo.mp4") to a URL usable by
  // raw DOM elements. Mirrors jibo's own view-asset resolution (PathUtils.getAssetUri
  // with this skill's assetPack) and falls back to the original path on any failure.
  public resolveAssetUri (uri: string): string {
    if (!uri || /^(https?:|file:|data:|\/)/.test(uri)) { return uri; }
    try {
      const PathUtils: any = (jibo as any).utils && (jibo as any).utils.PathUtils;
      if (PathUtils && typeof PathUtils.getAssetUri === 'function') {
        const resolved = PathUtils.getAssetUri(uri.replace(/^\.\//, ''), this.assetPack);
        if (resolved) {
          console.log('[recipe] resolved asset', uri, '->', resolved);
          return resolved;
        }
      }
    } catch (err) {
      console.warn('[recipe] resolveAssetUri failed for', uri, err);
    }
    return uri;
  }

  protected showOverlay (): void {
    if (this.overlay) {
      this.overlay.style.display = 'block';
    }
  }

  protected hideOverlay (): void {
    setTimeout(() => {
      if (this.overlay) {
        this.overlay.style.display = 'none';
      }
    }, 1000);
  }

  protected setSelectionNav (activeIndex?: number): void {
    if (activeIndex !== 0) {
      this.recipeSelectionView.moveToIndex(activeIndex);
    } else {
      this.blackboard.activeRecipe = this.blackboard.filteredRecipes[
        this.blackboard.activeIndex = 0
      ];
    }
  }

  protected registerClasses (): void {
    [ RecipeSelectionView, StepView, FavoriteBtn, RecipeMenuItem, StepMenuItem ]
      .forEach(Class => jibo.face.views.creator.registerClass(Class, Class.TYPE));
  }

}

module.exports = RecipeSkill;
