// jibo-flow-core/lib/jibo-flow-core.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboFlowCore = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Activity = (function () {
    function Activity(name, id, className, assetPack, options) {
        this.name = name;
        this.id = id;
        this.class = className;
        this.assetPack = assetPack;
        this.options = this.rejoinOptions(options);
    }
    Activity.prototype.rejoinOptions = function (options) {
        for (var prop in options) {
            var property = options[prop];
            if (Array.isArray(property)) {
                options[prop] = property.join('\n');
            }
        }
        return options;
    };
    Activity.prototype.set_transitions = function (transitions) {
        this.transitions = transitions;
    };
    Activity.prototype.set_exceptions = function (exceptions) {
        this.exceptions = exceptions;
    };
    Activity.prototype.set_procedure = function (procedure) {
        this.procedure = procedure;
    };
    return Activity;
}());
exports.default = Activity;

},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FlowStep_1 = require("./FlowStep");
var FlowHarness = (function () {
    function FlowHarness(options) {
        this.name = '';
        this._steps = [];
        if ('ASRService' in options) {
            this.SSM = options;
        }
        else {
            this.SSM = options.SSM;
            this.name = options.name || this.name;
            this.enableLogging = options.enableLogging;
        }
    }
    FlowHarness.prototype.when = function () {
        var step = new FlowStep_1.FlowStep();
        this._steps.push(step);
        return step;
    };
    FlowHarness.prototype.getSteps = function () {
        var stepInstructions = [];
        for (var _i = 0, _a = this._steps; _i < _a.length; _i++) {
            var step = _a[_i];
            stepInstructions.push(step.getStep());
        }
        return stepInstructions;
    };
    FlowHarness.prototype.getStep = function (n) {
        return this._steps[n].getStep();
    };
    FlowHarness.prototype.getLength = function () {
        return this._steps.length;
    };
    FlowHarness.prototype.checkSteps = function () {
        for (var _i = 0, _a = this._steps; _i < _a.length; _i++) {
            var step = _a[_i];
            step.checkStep();
        }
    };
    FlowHarness.prototype.attachTrace = function (activityTrace) {
        this._activityTrace = activityTrace;
    };
    FlowHarness.prototype.didExecuteAllSteps = function () {
        for (var _i = 0, _a = this._steps; _i < _a.length; _i++) {
            var step = _a[_i];
            if (!step.getStep().executed) {
                return false;
            }
        }
        return true;
    };
    FlowHarness.prototype.didExecuteActivity = function (activityName, conditions) {
        if (conditions === void 0) { conditions = {}; }
        var visits = {};
        var result = {};
        for (var _i = 0, _a = this._activityTrace; _i < _a.length; _i++) {
            var activity = _a[_i];
            if (!conditions.inFlowUri || activity.uri.toLowerCase().includes(conditions.inFlowUri.toLowerCase())) {
                visits[activity.name] = (visits[activity.name] || 0) + 1;
                result[activity.name] = activity.result;
            }
        }
        return (activityName in visits
            && (conditions.visits === undefined || visits[activityName] === conditions.visits)
            && (conditions.result === undefined || result[activityName] === conditions.result));
    };
    return FlowHarness;
}());
exports.FlowHarness = FlowHarness;

},{"./FlowStep":5}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FlowRoot = (function () {
    function FlowRoot(loader) {
        this.procedures = {};
        this.global_inits = {};
        this.loader = loader;
    }
    FlowRoot.prototype.setStartInfoFromUri = function (uri) {
        this.startUri = this.loader.getUri(uri);
        this.startProcedureOrName = this.loader.getStartProcedureOrName(uri);
    };
    FlowRoot.prototype.loadProcedure = function (currentUri, procedureObjectOrName) {
        return this.loader.loadProcedureRelative(currentUri, procedureObjectOrName);
    };
    FlowRoot.prototype.getStartActivity = function () {
        return this.getStartProcedure().getBeginActivity();
    };
    FlowRoot.prototype.getStartProcedure = function () {
        return this.loader.loadProcedure(this.startUri);
    };
    FlowRoot.prototype.getStartProcedureOrName = function () {
        return this.startProcedureOrName;
    };
    return FlowRoot;
}());
exports.default = FlowRoot;

},{}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FlowRoot_1 = require("./FlowRoot");
var TreeLoader_1 = require("./loaders/TreeLoader");
var GojsLoader_1 = require("./loaders/GojsLoader");
var FlowRootFactory = (function () {
    function FlowRootFactory() {
    }
    FlowRootFactory.create = function (uri, loader) {
        if (loader === undefined) {
            if (typeof (uri) === 'function') {
                loader = TreeLoader_1.default;
            }
            else if (uri.endsWith(".flow")) {
                loader = new GojsLoader_1.default();
            }
            else {
                console.error('FlowRootFactory encountered an unknown file type: ' + uri);
                throw new Error('FlowRootFactory encountered an unknown file type: ' + uri);
            }
        }
        var flowRoot = new FlowRoot_1.default(loader);
        flowRoot.setStartInfoFromUri(uri);
        return flowRoot;
    };
    return FlowRootFactory;
}());
exports.default = FlowRootFactory;

},{"./FlowRoot":3,"./loaders/GojsLoader":9,"./loaders/TreeLoader":10}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FlowStepInstruction_1 = require("./FlowStepInstruction");
var FlowStep = (function () {
    function FlowStep() {
        this._step = new FlowStepInstruction_1.FlowStepInstruction();
    }
    FlowStep.prototype.say = function (text) {
        this._ensureSettable('say', text);
        this._step.say = text;
        return this;
    };
    FlowStep.prototype.awaitingSpeechInput = function () {
        this._step.awaitingSpeechInput = true;
        return this;
    };
    FlowStep.prototype.stub = function (stubFunction) {
        this._ensureSettable('at', 'start');
        this._ensureSettable('stub');
        this._step.at = 'start';
        this._step.stub = stubFunction;
        return this;
    };
    FlowStep.prototype.do = function (func) {
        this._ensureSettable('do');
        this._step.do = func;
        return this;
    };
    FlowStep.prototype.delay = function (seconds) {
        this._step.delay = seconds;
        return this;
    };
    FlowStep.prototype.forceTransition = function (transition) {
        this._ensureSettable('at', 'end');
        this._ensureSettable('stub');
        this._step.at = 'end';
        this._step.stub = function () { return transition; };
        return this;
    };
    FlowStep.prototype.activeTransition = function (transition) {
        this._ensureSettable('at', 'update');
        this._ensureSettable('stub');
        this._step.at = 'update';
        this._step.stub = function () { return transition; };
        return this;
    };
    FlowStep.prototype.assert = function (assertionTest, message) {
        this._ensureSettable('at', 'end');
        this._ensureSettable('it');
        this._ensureSettable('assert');
        this._step.at = 'end';
        this._step.it = message;
        this._step.assert = assertionTest;
        return this;
    };
    FlowStep.prototype.activityName = function (activityName) {
        this._ensureSettable('activity');
        this._step.activity = activityName;
        return this;
    };
    FlowStep.prototype.before = function () {
        this._ensureSettable('at', 'start');
        this._step.at = 'start';
        return this;
    };
    FlowStep.prototype.at = function () {
        this._ensureSettable('at', 'update');
        this._step.at = 'update';
        return this;
    };
    FlowStep.prototype.after = function () {
        this._ensureSettable('at', 'end');
        this._step.at = 'end';
        return this;
    };
    FlowStep.prototype.inFlow = function (flowName) {
        this._ensureSettable('flow');
        this._step.flow = flowName;
        return this;
    };
    FlowStep.prototype.activityType = function (activityType) {
        this._ensureSettable('class');
        this._step.class = activityType;
        return this;
    };
    FlowStep.prototype.activityId = function (activityId) {
        this._ensureSettable('id');
        this._step.id = activityId;
        return this;
    };
    FlowStep.prototype.visits = function (count) {
        this._ensureSettable('visits', count);
        this._step.visits = count;
        return this;
    };
    FlowStep.prototype.checkStep = function () {
        if (this._step.delay && this._step.at && this._step.at !== 'update') {
            throw new Error('Delay operation only permitted during "update" cycle.');
        }
    };
    FlowStep.prototype._ensureSettable = function (prop, value) {
        if (this._step[prop] !== undefined && value === undefined) {
            throw new Error("FlowStep: Cannot set step property " + prop + " a second time");
        }
        if (this._step[prop] !== undefined && this._step[prop] !== value) {
            throw new Error("FlowStep: Cannot set step property " + prop + " to " + value + " after setting it to " + this._step[prop]);
        }
    };
    FlowStep.prototype.getStep = function () {
        return this._step;
    };
    return FlowStep;
}());
exports.FlowStep = FlowStep;

},{"./FlowStepInstruction":6}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FlowStepInstruction = (function () {
    function FlowStepInstruction() {
    }
    return FlowStepInstruction;
}());
exports.FlowStepInstruction = FlowStepInstruction;

},{}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Activity_1 = require("./Activity");
var path = require("path");
var findRoot = require("find-root");
var Procedure = (function () {
    function Procedure(name, uri, local_inits, param_inits) {
        this.name = name;
        this.uri = uri;
        this.activities = {};
        this.local_inits = local_inits;
        this.param_inits = param_inits;
        this.transitions = {};
        this.exceptions = {};
        this.connections = [];
    }
    Procedure.prototype.loadPropertiesFromTreeFunction = function (tree) {
        var treeObj = tree();
        for (var activityId in treeObj) {
            var t = treeObj[activityId];
            if (typeof t === 'function') {
                t = t();
            }
            var activity = new Activity_1.default(t.name, t.id, t.class, t['asset-pack'], t.options);
            activity.transitions = t.transitions;
            activity.exceptions = t.exceptions;
            this.activities[t.id] = activity;
        }
        this.name = treeObj.meta.name;
        this.uri = treeObj.meta.uri;
        this.relativeUri = path.relative(findRoot('.'), this.uri);
        this.local_inits = {};
        this.param_inits = {};
        this.procedureConstructor = tree;
    };
    Procedure.prototype.add_transition = function (transition) {
        var map = transition.value.startsWith('~') ? this.exceptions : this.transitions;
        if (!(transition.frm in map)) {
            map[transition.frm] = [];
        }
        map[transition.frm].push(transition);
        this.connections.push(transition);
    };
    Procedure.prototype.add_activity = function (activity) {
        var transitions = activity.id in this.transitions ? this.transitions[activity.id] : [];
        var exceptions = activity.id in this.exceptions ? this.exceptions[activity.id] : [];
        activity.set_transitions(transitions);
        activity.set_exceptions(exceptions);
        activity.set_procedure(this);
        this.activities[activity.id] = activity;
    };
    Procedure.prototype.getBeginActivity = function () {
        for (var key in this.activities) {
            var activity = this.activities[key];
            if (activity.class === 'Flow.Begin') {
                return activity;
            }
        }
        throw new Error("No Flow.Begin Activity in procedure " + this.name);
    };
    Procedure.prototype.getAsyncBeginActivities = function () {
        var asyncs = [];
        for (var key in this.activities) {
            var activity = this.activities[key];
            if (activity.class === 'Flow.Begin-Parallel') {
                asyncs.push(activity);
            }
        }
        return asyncs;
    };
    return Procedure;
}());
exports.default = Procedure;

},{"./Activity":1,"find-root":undefined,"path":undefined}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Transition = (function () {
    function Transition(frm, to, value) {
        this.frm = frm;
        this.to = to;
        this.value = value;
    }
    return Transition;
}());
exports.default = Transition;

},{}],9:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Activity_1 = require("../Activity");
var Procedure_1 = require("../Procedure");
var Transition_1 = require("../Transition");
var fs = require('fs');
var path = require('path');
var GojsLoader = (function () {
    function GojsLoader() {
        this.procedures = {};
    }
    GojsLoader.prototype.getStartProcedureOrName = function (uri) {
        return path.parse(uri).name;
    };
    GojsLoader.prototype.getUri = function (uri) {
        return uri;
    };
    GojsLoader.prototype.getBase = function (uri) {
        return path.parse(uri).dir;
    };
    GojsLoader.prototype.loadProcedureRelative = function (currentUri, name) {
        var baseUri = this.getBase(currentUri);
        var uri = path.join(baseUri, name + '.flow');
        return this.loadProcedure(uri);
    };
    GojsLoader.prototype.loadProcedure = function (uri) {
        var name = this.getStartProcedureOrName(uri);
        if (!this.procedures.hasOwnProperty(name)) {
            this.procedures[name] = this.readProcedure(name, uri);
        }
        return this.procedures[name];
    };
    GojsLoader.prototype.readProcedure = function (name, uri) {
        var data = fs.readFileSync(uri, 'utf8');
        var base = JSON.parse(data);
        var procedure = new Procedure_1.default(name, uri, {}, {});
        for (var i = 0; i < base.linkDataArray.length; i++) {
            var goLink = base.linkDataArray[i];
            var transitionValue = goLink.text === undefined ? "" : goLink.text;
            var transition = new Transition_1.default(goLink.from, goLink.to, transitionValue);
            procedure.add_transition(transition);
        }
        for (var i = 0; i < base.nodeDataArray.length; i++) {
            var goNode = base.nodeDataArray[i];
            var activity = new Activity_1.default(goNode.name, goNode.id, goNode.class, goNode['asset-pack'], goNode.options);
            procedure.add_activity(activity);
        }
        return procedure;
    };
    return GojsLoader;
}());
exports.default = GojsLoader;

},{"../Activity":1,"../Procedure":7,"../Transition":8,"fs":undefined,"path":undefined}],10:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Procedure_1 = require("../Procedure");
var TreeLoader = (function () {
    function TreeLoader() {
    }
    TreeLoader.getStartProcedureOrName = function (data) {
        return data;
    };
    TreeLoader.getUri = function (data) {
        return data().meta.uri;
    };
    TreeLoader.loadProcedureRelative = function (uri, procedureObject) {
        var procedure = new Procedure_1.default();
        procedure.loadPropertiesFromTreeFunction(procedureObject);
        return procedure;
    };
    return TreeLoader;
}());
exports.default = TreeLoader;

},{"../Procedure":7}],11:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GojsLoader_1 = require("./GojsLoader");
var TreeLoader_1 = require("./TreeLoader");
exports.default = {
    GojsLoader: GojsLoader_1.default,
    TreeLoader: TreeLoader_1.default
};

},{"./GojsLoader":9,"./TreeLoader":10}],12:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Activity_1 = require("./Activity");
exports.Activity = Activity_1.default;
var FlowRoot_1 = require("./FlowRoot");
exports.FlowRoot = FlowRoot_1.default;
var FlowRootFactory_1 = require("./FlowRootFactory");
exports.FlowRootFactory = FlowRootFactory_1.default;
var Procedure_1 = require("./Procedure");
exports.Procedure = Procedure_1.default;
var Transition_1 = require("./Transition");
exports.Transition = Transition_1.default;
var FlowHarness_1 = require("./FlowHarness");
exports.FlowHarness = FlowHarness_1.FlowHarness;
var FlowStep_1 = require("./FlowStep");
exports.FlowStep = FlowStep_1.FlowStep;
var FlowStepInstruction_1 = require("./FlowStepInstruction");
exports.FlowStepInstruction = FlowStepInstruction_1.FlowStepInstruction;
var loaders_1 = require("./loaders");
exports.loaders = loaders_1.default;

},{"./Activity":1,"./FlowHarness":2,"./FlowRoot":3,"./FlowRootFactory":4,"./FlowStep":5,"./FlowStepInstruction":6,"./Procedure":7,"./Transition":8,"./loaders":11}]},{},[12])(12)
});
