// jibo-emotion-system/src/api.ts
import {EmotionSystem, EmotionEvents} from "./EmotionSystem";
import {Emotion} from "./emotion/Emotion";
import {InitOptions, EmotionPoint, EmotionImpact} from "./common/Types";
import {ImpactValues} from "./emotion/EmotionDefinitions";

export * from './common/Types';

export { Emotion };

/*
 * Private variables
 */

/**
 * Various events emitted by the emotion system.
 * @name jibo.emotion#events
 * @type {jibo.emotion.EmotionEvents}
 */
export let events: EmotionEvents;

/**
 * The instance of the runtime. We expose it on the API just for
 * testing and special circumstances. Not meant to be accessed for standard use.
 * @type {jibo.emotion.EmotionSystem}
 * @internal
 * @private
 */
export let _runtime: EmotionSystem;

/*
 * Public API methods
 */

/**
 * Initializes the emotion system.
 * @method jibo.emotion#init
 * @param {jibo.emotion.InitOptions} initOptions options to init with jibo.
 * @returns {Promise<void>} A promise that is resolved when the emotion system has been initializd
 */
export function init(initOptions:InitOptions): Promise<void> {
    _runtime = new EmotionSystem();
    events = _runtime.events;
    return _runtime.init(initOptions);
}

/**
 * Get the nearest labeled emotion to Jibo's current emotional state.
 * (Mapped to a set of pre-defined states).
 * @method jibo.emotion#getNearestEmotion
 * @returns {jibo.emotion.Emotion}
 */
export function getNearestEmotion(): Emotion {
    return _runtime.getNearestEmotion();
}

/**
 * Get Jibo's exact current emotional state.
 * @method jibo.emotion#getCurrentEmotionalValues
 * @returns {jibo.emotion.EmotionPoint}
 */
export function getCurrentEmotionalValues(): EmotionPoint {
    return _runtime.getCurrentEmotionalValues();
}

/**
 * Supports skills impacting jibo's emotional state using a set of offered impacts
 * @param {jibo.emotion.EmotionImpact} impact The impact to effect Jibo's emotional state
 * @method jibo.emotion#triggerImpact
 */
export function triggerImpact(impact: EmotionImpact): void {
    let valenceImpact = ImpactValues.get(impact.valence) || 0;
    let confidenceImpact = ImpactValues.get(impact.confidence) || 0;
    const emoImpact = new EmotionPoint(valenceImpact, confidenceImpact);
    _runtime.triggerImpact(emoImpact);
}
