// jibo-emotion-system/src/EmotionSystem.ts
/*
 * The entrypoint for the emotion system library
 */

import {SensorySystem} from './sensing/SensorySystem';
import {Sensor} from './sensing/sensors/Sensor';
import {AppraisalSystem} from './appraisal/AppraisalSystem';
import {AppraisalRule} from './appraisal/rules/AppraisalRule';
import {ExpressionSystem} from './expression/ExpressionSystem';
import {EmotionSpace} from './emotion/EmotionSpace';
import {EmotionTrace} from './emotion/EmotionTrace';
import {Event, EventContainer} from 'jibo-typed-events';
import {MimRule, TouchRule, ExternalImpactRule, MotorFaultRule} from './appraisal/rules';
import {EmotionalExpression} from "./expression/EmotionalExpression";
import {Emotion} from "./emotion/Emotion";
import {TouchSensor} from "./sensing/sensors/TouchSensor";
import {MimSensor} from "./sensing/sensors/MimSensor";
import {ExternalImpactSensor} from "./sensing/sensors/ExternalImpactSensor";
import {MotorFaultSensor} from "./sensing/sensors/MotorFaultSensor";
// Only importing type definition
import JiboImport = require('../../jibo');
import {EmotionName, RuleNames, InitOptions, EmotionPoint, TouchOptions} from "./common/Types";
import {DEFAULT_EMOTIONS} from "./emotion/EmotionDefinitions";
import {IdentitySensor} from "./sensing/sensors/IdentitySensor";
import {IdentityRule} from "./appraisal/rules/IdentityRule";

import rootLog from "./log";
const log: typeof rootLog = rootLog.createChild("EmotionSystem");

/**
 * Emotional Events
 * @class EmotionEvents
 * @memberof jibo.emotion
 */
export class EmotionEvents extends EventContainer {
    /**
     * Jibo's emotion has changed. Gives the last available emotion.
     * @type {Event<jibo.emotion.Emotion>}
     * @name jibo.emotion.EmotionEvents#emotionChanged
     */
    emotionChanged = new Event<Emotion>('Emotion has changed');
}

/**
 * @class EmotionSystem
 * @description The entire Emotion System and management of Jibo's internal emotional state,
 * including the Appraisal System, Expression System, Sensory System, and Emotional Space.
 * @memberof jibo.emotion
 * @private
 * @internal
 */
export class EmotionSystem {

    /**
     * This is a record of all emotional events Jibo is aware of, containing Sensor data, appraisal rules
     * the previous and new emotional states, and what expressions if any occurred as a result
     * @type {Array<jibo.emotion.emotionTrace>}
     * @name jibo.emotion.EmotionSystem#history
     * @private
     */
    public history: EmotionTrace[] = [];
    /**
     * The emotional events of the EmotionSystem
     * @type {jibo.emotion.EmotionEvents}
     * @name jibo.emotion.EmotionSystem#events
     * @private
     */
    public events = new EmotionEvents();
    /**
     * Emotion Wrappers
     * @type {jibo.emotion.Emotion[]}
     * @name jibo.emotion.EmotionSystem#emotions
     * @private
     */
    public emotions: Emotion[];
    /**
     * Provides default rules given a sensory system
     * @type {jibo.emotion.appraisal.AppraisalRule[]}
     * @name jibo.emotion.EmotionSystem#rules
     * @private
     */
    public rules: AppraisalRule[];
    /**
     * The various sensors used to input emotional data and affect Jibo's
     * emotional landscape.
     * @const {Array.<Sensor>}
     * @memberOf jibo.emotion
     * @private
     */
    public sensors: Sensor<any>[];
    /**
     * This should not be a part of the external API
     * @internal
     * @int-docs
     * @private
     */
    public jibo: typeof JiboImport;
    /**
     * The current emotional P.A.D. space that represents what Jibo is feeling
     * in relation to his other possible emotional states.
     * @type {jibo.emotion.EmotionSpace}
     * @private
     * @memberOf jibo.emotion
     */
    private _emotionSpace: EmotionSpace;
    /**
     * The sensory system that aggregates all sensors and provides a single
     * point of interactions for events.
     * @type {jibo.emotion.SensorySystem}
     * @private
     * @memberOf jibo.emotion
     */
    private _sensorySystem: SensorySystem;
    /**
     * The perceptual appraisal system that interprets sensory input that
     * affects Jibo.
     * @type {jibo.emotion.AppraisalSystem}
     * @private
     * @memberOf jibo.emotion
     */
    private _appraisalSystem: AppraisalSystem;
    /**
     * The system that maps emotions and emotional impacts to actions via the Action System plugin
     * @type {jibo.emotion.ExpressionSystem}
     * @private
     * @memberOf jibo.emotion
     */
    private _expressionSystem: ExpressionSystem;
    /**
     * Map of expression names to Expression objects
     * @type {Map<jibo.emotion.EmotionName, jibo.emotion.EmotionalExpression>}
     * @private
     * @memberOf jibo.emotion
     */
    private _expressions: Map<EmotionName, EmotionalExpression>;

    constructor(private debug = true) {

        // create emotions this instance will be using
        this.emotions = [];
        DEFAULT_EMOTIONS.forEach((emotionValues, emotionName) => {
            let emotion = new Emotion(emotionName, emotionValues.valence, emotionValues.confidence);
            this.emotions.push(emotion);
        });

        // set up default emotions and default emotional states.
        this._emotionSpace = new EmotionSpace(this);
        this._expressions = new Map<EmotionName, EmotionalExpression>();
        this._expressionSystem = new ExpressionSystem(this);

        // set up default sensors
        // TODO: reinstate ASR Input Sensor once we have generic
        // sentiment analysis -- for now, comes through rules at an
        // ad-hoc basis.
        this.sensors = [
            new ExternalImpactSensor(this),
            new IdentitySensor(this),
            new TouchSensor(this),
            new MotorFaultSensor(this),
            new MimSensor(this)
        ];
        // create sensory system
        this._sensorySystem = new SensorySystem(this);
        // set up default appraisal rules
        // TODO: reinstate ASR Input rule once we have ASR sensor.
        this.rules = [
            new IdentityRule(RuleNames.IdentityRule, this),
            new TouchRule(RuleNames.TouchRule, this),
            new ExternalImpactRule(RuleNames.SkillInputRule, this),
            new MotorFaultRule(RuleNames.MotorFaultRule, this),
            new MimRule(RuleNames.MimRule, this)
        ];
        // set up default appraisal system and event listeners
        this._appraisalSystem = new AppraisalSystem(this);
    }

    /**
     * Initializes the EmotionSystem
     * @memberOf jibo.emotion
     */
    async init(options: InitOptions):Promise<any> {
        // initialize emotional expressions with our emotions
        this.jibo = options.jibo;
        DEFAULT_EMOTIONS.forEach((emotion, emotionName) => {
            this._expressions.set(emotionName, new EmotionalExpression(emotionName, this));
        });
        this._expressionSystem.init(this._expressions);
        this._emotionSpace.init(this.emotions);
        // initialize sensory system
        this._sensorySystem.init(this.sensors);
        this.rules.forEach((rule:AppraisalRule) => {
            rule.init(this._sensorySystem);
        });
        // initialize appraisal system
        this._appraisalSystem.init(this.rules);
        // initialize event listeners

        this._appraisalSystem.events.impactTriggered.on((emotionTrace:EmotionTrace) => {
            // Sensors will trigger an appraisal, which Jibo's appraisal system
            // will interpret as good/bad and various other things.
            //log.debug("impactTriggered", emotionTrace);
            this._emotionSpace.update(emotionTrace);
        });

        this._emotionSpace.events.updated.on((emotionTrace:EmotionTrace) => {
            //log.debug("updated", emotionTrace);
            this._expressionSystem.express(emotionTrace);
            // save in history
            // Disabling like this for beta3 because of a mem leak concern since this 
            // history is never cleared. A real fix for this will be to store a history
            // in a circular buffer with a fixed size. 
            // this.history.push(emotionTrace);
        });

        this._emotionSpace.events.changeEvent.on((emotion: Emotion) => {
            //log.debug("emotions changed", emotion);
            this.events.emotionChanged.emit(emotion);
        });
    }

    /**
     * Performs all cleanup work and unplugs all the things.
     * @memberOf jibo.emotion
     */
    public dispose() {
        this.events.removeAllListeners();
        this.rules.forEach((rule:AppraisalRule) => {
            rule.dispose();
        });
        this.rules = [];
        this.sensors = [];
        this._emotionSpace.dispose();
        this._appraisalSystem.dispose();
        this._sensorySystem.dispose();
        this._expressionSystem.dispose();
    }

    /**
     * Supports reading the current nearest Emotion
     * @param {jibo.emotion.EmotionPoint} [referencePoint] Point to get nearest emotion to. Defaults to current position.
     * @returns {Emotion}
     * @memberOf jibo.emotion
     */
    public getNearestEmotion(referencePoint?: EmotionPoint): Emotion {
        log.debug("getNearestEmotion", referencePoint);
        return this._emotionSpace.getNearestEmotion(referencePoint);
    }

    /**
     * Supports reading the exact current Emotional state of the robot
     * @returns {EmotionPoint}
     * @memberOf jibo.emotion
     */
    public getCurrentEmotionalValues(): EmotionPoint {
        log.debug("getCurrentEmotionalValues", this._emotionSpace.getCurrentPosition());
        return this._emotionSpace.getCurrentPosition();
    }

    /**
     * Supports skills impacting jibo's emotional state using a set of offered impacts
     * @param {jibo.emotion.EmotionPoint} impactPoint The impact to effect Jibo's emotional state
     * @memberOf jibo.emotion
     */
    public triggerImpact(impactPoint: EmotionPoint): void {
        log.debug("triggerImpact", impactPoint);
        const sensor = this._sensorySystem.getSensor(ExternalImpactSensor.SENSOR_NAME);
        sensor.supplyData(impactPoint);
    }

    /**
     * Trigger a touch sensor -- used internally for debugging.
     * @param {jibo.emotion.TouchOptions} options the options used to trigger a touch event.
     * @private
     * @memberOf jibo.emotion
     */
    public triggerTouch(options:TouchOptions) {
        log.debug("triggerTouch", options);
        const ts = <TouchSensor>this._sensorySystem.getSensor(TouchSensor.SENSOR_NAME);
        ts.jiboWasTouched(options);
    }
}
