// @be/recipe/src/models/RecipeModel.ts
import Recipe from '../entities/Recipe';
import Helper from '../utils/Helper';
import { stateModel } from './StateModel';

type SortType = 'ascending' | 'descending';
const ASC: SortType = 'ascending';
const DESC: SortType = 'descending';

const SOURCE = [
  require('../../data/source/beef/barbecuedbeef.json'),
  require('../../data/source/beef/beefbulgogi.json'),
  require('../../data/source/chicken/chickencordonblue.json'),
  require('../../data/source/chicken/chickenpotpieIX.json'),
  require('../../data/source/fish/fishchowder.json'),
  require('../../data/source/fish/fishtacos.json'),
  require('../../data/source/vegetarian/spinachenchiladas.json'),
  require('../../data/source/vegetarian/vegetariankalesoup.json'),
  require('../../data/source/dessert/dessertcrepes.json'),
  require('../../data/source/dessert/tiramisuII.json')
];
const CURATED = [
  require('../../data/curated/beef/barbecuedbeef.json'),
  require('../../data/curated/beef/beefbulgogi.json'),
  require('../../data/curated/chicken/chickencordonblue.json'),
  require('../../data/curated/chicken/chickenpotpieIX.json'),
  require('../../data/curated/fish/fishchowder.json'),
  require('../../data/curated/fish/fishtacos.json'),
  require('../../data/curated/vegetarian/spinachenchiladas.json'),
  require('../../data/curated/vegetarian/vegetariankalesoup.json'),
  require('../../data/curated/dessert/dessertcrepes.json'),
  require('../../data/curated/dessert/tiramisuII.json')
];

class RecipeModel {

  private data: Map<number, Recipe> = new Map<number, Recipe>();

  constructor () {
    this.setRecipes(SOURCE, CURATED);
  }

  get numRecipes (): number {
    return this.data.size;
  }

  get recipes (): Recipe[] {
    return Array.from(this.data.values());
  }

  public clear (): void {
    this.data.clear();
  }

  public setRecipes (source: any, curated?: any): void {
    if (Array.isArray(source)) {
      for (let i = 0, l = source.length; i < l; i++) {
        let data = source[i];
        let id = data.recipeID;
        let overwrite;
        if (curated) {
          // find the matching curated data
          for (let i = 0, l = curated.length; i < l; i++) {
            let data = curated[i];
            if (data.recipeID === id) {
              overwrite = data;
              break;
            }
          }
        }
        this.setData(data, overwrite);
      }
    } else {
      this.setData(source, curated);
    }
  }

  public sortRecipesByRating (order: SortType = ASC): Recipe[] {
    return this.sortBy('ratingAverage', order === ASC ? 1 : -1);
  }

  public sortRecipesByTitle (order: SortType = ASC): Recipe[] {
    return this.sortBy('title', order === ASC ? 1 : -1);
  }

  public getRecipesByRating (rating: number, andUp: boolean = true): Recipe[] {
    return this.recipes.filter(recipe => andUp ? recipe.ratingAverage >= rating : recipe.ratingAverage === rating);
  }

  public getRecipesByCategory (category: string, unique: boolean = true): Recipe[] {
    let arr = [];
    let terms = Helper.getCategoryTerms(category);
    terms.forEach(term => arr.push(...this.getRecipesByTitle(term), ...this.getRecipesByIngredient(term)));
    return unique ? this.unique(arr) : arr;
  }

  public getRecipesByIngredient (ingredient: string): Recipe[] {
    const regex = new RegExp(ingredient, 'i');
    return this.recipes.filter(recipe => {
      let ingredients = recipe.ingredients;
      for (let i = 0, l = ingredients.length; i < l; i++) {
        if (regex.test(ingredients[i].displayValue)) { return true; }
      }
      return false;
    });
  }

  public getRecipesByTitle (title: string): Recipe[] {
    let regex = new RegExp(title, 'i');
    return this.recipes.filter(recipe => regex.test(recipe.title));
  }

  public getRecipesByGenericFilter (token: string): Recipe[] {
    return this.unique(
      this.getRecipesByCategory(token, false).concat(
        this.getRecipesByTitle(token),
        this.getRecipesByIngredient(token)
      )
    );
  }

  public getRecipeTitles (): string[] {
    return this.recipes.map(recipe => recipe.title);
  }

  public getRecipeByTitle (title: string): Recipe {
    let regex = new RegExp(title, 'i');
    let recipes = this.recipes;
    for (let i = 0, l = recipes.length; i < l; i++) {
      let recipe = recipes[i];
      if (regex.test(recipe.title)) { return recipe; }
    }
    return null;
  }

  public getRecipeById (id: number): Recipe {
    return this.data.get(id);
  }

  public getRecipeByIndex (index: number): Recipe {
    return this.recipes[index];
  }

  public getRecommendations (): Recipe[] {
    return this.getRecipesByIngredient('chicken');
  }

  public getFavorites (): Recipe[] {
    return stateModel.get('favorites').map(this.getRecipeById.bind(this));
  }

  private setData (source: any, curated?: any): void {
    if (curated) {
      if (source.recipeID !== curated.recipeID) {
        console.warn("Curated data id does not match the source data id", source, curated);
      } else {
        // overwrite top level keys in the source object with curated data
        Object.assign(source, curated);
      }
    }
    let recipe = new Recipe(source);
    this.data.set(recipe.recipeID, recipe);
  }

  private sortBy (property: string, comparator: number): Recipe[] {
    return this.recipes.sort((a, b) => a[property] > b[property] ? -1 * comparator : a[property] < b[property] ? 1 * comparator : 0);
  }

  private unique (arr: Recipe[]): Recipe[] {
    return arr.filter((elem, pos) => {
      return arr.indexOf(elem) == pos;
    });
  }

}

export const recipeModel = new RecipeModel();
