// jibo-plugins/lib/jibo-plugins.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboPlugins = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
const path = require("path");
const Module = require("module");
const findRoot = require("find-root");
class PathUtils {
    static findRoot(start) {
        if (start) {
            let assetPack = PathUtils.getAssetPack(start);
            if (assetPack) {
                return PathUtils.resolveAssetPack(assetPack);
            }
        }
        return findRoot(start || process.cwd());
    }
    static getProjectName(dir) {
        return require(path.join(dir, 'package.json')).name;
    }
    static isURL(uri) {
        return uri.startsWith('https://') || uri.startsWith('http://') || uri.startsWith('file://');
    }
    static isImage(uri) {
        return uri && /^\.(jpg|png|gif|jpeg)$/i.test(path.extname(uri));
    }
    static setDefaultPath(parentDir, fileName) {
        let split = fileName.split(PathUtils.ASSET_TOKEN);
        if (split.length === 2) {
            return split[0] + PathUtils.ASSET_TOKEN + path.join(parentDir, split[1]);
        }
        else {
            return path.join(parentDir, fileName);
        }
    }
    static getAssetPack(fileName) {
        const split = fileName.split(PathUtils.ASSET_TOKEN);
        return (split.length === 1) ? "" : split[0];
    }
    static resolveAssetPack(assetPack) {
        if (assetPack && assetPack.endsWith(PathUtils.ASSET_TOKEN)) {
            assetPack = assetPack.substr(0, assetPack.length - PathUtils.ASSET_TOKEN.length);
        }
        if (assetPack === 'core' || assetPack === 'jibo') {
            assetPack = 'jibo';
            let packagePath = PathUtils.resolve(assetPack + "/package.json");
            if (!packagePath) {
                let uri = process.cwd();
                if (global.atom) {
                    uri = atom.workspace.getActivePaneItem().uri;
                }
                packagePath = findRoot(uri);
                return packagePath;
            }
            return path.dirname(packagePath);
        }
        const packagePath = PathUtils.resolve(assetPack + "/package.json");
        return path.dirname(packagePath);
    }
    static getAssetUri(fileName, assetPack = '', resourceRoot) {
        let split = fileName.split(PathUtils.ASSET_TOKEN);
        if (split.length === 2) {
            assetPack = split[0];
            fileName = split[1];
        }
        if (assetPack) {
            resourceRoot = PathUtils.resolveAssetPack(assetPack);
        }
        else if (path.isAbsolute(fileName)) {
            return fileName;
        }
        else if (!resourceRoot) {
            resourceRoot = PathUtils.findRoot();
        }
        return path.join(resourceRoot, fileName);
    }
    static getAudioUri(fileName, currentAssetPack = '', resourceRoot) {
        let audioPath = PathUtils.setDefaultPath('audio', fileName);
        return PathUtils.getAssetUri(audioPath, currentAssetPack, resourceRoot);
    }
    static getTimelineUri(fileName, currentAssetPack = '', resourceRoot) {
        let timelinePath = PathUtils.setDefaultPath('timelines', fileName);
        return PathUtils.getAssetUri(timelinePath, currentAssetPack, resourceRoot);
    }
    static resolve(moduleId, fromPath) {
        let fromDir = path.resolve(fromPath || process.cwd());
        let fromFile = path.join(fromDir, 'noop.js');
        try {
            return Module._resolveFilename(moduleId, {
                id: fromFile,
                filename: fromFile,
                paths: Module._nodeModulePaths(fromDir)
            });
        }
        catch (err) {
            return null;
        }
    }
}
PathUtils.ASSET_TOKEN = '://';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PathUtils;

},{"find-root":undefined,"module":undefined,"path":undefined}],2:[function(require,module,exports){
"use strict";
var PathUtils_1 = require("./PathUtils");
exports.PathUtils = PathUtils_1.default;

},{"./PathUtils":1}]},{},[2])(2)
});
