// @be/recipe/src/utils/Helper.ts
const CATEGORIES: { [id: string]: string[] } = {
  soup: [
    'chowder',
  ],
  meat: [
    'beef',
    'chicken',
    'fish'
  ],
  vegetarian: [
    'spinach',
    'kale'
  ],
  dessert: [
    'crepes',
    'tiramisu'
  ]
};

const STEP_MAP: { [id: string]: string | string[] } = {
    one: [ '1', 'won' ],
    two: [ '2', 'to', 'too' ],
    three: '3',
    four: [ '4', 'for' ],
    five: '5',
    six: '6',
    seven: '7',
    eight: [ '8', 'ate' ],
    nine: '9',
    ten: '10',
    eleven: '11',
    twelve: '12',
    thirteen: '13',
    fourteen: '14',
    fifteen: '15'
};

const MEAL_TIMES: { [id: number]: string } = {
  10: 'breakfast',
  14: 'lunch',
  24: 'dinner'
};

export default class Helper {

  constructor () {
    throw 'Helper is static and cannot be instantiated';
  }

  public static stepToString (step: string | number): string {
    if (typeof step === 'number') { step = step.toString(); }
    for (let key in STEP_MAP) {
      if (step === key) { return key; }
      let val = STEP_MAP[key];
      if (typeof val === 'string') {
        if (step === val) { return key; }
        continue;
      }
      for (let i = 0, l = val.length; i < l; i++) {
        if (val[i] === step) { return key; }
      }
    }
  }

  public static stepToNumber (step: string): number {
    let num = STEP_MAP[step];
    if (typeof num === 'string') { return Number(num); }
    return Number(num[0]);
  }

  public static getMeal (): string {
    let time = new Date().getHours() + 1;
    for (let hour in MEAL_TIMES) {
      if (Number(hour) >= time) { return MEAL_TIMES[hour]; }
    }
  }

  public static getLikeTerms (term: string): string[] {
    return this.getCategoryTerms(this.getCategoryFor(term));
  }

  public static getCategoryFor (term: string): string {
    for (let category in CATEGORIES) {
      if (~CATEGORIES[category].indexOf(term)) { return category; }
    }
    return '';
  }

  public static getCategoryTerms (category: string): string[] {
    return CATEGORIES[category] || [];
  }

}
