// jibo-esexpressions/lib/jibo-esexpressions.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboEsexpressions = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ArrayExpression {
    constructor() {
        this.expression = {
            type: "ArrayExpression",
            elements: []
        };
    }
    push(value) {
        this.expression.elements.push(value);
    }
}
exports.default = ArrayExpression;

},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const LiteralExpression_1 = require("./LiteralExpression");
exports.default = (relativePath) => {
    return {
        type: "ArrowFunctionExpression",
        id: null,
        params: [],
        defaults: [],
        body: {
            type: "BlockStatement",
            body: [
                {
                    type: "ReturnStatement",
                    argument: {
                        type: "CallExpression",
                        callee: {
                            type: "Identifier",
                            name: "require"
                        },
                        arguments: [LiteralExpression_1.default(relativePath)]
                    }
                }
            ]
        },
        generator: false,
        expression: false
    };
};

},{"./LiteralExpression":6}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FunctionExpression_1 = require("./FunctionExpression");
function CallExpression(objectExpression, hasSelf) {
    const args = hasSelf
        ? [
            {
                type: 'ObjectExpression',
                properties: []
            }
        ]
        : [];
    return {
        type: 'CallExpression',
        callee: FunctionExpression_1.default(objectExpression, hasSelf),
        arguments: args
    };
}
exports.default = CallExpression;

},{"./FunctionExpression":5}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const VariableDeclaration_1 = require("./VariableDeclaration");
exports.default = (name, className, args) => {
    return VariableDeclaration_1.default("const", name, {
        type: "NewExpression",
        callee: {
            type: "Identifier",
            name: className
        },
        arguments: args
    });
};

},{"./VariableDeclaration":10}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function FunctionExpression(objectExpression, hasSelf) {
    const params = hasSelf
        ? [
            {
                type: 'Identifier',
                name: 'self'
            }
        ]
        : [];
    return {
        type: 'FunctionExpression',
        id: null,
        params,
        defaults: [],
        body: {
            type: 'BlockStatement',
            body: [
                {
                    type: 'ReturnStatement',
                    argument: objectExpression
                }
            ]
        },
        rest: null,
        generator: false,
        expression: false
    };
}
exports.default = FunctionExpression;

},{}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function LiteralExpression(value) {
    return {
        type: "Literal",
        value: value,
        raw: "\"" + value + "\""
    };
}
exports.default = LiteralExpression;

},{}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const LiteralExpression_1 = require("./LiteralExpression");
class ObjectExpression {
    constructor(props) {
        this.expression = {
            type: "ObjectExpression",
            properties: props ? props : []
        };
    }
    addKeyValue(key, value) {
        this.expression.properties.push({
            type: "Property",
            key: LiteralExpression_1.default(key),
            value: value,
            kind: "init",
            method: false,
            shorthand: false
        });
    }
}
exports.default = ObjectExpression;

},{"./LiteralExpression":6}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const LiteralExpression_1 = require("./LiteralExpression");
const VariableDeclaration_1 = require("./VariableDeclaration");
exports.default = (name, moduleNameOrPath) => {
    return VariableDeclaration_1.default("const", name, {
        type: "CallExpression",
        callee: {
            type: "Identifier",
            name: "require"
        },
        arguments: [
            LiteralExpression_1.default(moduleNameOrPath)
        ]
    });
};

},{"./LiteralExpression":6,"./VariableDeclaration":10}],9:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = (objectIdentifier, propertyIdentifier) => {
    return {
        type: "MemberExpression",
        computed: false,
        object: {
            type: "Identifier",
            name: objectIdentifier
        },
        property: {
            type: "Identifier",
            name: propertyIdentifier
        }
    };
};

},{}],10:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = (kind, name, initializer) => {
    const declaration = {
        type: "VariableDeclaration",
        declarations: [
            {
                type: "VariableDeclarator",
                id: {
                    type: "Identifier",
                    name
                }
            }
        ],
        kind
    };
    if (initializer) {
        declaration.declarations[0].init = initializer;
    }
    return declaration;
};

},{}],11:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ArrayExpression_1 = require("./ArrayExpression");
exports.ArrayExpression = ArrayExpression_1.default;
const ArrowFunctionReturnRequire_1 = require("./ArrowFunctionReturnRequire");
exports.ArrowFunctionReturnRequire = ArrowFunctionReturnRequire_1.default;
const CallExpression_1 = require("./CallExpression");
exports.CallExpression = CallExpression_1.default;
const ConstFromNew_1 = require("./ConstFromNew");
exports.ConstFromNew = ConstFromNew_1.default;
const FunctionExpression_1 = require("./FunctionExpression");
exports.FunctionExpression = FunctionExpression_1.default;
const LiteralExpression_1 = require("./LiteralExpression");
exports.LiteralExpression = LiteralExpression_1.default;
const ObjectExpression_1 = require("./ObjectExpression");
exports.ObjectExpression = ObjectExpression_1.default;
const RequireModule_1 = require("./RequireModule");
exports.RequireModule = RequireModule_1.default;
const SimpleMemberExpression_1 = require("./SimpleMemberExpression");
exports.SimpleMemberExpression = SimpleMemberExpression_1.default;
const VariableDeclaration_1 = require("./VariableDeclaration");
exports.VariableDeclaration = VariableDeclaration_1.default;

},{"./ArrayExpression":1,"./ArrowFunctionReturnRequire":2,"./CallExpression":3,"./ConstFromNew":4,"./FunctionExpression":5,"./LiteralExpression":6,"./ObjectExpression":7,"./RequireModule":8,"./SimpleMemberExpression":9,"./VariableDeclaration":10}]},{},[11])(11)
});
